#########################
#
# Compound processing potential
#
# R code for the article:
# Bioenergetic mapping of ‘healthy microbiomes’ via compound processing potential imprinted in gut and soil metagenomes
#
# by Craig Liddicoat, Robert A. Edwards, Michael Roach, Jake M. Robinson, Andrew D. Barnes, Joel Brame, Anna Heintz-Buschart, 
# Timothy R. Cavagnaro, Elizabeth A. Dinsdale, Michael P. Doane, Nico Eisenhauer, Grace Mitchell, Bibishan Rai, Sunita Ramesh, 
# Kiri J. Wallace, Martin F. Breed 
#
#########################
# 
# Outline of code
# 
# For high-level description refer to: https://github.com/liddic/compound_potential
# 
# Case study datasets are analysed in order:
#   
# A) Soil metagenomes:
# - People, Cities and Nature (PCaN) urban forest ecosystem restoration (Barnes et al 2023: Aotearoa Genomic Data Repository, https://data.agdr.org.nz/; samples described by Mitchell 2022)
# - Australian Microbiome Initiative (AMI) disturbed vs. natural soils (https://data.bioplatforms.com/organization/australian-microbiome; Accessed in September 2022)
# - Post-mining ecosystem restoration (Sun and Badgley 2019)
# 
# B) Human gut metagenomes in health & disease:
# - Atherosclerotic cardiovascular disease (ACVD) (Jie et al 2017)
# - Colorectal cancer (Zeller et al 2014)
# - Type 2 diabetes (T2D) (Forslund et al 2015)
# - Childhood anxious-depressive behaviors by (Flannery et al 2020)
#
#########################

# record library and version info
.libPaths() # "/Library/Frameworks/R.framework/Versions/4.2/Resources/library"

R.Version()
# "R version 4.2.2 (2022-10-31)"
citation()
# R Core Team (2022). R: A language and environment for statistical computing. R Foundation for Statistical
# Computing, Vienna, Austria. URL https://www.R-project.org/.


library(readxl); packageVersion("readxl") # '1.4.1'
library(plyr); packageVersion("plyr") # '1.8.8'
library(dplyr); packageVersion("dplyr") # '1.0.10'
library(vegan);packageVersion("vegan") # '2.6.4'
library(phyloseq); packageVersion("phyloseq") # '1.42.0'
library(ggplot2); packageVersion("ggplot2") # '3.4.0'
library(grid); packageVersion("grid") #  '4.2.2'
library(reshape2); packageVersion("reshape2") # '1.4.4'
library(tidyr); packageVersion("tidyr") # '1.2.1'
library(corrr); packageVersion("corrr") # '0.4.4'
library(ggforce); packageVersion("ggforce") # '0.4.1'
library(ggrepel); packageVersion("ggrepel") # '0.9.2'
library(stringdist); packageVersion("stringdist") # ‘0.9.10’
library(stringr); packageVersion("stringr") # ‘1.5.0’
library(doParallel); packageVersion("doParallel") # '1.0.17'
library(zCompositions); packageVersion("zCompositions") # '1.4.0.1'
library(hexbin); packageVersion("hexbin") # '1.28.2'
library(RColorBrewer); packageVersion("RColorBrewer") # '1.1.3'
library(ggpp); packageVersion("ggpp") # ‘0.5.0’
library(corrplot); packageVersion("corrplot") #  '0.92'
library(caret); packageVersion("caret") # '6.0.93'
library(MASS); packageVersion("MASS") # ‘7.3.58.1’
library(ggsignif); packageVersion("ggsignif") # '0.6.4'
library(moments); packageVersion("moments") # ‘0.14.1’
library(ANCOMBC); packageVersion("ANCOMBC") # ‘2.0.1’
library(grDevices); packageVersion("grDevices") #  '4.2.2'
library(ggbiplot); packageVersion("ggbiplot") #  ‘0.55’
library(viridis); packageVersion("viridis") #  ‘0.6.2’
library(FSA); packageVersion("FSA") # '0.9.3'
library(rcompanion); packageVersion("rcompanion") # '2.4.18'
library(fields); packageVersion("fields") # ‘14.1’
library(car); packageVersion("car") # ‘3.1.1’
library(multcompView); packageVersion("multcompView") # ‘0.1.8’
library(gtools); packageVersion("gtools") # ‘3.9.4’
library(igraph); packageVersion("igraph") #  '1.4.2'
library(pheatmap); packageVersion("pheatmap") # '1.0.12'
library(colorspace); packageVersion("colorspace") # ‘2.1.0’


#########################
##save.image("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/WORKSPACE.RData")
##      load("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/WORKSPACE.RData")
#########################

workdir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
setwd(workdir)
getwd()

par.default <- par()


#### van Krevelen coords reference data
#-------------------------

# vK group rectangles for vanKrevelen plot
vkgrouprect <- read_excel(path = "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/vK-group-rectangles.xlsx", sheet = 1, range = "A1:D29")
vkgrouprect <- as.data.frame(vkgrouprect)
str(vkgrouprect)
vkgrouprect$label <- factor(vkgrouprect$label)
levels(vkgrouprect$label)
# [1] "Amino sugar"         "Carbohydrate"        "Condensed aromatics" "Lignin"              "Lipid"               "Protein"            
# [7] "Tannin"


# vK coords for health-associated molecules
df.vkcoords.refs <- read_excel(path = "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/oc-hc-ratios-copy.xlsx",
                               sheet = 1, range = "A1:N50")

names( df.vkcoords.refs )
# [1] "letter"                 "compound"               "KEGG_compound"          "alt_name"               "chem_formula"           "atomic_c"               "atomic_h"              
# [8] "atomic_o"               "oc_ratio_x"             "hc_ratio_y"             "figure_legend_category" "Studied_here"           "compound_type"          "role_or_comment"     

df.vkcoords.refs$compound
# [1] "Acetate"                         "Propionate"                      "Butyrate"                        "Riboflavin (Vit B2)"            
# [5] "Cobalamin (Vit B12)"             "Pyridoxal 5'-phosphate (Vit B6)" "Folate (Vit B9)"                 "Menaquinone (Vit K2)"           
# [9] "Trimethylamine"                  "p-cresol"                        "Indole"                          "Pyruvate"                       
# [13] "Acetyl-CoA"                      "Sucrose; Maltose; Lactose"       "Glycogen"                        "Cellulose; Amylose"             
# [17] "Chitin"                          "Stearic acid"                    "Oleic acid"                      "Linoleic acid"                  
# [21] "Ethanol"                         "Nicotine"                        "a-pinene; Camphene"              "Benzaldehyde"                   
# [25] "Wheat straw"                     "Corncob"                         "Softwood"                        "Switchgrass"                    
# [29] "Lignin"                          "Cellulose"                       "Wood chips"                      "Wood"                           
# [33] "Hardwood"                        "Switchgrass"                     "Corncob"                         "Sewage sludge"                  
# [37] "Sewage sludge"                   "Hemicellulose"                   "Hardwood"                        "Softwood"                       
# [41] "Humic acid (soil)"               "Fulvic acid (soil)"              "Meat protein"                    "Serotonin"                      
# [45] "Dopamine"                        "GABA"                            "Noradrenaline"                   "Glutamate"                      
# [49] "Glucose" 

# health-associated compounds
df.vk.ref <- data.frame(compound=df.vkcoords.refs$compound,
                        OC_x=df.vkcoords.refs$oc_ratio_x,
                        HC_y=df.vkcoords.refs$hc_ratio_y,
                        label=df.vkcoords.refs$figure_legend_category,
                        studied_here=df.vkcoords.refs$Studied_here,
                        group=df.vkcoords.refs$compound_type)

sel.na <- which(is.na(df.vk.ref$studied_here))
df.vk.ref$studied_here[sel.na] <- "No"


df.hac <- filter( df.vk.ref, studied_here == "Yes")
df.hac
#                           compound       OC_x     HC_y                         label studied_here                           group
# 1                          Acetate 1.00000000 2.000000 Health-associated biomolecule          Yes          Short chain fatty acid
# 2                       Propionate 0.66666667 2.000000 Health-associated biomolecule          Yes          Short chain fatty acid
# 3                         Butyrate 0.50000000 2.000000 Health-associated biomolecule          Yes          Short chain fatty acid
# 4              Riboflavin (Vit B2) 0.35294118 1.176471 Health-associated biomolecule          Yes                         Vitamin
# 5              Cobalamin (Vit B12) 0.22580645 1.419355 Health-associated biomolecule          Yes                         Vitamin
# 6  Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000 Health-associated biomolecule          Yes                         Vitamin
# 7                  Folate (Vit B9) 0.31578947 1.000000 Health-associated biomolecule          Yes                         Vitamin
# 8             Menaquinone (Vit K2) 0.06451613 1.290323 Health-associated biomolecule          Yes                         Vitamin
# 9                         Pyruvate 1.00000000 1.333333 Health-associated biomolecule          Yes        Intermediate biomolecule
# 10                       Glutamate 0.80000000 1.800000 Health-associated biomolecule          Yes Brain excitory neurotransmitter

unique( df.vk.ref$label )
df.vk.ref$label <- factor(df.vk.ref$label, levels = c("Health-associated biomolecule", "Dietary or environmental substrate"), ordered=TRUE)

col.points <- c("Health-associated biomolecule" = "#35978f" , "Dietary or environmental substrate" = "#fc8d62") # "#66c2a5"

p <- ggplot()+
  
  #xlim(0,1.8)+ ylim(0,3.1)+
  #xlim(0,2.1)+ ylim(0,3.1)+
  xlim(0,1.5)+ ylim(0,3.05)+
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #https://ggplot2-book.org/scale-colour.html#:~:text=11.3.&text=scale_colour_brewer()%20is%20a%20discrete,http%3A%2F%2Fcolorbrewer2.org%2F.
  
  theme_bw()+
  
  geom_segment(aes(x = 0, y = 0, xend = 1.5,  yend = 3), inherit.aes = FALSE , color = "#ffd92f",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ #  
  annotate(geom="text", x= 1.45, y= 3, label = "Hydration/\nCondensation", hjust=1, vjust=1, size = 2.75 , col="#ffd92f", lineheight = 0.8) +
  
  geom_segment(aes(x = 0, y = 1, xend = 1.5,  yend = 1), inherit.aes = FALSE , color = "#ffd92f",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ # , linetype = "dotted"
  annotate(geom="text", x= 1.5, y= 1.05, label = "Oxidation/\nReduction", hjust=1, vjust=0, size = 2.75 , col="#ffd92f", lineheight = 0.8) +
  
  geom_segment(aes(x = 0, y = 2, xend = 1.0,  yend = 0), inherit.aes = FALSE ,  color = "#ffd92f",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ # linetype = "dotted",
  annotate(geom="text", x= 0.95, y= 0.1, label = "Methylation/\nDemethylation", hjust=1, vjust=1, size = 2.75 , col="#ffd92f", lineheight = 0.8) +
  
  geom_segment(aes(x = 0.5, y = 0, xend = 0.5,  yend = 2.95), inherit.aes = FALSE , color = "#ffd92f",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ # , linetype = "dotted"
  annotate(geom="text", x= 0.51, y= 2.87, label = "Hydrogenation/\nDehydrogenation", hjust=0, vjust=1, size = 2.75 , col="#ffd92f", lineheight = 0.8) +
  
  geom_segment(aes(x = 0, y = 1.1, xend = 1.3,  yend = 0), inherit.aes = FALSE , linetype = "twodash", color = "#ffd92f" )+
  annotate(geom="text", x= 0.6, y= 0.57, label = "Increasing\naromaticity", angle = -17.8 , hjust=0.5, vjust=1, size = 2.75 , col="#ffd92f", lineheight = 0.8) +
  geom_segment(aes(x = 0.58, y = 0.46, xend = 0.565,  yend = 0.36), inherit.aes = FALSE , color = "#ffd92f", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "last") )+ # , linetype = "dotted"
  
  geom_segment(aes(x = 0.45, y = 1.95, xend = 0.15,  yend = 2.5), inherit.aes = FALSE , color = "#ffd92f",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "last"))+ #  
  annotate(geom="text", x= 0.15, y= 2.51, label = "Increasing\nenergy\ncontent", hjust=0.5, vjust=0, size = 2.75 , col="#ffd92f", lineheight = 0.8) +
  
  
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  #geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 1.5,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  geom_point(data = filter(df.vk.ref, studied_here=="No") ,aes(x=OC_x, y=HC_y, color = label))+ # , pch=8  "#80b1d3"
  geom_text(data = filter(df.vk.ref, studied_here=="No") ,aes(x=OC_x, y=HC_y, label=compound, color = label), hjust=0, vjust=1,size = 2.5, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  
  geom_point(data = filter(df.vk.ref, studied_here=="Yes") ,aes(x=OC_x, y=HC_y),size=3, color = "#35978f")+  # "#66c2a5"
  geom_point(data = filter(df.vk.ref, studied_here=="Yes") ,aes(x=OC_x, y=HC_y),size=3, shape = 21, stroke = 1, color = "#253494")+ # , pch=8  "#80b1d3"
  geom_text_repel(data = filter(df.vk.ref, studied_here=="Yes") ,aes(x=OC_x, y=HC_y, label=compound),color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01, nudge_y=0.05)+ # , pch=8.  color = "#e6550d",
  
  scale_color_manual(values = col.points, name = NULL)+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.8, 0.2), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.key.size = unit(0.9,"line"),
    #legend.title = element_text(size = rel(0.9))
    legend.title = element_blank()
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","vK-Ref-Figure-with-context-trends-v9.tiff"), width = 20, height = 18, units = "cm", res=450, compression="lzw",type="cairo")


## Alternative colour scheme

col.points <- c("Health-associated biomolecule" = "#35978f" , "Dietary or environmental substrate" = "#fb8072") # "#66c2a5"


p <- ggplot()+
  
  #xlim(0,1.8)+ ylim(0,3.1)+
  #xlim(0,2.1)+ ylim(0,3.1)+
  xlim(0,1.5)+ ylim(0,3.05)+
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #https://ggplot2-book.org/scale-colour.html#:~:text=11.3.&text=scale_colour_brewer()%20is%20a%20discrete,http%3A%2F%2Fcolorbrewer2.org%2F.
  
  theme_bw()+
  
  geom_segment(aes(x = 0, y = 0, xend = 1.5,  yend = 3), inherit.aes = FALSE , color = "#fdb462",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ #  
  annotate(geom="text", x= 1.45, y= 3, label = "Hydration/\nCondensation", hjust=1, vjust=1, size = 2.75 , col="#fdb462", lineheight = 0.8) +
  
  geom_segment(aes(x = 0, y = 1, xend = 1.5,  yend = 1), inherit.aes = FALSE , color = "#fdb462",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ # , linetype = "dotted"
  annotate(geom="text", x= 1.5, y= 1.05, label = "Oxidation/\nReduction", hjust=1, vjust=0, size = 2.75 , col="#fdb462", lineheight = 0.8) +
  
  geom_segment(aes(x = 0, y = 2, xend = 1.0,  yend = 0), inherit.aes = FALSE ,  color = "#fdb462",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ # linetype = "dotted",
  annotate(geom="text", x= 0.95, y= 0.1, label = "Methylation/\nDemethylation", hjust=1, vjust=1, size = 2.75 , col="#fdb462", lineheight = 0.8) +
  
  geom_segment(aes(x = 0.5, y = 0, xend = 0.5,  yend = 2.95), inherit.aes = FALSE , color = "#fdb462",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "both"))+ # , linetype = "dotted"
  annotate(geom="text", x= 0.51, y= 2.87, label = "Hydrogenation/\nDehydrogenation", hjust=0, vjust=1, size = 2.75 , col="#fdb462", lineheight = 0.8) +
  
  geom_segment(aes(x = 0, y = 1.1, xend = 1.3,  yend = 0), inherit.aes = FALSE , linetype = "twodash", color = "#fdb462" )+
  annotate(geom="text", x= 0.6, y= 0.57, label = "Increasing\naromaticity", angle = -17.8 , hjust=0.5, vjust=1, size = 2.75 , col="#fdb462", lineheight = 0.8) +
  geom_segment(aes(x = 0.58, y = 0.46, xend = 0.565,  yend = 0.36), inherit.aes = FALSE , color = "#fdb462", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = c("last")) )+ # , linetype = "dotted"
  
  geom_segment(aes(x = 0.45, y = 1.95, xend = 0.15,  yend = 2.5), inherit.aes = FALSE , color = "#fdb462",linetype = "dotdash", arrow = arrow(length = unit(0.3, "cm"), type = "closed", ends = "last"))+ #  
  annotate(geom="text", x= 0.15, y= 2.51, label = "Increasing\nenergy\ncontent", hjust=0.5, vjust=0, size = 2.75 , col="#fdb462", lineheight = 0.8) +
  
  
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  #geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 1.5,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  geom_point(data = filter(df.vk.ref, studied_here=="No") ,aes(x=OC_x, y=HC_y, color = label))+ # , pch=8  "#80b1d3"
  geom_text(data = filter(df.vk.ref, studied_here=="No") ,aes(x=OC_x, y=HC_y, label=compound, color = label), hjust=0, vjust=1,size = 2.5, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  
  geom_point(data = filter(df.vk.ref, studied_here=="Yes") ,aes(x=OC_x, y=HC_y),size=3, color = "#35978f")+  # "#66c2a5"
  geom_point(data = filter(df.vk.ref, studied_here=="Yes") ,aes(x=OC_x, y=HC_y),size=3, shape = 21, stroke = 1, color = "#253494")+ # , pch=8  "#80b1d3"
  geom_text_repel(data = filter(df.vk.ref, studied_here=="Yes") ,aes(x=OC_x, y=HC_y, label=compound),color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01, nudge_y=0.05)+ # , pch=8.  color = "#e6550d",
  
  scale_color_manual(values = col.points, name = NULL)+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.8, 0.2), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.key.size = unit(0.9,"line"),
    #legend.title = element_text(size = rel(0.9))
    legend.title = element_blank()
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","vK-Ref-Figure-with-context-trends-v9c.tiff"), width = 20, height = 18, units = "cm", res=450, compression="lzw",type="cairo")



#-------------------------

#### ModelSEED lookup tables
#-------------------------

# ModelSEED database tables from https://github.com/ModelSEED/ModelSEEDDatabase ; accessed 10 Aug 2022

seed_db_dir <- "~/WORKSPACE/DATA/ModelSEEDDB/select_files"


## Unique_ModelSEED_Reaction_ECs

rxn_ECs.lut <- read_excel(path = paste0(seed_db_dir,"/","Unique_ModelSEED_Reaction_ECs.xlsx") , range = "A1:C30834")
rxn_ECs.lut <- as.data.frame(rxn_ECs.lut)
head(rxn_ECs.lut)
names(rxn_ECs.lut) # "ModelSEED ID" "External ID"  "Source"
names(rxn_ECs.lut) <- c("rxn_id", "EC_id",  "Source")

dim(rxn_ECs.lut) # 30833     3
length(unique(rxn_ECs.lut$rxn_id)) # 25771
length(unique(rxn_ECs.lut$EC_id)) # 7354


## Unique_ModelSEED_Reaction_Pathways

# rxn_pathways.lut <- read_excel(path = paste0(seed_db_dir,"/","Unique_ModelSEED_Reaction_Pathways.xlsx") , range = "A1:C121445")
# rxn_pathways.lut <- as.data.frame(rxn_pathways.lut)
rxn_pathways.lut <- read.csv(file = paste0(seed_db_dir,"/","Unique_ModelSEED_Reaction_Pathways.txt"), header = TRUE, sep = "\t")
class(rxn_pathways.lut) # "data.frame"

head(rxn_pathways.lut)
names(rxn_pathways.lut) # "ModelSEED.ID" "External.ID"  "Source"
names(rxn_pathways.lut) <- c("rxn_id", "External_rxn_name",  "Source")

dim(rxn_pathways.lut) # 121444      3
length(unique(rxn_pathways.lut$rxn_id)) # 21734
length(unique(rxn_pathways.lut$External_rxn_name)) # 3819


## Model SEED Subsystems

subsys.lut <- read_excel(path = paste0(seed_db_dir,"/","ModelSEED_Subsystems.xlsx") , range = "A1:E9821")
subsys.lut <- as.data.frame(subsys.lut)
dim(subsys.lut) # 9820    5

head(subsys.lut)
tail(subsys.lut)
names(subsys.lut) # "Class"     "Sub-class" "Name"      "Role"      "Reaction" 
names(subsys.lut) <- c("Class",    "Subclass", "Name",  "Role",     "Reaction")

subsys.lut[sample(1:dim(subsys.lut)[1], 10,replace = FALSE), ]
#                                     Class                                    Subclass                                                Name                                                                           Role Reaction
# 2984          Clustering-based subsystems                                           -                              CBSS-196620.1.peg.2477                                      Maltose O-acetyltransferase (EC 2.3.1.79) rxn01133
# 6834 Fatty Acids, Lipids, and Isoprenoids                                 Fatty acids                       Fatty_Acid_Biosynthesis_FASII          (3R)-hydroxymyristoyl-[acyl carrier protein] dehydratase (EC 4.2.1.-) rxn05427
# 7057 Fatty Acids, Lipids, and Isoprenoids                                 Fatty acids                   Unsaturated_Fatty_Acid_Metabolism                  3-oxoacyl-[acyl-carrier-protein] synthase, KASI (EC 2.3.1.41) rxn05350
# 9384                      Stress Response                              Osmotic stress Choline_and_Betaine_Uptake_and_Betaine_Biosynthesis                   Glycine betaine ABC transport system, permease protein OpuAB rxn05181
# 5904              Experimental Subsystems                                           -                              Transporters_In_Models                 Oligopeptide transport ATP-binding protein oppF (TC 3.A.1.5.1) rxn05539
# 5267              Experimental Subsystems                                           -              Sugar_catabolome_in_Shewanella_species                                              Alpha-galactosidase (EC 3.2.1.22) rxn00818
# 2626                Cell Wall and Capsule   Capsular and extracellular polysacchrides                                 Alginate_metabolism                                      Acetoin (diacetyl) reductase (EC 1.1.1.5) rxn01685
# 8538                   Protein Metabolism                        Protein biosynthesis                       Translation_factors_bacterial                                       Methionine aminopeptidase (EC 3.4.11.18) rxn12635
# 1853                        Carbohydrates                                Fermentation                 Acetyl-CoA_fermentation_to_Butyrate                     Acyl-CoA dehydrogenase, short-chain specific (EC 1.3.99.2) rxn10012
# 776           Amino Acids and Derivatives Lysine, threonine, methionine, and cysteine                     Lysine_Biosynthesis_DAP_Pathway 2,3,4,5-tetrahydropyridine-2,6-dicarboxylate N-acetyltransferase (EC 2.3.1.89) rxn03030
## Correspondences:

# Superfocus (imported 'tab' object): Subsystem Level 1 = subsys.lut$Class
#                                     Subsystem Level 2 = subsys.lut$Subclass
#    (skip the subsys.lut$Name field)
#                                     Function = subsys.lut$Role  # replace " " with "_" to match Superfocus output

dim(subsys.lut) # 9820    5
length(unique(subsys.lut$Class )) # 29
length(unique(subsys.lut$Subclass )) # 143
length(unique(subsys.lut$Name )) # 666
length(unique(subsys.lut$Role )) # 1885
length(unique(subsys.lut$Reaction )) # 1986


## Reactions

rxns.lut <- read_excel(path = paste0(seed_db_dir,"/","reactions.xlsx") , range = "A1:V43775")
rxns.lut <- as.data.frame(rxns.lut)
dim(rxns.lut) # 43774    22
names(rxns.lut)
# [1] "id"                "abbreviation"      "name"              "code"              "stoichiometry"     "is_transport"      "equation"          "definition"       
# [9] "reversibility"     "direction"         "abstract_reaction" "pathways"          "aliases"           "ec_numbers"        "deltag"            "deltagerr"        
# [17] "compound_ids"      "status"            "is_obsolete"       "linked_reaction"   "notes"             "source" 

head(rxns.lut)

length(unique(rxns.lut$id)) # 43774
length(unique(rxns.lut$name)) # 27668
length(unique(rxns.lut$code)) # 36212
length(unique(rxns.lut$equation)) # 37310
length(unique(rxns.lut$aliases)) # 32718    # SEARCH aliases ???????
length(unique(rxns.lut$ec_numbers)) # 7607


## Compounds

compounds.lut <- read_excel(path = paste0(seed_db_dir,"/","compounds.xlsx") , range = "A1:T33993")
warnings()
compounds.lut <- as.data.frame(compounds.lut)
dim(compounds.lut) # 33992    20
names(compounds.lut)
# [1] "id"                "abbreviation"      "name"              "formula"           "mass"              "source"            "inchikey"          "charge"           
# [9] "is_core"           "is_obsolete"       "linked_compound"   "is_cofactor"       "deltag"            "deltagerr"         "pka"               "pkb"              
# [17] "abstract_compound" "comprised_of"      "aliases"           "smiles"

head(compounds.lut)

length(unique(compounds.lut$id)) # 33992
length(unique(compounds.lut$formula)) # 16763


#-------------------------

#### Atomic ratio lookup function? & map all ModelSEED compounds
#-------------------------

# from earlier
## Compounds

dim(compounds.lut) # 33992    20
names(compounds.lut)
# [1] "id"                "abbreviation"      "name"              "formula"           "mass"              "source"            "inchikey"          "charge"           
# [9] "is_core"           "is_obsolete"       "linked_compound"   "is_cofactor"       "deltag"            "deltagerr"         "pka"               "pkb"              
# [17] "abstract_compound" "comprised_of"      "aliases"           "smiles"

head(compounds.lut)
head(compounds.lut[ ,1:7])
#         id abbreviation  name       formula mass           source                    inchikey
# 1 cpd00001          h2o   H2O           H2O   18 Primary Database XLYOFNOQVPJJNP-UHFFFAOYSA-N
# 2 cpd00002          atp   ATP C10H13N5O13P3  504 Primary Database ZKHQWZAMYRWXGA-KQYNXXCUSA-K
# 3 cpd00003          nad   NAD C21H26N7O14P2  662 Primary Database BAWFJGJZGIEFAR-NNYOXOHSSA-M
# 4 cpd00004         nadh  NADH C21H27N7O14P2  663 Primary Database BOPGDPNILDQYTO-NNYOXOHSSA-L
# 5 cpd00005        nadph NADPH C21H26N7O17P3  742 Primary Database ACFIXJIJDZMPPO-NNYOXOHSSA-J
# 6 cpd00006         nadp  NADP C21H25N7O17P3  741 Primary Database XJLXINKUBYWONI-NNYOXOHSSA-K

length(unique(compounds.lut$id)) # 33992
length(unique(compounds.lut$formula)) # 16763
length(unique(compounds.lut$name)) # 33844


## get all combinations of 'O?' 'C?' 'H?' - to isolate oxygen, carbon, hydrogen containing compounds

temp <- compounds.lut

#.          1.      2.      3.       4.       5.       6.        7.        8.      9.     10.          11             12.        13
form <- c("NaCl", "H2O", "CaOH", "CaOCl2Rh","HgCl","C5H8O7PR", "MoOH2", "CoCH3", "OsR", "ScO2", "C10H12N4O12P2", "C15H27N5O5", "HgR")

atom <- "O" # avoid: Os, Og
atom <- "C" # avoid: Cs, Ca, Ce, Cr, Co, Cm, Cu, Cd, Cn, Cf, Cl, 
atom <- "H" # avoid: Hf, Hs, Hg, Ho, He



atomic_no <- function(form, atom) {
  ##form=form
  ##atom="O"
  coefs <- list()
  for (c in 1:length(form)) {
    #c<-1
    pos <- str_locate(string = form[c], pattern = atom)[1]
    if (is.na(pos)) { # not present
      ##print("zero")
      coef<-0
    } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A 
      ##print("A")
      coef<-1
    } else {
      # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
      ##print("B")
      checkstring <- substring(text = form[c], first = pos, last = pos+1)
      
      if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
        ##print("C")
        coef<-0
      } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
        ##print("D")
        coef<-0
      } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E 
        ##print("E")
        coef<-0
      } else { # F
        ##print("F")
        coef<-1
        coef.try <- coef
        pos.end <- pos+1
        
        while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
          # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
          coef <- coef.try
          coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
          pos.end <- pos.end+1
          ##print("G")
        } # END while loop
      } # END else
      
    } # END else
    
    coefs[[c]] <- coef
    
  } # END c loop
  return( unlist(coefs) )
}

form
# [1] "NaCl"          "H2O"           "CaOH"          "CaOCl2Rh"      "HgCl"          "C5H8O7PR"      "MoOH2"         "CoCH3"         "OsR"           "ScO2"         
# [11] "C10H12N4O12P2" "C15H27N5O5"    "HgR" 

atomic_no(form, "O")
# 0  1  1  1  0  7  1  0  0  2 12  5  0

atomic_no(form="NaCl",atom="C")
# [1] "B"
# [1] "D"
# [1] 0

atomic_no(form=form,atom="C")
#  0  0  0  0  0  5  0  0  0  0 10 15  0
# has issues if Co comes before C, however format of compounds database has C first

atomic_no(form=form,atom="H")
#   0  2  1  0  0  8  2  3  0  0 12 27  0



df.comp <- data.frame(id=compounds.lut$id, 
                      abbrev=compounds.lut$abbreviation,
                      name=compounds.lut$name,
                      form=compounds.lut$formula,
                      OC_ratio=NA,
                      HC_ratio=NA
)

# uses function:

atomic_no
# function(form, atom) {
#   ##form=form
#   ##atom="O"
#   coefs <- list()
#   for (c in 1:length(form)) {
#     #c<-1
#     pos <- str_locate(string = form[c], pattern = atom)[1]
#     if (is.na(pos)) { # not present
#       ##print("zero")
#       coef<-0
#     } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A 
#       ##print("A")
#       coef<-1
#     } else {
#       # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
#       ##print("B")
#       checkstring <- substring(text = form[c], first = pos, last = pos+1)
#       
#       if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
#         ##print("C")
#         coef<-0
#       } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
#         ##print("D")
#         coef<-0
#       } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E 
#         ##print("E")
#         coef<-0
#       } else { # F
#         ##print("F")
#         coef<-1
#         coef.try <- coef
#         pos.end <- pos+1
#         
#         while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
#           # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
#           coef <- coef.try
#           coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
#           pos.end <- pos.end+1
#           ##print("G")
#         } # END while loop
#       } # END else
#       
#     } # END else
#     
#     coefs[[c]] <- coef
#     
#   } # END c loop
#   return( unlist(coefs) )
# }
# <bytecode: 0x7fe50641d6f8>



for (i in 1:length(df.comp$id)) {
  #i<-1
  formx.char <- df.comp$form[i]
  
  # O:C ratio (replace Inf with NA when C not present)
  df.comp$OC_ratio[i] <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
  df.comp$OC_ratio[i][is.infinite(df.comp$OC_ratio[i])] <- NA
  
  # H:C ratio (replace Inf with NA when C not present)
  df.comp$HC_ratio[i] <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
  df.comp$HC_ratio[i][is.infinite(df.comp$HC_ratio[i])] <- NA
  
  print(paste0("completed ",i))
  
}


sel <- which(df.comp$form == "H2O")
df.comp[sel, ]
#             id abbrev          name form OC_ratio HC_ratio
# 1     cpd00001    h2o           H2O  H2O       NA       NA
# 13619 cpd15275    oh1 hydroxide ion  H2O       NA       NA


# remove NA, NaN rows

ok <- complete.cases(df.comp)
sel.ok <- which(ok==TRUE) # qty 30035

df.comp <- df.comp[sel.ok, ]

plot(df.comp$OC_ratio, df.comp$HC_ratio)


sel <- which(df.comp$OC_ratio >=4 | df.comp$HC_ratio >=4) # qty 98
sel <- which(df.comp$OC_ratio >=5 | df.comp$HC_ratio >=5) # qty 38

df.comp[sel, ]
#             id                                                              abbrev                                                                name       form OC_ratio HC_ratio
# 145   cpd00146                                                                 cbp                                                  Carbamoylphosphate    CH2NO5P      5.0 2.000000
# 186   cpd00187                                                                 mea                                                         Methanamine       CH6N      0.0 6.000000
# 1383  cpd01408                                                           Aminourea                                                           Aminourea     CH5N3O      1.0 5.000000
# 1565  cpd01598                                                    Formyl phosphate                                                    Formyl phosphate      CHO5P      5.0 1.000000
# 3796  cpd03874                                                           Foscarnet                                                           Foscarnet      CHO5P      5.0 1.000000
# 4424  cpd04506                                                      Methylarsonite                                                      Methylarsonite    CH5AsO2      2.0 5.000000
# 7816  cpd07918                                                                AMPA                                                                AMPA    CH5NO3P      3.0 5.000000
# 7978  cpd08081                                                  Thiocarbohydrazide                                                  Thiocarbohydrazide     CH6N4S      0.0 6.000000
# 11485 cpd11660                                                             RCH2NH2                                                             RCH2NH2      CH5NR      0.0 5.000000
# 11850 cpd12031                                                      Acyl phosphate                                                      Acyl phosphate      CO5PR      5.0 0.000000
# 11890 cpd12071                                                     Primary diamine                                                     Primary diamine      CH7N2      0.0 7.000000
# 11924 cpd12107                                                    Methylated amine                                                    Methylated amine      CH5NR      0.0 5.000000
# 12181 cpd12369                                          Alkane-alpha,omega-diamine                                          Alkane-alpha,omega-diamine      CH7N2      0.0 7.000000
# 14196 cpd15895                                                                 mma                                                     monomethylamine       CH6N      0.0 6.000000
# 16525 cpd19240                                                           Guanidine                                                           Guanidine      CH6N3      0.0 6.000000
# 16931 cpd19646                                        Ethylendiamine dihydroiodide                                        Ethylendiamine dihydroiodide  C2H10I2N2      0.0 5.000000
# 17843 cpd20558                                                   Tetranitromethane                                                   Tetranitromethane      CN4O8      8.0 0.000000
# 18061 cpd20776                                         Semicarbazide hydrochloride                                         Semicarbazide hydrochloride   CH5ClN3O      1.0 5.000000
# 19589 cpd22306                                                     Acyl-Phosphates                                                     Acyl-Phosphates      CO5PR      5.0 0.000000
# 19610 cpd22327                                                    Aliphatic-Amines                                                    Aliphatic-Amines      CH5NR      0.0 5.000000
# 19626 cpd22343                                                         Alkylamines                                                         Alkylamines      CH5NR      0.0 5.000000
# 19672 cpd22389                                                       Aralkylamines                                                       Aralkylamines      CH5NR      0.0 5.000000
# 19861 cpd22578                                                    hydroxyguanidine                                                    hydroxyguanidine     CH6N3O      1.0 6.000000
# 20002 cpd22719                                               aminomethanesulfonate                                               aminomethanesulfonate    CH5NO3S      3.0 5.000000
# 20342 cpd23061 4,5-bisdiphosphoinositol-1D-myo-inositol (1,2,3,6)tetrakisphosphate 4,5-bisdiphosphoinositol-1D-myo-inositol (1,2,3,6)tetrakisphosphate C6H19O30P8      5.0 3.166667
# 20431 cpd23150 1,5-bisdiphosphoinositol-1D-myo-inositol (2,3,4,6)tetrakisphosphate 1,5-bisdiphosphoinositol-1D-myo-inositol (2,3,4,6)tetrakisphosphate  C6H7O30P8      5.0 1.166667
# 20432 cpd23151 3,5-bisdiphosphoinositol-1D-myo-inositol (2,3,4,6)tetrakisphosphate 3,5-bisdiphosphoinositol-1D-myo-inositol (2,3,4,6)tetrakisphosphate  C6H7O30P8      5.0 1.166667
# 20577 cpd23297                                                      aminoguanidine                                                      aminoguanidine      CH7N4      0.0 7.000000
# 21712 cpd24433                                                   1,2-diaminoethane                                                   1,2-diaminoethane    C2H10N2      0.0 5.000000
# 21720 cpd24441                                                                 MDP                                                                 MDP    CH4O6P2      6.0 4.000000
# 21734 cpd24455                        (5,6)-bisdiphosphoinositol tetrakisphosphate                        (5,6)-bisdiphosphoinositol tetrakisphosphate  C6H7O30P8      5.0 1.166667
# 22101 cpd24822                                                 methylhydroxylamine                                                 methylhydroxylamine      CH5NO      1.0 5.000000
# 24869 cpd27599                                                 Methylhydroxylamine                                                 Methylhydroxylamine      CH5NO      1.0 5.000000
# 25979 cpd28711                            Bisdiphospho-myo-inositol-polyphosphates                            Bisdiphospho-myo-inositol-polyphosphates  C6H7O30P8      5.0 1.166667
# 26211 cpd28943                                                  O-methyl phosphate                                                  O-methyl phosphate     CH5O4P      4.0 5.000000
# 28213 cpd31520                                         glycolaldehyde triphosphate                                         glycolaldehyde triphosphate  C2H4O11P3      5.5 2.000000
# 29875 cpd33182                                                   N-methylhydrazine                                                   N-methylhydrazine      CH6N2      0.0 6.000000
# 32680 cpd35987                                                    carboxyphosphate                                                    carboxyphosphate      CHO6P      6.0 1.000000





# overlay vK ref data ...

# sel <- which(df.vk.ref$sample %in% c("Acetate", "Propionate",  "Butyrate", "Riboflavin (Vit B2)" ,           
#                                      "Cobalamin (Vit B12)", "Pyridoxal 5'-phosphate (Vit B6)", "Folate (Vit B9)", "Menaquinone (Vit K2)" ) )
# df.hac <- df.vk.ref[sel, ]
df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2

vkgrouprect


str(df.comp)
# 'data.frame':	30035 obs. of  6 variables:
# $ id      : chr  "cpd00002" "cpd00003" "cpd00004" "cpd00005" ...
# $ abbrev  : chr  "atp" "nad" "nadh" "nadph" ...
# $ name    : chr  "ATP" "NAD" "NADH" "NADPH" ...
# $ form    : chr  "C10H13N5O13P3" "C21H26N7O14P2" "C21H27N7O14P2" "C21H26N7O17P3" ...
# $ OC_ratio: num  1.3 0.667 0.667 0.81 0.81 ...
# $ HC_ratio: num  1.3 1.24 1.29 1.24 1.19 ...

library(scales)
library(viridis)

sel <- which(df.comp$OC_ratio >=0 & df.comp$OC_ratio <=2.1 & df.comp$HC_ratio >=0 & df.comp$HC_ratio <=3.1 )

p <- ggplot()+
  
  geom_bin2d(data = df.comp[sel, ], aes(x = OC_ratio, y = HC_ratio), bins = 100) + # , bins = 70
  
  scale_fill_continuous(type = "viridis", direction = -1) + # viridis inferno magma
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncompounds"))+
  
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  geom_point(data = df.hac ,aes(x=OC_x, y=HC_y), color = "#b10026")+ # , pch=8  "#80b1d3"
  geom_text(data = df.hac ,aes(x=OC_x, y=HC_y, label=compound), color = "#b10026", hjust=0, vjust=1,size = 3, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d", 
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = c(0.75, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p



grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","All-compounds-vK-space-scatterheatmap--v9d-viridis.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


p <- ggplot()+
  
  geom_bin2d(data = df.comp, aes(x = OC_ratio, y = HC_ratio), bins = 100) + # , bins = 70
  scale_fill_continuous(type = "viridis", direction = -1) + # viridis inferno magma
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncompounds"))+
  
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.75, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p


grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","All-compounds-vK-space-scatterheatmap-WIDEextent--v9d-viridis.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### cclasso.R function (Fang Huaying 2016)
#-------------------------
# Acknowledgement to:
# https://github.com/huayingfang/CCLasso/blob/master/R/cclasso.R
# File: cclasso.R
# Aim : Correlation inference for compositional data through lasso
# - - - - - - - - - - - - - - - - - - - - - 
# Author : Fang Huaying (Peking University)
# Email  : hyfang@pku.edu.cn
# Date   : 2016-01-08
# Version: 2.0
# - - - - - - - - - - - - - - - - - - - - - 
# Main function: cclasso(x, counts = FALSE, pseudo = 0.5, k_cv = 3, 
#                        lam_int = c(1e-4, 1), k_max = 20, n_boot = 20) 
#
#  Input:
#           x ------ n x p data matrix (row/column is sample/variable)
#                    n samples & p compositional variables
#      counts ------ Is the compositional data matrix a count matrix? 
#                    Default: FALSE
#      pseudo ------ pseudo count if counts = TRUE
#                    Default: 0.5
#        k_cv ------ folds of cross validation
#                    Default: 3     
#     lam_int ------ tuning parameter interval
#                    Default: [1e-4, 1]
#       k_max ------ maximum iterations for golden section method
#                    Default: 20
#      n_boot ------ Bootstrap times
#                    Default: 20
#  Output: 
#      A list structure contains:
#       var_w ------ variance estimation
#       cor_w ------ correlation estimation
#      p_vals ------ p-values for elements of cor_w equal 0 or not
#      lambda ------ final tuning parameter
#     info_cv ------ information for cross validation
# - - - - - - - - - - - - - - - - - - - - - 
cclasso <- function(x, counts = FALSE, pseudo = 0.5, k_cv = 3, 
                    lam_int = c(1e-4, 1), k_max = 20, n_boot = 20) {
  n <- nrow(x);
  p <- ncol(x);
  
  if(counts) {
    x <- x + pseudo;
    x <- x / rowSums(x);
  }
  x <- log(x);
  vx2 <- stats::var(x);
  
  # Diagonal weight for loss function
  rmean_vx2 <- rowMeans(vx2);
  wd <- 1/diag(vx2 - rmean_vx2 - rep(rmean_vx2, each = p) + mean(rmean_vx2));
  wd2 <- sqrt(wd);
  
  # Some global parameters for optimization with single lambda
  rho <- 1;
  u_f <- eigen(diag(p) - 1/p)$vectors;
  wd_u <- (t(u_f) %*% (wd * u_f))[-p, -p];
  wd_u_eig <- eigen(wd_u);
  d0_wd <- 1 / ( (rep(wd_u_eig$values, each = p-1) + wd_u_eig$values) / 
                   (2 * rho) + 1 );
  u0_wd <- wd_u_eig$vectors;
  
  # Golden section method for the selection of lambda (log10 scale)
  sigma <- vx2;
  lam_int2 <- log10(range(lam_int));
  a1 <- lam_int2[1]; 
  b1 <- lam_int2[2];
  # Store lambda and corresponding cross validation's loss
  lams <- NULL; 
  fvals <- NULL;
  # Two trial points in first 
  a2 <- a1 + 0.382 * (b1 - a1); 
  b2 <- a1 + 0.618 * (b1 - a1);
  fb2 <- cv_loss_cclasso(lambda2 = 10^b2 / rho, x = x, k_cv = k_cv, 
                         sigma = sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                         wd2 = wd2);
  lams <- c(lams, b2); 
  fvals <- c(fvals, fb2$cv_loss);
  fa2 <- cv_loss_cclasso(lambda2 = 10^a2 / rho, x = x, k_cv = k_cv, 
                         sigma = fb2$sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                         wd2 = wd2);
  lams <- c(lams, a2); 
  fvals <- c(fvals, fa2$cv_loss);
  # Error tolerance for convergence
  err_lam2 <- 1e-1 * max(1, lam_int2);
  err_fval <- 1e-4;
  
  err <- b1 - a1;
  k <- 0;
  while(err > err_lam2 && k < k_max) {
    fval_max <- max(fa2$cv_loss, fb2$cv_loss);
    
    if(fa2$cv_loss > fb2$cv_loss) {
      a1 <- a2;      
      a2 <- b2;
      fa2 <- fb2;
      b2 <- a1 + 0.618 * (b1 - a1);
      fb2 <- cv_loss_cclasso(lambda2 = 10^b2 / rho, x = x, k_cv = k_cv, 
                             sigma = fa2$sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                             wd2 = wd2);
      
      lams <- c(lams, b2); 
      fvals <- c(fvals, fb2$cv_loss);
    } else {
      b1 <- b2;
      b2 <- a2;
      fb2 <- fa2;
      a2 <- a1 + 0.382 * (b1 - a1);
      fa2 <- cv_loss_cclasso(lambda2 = 10^a2 / rho, x = x, k_cv = k_cv, 
                             sigma = fb2$sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                             wd2 = wd2);
      
      lams <- c(lams, a2); 
      fvals <- c(fvals, fa2$cv_loss);
    }
    fval_min <- min(fa2$cv_loss, fb2$cv_loss);      
    
    k <- k + 1;
    err <- b1 - a1;
    if(abs(fval_max - fval_min) / (1 + fval_min) <= err_fval) {
      break;
    }
  }
  info_cv <- list(lams = lams, fvals = fvals, k = k + 2, 
                  lam_int = 10^c(a1, b1)); 
  if(a1 == lam_int2[1] || b1 == lam_int2[2]) {
    cat("WARNING:\n", "\tOptimal lambda is near boundary! ([", 10^a1, ",", 
        10^b1, "])\n", sep = "");
  }
  
  lambda <- 10^((a2 + b2)/2);
  # Bootstrap for cclasso
  lambda2 <- lambda / rho;
  info_boot <- boot_cclasso(x = x, sigma = fb2$sigma, lambda2 = lambda2, 
                            n_boot = n_boot, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
  
  return(list(var_w = info_boot$var_w, cor_w = info_boot$cor_w, 
              p_vals = info_boot$p_vals, lambda = lambda, info_cv = info_cv));
}
# - - - - - - - - - - - - - - - - - - - - - 
# Bootstrap for cclasso
boot_cclasso <- function(x, sigma, lambda2, n_boot = 20, 
                         wd, u_f, u0_wd, d0_wd) {
  n <- nrow(x);
  p <- ncol(x);
  
  # Store the result of bootstrap
  cors_boot <- matrix(0, nrow = p * (p - 1)/2, ncol = n_boot + 1);
  vars_boot <- matrix(0, nrow = p, ncol = n_boot + 1);
  cors_mat <- matrix(0, p, p);
  ind_low <- lower.tri(cors_mat);
  
  # Bootstrap procedure
  sam_boot <- matrix(sample(1:n, size = n * n_boot, replace = T), 
                     ncol = n_boot);
  for(k in 1:n_boot) {
    ind_samp <- sam_boot[, k];
    sigma2 <- cclasso_sub(sigma = sigma, vx = var(x[ind_samp, ]), 
                          lambda2 = lambda2, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
    
    vars_boot[, k] <- diag(sigma2);
    Is <- 1 / sqrt(vars_boot[, k]);
    cors_mat <- Is * sigma2 * rep(Is, each = p);
    cors_boot[, k] <- cors_mat[ind_low];
  }
  Is <- 1 / sqrt(diag(sigma));
  cors_mat <- sigma * Is * rep(Is, each = p);
  cors_boot[, n_boot + 1] <- cors_mat[ind_low];
  vars_boot[, n_boot + 1] <- diag(sigma);
  
  # - - - - - - - - - - - - - - - - - - - - - 
  # Variance estimation via bootstrap
  vars2 <- rowMeans(vars_boot);
  # - - - - - - - - - - - - - - - - - - - - - 
  # Correlations' relationship for artificial null sample
  tol_cor <- 1e-3;
  sam_art0 <- matrix(rnorm(n * p), nrow = n) * rep(sqrt(vars2), each = n);
  cors_art0 <- cor(sam_art0)[ind_low];
  sam_art <- sam_art0 - log(rowSums(exp(sam_art0)));
  sigma_art <- cclasso_sub(sigma = sigma, vx = var(sam_art), 
                           lambda2 = lambda2, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
  Is <- 1 / sqrt(diag(sigma_art));
  cors_mat <- Is * sigma_art * rep(Is, each = p);
  cors_art2 <- cors_mat[ind_low];
  # Bias of estimation between absolute data and cclasso of compositional data
  cors0m2 <- log( ((1 + cors_art0) * (1 - cors_art2)) / ((1 + cors_art2) * 
                                                           (1 - cors_art0)) );
  tmp <- abs(cors_art2) >= tol_cor;
  bias02 <- ifelse(sum(tmp), median(abs(cors0m2)[tmp]), 0);
  # Modification of estimation for cclasso
  cors2 <- log( (1 + cors_boot) / (1 - cors_boot) );    
  cors2mod <- (cors_boot >= tol_cor) * (cors2 + bias02) + 
    (cors_boot <= -tol_cor) * (cors2 - bias02);
  cors2mod <- 1 - rowMeans(2 / (exp(cors2mod) + 1));
  cors2_mat <- diag(p);
  cors2_mat[ind_low] <- cors2mod;
  cors2_mat <- t(cors2_mat);
  cors2_mat[ind_low] <- cors2mod;
  # P-values with null distribution of correlation estimations of absolute data
  p_vals <- pt(cors2mod * sqrt((n - 2) / (1 - cors2mod^2)), df = n - 2);
  p_vals <- ifelse(p_vals <= 0.5, p_vals, 1 - p_vals);
  pval_mat <- diag(p);
  pval_mat[ind_low] <- p_vals;
  pval_mat <- t(pval_mat);
  pval_mat[ind_low] <- p_vals;
  # - - - - - - - - - - - - - - - - - - - - - 
  
  return(list(var_w = vars2, cor_w = cors2_mat, p_vals = pval_mat));    
}
# - - - - - - - - - - - - - - - - - - - - - 
# cross validation's loss of cclasso for single lambda
cv_loss_cclasso <- function(lambda2, x, k_cv, sigma, 
                            wd, u_f, u0_wd, d0_wd, wd2) {
  n <- nrow(x);
  p <- ncol(x);
  
  n_b <- floor(n / k_cv);
  cv_loss <- 0;
  for(k in 1:k_cv) {
    itest <- (n_b * (k - 1) + 1):(n_b * k);
    vxk <- stats::var(x[itest, ]);
    vx2k <- stats::var(x[-itest, ]);
    
    sigma <- cclasso_sub(sigma = sigma, vx = vx2k, lambda2 = lambda2, 
                         wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
    
    dsig <- sigma - vxk;
    tmp <- rowMeans(dsig);
    dsig <- dsig - tmp - rep(tmp, each = p) + mean(tmp);
    cv_loss <- cv_loss + base::norm(wd2 * dsig, "F")^2;
  }
  
  return(list(cv_loss = cv_loss, sigma = sigma));
}
# - - - - - - - - - - - - - - - - - - - - - 
# cclasso for single lambda
cclasso_sub <- function(sigma, vx, lambda2, 
                        wd, u_f, u0_wd, d0_wd, 
                        k_max = 200, x_tol = 1e-4) {
  p <- ncol(sigma);
  sigma2 <- sigma;
  LAMBDA <- matrix(0, p, p);
  lambda2 <- matrix(lambda2, p, p);
  diag(lambda2) <- 0;
  
  k <- 0;
  err <- 1;
  while(err > x_tol && k < k_max) {
    # Update sigma
    x_sigma <- t(u_f) %*% ((sigma2 - vx) - LAMBDA) %*% u_f;
    x_sigma[-p,-p] <- u0_wd %*% ((t(u0_wd) %*% x_sigma[-p, -p] %*% u0_wd) * 
                                   d0_wd) %*% t(u0_wd);
    sigma_new <- vx + u_f %*% x_sigma %*% t(u_f);
    # Update sigma2
    A <- LAMBDA + sigma_new;
    sigma2_new <- (A > lambda2) * (A - lambda2) + (A < -lambda2) * 
      (A + lambda2);
    # Update Lambda
    LAMBDA <- LAMBDA + (sigma_new - sigma2_new);
    
    err <- max( abs(sigma_new - sigma)/(abs(sigma) + 1), 
                abs(sigma2_new - sigma2)/(abs(sigma2) + 1) ); 
    k <- k + 1;
    sigma <- sigma_new;
    sigma2 <- sigma2_new;
  }
  
  if(k >= k_max) {
    cat("WARNING of cclasso_sub:\n", "\tMaximum Iteration:", k_max, 
        "&& Relative error:", err, "!\n");
  }
  
  return(sigma);
}
# - - - - - - - - - - - - - - - - - - - - - 

#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # PCaN data - NZ urban resto sites
# # # # # # # # #
# # # # # # # # #

#### PCaN - Metadata
#-------------------------

#### read in site info / metadata

meta <- read_excel(path= "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/Site information copy with coords site name - COPY.xlsx",
                  sheet=1, range="A1:AF20", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)
# 'data.frame':	19 obs. of  32 variables:
# $ sample                 : chr  "ChrMrs" "ChrRic" "DunIsP" "DunSiH" ...
# $ plot_number            : num  57 56 72 65 9 83 2 73 84 74 ...
# $ current_age            : num  14 41 11 31 14 NA 40 23 NA 13 ...
# $ age_category           : chr  "young" "old" "young" "old" ...
# $ site_name              : chr  "Marshland Road" "Riccarton Bush" "Island Park" "Signal Hill" ...
# $ city                   : chr  "Christchurch" "Christchurch" "Dunedin" "Dunedin" ...
# $ year_planted           : chr  "2006" "1979" "2009" "1989" ...
# $ latitude               : num  -43.5 -43.5 -45.9 -45.9 -37.8 ...
# $ longitude              : num  173 173 170 171 175 ...
# $ plot_airtemp_mean      : num  11.66 11.21 11.03 9.54 14.85 ...
# $ plot_airtemp_variance  : num  37.2 24.8 24.8 20.7 29.1 ...
# $ plot_rh_mean           : num  88.1 94.4 80.4 85 84.6 ...
# $ plot_rh_variance       : num  158.9 62.4 168.3 178.6 160.7 ...
# $ seedlings_plot         : num  48 1293 101 131 14 ...
# $ over15cm_seedlings     : num  30 811 38 46 8 NA 251 3 NA 228 ...
# $ canopy_openness_percent: num  55.22 10.61 25.79 8.11 22.1 ...
# $ soiltemp_mean          : num  9.98 NA 9.94 7.78 13.03 ...
# $ soiltemp_variance      : num  5.2 NA 6.31 4.57 6.14 ...
# $ herb_cover_percent     : num  67.38 7.5 28.75 0.25 17.81 ...
# $ total_BA_m2_ha_plot    : num  18.7 43.8 14.4 52.1 23.8 ...
# $ plot_mean_BA_m2        : num  0.00326 0.00985 0.00217 0.00347 0.00183 ...
# $ tree_density_100m2     : num  57.5 44.5 66.5 150 130 ...
# $ species_richness_trees : num  9 11 2 10 7 10 4 1 12 14 ...
# $ species_evenness_trees : num  0.882 0.803 0.664 0.736 0.725 ...
# $ litter_depth_cm        : num  2.07 1.8 1.27 1.33 1.8 ...
# $ WHC                    : num  56.2 59 63.2 53.5 54.9 ...
# $ patch_size_ha          : num  2.44 0.46 9.86 1.36 1.19 30 7.34 3.19 1.65 5 ...
# $ pH                     : num  5.55 5.51 4.3 NA 4.64 4.32 3.43 5.26 3.45 4.69 ...
# $ restoration_status     : chr  "restored" "restored" "restored" "restored" ...
# $ N                      : num  0.166 0.592 0.338 NA 0.451 ...
# $ C                      : num  1.8 8 5.5 NA 7.59 ...
# $ CN                     : num  10.9 13.5 16.3 NA 16.8 ...

row.names(meta) <- meta$sample

( varnames <- names(meta) )
# [1] "sample"                  "plot_number"             "current_age"             "age_category"           
# [5] "site_name"               "city"                    "year_planted"            "latitude"               
# [9] "longitude"               "plot_airtemp_mean"       "plot_airtemp_variance"   "plot_rh_mean"           
# [13] "plot_rh_variance"        "seedlings_plot"          "over15cm_seedlings"      "canopy_openness_percent"
# [17] "soiltemp_mean"           "soiltemp_variance"       "herb_cover_percent"      "total_BA_m2_ha_plot"    
# [21] "plot_mean_BA_m2"         "tree_density_100m2"      "species_richness_trees"  "species_evenness_trees" 
# [25] "litter_depth_cm"         "WHC"                     "patch_size_ha"           "pH"                     
# [29] "restoration_status"      "N"                       "C"                       "CN" 

dim(meta) # 19 32

unique(meta$age_category)
# "young"   "old"     "remnant"

meta$age_category <- factor(meta$age_category, levels = c("young","old","remnant"), ordered = TRUE)

sel <- which(varnames %in% c("current_age","year_planted",
                             "latitude","longitude",
                             "plot_airtemp_mean",      "plot_airtemp_variance" ,  "plot_rh_mean"  ,
                             "plot_rh_variance"  ,      "seedlings_plot"     ,     "over15cm_seedlings"  ,   "canopy_openness_percent",
                             "soiltemp_mean"     ,      "soiltemp_variance"   ,    "herb_cover_percent"  ,    "total_BA_m2_ha_plot"    ,
                             "plot_mean_BA_m2"    ,     "tree_density_100m2"  ,   "species_richness_trees" , "species_evenness_trees" ,
                             "litter_depth_cm"    ,     "WHC"       ,              "patch_size_ha"   ,        "pH"     ,
                             "N"         ,              "C"        ,               "CN" ))

plot_vars <- varnames[sel]
length(plot_vars) # 26

meta.melt <- melt(meta)
# Using sample, age_category, site_name, city, year_planted, restoration_status as id variables

names(meta.melt)
# [1] "sample"             "age_category"       "site_name"          "city"               "year_planted"       "restoration_status" "variable"          
# [8] "value" 

sel.plot <- which(meta.melt$variable %in% plot_vars)

p <- ggplot( meta.melt[sel.plot, ], aes(x = age_category, y = value )) +
  #geom_boxplot()+
  #geom_point() +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(size=1.5,width = 0.15, alpha=0.25) +
  facet_wrap(facets = vars(variable), ncol = 5, scales = "free_y")+
  theme_bw()+
  theme(
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    strip.text = element_text(size = rel(0.7)),
    #strip.text.y = element_text(size = rel(1.1)),
    strip.background = element_blank()
    )

p

dev.print(tiff, file = paste0(workdir,"/plots/","metadata-values.tiff"), width = 22, height = 22, units = "cm", res=600, compression="lzw", type = "cairo")


#-------------------------

#### PCaN - read in superfocus - fxn potential outputs
#-------------------------

manifest <- read.csv(file = "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/2_qc/input-fastq-files-tsv.txt", 
                     header = TRUE, sep = "\t")
class(manifest) # "data.frame"
str(manifest)
# 'data.frame':	19 obs. of  4 variables:
# $ SampleID    : int  10922 10923 10924 10925 10926 10927 10928 10929 10930 10931 ...
# $ Name        : chr  "ChrMrs" "ChrRic" "DunIsp" "DunSiH" ...
# $ R1_filenames: chr  "QGOJO329GC_20210506225159__S11_R1_001.fastq.gz" "QGOJO330GF_20210506225259__S12_R1_001.fastq.gz" "QGOJO327GU_20210506225159__S9_R1_001.fastq.gz" "QGOJO328G4_20210506225159__S10_R1_001.fastq.gz" ...
# $ R2_filenames: chr  "QGOJO329GC_20210506225259__S11_R2_001.fastq.gz" "QGOJO330GF_20210506225259__S12_R2_001.fastq.gz" "QGOJO327GU_20210506225159__S9_R2_001.fastq.gz" "QGOJO328G4_20210506225159__S10_R2_001.fastq.gz" ...

sampid <- manifest$SampleID
length(sampid) # 19
sampid
# 10922 10923 10924 10925 10926 10927 10928 10929 10930 10931 10932 10933 10934 10935 10936 10937 12007 12008 12009

site_names <- manifest$Name
site_names
# [1] "ChrMrs" "ChrRic" "DunIsp" "DunSiH" "HamAva" "HamMin" "InvEsW" "InvKBR" "InvWaR" "NelBot" "NelMRY" "NplHrs" "NplPer" "WeltA"  "WelOCH"
# [16] "WelOWR" "HamCla" "TauCaP" "TauMcB"

names(sampid) <- site_names

superfocus_out_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns"

list.dirs(superfocus_out_dir)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns"                     
# [2] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10922"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10923"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10924"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10925"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10926"
# [7] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10927"
# [8] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10928"
# [9] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10929"
# [10] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10930"
# [11] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10931"
# [12] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10932"
# [13] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10933"
# [14] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10934"
# [15] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10935"
# [16] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10936"
# [17] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10937"
# [18] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_12007"
# [19] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_12008"
# [20] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_12009"

( results_dirs <- list.dirs(superfocus_out_dir)[-1] )
# [1] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10922"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10923"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10924"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10925"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10926"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10927"
# [7] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10928"
# [8] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10929"
# [9] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10930"
# [10] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10931"
# [11] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10932"
# [12] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10933"
# [13] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10934"
# [14] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10935"
# [15] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10936"
# [16] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_10937"
# [17] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_12007"
# [18] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_12008"
# [19] "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_12009"

# check identical order
identical(as.character(sampid), gsub(pattern = "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/DT/PycharmProj-PCaN-NZ/5_superfocus_fxns/superfocus_out_", replacement = "", x = results_dirs) )
# TRUE

# collate results into a long-format table

sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)

for (i in 1:length(sampid)) {
  #i<-1
  this_samp <- sampid[i]
  sel.folder <- grep(pattern = this_samp, x = results_dirs)
  this_folder <- results_dirs[sel.folder]
  
  #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)
  
  tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
  # names(tab)
  # [1] "Subsystem.Level.1"                                                   
  # [2] "Subsystem.Level.2"                                                   
  # [3] "Subsystem.Level.3"                                                   
  # [4] "Function"                                                            
  # [5] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq"  
  # [6] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq.."
  
  tab$sampid <- this_samp
  names(tab)
  
  tab <- tab[,c(7,1,2,3,4,6)]
  names(tab) <- names(sfx.long)
  
  sfx.long <- rbind(sfx.long,tab)
  
  print(paste0("completed ",i," - sample ID: ",sampid[i]))
}

head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
head(sfx.long)
#   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 2    10922 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 5.565040e-04
# 3    10922 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 1.177039e-06
# 4    10922 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.254724e-04
# 5    10922 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 9.816505e-05
# 6    10922 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 2.389389e-03
# 7    10922 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 5.473231e-05

sfx.long$full_fxn_tax <- paste0(sfx.long$subsys_L1,"___", sfx.long$subsys_L2,"___", sfx.long$subsys_L3,"___", sfx.long$fxn)
head(sfx.long)
# sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 2    10922 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 5.565040e-04
# 3    10922 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 1.177039e-06
# 4    10922 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.254724e-04
# 5    10922 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 9.816505e-05
# 6    10922 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 2.389389e-03
# 7    10922 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 5.473231e-05
# full_fxn_tax
# 2                                         Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase
# 3                                                   Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase
# 4 Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)
# 5                       Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)
# 6                                       Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)
# 7                                    Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 36390    20

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 36390
length(unique(full_fxn_names)) # 36390

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 


tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)
# full_fxn_tax        10922        10923
# 1                                         Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase 5.565040e-04 3.655702e-04
# 2                                                   Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase 1.177039e-06 8.301553e-06
# 3 Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.254724e-04 6.445024e-05
# 4                       Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 9.816505e-05 6.097868e-05
# 5                                       Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8) 2.389389e-03 1.965204e-03
# 6                                    Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 5.473231e-05 4.226245e-05
# 10924        10925        10926        10927        10928        10929        10930        10931        10932        10933        10934        10935
# 1 6.641186e-04 3.864582e-04 2.554680e-04 1.468410e-04 6.293564e-04 1.506456e-04 5.844038e-04 2.079619e-04 3.136727e-04 2.956237e-04 2.990754e-04 4.689111e-04
# 2 1.231903e-05 2.974623e-05 4.123620e-06 1.674993e-06 1.591530e-05 8.898894e-06 5.134538e-06 4.358075e-06 5.900641e-06 9.335484e-06 6.928388e-06 1.084048e-05
# 3 1.347701e-04 6.920551e-05 4.045074e-05 3.740817e-05 1.454221e-04 4.801165e-05 1.197948e-04 3.105128e-05 2.789394e-05 4.232086e-05 7.119743e-05 9.166230e-05
# 4 2.240303e-04 1.205113e-04 5.694522e-05 5.359977e-05 1.193789e-04 5.360524e-05 1.602209e-04 4.494265e-05 9.011888e-05 4.418796e-05 6.836009e-05 1.108138e-04
# 5 2.456414e-03 1.932898e-03 1.708357e-03 2.350573e-03 2.262079e-03 2.152261e-03 1.906781e-03 1.902572e-03 2.046450e-03 1.646157e-03 1.866623e-03 2.504151e-03
# 6 4.866015e-05 2.913916e-05 4.663617e-05 1.395827e-05 6.483143e-05 3.775674e-05 5.799694e-05 4.902834e-05 3.285584e-05 3.656398e-05 4.642020e-05 5.143207e-05
# 10936        10937        12007        12008        12009
# 1 4.936050e-04 2.956914e-04 3.450605e-04 4.956079e-04 3.774734e-04
# 2 6.055074e-06 3.677753e-06 2.439374e-06 1.207356e-05 2.037893e-05
# 3 7.924828e-05 4.906264e-05 7.074184e-05 9.386603e-05 1.059087e-04
# 4 1.000252e-04 7.842809e-05 9.934905e-05 6.948217e-05 8.799992e-05
# 5 2.052903e-03 2.068736e-03 1.637707e-03 1.837075e-03 1.940630e-03
# 6 3.085759e-05 3.861641e-05 3.714501e-05 4.142889e-05 3.859646e-05

names(sfx.wide)
# [1] "full_fxn_tax" "10922"        "10923"        "10924"        "10925"        "10926"        "10927"        "10928"        "10929"        "10930"        "10931"        "10932"       
# [13] "10933"        "10934"        "10935"        "10936"        "10937"        "12007"        "12008"        "12009"

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide) # "10922" "10923" "10924" etc
head(sampid)
# ChrMrs ChrRic DunIsp DunSiH HamAva HamMin 
# 10922  10923  10924  10925  10926  10927 

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # TRUE

names(sfx.wide) <- names(sampid)



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 36390     4

length(unique(tax.fxn$subsys_L1)) # 35
length(unique(tax.fxn$subsys_L2)) # 195
length(unique(tax.fxn$subsys_L3)) # 1268
length(unique(tax.fxn$fxn)) # 17218

sel.dup <- which(duplicated.default(tax.fxn$fxn)==TRUE)
head(tax.fxn$fxn[sel.dup])



# examine last example 'tab'

head(tab)
#   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 1    12009 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 3.774734e-04
# 2    12009 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 2.037893e-05
# 3    12009 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.059087e-04
# 4    12009 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 8.799992e-05
# 5    12009 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.940630e-03
# 6    12009 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 3.859646e-05

length(unique(tab$subsys_L1)) # 35
length(unique(tab$subsys_L2)) # 193
length(unique(tab$subsys_L3)) # 1235
length(unique(tab$fxn)) # 13358

#-------------------------

#### PCaN - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 36390     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 36390    19

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 36390 taxa and 19 samples ]
# tax_table()   Taxonomy Table:    [ 36390 taxa by 4 taxonomic ranks ]

sample_names(phy)
# [1] "ChrMrs" "ChrRic" "DunIsp" "DunSiH" "HamAva" "HamMin" "InvEsW" "InvKBR" "InvWaR" "NelBot" "NelMRY" "NplHrs" "NplPer" "WelMtA" "WelOCH" "WelOWR"
# [17] "HamCla" "TauCaP" "TauMcB"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta) # 19 32
identical( row.names(meta), sample_names(phy) ) # FALSE

sel <- which(!row.names(meta) %in% sample_names(phy)) # empty

length( which(row.names(meta) %in% sample_names(phy)) ) # 19


meta2 <- meta[ sample_names(phy), ]
identical( row.names(meta2), sample_names(phy) ) # TRUE

SAMP <- sample_data(meta2)

#phy@sam_data <- SAMP


### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 36390 taxa and 19 samples ]
# sample_data() Sample Data:       [ 19 samples by 32 sample variables ]
# tax_table()   Taxonomy Table:    [ 36390 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#       subsys_L1                     subsys_L2 subsys_L3             fxn                                                                            
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylaconitate_cis-trans_isomerase"                                        
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylaconitate_isomerase"                                                  
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)"
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)"                      
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                      
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 


getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-PCaN-pilot.RDS")

phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")


#-------------------------

#### PCaN - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 36390     4


get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)
        
        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1
          
          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else
        
      } # END else
      
      coefs[[c]] <- coef
      
    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  
  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      

    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/PCaN-NZ-pilot-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%  #
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()



time.start #  "2023-04-05 22:09:12 ACST"
time.finish # "2023-04-06 01:08:10 ACST"


# 36390 rows == 3 hrs


## assemble results

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/PCaN-NZ-pilot-R-working-files"


# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# total results written to disk - 36390
dim(df.tax) # 36390     4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 43977     9

table(df.out$matching_method)
# EC number      Exact hierachy match Matched Reactions aliases    Matched Reactions name     Matched Subsytem role 
# 16445                      5280                       129                       872                      1046 
# No match found 
# 20205 

100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 45.9% No match found

# 54% match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.610   0.800   0.846   1.048   3.333   17654 

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.242   1.400   1.412   1.573   4.000   17654 

table(df.out.distil$matching_method)
# EC number                                  EC number;Exact hierachy match 
# 11401                                                            1131 
# EC number;Exact hierachy match;Matched Reactions name            EC number;Exact hierachy match;Matched Subsytem role 
# 2                                                               6 
# EC number;Exact hierachy match;No match found                             EC number;Matched Reactions aliases 
# 41                                                              12 
# EC number;Matched Reactions name                 EC number;Matched Reactions name;No match found 
# 56                                                               4 
# EC number;Matched Subsytem role                  EC number;Matched Subsytem role;No match found 
# 149                                                               8 
# EC number;No match found                                            Exact hierachy match 
# 1282                                                            2999 
# Exact hierachy match;Matched Reactions aliases                     Exact hierachy match;Matched Reactions name 
# 1                                                              19 
# Exact hierachy match;Matched Subsytem role       Exact hierachy match;Matched Subsytem role;No match found 
# 48                                                               4 
# Exact hierachy match;No match found                                       Matched Reactions aliases 
# 467                                                             104 
# Matched Reactions aliases;Matched Reactions name;No match found                        Matched Reactions aliases;No match found 
# 3                                                               9 
# Matched Reactions name                    Matched Reactions name;Matched Subsytem role 
# 748                                                               4 
# Matched Reactions name;No match found                                           Matched Subsytem role 
# 36                                                             693 
# Matched Subsytem role;No match found                                                  No match found 
# 115                                                           17048

 
100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 38.76572% No match found

100-38.8 # i.e. in 61.2% of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_PCan-pilot.RDS")

df.out.distil <- readRDS(file = "df.out.distil_PCan-pilot.RDS")


#-------------------------

#### PCaN - inspect & clean data
#-------------------------

this_study <- "-PCaN-"
phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")
df.out.distil <- readRDS(file = "df.out.distil_PCan-pilot.RDS")

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 36390     4

phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE


df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 36390    19
df.OTU[1:5, 1:8]
# ChrMrs       ChrRic       DunIsp       DunSiH       HamAva       HamMin       InvEsW       InvKBR
# fxn_1 5.565040e-04 3.655702e-04 6.641186e-04 3.864582e-04 2.554680e-04 1.468410e-04 0.0006293564 1.506456e-04
# fxn_2 1.177039e-06 8.301553e-06 1.231903e-05 2.974623e-05 4.123620e-06 1.674993e-06 0.0000159153 8.898894e-06
# fxn_3 1.254724e-04 6.445024e-05 1.347701e-04 6.920551e-05 4.045074e-05 3.740817e-05 0.0001454221 4.801165e-05
# fxn_4 9.816505e-05 6.097868e-05 2.240303e-04 1.205113e-04 5.694522e-05 5.359977e-05 0.0001193789 5.360524e-05
# fxn_5 2.389389e-03 1.965204e-03 2.456414e-03 1.932898e-03 1.708357e-03 2.350573e-03 0.0022620787 2.152261e-03

mean( df.OTU[ , "ChrMrs"] ) # 0.002748008
sum( df.OTU[ , "ChrMrs"] ) # 100

sample_sums(phy) # all values of 100


identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
#       tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.8571429     0.7142857
# fxn_3     0.9285714     0.8571429
# fxn_4     0.9285714     0.8571429
# fxn_5     0.6000000     1.8000000
# fxn_6     1.0000000     0.4285714



dim(df.fxn_vk_coords) # 36390     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.610   0.800   0.846   1.048   3.333   17654
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.242   1.400   1.412   1.573   4.000   17654 

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.8571429     0.7142857
# fxn_3     0.9285714     0.8571429
# fxn_4     0.9285714     0.8571429
# fxn_5     0.6000000     1.8000000
# fxn_6     1.0000000     0.4285714
# fxn_7     0.6666667     2.3333333


## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 19
dim(df.fxn_vk_coords.noNAs) # 18736     2
dim(df.OTU.noNAs) # 18736  19

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# DunSiH   InvEsW   NplHrs   WelOCH   DunIsp   NelMRY   ChrMrs   InvWaR   ChrRic   WelMtA   NplPer   HamCla   TauMcB   HamAva 
# 60.98754 61.08520 61.44005 61.52752 61.56644 61.61691 61.64620 61.65163 61.65770 61.69973 61.73427 61.84207 61.90441 62.02975 
# TauCaP   NelBot   WelOWR   HamMin   InvKBR 
# 62.03429 62.06474 62.07132 62.23924 62.50473
# this reflects that >61-62% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

colSums(df.OTU)
# all values of 100


dim(df.tax) # 36390     4


# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE



row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  group=NA, 
                                  samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"],
                         group= rep( phy@sam_data$current_age[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}



dim(vk.samp.noNAs.long) # 355985      8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok) # 299776 elements

head(vk.samp.noNAs.long[-sel.ok, ])
#        sample fxn_id         abun abun_type      OC_x      HC_y group   samplabel
# 131259 InvKBR  fxn_2 8.898894e-06   relabun 0.8571429 0.7142857    NA InvKBR (NA)
# 131260 InvKBR  fxn_3 4.801165e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131261 InvKBR  fxn_4 5.360524e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131262 InvKBR  fxn_5 2.152261e-03   relabun 0.6000000 1.8000000    NA InvKBR (NA)
# 131263 InvKBR  fxn_6 3.775674e-05   relabun 1.0000000 0.4285714    NA InvKBR (NA)
# 131264 InvKBR  fxn_8 2.990937e-03   relabun 0.6666667 2.3333333    NA InvKBR (NA)


## don't remove Ref (group = NA) yet
# # removing NA cases will remove Reference group
# vk.samp.noNAs.long <- vk.samp.noNAs.long[sel.ok, ]


str(vk.samp.noNAs.long) 
# 'data.frame':	355984 obs. of  8 variables:
unique(vk.samp.noNAs.long$group) # 14 41 11 31 40 23 NA 13 10 48 28 33

class(vk.samp.noNAs.long$group) # "numeric"
#vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c("6","12","22","31","UM"), ordered = TRUE)
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	355984 obs. of  8 variables:
# $ sample   : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ fxn_id   : Factor w/ 18736 levels "fxn_10","fxn_100",..: 6320 11079 14428 15093 15917 17316 18079 1 602 1054 ...
# $ abun     : num  1.18e-06 1.25e-04 9.82e-05 2.39e-03 5.47e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : num  14 14 14 14 14 14 14 14 14 14 ...
# $ samplabel: chr  "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" ...



saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-PCaN-pilot.RDS")


vk.samp.noNAs.long <- readRDS("vk.samp.noNAs.long-PCaN-pilot.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "PCaN restoration soils", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")




## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/PCaN-NZ-pilot-R-working-files"



i<-2
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                       f__in        matching_method min_adist_modelSEED min_amatch_modelSEED     rxns tot_mean_OC_x tot_mean_HC_y
# 1          fxn_2 1 2-methylaconitate isomerase Matched Reactions name                  NA                   NA rxn25278     0.8571429     0.7142857

temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                       f__in   rxn_id                    rxn_name                             rxn_eqn
# 1          fxn_2 1 2-methylaconitate isomerase rxn25278 2-methylaconitate isomerase (1) cpd25681[0] <=> (1) cpd02597[0]
# rxn_defn            compds compd_coef    chem_formx                           OC_ratios
# 1 (1) 2-methyl-trans-aconitate[0] <=> (1) cis-2-Methylaconitate[0] cpd25681;cpd02597        1;1 C7H5O6;C7H5O6 0.857142857142857;0.857142857142857
# HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 0.714285714285714;0.714285714285714       0.8571429       0.7142857


i<-20000
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                               f__in matching_method min_adist_modelSEED min_amatch_modelSEED
# 1      fxn_20000 1 Lycopene beta cyclase (EC 1.14.-.-)       EC number                  NA                   NA
# rxns
# 1 rxn00447;rxn02968;rxn03577;rxn03582;rxn04267;rxn04326;rxn04328;rxn04329;rxn04331;rxn04692;rxn04697;rxn06793;rxn06925;rxn07105;rxn07109;rxn07275;rxn07277;rxn07281;rxn07282;rxn07283;rxn07284;rxn07288;rxn07289;rxn07290;rxn07303;rxn07501;rxn07511;rxn07514;rxn07520;rxn07600;rxn11529;rxn11686;rxn12018;rxn12022;rxn14064;rxn14086;rxn14107;rxn14195;rxn14251;rxn14259;rxn14273;rxn14331;rxn15608;rxn15609;rxn15610;rxn15611;rxn15755;rxn15967;rxn15969;rxn16008;rxn16009;rxn16023;rxn16029;rxn16031;rxn16033;rxn16036;rxn16040;rxn16052;rxn16072;rxn16290;rxn16292;rxn16298;rxn16299;rxn16301;rxn16605;rxn16606;rxn20720;rxn20723;rxn21585;rxn21587;rxn21753;rxn21756;rxn21760;rxn21763;rxn21766;rxn21769;rxn22061;rxn23195;rxn24247;rxn24250;rxn24256;rxn24257;rxn24332;rxn24334;rxn25115;rxn25464;rxn25469;rxn26751;rxn28816;rxn28825;rxn28826;rxn28827;rxn28831;rxn30826;rxn31246;rxn31345;rxn32218;rxn32219;rxn33025;rxn33027;rxn33028;rxn33074;rxn33075;rxn33077;rxn33293;rxn33294;rxn33544;rxn33547;rxn33771;rxn33801;rxn37063;rxn37064;rxn37065;rxn37066;rxn37067;rxn37068;rxn37069;rxn37070;rxn37071;rxn37072;rxn37073;rxn37074
# tot_mean_OC_x tot_mean_HC_y
# 1     0.2892254      1.358621

temp[[1]][["rxns"]]
# very large number of reactions !!!



sel <- which(vk.samp.noNAs.long$OC_x > 1.3 & vk.samp.noNAs.long$HC_y < 1) # qty 4161
head( vk.samp.noNAs.long[ sel, ] )
#     sample  fxn_id         abun abun_type     OC_x     HC_y group   samplabel
# 502 ChrMrs fxn_772 2.010971e-02   relabun 1.833333 0.500000    14 ChrMrs (14)
# 503 ChrMrs fxn_773 4.590452e-05   relabun 1.667642 0.838357    14 ChrMrs (14)
# 504 ChrMrs fxn_774 3.531117e-06   relabun 1.833333 0.500000    14 ChrMrs (14)
# 505 ChrMrs fxn_775 2.824893e-05   relabun 1.833333 0.500000    14 ChrMrs (14)
# 506 ChrMrs fxn_776 3.531117e-06   relabun 1.833333 0.500000    14 ChrMrs (14)
# 632 ChrMrs fxn_973 8.104754e-06   relabun 1.513095 0.840873    14 ChrMrs (14)


i<-772
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                               f__in matching_method min_adist_modelSEED min_amatch_modelSEED                       rxns tot_mean_OC_x tot_mean_HC_y
# 1        fxn_772 1 Allophanate hydrolase (EC 3.5.1.54)       EC number                  NA                   NA rxn00002;rxn30346;rxn35525      1.833333           0.5

temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                               f__in   rxn_id                            rxn_name                                                                                  rxn_eqn
# 1        fxn_772 1 Allophanate hydrolase (EC 3.5.1.54) rxn00002   urea-1-carboxylate amidohydrolase (1) cpd00001[0] + (3) cpd00067[0] + (1) cpd00742[0] => (2) cpd00011[0] + (2) cpd00013[0]
# 2        fxn_772 1 Allophanate hydrolase (EC 3.5.1.54) rxn30346               allophanate hydrolase (1) cpd00001[0] + (3) cpd00067[0] + (1) cpd00742[0] => (2) cpd00011[0] + (2) cpd00013[0]
# 3        fxn_772 1 Allophanate hydrolase (EC 3.5.1.54) rxn35525 allophanate hydrolase, mitochondria (1) cpd00001[0] + (3) cpd00067[0] + (1) cpd00742[0] => (2) cpd00011[0] + (2) cpd00013[0]
#                                                                 rxn_defn                                       compds compd_coef             chem_formx        OC_ratios      HC_ratios coefwtmean_OC_x
# 1 (1) H2O[0] + (3) H+[0] + (1) Allophanate[0] => (2) CO2[0] + (2) NH3[0] cpd00001;cpd00067;cpd00742;cpd00011;cpd00013  1;3;1;2;2 H2O;H;C2H3N2O3;CO2;H4N NA;NaN;1.5;2;NaN NA;NA;1.5;0;NA        1.833333
# 2 (1) H2O[0] + (3) H+[0] + (1) Allophanate[0] => (2) CO2[0] + (2) NH3[0] cpd00001;cpd00067;cpd00742;cpd00011;cpd00013  1;3;1;2;2 H2O;H;C2H3N2O3;CO2;H4N NA;NaN;1.5;2;NaN NA;NA;1.5;0;NA        1.833333
# 3 (1) H2O[0] + (3) H+[0] + (1) Allophanate[0] => (2) CO2[0] + (2) NH3[0] cpd00001;cpd00067;cpd00742;cpd00011;cpd00013  1;3;1;2;2 H2O;H;C2H3N2O3;CO2;H4N NA;NaN;1.5;2;NaN NA;NA;1.5;0;NA        1.833333
# coefwtmean_HC_y
# 1             0.5
# 2             0.5
# 3             0.5


## check anomalous high value in Lignin zone in PCaN density plot
p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405) # qty 9633

length(sel) # 9633
dim(vk.samp.noNAs.long) # 355984      8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 2.7% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
#     sample  fxn_id         abun abun_type      OC_x     HC_y group   samplabel
# 18  ChrMrs  fxn_22 0.000000e+00   relabun 0.5206427 1.400701    14 ChrMrs (14)
# 32  ChrMrs  fxn_43 2.824893e-05   relabun 0.5206427 1.400701    14 ChrMrs (14)
# 85  ChrMrs fxn_111 3.783339e-06   relabun 0.5206427 1.400701    14 ChrMrs (14)
# 119 ChrMrs fxn_176 1.082876e-05   relabun 0.5206427 1.400701    14 ChrMrs (14)
# 158 ChrMrs fxn_248 7.062234e-06   relabun 0.5206427 1.400701    14 ChrMrs (14)
# 189 ChrMrs fxn_287 7.003382e-06   relabun 0.5206427 1.400701    14 ChrMrs (14)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.04619569 = 0.046% = less than 0.05% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 61.75283 = 61.8% per sample

# remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702) # qty 9633


#fxn_22
i<-22
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED                                                                             rxns tot_mean_OC_x
# 1         fxn_22 1 hypothetical protein Matched Reactions name                  NA                   NA rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427
# tot_mean_HC_y
# 1      1.400701
temp[[1]][["rxns"]]
# [[1]]
#   superfocus_fxn f                f__in   rxn_id                                                      rxn_name
# 1         fxn_22 1 hypothetical protein rxn01869     CHLREDRAFT_140509 hypothetical protein  .SP:A8JHB7_CHLRE.
# 2         fxn_22 1 hypothetical protein rxn05992     CHLREDRAFT_142823 hypothetical protein  .SP:A8I7C1_CHLRE.
# 3         fxn_22 1 hypothetical protein rxn06889     CHLREDRAFT_121409 hypothetical protein  .SP:A8JB38_CHLRE.
# 4         fxn_22 1 hypothetical protein rxn11873     CHLREDRAFT_187700 hypothetical protein  .SP:A8I0K5_CHLRE.
# 5         fxn_22 1 hypothetical protein rxn11874     CHLREDRAFT_187700 hypothetical protein  .SP:A8I0K5_CHLRE.
# 6         fxn_22 1 hypothetical protein rxn31711 glycosyltransferase family protein 28 or hypothetical protein
# 7         fxn_22 1 hypothetical protein rxn34308     CHLREDRAFT_121409 hypothetical protein  .SP:A8JB38_CHLRE.
# 8         fxn_22 1 hypothetical protein rxn34312     CHLREDRAFT_142782 hypothetical protein  .SP:A8I707_CHLRE.
# 9         fxn_22 1 hypothetical protein rxn34322     CHLREDRAFT_186686 hypothetical protein  .SP:A8I9K8_CHLRE.
# rxn_eqn
# 1 (1) cpd00001[0] + (1) cpd00015[0] + (1) cpd00447[0] => (1) cpd00067[0] + (1) cpd00540[0] + (1) cpd00982[0]
# 2                                    (1) cpd00144[0] + (1) cpd11609[0] <=> (1) cpd00014[0] + (1) cpd12241[0]
# 3                                                       (1) cpd04016[0] => (2) cpd00067[0] + (1) cpd15064[0]
# 4                                                                         (1) cpd04740[0] => (1) cpd16416[0]
# 5                                                                         (1) cpd16412[0] => (1) cpd16416[0]
# 6                                     (1) cpd00037[0] + (1) cpd12559[0] <= (1) cpd00014[0] + (1) cpd29932[0]
# 7                                                       (1) cpd04015[0] <= (2) cpd00067[0] + (1) cpd30045[0]
# 8                                     (1) cpd12407[0] + (1) cpd30050[0] <= (1) cpd11619[0] + (1) cpd29780[0]
# 9                                     (1) cpd11456[0] + (1) cpd29781[0] <= (1) cpd11423[0] + (1) cpd30057[0]
# rxn_defn                                                compds  compd_coef
# 1                     (1) H2O[0] + (1) FAD[0] + (1) Betaine aldehyde[0] => (1) H+[0] + (1) BET[0] + (1) FADH2[0] cpd00001;cpd00015;cpd00447;cpd00067;cpd00540;cpd00982 1;1;1;1;1;1
# 2                                           (1) UDPglucuronate[0] + (1) A[0] <=> (1) UDP[0] + (1) Glucuronide[0]                   cpd00144;cpd11609;cpd00014;cpd12241     1;1;1;1
# 3                            (1) cis-3-Chloro-2-propene-1-ol[0] => (2) H+[0] + (1) cis-3-Chloroallyl aldehyde[0]                            cpd04016;cpd00067;cpd15064       1;2;1
# 4                                                                (1) Citalopram[0] => (1) Citalopram aldehyde[0]                                     cpd04740;cpd16416         1;1
# 5                                                        (1) Demethylcitalopram[0] => (1) Citalopram aldehyde[0]                                     cpd16412;cpd16416         1;1
# 6 (1) UDP-N-acetylglucosamine[0] + (1) N-Acetyl-D-glucosaminyldiphosphodolichol[0] <= (1) UDP[0] + (1) G00002[0]                   cpd00037;cpd12559;cpd00014;cpd29932     1;1;1;1
# 7                              (1) trans-3-Chloro-2-propene-1-ol[0] <= (2) H+[0] + (1) 3-Chloroallyl aldehyde[0]                            cpd04015;cpd00067;cpd30045       1;2;1
# 8               (1) Dolichyl phosphate D-mannose[0] + (1) G00147[0] <= (1) Dolichyl phosphate[0] + (1) G00148[0]                   cpd12407;cpd30050;cpd11619;cpd29780     1;1;1;1
# 9               (1) Phosphatidylethanolamine[0] + (1) G00149[0] <= (1) 1,2-Diacyl-sn-glycerol[0] + (1) G13044[0]                   cpd11456;cpd29781;cpd11423;cpd30057     1;1;1;1
# chem_formx                                          OC_ratios                                              HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 H2O;C27H31N9O15P2;C5H12NO;H;C5H11NO2;C27H33N9O15P2 NA;0.555555555555556;0.2;NaN;0.4;0.555555555555556        NA;1.14814814814815;2.4;NA;2.2;1.22222222222222      0.42777778       1.7425926
# 2            C15H19N2O18P2;H2NR;C9H12N2O12P2;C6H8O7R          1.2;NaN;1.33333333333333;1.16666666666667  1.26666666666667;NA;1.33333333333333;1.33333333333333      1.23333333       1.3111111
# 3                                  C3H5ClO;H;C3H3ClO            0.333333333333333;NaN;0.333333333333333                                  1.66666666666667;NA;1      0.33333333       1.3333333
# 4                              C20H22FN2O;C18H14FNO2                             0.05;0.111111111111111                                  1.1;0.777777777777778      0.08055556       0.9388889
# 5                              C19H20FN2O;C18H14FNO2               0.0526315789473684;0.111111111111111                     1.05263157894737;0.777777777777778      0.08187135       0.9152047
# 6       C17H25N3O17P2;C33H57NO12P2;C9H12N2O12P2;null           1;0.363636363636364;1.33333333333333;NaN 1.47058823529412;1.72727272727273;1.33333333333333;NaN      0.89898990       1.5103981
# 7                                     C3H5ClO;H;null                          0.333333333333333;NaN;NaN                                1.66666666666667;NA;NaN      0.33333333       1.6666667
# 8                      C31H54O9P;null;C25H43O4P;null                     0.290322580645161;NaN;0.16;NaN                          1.74193548387097;NaN;1.72;NaN      0.22516129       1.7309677
# 9                     C7H12NO8PR2;null;C5H6O5R2;null                         1.14285714285714;NaN;1;NaN                           1.71428571428571;NaN;1.2;NaN      1.07142857       1.4571429

# fxn_43
i<-43
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED                                                                             rxns tot_mean_OC_x
# 1         fxn_43 1 hypothetical protein Matched Reactions name                  NA                   NA rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427
# tot_mean_HC_y
# 1      1.400701
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                f__in   rxn_id                                                      rxn_name
# 1         fxn_43 1 hypothetical protein rxn01869     CHLREDRAFT_140509 hypothetical protein  .SP:A8JHB7_CHLRE.
# 2         fxn_43 1 hypothetical protein rxn05992     CHLREDRAFT_142823 hypothetical protein  .SP:A8I7C1_CHLRE.
# 3         fxn_43 1 hypothetical protein rxn06889     CHLREDRAFT_121409 hypothetical protein  .SP:A8JB38_CHLRE.
# 4         fxn_43 1 hypothetical protein rxn11873     CHLREDRAFT_187700 hypothetical protein  .SP:A8I0K5_CHLRE.
# 5         fxn_43 1 hypothetical protein rxn11874     CHLREDRAFT_187700 hypothetical protein  .SP:A8I0K5_CHLRE.
# 6         fxn_43 1 hypothetical protein rxn31711 glycosyltransferase family protein 28 or hypothetical protein
# 7         fxn_43 1 hypothetical protein rxn34308     CHLREDRAFT_121409 hypothetical protein  .SP:A8JB38_CHLRE.
# 8         fxn_43 1 hypothetical protein rxn34312     CHLREDRAFT_142782 hypothetical protein  .SP:A8I707_CHLRE.
# 9         fxn_43 1 hypothetical protein rxn34322     CHLREDRAFT_186686 hypothetical protein  .SP:A8I9K8_CHLRE.
# rxn_eqn
# 1 (1) cpd00001[0] + (1) cpd00015[0] + (1) cpd00447[0] => (1) cpd00067[0] + (1) cpd00540[0] + (1) cpd00982[0]
# 2                                    (1) cpd00144[0] + (1) cpd11609[0] <=> (1) cpd00014[0] + (1) cpd12241[0]
# 3                                                       (1) cpd04016[0] => (2) cpd00067[0] + (1) cpd15064[0]
# 4                                                                         (1) cpd04740[0] => (1) cpd16416[0]
# 5                                                                         (1) cpd16412[0] => (1) cpd16416[0]
# 6                                     (1) cpd00037[0] + (1) cpd12559[0] <= (1) cpd00014[0] + (1) cpd29932[0]
# 7                                                       (1) cpd04015[0] <= (2) cpd00067[0] + (1) cpd30045[0]
# 8                                     (1) cpd12407[0] + (1) cpd30050[0] <= (1) cpd11619[0] + (1) cpd29780[0]
# 9                                     (1) cpd11456[0] + (1) cpd29781[0] <= (1) cpd11423[0] + (1) cpd30057[0]
# rxn_defn                                                compds  compd_coef
# 1                     (1) H2O[0] + (1) FAD[0] + (1) Betaine aldehyde[0] => (1) H+[0] + (1) BET[0] + (1) FADH2[0] cpd00001;cpd00015;cpd00447;cpd00067;cpd00540;cpd00982 1;1;1;1;1;1
# 2                                           (1) UDPglucuronate[0] + (1) A[0] <=> (1) UDP[0] + (1) Glucuronide[0]                   cpd00144;cpd11609;cpd00014;cpd12241     1;1;1;1
# 3                            (1) cis-3-Chloro-2-propene-1-ol[0] => (2) H+[0] + (1) cis-3-Chloroallyl aldehyde[0]                            cpd04016;cpd00067;cpd15064       1;2;1
# 4                                                                (1) Citalopram[0] => (1) Citalopram aldehyde[0]                                     cpd04740;cpd16416         1;1
# 5                                                        (1) Demethylcitalopram[0] => (1) Citalopram aldehyde[0]                                     cpd16412;cpd16416         1;1
# 6 (1) UDP-N-acetylglucosamine[0] + (1) N-Acetyl-D-glucosaminyldiphosphodolichol[0] <= (1) UDP[0] + (1) G00002[0]                   cpd00037;cpd12559;cpd00014;cpd29932     1;1;1;1
# 7                              (1) trans-3-Chloro-2-propene-1-ol[0] <= (2) H+[0] + (1) 3-Chloroallyl aldehyde[0]                            cpd04015;cpd00067;cpd30045       1;2;1
# 8               (1) Dolichyl phosphate D-mannose[0] + (1) G00147[0] <= (1) Dolichyl phosphate[0] + (1) G00148[0]                   cpd12407;cpd30050;cpd11619;cpd29780     1;1;1;1
# 9               (1) Phosphatidylethanolamine[0] + (1) G00149[0] <= (1) 1,2-Diacyl-sn-glycerol[0] + (1) G13044[0]                   cpd11456;cpd29781;cpd11423;cpd30057     1;1;1;1
# chem_formx                                          OC_ratios                                              HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 H2O;C27H31N9O15P2;C5H12NO;H;C5H11NO2;C27H33N9O15P2 NA;0.555555555555556;0.2;NaN;0.4;0.555555555555556        NA;1.14814814814815;2.4;NA;2.2;1.22222222222222      0.42777778       1.7425926
# 2            C15H19N2O18P2;H2NR;C9H12N2O12P2;C6H8O7R          1.2;NaN;1.33333333333333;1.16666666666667  1.26666666666667;NA;1.33333333333333;1.33333333333333      1.23333333       1.3111111
# 3                                  C3H5ClO;H;C3H3ClO            0.333333333333333;NaN;0.333333333333333                                  1.66666666666667;NA;1      0.33333333       1.3333333
# 4                              C20H22FN2O;C18H14FNO2                             0.05;0.111111111111111                                  1.1;0.777777777777778      0.08055556       0.9388889
# 5                              C19H20FN2O;C18H14FNO2               0.0526315789473684;0.111111111111111                     1.05263157894737;0.777777777777778      0.08187135       0.9152047
# 6       C17H25N3O17P2;C33H57NO12P2;C9H12N2O12P2;null           1;0.363636363636364;1.33333333333333;NaN 1.47058823529412;1.72727272727273;1.33333333333333;NaN      0.89898990       1.5103981
# 7                                     C3H5ClO;H;null                          0.333333333333333;NaN;NaN                                1.66666666666667;NA;NaN      0.33333333       1.6666667
# 8                      C31H54O9P;null;C25H43O4P;null                     0.290322580645161;NaN;0.16;NaN                          1.74193548387097;NaN;1.72;NaN      0.22516129       1.7309677
# 9                     C7H12NO8PR2;null;C5H6O5R2;null                         1.14285714285714;NaN;1;NaN                           1.71428571428571;NaN;1.2;NaN      1.07142857       1.4571429


## check anomaly in Amino sugars ??

p + xlim(0.5, 0.6) + ylim(2,2.3)
p + xlim(0.575, 0.6) + ylim(2.1,2.2)
sel <- which(vk.samp.noNAs.long$OC_x > 0.575 & vk.samp.noNAs.long$OC_x < 0.6 & vk.samp.noNAs.long$HC_y > 2.1 & vk.samp.noNAs.long$HC_y < 2.2)

length(sel) # 3135
dim(vk.samp.noNAs.long) # 355984      8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 0.88% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
#    sample fxn_id         abun abun_type      OC_x     HC_y group   samplabel
# 68 ChrMrs fxn_90 0.000000e+00   relabun 0.5860141 2.159295    14 ChrMrs (14)
# 69 ChrMrs fxn_91 4.177370e-03   relabun 0.5860141 2.159295    14 ChrMrs (14)
# 70 ChrMrs fxn_92 7.062234e-06   relabun 0.5860141 2.159295    14 ChrMrs (14)
# 71 ChrMrs fxn_93 3.079919e-06   relabun 0.5860141 2.159295    14 ChrMrs (14)
# 72 ChrMrs fxn_94 6.406297e-04   relabun 0.5860141 2.159295    14 ChrMrs (14)
# 73 ChrMrs fxn_95 2.154932e-02   relabun 0.5860141 2.159295    14 ChrMrs (14)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.1942382 = 0.194% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 61.75283 = 61.8% per sample


# fxn_90
i<-90
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                                          f__in matching_method min_adist_modelSEED min_amatch_modelSEED
# 1         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift       EC number                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn07292;rxn23023;rxn25837;rxn26037;rxn28193;rxn28649;rxn31336;rxn38508;rxn45384     0.5860141      2.159295
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                                          f__in   rxn_id
# 1         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn07292
# 2         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn23023
# 3         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn25837
# 4         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn26037
# 5         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn28193
# 6         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn28649
# 7         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn31336
# 8         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn38508
# 9         fxn_90 1 Cysteine desulfurase (EC 2.8.1.7) # frameshift rxn45384

# fxn_91
i<-91
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                                             f__in matching_method min_adist_modelSEED min_amatch_modelSEED
# 1         fxn_91 1 Cysteine desulfurase (EC 2.8.1.7), IscS subfamily       EC number                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn07292;rxn23023;rxn25837;rxn26037;rxn28193;rxn28649;rxn31336;rxn38508;rxn45384     0.5860141      2.159295
temp[[1]][["rxns"]]



## check anomalous high value in Other demethylated compounds zone in PCaN density plot

p + xlim(1, 1.3) + ylim(1,1.5)
p + xlim(1.1, 1.2) + ylim(1.25,1.32)
p + xlim(1.14, 1.155) + ylim(1.29,1.305)
sel <- which(vk.samp.noNAs.long$OC_x > 1.145 & vk.samp.noNAs.long$OC_x < 1.155 & vk.samp.noNAs.long$HC_y > 1.295 & vk.samp.noNAs.long$HC_y < 1.305)

length(sel) # 4085
dim(vk.samp.noNAs.long) # 355984      8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 1.147523% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
#      sample   fxn_id         abun abun_type OC_x HC_y group   samplabel
# 481  ChrMrs  fxn_732 1.341824e-05   relabun 1.15  1.3    14 ChrMrs (14)
# 2131 ChrMrs fxn_2838 0.000000e+00   relabun 1.15  1.3    14 ChrMrs (14)
# 2148 ChrMrs fxn_2863 1.601362e-02   relabun 1.15  1.3    14 ChrMrs (14)
# 2149 ChrMrs fxn_2864 1.297323e-02   relabun 1.15  1.3    14 ChrMrs (14)
# 2150 ChrMrs fxn_2865 1.765558e-06   relabun 1.15  1.3    14 ChrMrs (14)
# 2151 ChrMrs fxn_2868 1.163020e-02   relabun 1.15  1.3    14 ChrMrs (14)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 1.590633 = 1.59% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 61.75283 = 61.8% per sample



#fxn_732
i<-732
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                                                                 f__in matching_method min_adist_modelSEED
# 1        fxn_732 1 Sulfate and thiosulfate import ATP-binding protein CysA (EC 3.6.3.25)       EC number                  NA
# min_amatch_modelSEED                       rxns tot_mean_OC_x tot_mean_HC_y
# 1                   NA rxn05153;rxn09276;rxn29239          1.15           1.3
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                                                                 f__in   rxn_id                    rxn_name
# 1        fxn_732 1 Sulfate and thiosulfate import ATP-binding protein CysA (EC 3.6.3.25) rxn05153 sulfate-transporting ATPase
# 2        fxn_732 1 Sulfate and thiosulfate import ATP-binding protein CysA (EC 3.6.3.25) rxn09276               ABC-70-RXN.cp
# 3        fxn_732 1 Sulfate and thiosulfate import ATP-binding protein CysA (EC 3.6.3.25) rxn29239        TRANS-RXN4LZ-6845.cd
# rxn_eqn
# 1  (1) cpd00001[0] + (1) cpd00002[0] + (1) cpd00048[1] => (1) cpd00008[0] + (1) cpd00009[0] + (1) cpd00048[0] + (1) cpd00067[0]
# 2  (1) cpd00001[0] + (1) cpd00002[0] + (1) cpd00048[1] => (1) cpd00008[0] + (1) cpd00009[0] + (1) cpd00048[0] + (1) cpd00067[0]
# 3 (1) cpd00001[0] + (1) cpd00002[0] + (1) cpd00048[1] <=> (1) cpd00008[0] + (1) cpd00009[0] + (1) cpd00048[0] + (1) cpd00067[1]
# rxn_defn
# 1  (1) H2O[0] + (1) ATP[0] + (1) Sulfate[1] => (1) ADP[0] + (1) Phosphate[0] + (1) Sulfate[0] + (1) H+[0]
# 2  (1) H2O[0] + (1) ATP[0] + (1) Sulfate[1] => (1) ADP[0] + (1) Phosphate[0] + (1) Sulfate[0] + (1) H+[0]
# 3 (1) H2O[0] + (1) ATP[0] + (1) Sulfate[1] <=> (1) ADP[0] + (1) Phosphate[0] + (1) Sulfate[0] + (1) H+[1]
# compds    compd_coef                                     chem_formx
# 1 cpd00001;cpd00002;cpd00048;cpd00008;cpd00009;cpd00048;cpd00067 1;1;1;1;1;1;1 H2O;C10H13N5O13P3;O4S;C10H13N5O10P2;HO4P;O4S;H
# 2 cpd00001;cpd00002;cpd00048;cpd00008;cpd00009;cpd00048;cpd00067 1;1;1;1;1;1;1 H2O;C10H13N5O13P3;O4S;C10H13N5O10P2;HO4P;O4S;H
# 3 cpd00001;cpd00002;cpd00048;cpd00008;cpd00009;cpd00048;cpd00067 1;1;1;1;1;1;1 H2O;C10H13N5O13P3;O4S;C10H13N5O10P2;HO4P;O4S;H
# OC_ratios                HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 NA;1.3;NA;1;NA;NA;NaN NA;1.3;NaN;1.3;NA;NaN;NA            1.15             1.3
# 2 NA;1.3;NA;1;NA;NA;NaN NA;1.3;NaN;1.3;NA;NaN;NA            1.15             1.3
# 3 NA;1.3;NA;1;NA;NA;NaN NA;1.3;NaN;1.3;NA;NaN;NA            1.15             1.3


# fxn_2838
i<-2838
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                                                           f__in       matching_method min_adist_modelSEED
# 1       fxn_2838 1 Phosphate transport system permease protein PstC (TC 3.A.1.7.1) Matched Subsytem role                  NA
# min_amatch_modelSEED     rxns tot_mean_OC_x tot_mean_HC_y
# 1                   NA rxn05145          1.15           1.3
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                                                           f__in   rxn_id                      rxn_name
# 1       fxn_2838 1 Phosphate transport system permease protein PstC .TC 3.A.1.7.1. rxn05145 phosphate-transporting ATPase
# rxn_eqn
# 1 (1) cpd00001[0] + (1) cpd00002[0] + (1) cpd00009[1] => (1) cpd00008[0] + (2) cpd00009[0] + (1) cpd00067[0]
# rxn_defn                                                compds
# 1 (1) H2O[0] + (1) ATP[0] + (1) Phosphate[1] => (1) ADP[0] + (2) Phosphate[0] + (1) H+[0] cpd00001;cpd00002;cpd00009;cpd00008;cpd00009;cpd00067
# compd_coef                                  chem_formx          OC_ratios           HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 1;1;1;1;2;1 H2O;C10H13N5O13P3;HO4P;C10H13N5O10P2;HO4P;H NA;1.3;NA;1;NA;NaN NA;1.3;NA;1.3;NA;NA            1.15             1.3



### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 9633


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-PCaN-pilot.RDS")


#-------------------------

#### PCaN - CPP-class: total rel abun in vk spatial regions -- Split into strongly acidic (n=8) and acidic to neutral (n=10) pH groups
#-------------------------

this_study <- "-PCaN-"
header <- "-vk-zones-"

phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")
df.out.distil <- readRDS(file = "df.out.distil_PCan-pilot.RDS")
vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-PCaN-pilot.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 36390     4


ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok) # 299776 elements

head(vk.samp.noNAs.long[-sel.ok, ])
#        sample fxn_id         abun abun_type      OC_x      HC_y group   samplabel
# 131259 InvKBR  fxn_2 8.898894e-06   relabun 0.8571429 0.7142857    NA InvKBR (NA)
# 131260 InvKBR  fxn_3 4.801165e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131261 InvKBR  fxn_4 5.360524e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131262 InvKBR  fxn_5 2.152261e-03   relabun 0.6000000 1.8000000    NA InvKBR (NA)
# 131263 InvKBR  fxn_6 3.775674e-05   relabun 1.0000000 0.4285714    NA InvKBR (NA)
# 131264 InvKBR  fxn_8 2.990937e-03   relabun 0.6666667 2.3333333    NA InvKBR (NA)


## don't remove Ref (group = NA) yet
# # removing NA cases will remove Reference group
# vk.samp.noNAs.long <- vk.samp.noNAs.long[sel.ok, ]


str(vk.samp.noNAs.long) 
# 'data.frame':	346351 obs. of  8 variables:
unique(vk.samp.noNAs.long$group) # 14 41 11 31 40 23 NA 13 10 48 28 33

class(vk.samp.noNAs.long$group) # "numeric"
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	346351 obs. of  8 variables:
# $ sample   : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ fxn_id   : Factor w/ 18229 levels "fxn_10","fxn_100",..: 6176 10762 14006 14662 15473 16843 17592 1 588 1021 ...
# $ abun     : num  1.18e-06 1.25e-04 9.82e-05 2.39e-03 5.47e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : num  14 14 14 14 14 14 14 14 14 14 ...
# $ samplabel: chr  "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" ...



# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 


spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) #  278281      5

table(spdf$group, useNA = "ifany")
# 10    11    13    14    23    28    31    33    40    41    48  <NA> 
#   29017 30000 14846 45164 15131 14747 27858 14785 14123 14763 14131 43716

head(spdf)
# sample group      OC_x      HC_y         abun
# 2 ChrMrs    14 0.8571429 0.7142857 1.177039e-06
# 3 ChrMrs    14 0.9285714 0.8571429 1.254724e-04
# 4 ChrMrs    14 0.9285714 0.8571429 9.816505e-05
# 5 ChrMrs    14 0.6000000 1.8000000 2.389389e-03
# 6 ChrMrs    14 1.0000000 0.4285714 5.473231e-05
# 7 ChrMrs    14 0.6666667 2.3333333 3.088441e-03

#dim(spdf) # 


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 19

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 60.94   61.55   61.65   61.71   61.98   62.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"

## now want to sum total rela abun within vK spatial zones

## to de-duplicate get total relabun within a sample, then mean across samples

sum_relabun_in_vk_zones_env <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  # lipids
  subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  subsel.all <- c(subsel.lipids)
  # proteins
  subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.proteins)
  # amino_sugars
  subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.amino_sugars)
  # carbs - Carbohydrates
  subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  subsel.all <- c(subsel.all, subsel.carbs)
  # cond_aromtx - Condensed aromatics
  subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  subsel.all <- c(subsel.all, subsel.cond_aromtx)
  # lignin
  subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.lignin)
  # tannins
  subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.tannins)
  # Other compounds ...
  
  # methylated
  subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
  # and not already used?
  sel.dup <- which(subsel.methylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.methylated)
  
  # hydrated
  subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
  # and not already used?
  sel.dup <- which(subsel.hydrated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.hydrated)
  
  # demethylated / Oxidised-dehydrogenated
  subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.demethylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.demethylated)
  
  # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
  #subsel.condensed = which(                   spdf_in$OC_x <= 0.97                       & spdf_in$HC_y < 0.75)
  subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
  
  # and not already used?
  sel.dup <- which(subsel.condensed %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.condensed)
  
  # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
  subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.k2_b12 %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
  
  df_out <- data.frame(
    sample = this_samp,
    group = unique(spdf_in$group),
    
    lipids = sum(spdf_in$abun[ subsel.lipids ]),
    proteins = sum(spdf_in$abun[ subsel.proteins ]),
    amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
    carbs = sum(spdf_in$abun[ subsel.carbs ]),
    cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
    lignin = sum(spdf_in$abun[ subsel.lignin ]),
    tannins = sum(spdf_in$abun[ subsel.tannins ]),
    # Other compounds ...
    methylated = sum(spdf_in$abun[ subsel.methylated ]),
    hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
    demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
    condensed = sum(spdf_in$abun[ subsel.condensed ]),
    k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
  )
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones_env( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 19
out[[1]]
# sample group    lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1 ChrMrs    14 0.2373312 2.670232     4.661897 9.685132   0.9440172 8.331892 15.91626  0.4793873 4.710633      13.0427 0.4759027 0.4448641

names(out[[1]])
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"

df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample group    lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1 ChrMrs    14 0.2373312 2.670232     4.661897 9.685132   0.9440172 8.331892 15.91626  0.4793873 4.710633     13.04270 0.4759027 0.4448641
# 2 ChrRic    41 0.3093743 2.606441     4.546414 9.223352   1.0282546 8.273361 16.12118  0.4859519 4.548059     13.24971 0.7208332 0.4969022
# 3 DunIsp    11 0.2951763 2.832572     4.892685 9.705331   0.8589427 8.275916 15.69230  0.4542437 4.749909     12.87591 0.4095038 0.4852845
# 4 DunSiH    31 0.3203347 2.561207     4.778400 9.868985   1.0619048 8.128550 15.32182  0.4728172 4.649300     12.78093 0.5326770 0.4654128
# 5 HamAva    14 0.3019526 2.622486     4.709111 9.473698   1.0366787 8.315531 16.05144  0.4624811 4.696233     13.31069 0.4991512 0.5025764
# 6 HamMin    40 0.3888312 2.578161     4.560157 9.400301   1.0067409 8.174625 15.98705  0.4491056 4.960649     13.74471 0.4060273 0.5372105

names(df.zones)
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"
names(df.zones[ ,c(3:14)])
# [1] "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"   "hydrated"     "demethylated"
# [11] "condensed"    "k2_b12"

df.zones$all <- apply(X = df.zones[ ,c(3:14)], MARGIN = 1, FUN = sum)


# set Ref group
sel <- which(is.na(df.zones$group)==TRUE)
df.zones$group[sel] <- "Rem"

df.zones$pH <- NA

# add pH data
for (i in 1:dim(df.zones)[1]) {
  #i<-1
  this_samp <- df.zones$sample[i]
  sel.phy <- which(phy@sam_data$sample == this_samp)
  df.zones$pH[i] <- phy@sam_data$pH[sel.phy]
  print(paste0("completed ",i))
}

df.zones$pH_class <- NA
sel <- which(df.zones$pH <4.5) # <5.5 qty 15; <5 qty 11; <4.5 qty 8
df.zones$pH_class[sel] <- "low"
sel <- which(df.zones$pH >=4.5) # >=4.5 qty 10
df.zones$pH_class[sel] <- "mod"

ok <- complete.cases(df.zones)
sel.ok <- which(ok==TRUE)
df.zones <- df.zones[sel.ok, ]

df.zones$group <- factor(df.zones$group, levels = c(as.character(1:50),"Rem"), ordered = TRUE )

saveRDS(object = df.zones, file = "df.zones--PCaN-pilot-Extra-pH-level.RDS")

df.zones <- readRDS("df.zones--PCaN-pilot-Extra-pH-level.RDS")


table(df.zones$pH_class)
# low mod 
# 8  10


str(df.zones)
# 'data.frame':	18 obs. of  17 variables:
# $ sample      : chr  "ChrMrs" "ChrRic" "DunIsp" "HamAva" ...
# $ group       : Ord.factor w/ 51 levels "1"<"2"<"3"<"4"<..: 14 41 11 14 40 23 51 13 31 10 ...
# $ lipids      : num  0.237 0.309 0.295 0.302 0.389 ...
# $ proteins    : num  2.67 2.61 2.83 2.62 2.58 ...
# $ amino_sugars: num  4.66 4.55 4.89 4.71 4.56 ...
# $ carbs       : num  9.69 9.22 9.71 9.47 9.4 ...
# $ cond_aromtx : num  0.944 1.028 0.859 1.037 1.007 ...
# $ lignin      : num  8.33 8.27 8.28 8.32 8.17 ...
# $ tannins     : num  15.9 16.1 15.7 16.1 16 ...
# $ methylated  : num  0.479 0.486 0.454 0.462 0.449 ...
# $ hydrated    : num  4.71 4.55 4.75 4.7 4.96 ...
# $ demethylated: num  13 13.2 12.9 13.3 13.7 ...
# $ condensed   : num  0.476 0.721 0.41 0.499 0.406 ...
# $ k2_b12      : num  0.445 0.497 0.485 0.503 0.537 ...
# $ all         : num  61.6 61.6 61.5 62 62.2 ...
# $ pH          : num  5.55 5.51 4.3 4.64 3.43 5.26 3.45 4.69 3.81 4.17 ...
# $ pH_class    : chr  "mod" "mod" "low" "mod" ...


df.zones.melt <- melt(df.zones, id.vars = c("sample","group","pH","pH_class"))
head(df.zones.melt)
#   sample group   pH pH_class variable     value
# 1 ChrMrs    14 5.55      mod   lipids 0.2373312
# 2 ChrRic    41 5.51      mod   lipids 0.3093743
# 3 DunIsp    11 4.30      low   lipids 0.2951763
# 4 HamAva    14 4.64      mod   lipids 0.3019526
# 5 HamMin    40 3.43      low   lipids 0.3888312
# 6 InvEsW    23 5.26      mod   lipids 0.2660408



str(df.zones.melt)
# 'data.frame':	234 obs. of  6 variables:
# $ sample  : chr  "ChrMrs" "ChrRic" "DunIsp" "HamAva" ...
# $ group   : Ord.factor w/ 51 levels "1"<"2"<"3"<"4"<..: 14 41 11 14 40 23 51 13 31 10 ...
# $ pH      : num  5.55 5.51 4.3 4.64 3.43 5.26 3.45 4.69 3.81 4.17 ...
# $ pH_class: chr  "mod" "mod" "low" "mod" ...
# $ variable: Factor w/ 13 levels "lipids","proteins",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0.237 0.309 0.295 0.302 0.389 ...


saveRDS(object = df.zones.melt, file = "df.zones.melt.vkZones-PCaN-pilot-Extra-pH-level.RDS")

#df.zones.melt <- readRDS("df.zones.melt.vkZones-PCaN-pilot-Extra-pH-level.RDS")


summary(filter(df.zones.melt, pH_class == "low")[ ,"pH" ])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 3.430   3.465   3.990   3.901   4.270   4.320 
unique(sort(filter(df.zones.melt, pH_class == "low")[ ,"group" ]))
# 10  11  31  40  Rem

summary(filter(df.zones.melt, pH_class == "mod")[ ,"pH" ])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 4.640   4.840   5.235   5.309   5.510   6.950
unique(sort(filter(df.zones.melt, pH_class == "mod")[ ,"group" ]))
# 11 13 14 23 28 33 41 48


## create named numeric vector for x-axis display in facet_grid

dat <- df.zones.melt

dat$age_vec_plot <- as.character( dat$group )

sort(unique( dat$age_vec_plot ))
# "10"  "11"  "13"  "14"  "23"  "28"  "31"  "33"  "40"  "41"  "48"  "Rem"

dat$age_vec_plot <- factor(dat$age_vec_plot,
                           levels = c(
                             "10",  "11",  "13",  "14",  "23",  "28",  "31" , "33",  "40",  "41" , "48" , "Rem"
                           ),
                           labels = c(
                             "10",  "11",  "13", "14",  "23",  "28",  "31" , "33" , "40", "41" , "48", "58"
                             
                           ), 
                           ordered = TRUE)

dat$age_vec_plot_num <- as.integer(as.character(dat$age_vec_plot))
sort(unique( dat$age_vec_plot_num))
#   10 11 13 14 23 28 31 33 40 41 48 58


age_vec_plot_num <- c(1:60)

x <- c(
  rep(NA, times=9),"10", # 1:10
  rep(NA, times=9),"20", # 11:20
  rep(NA, times=9),"30", # 21:30
  rep(NA, times=9),"40", # 31:40
  rep(NA, times=10),     # 41:50
  NA,NA,NA,NA,NA,NA,NA,"Rem",NA,NA  # 51:60
  )

names(age_vec_plot_num) <- x

label_breaks <- which( is.na( names(age_vec_plot_num) ) == FALSE)
label_breaks
# 10 20 30 40 58

names( age_vec_plot_num[ label_breaks ] )
# "10"  "20"  "30"  "40"  "Rem"


str(dat)

p <- ggplot(data = dat, aes(x = age_vec_plot_num, y = value))+
  #geom_point()+
  #geom_smooth(method="lm")+
  
  geom_point(aes(color=pH_class))+
  geom_smooth(aes(color=pH_class),method="lm")+
  
  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+
  
  facet_wrap(facets = vars(variable), scales = "free_y")

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-Trendlines-",this_study,header,"-v9d-Extra-pH-level.tiff"), width = 20, height = 20, units = "cm", res=350, compression="lzw",type="cairo")



unique(dat$group)
# [1] 14  41  11  40  23  Rem 13  31  10  48  28  33 
# 51 Levels: 1 < 2 < 3 < 4 < 5 < 6 < 7 < 8 < 9 < 10 < 11 < 12 < 13 < 14 < 15 < 16 < 17 < 18 < 19 < 20 < 21 < ... < Rem


str(dat)


## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"


head(dat)
#   sample group   pH pH_class variable     value age_vec_plot age_vec_plot_num
# 1 ChrMrs    14 5.55      mod   Lipids 0.2373312           14               14
# 2 ChrRic    41 5.51      mod   Lipids 0.3093743           41               41
# 3 DunIsp    11 4.30      low   Lipids 0.2951763           11               11
# 4 HamAva    14 4.64      mod   Lipids 0.3019526           14               14
# 5 HamMin    40 3.43      low   Lipids 0.3888312           40               40
# 6 InvEsW    23 5.26      mod   Lipids 0.2660408           23               23


names(dat)
# [1] "sample"           "group"            "pH"               "pH_class"         "variable"         "value"           
# [7] "age_vec_plot"     "age_vec_plot_num"


sig_symb
# function(x) {
#   out <- character(length = length(x))
#   for (i in seq_along(x)) {
#     if (is.na(x[i])) {
#       out[i] <- NA
#     } else if (x[i] < 0.001) { 
#       out[i] <- "***"
#     } else if (x[i] >= 0.001 & x[i] < 0.01) { 
#       out[i] <- "**"
#     } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
#       out[i] <- "*" 
#     } else { 
#       out[i] <- "ns"
#     }
#   } # END loop
#   return(out)
# }
# <bytecode: 0x7fe3b2960ef0>


pos_anotations <- list()
sig_anotations <- list()
Kendall_Tau <- list()



## "Lipids"
this_var <- "Lipids"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 1.5407, p-value = 0.1234
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
# tau 
# 0.46291

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -1.0894, p-value = 0.276
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
# tau 
# -0.2760262 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Proteins"
this_var <- "Proteins"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -0.5447, p-value = 0.586
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.1380131

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Amino sugars"
this_var <- "Amino sugars"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -0.18157, p-value = 0.8559
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.04600437

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Carbohydrates"                
this_var <- "Carbohydrates"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = -1.7974, p-value = 0.07227
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.5400617 
plot(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"])

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -1.8157, p-value = 0.06942
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.4600437 
plot(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"])

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = -0.25678, p-value = 0.7974
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.07715167 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = 1.271, p-value = 0.2037
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3220306 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Lignin" 
this_var <- "Lignin"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = -0.51355, p-value = 0.6076
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.1543033

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = 0.5447, p-value = 0.586
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1380131

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Tannins"
this_var <- "Tannins"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 2.0542, p-value = 0.03996
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.6172134

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = 1.0894, p-value = 0.276
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.2760262 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other methylated\ncompounds"  
this_var <- "Other methylated\ncompounds"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -0.36314, p-value = 0.7165
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.09200874 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other hydrated\ncompounds"
this_var <- "Other hydrated\ncompounds"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 1.2839, p-value = 0.1992
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3857584

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -3.2682, p-value = 0.001082
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.8280787

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 1.0271, p-value = 0.3044
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3086067 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -0.5447, p-value = 0.586
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.1380131 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 0.25678, p-value = 0.7974
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.07715167

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = 1.6341, p-value = 0.1022
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
# tau 
# 0.4140393

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 2.0542, p-value = 0.03996
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
# tau 
# 0.6172134 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -2.1788, p-value = 0.02935
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.5520524 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "All compounds"
this_var <- "All compounds"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "value"]
# z = 2.0542, p-value = 0.03996
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.6172134

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "value"]
# z = -1.271, p-value = 0.2037
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.3220306 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



# all
annotation_df.all <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  npcx = c( pos_anotations[["Lipids"]][1],
            pos_anotations[["Proteins"]][1],
            pos_anotations[["Amino sugars"]][1],
            pos_anotations[["Carbohydrates"]][1],
            pos_anotations[["Condensed\naromatics"]][1],
            pos_anotations[["Lignin"]][1],
            pos_anotations[["Tannins"]][1],
            pos_anotations[["Other methylated\ncompounds"]][1],
            pos_anotations[["Other hydrated\ncompounds"]][1],
            pos_anotations[["Other demethylated\ncompounds"]][1],
            pos_anotations[["Other condensed\ncompounds"]][1],
            pos_anotations[["Near VitK2-VitB12\ncompounds"]][1],
            pos_anotations[["All compounds"]][1]
  ),
  npcy = c( pos_anotations[["Lipids"]][2],
            pos_anotations[["Proteins"]][2],
            pos_anotations[["Amino sugars"]][2],
            pos_anotations[["Carbohydrates"]][2],
            pos_anotations[["Condensed\naromatics"]][2],
            pos_anotations[["Lignin"]][2],
            pos_anotations[["Tannins"]][2],
            pos_anotations[["Other methylated\ncompounds"]][2],
            pos_anotations[["Other hydrated\ncompounds"]][2],
            pos_anotations[["Other demethylated\ncompounds"]][2],
            pos_anotations[["Other condensed\ncompounds"]][2],
            pos_anotations[["Near VitK2-VitB12\ncompounds"]][2],
            pos_anotations[["All compounds"]][2]
  ),
  tau1 = c(Kendall_Tau[["Lipids"]][1],
          Kendall_Tau[["Proteins"]][1],
          Kendall_Tau[["Amino sugars"]][1],
          Kendall_Tau[["Carbohydrates"]][1],
          Kendall_Tau[["Condensed\naromatics"]][1],
          Kendall_Tau[["Lignin"]][1],
          Kendall_Tau[["Tannins"]][1],
          Kendall_Tau[["Other methylated\ncompounds"]][1],
          Kendall_Tau[["Other hydrated\ncompounds"]][1],
          Kendall_Tau[["Other demethylated\ncompounds"]][1],
          Kendall_Tau[["Other condensed\ncompounds"]][1],
          Kendall_Tau[["Near VitK2-VitB12\ncompounds"]][1],
          Kendall_Tau[["All compounds"]][1]
  ),
  tau2 = c(Kendall_Tau[["Lipids"]][2],
          Kendall_Tau[["Proteins"]][2],
          Kendall_Tau[["Amino sugars"]][2],
          Kendall_Tau[["Carbohydrates"]][2],
          Kendall_Tau[["Condensed\naromatics"]][2],
          Kendall_Tau[["Lignin"]][2],
          Kendall_Tau[["Tannins"]][2],
          Kendall_Tau[["Other methylated\ncompounds"]][2],
          Kendall_Tau[["Other hydrated\ncompounds"]][2],
          Kendall_Tau[["Other demethylated\ncompounds"]][2],
          Kendall_Tau[["Other condensed\ncompounds"]][2],
          Kendall_Tau[["Near VitK2-VitB12\ncompounds"]][2],
          Kendall_Tau[["All compounds"]][2]
  ),
  siglabel1 = c(sig_anotations[["Lipids"]][1],
               sig_anotations[["Proteins"]][1],
               sig_anotations[["Amino sugars"]][1],
               sig_anotations[["Carbohydrates"]][1],
               sig_anotations[["Condensed\naromatics"]][1],
               sig_anotations[["Lignin"]][1],
               sig_anotations[["Tannins"]][1],
               sig_anotations[["Other methylated\ncompounds"]][1],
               sig_anotations[["Other hydrated\ncompounds"]][1],
               sig_anotations[["Other demethylated\ncompounds"]][1],
               sig_anotations[["Other condensed\ncompounds"]][1],
               sig_anotations[["Near VitK2-VitB12\ncompounds"]][1],
               sig_anotations[["All compounds"]][1]
  ),
  siglabel2 = c(sig_anotations[["Lipids"]][2],
                  sig_anotations[["Proteins"]][2],
                  sig_anotations[["Amino sugars"]][2],
                  sig_anotations[["Carbohydrates"]][2],
                  sig_anotations[["Condensed\naromatics"]][2],
                  sig_anotations[["Lignin"]][2],
                  sig_anotations[["Tannins"]][2],
                  sig_anotations[["Other methylated\ncompounds"]][2],
                  sig_anotations[["Other hydrated\ncompounds"]][2],
                  sig_anotations[["Other demethylated\ncompounds"]][2],
                  sig_anotations[["Other condensed\ncompounds"]][2],
                  sig_anotations[["Near VitK2-VitB12\ncompounds"]][2],
                  sig_anotations[["All compounds"]][2]
  ), stringsAsFactors = FALSE)
str(annotation_df.all)

# ensure no conflict in factor levels
annotation_df.all$variable <- factor(annotation_df.all$variable, 
                                     levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                                "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                     ordered = TRUE)

annotation_df.all$siglabel1 # "ns" "ns" "ns" "ns" "ns" "ns" "*"  "ns" "ns" "ns" "ns" "*"  "*" 


str(dat)
# 'data.frame':	234 obs. of  8 variables:
# $ sample          : chr  "ChrMrs" "ChrRic" "DunIsp" "HamAva" ...
# $ group           : Ord.factor w/ 51 levels "1"<"2"<"3"<"4"<..: 14 41 11 14 40 23 51 13 31 10 ...
# $ pH              : num  5.55 5.51 4.3 4.64 3.43 5.26 3.45 4.69 3.81 4.17 ...
# $ pH_class        : chr  "mod" "mod" "low" "mod" ...
# $ variable        : Ord.factor w/ 13 levels "Lipids"<"Proteins"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value           : num  0.237 0.309 0.295 0.302 0.389 ...
# $ age_vec_plot    : Ord.factor w/ 12 levels "10"<"11"<"13"<..: 4 10 2 4 9 5 12 3 7 1 ...
# $ age_vec_plot_num: int  14 41 11 14 40 23 58 13 31 10 ...


annotation_df.all$displaylabel <- paste0("Kendall's tau\n",
                                         "1) ",annotation_df.all$tau1, annotation_df.all$siglabel1,
                                         "; 2) ",annotation_df.all$tau2, annotation_df.all$siglabel2)

annotation_df.all$displaylabel <- gsub(pattern = "ns", replacement = "", x = annotation_df.all$displaylabel)


p <- ggplot(data = dat, aes(x = age_vec_plot_num, y = value))+
  geom_point(aes(color=pH_class))+
  geom_smooth(aes(color=pH_class),method="lm")+
  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+
  facet_wrap(facets = vars(variable), scales = "free_y")



p <- ggplot(data = dat, aes(x = age_vec_plot_num, y = value))+ # y = value   y = win95abun
  geom_point(aes(color=pH_class), alpha=0.5)+
  geom_smooth(aes(color=pH_class),method="lm")+
  geom_text_npc(data = annotation_df.all, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 3, lineheight=0.875 )+ # size = 3
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)+ # scales = "free_y",
  theme_classic() +
  labs(x = "Age (years)", y = "Function % rel. abun. in vK zones") +  # "Function % rel. abun. in vK zones (Winsorized)"
  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+
  scale_color_discrete(labels=c("1) Strongly acidic", "2) Acidic to neutral"), name = "pH level" )+
  theme(
    legend.position= c(0.825,0.16),
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-Trendplots-REVEG-AGE-",this_study,header,"-v9d-Extra-pH-level.tiff"), width = 24, height = 18, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### PCaN - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25) -- Split into strongly acidic (n=8) and acidic to neutral (n=10) pH groups
#-------------------------

this_study <- "-PCaN-"
header <- "-vk-hac-buffers-"

phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")

df.out.distil <- readRDS(file = "df.out.distil_PCaN-pilot.RDS")

vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-PCaN-pilot.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 36390     4


ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok) # 291664 elements

head(vk.samp.noNAs.long[-sel.ok, ])
#        sample fxn_id         abun abun_type      OC_x      HC_y group   samplabel
# 131259 InvKBR  fxn_2 8.898894e-06   relabun 0.8571429 0.7142857    NA InvKBR (NA)
# 131260 InvKBR  fxn_3 4.801165e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131261 InvKBR  fxn_4 5.360524e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131262 InvKBR  fxn_5 2.152261e-03   relabun 0.6000000 1.8000000    NA InvKBR (NA)
# 131263 InvKBR  fxn_6 3.775674e-05   relabun 1.0000000 0.4285714    NA InvKBR (NA)
# 131264 InvKBR  fxn_8 2.990937e-03   relabun 0.6666667 2.3333333    NA InvKBR (NA)

## don't remove Ref (group = NA) yet

str(vk.samp.noNAs.long) 
# 'data.frame':	346351 obs. of  8 variables:
unique(vk.samp.noNAs.long$group) # 14 41 11 31 40 23 NA 13 10 48 28 33

class(vk.samp.noNAs.long$group) # "numeric"
#vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c("6","12","22","31","UM"), ordered = TRUE)
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	346351 obs. of  8 variables:
# $ sample   : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ fxn_id   : Factor w/ 18229 levels "fxn_10","fxn_100",..: 6176 10762 14006 14662 15473 16843 17592 1 588 1021 ...
# $ abun     : num  1.18e-06 1.25e-04 9.82e-05 2.39e-03 5.47e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : num  14 14 14 14 14 14 14 14 14 14 ...
# $ samplabel: chr  "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" ...


# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) #  278281      5

table(spdf$group, useNA = "ifany")
# 10    11    13    14    23    28    31    33    40    41    48  <NA> 
#   29017 30000 14846 45164 15131 14747 27858 14785 14123 14763 14131 43716

head(spdf)
# sample group      OC_x      HC_y         abun
# 2 ChrMrs    14 0.8571429 0.7142857 1.177039e-06
# 3 ChrMrs    14 0.9285714 0.8571429 1.254724e-04
# 4 ChrMrs    14 0.9285714 0.8571429 9.816505e-05
# 5 ChrMrs    14 0.6000000 1.8000000 2.389389e-03
# 6 ChrMrs    14 1.0000000 0.4285714 5.473231e-05
# 7 ChrMrs    14 0.6666667 2.3333333 3.088441e-03

temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 19

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 60.94   61.55   61.65   61.71   61.98   62.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"


## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2
# df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
# df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)

sum_relabun_in_vk_hac_buffers_env <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  df_out <- data.frame(sample=NA, group=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      group = unique(spdf_in$group),
      #sex = unique(spdf_in$sex),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
    )
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers_env( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 19
out[[1]]
# sample group  rad  Acetate Propionate Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2 ChrMrs    14 0.05 21.11034  23.558398 0.000000 41.69772 16.845892 145.13537  1.488226 0.6255147  24.71484 66.57648
# 3 ChrMrs    14 0.10 17.13976   9.287203 5.616337 15.57725  7.388892 124.18431  3.645856 1.0791590  46.60457 58.16535
# 4 ChrMrs    14 0.15 13.84277   6.808156 6.469835 14.42792  8.761338  88.37149 11.752562 0.4913157  43.50636 63.96587
# 5 ChrMrs    14 0.20 12.08568  10.657949 7.700054 20.51997 19.848393  80.45936 12.839707 3.5579941  35.56173 81.24206
# 6 ChrMrs    14 0.25 11.92582  12.485889 7.989049 24.58022 18.764374  84.61773 10.720696 4.0653868  34.09802 79.48079


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample group  rad  Acetate Propionate Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2  ChrMrs    14 0.05 21.11034  23.558398 0.000000 41.69772 16.845892 145.13537  1.488226 0.6255147  24.71484 66.57648
# 3  ChrMrs    14 0.10 17.13976   9.287203 5.616337 15.57725  7.388892 124.18431  3.645856 1.0791590  46.60457 58.16535
# 4  ChrMrs    14 0.15 13.84277   6.808156 6.469835 14.42792  8.761338  88.37149 11.752562 0.4913157  43.50636 63.96587
# 5  ChrMrs    14 0.20 12.08568  10.657949 7.700054 20.51997 19.848393  80.45936 12.839707 3.5579941  35.56173 81.24206
# 6  ChrMrs    14 0.25 11.92582  12.485889 7.989049 24.58022 18.764374  84.61773 10.720696 4.0653868  34.09802 79.48079
# 21 ChrRic    41 0.05 17.76957  23.778397 0.000000 39.46803 16.573247 144.73210  1.623608 0.7008878  20.83100 64.57633

dim(df.buffers) # 95 13

table(df.buffers$group[ which(df.buffers$rad==0.05) ], useNA= "ifany") # count at single radius only
# 10   11   13   14   23   28   31   33   40   41   48 <NA> 
#   2    2    1    3    1    1    2    1    1    1    1    3  


# set Ref group
sel <- which(is.na(df.buffers$group)==TRUE)
df.buffers$age_vec_plot <- as.character( df.buffers$group )
df.buffers$age_vec_plot[sel] <- "Rem"


# add pH data

df.buffers$pH <- NA

for (i in 1:dim(df.buffers)[1]) {
  #i<-1
  this_samp <- df.buffers$sample[i]
  sel.phy <- which(phy@sam_data$sample == this_samp)
  df.buffers$pH[i] <- phy@sam_data$pH[sel.phy]
  print(paste0("completed ",i))
}

df.buffers$pH_class <- NA
sel <- which(df.buffers$pH <4.5) # <4.5 qty 40
df.buffers$pH_class[sel] <- "low"
sel <- which(df.buffers$pH >=4.5) # >=4.5 qty 50
df.buffers$pH_class[sel] <- "mod"

ok <- complete.cases(df.buffers[ ,"pH"])
sel.ok <- which(ok==TRUE) # 90
df.buffers <- df.buffers[sel.ok, ]

##df.buffers$group <- factor(df.buffers$group, levels = c(as.character(1:50),"Rem"), ordered = TRUE )


saveRDS(object = df.buffers, file = "df.buffers--PCaN-pilot-Extra-pH-level.RDS")

str(df.buffers)
# 'data.frame':	90 obs. of  16 variables:
# $ sample      : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ group       : num  14 14 14 14 14 41 41 41 41 41 ...
# $ rad         : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate     : num  21.1 17.1 13.8 12.1 11.9 ...
# $ Propionate  : num  23.56 9.29 6.81 10.66 12.49 ...
# $ Butyrate    : num  0 5.62 6.47 7.7 7.99 ...
# $ VitB2       : num  41.7 15.6 14.4 20.5 24.6 ...
# $ VitB12      : num  16.85 7.39 8.76 19.85 18.76 ...
# $ VitB6       : num  145.1 124.2 88.4 80.5 84.6 ...
# $ VitB9       : num  1.49 3.65 11.75 12.84 10.72 ...
# $ VitK2       : num  0.626 1.079 0.491 3.558 4.065 ...
# $ Glutamate   : num  24.7 46.6 43.5 35.6 34.1 ...
# $ Pyruvate    : num  66.6 58.2 64 81.2 79.5 ...
# $ age_vec_plot: chr  "14" "14" "14" "14" ...
# $ pH          : num  5.55 5.55 5.55 5.55 5.55 5.51 5.51 5.51 5.51 5.51 ...
# $ pH_class    : chr  "mod" "mod" "mod" "mod" ...


df.buffers.melt <- melt(df.buffers, id.vars = c("sample","group","age_vec_plot","rad","pH","pH_class"))
head(df.buffers.melt)
#   sample group age_vec_plot  rad   pH pH_class variable    value
# 1 ChrMrs    14           14 0.05 5.55      mod  Acetate 21.11034
# 2 ChrMrs    14           14 0.10 5.55      mod  Acetate 17.13976
# 3 ChrMrs    14           14 0.15 5.55      mod  Acetate 13.84277
# 4 ChrMrs    14           14 0.20 5.55      mod  Acetate 12.08568
# 5 ChrMrs    14           14 0.25 5.55      mod  Acetate 11.92582
# 6 ChrRic    41           41 0.05 5.51      mod  Acetate 17.76957

unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	900 obs. of  8 variables:
#   $ sample      : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ group       : num  14 14 14 14 14 41 41 41 41 41 ...
# $ age_vec_plot: chr  "14" "14" "14" "14" ...
# $ rad         : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ pH          : num  5.55 5.55 5.55 5.55 5.55 5.51 5.51 5.51 5.51 5.51 ...
# $ pH_class    : chr  "mod" "mod" "mod" "mod" ...
# $ variable    : Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value       : num  21.1 17.1 13.8 12.1 11.9 ...


saveRDS(object = df.buffers.melt, file = "df.buffers.melt.vkBuffers-PCaN-pilot-Extra-pH-level.RDS")


hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))


## create named numeric vector for x-axis display in facet_grid


dat <- df.buffers.melt

unique(dat$group)
# 14 41 11 31 40 23 NA 13 10 48 28 33

unique(dat$age_vec_plot)
# "14"  "41"  "11"  "40"  "23"  "Rem" "13"  "31"  "10"  "48"  "28"  "33"


dat$age_vec_plot <- factor(dat$age_vec_plot,
                           levels = c(
                             "10",  "11",  "13",  "14",  "23",  "28",  "31" , "33",  "40",  "41" , "48" , "Rem"
                           ),
                           labels = c(
                             "10",  "11",  "13", "14",  "23",  "28",  "31" , "33" , "40", "41" , "48", "58"
                             
                           ), 
                           ordered = TRUE)

dat$age_vec_plot_num <- as.integer(as.character(dat$age_vec_plot))
sort(unique( dat$age_vec_plot_num))
#   10 11 13 14 23 28 31 33 40 41 48 58


age_vec_plot_num <- c(1:60)

x <- c(
  rep(NA, times=9),"10", # 1:10
  rep(NA, times=9),"20", # 11:20
  rep(NA, times=9),"30", # 21:30
  rep(NA, times=9),"40", # 31:40
  rep(NA, times=10),     # 41:50
  NA,NA,NA,NA,NA,NA,NA,"Rem",NA,NA  # 51:60
)

names(age_vec_plot_num) <- x

label_breaks <- which( is.na( names(age_vec_plot_num) ) == FALSE)
label_breaks
# 10 20 30 40 58

names( age_vec_plot_num[ label_breaks ] )
# "10"  "20"  "30"  "40"  "Rem"


dat$log10abun <- dat$value # then over-write later


vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}

# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.225  12.040  13.577  15.140  17.612  27.844 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.965   1.081   1.133   1.164   1.246   1.445 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 6.493   9.331  10.834  13.061  13.018  27.404 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.8124  0.9699  1.0348  1.0706  1.1145  1.4378 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   5.852   6.669   5.660   7.723   8.405 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.4066  0.7673  0.8241  0.7587  0.8878  0.9246 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 13.61   15.28   20.59   22.93   24.62   45.01 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.134   1.184   1.314   1.330   1.391   1.653 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 6.015   8.604  14.550  13.939  19.149  22.197 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7792  0.9347  1.1629  1.1062  1.2822  1.3463 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 77.18   84.23   86.74  104.29  124.06  151.46 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.887   1.925   1.938   2.005   2.094   2.180 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.140   3.646  10.659   8.052  11.964  13.144 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.05697 0.56183 1.02772 0.78776 1.07788 1.11874 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.2839  0.7021  1.0203  2.0941  3.9507  5.2008 
# [1] "log10 values:"
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# -0.546871 -0.153615  0.008731  0.156568  0.596673  0.716070 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 16.44   33.08   35.73   37.27   45.93   51.42 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.216   1.520   1.553   1.556   1.662   1.711 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54.96   62.86   66.22   69.79   79.91   85.23 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.740   1.798   1.821   1.840   1.903   1.931


head(dat)
# sample group age_vec_plot  rad   pH pH_class variable    value age_vec_plot_num log10abun
# 1 ChrMrs    14           14 0.05 5.55      mod  Acetate 21.11034               14  1.324495
# 2 ChrMrs    14           14 0.10 5.55      mod  Acetate 17.13976               14  1.234005
# 3 ChrMrs    14           14 0.15 5.55      mod  Acetate 13.84277               14  1.141223
# 4 ChrMrs    14           14 0.20 5.55      mod  Acetate 12.08568               14  1.082271
# 5 ChrMrs    14           14 0.25 5.55      mod  Acetate 11.92582               14  1.076488
# 6 ChrRic    41           41 0.05 5.51      mod  Acetate 17.76957               41  1.249677

str(dat)
# 'data.frame':	900 obs. of  10 variables:
# $ sample          : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ group           : num  14 14 14 14 14 41 41 41 41 41 ...
# $ age_vec_plot    : Ord.factor w/ 12 levels "10"<"11"<"13"<..: 4 4 4 4 4 10 10 10 10 10 ...
# $ rad             : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ pH              : num  5.55 5.55 5.55 5.55 5.55 5.51 5.51 5.51 5.51 5.51 ...
# $ pH_class        : chr  "mod" "mod" "mod" "mod" ...
# $ variable        : Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value           : num  21.1 17.1 13.8 12.1 11.9 ...
# $ age_vec_plot_num: int  14 14 14 14 14 41 41 41 41 41 ...
# $ log10abun       : num  1.32 1.23 1.14 1.08 1.08 ...


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free")
p


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  guides(colour = guide_colorbar(title = "Age (years)", barheight = 0.7))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    legend.position="bottom",
    #legend.title = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

this_study # "-PCaN-"
header     # "-vk-hac-buffers-"

saveRDS(object = dat, file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9d-Extra-pH-level.RDS"))

#dat <- readRDS(file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9d-Extra-pH-level.RDS"))


## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]


head(dat)
#    sample group age_vec_plot  rad   pH pH_class variable    value age_vec_plot_num log10abun
# 1  ChrMrs    14           14 0.05 5.55      mod  Acetate 21.11034               14  1.324495
# 6  ChrRic    41           41 0.05 5.51      mod  Acetate 17.76957               41  1.249677
# 11 DunIsp    11           11 0.05 4.30      low  Acetate 25.92074               11  1.413647
# 16 HamAva    14           14 0.05 4.64      mod  Acetate 21.30437               14  1.328469
# 21 HamMin    40           40 0.05 3.43      low  Acetate 27.11873               40  1.433269
# 26 InvEsW    23           23 0.05 5.26      mod  Acetate 20.37210               23  1.309036


p <- ggplot(data = dat, aes(x = age_vec_plot_num, y = log10abun))+ # y = value
  geom_point(aes(color=pH_class), alpha=0.5)+
  geom_smooth(aes(color=pH_class),method="lm")+
  labs(x = "Age (years)", y = "log10 fxn % rel. abun. in 0.05 radius buffer (per vK unit area)") +
  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+
  facet_wrap(facets = vars(variable), scales = "free")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-rough-trendlines",this_study,header,"-v9d-Extra-pH-level.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")



## perform sig tests for annotations !!
levels( dat$variable )
# "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"      "Glutamate"  "Pyruvate"  


names(dat)
# [1] "sample"           "group"            "age_vec_plot"     "rad"              "pH"               "pH_class"         "variable"        
# [8] "value"            "age_vec_plot_num" "log10abun"   


sig_symb
# function(x) {
#   out <- character(length = length(x))
#   for (i in seq_along(x)) {
#     if (is.na(x[i])) {
#       out[i] <- NA
#     } else if (x[i] < 0.001) { 
#       out[i] <- "***"
#     } else if (x[i] >= 0.001 & x[i] < 0.01) { 
#       out[i] <- "**"
#     } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
#       out[i] <- "*" 
#     } else { 
#       out[i] <- "ns"
#     }
#   } # END loop
#   return(out)
# }
# <bytecode: 0x7fe3b2960ef0>



pos_anotations <- list()
sig_anotations <- list()
Kendall_Tau <- list()



## "Acetate"
this_var <- "Acetate"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 0.77033, p-value = 0.4411
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.231455

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = -2.7235, p-value = 0.006459
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.6900656

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Propionate"
this_var <- "Propionate"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 0.25678, p-value = 0.7974
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.07715167

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = 1.271, p-value = 0.2037
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3220306 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Butyrate"
this_var <- "Butyrate"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# T = NA, p-value = NA
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# NA 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# T = NA, p-value = NA
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# NA 

pos_anotations[[this_var]] <- c("centre", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "VitB2"
this_var <- "VitB2"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = -1.7974, p-value = 0.07227
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.5400617 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = -0.90784, p-value = 0.364
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.2300219 

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "VitB12"
this_var <- "VitB12"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = -2.0542, p-value = 0.03996
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.6172134

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = 0.36314, p-value = 0.7165
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.09200874 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "VitB6"
this_var <- "VitB6"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 1.2839, p-value = 0.1992
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3857584

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = -1.4525, p-value = 0.1463
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.368035

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "VitB9"
this_var <- "VitB9"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 1.7974, p-value = 0.07227
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.5400617 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = 0.18157, p-value = 0.8559
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.04600437

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "VitK2" 
this_var <- "VitK2"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 1.7974, p-value = 0.07227
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.5400617

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = 0.36314, p-value = 0.7165
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.09200874

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Glutamate" 
this_var <- "Glutamate"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0 

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = -2.3604, p-value = 0.01826
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.5980568 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Pyruvate"  
this_var <- "Pyruvate"

sel1 <- which(dat$variable == this_var & dat$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat$age_vec_plot_num[sel1], y = dat[sel1 ,"log10abun"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel1] and dat[sel1, "log10abun"]
# z = 0.77033, p-value = 0.4411
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.231455

sel2 <- which(dat$variable == this_var & dat$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat$age_vec_plot_num[sel2], y = dat[sel2 ,"log10abun"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat$age_vec_plot_num[sel2] and dat[sel2, "log10abun"]
# z = -1.4525, p-value = 0.1463
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.368035

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



# Add significance annotation

# all
annotation_df.all <- data.frame(
  
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
  npcx = c( pos_anotations[["Acetate"]][1],
            pos_anotations[["Propionate"]][1],
            pos_anotations[["Butyrate"]][1],
            pos_anotations[["VitB2"]][1],
            pos_anotations[["VitB12"]][1],
            pos_anotations[["VitB6"]][1],
            pos_anotations[["VitB9"]][1],
            pos_anotations[["VitK2"]][1],
            pos_anotations[["Glutamate"]][1],
            pos_anotations[["Pyruvate"]][1]
  ),
  
  npcy = c( pos_anotations[["Acetate"]][2],
            pos_anotations[["Propionate"]][2],
            pos_anotations[["Butyrate"]][2],
            pos_anotations[["VitB2"]][2],
            pos_anotations[["VitB12"]][2],
            pos_anotations[["VitB6"]][2],
            pos_anotations[["VitB9"]][2],
            pos_anotations[["VitK2"]][2],
            pos_anotations[["Glutamate"]][2],
            pos_anotations[["Pyruvate"]][2]
  ),
  
  tau1 = c(Kendall_Tau[["Acetate"]][1],
          Kendall_Tau[["Propionate"]][1],
          Kendall_Tau[["Butyrate"]][1],
          Kendall_Tau[["VitB2"]][1],
          Kendall_Tau[["VitB12"]][1],
          Kendall_Tau[["VitB6"]][1],
          Kendall_Tau[["VitB9"]][1],
          Kendall_Tau[["VitK2"]][1],
          Kendall_Tau[["Glutamate"]][1],
          Kendall_Tau[["Pyruvate"]][1]
  ),
  tau2 = c(Kendall_Tau[["Acetate"]][2],
           Kendall_Tau[["Propionate"]][2],
           Kendall_Tau[["Butyrate"]][2],
           Kendall_Tau[["VitB2"]][2],
           Kendall_Tau[["VitB12"]][2],
           Kendall_Tau[["VitB6"]][2],
           Kendall_Tau[["VitB9"]][2],
           Kendall_Tau[["VitK2"]][2],
           Kendall_Tau[["Glutamate"]][2],
           Kendall_Tau[["Pyruvate"]][2]
  ),
  
  siglabel1 = c(sig_anotations[["Acetate"]][1],
               sig_anotations[["Propionate"]][1],
               sig_anotations[["Butyrate"]][1],
               sig_anotations[["VitB2"]][1],
               sig_anotations[["VitB12"]][1],
               sig_anotations[["VitB6"]][1],
               sig_anotations[["VitB9"]][1],
               sig_anotations[["VitK2"]][1],
               sig_anotations[["Glutamate"]][1],
               sig_anotations[["Pyruvate"]][1]
  ),
  siglabel2 = c(sig_anotations[["Acetate"]][2],
                sig_anotations[["Propionate"]][2],
                sig_anotations[["Butyrate"]][2],
                sig_anotations[["VitB2"]][2],
                sig_anotations[["VitB12"]][2],
                sig_anotations[["VitB6"]][2],
                sig_anotations[["VitB9"]][2],
                sig_anotations[["VitK2"]][2],
                sig_anotations[["Glutamate"]][2],
                sig_anotations[["Pyruvate"]][2]
  ),
  stringsAsFactors = FALSE)
str(annotation_df.all)

# ensure no conflict in factor levels
annotation_df.all$variable <- factor(annotation_df.all$variable, 
                                     levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
                                     ordered = TRUE)

annotation_df.all$siglabel1 # "ns" "ns" NA   "ns" "*"  "ns" "ns" "ns" "ns" "ns"

annotation_df.all$displaylabel <- paste0("Kendall's tau\n",
                                         "1) ",annotation_df.all$tau1, annotation_df.all$siglabel1,
                                         "; 2) ",annotation_df.all$tau2, annotation_df.all$siglabel2)

annotation_df.all$displaylabel <- gsub(pattern = "ns", replacement = "", x = annotation_df.all$displaylabel)

unique(annotation_df.all$displaylabel)
# [1] "Kendall's tau\n1) 0.231; 2) -0.69**" "Kendall's tau\n1) 0.077; 2) 0.322"   "Kendall's tau\n1) NANA; 2) NANA"    
# [4] "Kendall's tau\n1) -0.54; 2) -0.23"   "Kendall's tau\n1) -0.617*; 2) 0.092" "Kendall's tau\n1) 0.386; 2) -0.368" 
# [7] "Kendall's tau\n1) 0.54; 2) 0.046"    "Kendall's tau\n1) 0.54; 2) 0.092"    "Kendall's tau\n1) 0; 2) -0.598*"    
# [10] "Kendall's tau\n1) 0.231; 2) -0.368" 

annotation_df.all$displaylabel <- gsub(pattern = "NANA", replacement = "N/A", x = annotation_df.all$displaylabel)

str(annotation_df.all)
# 'data.frame':	10 obs. of  8 variables:
# $ variable    : Ord.factor w/ 10 levels "Acetate"<"Propionate"<..: 1 2 3 4 5 6 7 8 9 10
# $ npcx        : chr  "left" "right" "centre" "right" ...
# $ npcy        : chr  "bottom" "bottom" "bottom" "top" ...
# $ tau1        : num  0.231 0.077 NA -0.54 -0.617 0.386 0.54 0.54 0 0.231
# $ tau2        : num  -0.69 0.322 NA -0.23 0.092 -0.368 0.046 0.092 -0.598 -0.368
# $ siglabel1   : chr  "ns" "ns" NA "ns" ...
# $ siglabel2   : chr  "**" "ns" NA "ns" ...
# $ displaylabel: chr  "Kendall's tau\nLow pH: 0.231\nMod pH: -0.69**" "Kendall's tau\nLow pH: 0.077\nMod pH: 0.322" "Kendall's tau\nLow pH: N/A\nMod pH: N/A" "Kendall's tau\nLow pH: -0.54\nMod pH: -0.23" ...

str(dat)
# 'data.frame':	180 obs. of  10 variables:
# $ sample          : chr  "ChrMrs" "ChrRic" "DunIsp" "HamAva" ...
# $ group           : num  14 41 11 14 40 23 NA 13 31 10 ...
# $ age_vec_plot    : Ord.factor w/ 12 levels "10"<"11"<"13"<..: 4 10 2 4 9 5 12 3 7 1 ...
# $ rad             : num  0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 ...
# $ pH              : num  5.55 5.51 4.3 4.64 3.43 5.26 3.45 4.69 3.81 4.17 ...
# $ pH_class        : chr  "mod" "mod" "low" "mod" ...
# $ variable        : Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value           : num  21.1 17.8 25.9 21.3 27.1 ...
# $ age_vec_plot_num: int  14 41 11 14 40 23 58 13 31 10 ...
# $ log10abun       : num  1.32 1.25 1.41 1.33 1.43 ...



p <- ggplot(data = dat, aes(x = age_vec_plot_num, y = log10abun))+ # y = value
  geom_point(aes(color=pH_class), alpha=0.5)+
  geom_smooth(aes(color=pH_class),method="lm")+
  geom_text_npc(data = annotation_df.all, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 3, lineheight=0.875 )+ # size = 3
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = "Age (years)", y = "Function log10 % rel. abun. in 0.05 radius vK buffer zones") +
  
  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+
  
  scale_color_discrete(labels=c("1) Strongly acidic", "2) Acidic to neutral"), name = "pH level" )+
  theme(
    legend.position="bottom",
    legend.margin=margin(0,0,0,0),
    legend.box.margin=margin(0,-5,0,0), 
    strip.background = element_rect(fill="white", linetype = "blank"),
    #strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.9))
    )
p

this_study # "-PCaN-"
header     <- "-vk-hac-buffers-"

dev.print(tiff, file = paste0(workdir,"/plots/","All-Env-vkBuffers-log1oabun-boxplots-",this_study,header,"-v9d-Extra-pH-level.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### PCaN - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

# Determines weighted coordinates in each sample, then later assess stats by pH class

this_study <- "-PCaN-"
header <- "-vk-coords-wtmean-"

phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")

df.out.distil <- readRDS(file = "df.out.distil_PCaN-pilot.RDS")

vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-PCaN-pilot.RDS")

# use clean data
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 36390     4


ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok) # 291664 elements

head(vk.samp.noNAs.long[-sel.ok, ])
#        sample fxn_id         abun abun_type      OC_x      HC_y group   samplabel
# 131259 InvKBR  fxn_2 8.898894e-06   relabun 0.8571429 0.7142857    NA InvKBR (NA)
# 131260 InvKBR  fxn_3 4.801165e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131261 InvKBR  fxn_4 5.360524e-05   relabun 0.9285714 0.8571429    NA InvKBR (NA)
# 131262 InvKBR  fxn_5 2.152261e-03   relabun 0.6000000 1.8000000    NA InvKBR (NA)
# 131263 InvKBR  fxn_6 3.775674e-05   relabun 1.0000000 0.4285714    NA InvKBR (NA)
# 131264 InvKBR  fxn_8 2.990937e-03   relabun 0.6666667 2.3333333    NA InvKBR (NA)


## don't remove Ref (group = NA) yet


str(vk.samp.noNAs.long) 
# 'data.frame':	346351 obs. of  8 variables:
unique(vk.samp.noNAs.long$group) # 14 41 11 31 40 23 NA 13 10 48 28 33

class(vk.samp.noNAs.long$group) # "numeric"

vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	346351 obs. of  8 variables:
# $ sample   : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ fxn_id   : Factor w/ 18229 levels "fxn_10","fxn_100",..: 6176 10762 14006 14662 15473 16843 17592 1 588 1021 ...
# $ abun     : num  1.18e-06 1.25e-04 9.82e-05 2.39e-03 5.47e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : num  14 14 14 14 14 14 14 14 14 14 ...
# $ samplabel: chr  "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" "ChrMrs (14)" ...



# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) #  278281      5

table(spdf$group, useNA = "ifany")
# 10    11    13    14    23    28    31    33    40    41    48  <NA> 
#   29017 30000 14846 45164 15131 14747 27858 14785 14123 14763 14131 43716

head(spdf)
# sample group      OC_x      HC_y         abun
# 2 ChrMrs    14 0.8571429 0.7142857 1.177039e-06
# 3 ChrMrs    14 0.9285714 0.8571429 1.254724e-04
# 4 ChrMrs    14 0.9285714 0.8571429 9.816505e-05
# 5 ChrMrs    14 0.6000000 1.8000000 2.389389e-03
# 6 ChrMrs    14 1.0000000 0.4285714 5.473231e-05
# 7 ChrMrs    14 0.6666667 2.3333333 3.088441e-03


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 19

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 60.94   61.55   61.65   61.71   61.98   62.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"


get_wtmean_vk_coords <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  subsel <- list()
  
  # lipids
  subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  # proteins
  subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # amino_sugars
  subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # carbs - Carbohydrates
  subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  # cond_aromtx - Condensed aromatics
  subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  # lignin
  subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  # tannins
  subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  
  vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")
  
  # establish blank template data frame
  df_out <- data.frame(sample = NA, group = NA, no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)
  
  for (v in 1:length(vars)) {
    #v<-1
    this_var <- vars[v]
    if (length(subsel[[ this_var ]]) > 0) {
      df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
    } else {
      df_out$wtmean_OC_x[v] <- NA
      df_out$wtmean_HC_y[v] <- NA
      
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
      
    }
    print(paste0("completed ",v))
  }
  
  df_out$sample <- this_samp
  df_out$group <- unique(spdf_in$group)
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 19
out[[1]]
#   sample group no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1 ChrMrs    14     142       lipids  0.09344992   1.6482889
# 2 ChrMrs    14     749     proteins  0.38449788   1.7139833
# 3 ChrMrs    14     807 amino_sugars  0.62743492   1.6724291
# 4 ChrMrs    14    2185        carbs  0.87741899   1.7138420
# 5 ChrMrs    14     193  cond_aromtx  0.14218921   0.9302769
# 6 ChrMrs    14    2610       lignin  0.49641001   1.2694264
# 7 ChrMrs    14    3896      tannins  0.80276129   1.3206590

names(out[[1]])
# "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 133 6



# set Ref group
sel <- which(is.na(df.wtcoords$group)==TRUE)
df.wtcoords[sel, ]
df.wtcoords$age_vec_plot <- as.character( df.wtcoords$group )
df.wtcoords$age_vec_plot[sel] <- "Rem"


# add pH data

df.wtcoords$pH <- NA

for (i in 1:dim(df.wtcoords)[1]) {
  #i<-1
  this_samp <- df.wtcoords$sample[i]
  sel.phy <- which(phy@sam_data$sample == this_samp)
  df.wtcoords$pH[i] <- phy@sam_data$pH[sel.phy]
  print(paste0("completed ",i))
}

df.wtcoords$pH_class <- NA
sel <- which(df.wtcoords$pH <4.5) # <4.5 qty 56 (7*8)
df.wtcoords$pH_class[sel] <- "low"
sel <- which(df.wtcoords$pH >=4.5) # >=4.5 qty 70 (7*10)
df.wtcoords$pH_class[sel] <- "mod"

#ok <- complete.cases(df.wtcoords[ ,"pH"])
# any cases with missing coordinate data?
ok <- complete.cases(df.wtcoords[ ,"wtmean_OC_x"])
sel.ok <- which(ok==TRUE) # 133
df.wtcoords <- df.wtcoords[sel.ok, ]

df.wtcoords$age_vec_plot <- factor(df.wtcoords$age_vec_plot, levels = c(as.character(1:50),"Rem"), ordered = TRUE )
sort( unique(df.wtcoords$age_vec_plot)) # 
# [1] 10  11  13  14  23  28  31  33  40  41  48  Rem

df.wtcoords$group_rank <- factor(df.wtcoords$age_vec_plot, levels = c("10","11","13","14","23","28","31","33","40","41","48","Rem"), ordered = TRUE)
df.wtcoords$group_rank <- as.numeric(df.wtcoords$group_rank)

saveRDS(object = df.wtcoords, file = "df.wtcoords--PCaN-pilot-Extra-pH-level.RDS")



str(df.wtcoords)
# 'data.frame':	133 obs. of  10 variables:
# $ sample      : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ group       : num  14 14 14 14 14 14 14 41 41 41 ...
# $ no_fxns     : int  142 749 807 2185 193 2610 3896 150 687 782 ...
# $ variable    : chr  "lipids" "proteins" "amino_sugars" "carbs" ...
# $ wtmean_OC_x : num  0.0934 0.3845 0.6274 0.8774 0.1422 ...
# $ wtmean_HC_y : num  1.65 1.71 1.67 1.71 0.93 ...
# $ age_vec_plot: Ord.factor w/ 51 levels "1"<"2"<"3"<"4"<..: 14 14 14 14 14 14 14 41 41 41 ...
# $ pH          : num  5.55 5.55 5.55 5.55 5.55 5.55 5.55 5.51 5.51 5.51 ...
# $ pH_class    : chr  "mod" "mod" "mod" "mod" ...
# $ group_rank  : num  4 4 4 4 4 4 4 10 10 10 ...



dat <- df.wtcoords

unique(dat$variable)
# "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"


dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)

names(dat)
# [1] "sample"       "group"        "no_fxns"      "variable"     "wtmean_OC_x"  "wtmean_HC_y"  "age_vec_plot" "pH"           "pH_class"    
# [10] "group_ord


# Strongly acidic

p <- ggplot(data = filter(dat, pH_class == "low"), aes(x = wtmean_OC_x, y = wtmean_HC_y, color = age_vec_plot) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  #geom_density_2d(contour_var = "ndensity", bins = 3, alpha=.5) + # , alpha=.5
  
  geom_point()+ # alpha=0.5
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 2,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Reveg age\n(strongly acidic)"))

p


this_study # "-PCaN-"
header # "-vk-coords-wtmean-"

grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9e-a.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


# Acidic to Neutral

p <- ggplot(data = filter(dat, pH_class == "mod"), aes(x = wtmean_OC_x, y = wtmean_HC_y, color = age_vec_plot) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  #geom_density_2d(contour_var = "ndensity", bins = 3, alpha=.5) + # , alpha=.5
  
  geom_point()+ # alpha=0.5
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    
    legend.key.height = unit(0.85, 'line'),
    
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 2,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Reveg age\n(Acidic to neutral)"))

p


this_study # "-PCaN-"
header # "-vk-coords-wtmean-"

grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9e-b.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



saveRDS(object = dat, file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9e.RDS"))

#dat <- readRDS(file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9e.RDS"))


str(dat)
# 'data.frame':	133 obs. of  10 variables:
# $ sample      : chr  "ChrMrs" "ChrMrs" "ChrMrs" "ChrMrs" ...
# $ group       : num  14 14 14 14 14 14 14 41 41 41 ...
# $ no_fxns     : int  142 749 807 2185 193 2610 3896 150 687 782 ...
# $ variable    : Ord.factor w/ 7 levels "Lipids"<"Proteins"<..: 1 2 3 4 5 6 7 1 2 3 ...
# $ wtmean_OC_x : num  0.0934 0.3845 0.6274 0.8774 0.1422 ...
# $ wtmean_HC_y : num  1.65 1.71 1.67 1.71 0.93 ...
# $ age_vec_plot: Ord.factor w/ 51 levels "1"<"2"<"3"<"4"<..: 14 14 14 14 14 14 14 41 41 41 ...
# $ pH          : num  5.55 5.55 5.55 5.55 5.55 5.55 5.55 5.51 5.51 5.51 ...
# $ pH_class    : chr  "mod" "mod" "mod" "mod" ...
# $ group_rank  : num  4 4 4 4 4 4 4 10 10 10 ...



table(dat$pH_class)/length(unique(dat$variable))
# low mod 
# 8  10

# Females

dat.low <- filter(dat, pH_class == "low")



## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.low) # "sample"      "behav_PC1"   "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group"       "group_mf"   


levels(dat.low$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00000213 0.00528 0.0319  0.912
# Residual     6 0.00040031 0.99472              
# Total        7 0.00040243 1.00000 

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 5.4630e-06 0.06231 0.3987  0.529
# Residual     6 8.2221e-05 0.93769              
# Total        7 8.7685e-05 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2     F Pr(>F)
# samps.class  1 1.6090e-06 0.01105 0.067  0.813
# Residual     6 1.4402e-04 0.98895             
# Total        7 1.4563e-04 1.00000 

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 1.5297e-05 0.08529 0.5594  0.628
# Residual     6 1.6406e-04 0.91471              
# Total        7 1.7936e-04 1.00000 

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0000029 0.00285 0.0172  0.919
# Residual     6 0.0010127 0.99715              
# Total        7 0.0010156 1.00000 


# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 7.2812e-06 0.30295 2.6077  0.139
# Residual     6 1.6753e-05 0.69705              
# Total        7 2.4035e-05 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta)) 



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.low$variable == this_var)
x <- dat.low[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.low$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 6.8770e-06 0.19392 1.4434  0.288
# Residual     6 2.8586e-05 0.80608              
# Total        7 3.5463e-05 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



# # # # # # # # # # # # #


# Acidic to neutral soils

dat.mod <- filter(dat, pH_class == "mod")


## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.mod) 
# [1] "sample"       "group"        "no_fxns"      "variable"     "wtmean_OC_x"  "wtmean_HC_y"  "age_vec_plot" "pH"           "pH_class"    
# [10] "group_rank"

levels(dat.mod$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]


set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00002063 0.03943 0.3284  0.653
# Residual     8 0.00050260 0.96057              
# Total        9 0.00052323 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 6.0290e-06 0.12152 1.1066  0.344
# Residual     8 4.3589e-05 0.87848              
# Total        9 4.9619e-05 1.00000 

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2     F Pr(>F)
# samps.class  1 1.8130e-06 0.01648 0.134  0.733
# Residual     8 1.0818e-04 0.98352             
# Total        9 1.0999e-04 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)   
# samps.class  1 0.00016041 0.56915 10.568  0.008 **
# Residual     8 0.00012143 0.43085                 
# Total        9 0.00028184 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)  
# samps.class  1 0.00040578 0.54146 9.4465  0.019 *
# Residual     8 0.00034364 0.45854                
# Total        9 0.00074942 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 3.4620e-06 0.08567 0.7495  0.407
# Residual     8 3.6954e-05 0.91433              
# Total        9 4.0416e-05 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.mod$variable == this_var)
x <- dat.mod[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.mod$group_rank[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 9.4540e-06 0.23775 2.4952  0.117
# Residual     8 3.0312e-05 0.76225              
# Total        9 3.9767e-05 1.00000

# beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
# set.seed(123)
# (betadisp <- permutest(beta))

#-------------------------

#### PCaN - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-PCaN-pilot.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 36390 taxa and 19 samples ]
# sample_data() Sample Data:       [ 19 samples by 32 sample variables ]
# tax_table()   Taxonomy Table:    [ 36390 taxa by 4 taxonomic ranks ]
df.zones <- readRDS("df.zones--PCaN-pilot-Extra-pH-level.RDS")
vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-PCaN-pilot.RDS")

samps <- unique(df.zones$sample) # qty 18
phy <- prune_samples( samples = samps, phy )
# prune taxa with zero reads
phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 36324 taxa and 18 samples ]
# sample_data() Sample Data:       [ 18 samples by 32 sample variables ]
# tax_table()   Taxonomy Table:    [ 36324 taxa by 4 taxonomic ranks ]
head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 18
NonZeroFxns
# ChrMrs ChrRic DunIsp HamAva HamMin InvEsW InvKBR InvWaR NelBot NelMRY NplHrs NplPer WelMtA WelOCH WelOWR HamCla TauCaP TauMcB 
# 29513  28127  27976  28903  26484  29035  26747  28284  26167  27465  26661  28518  28049  27579  27850  28088  29823  28165
mean(NonZeroFxns) # 27968.56
sd(NonZeroFxns) # 1011.447
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 18197
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 3302
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 14689.67
sd(fxns_converted) # 396.2955

mean(fxns_relabun_converted) # 61.74909
sd(fxns_relabun_converted) # 0.3299077

mean(vks) # 2964.556
sd(vks) # 48.60915


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # AMI data - natural vs disturbed soils; Surface 0-10cm, temperate climate, 7.5 - 45% clay content
# # # # # # # # #
# # # # # # # # #

#### AMI - Metadata
#-------------------------

#### read in site info / metadata

# all samples from temperate climate zone, clay content 7.5 to 45%
meta <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/AMI-soil-metg/contextual-temperate-loams-7_5-45-claycontent-copy.xlsx",
                   sheet=1, range="A1:CK105", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)
# 'data.frame':	104 obs. of  89 variables:
# $ Sample ID                          : chr  "102.100.100/12465" "102.100.100/12469" "102.100.100/12471" "102.100.100/12473" ...
# $ Am Environment                     : chr  "Soil" "Soil" "Soil" "Soil" ...
# $ Ammonium Nitrogen Wt [mg/kg]       : num  22 7 13 8 16 9 6 6 5 8 ...
# $ Ammonium Nitrogen Wt Meth          : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Biotic Relationship                : chr  "Free-living" "Free-living" "Free-living" "Free-living" ...
# $ Boron Hot Cacl2 [mg/kg]            : num  1.35 0.49 0.33 0.25 0.47 0.49 0.33 0.53 0.45 0.33 ...
# $ Boron Hot Cacl2 Meth               : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Clay [%]                           : num  12.7 16.9 13.5 13.1 16.1 ...
# $ Clay Meth                          : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Coarse Sand [%]                    : num  46.1 50.1 43.1 48.8 53.2 ...
# $ Coarse Sand Meth                   : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Collection Date                    : POSIXct, format: "2013-08-14" "2013-08-14" "2013-08-14" "2013-08-14" ...
# $ Color                              : chr  "BR" "BR" "BRYW" "GRYW" ...
# $ Conductivity [dS/m]                : num  0.064 0.081 0.05 0.08 0.101 0.075 0.058 0.054 0.036 0.101 ...
# $ Conductivity Meth                  : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Database Schema Definitions Url    : chr  "https://github.com/AusMicrobiome/contextualdb_doc/raw/3.2.0/db_schema_definitions/db_schema_definitions.xlsx" "https://github.com/AusMicrobiome/contextualdb_doc/raw/3.2.0/db_schema_definitions/db_schema_definitions.xlsx" "https://github.com/AusMicrobiome/contextualdb_doc/raw/3.2.0/db_schema_definitions/db_schema_definitions.xlsx" "https://github.com/AusMicrobiome/contextualdb_doc/raw/3.2.0/db_schema_definitions/db_schema_definitions.xlsx" ...
# $ Depth [m]                          : num  0 0 0 0 0 0 0 0 0 0 ...
# $ Depth Lower [m]                    : num  0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 ...
# $ Depth Upper [m]                    : num  0 0 0 0 0 0 0 0 0 0 ...
# $ Dtpa Copper [mg/kg]                : num  0.89 0.62 0.73 0.91 0.62 0.85 0.65 1.79 0.42 0.72 ...
# $ Dtpa Copper Meth                   : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Dtpa Iron [mg/kg]                  : num  133.1 99.5 132.8 129.5 98.7 ...
# $ Dtpa Iron Meth                     : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Dtpa Manganese [mg/kg]             : num  167.9 100.1 110 88.2 191.3 ...
# $ Dtpa Manganese Meth                : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Dtpa Zinc [mg/kg]                  : num  8.88 4.24 3.37 5.79 4.57 ...
# $ Dtpa Zinc Meth                     : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Elev [m]                           : num  1262 1278 1267 1300 1292 ...
# $ Env Broad Scale                    : chr  "1 Conservation and Natural Environments" "1 Conservation and Natural Environments" "1 Conservation and Natural Environments" "1 Conservation and Natural Environments" ...
# $ Env Local Scale                    : chr  "1.1.3 National park" "1.1.3 National park" "1.1.3 National park" "1.1.3 National park" ...
# $ Env Medium                         : chr  "Soil [ENVO_00001998]" "Soil [ENVO_00001998]" "Soil [ENVO_00001998]" "Soil [ENVO_00001998]" ...
# $ Exc Aluminium [meq/100g]           : num  0.2 0.115 0.275 1.019 0.364 ...
# $ Exc Aluminium Meth                 : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Exc Calcium [meq/100g]             : num  18.6 12.4 10.7 4.2 11.2 ...
# $ Exc Calcium Meth                   : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Exc Magnesium [meq/100g]           : num  2.4 1.63 1.48 0.83 1.48 ...
# $ Exc Magnesium Meth                 : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Exc Potassium [meq/100g]           : num  0.74 0.92 0.44 0.42 0.62 0.6 0.33 0.49 0.45 0.66 ...
# $ Exc Potassium Meth                 : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Exc Sodium [meq/100g]              : num  0.05 0.05 0.04 0.04 0.05 0.03 0.04 0.04 0.02 0.55 ...
# $ Exc Sodium Meth                    : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Fine Sand [%]                      : num  21.1 11.7 18.1 18.9 14.5 ...
# $ Fine Sand Meth                     : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ General Env Feature                : chr  "Temperate" "Temperate" "Temperate" "Temperate" ...
# $ Geo Loc Name                       : chr  "Australia" "Australia" "Australia" "Australia" ...
# $ Gravel [%]                         : num  0 0 0 0 0 0 0 0 5 5 ...
# $ Gravel Meth                        : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Hyperspectral Analysis             : chr  "DOI 10.5281/zenodo.6265730" "DOI 10.5281/zenodo.6265730" "DOI 10.5281/zenodo.6265730" "DOI 10.5281/zenodo.6265730" ...
# $ Hyperspectral Analysis Meth        : num  2.26 2.26 2.26 2.26 2.26 2.26 2.26 2.26 2.26 2.26 ...
# $ Lat Lon [decimal_degrees]          : chr  "-35.4214 148.79855" "-35.4235 148.7968" "-35.42386667 148.7971667" "-35.42325 148.7967667" ...
# $ Latitude [decimal_degrees]         : num  -35.4 -35.4 -35.4 -35.4 -35.4 ...
# $ Local Class                        : chr  "Tenosols" "Tenosols" "Tenosols" "Tenosols" ...
# $ Longitude [decimal_degrees]        : num  149 149 149 149 149 ...
# $ Nitrate Nitrogen [mg/kg]           : num  1e-10 1e-10 1e+00 1e-10 1e-10 ...
# $ Nitrate Nitrogen Meth              : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Nucl Acid Ext                      : chr  "3.1.1" "3.1.1" "3.1.1" "3.1.1" ...
# $ Organic Carbon [%]                 : num  4.82 3.75 4.38 3.19 3.42 3.4 3.16 4.07 2.62 5.35 ...
# $ Organic Carbon Meth                : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Ph                                 : num  4.7 5.2 4.8 4.5 4.7 4.7 4.5 5.2 5.3 5.1 ...
# $ Ph Meth                            : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Ph Solid H2o                       : num  5.7 6.1 5.7 5.5 5.5 5.6 5.4 5.8 5.8 5.7 ...
# $ Ph Solid H2o Meth                  : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Phosphorus Colwell [mg/kg]         : num  68 51 70 22 34 13 68 24 26 26 ...
# $ Phosphorus Colwell Meth            : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Potassium Colwell [mg/kg]          : num  320 361 202 182 289 249 139 230 220 258 ...
# $ Potassium Colwell Meth             : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Sample Database File               : chr  "AM_db_v3.2_202203021355.db" "AM_db_v3.2_202203021355.db" "AM_db_v3.2_202203021355.db" "AM_db_v3.2_202203021355.db" ...
# $ Sample Metadata Ingest Date        : POSIXct, format: "2021-04-23" "2021-04-23" "2021-04-23" "2021-04-23" ...
# $ Sample Metadata Ingest File        : chr  "BASE_NEW_CKAN_Curated_depth_xlsx" "BASE_NEW_CKAN_Curated_depth_xlsx" "BASE_NEW_CKAN_Curated_depth_xlsx" "BASE_NEW_CKAN_Curated_depth_xlsx" ...
# $ Sample Metadata Update History     : chr  "All_soil_noAntarctic_Methods_FILLED_UPDATE_xlsx; AM_visNIR_UPDATE_xlsx;" "All_soil_noAntarctic_Methods_FILLED_UPDATE_xlsx; AM_visNIR_UPDATE_xlsx;" "All_soil_noAntarctic_Methods_FILLED_UPDATE_xlsx; AM_visNIR_UPDATE_xlsx;" "All_soil_noAntarctic_Methods_FILLED_UPDATE_xlsx; AM_visNIR_UPDATE_xlsx;" ...
# $ Sample Site Location Description   : chr  "Namadgi NP" "Namadgi NP" "Namadgi NP" "Namadgi NP" ...
# $ Sample Type                        : chr  "Soil" "Soil" "Soil" "Soil" ...
# $ Sand [%]                           : num  67.2 61.9 61.2 67.7 67.6 ...
# $ Sand Meth                          : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Silt [%]                           : num  20.1 21.2 25.3 19.2 16.2 ...
# $ Silt Meth                          : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Slope Aspect [direction_or_degrees]: num  NA NA NA NA NA NA NA NA NA NA ...
# $ Slope Gradient [%]                 : num  0 2 5 2 5 7 5 0 15 0 ...
# $ Source Mat                         : chr  "102.100.100/12465" "102.100.100/12469" "102.100.100/12471" "102.100.100/12473" ...
# $ Store Cond                         : chr  "Frozen" "Frozen" "Frozen" "Frozen" ...
# $ Sulphur [mg/kg]                    : num  5 5.4 4.2 13.7 6.1 9.9 4.9 5.8 3.2 15 ...
# $ Sulphur Meth                       : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Texture                            : num  1 1 1.5 1 1 1.5 1 2.5 2.5 3.5 ...
# $ Texture Meth                       : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...
# $ Url                                : chr  "https://data.bioplatforms.com//organization/australian-microbiome?q=sample_id:102.100.100/12465" "https://data.bioplatforms.com//organization/australian-microbiome?q=sample_id:102.100.100/12469" "https://data.bioplatforms.com//organization/australian-microbiome?q=sample_id:102.100.100/12471" "https://data.bioplatforms.com//organization/australian-microbiome?q=sample_id:102.100.100/12473" ...
# $ Utc Date Sampled                   : POSIXct, format: "2013-08-14" "2013-08-14" "2013-08-14" "2013-08-14" ...
# $ Vegetation Type                    : chr  "Forest" "Forest" "Forest" "Forest" ...
# $ Water Content [%]                  : num  39.6 25.8 23.4 22.9 29 ...
# $ Water Content Soil Meth            : num  2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 2.1 ...

meta$sample <- gsub(pattern = "102.100.100/", replacement = "", x = meta$`Sample ID`)

row.names(meta) <- meta$sample

( varnames <- names(meta) )
# [1] "Sample ID"                           "Am Environment"                      "Ammonium Nitrogen Wt [mg/kg]"        "Ammonium Nitrogen Wt Meth"          
# [5] "Biotic Relationship"                 "Boron Hot Cacl2 [mg/kg]"             "Boron Hot Cacl2 Meth"                "Clay [%]"                           
# [9] "Clay Meth"                           "Coarse Sand [%]"                     "Coarse Sand Meth"                    "Collection Date"                    
# [13] "Color"                               "Conductivity [dS/m]"                 "Conductivity Meth"                   "Database Schema Definitions Url"    
# [17] "Depth [m]"                           "Depth Lower [m]"                     "Depth Upper [m]"                     "Dtpa Copper [mg/kg]"                
# [21] "Dtpa Copper Meth"                    "Dtpa Iron [mg/kg]"                   "Dtpa Iron Meth"                      "Dtpa Manganese [mg/kg]"             
# [25] "Dtpa Manganese Meth"                 "Dtpa Zinc [mg/kg]"                   "Dtpa Zinc Meth"                      "Elev [m]"                           
# [29] "Env Broad Scale"                     "Env Local Scale"                     "Env Medium"                          "Exc Aluminium [meq/100g]"           
# [33] "Exc Aluminium Meth"                  "Exc Calcium [meq/100g]"              "Exc Calcium Meth"                    "Exc Magnesium [meq/100g]"           
# [37] "Exc Magnesium Meth"                  "Exc Potassium [meq/100g]"            "Exc Potassium Meth"                  "Exc Sodium [meq/100g]"              
# [41] "Exc Sodium Meth"                     "Fine Sand [%]"                       "Fine Sand Meth"                      "General Env Feature"                
# [45] "Geo Loc Name"                        "Gravel [%]"                          "Gravel Meth"                         "Hyperspectral Analysis"             
# [49] "Hyperspectral Analysis Meth"         "Lat Lon [decimal_degrees]"           "Latitude [decimal_degrees]"          "Local Class"                        
# [53] "Longitude [decimal_degrees]"         "Nitrate Nitrogen [mg/kg]"            "Nitrate Nitrogen Meth"               "Nucl Acid Ext"                      
# [57] "Organic Carbon [%]"                  "Organic Carbon Meth"                 "Ph"                                  "Ph Meth"                            
# [61] "Ph Solid H2o"                        "Ph Solid H2o Meth"                   "Phosphorus Colwell [mg/kg]"          "Phosphorus Colwell Meth"            
# [65] "Potassium Colwell [mg/kg]"           "Potassium Colwell Meth"              "Sample Database File"                "Sample Metadata Ingest Date"        
# [69] "Sample Metadata Ingest File"         "Sample Metadata Update History"      "Sample Site Location Description"    "Sample Type"                        
# [73] "Sand [%]"                            "Sand Meth"                           "Silt [%]"                            "Silt Meth"                          
# [77] "Slope Aspect [direction_or_degrees]" "Slope Gradient [%]"                  "Source Mat"                          "Store Cond"                         
# [81] "Sulphur [mg/kg]"                     "Sulphur Meth"                        "Texture"                             "Texture Meth"                       
# [85] "Url"                                 "Utc Date Sampled"                    "Vegetation Type"                     "Water Content [%]"                  
# [89] "Water Content Soil Meth"             "sample" 

dim(meta) # 104 90

unique(meta$`Env Local Scale`)
# [1] "1.1.3 National park"                   "3.3.6 Cotton"                          "3.4.4 Vine fruits"                    
# [4] "3.4.1 Tree fruits"                     "4.5 Irrigated seasonal horticulture"   "4.2.3 Irrigated legume/grass mixtures"
# [7] "1.1 Nature conservation"               "3.2.1 Native/exotic pasture mosaic"    "6.5.1 Marsh/wetland‚Äîconservation"   
# [10] "1.1.5 Habitat/species management area" "3.3.1 Cereals"                         "1.3.4 Rehabilitation"                 
# [13] "1.3.3 Residual native cover"           "1.1.6 Protected landscape"    

keep_samptype <- c("1.1 Nature conservation", "1.1.3 National park", "3.3.1 Cereals", "3.3.6 Cotton",
                   "3.4.1 Tree fruits", "3.4.4 Vine fruits", "4.2.3 Irrigated legume/grass mixtures", "4.5 Irrigated seasonal horticulture"
                   ,"6.5.1 Marsh/wetland‚Äîconservation" 
                   )

sel <- which(meta$`Env Local Scale`=="6.5.1 Marsh/wetland‚Äîconservation")
meta[sel, ] # 19471

# i.e. exclude ambiguous and outlying sample types: 
# "3.2.1 Native/exotic pasture mosaic", "1.1.5 Habitat/species management area",
# "1.3.4 Rehabilitation", "1.3.3 Residual native cover", "1.1.6 Protected landscape" 

meta <- filter(.data = meta, `Env Local Scale` %in% keep_samptype)

table (meta$`Env Broad Scale`)
# 1 Conservation and Natural Environments   3 Production from Dryland Agriculture and Plantations 
# 55                                                      28 
# 4 Production from Irrigated Agriculture and Plantations 
# 1

meta$nat_vs_dist <- "Disturbed"

sel <- which(meta$`Env Broad Scale` == "1 Conservation and Natural Environments")
meta$nat_vs_dist[sel] <- "Natural"

table(meta$nat_vs_dist)
# Disturbed   Natural 
# 29        55 

names(meta)

# month of sampling?
meta$sample_month <- as.character(meta$`Collection Date`)
meta$sample_month <- gsub(pattern = "^....-", replacement = "", x = meta$sample_month)
meta$sample_month <- gsub(pattern = "-..$", replacement = "", x = meta$sample_month)
meta$sample_month <- as.numeric(meta$sample_month)

varnames <- names(meta)

sel <- which(varnames %in% c("nat_vs_dist",
                             # "sample",
                             "Latitude [decimal_degrees]","Longitude [decimal_degrees]",
                             "Ammonium Nitrogen Wt [mg/kg]",
                             "Boron Hot Cacl2 [mg/kg]", "Clay [%]",
                             "Coarse Sand [%]",
                             "Conductivity [dS/m]",     
                             "Dtpa Copper [mg/kg]",
                             "Dtpa Iron [mg/kg]",
                             "Dtpa Manganese [mg/kg]",
                             "Dtpa Zinc [mg/kg]", 
                             "Elev [m]",
                             "Exc Aluminium [meq/100g]",
                             "Exc Calcium [meq/100g]",
                             "Exc Magnesium [meq/100g]",         
                             "Exc Potassium [meq/100g]",
                             "Exc Sodium [meq/100g]",
                             "Fine Sand [%]",
                             "Gravel [%]",
                             # "Local Class"                        
                             "Nitrate Nitrogen [mg/kg]" ,
                             "Organic Carbon [%]",
                             "Ph",
                             "Ph Solid H2o", 
                             "Phosphorus Colwell [mg/kg]",
                             "Potassium Colwell [mg/kg]",
                             # "Sample Site Location Description"
                             "Sand [%]",
                             "Silt [%]",
                             "Slope Gradient [%]",
                             "Sulphur [mg/kg]",
                             #"Texture"   
                             "sample_month",
                             #"Vegetation Type",
                             "Water Content [%]"))

table(meta$`Vegetation Type`)
# Forest Grassland Heathland Marsh/bog     Other  Woodland 
# 15        15         1         3        26        24 

plot_vars <- varnames[sel]
length(plot_vars) # 32

meta.melt <- melt(meta[ , plot_vars])
# Using nat_vs_dist as id variables

names(meta.melt)
# "nat_vs_dist" "variable"    "value" 

sel.plot <- which(meta.melt$variable %in% plot_vars)

p <- ggplot( meta.melt[sel.plot, ], aes(x = nat_vs_dist, y = value )) +
  #geom_boxplot()+
  #geom_point() +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(size=1.5,width = 0.15, alpha=0.25) +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free_y")+
  theme_bw()+
  theme(
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    strip.text = element_text(size = rel(0.7)),
    #strip.text.y = element_text(size = rel(1.1)),
    strip.background = element_blank()
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","AMI-temperate-7.5to45clay-metadata-values.tiff"), width = 22, height = 28, units = "cm", res=600, compression="lzw", type = "cairo")


#-------------------------

#### AMI - read in superfocus - fxn potential outputs
#-------------------------

sampid <- meta$sample

length(sampid) # 84
sampid
# [1] "12465" "12469" "12471" "12473" "12475" "12477" "12479" "12491" "12493" "12495" "12497" "12501" "12503" "12505" "12507" "12509" "12511" "12513" "12515"
# [20] "12517" "12519" "12521" "12523" "12525" "12527" "12560" "12562" "12564" "12566" "12568" "12570" "12572" "12574" "12576" "12578" "12580" "12582" "12590"
# [39] "12614" "12616" "12618" "12620" "12622" "12624" "13260" "13274" "13278" "13729" "13731" "13733" "13735" "13906" "19471" "7845"  "7861"  "7863"  "7888" 
# [58] "7890"  "7900"  "7902"  "8076"  "8078"  "8084"  "8088"  "8126"  "8138"  "8140"  "8142"  "8144"  "8146"  "8150"  "8152"  "9430"  "9434"  "9436"  "9438" 
# [77] "9440"  "9442"  "9446"  "9458"  "9462"  "9464"  "9466"  "9468" 


superfocus_out_dir <- "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami"

list.dirs(superfocus_out_dir)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami"                     
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/fastqc_reports"      
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12465"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12469"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12471"
# etc ...

# don't keep 1st two 
( results_dirs <- list.dirs(superfocus_out_dir)[-c(1,2)] )
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12465"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12469"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12471"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12473"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_12475"
# etc ...

# check identical order
identical(as.character(sampid), gsub(pattern = "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/ami/superfocus_out_", replacement = "", x = results_dirs) )
# TRUE

# collate results into a long-format table

sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)

for (i in 1:length(sampid)) {
  #i<-3
  this_samp <- sampid[i]
  sel.folder <- grep(pattern = this_samp, x = results_dirs)
  this_folder <- results_dirs[sel.folder]
  
  #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)
  
  tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
  # names(tab)
  # [1] "Subsystem.Level.1"                                                   
  # [2] "Subsystem.Level.2"                                                   
  # [3] "Subsystem.Level.3"                                                   
  # [4] "Function"                                                            
  # [5] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq"  
  # [6] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq.."
  
  
  # [1] "Subsystem.Level.1"                                                                                                 
  # [2] "Subsystem.Level.2"                                                                                                 
  # [3] "Subsystem.Level.3"                                                                                                 
  # [4] "Function"                                                                                                          
  # [5] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"  
  # [6] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"  
  # [7] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"  
  # [8] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"  
  # [9] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [10] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  # [11] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [12] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  
  
  tab$sampid <- this_samp
  names(tab)
  
  #tab <- tab[,c(7,1,2,3,4,6)]
  
  # last column is sampid
  # take average of percentages
  
  sel.col.percent <- grep(pattern = "R1.good.fastq..$", x = names(tab))
  if (length(sel.col.percent)>1) {
    tab$percent_abun <- apply(X = tab[ ,sel.col.percent], MARGIN = 1, FUN = mean )
  } else {
    tab$percent_abun <- tab[ ,sel.col.percent]
  }
  
  # sum(tab$percent_abun) # 100
  # mean(tab$percent_abun) # 0.004338583
  
  names(sfx.long) # "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"    "percent_abun"
  # names(tab)
  # [1] "Subsystem.Level.1"                                                                                                 
  # [2] "Subsystem.Level.2"                                                                                                 
  # [3] "Subsystem.Level.3"                                                                                                 
  # [4] "Function" 
  # ...
  # [13] "sampid"                                                                                                            
  # [14] "percent_abun"   
  
  tab <- tab[ ,c("sampid","Subsystem.Level.1","Subsystem.Level.2","Subsystem.Level.3","Function","percent_abun")]
  names(tab) <- names(sfx.long)
  
  sfx.long <- rbind(sfx.long,tab)
  
  print(paste0("completed ",i," - sample ID: ",sampid[i]))
}

head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
head(sfx.long)
#   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 2    12465 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 4.085207e-04
# 3    12465 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 2.288676e-05
# 4    12465 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 5.276470e-05
# 5    12465 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.152034e-05
# 6    12465 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.636470e-03
# 7    12465 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 5.017155e-06

sfx.long$full_fxn_tax <- paste0(sfx.long$subsys_L1,"___", sfx.long$subsys_L2,"___", sfx.long$subsys_L3,"___", sfx.long$fxn)
head(sfx.long)
#   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 2    12465 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 4.085207e-04
# 3    12465 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 2.288676e-05
# 4    12465 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 5.276470e-05
# 5    12465 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.152034e-05
# 6    12465 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.636470e-03
# 7    12465 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 5.017155e-06
#                                                                                                                           full_fxn_tax
# 2                                         Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase
# 3                                                   Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase
# 4 Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)
# 5                       Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)
# 6                                       Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)
# 7                                    Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 37335    85

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 37335
length(unique(full_fxn_names)) # 37335

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 


tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)
# full_fxn_tax        12465        12469        12471
# 1                                         Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase 4.085207e-04 4.205200e-04 2.964560e-04
# 2                                                   Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase 2.288676e-05 2.821954e-05 0.000000e+00
# 3 Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 5.276470e-05 5.603135e-05 1.824345e-05
# 4                       Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.152034e-05 2.634775e-05 3.420646e-05
# 5                                       Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.636470e-03 1.417673e-03 1.334052e-03
# 6                                    Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 5.017155e-06 6.811763e-05 0.000000e+00
# 12473        12475        12477        12479        12491        12493        12495        12497        12501        12503        12505        12507        12509
# 1 3.198079e-04 3.852306e-04 3.076140e-04 4.778209e-04 2.779646e-04 2.142208e-04 2.308746e-04 3.619040e-04 0.0004374537 5.517850e-04 2.631182e-04 9.482282e-05 2.618330e-04
# 2 1.035635e-05 4.866925e-05 3.717924e-05 2.275338e-05 1.166878e-05 3.481088e-05 0.000000e+00 7.000745e-05 0.0000000000 1.718886e-05 2.517006e-05 0.000000e+00 1.095440e-05
# 3 2.051317e-05 1.165885e-05 2.124528e-05 6.929437e-05 0.000000e+00 5.087744e-05 2.823167e-05 1.410095e-05 0.0001044173 5.884055e-05 1.608532e-05 0.000000e+00 7.018106e-05
# 4 2.616284e-05 1.391540e-04 7.170283e-05 1.965064e-04 0.000000e+00 5.355520e-05 3.347418e-05 3.241173e-05 0.0001551178 3.477614e-05 6.081880e-05 3.189143e-05 2.684210e-05
# 5 1.783856e-03 2.032855e-03 1.633231e-03 1.566880e-03 1.736613e-03 1.569837e-03 1.498095e-03 1.832041e-03 0.0022915319 1.683597e-03 1.676688e-03 1.067395e-03 1.864685e-03
# 6 3.883362e-05 3.669575e-05 4.204795e-05 5.946905e-05 5.844942e-05 4.016640e-05 2.572101e-05 6.472313e-05 0.0000000000 7.034909e-05 1.118337e-05 3.132379e-05 1.878947e-05
# etc ....

names(sfx.wide)
# [1] "full_fxn_tax" "12465"        "12469"        "12471"        "12473"        "12475"        "12477"        "12479"        "12491"        "12493"        "12495"       
# [12] "12497"        "12501"        "12503"        "12505"        "12507"        "12509"        "12511"        "12513"        "12515"        "12517"        "12519"       
# [23] "12521"        "12523"        "12525"        "12527"        "12560"        "12562"        "12564"        "12566"        "12568"        "12570"        "12572"       
# [34] "12574"        "12576"        "12578"        "12580"        "12582"        "12590"        "12614"        "12616"        "12618"        "12620"        "12622"       
# [45] "12624"        "13260"        "13274"        "13278"        "13729"        "13731"        "13733"        "13735"        "13906"        "19471"        "7845"        
# [56] "7861"         "7863"         "7888"         "7890"         "7900"         "7902"         "8076"         "8078"         "8084"         "8088"         "8126"        
# [67] "8138"         "8140"         "8142"         "8144"         "8146"         "8150"         "8152"         "9430"         "9434"         "9436"         "9438"        
# [78] "9440"         "9442"         "9446"         "9458"         "9462"         "9464"         "9466"         "9468"

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide)
# [1] "12465" "12469" "12471" "12473" "12475" "12477" "12479" "12491" "12493" "12495" "12497" "12501" "12503" "12505" "12507" "12509" "12511" "12513" "12515" "12517" "12519" "12521"
# [23] "12523" "12525" "12527" "12560" "12562" "12564" "12566" "12568" "12570" "12572" "12574" "12576" "12578" "12580" "12582" "12590" "12614" "12616" "12618" "12620" "12622" "12624"
# [45] "13260" "13274" "13278" "13729" "13731" "13733" "13735" "13906" "19471" "7845"  "7861"  "7863"  "7888"  "7890"  "7900"  "7902"  "8076"  "8078"  "8084"  "8088"  "8126"  "8138" 
# [67] "8140"  "8142"  "8144"  "8146"  "8150"  "8152"  "9430"  "9434"  "9436"  "9438"  "9440"  "9442"  "9446"  "9458"  "9462"  "9464"  "9466"  "9468" 

head(sampid)
# "12465" "12469" "12471" "12473" "12475" "12477"

names(sampid) # NULL - in this case there is NOT an alternative sample name being used

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # TRUE

#NOT RUN THIS TIME
#names(sfx.wide) <- names(sampid)



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 37335     4

length(unique(tax.fxn$subsys_L1)) # 35
length(unique(tax.fxn$subsys_L2)) # 194
length(unique(tax.fxn$subsys_L3)) # 1268
length(unique(tax.fxn$fxn)) # 17657

sel.dup <- which(duplicated.default(tax.fxn$fxn)==TRUE)
head(tax.fxn$fxn[sel.dup])



# examine last example 'tab'

head(tab)
#   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 1     9468 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 3.819676e-04
# 2     9468 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 1.766030e-05
# 3     9468 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 2.587839e-05
# 4     9468 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.514521e-05
# 5     9468 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.287010e-03
# 6     9468 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 4.401921e-05


length(unique(tab$subsys_L1)) # 35
length(unique(tab$subsys_L2)) # 191
length(unique(tab$subsys_L3)) # 1184
length(unique(tab$fxn)) # 10975

#-------------------------

#### AMI - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 37335     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 37335    84

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 37335 taxa and 84 samples ]
# tax_table()   Taxonomy Table:    [ 37335 taxa by 4 taxonomic ranks ]

sample_names(phy)
# [1] "12465" "12469" "12471" "12473" "12475" "12477" "12479" "12491" "12493" "12495" "12497" "12501" "12503" "12505" "12507" "12509" "12511" "12513" "12515" "12517" "12519" "12521"
# [23] "12523" "12525" "12527" "12560" "12562" "12564" "12566" "12568" "12570" "12572" "12574" "12576" "12578" "12580" "12582" "12590" "12614" "12616" "12618" "12620" "12622" "12624"
# [45] "13260" "13274" "13278" "13729" "13731" "13733" "13735" "13906" "19471" "7845"  "7861"  "7863"  "7888"  "7890"  "7900"  "7902"  "8076"  "8078"  "8084"  "8088"  "8126"  "8138" 
# [67] "8140"  "8142"  "8144"  "8146"  "8150"  "8152"  "9430"  "9434"  "9436"  "9438"  "9440"  "9442"  "9446"  "9458"  "9462"  "9464"  "9466"  "9468"


### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta) # 84  92
identical( row.names(meta), sample_names(phy) ) # TRUE
head(row.names(meta)) # "12465" "12469" "12471" "12473" "12475" "12477"
head(sample_names(phy)) # "12465" "12469" "12471" "12473" "12475" "12477"



SAMP <- sample_data(meta)


### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 37335 taxa and 84 samples ]
# sample_data() Sample Data:       [ 84 samples by 92 sample variables ]
# tax_table()   Taxonomy Table:    [ 37335 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#       subsys_L1                     subsys_L2 subsys_L3             fxn                                                                            
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylaconitate_cis-trans_isomerase"                                        
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylaconitate_isomerase"                                                  
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)"
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)"                      
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                      
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 


getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-AMI.RDS")

phy <- readRDS("phy-phyloseq-object-AMI.RDS")

#-------------------------

#### AMI - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-AMI.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 37335     4


## use this sub-function later
atomic_no
# function(form, atom) {
#   ##form=form
#   ##atom="O"
#   coefs <- list()
#   for (c in 1:length(form)) {
#     #c<-1
#     pos <- str_locate(string = form[c], pattern = atom)[1]
#     if (is.na(pos)) { # not present
#       ##print("zero")
#       coef<-0
#     } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A 
#       ##print("A")
#       coef<-1
#     } else {
#       # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
#       ##print("B")
#       checkstring <- substring(text = form[c], first = pos, last = pos+1)
#       
#       if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
#         ##print("C")
#         coef<-0
#       } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
#         ##print("D")
#         coef<-0
#       } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E 
#         ##print("E")
#         coef<-0
#       } else { # F
#         ##print("F")
#         coef<-1
#         coef.try <- coef
#         pos.end <- pos+1
#         
#         while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
#           # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
#           coef <- coef.try
#           coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
#           pos.end <- pos.end+1
#           ##print("G")
#         } # END while loop
#       } # END else
#       
#     } # END else
#     
#     coefs[[c]] <- coef
#     
#   } # END c loop
#   return( unlist(coefs) )
# }
# <bytecode: 0x7fe50641d6f8>



get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)
        
        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1
          
          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else
        
      } # END else
      
      coefs[[c]] <- coef
      
    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  

  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }

            
    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/AMI-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%  #
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()



time.start #  "2023-04-06 22:22:53 ACST"
time.finish # "2023-04-07 01:25:30 ACST"

# 3 hrs


## assemble results

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/AMI-R-working-files"



# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# total results written to disk - 37335
dim(df.tax) # 37335     4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 45026     9

table(df.out$matching_method)
# EC number                Exact hierachy match Matched ModelSEED Reaction pathways           Matched Reactions aliases 
# 16755                                5339                                   4                                 136 
# Matched Reactions name               Matched Subsytem role                      No match found 
# 918                                1059                               20815 

100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 46.22885% No match found

# 54% match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.605   0.800   0.844   1.048   3.333   18242

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.240   1.400   1.411   1.570   4.000   18242 

table(df.out.distil$matching_method)
# EC number                                  EC number;Exact hierachy match 
# 11636                                                            1133 
# EC number;Exact hierachy match;Matched Reactions name            EC number;Exact hierachy match;Matched Subsytem role 
# 2                                                               6 
# EC number;Exact hierachy match;No match found    EC number;Matched ModelSEED Reaction pathways;No match found 
# 41                                                               4 
# EC number;Matched Reactions aliases                                EC number;Matched Reactions name 
# 14                                                              56 
# EC number;Matched Reactions name;No match found                                 EC number;Matched Subsytem role 
# 4                                                             150 
# EC number;Matched Subsytem role;No match found                                        EC number;No match found 
# 8                                                            1310 
# Exact hierachy match                  Exact hierachy match;Matched Reactions aliases 
# 3033                                                               3 
# Exact hierachy match;Matched Reactions name                      Exact hierachy match;Matched Subsytem role 
# 19                                                              50 
# Exact hierachy match;Matched Subsytem role;No match found                             Exact hierachy match;No match found 
# 4                                                             476 
# Matched Reactions aliases Matched Reactions aliases;Matched Reactions name;No match found 
# 107                                                               3 
# Matched Reactions aliases;No match found                                          Matched Reactions name 
# 9                                                             794 
# Matched Reactions name;Matched Subsytem role                           Matched Reactions name;No match found 
# 4                                                              36 
# Matched Subsytem role                            Matched Subsytem role;No match found 
# 707                                                             111 
# No match found 
# 17615 


100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 39.12184% No match found

100-39 # i.e. in 61% of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_AMI.RDS")

df.out.distil <-readRDS( "df.out.distil_AMI.RDS" )


#-------------------------

#### AMI - inspect & clean data
#-------------------------

this_study <- "-AMI-"
phy <- readRDS("phy-phyloseq-object-AMI.RDS")
df.out.distil <-readRDS( "df.out.distil_AMI.RDS" )

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 37335    4

phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE

df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 37225 84
df.OTU[1:5, 1:8]
# 12465        12469        12471        12473        12475        12477        12479        12491
# fxn_1 4.085207e-04 4.205200e-04 2.964560e-04 3.198079e-04 3.852306e-04 3.076140e-04 4.778209e-04 2.779646e-04
# fxn_2 2.288676e-05 2.821954e-05 0.000000e+00 1.035635e-05 4.866925e-05 3.717924e-05 2.275338e-05 1.166878e-05
# fxn_3 5.276470e-05 5.603135e-05 1.824345e-05 2.051317e-05 1.165885e-05 2.124528e-05 6.929437e-05 0.000000e+00
# fxn_4 3.152034e-05 2.634775e-05 3.420646e-05 2.616284e-05 1.391540e-04 7.170283e-05 1.965064e-04 0.000000e+00
# fxn_5 1.636470e-03 1.417673e-03 1.334052e-03 1.783856e-03 2.032855e-03 1.633231e-03 1.566880e-03 1.736613e-03

mean( df.OTU[ , "12465"] ) # 0.002678452
sum( df.OTU[ , "12465"] ) # 100

sample_sums(phy) # all values of 100


identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
# tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.8571429     0.7142857
# fxn_3     0.9285714     0.8571429
# fxn_4     0.9285714     0.8571429
# fxn_5     0.6000000     1.8000000
# fxn_6     1.0000000     0.4285714


dim(df.fxn_vk_coords) # 37335     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.605   0.800   0.844   1.048   3.333   18242
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.240   1.400   1.411   1.570   4.000   18242

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.8571429     0.7142857
# fxn_3     0.9285714     0.8571429
# fxn_4     0.9285714     0.8571429
# fxn_5     0.6000000     1.8000000
# fxn_6     1.0000000     0.4285714
# fxn_8     0.6666667     2.3333333


## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 84
dim(df.fxn_vk_coords.noNAs) # 19093     2
dim(df.OTU.noNAs) # 19093    84

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# 12469    12473     9462    12475    12479    12477    12471     9458    12465     9442     9440     9438     9468    12570    13906     8142 
# 58.94168 59.02717 59.07146 59.12597 59.14566 59.15662 59.33078 59.38991 59.52911 59.55300 59.55650 59.58615 59.58722 59.72424 59.73118 59.77067 
# 9430     9446    12493     8146    12566    12582     7845    12562    12497     9466    12574     8152    12578    19471    12568     8150 
# 59.80635 59.89128 59.90420 59.95248 60.01637 60.02838 60.03031 60.06046 60.08412 60.10192 60.18361 60.24211 60.24603 60.24916 60.47393 60.49589 
# 13278    12580     9464    12576    12564    12495    12622     9436    12491    12618     8144    12560    12507     9434    12624    12620 
# 60.52927 60.54084 60.54711 60.59718 60.61393 60.62846 60.64361 60.71522 60.75624 60.79506 60.80340 60.84723 60.87051 60.88415 60.92412 60.94276 
# 12614    13274    12572     8126    12616     8076    12523    12525     8084    12515    12501    12590    12527    12517    12503     7888 
# 61.01824 61.05406 61.11714 61.13596 61.14912 61.21357 61.51069 61.53974 61.54131 61.54900 61.59474 61.64000 61.66101 61.68093 61.71150 61.73959 
# 7861     8138    12509    12513     7863    12505    12521    12519    13260     8140     8088     8078     7902    12511    13735     7900 
# 61.77916 61.82820 61.86047 61.92137 61.92260 61.93542 61.97081 62.06440 62.06751 62.08928 62.09331 62.14191 62.28928 62.48257 62.53797 62.55450 
# 13729    13731    13733     7890 
# 62.60352 62.71204 62.86212 62.88386

summary( colSums(df.OTU.noNAs) )
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.94   60.03   60.80   60.85   61.72   62.88

# this reflects that 59-63% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

## PREVIOUSLY for env/soil sample
## this reflects the just >60% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

colSums(df.OTU)
# all values of 100



# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE


dim(df.fxn_vk_coords.noNAs) # 19093     2
dim(df.in) # 19093    84

row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  group=NA, 
                                  samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"],
                         group= rep( phy@sam_data$nat_vs_dist[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}



dim(vk.samp.noNAs.long) # 1603813       8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok)

str(vk.samp.noNAs.long)
unique(vk.samp.noNAs.long$group) # "Natural"   "Disturbed"
class(vk.samp.noNAs.long$group) # "character"
vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c("Disturbed","Natural"), ordered = TRUE) # CHECK GROUP VARIABLE !!
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	1603812 obs. of  8 variables:
#   $ sample   : chr  "12465" "12465" "12465" "12465" ...
# $ fxn_id   : Factor w/ 19093 levels "fxn_10","fxn_100",..: 6152 11143 14787 15443 16271 17690 18441 1 583 1103 ...
# $ abun     : num  2.29e-05 5.28e-05 3.15e-05 1.64e-03 5.02e-06 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" ...


saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-object-AMI.RDS")


#vk.samp.noNAs.long <- readRDS(file = "vk.samp.noNAs.long-object-AMI.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.text = element_text(angle=25),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "AMI disturbed & natural soils", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust =0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")



## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/AMI-R-working-files"

i<-2
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
names(temp)
names( temp[[1]] ) # "fxns" "rxns"
temp[[1]][["fxns"]]
# superfocus_fxn f                       f__in        matching_method min_adist_modelSEED min_amatch_modelSEED     rxns tot_mean_OC_x tot_mean_HC_y
# 1          fxn_2 1 2-methylaconitate isomerase Matched Reactions name                  NA                   NA rxn25278     0.8571429     0.7142857
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                       f__in   rxn_id                    rxn_name                             rxn_eqn
# 1          fxn_2 1 2-methylaconitate isomerase rxn25278 2-methylaconitate isomerase (1) cpd25681[0] <=> (1) cpd02597[0]
# rxn_defn            compds compd_coef    chem_formx                           OC_ratios
# 1 (1) 2-methyl-trans-aconitate[0] <=> (1) cis-2-Methylaconitate[0] cpd25681;cpd02597        1;1 C7H5O6;C7H5O6 0.857142857142857;0.857142857142857
# HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 0.714285714285714;0.714285714285714       0.8571429       0.7142857



## check anomalous high value in Lignin zone in PCaN density plot
p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405)
length(sel) # 45528
dim(vk.samp.noNAs.long) # 1603812       8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 2.838737% of all vK-coordinates - on average per sample

head( vk.samp.noNAs.long[ sel, ] )
# sample  fxn_id         abun abun_type      OC_x     HC_y   group       samplabel
# 19   12465  fxn_24 0.000000e+00   relabun 0.5206427 1.400701 Natural 12465 (Natural)
# 33   12465  fxn_45 1.505146e-05   relabun 0.5206427 1.400701 Natural 12465 (Natural)
# 84   12465 fxn_112 1.316177e-05   relabun 0.5206427 1.400701 Natural 12465 (Natural)
# 117  12465 fxn_175 1.486617e-05   relabun 0.5206427 1.400701 Natural 12465 (Natural)
# 156  12465 fxn_246 0.000000e+00   relabun 0.5206427 1.400701 Natural 12465 (Natural)
# 188  12465 fxn_286 0.000000e+00   relabun 0.5206427 1.400701 Natural 12465 (Natural)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.06223107 = 0.062% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 60.84631 = 61.8% per sample

# remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702) # qty 45528


#fxn_24
i<-24
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED
# 1         fxn_24 1 hypothetical protein Matched Reactions name                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427      1.400701



### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 45528


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-AMI.RDS")


#-------------------------

#### AMI - CPP-class: total rel abun in vk spatial regions
#-------------------------

phy <- readRDS("phy-phyloseq-object-AMI.RDS")
df.out.distil <-readRDS( "df.out.distil_AMI.RDS" )
df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-AMI.RDS")

# use cleaned version
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

this_study <- "-AMI-"
header <- "-vk-zones-"


str(vk.samp.noNAs.long)
# 'data.frame':	1558284 obs. of  8 variables:
# $ sample   : chr  "12465" "12465" "12465" "12465" ...
# $ fxn_id   : Factor w/ 19093 levels "fxn_10","fxn_100",..: 6152 11143 14787 15443 16271 17690 18441 1 583 1103 ...
# $ abun     : num  2.29e-05 5.28e-05 3.15e-05 1.64e-03 5.02e-06 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" ...

# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 


spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) #   1025800       5

table(spdf$group)
# Disturbed   Natural 
# 367598    680312 

head(spdf)
# sample   group      OC_x      HC_y         abun
# 2  12465 Natural 0.8571429 0.7142857 2.288676e-05
# 3  12465 Natural 0.9285714 0.8571429 5.276470e-05
# 4  12465 Natural 0.9285714 0.8571429 3.152034e-05
# 5  12465 Natural 0.6000000 1.8000000 1.636470e-03
# 6  12465 Natural 1.0000000 0.4285714 5.017155e-06
# 7  12465 Natural 0.6666667 2.3333333 3.418492e-03


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 84

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.89   59.97   60.74   60.78   61.65   62.84

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"

## now want to sum total rela abun within vK spatial zones

## to de-duplicate get total relabun within a sample, then mean across samples

sum_relabun_in_vk_zones_env <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  # lipids
  subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  subsel.all <- c(subsel.lipids)
  # proteins
  subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.proteins)
  # amino_sugars
  subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.amino_sugars)
  # carbs - Carbohydrates
  subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  subsel.all <- c(subsel.all, subsel.carbs)
  # cond_aromtx - Condensed aromatics
  subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  subsel.all <- c(subsel.all, subsel.cond_aromtx)
  # lignin
  subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.lignin)
  # tannins
  subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.tannins)
  # Other compounds ...
  
  # methylated
  subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
  # and not already used?
  sel.dup <- which(subsel.methylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.methylated)
  
  # hydrated
  subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
  # and not already used?
  sel.dup <- which(subsel.hydrated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.hydrated)
  
  # demethylated / Oxidised-dehydrogenated
  subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.demethylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.demethylated)
  
  # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
  #subsel.condensed = which(                   spdf_in$OC_x <= 0.97                       & spdf_in$HC_y < 0.75)
  subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
  
  # and not already used?
  sel.dup <- which(subsel.condensed %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.condensed)
  
  # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
  subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.k2_b12 %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
  
  df_out <- data.frame(
    sample = this_samp,
    group = unique(spdf_in$group),
    
    lipids = sum(spdf_in$abun[ subsel.lipids ]),
    proteins = sum(spdf_in$abun[ subsel.proteins ]),
    amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
    carbs = sum(spdf_in$abun[ subsel.carbs ]),
    cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
    lignin = sum(spdf_in$abun[ subsel.lignin ]),
    tannins = sum(spdf_in$abun[ subsel.tannins ]),
    # Other compounds ...
    methylated = sum(spdf_in$abun[ subsel.methylated ]),
    hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
    demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
    condensed = sum(spdf_in$abun[ subsel.condensed ]),
    k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
  )
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones_env( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 84
out[[1]]
# sample   group   lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1  12465 Natural 0.283273 2.602413     4.670389 9.217376    1.101779 8.077073 15.23948  0.4211606 4.397567      12.5363   0.49295 0.4325103

names(out[[1]])
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"

df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample   group    lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1  12465 Natural 0.2832730 2.602413     4.670389 9.217376    1.101779 8.077073 15.23948  0.4211606 4.397567     12.53630 0.4929500 0.4325103
# 2  12469 Natural 0.2880593 2.591837     4.800173 9.044561    1.136709 8.035447 15.13495  0.4148774 4.247478     12.25814 0.4662095 0.4671852
# 3  12471 Natural 0.3003381 2.540432     4.737506 8.907222    1.198854 8.040201 15.36153  0.4111766 4.292015     12.56650 0.4861662 0.4288692
# 4  12473 Natural 0.3040296 2.525280     4.746121 9.150795    1.216679 8.002563 15.05096  0.4090416 4.313162     12.31675 0.4687271 0.4702917
# 5  12475 Natural 0.2966180 2.568097     4.740509 9.281669    1.112116 8.017991 15.08434  0.4176426 4.344765     12.26911 0.4648987 0.4709509
# 6  12477 Natural 0.3047254 2.553195     4.706102 9.335030    1.223663 7.990108 14.92098  0.4220065 4.337522     12.36609 0.4705839 0.4709797

names(df.zones)
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"
names(df.zones[ ,c(3:14)])
# [1] "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"   "hydrated"     "demethylated"
# [11] "condensed"    "k2_b12"

df.zones$all <- apply(X = df.zones[ ,c(3:14)], MARGIN = 1, FUN = sum)

table(df.zones$group)
# Disturbed   Natural 
#        29        55

saveRDS(object = df.zones, file = "df.zones--AMI.RDS")

str(df.zones)
# 'data.frame':	84 obs. of  15 variables:
# $ sample      : chr  "12465" "12469" "12471" "12473" ...
# $ group       : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ lipids      : num  0.283 0.288 0.3 0.304 0.297 ...
# $ proteins    : num  2.6 2.59 2.54 2.53 2.57 ...
# $ amino_sugars: num  4.67 4.8 4.74 4.75 4.74 ...
# $ carbs       : num  9.22 9.04 8.91 9.15 9.28 ...
# $ cond_aromtx : num  1.1 1.14 1.2 1.22 1.11 ...
# $ lignin      : num  8.08 8.04 8.04 8 8.02 ...
# $ tannins     : num  15.2 15.1 15.4 15.1 15.1 ...
# $ methylated  : num  0.421 0.415 0.411 0.409 0.418 ...
# $ hydrated    : num  4.4 4.25 4.29 4.31 4.34 ...
# $ demethylated: num  12.5 12.3 12.6 12.3 12.3 ...
# $ condensed   : num  0.493 0.466 0.486 0.469 0.465 ...
# $ k2_b12      : num  0.433 0.467 0.429 0.47 0.471 ...
# $ all         : num  59.5 58.9 59.3 59 59.1 ...


df.zones.melt <- melt(df.zones, id.vars = c("sample","group"))
head(df.zones.melt)
# sample   group variable     value
# 1  12465 Natural   lipids 0.2832730
# 2  12469 Natural   lipids 0.2880593
# 3  12471 Natural   lipids 0.3003381
# 4  12473 Natural   lipids 0.3040296
# 5  12475 Natural   lipids 0.2966180
# 6  12477 Natural   lipids 0.3047254

unique( df.zones.melt$group )
# [1] Natural   Disturbed
# Levels: Disturbed < Natural

str(df.zones.melt)
# 'data.frame':	1092 obs. of  4 variables:
#   $ sample  : chr  "12465" "12469" "12471" "12473" ...
# $ group   : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ variable: Factor w/ 13 levels "lipids","proteins",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0.283 0.288 0.3 0.304 0.297 ...

hist(df.zones.melt$value)
hist(log10( df.zones.melt$value))

p <- ggplot(data = df.zones.melt, aes(x = group, y = value))+
  geom_boxplot(outlier.shape = NA)+
  #geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  geom_jitter(alpha=0.6)+
  facet_wrap(facets = vars(variable), scales = "free_y")

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=350, compression="lzw",type="cairo")


dat <- df.zones.melt



str(dat)


p <- ggplot(data = dat, aes(x = group, y = value))+ # y = value   y = win95abun
  geom_violin(width = 1.1)+
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(alpha=0.6)+
  facet_wrap(facets = vars(variable), scales = "free_y")
p



## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"


unique(dat$group)
# [1] Natural   Disturbed
# Levels: Disturbed < Natural


y_anotations <- list()
sig_anotations <- list()

# sig converter - Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
sig_symb <- function(x) {
  out <- character(length = length(x))
  for (i in seq_along(x)) {
    if (is.na(x[i])) {
      out[i] <- NA
    } else if (x[i] < 0.001) { 
      out[i] <- "***"
    } else if (x[i] >= 0.001 & x[i] < 0.01) { 
      out[i] <- "**"
    } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
      out[i] <- "*" 
    } else { 
      out[i] <- "ns"
    }
  } # END loop
  return(out)
}


names(dat) # "sample"   "group"    "variable" "value"  



## lipids
this_var <- "Lipids"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.6832, num df = 28, denom df = 54, p-value = 0.1007

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -5.5146, df = 82, p-value = 1.981e-07
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.03365079
# sample estimates:
#   mean of x mean of y 
# 0.2442399 0.2924282

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Proteins"
this_var <- "Proteins"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.69314, num df = 28, denom df = 54, p-value = 0.2949

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 4.2965, df = 82, p-value = 2.369e-05
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.04284445        Inf
# sample estimates:
#   mean of x mean of y 
# 2.644541  2.574625 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.55268, num df = 28, denom df = 54, p-value = 0.0911
# variances are different

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# data:  x and y
# t = -0.87748, df = 82, p-value = 0.1914
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf 0.03262261
# sample estimates:
#   mean of x mean of y 
# 4.638892  4.675303 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 3.6072, num df = 28, denom df = 54, p-value = 5.225e-05

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1385, p-value = 1.671e-08
# alternative hypothesis: true location shift is greater than 0

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.61025, num df = 28, denom df = 54, p-value = 0.1582

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -4.3011, df = 82, p-value = 2.33e-05
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.04649699
# sample estimates:
#   mean of x mean of y 
# 1.002717  1.078544 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.89588, num df = 28, denom df = 54, p-value = 0.7681

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 7.9102, df = 82, p-value = 5.259e-12
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.2204001       Inf
# sample estimates:
#   mean of x mean of y 
# 8.444892  8.165793 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Tannins"
this_var <- "Tannins"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.2033, num df = 28, denom df = 54, p-value = 2.147e-05

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1193, p-value = 0.0001011
# alternative hypothesis: true location shift is greater than 0

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other methylated\ncompounds"  
this_var <- "Other methylated\ncompounds"  

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.2116, num df = 28, denom df = 54, p-value = 0.5355

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 5.784, df = 82, p-value = 6.416e-08
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.04377021        Inf
# sample estimates:
#   mean of x mean of y 
# 0.5081926 0.4467496 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other hydrated\ncompounds"  
this_var <- "Other hydrated\ncompounds" 

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.52118, num df = 28, denom df = 54, p-value = 0.06396

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 5.1968, df = 82, p-value = 7.276e-07
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.1501817       Inf
# sample estimates:
#   mean of x mean of y 
# 4.758157  4.537261 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.23823, num df = 28, denom df = 54, p-value = 0.0001071

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1037, p-value = 0.01227

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.60017, num df = 28, denom df = 54, p-value = 0.1447

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -2.8093, df = 82, p-value = 0.003102
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.03302483
# sample estimates:
#   mean of x mean of y 
# 0.3968897 0.4778698 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds" 

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.88688, num df = 28, denom df = 54, p-value = 0.7453

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -1.3517, df = 82, p-value = 0.0901
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf 0.003116398
# sample estimates:
#   mean of x mean of y 
# 0.4249559 0.4384574 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "All compounds"
this_var <- "All compounds"

x=filter(dat, group == "Disturbed" & variable == this_var)$value  # win95abun
y=filter(dat, group == "Natural" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = value))+
  geom_boxplot()+
  geom_jitter(alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.34251, num df = 28, denom df = 54, p-value = 0.002961

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1323, p-value = 3.921e-07

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"


# all
annotation_df.all <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  start=rep(c(1), times=13), end = rep(c(2), times=13),
  y = c( y_anotations[["Lipids"]],
         y_anotations[["Proteins"]],
         y_anotations[["Amino sugars"]],
         y_anotations[["Carbohydrates"]],
         y_anotations[["Condensed\naromatics"]],
         y_anotations[["Lignin"]],
         y_anotations[["Tannins"]],
         y_anotations[["Other methylated\ncompounds"]],
         y_anotations[["Other hydrated\ncompounds"]],
         y_anotations[["Other demethylated\ncompounds"]],
         y_anotations[["Other condensed\ncompounds"]],
         y_anotations[["Near VitK2-VitB12\ncompounds"]],
         y_anotations[["All compounds"]]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Lipids"]],
            sig_anotations[["Proteins"]],
            sig_anotations[["Amino sugars"]],
            sig_anotations[["Carbohydrates"]],
            sig_anotations[["Condensed\naromatics"]],
            sig_anotations[["Lignin"]],
            sig_anotations[["Tannins"]],
            sig_anotations[["Other methylated\ncompounds"]],
            sig_anotations[["Other hydrated\ncompounds"]],
            sig_anotations[["Other demethylated\ncompounds"]],
            sig_anotations[["Other condensed\ncompounds"]],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]],
            sig_anotations[["All compounds"]]
  ), stringsAsFactors = FALSE)
str(annotation_df.all)

# ensure no conflict in factor levels
annotation_df.all$variable <- factor(annotation_df.all$variable, 
                                     levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                                "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                     ordered = TRUE)


annotation_df.all$label # "***" "***" "ns"  "***" "***" "***" "***" "***" "***" "*"   "**"  "ns"  "***"

sel <- which(annotation_df.all$label == "ns")
annotation_df.all$label[sel] <- NA
annotation_df.all$start[sel] <- NA
annotation_df.all$end[sel] <- NA
annotation_df.all$y[sel] <- NA



p <- ggplot(data = dat, aes(x = group, y = value))+ # y = value   y = win95abun
  #geom_violin(width = 1.1)+
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(width = 0.2, alpha=0.5)+
  
  # all
  geom_text(data=annotation_df.all, mapping = aes(x = 1.5, y = y, label = label), size = 5, vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=annotation_df.all, mapping = aes(x = start, y = y, xend = end, yend = y))+
  
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function % rel. abun. in vK zones") +  # "Function % rel. abun. in vK zones (Winsorized)"
  theme(
    #legend.position="bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.7))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-boxplots-",this_study,header,"-v9.tiff"), width = 24, height = 14, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### AMI - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25)
#-------------------------

phy <- readRDS("phy-phyloseq-object-AMI.RDS")
df.out.distil <-readRDS( "df.out.distil_AMI.RDS" )
df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-AMI.RDS")

# use cleaned version
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

this_study <- "-AMI-"
header <- "-vk-hac-buffers-"


str(vk.samp.noNAs.long)
# 'data.frame':	1558284 obs. of  8 variables:
# $ sample   : chr  "12465" "12465" "12465" "12465" ...
# $ fxn_id   : Factor w/ 19093 levels "fxn_10","fxn_100",..: 6152 11143 14787 15443 16271 17690 18441 1 583 1103 ...
# $ abun     : num  2.29e-05 5.28e-05 3.15e-05 1.64e-03 5.02e-06 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" ...

# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

head(spdf)
# sample   group      OC_x      HC_y         abun
# 2  12465 Natural 0.8571429 0.7142857 2.288676e-05
# 3  12465 Natural 0.9285714 0.8571429 5.276470e-05
# 4  12465 Natural 0.9285714 0.8571429 3.152034e-05
# 5  12465 Natural 0.6000000 1.8000000 1.636470e-03
# 6  12465 Natural 1.0000000 0.4285714 5.017155e-06
# 7  12465 Natural 0.6666667 2.3333333 3.418492e-03

dim(spdf) # 1025800       5


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 84

hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000010 0.0000415 0.0002763 0.0049774 0.0027449 1.0768881

12.5/(pi*0.05^2) # 1591.549

add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.89   59.97   60.74   60.78   61.65   62.84

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"  

## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2
# df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
# df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)

sum_relabun_in_vk_hac_buffers_env <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  df_out <- data.frame(sample=NA, group=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      group = unique(spdf_in$group),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
    )
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers_env( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 84
out[[1]]
# sample   group  rad  Acetate Propionate Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2  12465 Natural 0.05 17.42657  22.169283 0.000000 41.71924 13.727015 132.95744  2.177403 1.1463411  18.89142 57.41581
# 3  12465 Natural 0.10 13.12582   8.803963 7.594442 16.06896  6.795372 113.18799  3.615191 0.8336420  50.37943 53.92664
# 4  12465 Natural 0.15 10.97779   6.413177 7.816726 14.71928  8.846629  80.91578 11.233859 0.3782838  45.44730 68.01910
# 5  12465 Natural 0.20 10.29840  10.284188 7.903941 19.81265 18.763375  74.22164 12.386976 3.3708981  34.93510 80.94928
# 6  12465 Natural 0.25 10.71812  12.307514 7.830025 23.30090 18.352456  80.28088 10.776519 3.7632441  32.56902 76.12610


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample   group  rad  Acetate Propionate Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2   12465 Natural 0.05 17.42657  22.169283 0.000000 41.71924 13.727015 132.95744  2.177403 1.1463411  18.89142 57.41581
# 3   12465 Natural 0.10 13.12582   8.803963 7.594442 16.06896  6.795372 113.18799  3.615191 0.8336420  50.37943 53.92664
# 4   12465 Natural 0.15 10.97779   6.413177 7.816726 14.71928  8.846629  80.91578 11.233859 0.3782838  45.44730 68.01910
# 5   12465 Natural 0.20 10.29840  10.284188 7.903941 19.81265 18.763375  74.22164 12.386976 3.3708981  34.93510 80.94928
# 6   12465 Natural 0.25 10.71812  12.307514 7.830025 23.30090 18.352456  80.28088 10.776519 3.7632441  32.56902 76.12610
# 21  12469 Natural 0.05 16.49383  21.510944 0.000000 44.65537 15.350703 132.42045  2.199757 1.4182089  17.35084 55.11533

table(df.buffers$group[ which(df.buffers$rad==0.05) ]) # count at single radius only
# Disturbed   Natural 
#        29        55


saveRDS(object = df.buffers, file = "df.buffers--AMI.RDS")

str(df.buffers)
# 'data.frame':	420 obs. of  13 variables:
# $ sample    : chr  "12465" "12465" "12465" "12465" ...
# $ group     : chr  "Natural" "Natural" "Natural" "Natural" ...
# $ rad       : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate   : num  17.4 13.1 11 10.3 10.7 ...
# $ Propionate: num  22.17 8.8 6.41 10.28 12.31 ...
# $ Butyrate  : num  0 7.59 7.82 7.9 7.83 ...
# $ VitB2     : num  41.7 16.1 14.7 19.8 23.3 ...
# $ VitB12    : num  13.73 6.8 8.85 18.76 18.35 ...
# $ VitB6     : num  133 113.2 80.9 74.2 80.3 ...
# $ VitB9     : num  2.18 3.62 11.23 12.39 10.78 ...
# $ VitK2     : num  1.146 0.834 0.378 3.371 3.763 ...
# $ Glutamate : num  18.9 50.4 45.4 34.9 32.6 ...
# $ Pyruvate  : num  57.4 53.9 68 80.9 76.1 ...

df.buffers.melt <- melt(df.buffers, id.vars = c("sample","group","rad"))
head(df.buffers.melt)
# sample   group  rad variable    value
# 1  12465 Natural 0.05  Acetate 17.42657
# 2  12465 Natural 0.10  Acetate 13.12582
# 3  12465 Natural 0.15  Acetate 10.97779
# 4  12465 Natural 0.20  Acetate 10.29840
# 5  12465 Natural 0.25  Acetate 10.71812
# 6  12469 Natural 0.05  Acetate 16.49383

unique( df.buffers.melt$group ) # "Natural"   "Disturbed"
df.buffers.melt$group <- factor(df.buffers.melt$group, levels = c("Disturbed","Natural"), ordered = TRUE)

unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	4200 obs. of  5 variables:
# $ sample  : chr  "12465" "12465" "12465" "12465" ...
# $ group   : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ rad     : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ variable: Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  17.4 13.1 11 10.3 10.7 ...

hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))

dat <- df.buffers.melt


dat$log10abun <- dat$value # then over-write later


vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}

# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.549  11.496  13.107  14.154  16.150  24.491 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.980   1.061   1.117   1.139   1.208   1.389 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 5.971   8.741  10.550  12.555  12.369  31.518 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7761  0.9416  1.0233  1.0545  1.0923  1.4986 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   4.678   7.351   5.767   7.846   9.477 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.3116  0.6700  0.8663  0.7439  0.8947  0.9767 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 13.21   15.76   20.34   23.31   25.23   56.10 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.121   1.197   1.308   1.336   1.402   1.749 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 5.694   8.376  14.685  13.707  18.617  21.806 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7554  0.9230  1.1669  1.1007  1.2699  1.3386 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 70.19   80.38   85.52  100.77  123.58  163.16 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.846   1.905   1.932   1.990   2.092   2.213 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7572  3.8194 11.0689  8.5074 12.3873 14.7025 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.1208  0.5820  1.0441  0.8241  1.0930  1.1674 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.3050  0.7094  1.0317  1.9301  3.5838  4.8805 
# [1] "log10 values:"
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# -0.51575 -0.14911  0.01355  0.12751  0.55434  0.68846 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 16.85   32.88   35.66   36.67   45.71   55.00 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.226   1.517   1.552   1.546   1.660   1.740 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 49.96   59.18   67.49   69.08   79.45   88.29 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.699   1.772   1.829   1.834   1.900   1.946



head(dat)
# sample   group  rad variable    value log10abun
# 1  12465 Natural 0.05  Acetate 17.42657  1.241212
# 2  12465 Natural 0.10  Acetate 13.12582  1.118127
# 3  12465 Natural 0.15  Acetate 10.97779  1.040515
# 4  12465 Natural 0.20  Acetate 10.29840  1.012770
# 5  12465 Natural 0.25  Acetate 10.71812  1.030119
# 6  12469 Natural 0.05  Acetate 16.49383  1.217322

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value.  y = log10abun
  geom_violin(width = 1.1)+
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(alpha =0.5)+
  facet_wrap(facets = vars(variable), scales = "free_y")
p


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free")
p


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    legend.position="bottom",
    legend.title = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

this_study # "-AMI-"
header     # "-vk-hac-buffers-"

saveRDS(object = dat, file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))


## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]


head(dat)
# sample   group  rad variable    value log10abun
# 1   12465 Natural 0.05  Acetate 17.42657  1.241212
# 6   12469 Natural 0.05  Acetate 16.49383  1.217322
# 11  12471 Natural 0.05  Acetate 15.35255  1.186181
# 16  12473 Natural 0.05  Acetate 17.60781  1.245705
# 21  12475 Natural 0.05  Acetate 18.99181  1.278566
# 26  12477 Natural 0.05  Acetate 18.37200  1.264157

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(height = 0, alpha=0.5)+
  labs(x = NULL, y = "log10 fxn % rel. abun. in 0.05 radius buffer (per vK unit area)") +
  facet_wrap(facets = vars(variable), scales = "free")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-rough-boxplots",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")




## perform sig tests for annotations !!

levels( dat$variable )
# "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"      "Glutamate"  "Pyruvate"

unique(dat$group)
# [1] Natural   Disturbed
# Levels: Disturbed < Natural

names(dat) # "sample"    "group"     "rad"       "variable"  "value"     "log10abun"


y_anotations <- list()
sig_anotations <- list()

# sig converter - Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
sig_symb <- function(x) {
  out <- character(length = length(x))
  for (i in seq_along(x)) {
    if (is.na(x[i])) {
      out[i] <- NA
    } else if (x[i] < 0.001) { 
      out[i] <- "***"
    } else if (x[i] >= 0.001 & x[i] < 0.01) { 
      out[i] <- "**"
    } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
      out[i] <- "*" 
    } else { 
      out[i] <- "ns"
    }
  } # END loop
  return(out)
}



## Acetate
this_var <- "Acetate"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.254, num df = 28, denom df = 54, p-value = 0.4682

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 4.5397, df = 82, p-value = 9.559e-06
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.02789299        Inf
# sample estimates:
#   mean of x mean of y 
# 1.317251  1.273224 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Propionate"
this_var <- "Propionate"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.60336, num df = 28, denom df = 54, p-value = 0.1489

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 7.2317, df = 82, p-value = 1.136e-10
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.03811395        Inf
# sample estimates:
#   mean of x mean of y 
# 1.413238  1.363736 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Butyrate"
this_var <- "Butyrate"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = NaN, num df = 28, denom df = 54, p-value = NA

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 797.5, p-value = 1

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB2"
this_var <- "VitB2"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.21121, num df = 28, denom df = 54, p-value = 3.186e-05

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "two.sided") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 772, p-value = 0.8141

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB12"
this_var <- "VitB12"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.0656, num df = 28, denom df = 54, p-value = 0.8201

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 6.3001, df = 82, p-value = 7.026e-09
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.04297671        Inf
# sample estimates:
#   mean of x mean of y 
# 1.212794  1.154396 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB6"
this_var <- "VitB6"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.3097, num df = 28, denom df = 54, p-value = 0.001246

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1384, p-value = 1.763e-08
# alternative hypothesis: true location shift is greater than 0

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB9"
this_var <- "VitB9"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.3801, num df = 28, denom df = 54, p-value = 0.307

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -4.3606, df = 82, p-value = 1.87e-05
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.05807606
# sample estimates:
#   mean of x mean of y 
# 0.2000129 0.2939134 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitK2"
this_var <- "VitK2"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 2.9184, num df = 28, denom df = 54, p-value = 0.0007279

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 305, p-value = 1.839e-06

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



##   "Glutamate" 
this_var <- "Glutamate" 

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.53378, num df = 28, denom df = 54, p-value = 0.07404

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 3.3178, df = 82, p-value = 0.0006772
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.01524921        Inf
# sample estimates:
#   mean of x mean of y 
# 1.335748  1.305162 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
sig_anotations[[this_var]] <- sig_symb( tt$p.value)
#sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Pyruvate"
this_var <- "Pyruvate"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.304, num df = 28, denom df = 54, p-value = 0.001057

# overall
# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x = x,y = y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1188, p-value = 0.0001217

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)

y_anotations[[this_var]] <- y_anot.all
#sig_anotations[[this_var]] <- sig_symb( tt$p.value)
sig_anotations[[this_var]] <- sig_symb( wt$p.value)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



# Add significance annotation

# all
annotation_df.all <- data.frame(
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2","Glutamate", "Pyruvate"),
  start=rep(1, times=10), end = rep(2, times=10),
  y = c( y_anotations[["Acetate"]],
         y_anotations[["Propionate"]],
         y_anotations[["Butyrate"]],
         y_anotations[["VitB2"]],
         y_anotations[["VitB12"]],
         y_anotations[["VitB6"]],
         y_anotations[["VitB9"]],
         y_anotations[["VitK2"]],
         y_anotations[["Glutamate"]],
         y_anotations[["Pyruvate"]]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Acetate"]],
            sig_anotations[["Propionate"]],
            sig_anotations[["Butyrate"]],
            sig_anotations[["VitB2"]],
            sig_anotations[["VitB12"]],
            sig_anotations[["VitB6"]],
            sig_anotations[["VitB9"]],
            sig_anotations[["VitK2"]],
            sig_anotations[["Glutamate"]],
            sig_anotations[["Pyruvate"]]
  ),
  stringsAsFactors = FALSE)

str(annotation_df.all)

# ensure no conflict in factor levels
annotation_df.all$variable <- factor(annotation_df.all$variable, 
                                     levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2","Glutamate", "Pyruvate"),
                                     ordered = TRUE)

sel <- which(annotation_df.all$label == "ns")
annotation_df.all$label[sel] <- NA
annotation_df.all$start[sel] <- NA
annotation_df.all$end[sel] <- NA
annotation_df.all$y[sel] <- NA

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(height = 0, alpha=0.5)+
  
  # all
  geom_text(data=annotation_df.all, mapping = aes(x = 1.5, y = y, label = label), size = 5, vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=annotation_df.all, mapping = aes(x = start, y = y, xend = end, yend = y) )+
  
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function log10 % rel. abun. in 0.05 radius vK buffer zones") +
  theme(
    legend.position="bottom",
    strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.8))
  )
p

this_study # "-AMI-"
header     # "-vk-hac-buffers-"

dev.print(tiff, file = paste0(workdir,"/plots/","All-Env-vkBuffers-log1oabun-boxplots-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



## Butyrate only at radius 0.1 ??

#dat <- readRDS(file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))

sel <- which( dat$rad == 0.1 & dat$variable == "Butyrate")

dat <- dat[sel, ]



## "Butyrate"
this_var <- "Butyrate"

x=filter(dat, group == "Disturbed" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group == "Natural" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+ #   y = value
  geom_boxplot()+
  geom_jitter(height = 0, alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.85991, num df = 28, denom df = 54, p-value = 0.6769

# overall
tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -5.0531, df = 82, p-value = 1.296e-06
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.07089537
# sample estimates:
#   mean of x mean of y 
# 0.7073345 0.8130271 

# wt <- wilcox.test(x = x,y = y, alternative = "less") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

y_anot.all <- 0.02*(max(x,y) - min(x,y)) + max(x,y)
sig_symb( tt$p.value) # "***"


# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(height = 0, alpha=0.5)+
  
  geom_text(aes(x = 1.5, y = y_anot.all, label = sig_symb( tt$p.value)), size = 5, vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(aes(x = 1, y = y_anot.all, xend = 2, yend = y_anot.all) )+
  
  theme_classic() +
  labs(x = NULL, y = "Function log10 % rel. abun.\nin 0.10 radius vK buffer zone") +
  theme(
    legend.position="bottom",
    strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.8))
  )
p

this_study # "-AMI-"
header     # "-vk-hac-buffers-"

dev.print(tiff, file = paste0(workdir,"/plots/","Butyrate-only-Rad-0_1--Env-vkBuffers-log1oabun-boxplots-",this_study,header,"-v9.tiff"), width = 7, height = 6, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### AMI - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

# Determines weighted coordinates in each sample

phy <- readRDS("phy-phyloseq-object-AMI.RDS")

this_study <- "-AMI-"
header <- "-vk-coords-wtmean-"

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-AMI.RDS")

# use cleaned version
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

# from earlier
str(vk.samp.noNAs.long)
# 'data.frame':	1558284 obs. of  8 variables:
# $ sample   : chr  "12465" "12465" "12465" "12465" ...
# $ fxn_id   : Factor w/ 19093 levels "fxn_10","fxn_100",..: 6152 11143 14787 15443 16271 17690 18441 1 583 1103 ...
# $ abun     : num  2.29e-05 5.28e-05 3.15e-05 1.64e-03 5.02e-06 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" "12465 (Natural)" ...

# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) # 1025800       5

head(spdf)
# sample   group      OC_x      HC_y         abun
# 2  12465 Natural 0.8571429 0.7142857 2.288676e-05
# 3  12465 Natural 0.9285714 0.8571429 5.276470e-05
# 4  12465 Natural 0.9285714 0.8571429 3.152034e-05
# 5  12465 Natural 0.6000000 1.8000000 1.636470e-03
# 6  12465 Natural 1.0000000 0.4285714 5.017155e-06
# 7  12465 Natural 0.6666667 2.3333333 3.418492e-03

temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 84

hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000010 0.0000415 0.0002763 0.0049774 0.0027449 1.0768881

add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.89   59.97   60.74   60.78   61.65   62.84 

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"


get_wtmean_vk_coords_env <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)

  spdf_in <- spdf_in[sel, ]

  subsel <- list()

  # lipids
  subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  # proteins
  subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # amino_sugars
  subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # carbs - Carbohydrates
  subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  # cond_aromtx - Condensed aromatics
  subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  # lignin
  subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  # tannins
  subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)

  vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")

  # establish blank template data frame
  df_out <- data.frame(sample = NA,group = NA,no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)

  for (v in 1:length(vars)) {
    #v<-1
    this_var <- vars[v]
    if (length(subsel[[ this_var ]]) > 0) {
      df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
    } else {
      df_out$wtmean_OC_x[v] <- NA
      df_out$wtmean_HC_y[v] <- NA
    }
    print(paste0("completed ",v))
  }

  df_out$sample <- this_samp
  df_out$group <- unique(spdf_in$group)
  #df_out$sex <- unique(spdf_in$sex)

  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords_env( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 84
out[[1]]
# sample   group no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1  12465 Natural     104       lipids  0.08809081   1.6592823
# 2  12465 Natural     573     proteins  0.38830136   1.7192978
# 3  12465 Natural     649 amino_sugars  0.62651445   1.6578977
# 4  12465 Natural    1734        carbs  0.87321181   1.7127491
# 5  12465 Natural     150  cond_aromtx  0.12801555   0.9044642
# 6  12465 Natural    2121       lignin  0.49608229   1.2673227
# 7  12465 Natural    3209      tannins  0.80677766   1.3146600

names(out[[1]])
# "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 588 6

str(df.wtcoords)
# 'data.frame':	588 obs. of  6 variables:
# $ sample     : chr  "12465" "12465" "12465" "12465" ...
# $ group      : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ no_fxns    : int  104 573 649 1734 150 2121 3209 97 526 639 ...
# $ variable   : chr  "lipids" "proteins" "amino_sugars" "carbs" ...
# $ wtmean_OC_x: num  0.0881 0.3883 0.6265 0.8732 0.128 ...
# $ wtmean_HC_y: num  1.659 1.719 1.658 1.713 0.904 ...

saveRDS(object = df.wtcoords, file = "df.wtcoords--AMI.RDS")

#df.wtcoords <- readRDS("df.wtcoords--AMI.RDS")

dat <- df.wtcoords

unique(dat$variable)
# "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"
dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)

dat$group

table(dat$group)
# Disturbed   Natural
# 203       385

table(dat$group)/length(levels(dat$variable))
# Disturbed   Natural
# 29        55


dim(dat) # 588   6

ok <- complete.cases(dat)
sel.ok <- which(ok==TRUE) # qty 588


names(dat) # "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

cols = c("Disturbed" = "#f46d43",  "Natural" = "#66c2a5") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9


p <- ggplot(data = dat, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  
  geom_density_2d(contour_var = "ndensity", bins = 3, alpha=.5) + # , alpha=.5

  geom_point()+ # alpha=0.5
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+

  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 2,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Sample\ntype"))

p

this_study # "-AMI-"
header # "-vk-coords-wtmean-"

grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9d-b.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat, file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9.RDS"))



## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat) # "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

levels(dat$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)    
# samps.class  1 0.0021562 0.19215 19.503  0.001 ***
# Residual    82 0.0090654 0.80785                  
# Total       83 0.0112215 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0000013 1.3340e-06 0.0231    999  0.887
# Residuals 82 0.0047244 5.7615e-05



## "Proteins"
this_var <- "Proteins"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)    
# samps.class  1 0.00019932 0.11205 10.347  0.001 ***
# Residual    82 0.00157957 0.88795                  
# Total       83 0.00177888 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)   
# Groups     1 0.00009084 9.0843e-05 13.044    999  0.002 **
# Residuals 82 0.00057105 6.9640e-06                        
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)    
# samps.class  1 0.0021699 0.26602 29.719  0.001 ***
# Residual    82 0.0059871 0.73398                  
# Total       83 0.0081570 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)   
# Groups     1 0.00022356 2.2356e-04 8.2994    999  0.008 **
# Residuals 82 0.00220878 2.6936e-05                        
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)    
# samps.class  1 0.00030853 0.15317 14.831  0.001 ***
# Residual    82 0.00170579 0.84683                  
# Total       83 0.00201432 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.00000194 1.9417e-06 0.2067    999  0.671
# Residuals 82 0.00077046 9.3959e-06



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)    
# samps.class  1 0.0036503 0.19481 19.839  0.001 ***
# Residual    82 0.0150878 0.80519                  
# Total       83 0.0187380 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.0003134 3.1344e-04 3.8104    999  0.056 .
# Residuals 82 0.0067452 8.2258e-05                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)  
# samps.class  1 0.00005142 0.06489 5.6898  0.014 *
# Residual    82 0.00074104 0.93511                
# Total       83 0.00079246 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 1.7130e-06 1.7135e-06 0.4848    999  0.506
# Residuals 82 2.8984e-04 3.5347e-06



## "Tannins"
this_var <- "Tannins"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)   
# samps.class  1 0.00010247 0.09133 8.2414  0.002 **
# Residual    82 0.00101956 0.90867                 
# Total       83 0.00112203 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.00002250 2.2495e-05 4.2101    999  0.039 *
# Residuals 82 0.00043814 5.3431e-06                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



annot_df.ami.coord <- data.frame(
  variable = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
  p_value = c(0.001, 0.001, 0.001, 0.001, 0.001, 0.014, 0.002),
  r_squared = c(0.192, 0.112, 0.266, 0.153, 0.195, 0.065, 0.091),
  npcx = c("left", "left", "right",  "left", "left", "right", "right"),
  npcy = c("top",  "top",  "bottom", "top",  "top",  "top",   "top")
)

annot_df.ami.coord$variable <- factor(annot_df.ami.coord$variable, 
                                            levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                                            ordered = TRUE)
annot_df.ami.coord$displaylabel <- paste0("PERMANOVA\n","P = ",annot_df.ami.coord$p_value,"\nR^2 = ",annot_df.ami.coord$r_squared)

p +
  geom_text_npc(data = annot_df.ami.coord, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 2.5, lineheight=0.9 )


grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9e-b-with-Permanova.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### AMI - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-AMI.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 37335 taxa and 84 samples ]
# sample_data() Sample Data:       [ 84 samples by 92 sample variables ]
# tax_table()   Taxonomy Table:    [ 37335 taxa by 4 taxonomic ranks ]

df.zones <- readRDS("df.zones--AMI.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-AMI.RDS")

samps <- unique(df.zones$sample) # qty 84
# phy <- prune_samples( samples = samps, phy )
# # prune taxa with zero reads
# phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
# phy

head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 84
NonZeroFxns
mean(NonZeroFxns) # 22958.61
sd(NonZeroFxns) # 1171.752
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 18551
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 3326
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 12211.9
sd(fxns_converted) # 552.6064

mean(fxns_relabun_converted) # 60.78408
sd(fxns_relabun_converted) # 1.044727

mean(vks) # 2699.917
sd(vks) # 69.23614


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Sun & Badgley - post-mining ecosystem restoration
# # # # # # # # #
# # # # # # # # #

#### Sun & Badgley - Metadata
#-------------------------

#### read in site info / metadata

meta <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/mgp16379_metadata.xlsx",
                   sheet="COPY for METADATA CSV", range="A1:H16", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)
# 'data.frame':	15 obs. of  7 variables:
# $ sample_name       : chr  "10A" "10B" "10C" "20A" ...
# $ mgrast_id         : chr  "mgl422984" "mgl422987" "mgl422990" "mgl422993" ...
# $ metagenome_id     : chr  "mgm4679669.3" "mgm4679662.3" "mgm4679664.3" "mgm4679665.3" ...
# $ metagenome_name   : chr  "10A.fastq" "10B.fastq" "10C.fastq" "20A.fastq" ...
# $ investigation_type: chr  "metagenome" "metagenome" "metagenome" "metagenome" ...
# $ seq_meth          : chr  "pyrosequencing" "pyrosequencing" "pyrosequencing" "pyrosequencing" ...
# $ file_name         : chr  "10A.fastq" "10B.fastq" "10C.fastq" "20A.fastq" ...
# $ age               : chr  "12" "12" "12" "22" ...

row.names(meta) <- meta$metagenome_id


dim(meta) # 15  8

unique(meta$age)
# "12" "22" "31" "6"  "UM"

table(meta$age)
# 12 22 31  6 UM 
# 3  3  3  3  3 

#-------------------------

#### Sun & Badgley - read in superfocus - fxn potential outputs
#-------------------------

sampid <- meta$metagenome_id
  
length(sampid) # 15
sampid
# [1] "mgm4679669.3" "mgm4679662.3" "mgm4679664.3" "mgm4679665.3" "mgm4679666.3" "mgm4679658.3" "mgm4679660.3" "mgm4679659.3"
# [9] "mgm4679672.3" "mgm4679663.3" "mgm4679667.3" "mgm4679670.3" "mgm4679661.3" "mgm4679668.3" "mgm4679671.3"

meta$sample_name
# "10A" "10B" "10C" "20A" "20B" "20C" "30A" "30B" "30C" "5A"  "5B"  "5C"  "UMA" "UMB" "UMC"

names(sampid) <- meta$sample_name

sampid <- sort(sampid)

superfocus_out_dir <- "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus"

list.dirs(superfocus_out_dir)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus"                            
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679658.3"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679659.3"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679660.3"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679661.3"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679662.3"
# [7] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679663.3"
# [8] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679664.3"
# [9] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679665.3"
# [10] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679666.3"
# [11] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679667.3"
# [12] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679668.3"
# [13] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679669.3"
# [14] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679670.3"
# [15] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679671.3"
# [16] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679672.3"

( results_dirs <- list.dirs(superfocus_out_dir)[-1] )
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679658.3"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679659.3"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679660.3"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679661.3"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679662.3"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679663.3"
# [7] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679664.3"
# [8] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679665.3"
# [9] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679666.3"
# [10] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679667.3"
# [11] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679668.3"
# [12] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679669.3"
# [13] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679670.3"
# [14] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679671.3"
# [15] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_mgm4679672.3"

# check identical order
identical(as.character(sampid), gsub(pattern = "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/sunbad-resto/superfocus/superfocus_out_", replacement = "", x = results_dirs) )
# TRUE

# collate results into a long-format table

sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)

for (i in 1:length(sampid)) {
  #i<-1
  this_samp <- sampid[i]
  sel.folder <- grep(pattern = this_samp, x = results_dirs)
  this_folder <- results_dirs[sel.folder]
  
  #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)
  
  tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
  # names(tab)
  # [1] "Subsystem.Level.1"                                                   
  # [2] "Subsystem.Level.2"                                                   
  # [3] "Subsystem.Level.3"                                                   
  # [4] "Function"                                                            
  # [5] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq"  
  # [6] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq.."
  
  tab$sampid <- this_samp
  names(tab)
  
  tab <- tab[,c(7,1,2,3,4,6)]
  names(tab) <- names(sfx.long)
  
  sfx.long <- rbind(sfx.long,tab)
  
  print(paste0("completed ",i," - sample ID: ",sampid[i]))
}

head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
head(sfx.long)
#       sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 2 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 5.233125e-04
# 3 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 2.012740e-05
# 4 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.363632e-04
# 5 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.975162e-05
# 6 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 3.371340e-03
# 7 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 6.038221e-05

sfx.long$full_fxn_tax <- paste0(sfx.long$subsys_L1,"___", sfx.long$subsys_L2,"___", sfx.long$subsys_L3,"___", sfx.long$fxn)
head(sfx.long)
# sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# 2 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 5.233125e-04
# 3 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 2.012740e-05
# 4 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.363632e-04
# 5 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.975162e-05
# 6 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 3.371340e-03
# 7 mgm4679658.3 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 6.038221e-05
# full_fxn_tax
# 2                                         Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase
# 3                                                   Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase
# 4 Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)
# 5                       Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)
# 6                                       Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)
# 7                                    Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 30125    16

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 30125
length(unique(full_fxn_names)) # 30125

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 


tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)
# full_fxn_tax mgm4679658.3 mgm4679659.3 mgm4679660.3 mgm4679661.3
# 1                                         Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase 5.233125e-04 5.972645e-05 1.551835e-04 1.195234e-04
# 2                                                   Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_isomerase 2.012740e-05 0.000000e+00 0.000000e+00 0.000000e+00
# 3 Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 1.363632e-04 2.526888e-05 4.877197e-05 2.988084e-05
# 4                       Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.975162e-05 1.148586e-05 6.650723e-05 8.964252e-05
# 5                                       Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8) 3.371340e-03 2.216770e-03 2.194738e-03 1.942255e-03
# 6                                    Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 6.038221e-05 2.297171e-05 6.650723e-05 3.486098e-05
# mgm4679662.3 mgm4679663.3 mgm4679664.3 mgm4679665.3 mgm4679666.3 mgm4679667.3 mgm4679668.3 mgm4679669.3 mgm4679670.3 mgm4679671.3 mgm4679672.3
# 1 2.501741e-04 5.509386e-04 3.930307e-04 4.019769e-04 4.197880e-04 3.839742e-04 2.623760e-04 6.503003e-04 5.016445e-04 3.828717e-04 5.379590e-04
# 2 2.204177e-05 1.756616e-05 0.000000e+00 3.122151e-05 2.332155e-05 1.777658e-05 0.000000e+00 0.000000e+00 1.254111e-05 1.002282e-05 0.000000e+00
# 3 1.146172e-04 1.250163e-04 9.707384e-05 1.951344e-05 5.597173e-05 1.315467e-04 1.301038e-05 9.024576e-05 1.964774e-04 1.182693e-04 9.047492e-05
# 4 2.204177e-05 7.984617e-05 8.286791e-05 5.854033e-05 2.332155e-05 1.066595e-04 5.637832e-05 1.681048e-04 0.000000e+00 3.006846e-05 3.667902e-05
# 5 2.182136e-03 2.626939e-03 2.379493e-03 2.566018e-03 2.658657e-03 3.030907e-03 2.049135e-03 2.441944e-03 2.696339e-03 2.495682e-03 2.445268e-03
# 6 2.204177e-05 8.783079e-05 1.775741e-05 3.902689e-05 2.332155e-05 0.000000e+00 4.336794e-05 4.866193e-05 3.657824e-05 2.004564e-05 2.445268e-05

names(sfx.wide)
# [1] "full_fxn_tax" "mgm4679658.3" "mgm4679659.3" "mgm4679660.3" "mgm4679661.3" "mgm4679662.3" "mgm4679663.3" "mgm4679664.3" "mgm4679665.3" "mgm4679666.3" "mgm4679667.3" "mgm4679668.3"
# [13] "mgm4679669.3" "mgm4679670.3" "mgm4679671.3" "mgm4679672.3"

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide)
# [1] "mgm4679658.3" "mgm4679659.3" "mgm4679660.3" "mgm4679661.3" "mgm4679662.3" "mgm4679663.3" "mgm4679664.3" "mgm4679665.3" "mgm4679666.3" "mgm4679667.3" "mgm4679668.3" "mgm4679669.3"
# [13] "mgm4679670.3" "mgm4679671.3" "mgm4679672.3"

head(sampid)
# 20C            30B            30A            UMA            10B             5A 
# "mgm4679658.3" "mgm4679659.3" "mgm4679660.3" "mgm4679661.3" "mgm4679662.3" "mgm4679663.3" 

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # TRUE

names(sfx.wide) <- names(sampid)



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 30125     4

length(unique(tax.fxn$subsys_L1)) # 35
length(unique(tax.fxn$subsys_L2)) # 193
length(unique(tax.fxn$subsys_L3)) # 1244
length(unique(tax.fxn$fxn)) # 14387

sel.dup <- which(duplicated.default(tax.fxn$fxn)==TRUE)
head(tax.fxn$fxn[sel.dup])


#-------------------------

#### Sun & Badgley - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 30125     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 30125  15

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30125 taxa and 15 samples ]
# tax_table()   Taxonomy Table:    [ 30125 taxa by 4 taxonomic ranks ]

sample_names(phy)
# "20C" "30B" "30A" "UMA" "10B" "5A"  "10C" "20A" "20B" "5B"  "UMB" "10A" "5C"  "UMC" "30C"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta) # 15  8
identical( row.names(meta), sample_names(phy) ) # FALSE

identical( names(sampid), sample_names(phy) ) # TRUE

meta2 <- meta[ as.character(sampid) , ]
row.names(meta2) <- meta2$sample_name

identical( row.names(meta2), sample_names(phy) ) # TRUE

SAMP <- sample_data(meta2)

#phy@sam_data <- SAMP


### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30125 taxa and 15 samples ]
# sample_data() Sample Data:       [ 15 samples by 8 sample variables ]
# tax_table()   Taxonomy Table:    [ 30125 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#   subsys_L1                     subsys_L2 subsys_L3            
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn                                                                            
# fxn_1 "2-methylaconitate_cis-trans_isomerase"                                        
# fxn_2 "2-methylaconitate_isomerase"                                                  
# fxn_3 "2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)"
# fxn_4 "2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)"                      
# fxn_5 "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                      
# fxn_6 "4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 

getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-sunbad-resto.RDS")

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")


#-------------------------

#### Sun & Badgley - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 30125    4



get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)
        
        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1
          
          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else
        
      } # END else
      
      coefs[[c]] <- coef
      
    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  
  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      

    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Sunbad-resto-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%  #
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()

  



time.start #  "2023-04-21 20:52:46 ACST"
time.finish # "2023-04-21 23:16:05 ACST"

# 2+ hrs

## assemble results

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Sunbad-resto-R-working-files"

# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# total results written to disk 
dim(df.tax) # 30125     4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 36270     9

table(df.out$matching_method)
# EC number      Exact hierachy match Matched Reactions aliases    Matched Reactions name     Matched Subsytem role            No match found 
# 13716                      4803                       105                       712                       873                     16061 

100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 44.28178 % No match found

100-44.28178
# 55.718 % match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.608   0.796   0.845   1.048   3.333   14135

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.238   1.400   1.410   1.569   4.000   14135 

table(df.out.distil$matching_method)
# EC number                                  EC number;Exact hierachy match 
# 9639                                                             956 
# EC number;Exact hierachy match;Matched Reactions name            EC number;Exact hierachy match;Matched Subsytem role 
# 2                                                               6 
# EC number;Exact hierachy match;No match found                             EC number;Matched Reactions aliases 
# 29                                                              12 
# EC number;Matched Reactions name                 EC number;Matched Reactions name;No match found 
# 54                                                               4 
# EC number;Matched Subsytem role                  EC number;Matched Subsytem role;No match found 
# 123                                                               7 
# EC number;No match found                                            Exact hierachy match 
# 985                                                            2863 
# Exact hierachy match;Matched Reactions aliases                     Exact hierachy match;Matched Reactions name 
# 1                                                              16 
# Exact hierachy match;Matched Subsytem role       Exact hierachy match;Matched Subsytem role;No match found 
# 40                                                               4 
# Exact hierachy match;No match found                                       Matched Reactions aliases 
# 380                                                              80 
# Matched Reactions aliases;Matched Reactions name;No match found                        Matched Reactions aliases;No match found 
# 3                                                               9 
# Matched Reactions name                    Matched Reactions name;Matched Subsytem role 
# 602                                                               4 
# Matched Reactions name;No match found                                           Matched Subsytem role 
# 27                                                             594 
# Matched Subsytem role;No match found                                                  No match found 
# 81                                                           13604


100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 37.50758 % No match found

100-37.50758 # i.e. in 62.49242 % of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_sunbad-resto.RDS")

df.out.distil <- readRDS(file = "df.out.distil_sunbad-resto.RDS")

#-------------------------

#### Sun & Badgley - inspect & clean data
#-------------------------

this_study <- "-SunBadgley-"

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")

df.out.distil <- readRDS(file = "df.out.distil_sunbad-resto.RDS")

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 30125     4



phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE

df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 30125    15
df.OTU[1:5, 1:8]
# 20C          30B          30A          UMA          10B           5A          10C          20A
# fxn_1 5.233125e-04 5.972645e-05 1.551835e-04 1.195234e-04 2.501741e-04 5.509386e-04 3.930307e-04 4.019769e-04
# fxn_2 2.012740e-05 0.000000e+00 0.000000e+00 0.000000e+00 2.204177e-05 1.756616e-05 0.000000e+00 3.122151e-05
# fxn_3 1.363632e-04 2.526888e-05 4.877197e-05 2.988084e-05 1.146172e-04 1.250163e-04 9.707384e-05 1.951344e-05
# fxn_4 3.975162e-05 1.148586e-05 6.650723e-05 8.964252e-05 2.204177e-05 7.984617e-05 8.286791e-05 5.854033e-05
# fxn_5 3.371340e-03 2.216770e-03 2.194738e-03 1.942255e-03 2.182136e-03 2.626939e-03 2.379493e-03 2.566018e-03

mean( df.OTU[ , "20C"] ) # 0.003319502
sum( df.OTU[ , "20C"] ) # 100

sample_sums(phy) # all values of 100


identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
# tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.8571429     0.7142857
# fxn_3     0.9285714     0.8571429
# fxn_4     0.9285714     0.8571429
# fxn_5     0.6000000     1.8000000
# fxn_6     1.0000000     0.4285714



dim(df.fxn_vk_coords) # 30125     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.608   0.796   0.845   1.048   3.333   14135
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.238   1.400   1.410   1.569   4.000   14135 

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.8571429     0.7142857
# fxn_3     0.9285714     0.8571429
# fxn_4     0.9285714     0.8571429
# fxn_5     0.6000000     1.8000000
# fxn_6     1.0000000     0.4285714
# fxn_7     0.6666667     2.3333333


## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 15
dim(df.fxn_vk_coords.noNAs) # 15990     2
dim(df.OTU.noNAs) # 15990    15

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# 10A      UMA      30B      UMB      UMC       5B      30A       5A      30C      20A      10B      20C      10C      20B       5C 
# 61.37220 61.65195 61.65362 61.69752 61.86026 61.88909 61.89054 61.92262 62.08157 62.08236 62.09993 62.16566 62.26411 62.27504 62.29131

summary( colSums(df.OTU.noNAs) )
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 61.37   61.78   61.92   61.95   62.13   62.29 

# this reflects that 61-62% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

## PREVIOUSLY for env/soil sample
## this reflects the just >60% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

colSums(df.OTU)
# all values of 100


dim(df.tax) # 30125     4


# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE


dim(df.fxn_vk_coords.noNAs) # 15990     2
dim(df.in) # 15990    15

row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  group=NA, 
                                  samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"],
                         group= rep( phy@sam_data$age[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}


dim(vk.samp.noNAs.long) # 239851      8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok)

str(vk.samp.noNAs.long)
unique(vk.samp.noNAs.long$group) #  "22" "31" "UM" "12" "6"
class(vk.samp.noNAs.long$group) # "character"
vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c("6","12","22","31","UM"), ordered = TRUE)
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	239850 obs. of  8 variables:
# $ sample   : chr  "20C" "20C" "20C" "20C" ...
# $ fxn_id   : Factor w/ 15990 levels "fxn_10","fxn_1000",..: 6212 11061 11807 12660 13412 14114 14824 15467 1 505 ...
# $ abun     : num  2.01e-05 1.36e-04 3.98e-05 3.37e-03 6.04e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel: chr  "20C (22)" "20C (22)" "20C (22)" "20C (22)" ...


saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-Sunbad-resto.RDS")


#vk.samp.noNAs.long <- readRDS(file = "vk.samp.noNAs.long-Sunbad-resto.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
 
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    #legend.text = element_text(angle=25),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "Post-mining forest restoration soils", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust =0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")




## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Sunbad-resto-R-working-files"


## check anomalous high value in Lignin zone in PCaN density plot
#p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405)
length(sel) # 6210
dim(vk.samp.noNAs.long) # 239850      8

head( vk.samp.noNAs.long[ sel, ] )
#     sample  fxn_id         abun abun_type      OC_x     HC_y group samplabel
# 30     20C  fxn_39 0.0000301911   relabun 0.5206427 1.400701    22  20C (22)
# 104    20C fxn_151 0.0000000000   relabun 0.5206427 1.400701    22  20C (22)
# 139    20C fxn_205 0.0000000000   relabun 0.5206427 1.400701    22  20C (22)
# 166    20C fxn_237 0.0000000000   relabun 0.5206427 1.400701    22  20C (22)
# 208    20C fxn_296 0.0000000000   relabun 0.5206427 1.400701    22  20C (22)
# 263    20C fxn_383 0.0021244473   relabun 0.5206427 1.400701    22  20C (22)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.04606002 % per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 61.94652 % per sample



#fxn_39
i<-39
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED
# 1         fxn_39 1 hypothetical protein Matched Reactions name                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427      1.400701


### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 6210


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-Sunbad-resto.RDS")


#-------------------------

#### Sun & Badgley - CPP-class: total rel abun in vk spatial regions
#-------------------------

this_study <- "-SunBadgley-"
header <- "-vk-zones-"

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")

df.out.distil <- readRDS(file = "df.out.distil_sunbad-resto.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Sunbad-resto.RDS")

# used cleaned dataset
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

str(vk.samp.noNAs.long)
# 'data.frame':	233640 obs. of  8 variables:
# $ sample   : chr  "20C" "20C" "20C" "20C" ...
# $ fxn_id   : Factor w/ 15990 levels "fxn_10","fxn_1000",..: 6212 11061 11807 12660 13412 14114 14824 15467 1 505 ...
# $ abun     : num  2.01e-05 1.36e-04 3.98e-05 3.37e-03 6.04e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel: chr  "20C (22)" "20C (22)" "20C (22)" "20C (22)" ...


# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 


spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) #  165763      5

table(spdf$group)
# 6    12    22    31    UM 
# 34503 33124 33330 32194 32612

head(spdf)
# sample group      OC_x      HC_y         abun
# 2    20C    22 0.8571429 0.7142857 2.012740e-05
# 3    20C    22 0.9285714 0.8571429 1.363632e-04
# 4    20C    22 0.9285714 0.8571429 3.975162e-05
# 5    20C    22 0.6000000 1.8000000 3.371340e-03
# 6    20C    22 1.0000000 0.4285714 6.038221e-05
# 7    20C    22 0.6666667 2.3333333 1.368663e-03

#dim(spdf) # 


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 15

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 61.33   61.73   61.88   61.90   62.09   62.25 

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"

## now want to sum total rela abun within vK spatial zones

## to de-duplicate get total relabun within a sample, then mean across samples

sum_relabun_in_vk_zones_env <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  # lipids
  subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  subsel.all <- c(subsel.lipids)
  # proteins
  subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.proteins)
  # amino_sugars
  subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.amino_sugars)
  # carbs - Carbohydrates
  subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  subsel.all <- c(subsel.all, subsel.carbs)
  # cond_aromtx - Condensed aromatics
  subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  subsel.all <- c(subsel.all, subsel.cond_aromtx)
  # lignin
  subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.lignin)
  # tannins
  subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.tannins)
  # Other compounds ...
  
  # methylated
  subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
  # and not already used?
  sel.dup <- which(subsel.methylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.methylated)
  
  # hydrated
  subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
  # and not already used?
  sel.dup <- which(subsel.hydrated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.hydrated)
  
  # demethylated / Oxidised-dehydrogenated
  subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.demethylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.demethylated)
  
  # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
  #subsel.condensed = which(                   spdf_in$OC_x <= 0.97                       & spdf_in$HC_y < 0.75)
  subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
  
  # and not already used?
  sel.dup <- which(subsel.condensed %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.condensed)
  
  # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
  subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.k2_b12 %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
  
  df_out <- data.frame(
    sample = this_samp,
    group = unique(spdf_in$group),
    
    #sex = unique(spdf_in$sex),
    
    lipids = sum(spdf_in$abun[ subsel.lipids ]),
    proteins = sum(spdf_in$abun[ subsel.proteins ]),
    amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
    carbs = sum(spdf_in$abun[ subsel.carbs ]),
    cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
    lignin = sum(spdf_in$abun[ subsel.lignin ]),
    tannins = sum(spdf_in$abun[ subsel.tannins ]),
    # Other compounds ...
    methylated = sum(spdf_in$abun[ subsel.methylated ]),
    hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
    demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
    condensed = sum(spdf_in$abun[ subsel.condensed ]),
    k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
  )
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones_env( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 15
out[[1]]
# sample group    lipids proteins amino_sugars   carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1    20C    22 0.2023508   2.6656     4.506585 9.48928    1.079101 8.254477 16.21936  0.4267019 4.671542      13.2689 0.8907805 0.4501326

names(out[[1]])
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"

df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample group    lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1    20C    22 0.2023508 2.665600     4.506585 9.489280    1.079101 8.254477 16.21936  0.4267019 4.671542     13.26890 0.8907805 0.4501326
# 2    30B    31 0.2918326 2.516996     4.411393 9.261701    1.327611 8.011814 15.94664  0.4074906 4.809269     13.29322 0.8267715 0.5010569
# 3    30A    31 0.2638530 2.564836     4.420543 9.325482    1.191961 8.115693 16.15800  0.4175346 4.678004     13.36511 0.8538597 0.4881222
# 4    UMA    UM 0.2757619 2.617153     4.514712 9.161785    1.250600 8.105057 16.02163  0.4049840 4.708850     13.25461 0.7779362 0.5130331
# 5    10B    12 0.2174048 2.602519     4.523768 9.373880    1.114075 8.223862 16.30381  0.4429146 4.612926     13.22297 0.9494494 0.4673256
# 6     5A     6 0.1935862 2.686387     4.657028 9.725410    1.000716 8.288855 16.00715  0.4495674 4.629322     13.10209 0.6817673 0.4559018

names(df.zones)
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"
names(df.zones[ ,c(3:14)])
# [1] "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"   "hydrated"     "demethylated"
# [11] "condensed"    "k2_b12"

df.zones$all <- apply(X = df.zones[ ,c(3:14)], MARGIN = 1, FUN = sum)

table(df.zones$group)
# 6 12 22 31 UM 
# 3  3  3  3  3

saveRDS(object = df.zones, file = "df.zones--SunBadgley.RDS")

str(df.zones)
# 'data.frame':	15 obs. of  15 variables:
# $ sample      : chr  "20C" "30B" "30A" "UMA" ...
# $ group       : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ lipids      : num  0.202 0.292 0.264 0.276 0.217 ...
# $ proteins    : num  2.67 2.52 2.56 2.62 2.6 ...
# $ amino_sugars: num  4.51 4.41 4.42 4.51 4.52 ...
# $ carbs       : num  9.49 9.26 9.33 9.16 9.37 ...
# $ cond_aromtx : num  1.08 1.33 1.19 1.25 1.11 ...
# $ lignin      : num  8.25 8.01 8.12 8.11 8.22 ...
# $ tannins     : num  16.2 15.9 16.2 16 16.3 ...
# $ methylated  : num  0.427 0.407 0.418 0.405 0.443 ...
# $ hydrated    : num  4.67 4.81 4.68 4.71 4.61 ...
# $ demethylated: num  13.3 13.3 13.4 13.3 13.2 ...
# $ condensed   : num  0.891 0.827 0.854 0.778 0.949 ...
# $ k2_b12      : num  0.45 0.501 0.488 0.513 0.467 ...
# $ all         : num  62.1 61.6 61.8 61.6 62.1 ...


df.zones.melt <- melt(df.zones, id.vars = c("sample","group"))
head(df.zones.melt)
# sample group variable     value
# 1    20C    22   lipids 0.2023508
# 2    30B    31   lipids 0.2918326
# 3    30A    31   lipids 0.2638530
# 4    UMA    UM   lipids 0.2757619
# 5    10B    12   lipids 0.2174048
# 6     5A     6   lipids 0.1935862

unique( df.zones.melt$group )
# [1] 22 31 UM 12 6 
# Levels: 6 < 12 < 22 < 31 < UM

str(df.zones.melt)
# 'data.frame':	195 obs. of  4 variables:
#   $ sample  : chr  "20C" "30B" "30A" "UMA" ...
# $ group   : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ variable: Factor w/ 13 levels "lipids","proteins",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0.202 0.292 0.264 0.276 0.217 ...

hist(df.zones.melt$value)
hist(log10( df.zones.melt$value))

p <- ggplot(data = df.zones.melt, aes(x = group, y = value))+
  geom_boxplot(outlier.shape = NA)+
  #geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  geom_jitter(alpha=0.6)+
  facet_wrap(facets = vars(variable), scales = "free_y")

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=350, compression="lzw",type="cairo")


saveRDS(object = df.zones.melt, file = "df.zones.melt.vkZones-SunBadgley.RDS")


dat <- df.zones.melt


str(dat)


p <- ggplot(data = dat, aes(x = group, y = value))+ # y = value   y = win95abun
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(alpha=0.6)+
  facet_wrap(facets = vars(variable), scales = "free_y")
p



## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"


head(dat)
# sample group variable     value
# 1    20C    22   Lipids 0.2023508
# 2    30B    31   Lipids 0.2918326
# 3    30A    31   Lipids 0.2638530
# 4    UMA    UM   Lipids 0.2757619
# 5    10B    12   Lipids 0.2174048
# 6     5A     6   Lipids 0.1935862

unique(dat$group)
# [1] 22 31 UM 12 6 
# Levels: 6 < 12 < 22 < 31 < UM


dat$Age <- factor(dat$group, levels = c("6","12","22","31","UM"), ordered = TRUE)
dat$Age
# [1] 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12
# [36] 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6 
# [71] UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31
# [106] 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12
# [141] 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6 
# [176] UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31
# Levels: 6 < 12 < 22 < 31 < UM

dat$age_vec <- as.integer( dat$Age ) # convert factor to integer
dat$age_vec
# [1] 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3
# [54] 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3
# [107] 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3
# [160] 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4

# store ages as a character vector
age_vec <- c(1:5)
names(age_vec) <- c("6","12","22","31","UM")

names(dat)
# "sample"   "group"    "variable" "value"    "Age"      "age_vec" 


sig_symb
# function(x) {
#   out <- character(length = length(x))
#   for (i in seq_along(x)) {
#     if (is.na(x[i])) {
#       out[i] <- NA
#     } else if (x[i] < 0.001) { 
#       out[i] <- "***"
#     } else if (x[i] >= 0.001 & x[i] < 0.01) { 
#       out[i] <- "**"
#     } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
#       out[i] <- "*" 
#     } else { 
#       out[i] <- "ns"
#     }
#   } # END loop
#   return(out)
# }
# <bytecode: 0x7fe3b2960ef0>



pos_anotations <- list()
sig_anotations <- list()
Kendall_Tau <- list()


## "Lipids"
this_var <- "Lipids"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = 3.1395, p-value = 0.001692
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.6377872 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Proteins"
this_var <- "Proteins"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -2.4306, p-value = 0.01507
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.4937707

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Proteins")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -1.2153, p-value = 0.2243
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.2468854

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Amino sugars")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "centre", npcy = "top", label = test_result, size = 3 ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("centre", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Carbohydrates"                
this_var <- "Carbohydrates"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -4.2535, p-value = 2.104e-05
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.8640988

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Carbohydrates")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3  , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = 3.4433, p-value = 0.0005746
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.6995085

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%):\nCondensed aromatics")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "top", label = test_result, size = 3  , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Lignin" 
this_var <- "Lignin"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -3.1395, p-value = 0.001692
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.6377872

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lignin")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3  , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Tannins"
this_var <- "Tannins"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -0.30382, p-value = 0.7613
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.06172134

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Tannins")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "bottom", label = test_result, size = 3  ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other methylated\ncompounds"  
this_var <- "Other methylated\ncompounds"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -3.7471, p-value = 0.0001789
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.7612299 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%):\nOther methylated compounds")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "bottom", label = test_result, size = 3  ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other hydrated\ncompounds"
this_var <- "Other hydrated\ncompounds"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
#	Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = 0.81019, p-value = 0.4178
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1645902 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%):\nOther hydrated compounds")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3 , color = "red" ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = 1.0127, p-value = 0.3112
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.2057378 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%):\nOther demethylated compounds")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "bottom", label = test_result, size = 3 ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = 1.5191, p-value = 0.1287
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3086067 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%):\nOther condensed compounds")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "bottom", label = test_result, size = 3 ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = 2.4306, p-value = 0.01507
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.4937707

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%):\nNear VitK2-VitB12 compounds")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "bottom", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "All compounds"
this_var <- "All compounds"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"value"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"value"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "value"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "value"]
# z = -1.8229, p-value = 0.06831
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.370328

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = value))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): All compounds")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "bottom", label = test_result, size = 3 ) # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



# all
annotation_df.all <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  npcx = c( pos_anotations[["Lipids"]][1],
            pos_anotations[["Proteins"]][1],
            pos_anotations[["Amino sugars"]][1],
            pos_anotations[["Carbohydrates"]][1],
            pos_anotations[["Condensed\naromatics"]][1],
            pos_anotations[["Lignin"]][1],
            pos_anotations[["Tannins"]][1],
            pos_anotations[["Other methylated\ncompounds"]][1],
            pos_anotations[["Other hydrated\ncompounds"]][1],
            pos_anotations[["Other demethylated\ncompounds"]][1],
            pos_anotations[["Other condensed\ncompounds"]][1],
            pos_anotations[["Near VitK2-VitB12\ncompounds"]][1],
            pos_anotations[["All compounds"]][1]
  ),
  npcy = c( pos_anotations[["Lipids"]][2],
            pos_anotations[["Proteins"]][2],
            pos_anotations[["Amino sugars"]][2],
            pos_anotations[["Carbohydrates"]][2],
            pos_anotations[["Condensed\naromatics"]][2],
            pos_anotations[["Lignin"]][2],
            pos_anotations[["Tannins"]][2],
            pos_anotations[["Other methylated\ncompounds"]][2],
            pos_anotations[["Other hydrated\ncompounds"]][2],
            pos_anotations[["Other demethylated\ncompounds"]][2],
            pos_anotations[["Other condensed\ncompounds"]][2],
            pos_anotations[["Near VitK2-VitB12\ncompounds"]][2],
            pos_anotations[["All compounds"]][2]
  ),
  tau = c(Kendall_Tau[["Lipids"]],
          Kendall_Tau[["Proteins"]],
          Kendall_Tau[["Amino sugars"]],
          Kendall_Tau[["Carbohydrates"]],
          Kendall_Tau[["Condensed\naromatics"]],
          Kendall_Tau[["Lignin"]],
          Kendall_Tau[["Tannins"]],
          Kendall_Tau[["Other methylated\ncompounds"]],
          Kendall_Tau[["Other hydrated\ncompounds"]],
          Kendall_Tau[["Other demethylated\ncompounds"]],
          Kendall_Tau[["Other condensed\ncompounds"]],
          Kendall_Tau[["Near VitK2-VitB12\ncompounds"]],
          Kendall_Tau[["All compounds"]]
    ),
  siglabel = c(sig_anotations[["Lipids"]],
            sig_anotations[["Proteins"]],
            sig_anotations[["Amino sugars"]],
            sig_anotations[["Carbohydrates"]],
            sig_anotations[["Condensed\naromatics"]],
            sig_anotations[["Lignin"]],
            sig_anotations[["Tannins"]],
            sig_anotations[["Other methylated\ncompounds"]],
            sig_anotations[["Other hydrated\ncompounds"]],
            sig_anotations[["Other demethylated\ncompounds"]],
            sig_anotations[["Other condensed\ncompounds"]],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]],
            sig_anotations[["All compounds"]]
  ), stringsAsFactors = FALSE)
str(annotation_df.all)

# ensure no conflict in factor levels
annotation_df.all$variable <- factor(annotation_df.all$variable, 
                                     levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                                "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                     ordered = TRUE)

annotation_df.all$siglabel #"**"  "*"   "ns"  "***" "***" "**"  "ns"  "***" "ns"  "ns"  "ns"  "*"   "ns" 


str(dat)
# 'data.frame':	195 obs. of  6 variables:
# $ sample  : chr  "20C" "30B" "30A" "UMA" ...
# $ group   : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ variable: Ord.factor w/ 13 levels "Lipids"<"Proteins"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0.202 0.292 0.264 0.276 0.217 ...
# $ Age     : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ age_vec : int  3 4 4 5 2 1 2 3 3 1 ...

paste0("Kendall's tau: ",annotation_df.all$tau[1],annotation_df.all$siglabel[1])

annotation_df.all$displaylabel <- paste0("Kendall's\ntau: ", annotation_df.all$tau, annotation_df.all$siglabel)
annotation_df.all$displaylabel <- gsub(pattern = "ns", replacement = "", x = annotation_df.all$displaylabel)

p <- ggplot(data = dat, aes(x = age_vec, y = value))+ # y = value   y = win95abun
  
  geom_smooth(method="lm")+
  geom_point(alpha=0.5)+ # shape = 1
  
  scale_x_continuous(labels= names(age_vec))+
  
  geom_text_npc(data = annotation_df.all, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 3.25, lineheight=0.9 )+ # size = 3
  
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)+ # scales = "free_y",
  theme_classic() +
  labs(x = "Age (years)", y = "Function % rel. abun. in vK zones") +  # "Function % rel. abun. in vK zones (Winsorized)"
  theme(
    #legend.position="bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-Trendplots-",this_study,header,"-v9.tiff"), width = 24, height = 20, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Sun & Badgley - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25)
#-------------------------

this_study <- "-SunBadgley-"
header <- "-vk-hac-buffers-"

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")

df.out.distil <- readRDS(file = "df.out.distil_sunbad-resto.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Sunbad-resto.RDS")

# used cleaned dataset
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

str(vk.samp.noNAs.long)
# 'data.frame':	233640 obs. of  8 variables:
# $ sample   : chr  "20C" "20C" "20C" "20C" ...
# $ fxn_id   : Factor w/ 15990 levels "fxn_10","fxn_1000",..: 6212 11061 11807 12660 13412 14114 14824 15467 1 505 ...
# $ abun     : num  2.01e-05 1.36e-04 3.98e-05 3.37e-03 6.04e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel: chr  "20C (22)" "20C (22)" "20C (22)" "20C (22)" ...


# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data


spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) #  165763      5

table(spdf$group)
# 6    12    22    31    UM 
# 34503 33124 33330 32194 32612 

head(spdf)
# sample group      OC_x      HC_y         abun
# 2    20C    22 0.8571429 0.7142857 2.012740e-05
# 3    20C    22 0.9285714 0.8571429 1.363632e-04
# 4    20C    22 0.9285714 0.8571429 3.975162e-05
# 5    20C    22 0.6000000 1.8000000 3.371340e-03
# 6    20C    22 1.0000000 0.4285714 6.038221e-05
# 7    20C    22 0.6666667 2.3333333 1.368663e-03

#dim(spdf) # 


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 15

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 61.33   61.73   61.88   61.90   62.09   62.25 

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"
 

## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2
# df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
# df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)

sum_relabun_in_vk_hac_buffers_env <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  #df_out <- data.frame(sample=NA, group=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA )
  df_out <- data.frame(sample=NA, group=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      group = unique(spdf_in$group),
      #sex = unique(spdf_in$sex),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
    )
    #names(temp_out) # "sample"     "group"      "sex"        "rad"        "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"     
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers_env( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 15
out[[1]]
# sample group  rad  Acetate Propionate Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2    20C    22 0.05 20.51877  22.412225 0.000000 39.40585 15.118030 143.75162  1.752246 0.8126323  23.88519 65.41281
# 3    20C    22 0.10 15.69235   8.738809 6.837404 14.37275  7.084358 124.24832  2.681706 1.0448450  48.32207 57.08007
# 4    20C    22 0.15 12.96401   6.868934 6.703042 13.76449  8.689020  87.82770 12.191477 0.4729179  44.10149 64.48203
# 5    20C    22 0.20 11.28212  10.805371 7.590824 20.00646 19.901044  81.51506 12.696264 3.5844011  35.91803 81.99349
# 6    20C    22 0.25 11.47991  12.720300 7.775856 24.74038 18.470811  85.91444 10.789023 3.9069647  34.01327 80.31015


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample group  rad  Acetate Propionate Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2     20C    22 0.05 20.51877  22.412225 0.000000 39.40585 15.118030 143.75162  1.752246 0.8126323  23.88519 65.41281
# 3     20C    22 0.10 15.69235   8.738809 6.837404 14.37275  7.084358 124.24832  2.681706 1.0448450  48.32207 57.08007
# 4     20C    22 0.15 12.96401   6.868934 6.703042 13.76449  8.689020  87.82770 12.191477 0.4729179  44.10149 64.48203
# 5     20C    22 0.20 11.28212  10.805371 7.590824 20.00646 19.901044  81.51506 12.696264 3.5844011  35.91803 81.99349
# 6     20C    22 0.25 11.47991  12.720300 7.775856 24.74038 18.470811  85.91444 10.789023 3.9069647  34.01327 80.31015
# 21    30B    31 0.05 21.22592  22.084075 0.000000 34.35587 13.077681 136.92804  1.460231 0.8561034  22.39011 66.64165

table(df.buffers$group[ which(df.buffers$rad==0.05) ]) # count at single radius only
# 12 22 31  6 UM 
# 3  3  3  3  3


saveRDS(object = df.buffers, file = "df.buffers--SunBadgley.RDS")

str(df.buffers)
# 'data.frame':	75 obs. of  13 variables:
# $ sample    : chr  "20C" "20C" "20C" "20C" ...
# $ group     : chr  "22" "22" "22" "22" ...
# $ rad       : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate   : num  20.5 15.7 13 11.3 11.5 ...
# $ Propionate: num  22.41 8.74 6.87 10.81 12.72 ...
# $ Butyrate  : num  0 6.84 6.7 7.59 7.78 ...
# $ VitB2     : num  39.4 14.4 13.8 20 24.7 ...
# $ VitB12    : num  15.12 7.08 8.69 19.9 18.47 ...
# $ VitB6     : num  143.8 124.2 87.8 81.5 85.9 ...
# $ VitB9     : num  1.75 2.68 12.19 12.7 10.79 ...
# $ VitK2     : num  0.813 1.045 0.473 3.584 3.907 ...
# $ Glutamate : num  23.9 48.3 44.1 35.9 34 ...
# $ Pyruvate  : num  65.4 57.1 64.5 82 80.3 ...

df.buffers.melt <- melt(df.buffers, id.vars = c("sample","group","rad"))
head(df.buffers.melt)
# sample group  rad variable    value
# 1    20C    22 0.05  Acetate 20.51877
# 2    20C    22 0.10  Acetate 15.69235
# 3    20C    22 0.15  Acetate 12.96401
# 4    20C    22 0.20  Acetate 11.28212
# 5    20C    22 0.25  Acetate 11.47991
# 6    30B    31 0.05  Acetate 21.22592

unique( df.buffers.melt$group ) # "22" "31" "UM" "12" "6" 
df.buffers.melt$group <- factor(df.buffers.melt$group, levels = c("6","12","22","31","UM"), ordered = TRUE)

unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	750 obs. of  5 variables:
# $ sample  : chr  "20C" "20C" "20C" "20C" ...
# $ group   : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 3 3 3 3 4 4 4 4 4 ...
# $ rad     : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ variable: Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  20.5 15.7 13 11.3 11.5 ...

hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))
#hist(sqrt( df.buffers.melt$value))

dat <- df.buffers.melt

dat$log10abun <- dat$value # then over-write later

vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}


# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 10.28   11.30   12.40   14.06   15.95   22.27 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.012   1.053   1.093   1.136   1.203   1.348 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 6.125   8.477  10.635  12.168  12.560  23.902 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7871  0.9282  1.0267  1.0463  1.0990  1.3784 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   6.086   7.052   5.710   7.526   8.402 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.4688  0.7843  0.8483  0.7752  0.8766  0.9244 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 12.73   14.00   19.78   21.97   24.74   40.52 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.105   1.146   1.296   1.309   1.393   1.608 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 6.324   8.466  14.959  13.785  18.832  20.789 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.8010  0.9277  1.1749  1.1019  1.2749  1.3178 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 77.05   83.28   86.16  102.86  124.06  149.09 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.887   1.921   1.935   2.000   2.094   2.173 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.452   2.713  10.831   8.109  12.462  13.223 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.1621  0.4334  1.0347  0.7870  1.0956  1.1213 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.3745  0.8019  1.0729  2.0618  3.8818  4.2409 
# [1] "log10 values:"
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# -0.42652 -0.09594  0.03057  0.16526  0.58902  0.62745 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 18.16   33.22   35.75   37.04   45.54   50.56 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.259   1.521   1.553   1.553   1.658   1.704 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.42   62.40   67.18   69.46   80.32   84.02 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.711   1.795   1.827   1.837   1.905   1.924 


head(dat)
# sample group  rad variable    value log10abun
# 1    20C    22 0.05  Acetate 20.51877  1.312151
# 2    20C    22 0.10  Acetate 15.69235  1.195688
# 3    20C    22 0.15  Acetate 12.96401  1.112739
# 4    20C    22 0.20  Acetate 11.28212  1.052391
# 5    20C    22 0.25  Acetate 11.47991  1.059938
# 6    30B    31 0.05  Acetate 21.22592  1.326867


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free")
p



p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    legend.position="bottom",
    legend.title = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

this_study # "-SunBadgley-"
header     # "-vk-hac-buffers-"

saveRDS(object = dat, file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--All-Env-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))


## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]


head(dat)
# sample group  rad variable    value log10abun
# 1     20C    22 0.05  Acetate 20.51877  1.312151
# 6     30B    31 0.05  Acetate 21.22592  1.326867
# 11    30A    31 0.05  Acetate 20.50955  1.311956
# 16    UMA    UM 0.05  Acetate 18.92960  1.277141
# 21    10B    12 0.05  Acetate 19.26176  1.284696
# 26     5A     6 0.05  Acetate 21.60284  1.334511

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(height = 0, alpha=0.5)+
  labs(x = NULL, y = "log10 fxn % rel. abun. in 0.05 radius buffer (per vK unit area)") +
  facet_wrap(facets = vars(variable), scales = "free")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-rough-trends",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")




## perform sig tests for annotations !!

unique( dat$variable )
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

unique(dat$group)
# [1] 22 31 UM 12 6 
# Levels: 6 < 12 < 22 < 31 < UM

dat$Age <- factor(dat$group, levels = c("6","12","22","31","UM"), ordered = TRUE)
dat$Age
# [1] 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31
# [61] 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31 22 31 31 UM 12 6  12 22 22 6  UM 12 6  UM 31
# Levels: 6 < 12 < 22 < 31 < UM

dat$age_vec <- as.integer( dat$Age ) # convert factor to integer
dat$age_vec
# [1] 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3
# [92] 4 4 5 2 1 2 3 3 1 5 2 1 5 4 3 4 4 5 2 1 2 3 3 1 5 2 1 5 4

# store ages as a character vector
age_vec <- c(1:5)
names(age_vec) <- c("6","12","22","31","UM")

names(dat)
# [1] "sample"    "group"     "rad"       "variable"  "value"     "log10abun" "Age"       "age_vec" 


sig_symb
# function(x) {
#   out <- character(length = length(x))
#   for (i in seq_along(x)) {
#     if (is.na(x[i])) {
#       out[i] <- NA
#     } else if (x[i] < 0.001) { 
#       out[i] <- "***"
#     } else if (x[i] >= 0.001 & x[i] < 0.01) { 
#       out[i] <- "**"
#     } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
#       out[i] <- "*" 
#     } else { 
#       out[i] <- "ns"
#     }
#   } # END loop
#   return(out)
# }
# <bytecode: 0x7fe3b2960ef0>



pos_anotations <- list()
sig_anotations <- list()
Kendall_Tau <- list()



## "Acetate"
this_var <- "Acetate"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = -2.7344, p-value = 0.006249
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.5554921 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Propionate"
this_var <- "Propionate"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = 0.60764, p-value = 0.5434
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1234427

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Butyrate"
this_var <- "Butyrate"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# 
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# T = NA, p-value = NA
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
# tau 
#  NA 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "centre", npcy = "bottom", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("centre", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "VitB2"
this_var <- "VitB2"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = -3.1395, p-value = 0.001692
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.6377872 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "bottom", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "VitB12"
this_var <- "VitB12"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = -2.4306, p-value = 0.01507
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.4937707

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "bottom", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "VitB6"
this_var <- "VitB6"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = -3.2408, p-value = 0.001192
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.658361

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "right", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "VitB9"
this_var <- "VitB9"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = 1.9242, p-value = 0.05433
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3909018

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "VitK2" 
this_var <- "VitK2"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = 2.3293, p-value = 0.01984
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.4731969 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Glutamate" 
this_var <- "Glutamate"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = -3.342, p-value = 0.0008317
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.6789347 

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Pyruvate" 
this_var <- "Pyruvate"

sel <- which(dat$variable == this_var)
plot(dat$age_vec[sel], dat[sel ,"log10abun"])

# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec[sel], y = dat[sel ,"log10abun"], method = "kendall")
# Warning message:
# In cor.test.default(x = dat$age_vec[sel], y = dat[sel, "log10abun"],  :
#   Cannot compute exact p-value with ties
ktcor
# Kendall's rank correlation tau
# data:  dat$age_vec[sel] and dat[sel, "log10abun"]
# z = -1.5191, p-value = 0.1287
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.3086067

pval <- ifelse(test = ktcor$p.value < 0.001, yes = paste0("P-value < 0.001"), no = paste0("P-value = ",round(ktcor$p.value,3)) )
test_result <- paste0("Kendall's tau-b ~Age\nz = ",round(ktcor$statistic,3),"; ",pval)

p <- ggplot(data = filter(dat, variable == this_var) , aes(x=age_vec, y = log10abun))+
  geom_point()+
  geom_point(shape = 1)+
  geom_smooth(method="lm")+
  theme_bw()+
  xlab("Age (years)")+ ylab("Fxn rel. abun. (%): Lipids")+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(size = rel(1.1)))+
  scale_x_continuous(labels= names(age_vec))+
  annotate(geom="text_npc", npcx = "left", npcy = "top", label = test_result, size = 3 , color = "red") # , color = "red"
p

dev.print(tiff, file = paste0(workdir,"/plots/","Kendall-tau-b-Age-vs-",this_var,"-sunbad-resto-",header,"-v7.tiff"), width = 8, height = 8, units = "cm", res=600, compression="lzw",type="cairo")

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



# Add significance annotation

# all
annotation_df.all <- data.frame(
  
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
  npcx = c( pos_anotations[["Acetate"]][1],
            pos_anotations[["Propionate"]][1],
            pos_anotations[["Butyrate"]][1],
            pos_anotations[["VitB2"]][1],
            pos_anotations[["VitB12"]][1],
            pos_anotations[["VitB6"]][1],
            pos_anotations[["VitB9"]][1],
            pos_anotations[["VitK2"]][1],
            pos_anotations[["Glutamate"]][1],
            pos_anotations[["Pyruvate"]][1]
            ),
  
  npcy = c( pos_anotations[["Acetate"]][2],
            pos_anotations[["Propionate"]][2],
            pos_anotations[["Butyrate"]][2],
            pos_anotations[["VitB2"]][2],
            pos_anotations[["VitB12"]][2],
            pos_anotations[["VitB6"]][2],
            pos_anotations[["VitB9"]][2],
            pos_anotations[["VitK2"]][2],
            pos_anotations[["Glutamate"]][2],
            pos_anotations[["Pyruvate"]][2]
            ),
  
  tau = c(Kendall_Tau[["Acetate"]],
          Kendall_Tau[["Propionate"]],
          Kendall_Tau[["Butyrate"]],
          Kendall_Tau[["VitB2"]],
          Kendall_Tau[["VitB12"]],
          Kendall_Tau[["VitB6"]],
          Kendall_Tau[["VitB9"]],
          Kendall_Tau[["VitK2"]],
          Kendall_Tau[["Glutamate"]],
          Kendall_Tau[["Pyruvate"]]
          ),
  
  siglabel = c(sig_anotations[["Acetate"]],
               sig_anotations[["Propionate"]],
               sig_anotations[["Butyrate"]],
               sig_anotations[["VitB2"]],
               sig_anotations[["VitB12"]],
               sig_anotations[["VitB6"]],
               sig_anotations[["VitB9"]],
               sig_anotations[["VitK2"]],
               sig_anotations[["Glutamate"]],
               sig_anotations[["Pyruvate"]]
               ),
  stringsAsFactors = FALSE)
str(annotation_df.all)

# ensure no conflict in factor levels
annotation_df.all$variable <- factor(annotation_df.all$variable, 
                                     levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate","Pyruvate"),
                                     ordered = TRUE)

annotation_df.all$siglabel # "**"  "ns"  NA    "**"  "*"   "**"  "ns"  "*"   "***" "ns" 

annotation_df.all$displaylabel <- paste0("Kendall's\ntau: ", annotation_df.all$tau, annotation_df.all$siglabel)
unique(annotation_df.all$displaylabel)
annotation_df.all$displaylabel <- gsub(pattern = "ns", replacement = "", x = annotation_df.all$displaylabel)
annotation_df.all$displaylabel <- gsub(pattern = "NANA", replacement = "N/A", x = annotation_df.all$displaylabel)


str(annotation_df.all)
# 'data.frame':	10 obs. of  6 variables:
# $ variable    : Ord.factor w/ 10 levels "Acetate"<"Propionate"<..: 1 2 3 4 5 6 7 8 9 10
# $ npcx        : chr  "left" "right" "centre" "left" ...
# $ npcy        : chr  "bottom" "top" "bottom" "bottom" ...
# $ tau         : num  -0.555 0.123 NA -0.638 -0.494 -0.658 0.391 0.473 -0.679 -0.309
# $ siglabel    : chr  "**" "ns" NA "**" ...
# $ displaylabel: chr  "Kendall's\ntau: -0.555**" "Kendall's\ntau: 0.123" "Kendall's\ntau: N/A" "Kendall's\ntau: -0.638**" ...


str(dat)
# 'data.frame':	150 obs. of  8 variables:
# $ sample   : chr  "20C" "30B" "30A" "UMA" ...
# $ group    : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ rad      : num  0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 ...
# $ variable : Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value    : num  20.5 21.2 20.5 18.9 19.3 ...
# $ log10abun: num  1.31 1.33 1.31 1.28 1.28 ...
# $ Age      : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ age_vec  : int  3 4 4 5 2 1 2 3 3 1 ...

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = dat, file = "dat-object-HACBuffer_0_05--SunBadgley-v9.RDS")



p <- ggplot(data = dat, aes(x = age_vec, y = log10abun))+ # y = value
  geom_smooth(method="lm")+
  geom_point(alpha=0.5)+ # shape = 1
  
  scale_x_continuous(labels= names(age_vec))+
  
  geom_text_npc(data = annotation_df.all, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 3.25, lineheight=0.9 )+ # size = 3
  
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = "Age (years)", y = "Function log10 % rel. abun. in 0.05 radius vK buffer zones") +
  theme(
    #legend.position="bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    #strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.9))
  )
p

this_study # "-SunBadgley-"
header    #  "-vk-hac-buffers-"

dev.print(tiff, file = paste0(workdir,"/plots/","All-Env-vkBuffers-log1oabun-boxplots-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### Sun & Badgley - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

# Determines weighted coordinates in each sample

this_study <- "-SunBadgley-"
header <- "-vk-coords-wtmean-"

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")

df.out.distil <- readRDS(file = "df.out.distil_sunbad-resto.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Sunbad-resto.RDS")

# used cleaned dataset
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


str(vk.samp.noNAs.long)
# 'data.frame':	233640 obs. of  8 variables:
# $ sample   : chr  "20C" "20C" "20C" "20C" ...
# $ fxn_id   : Factor w/ 15990 levels "fxn_10","fxn_1000",..: 6212 11061 11807 12660 13412 14114 14824 15467 1 505 ...
# $ abun     : num  2.01e-05 1.36e-04 3.98e-05 3.37e-03 6.04e-05 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.857 0.929 0.929 0.6 1 ...
# $ HC_y     : num  0.714 0.857 0.857 1.8 0.429 ...
# $ group    : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel: chr  "20C (22)" "20C (22)" "20C (22)" "20C (22)" ...


# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]

dim(spdf) # 165763      5

head(spdf)
# sample group      OC_x      HC_y         abun
# 2    20C    22 0.8571429 0.7142857 2.012740e-05
# 3    20C    22 0.9285714 0.8571429 1.363632e-04
# 4    20C    22 0.9285714 0.8571429 3.975162e-05
# 5    20C    22 0.6000000 1.8000000 3.371340e-03
# 6    20C    22 1.0000000 0.4285714 6.038221e-05
# 7    20C    22 0.6666667 2.3333333 1.368663e-03

temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 15

hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000022 0.0000627 0.0003742 0.0056014 0.0033075 0.7939269


add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 61.33   61.73   61.88   61.90   62.09   62.25

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"



get_wtmean_vk_coords_env <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)

  spdf_in <- spdf_in[sel, ]

  subsel <- list()

  # lipids
  subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  # proteins
  subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # amino_sugars
  subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # carbs - Carbohydrates
  subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  # cond_aromtx - Condensed aromatics
  subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  # lignin
  subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  # tannins
  subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)

  vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")

  # establish blank template data frame
  df_out <- data.frame(sample = NA,group = NA,no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)

  for (v in 1:length(vars)) {
    #v<-1
    this_var <- vars[v]
    if (length(subsel[[ this_var ]]) > 0) {
      df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
    } else {
      df_out$wtmean_OC_x[v] <- NA
      df_out$wtmean_HC_y[v] <- NA
    }
    print(paste0("completed ",v))
  }

  df_out$sample <- this_samp
  df_out$group <- unique(spdf_in$group)
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords_env( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 15
out[[1]]
# sample group no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1    20C    22     109       lipids  0.09770531   1.6842892
# 2    20C    22     496     proteins  0.38895012   1.7159819
# 3    20C    22     622 amino_sugars  0.62812632   1.6737243
# 4    20C    22    1613        carbs  0.87593122   1.7123885
# 5    20C    22     144  cond_aromtx  0.13641647   0.9245388
# 6    20C    22    2002       lignin  0.49517499   1.2702310
# 7    20C    22    2865      tannins  0.80336246   1.3201894

names(out[[1]])
# "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 105 6

str(df.wtcoords)
# 'data.frame':	105 obs. of  6 variables:
# $ sample     : chr  "20C" "20C" "20C" "20C" ...
# $ group      : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 3 3 3 3 3 3 4 4 4 ...
# $ no_fxns    : int  109 496 622 1613 144 2002 2865 93 482 556 ...
# $ variable   : chr  "lipids" "proteins" "amino_sugars" "carbs" ...
# $ wtmean_OC_x: num  0.0977 0.389 0.6281 0.8759 0.1364 ...
# $ wtmean_HC_y: num  1.684 1.716 1.674 1.712 0.925 ...

saveRDS(object = df.wtcoords, file = "df.wtcoords-SunBad-resto.RDS")


#df.wtcoords <- readRDS("df.wtcoords-SunBad-resto.RDS")


dat <- df.wtcoords

unique(dat$variable)
# "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"
dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)

dat$group
# [1] 22 22 22 22 22 22 22 31 31 31 31 31 31 31 31 31 31 31 31 31 31 UM UM UM UM UM UM UM 12 12 12 12 12 12 12 6  6  6
# [39] 6  6  6  6  12 12 12 12 12 12 12 22 22 22 22 22 22 22 22 22 22 22 22 22 22 6  6  6  6  6  6  6  UM UM UM UM UM UM
# [77] UM 12 12 12 12 12 12 12 6  6  6  6  6  6  6  UM UM UM UM UM UM UM 31 31 31 31 31 31 31
# Levels: 6 < 12 < 22 < 31 < UM

table(dat$group)
# 6 12 22 31 UM
# 21 21 21 21 21

dim(dat) # 105   6

ok <- complete.cases(dat)
sel.ok <- which(ok==TRUE) # qty 105


names(dat) # "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

#cols = c("Young" = "#fc8d62", "Old" = "#66c2a5", "Remnant" = "#8da0cb")

#cols = c("6" = "", "12" = "", "22" = "", "31" = "", "UM" = "")

p <- ggplot(data = dat, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1

  geom_density_2d(contour_var = "ndensity", bins = 3, alpha=.5) + # , alpha=.5

  geom_point()+ # alpha=0.5
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+

  theme(
    legend.position = c(0.9, 0.2), # as fraction of plot dimensions
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 2,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Reveg age\n(years)"))

p

this_study # "-SunBadgley-"
header # "-vk-coords-wtmean-"

grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9d-a.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat, file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9.RDS"))

dim(dat) # 105 6


## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat) # "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"


levels(dat$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)  
# samps.class  4 0.00073620 0.65524 4.7514  0.014 *
# Residual    10 0.00038736 0.34476                
# Total       14 0.00112356 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 9.8360e-06 2.4590e-06 0.1414    999  0.944
# Residuals 10 1.7391e-04 1.7391e-05



## "Proteins"
this_var <- "Proteins"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)   
# samps.class  4 1.1091e-04 0.88373 19.001  0.002 **
# Residual    10 1.4592e-05 0.11627                 
# Total       14 1.2550e-04 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 2.2983e-06 5.7458e-07 0.8123    999  0.546
# Residuals 10 7.0732e-06 7.0732e-07



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)   
# samps.class  4 0.00031140 0.85405 14.629   0.01 **
# Residual    10 0.00005321 0.14595                 
# Total       14 0.00036461 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 5.7604e-06 1.4401e-06 0.5153    999   0.74
# Residuals 10 2.7947e-05 2.7947e-06 



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)   
# samps.class  4 6.2916e-05 0.60708 3.8627  0.008 **
# Residual    10 4.0720e-05 0.39292                 
# Total       14 1.0364e-04 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 1.6533e-06 4.1332e-07 0.1647    999  0.947
# Residuals 10 2.5091e-05 2.5091e-06



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)   
# samps.class  4 0.00141084 0.92464 30.676  0.002 **
# Residual    10 0.00011498 0.07536                 
# Total       14 0.00152581 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 1.3816e-05 3.4541e-06 0.5661    999  0.686
# Residuals 10 6.1016e-05 6.1016e-06 



## "Lignin"
this_var <- "Lignin"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)    
# samps.class  4 3.9943e-05 0.84079 13.203  0.001 ***
# Residual    10 7.5630e-06 0.15921                  
# Total       14 4.7506e-05 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 1.8266e-07 4.5666e-08 0.1813    999  0.942
# Residuals 10 2.5186e-06 2.5186e-07



## "Tannins"
this_var <- "Tannins"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs    R2      F Pr(>F)    
# samps.class  4 8.8796e-05 0.795 9.6951  0.001 ***
# Residual    10 2.2897e-05 0.205                  
# Total       14 1.1169e-04 1.000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     4 3.7471e-06 9.3676e-07 1.0126    999   0.46
# Residuals 10 9.2512e-06 9.2512e-07 



annot_df.sunbad.coord <- data.frame(
  variable = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
  p_value = c(0.014, 0.002, 0.01, 0.008, 0.002, 0.001, 0.001),
  r_squared = c(0.655, 0.884, 0.854, 0.607, 0.925, 0.841, 0.795),
  npcx = c("left", "left", "right",  "left", "right",  "left", "right"),
  npcy = c("top",  "top",  "bottom", "top",  "bottom", "centre",   "top")
)

annot_df.sunbad.coord$variable <- factor(annot_df.sunbad.coord$variable, 
                                      levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                                      ordered = TRUE)
annot_df.sunbad.coord$displaylabel <- paste0("PERMANOVA\n","P = ",annot_df.sunbad.coord$p_value,"\nR^2 = ",annot_df.sunbad.coord$r_squared)


p +
  geom_text_npc(data = annot_df.sunbad.coord, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 2.5, lineheight=0.9 )


grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9e-a-with-Permanova.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Sun & Badgley - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-sunbad-resto.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30125 taxa and 15 samples ]
# sample_data() Sample Data:       [ 15 samples by 8 sample variables ]
# tax_table()   Taxonomy Table:    [ 30125 taxa by 4 taxonomic ranks ]

df.zones <- readRDS("df.zones--SunBadgley.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Sunbad-resto.RDS")

samps <- unique(df.zones$sample) # qty 15
# phy <- prune_samples( samples = samps, phy )
# # prune taxa with zero reads
# phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
# phy

head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 15
NonZeroFxns
# 20C   30B   30A   UMA   10B    5A   10C   20A   20B    5B   UMB   10A    5C   UMC   30C 
# 20971 19041 20134 19760 20395 21895 19969 20498 19793 21392 19776 20897 20875 20085 19438 
mean(NonZeroFxns) # 20327.93
sd(NonZeroFxns) # 767.4159
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 15576
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 3076
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 11050.87
sd(fxns_converted) # 333.8511

mean(fxns_relabun_converted) # 61.90046
sd(fxns_relabun_converted) # 0.2700508

mean(vks) # 2536.8
sd(vks) # 53.97645


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Jie et al 2017 - Atherosclerotic cardiovascular disease vs normal gut metagenomes
# # # # # # # # #
# # # # # # # # #

#### Jie ACVD - prepare for data download from SRA
#-------------------------

# Jie, Z. et al. The gut microbiome in atherosclerotic cardiovascular disease. 
# Nature Communications 8, 845, doi:10.1038/s41467-017-00900-1 (2017).

 
# Jie-2017-Metadata_Supp_data1_41467_2017_900_MOESM3_ESM.xlsx
# A3:BC408
# Sheet 1

subjects <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Jie-2017-Cardiovascular/Jie-2017-Metadata_Supp_data1_41467_2017_900_MOESM3_ESM.xlsx",
                       sheet=1, range="A3:BC408", col_names = TRUE)
subjects <- as.data.frame(subjects)
str(subjects)
length(unique( subjects$`Sample ID` )) # 405
table(subjects$`ACVD status`)
# 0   1    # 0: Normal   1:ACVD
# 187 218

187+218 # 405

table(subjects$`Coronary type`)
# AMI              NA   Stable angina Unstable angina 
# 5             187             205               8 



# Jie-2017-Cardiovascular-SRA-Metadata--SraRunTable.xlsx
# A1:AV406
# Sheet 1
# 405 samples/Runs

meta <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Jie-2017-Cardiovascular/Jie-2017-Cardiovascular-SRA-Metadata--SraRunTable.xlsx",
                   sheet=1, range="A1:AV406", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)
# 'data.frame':	405 obs. of  48 variables:

names(meta)
# [1] "Run"                                      "Assay Type"                               "BioProject"                              
# [4] "BioSample"                                "Center Name"                              "Collection_Date"                         
# [7] "Consent"                                  "ENA-FIRST-PUBLIC (run)"                   "ENA-FIRST-PUBLIC"                        
# [10] "ENA-LAST-UPDATE (run)"                    "ENA-LAST-UPDATE"                          "environment_(biome)"                     
# [13] "environment_(feature)"                    "environment_(material)"                   "Experiment"                              
# [16] "External_Id"                              "geo_loc_name_country"                     "geo_loc_name_country_continent"          
# [19] "geographic_location_(country_and/or_sea)" "geographic_location_(latitude)"           "geographic_location_(longitude)"         
# [22] "human_gut_environmental_package"          "INSDC_center_alias"                       "INSDC_center_name"                       
# [25] "INSDC_first_public"                       "INSDC_last_update"                        "INSDC_status"                            
# [28] "Instrument"                               "Investigation_type"                       "Library Name"                            
# [31] "LibraryLayout"                            "LibrarySelection"                         "LibrarySource"                           
# [34] "Organism"                                 "Platform"                                 "project_name"                            
# [37] "ReleaseDate"                              "Sample Name"                              "Sample_name"                             
# [40] "Sequencing_method"                        "SRA Study"                                "Submitter_Id"                            
# [43] "DATASTORE filetype"                       "DATASTORE provider"                       "DATASTORE region"                        
# [46] "AvgSpotLen"                               "Bases"                                    "Bytes" 

length(meta$Run) # 405

length(which(subjects$`Sample ID` %in% meta$Submitter_Id)) # 405


row.names(meta) <- meta$Run
names(meta)

head(sort(meta$Run)) # "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" "ERR2017415" "ERR2017416"

# add Diagnosis to metadata
meta$acvd_status <- NA
meta$coronary_type <- NA

for (i in 1:length(meta$Submitter_Id)) {
  #i<-1
  this_submitter_id <- meta$Submitter_Id[i]
  sel <- which(subjects$`Sample ID`==this_submitter_id)
  meta$acvd_status[i] <- subjects$`ACVD status`[sel]
  meta$coronary_type[i] <- subjects$`Coronary type`[sel]
  
  print(paste0(i,": ",this_submitter_id," == ",subjects$`Sample ID`[sel]))
}

head(meta)

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

write.csv(x = meta[ sort(meta$Run), c("Run","Submitter_Id","acvd_status","coronary_type") ], file = "Jie-ACVD-sra-run-list.txt", quote = FALSE, row.names = FALSE)

# # Download from command line using format:
# fastq-dump --outdir /scratch/user/lidd0026/gmrepo_1_meta_raw --skip-technical --readids --read-filter pass --dumpbase --clip --split-3 ERR719096

write.csv(x = paste0("fastq-dump --outdir /scratch/user/lidd0026/jie-acvd/jacvd_1_meta_raw --skip-technical --readids --read-filter pass --dumpbase --clip --split-3 ",sort(meta$Run)), file = "jie-acvd-sra-runs.txt", quote = FALSE, row.names = FALSE)


## a number of SRA Runs were not available for download
# a number of fastq files failed to download. After investigating, it seems all the runs marked in SRA Run Table as DATASTORE filetype == 'fastq' (but not 'fastq,sra') were not available on SRA or ENA. Not in ENA ftp Submitter files, or stored fastq files.
# Checked at ENA ftp site as follows:
#   https://ena-docs.readthedocs.io/en/latest/retrieval/file-download/sra-ftp-structure.html
# The root address of the SRA FTP server is:
#   ftp://ftp.sra.ebi.ac.uk/vol1/
#   This server contains all raw read data submitted to ENA and INSDC, as well a subset of analysis data.
# 
# ftp.sra.ebi.ac.uk
# Name: anonymous
# Password: <blank>
#   
#   ENA-SRA ftp
# --Fastq files:
#   http://ftp.sra.ebi.ac.uk/vol1/fastq/<accession-prefix>/<00-last-digit-of-full-accession>/<full-accession>/
#   e.g.
# ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR164/007/ERR164407/
#   vol1/fastq/ERR201/005/ERR201/005/ERR2017425 - not available
# 
# --Submitted files
# ftp://ftp.sra.ebi.ac.uk/vol1/run/<accession-prefix>/<full-accession>/
#   e.g.
# ftp://ftp.sra.ebi.ac.uk/vol1/run/ER201/ERR2017425/ - not available


# identify files that were NOT available
unique(meta$`DATASTORE filetype`) # "fastq,sra" "sra,fastq" "fastq"     NA  
unique(meta$`DATASTORE provider`) # "ena,s3" "s3,ena" "ena"    NA 

table( meta$`DATASTORE provider` )
# ena ena,s3 s3,ena 
# 21    206    177

# sel <- which(meta$`DATASTORE provider` == "ena")
# length(sel) # 21

table( meta$`DATASTORE filetype`, useNA = "ifany" )
# fastq fastq,sra sra,fastq      <NA> 
#   19       211       174         1 

sel1 <- which(meta$`DATASTORE filetype` %in% c("fastq",NA))
length(sel1) # 20

405 - 20 # 385

sel2 <- which(is.na(meta$Bases))
length(sel2) # 20

identical(sel1,sel2) #TRUE

meta[sel, ]

write.csv(x = meta[ -sel1 , c("Run","Submitter_Id","acvd_status","coronary_type") ], file = "jie-ACVD-sra-run-list-AVAILABLE.txt", quote = FALSE, row.names = FALSE)


#-------------------------

#### Jie ACVD - Metadata
#-------------------------

# metadata as per above

str(meta)
# 'data.frame':	405 obs. of  50 variables:
#   $ Run                                     : chr  "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" ...
# $ Assay Type                              : chr  "WGS" "WGS" "WGS" "WGS" ...
# $ BioProject                              : chr  "PRJEB21528" "PRJEB21528" "PRJEB21528" "PRJEB21528" ...
# $ BioSample                               : chr  "SAMEA104142374" "SAMEA104142373" "SAMEA104142372" "SAMEA104142371" ...
# $ Center Name                             : chr  "BEIJING GENOME INSTITUTE" "BEIJING GENOME INSTITUTE" "BEIJING GENOME INSTITUTE" "BEIJING GENOME INSTITUTE" ...
# $ Collection_Date                         : num  2011 2011 2011 2011 2011 ...
# $ Consent                                 : chr  "public" "public" "public" "public" ...
# $ ENA-FIRST-PUBLIC (run)                  : POSIXct, format: "2017-12-14" "2017-12-14" "2017-12-14" "2017-12-14" ...
# $ ENA-FIRST-PUBLIC                        : POSIXct, format: "2017-08-26" "2017-08-26" "2017-08-26" "2017-08-26" ...
# $ ENA-LAST-UPDATE (run)                   : POSIXct, format: "2018-11-16" "2018-11-16" "2018-11-16" "2018-11-16" ...
# $ ENA-LAST-UPDATE                         : POSIXct, format: "2017-06-27" "2017-06-27" "2017-06-27" "2017-06-27" ...
# $ environment_(biome)                     : chr  "human gut" "human gut" "human gut" "human gut" ...
# $ environment_(feature)                   : chr  "human-associated habitat" "human-associated habitat" "human-associated habitat" "human-associated habitat" ...
# $ environment_(material)                  : chr  "faeces" "faeces" "faeces" "faeces" ...
# $ Experiment                              : chr  "ERX2076993" "ERX2076994" "ERX2076995" "ERX2076996" ...
# $ External_Id                             : chr  "SAMEA104142374" "SAMEA104142373" "SAMEA104142372" "SAMEA104142371" ...
# $ geo_loc_name_country                    : chr  "China" "China" "China" "China" ...
# $ geo_loc_name_country_continent          : chr  "Asia" "Asia" "Asia" "Asia" ...
# $ geographic_location_(country_and/or_sea): chr  "China" "China" "China" "China" ...
# $ geographic_location_(latitude)          : num  55.7 55.7 55.7 55.7 55.7 ...
# $ geographic_location_(longitude)         : num  12.6 12.6 12.6 12.6 12.6 ...
# $ human_gut_environmental_package         : chr  "human-gut" "human-gut" "human-gut" "human-gut" ...
# $ INSDC_center_alias                      : chr  "BGI" "BGI" "BGI" "BGI" ...
# $ INSDC_center_name                       : chr  "Beijing Genome Institute" "Beijing Genome Institute" "Beijing Genome Institute" "Beijing Genome Institute" ...
# $ INSDC_first_public                      : chr  "2017-08-26T17:01:31Z" "2017-08-26T17:01:31Z" "2017-08-26T17:01:31Z" "2017-08-26T17:01:31Z" ...
# $ INSDC_last_update                       : chr  "2017-06-27T07:34:11Z" "2017-06-27T07:34:11Z" "2017-06-27T07:34:11Z" "2017-06-27T07:34:11Z" ...
# $ INSDC_status                            : chr  "public" "public" "public" "public" ...
# $ Instrument                              : chr  "Illumina HiSeq 2000" "Illumina HiSeq 2000" "Illumina HiSeq 2000" "Illumina HiSeq 2000" ...
# $ Investigation_type                      : chr  "metagenome" "metagenome" "metagenome" "metagenome" ...
# $ Library Name                            : chr  "SZAXPI001356-43" "SZAXPI001357-44" "SZAXPI001358-45" "SZAXPI001359-46" ...
# $ LibraryLayout                           : chr  "PAIRED" "PAIRED" "PAIRED" "PAIRED" ...
# $ LibrarySelection                        : chr  "other" "other" "other" "other" ...
# $ LibrarySource                           : chr  "METAGENOMIC" "METAGENOMIC" "METAGENOMIC" "METAGENOMIC" ...
# $ Organism                                : chr  "human gut metagenome" "human gut metagenome" "human gut metagenome" "human gut metagenome" ...
# $ Platform                                : chr  "ILLUMINA" "ILLUMINA" "ILLUMINA" "ILLUMINA" ...
# $ project_name                            : chr  "Acvd" "Acvd" "Acvd" "Acvd" ...
# $ ReleaseDate                             : chr  "2017-12-15T00:00:00Z" "2017-12-15T00:00:00Z" "2017-12-15T00:00:00Z" "2017-12-15T00:00:00Z" ...
# $ Sample Name                             : chr  "SAMEA104142374" "SAMEA104142373" "SAMEA104142372" "SAMEA104142371" ...
# $ Sample_name                             : chr  "N001" "N002" "N003" "N005" ...
# $ Sequencing_method                       : chr  "Illumina" "Illumina" "Illumina" "Illumina" ...
# $ SRA Study                               : chr  "ERP023788" "ERP023788" "ERP023788" "ERP023788" ...
# $ Submitter_Id                            : chr  "N001" "N002" "N003" "N005" ...
# $ DATASTORE filetype                      : chr  "fastq,sra" "sra,fastq" "fastq,sra" "sra,fastq" ...
# $ DATASTORE provider                      : chr  "ena,s3" "ena,s3" "s3,ena" "ena,s3" ...
# $ DATASTORE region                        : chr  "s3.us-east-1,ena" "ena,s3.us-east-1" "s3.us-east-1,ena" "s3.us-east-1,ena" ...
# $ AvgSpotLen                              : num  177 157 158 158 158 156 157 158 176 156 ...
# $ Bases                                   : num  3.16e+09 2.81e+09 2.81e+09 3.14e+09 3.67e+09 ...
# $ Bytes                                   : num  2.41e+09 2.46e+09 2.45e+09 2.76e+09 3.21e+09 ...
# $ acvd_status                             : num  0 0 0 0 0 0 0 0 0 0 ...
# $ coronary_type                           : chr  "NA" "NA" "NA" "NA" ...

table(meta$acvd_status)
# 0   1 
# 187 218

meta$acvd_status # heart disease = 1; normal = 0


table(meta$coronary_type)
# AMI              NA   Stable angina Unstable angina 
# 5             187             205               8 

str(subjects)
# 'data.frame':	405 obs. of  55 variables:
#   $ Sample ID                                : chr  "N001" "N002" "N003" "N005" ...
# $ ACVD status                              : num  0 0 0 0 0 0 0 0 0 0 ...
# $ Coronary type                            : chr  "NA" "NA" "NA" "NA" ...
# $ Age (year)                               : chr  "52" "53" "48" "53" ...
# $ Gender                                   : chr  "male" "male" "male" "female" ...
# $ Height (cm)                              : chr  "165" "160" "166" "158" ...
# $ Weight (kg)                              : chr  "65" "63" "70" "57.5" ...
# $ Body Mass Index (BMI)                    : chr  "23.88" "24.61" "25.4" "23.03" ...
# $ Waist (cm)                               : chr  "NA" "78" "92" "NA" ...
# $ Hip (cm)                                 : chr  "NA" "90" "98" "NA" ...
# $ Waist Hip Ratio (WHR)                    : chr  "NA" "0.86699999999999999" "0.93899999999999995" "NA" ...
# $ Smoking                                  : chr  "NA" "smoking" "NA" "NA" ...
# $ Systolic Blood Pressure (mmHg)           : chr  "NA" "123" "128" "NA" ...
# $ Diastolic Blood Pressure (mmHg)          : chr  "NA" "67" "72" "NA" ...
# $ FBG                                      : chr  "8.9700000000000006" "4.46" "6.3" "5.19" ...
# $ TRIG (mmol/L)                            : chr  "NA" "1.88" "1.22" "2.2000000000000002" ...
# $ LDLC (mmol/L)                            : chr  "NA" "2.89" "4.16" "0.72" ...
# $ CHOL (mmol/L)                            : chr  "NA" "4.58" "6.38" "4.67" ...
# $ HDLC (mmol/L)                            : chr  "NA" "0.97" "1.1499999999999999" "1.23" ...
# $ APOB (mg/dl)                             : chr  "NA" "0.76" "1.1100000000000001" "NA" ...
# $ APOA (mg/dl)                             : chr  "NA" "0.99" "2.27" "2.73" ...
# $ Lpa (mg/dl)                              : chr  "NA" "53" "192" "57.92" ...
# $ HbA1c (%)                                : chr  "NA" "NA" "NA" "NA" ...
# $ CKMB (U/L)                               : chr  "NA" "NA" "NA" "NA" ...
# $ ALB (U/L)                                : chr  "NA" "NA" "NA" "NA" ...
# $ ALT (U/L)                                : chr  "NA" "NA" "NA" "NA" ...
# $ TP (g/L)                                 : chr  "NA" "NA" "NA" "NA" ...
# $ AST (U/L)                                : chr  "NA" "NA" "NA" "NA" ...
# $ CREA (umol/L)                            : chr  "NA" "NA" "NA" "NA" ...
# $ CK (U/L)                                 : chr  "NA" "NA" "NA" "NA" ...
# $ ProBNP E (Pg/ml)                         : chr  "NA" "NA" "NA" "NA" ...
# $ CRP (mg/l)                               : chr  "NA" "NA" "NA" "NA" ...
# $ HBDH (U/L）                              : chr  "NA" "NA" "NA" "NA" ...
# $ DBIL (umol/L)                            : chr  "NA" "NA" "NA" "NA" ...
# $ BUN (nmol/L)                             : chr  "NA" "NA" "NA" "NA" ...
# $ TBIL (umol/L)                            : chr  "NA" "NA" "NA" "NA" ...
# $ URIC (umol/L)                            : chr  "NA" "NA" "NA" "NA" ...
# $ Clopidogrel Hydrogen Sulphate Tablets_117: chr  "NA" "NA" "NA" "NA" ...
# $ Aspirin_33                               : chr  "NA" "NA" "NA" "NA" ...
# $ Atorvastatin_25                          : chr  "NA" "NA" "NA" "NA" ...
# $ Esomeprazole_9                           : chr  "NA" "NA" "NA" "NA" ...
# $ Isosorbide Mononitrate_11                : chr  "NA" "NA" "NA" "NA" ...
# $ Potassium Citrate_14                     : chr  "NA" "NA" "NA" "NA" ...
# $ Perindopril_15                           : chr  "NA" "NA" "NA" "NA" ...
# $ Heparin Sodium_10                        : chr  "NA" "NA" "NA" "NA" ...
# $ Bisoprolol_9                             : chr  "NA" "NA" "NA" "NA" ...
# $ Fondaparinux Sodium_10                   : chr  "NA" "NA" "NA" "NA" ...
# $ Metoprolol_19                            : chr  "NA" "NA" "NA" "NA" ...
# $ Insulin aspart_6                         : chr  "NA" "NA" "NA" "NA" ...
# $ Isosorbide dinitrate_10                  : chr  "NA" "NA" "NA" "NA" ...
# $ Acarbose_7                               : chr  "NA" "NA" "NA" "NA" ...
# $ Captopril Tablets_6                      : chr  "NA" "NA" "NA" "NA" ...
# $ Estazolam_6                              : chr  "NA" "NA" "NA" "NA" ...
# $ Nitroglycerin Tablets_8                  : chr  "NA" "NA" "NA" "NA" ...
# $ Potassium chloride_32                    : chr  "NA" "NA" "NA" "NA" ...

# convert characters to numeric
subjects$`Body Mass Index (BMI)` <- as.numeric(subjects$`Body Mass Index (BMI)`)

table(meta$geo_loc_name_country)
# China 
# 405 

length(unique(meta$Sample_name)) # 114
head(meta$Sample_name) # "N001" "N002" "N003" "N005" "N006" "N009"
tail(meta$Sample_name) # "ZSL_56" "ZSL_58" "ZSL_67" "ZSL_6"  "ZSL_82" "ZSL_96"
length(unique(meta$Sample_Name)) # 114

identical(meta$Sample_name, subjects$`Sample ID`) # FALSE

length(meta$Sample_name %in% subjects$`Sample ID`) # 405

names(subjects)
# [1] "Sample ID"                                 "ACVD status"                              
# [3] "Coronary type"                             "Age (year)"                               
# [5] "Gender"                                    "Height (cm)"                              
# [7] "Weight (kg)"                               "Body Mass Index (BMI)"                    
# [9] "Waist (cm)"                                "Hip (cm)"                                 
# [11] "Waist Hip Ratio (WHR)"                     "Smoking"                                  
# [13] "Systolic Blood Pressure (mmHg)"            "Diastolic Blood Pressure (mmHg)"          
# [15] "FBG"                                       "TRIG (mmol/L)"                            
# [17] "LDLC (mmol/L)"                             "CHOL (mmol/L)"                            
# [19] "HDLC (mmol/L)"                             "APOB (mg/dl)"                             
# [21] "APOA (mg/dl)"                              "Lpa (mg/dl)"                              
# [23] "HbA1c (%)"                                 "CKMB (U/L)"                               
# [25] "ALB (U/L)"                                 "ALT (U/L)"                                
# [27] "TP (g/L)"                                  "AST (U/L)"                                
# [29] "CREA (umol/L)"                             "CK (U/L)"                                 
# [31] "ProBNP E (Pg/ml)"                          "CRP (mg/l)"                               
# [33] "HBDH (U/L）"                               "DBIL (umol/L)"                            
# [35] "BUN (nmol/L)"                              "TBIL (umol/L)"                            
# [37] "URIC (umol/L)"                             "Clopidogrel Hydrogen Sulphate Tablets_117"
# [39] "Aspirin_33"                                "Atorvastatin_25"                          
# [41] "Esomeprazole_9"                            "Isosorbide Mononitrate_11"                
# [43] "Potassium Citrate_14"                      "Perindopril_15"                           
# [45] "Heparin Sodium_10"                         "Bisoprolol_9"                             
# [47] "Fondaparinux Sodium_10"                    "Metoprolol_19"                            
# [49] "Insulin aspart_6"                          "Isosorbide dinitrate_10"                  
# [51] "Acarbose_7"                                "Captopril Tablets_6"                      
# [53] "Estazolam_6"                               "Nitroglycerin Tablets_8"                  
# [55] "Potassium chloride_32"


names(subjects)[1] # "Sample ID"
names(subjects)[1] <- "Sample_ID"

subjects$Sample_ID

length(unique( gsub(pattern = "-", replacement = "_", x = subjects$Sample_ID) )) # 405

subjects$Sample_ID <- gsub(pattern = "-", replacement = "_", x = subjects$Sample_ID)

row.names(subjects) <- subjects$Sample_ID

( varnames <- c("Age (year)" , 
                  "Height (cm)"         ,                     
                "Weight (kg)"            ,                  "Body Mass Index (BMI)"    ,                
                #"Waist (cm)"              ,                  "Hip (cm)"               ,           Waist, Hip, Waist-Hip ratio not available for ACVD cases       
                #"Waist Hip Ratio (WHR)"     ,               # "Smoking"                ,                  
                "Systolic Blood Pressure (mmHg)"   ,         "Diastolic Blood Pressure (mmHg)"          ,
                "FBG"                        ,               "TRIG (mmol/L)"                          ,  
                "LDLC (mmol/L)"              ,               "CHOL (mmol/L)"                          ,  
                "HDLC (mmol/L)"             ,                "APOB (mg/dl)"                           ,  
                "APOA (mg/dl)"              ,                "Lpa (mg/dl)"                            ,  
                #"HbA1c (%)"                 ,    
                "CKMB (U/L)"                             ,  
                "ALB (U/L)"                 ,                "ALT (U/L)"                              ,  
                "TP (g/L)"                  ,                "AST (U/L)"                              ,  
                "CREA (umol/L)" ) )

# [1] "Age (year)"                      "Height (cm)"                     "Weight (kg)"                    
# [4] "Body Mass Index (BMI)"           "Systolic Blood Pressure (mmHg)"  "Diastolic Blood Pressure (mmHg)"
# [7] "FBG"                             "TRIG (mmol/L)"                   "LDLC (mmol/L)"                  
# [10] "CHOL (mmol/L)"                   "HDLC (mmol/L)"                   "APOB (mg/dl)"                   
# [13] "APOA (mg/dl)"                    "Lpa (mg/dl)"                     "CKMB (U/L)"                     
# [16] "ALB (U/L)"                       "ALT (U/L)"                       "TP (g/L)"                       
# [19] "AST (U/L)"                       "CREA (umol/L)" 

# convert to numeric
subjects[ ,varnames] <- lapply(X = subjects[ ,varnames], FUN = as.numeric)


# reduce to key data

subjects2 <- subjects[ ,c("Sample_ID", "ACVD status", "Coronary type", "Gender", varnames )]

plot_vars <- c("Gender",varnames)
length(plot_vars) # 21

subjects.melt <- melt(subjects2[ ,c("Sample_ID", "ACVD status", plot_vars)], id.vars = c("Sample_ID", "ACVD status"))

names(subjects.melt)
#   "Sample_ID"   "ACVD status" "variable"    "value"      

unique(subjects.melt$`ACVD status`) # 0 1
subjects.melt$`ACVD status` <- factor(subjects.melt$`ACVD status`, levels = c(1,0), labels = c("ACVD","Normal"),ordered = TRUE )

# plot numeric & categorical separately

unique(subjects.melt$variable)
# [1] Gender                          Age (year)                      Height (cm)                     Weight (kg)                    
# [5] Body Mass Index (BMI)           Systolic Blood Pressure (mmHg)  Diastolic Blood Pressure (mmHg) FBG                            
# [9] TRIG (mmol/L)                   LDLC (mmol/L)                   CHOL (mmol/L)                   HDLC (mmol/L)                  
# [13] APOB (mg/dl)                    APOA (mg/dl)                    Lpa (mg/dl)                     CKMB (U/L)                     
# [17] ALB (U/L)                       ALT (U/L)                       TP (g/L)                        AST (U/L)                      
# [21] CREA (umol/L)                  
# 21 Levels: Gender Age (year) Height (cm) Weight (kg) Body Mass Index (BMI) ... CREA (umol/L)

sel.cat <- which(subjects.melt$variable == "Gender")

subjects.melt.num <- subjects.melt[-sel.cat, ]
subjects.melt.num$value <- as.numeric(subjects.melt.num$value)


#sel.plot <- which(meta.melt$variable %in% plot#_vars)

p <- ggplot( subjects.melt.num, aes(x = `ACVD status`, y = value )) +
  geom_jitter(size=1.5,width = 0.15, alpha=0.25) +
  facet_wrap(facets = vars(variable), ncol = 5, scales = "free_y")+
  theme_bw()+
  theme(
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    strip.text = element_text(size = rel(0.7)),
    #strip.text.y = element_text(size = rel(1.1)),
    strip.background = element_blank()
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","Jie-ACVD-metadata-values.tiff"), width = 22, height = 24, units = "cm", res=600, compression="lzw", type = "cairo")


# sex

p <- ggplot( subjects.melt[sel.cat, ], aes(x = `ACVD status`, y = value )) +
  geom_jitter(size=1.5,width = 0.15, alpha=0.25) +
  theme_bw()+
  xlab(NULL) + ylab(NULL)+
  theme(
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    strip.text = element_text(size = rel(0.7)),
    #strip.text.y = element_text(size = rel(1.1)),
    strip.background = element_blank()
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","Jie-ACVD-metadata-values-GENDER-ONLY.tiff"), width = 5, height = 6, units = "cm", res=600, compression="lzw", type = "cairo")


#-------------------------

#### Jie ACVD - read in superfocus - fxn potential outputs
#-------------------------

head( subjects$Sample_ID )
# "N001" "N002" "N003" "N005" "N006" "N009"

## use the other metadata table to compile Superfocus data ...

head(meta)
# Run Assay Type BioProject      BioSample              Center Name Collection_Date Consent ENA-FIRST-PUBLIC (run)
# ERR2017411 ERR2017411        WGS PRJEB21528 SAMEA104142374 BEIJING GENOME INSTITUTE            2011  public             2017-12-14
# ERR2017412 ERR2017412        WGS PRJEB21528 SAMEA104142373 BEIJING GENOME INSTITUTE            2011  public             2017-12-14
# ERR2017413 ERR2017413        WGS PRJEB21528 SAMEA104142372 BEIJING GENOME INSTITUTE            2011  public             2017-12-14
# ERR2017414 ERR2017414        WGS PRJEB21528 SAMEA104142371 BEIJING GENOME INSTITUTE            2011  public             2017-12-14
# ERR2017415 ERR2017415        WGS PRJEB21528 SAMEA104142370 BEIJING GENOME INSTITUTE            2011  public             2017-12-14
# ERR2017416 ERR2017416        WGS PRJEB21528 SAMEA104142369 BEIJING GENOME INSTITUTE            2011  public             2017-12-14
# ENA-FIRST-PUBLIC ENA-LAST-UPDATE (run) ENA-LAST-UPDATE environment_(biome)    environment_(feature) environment_(material)
# ERR2017411       2017-08-26            2018-11-16      2017-06-27           human gut human-associated habitat                 faeces
# ERR2017412       2017-08-26            2018-11-16      2017-06-27           human gut human-associated habitat                 faeces
# ERR2017413       2017-08-26            2018-11-16      2017-06-27           human gut human-associated habitat                 faeces
# ERR2017414       2017-08-26            2018-11-16      2017-06-27           human gut human-associated habitat                 faeces
# ERR2017415       2017-08-26            2018-11-16      2017-06-27           human gut human-associated habitat                 faeces
# ERR2017416       2017-08-26            2018-11-16      2017-06-27           human gut human-associated habitat                 faeces
# Experiment    External_Id geo_loc_name_country geo_loc_name_country_continent geographic_location_(country_and/or_sea)
# ERR2017411 ERX2076993 SAMEA104142374                China                           Asia                                    China
# ERR2017412 ERX2076994 SAMEA104142373                China                           Asia                                    China
# ERR2017413 ERX2076995 SAMEA104142372                China                           Asia                                    China
# ERR2017414 ERX2076996 SAMEA104142371                China                           Asia                                    China
# ERR2017415 ERX2076997 SAMEA104142370                China                           Asia                                    China
# ERR2017416 ERX2076998 SAMEA104142369                China                           Asia                                    China
# geographic_location_(latitude) geographic_location_(longitude) human_gut_environmental_package INSDC_center_alias
# ERR2017411                        55.6761                        12.56834                       human-gut                BGI
# ERR2017412                        55.6761                        12.56834                       human-gut                BGI
# ERR2017413                        55.6761                        12.56834                       human-gut                BGI
# ERR2017414                        55.6761                        12.56834                       human-gut                BGI
# ERR2017415                        55.6761                        12.56834                       human-gut                BGI
# ERR2017416                        55.6761                        12.56834                       human-gut                BGI
# INSDC_center_name   INSDC_first_public    INSDC_last_update INSDC_status          Instrument Investigation_type
# ERR2017411 Beijing Genome Institute 2017-08-26T17:01:31Z 2017-06-27T07:34:11Z       public Illumina HiSeq 2000         metagenome
# ERR2017412 Beijing Genome Institute 2017-08-26T17:01:31Z 2017-06-27T07:34:11Z       public Illumina HiSeq 2000         metagenome
# ERR2017413 Beijing Genome Institute 2017-08-26T17:01:31Z 2017-06-27T07:34:11Z       public Illumina HiSeq 2000         metagenome
# ERR2017414 Beijing Genome Institute 2017-08-26T17:01:31Z 2017-06-27T07:34:11Z       public Illumina HiSeq 2000         metagenome
# ERR2017415 Beijing Genome Institute 2017-08-26T17:01:31Z 2017-06-27T07:34:11Z       public Illumina HiSeq 2000         metagenome
# ERR2017416 Beijing Genome Institute 2017-08-26T17:01:31Z 2017-06-27T07:34:11Z       public Illumina HiSeq 2000         metagenome
# Library Name LibraryLayout LibrarySelection LibrarySource             Organism Platform project_name
# ERR2017411 SZAXPI001356-43        PAIRED            other   METAGENOMIC human gut metagenome ILLUMINA         Acvd
# ERR2017412 SZAXPI001357-44        PAIRED            other   METAGENOMIC human gut metagenome ILLUMINA         Acvd
# ERR2017413 SZAXPI001358-45        PAIRED            other   METAGENOMIC human gut metagenome ILLUMINA         Acvd
# ERR2017414 SZAXPI001359-46        PAIRED            other   METAGENOMIC human gut metagenome ILLUMINA         Acvd
# ERR2017415 SZAXPI001360-47        PAIRED            other   METAGENOMIC human gut metagenome ILLUMINA         Acvd
# ERR2017416 SZAXPI001362-57        PAIRED            other   METAGENOMIC human gut metagenome ILLUMINA         Acvd
# ReleaseDate    Sample Name Sample_name Sequencing_method SRA Study Submitter_Id DATASTORE filetype
# ERR2017411 2017-12-15T00:00:00Z SAMEA104142374        N001          Illumina ERP023788         N001          fastq,sra
# ERR2017412 2017-12-15T00:00:00Z SAMEA104142373        N002          Illumina ERP023788         N002          sra,fastq
# ERR2017413 2017-12-15T00:00:00Z SAMEA104142372        N003          Illumina ERP023788         N003          fastq,sra
# ERR2017414 2017-12-15T00:00:00Z SAMEA104142371        N005          Illumina ERP023788         N005          sra,fastq
# ERR2017415 2017-12-15T00:00:00Z SAMEA104142370        N006          Illumina ERP023788         N006          fastq,sra
# ERR2017416 2017-12-15T00:00:00Z SAMEA104142369        N009          Illumina ERP023788         N009          sra,fastq
# DATASTORE provider DATASTORE region AvgSpotLen      Bases      Bytes acvd_status coronary_type
# ERR2017411             ena,s3 s3.us-east-1,ena        177 3161133051 2409737743           0            NA
# ERR2017412             ena,s3 ena,s3.us-east-1        157 2808375061 2461690116           0            NA
# ERR2017413             s3,ena s3.us-east-1,ena        158 2807517536 2453740943           0            NA
# ERR2017414             ena,s3 s3.us-east-1,ena        158 3141515009 2756007878           0            NA
# ERR2017415             ena,s3 ena,s3.us-east-1        158 3671068081 3213635111           0            NA
# ERR2017416             s3,ena s3.us-east-1,ena        156 3698292976 3284925472           0            NA

table( meta$`DATASTORE filetype`, useNA = "ifany" )
# fastq fastq,sra sra,fastq      <NA> 
#   19       211       174         1 

sel1 <- which(meta$`DATASTORE filetype` %in% c("fastq",NA))
length(sel1) # 20

405 - 20 # 385

length( unique(meta$Sample_name) ) # 405

meta[sel1, ]

meta.ok <- meta[-sel1, ]


sampid <- meta.ok$Run
length(meta.ok$Run) # 385
length(unique(sampid)) # 385

superfocus_out_dir <- "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus"

list.dirs(superfocus_out_dir)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus"                          
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017411"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017412"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017413"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017414"
# etc ...


# # don't keep 1st directory
( results_dirs <- list.dirs(superfocus_out_dir)[-c(1)] )

head(results_dirs)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017411"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017412"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017413"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017414"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017415"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_ERR2017416"

# check identical order
identical(as.character(sampid), gsub(pattern = "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/jie-acvd/superfocus/superfocus_out_", replacement = "", x = results_dirs) )
# TRUE

# In this data one Run corresponds to a single Sample_ID !!!

# collate results into a long-format table

sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)

for (i in 1:length(sampid)) {
  #i<-1
  this_samp <- sampid[i]
  sel.folder <- grep(pattern = this_samp, x = results_dirs)
  this_folder <- results_dirs[sel.folder]

  #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)

  tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
  # names(tab)
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # [5] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq"
  # [6] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq.."


  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # [5] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"
  # [6] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"
  # [7] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"
  # [8] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"
  # [9] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [10] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  # [11] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [12] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %


  tab$sampid <- this_samp
  names(tab)

  #tab <- tab[,c(7,1,2,3,4,6)]

  # last column is sampid
  # take average of percentages

  sel.col.percent <- grep(pattern = "R1.good.fastq..$", x = names(tab))
  if (length(sel.col.percent)>1) {
    tab$percent_abun <- apply(X = tab[ ,sel.col.percent], MARGIN = 1, FUN = mean )
  } else {
    tab$percent_abun <- tab[ ,sel.col.percent]
  }

  # sum(tab$percent_abun) # 100
  # mean(tab$percent_abun) # 0.004338583

  names(sfx.long) # "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"    "percent_abun"
  # names(tab)
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # ...
  # [13] "sampid"
  # [14] "percent_abun"

  tab <- tab[ ,c("sampid","Subsystem.Level.1","Subsystem.Level.2","Subsystem.Level.3","Function","percent_abun")]
  names(tab) <- names(sfx.long)

  sfx.long <- rbind(sfx.long,tab)

  print(paste0("completed ",i," - sample ID: ",sampid[i]))
}





head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
dim(sfx.long) # 3581616       6
head(sfx.long)
# sampleID                   subsys_L1                    subsys_L2                      subsys_L3
# 2 ERR2017411 Amino Acids and Derivatives Alanine, serine, and glycine           Alanine biosynthesis
# 3 ERR2017411 Amino Acids and Derivatives Alanine, serine, and glycine           Glycine Biosynthesis
# 4 ERR2017411 Amino Acids and Derivatives Alanine, serine, and glycine Glycine and Serine Utilization
# 5 ERR2017411 Amino Acids and Derivatives Alanine, serine, and glycine Glycine and Serine Utilization
# 6 ERR2017411 Amino Acids and Derivatives Alanine, serine, and glycine Glycine and Serine Utilization
# 7 ERR2017411 Amino Acids and Derivatives Alanine, serine, and glycine        Glycine cleavage system
# fxn percent_abun
# 2                                             Branched-chain_amino_acid_aminotransferase_(EC_2.6.1.42)  0.004222260
# 3                                                           L-threonine_3-dehydrogenase_(EC_1.1.1.103)  0.042222598
# 4                                                     D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)  0.023644655
# 5 L-serine_dehydratase,_beta_subunit_(EC_4.3.1.17)_/_L-serine_dehydratase,_alpha_subunit_(EC_4.3.1.17)  0.060800540
# 6                                                                   L-serine_dehydratase_(EC_4.3.1.17)  0.031244722
# 7                                                                   L-serine_dehydratase_(EC_4.3.1.17)  0.001688904


sfx.long$full_fxn_tax <- paste0(sfx.long$subsys_L1,"___", sfx.long$subsys_L2,"___", sfx.long$subsys_L3,"___", sfx.long$fxn)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 21128   386

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 21128
length(unique(full_fxn_names)) # 21128

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_##_catabolic"


tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)


names(sfx.wide)
# [1] "full_fxn_tax" "ERR2017411"   "ERR2017412"   "ERR2017413"   "ERR2017414"   "ERR2017415"   "ERR2017416"   "ERR2017417"   "ERR2017418"  
# [10] "ERR2017419"   "ERR2017420"   "ERR2017421"   "ERR2017422"   "ERR2017423"   "ERR2017424"   "ERR2017426"   "ERR2017427"   "ERR2017428"  
# [19] "ERR2017429"   "ERR2017430"   "ERR2017431"   "ERR2017432"   "ERR2017433"   "ERR2017434"   "ERR2017435"   "ERR2017436"   "ERR2017437"  
# [28] "ERR2017438"   "ERR2017439"   "ERR2017440"   "ERR2017441"   "ERR2017442"   "ERR2017443"   "ERR2017444"   "ERR2017445"   "ERR2017446"  
# [37] "ERR2017447"   "ERR2017448"   "ERR2017449"   "ERR2017450"   "ERR2017451"   "ERR2017452"   "ERR2017453"   "ERR2017454"   "ERR2017455"  
# etc ...

#names(sfx.wide) <- gsub(pattern = "-", replacement = "_", x = names(sfx.wide))

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide)


head(sampid)
# "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" "ERR2017415" "ERR2017416"

length(sampid) # 385

names(sampid) # NULL - in this case there is NOT an alternative sample name being used

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # TRUE



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 21128     4

length(unique(tax.fxn$subsys_L1)) # 35
length(unique(tax.fxn$subsys_L2)) # 188
length(unique(tax.fxn$subsys_L3)) # 1150
length(unique(tax.fxn$fxn)) # 11143

# sel.dup <- which(duplicated.default(tax.fxn$fxn)==TRUE)
# head(tax.fxn$fxn[sel.dup])
# 
# 
# 
# # examine last example 'tab'
# 
# head(tab)
# #   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# # 1     9468 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 3.819676e-04
# # 2     9468 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 1.766030e-05
# # 3     9468 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 2.587839e-05
# # 4     9468 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.514521e-05
# # 5     9468 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.287010e-03
# # 6     9468 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 4.401921e-05
# 
# 
# length(unique(tab$subsys_L1)) # 35
# length(unique(tab$subsys_L2)) # 191
# length(unique(tab$subsys_L3)) # 1184
# length(unique(tab$fxn)) # 10975

#-------------------------

#### Jie ACVD - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 21128     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 21128  385

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 21128 taxa and 385 samples ]
# tax_table()   Taxonomy Table:    [ 21128 taxa by 4 taxonomic ranks ]

sample_names(phy)
# [1] "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" etc ...

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta.ok) # 385  50
dim(subjects) # 405  55

keep_subjects <- unique(meta.ok$Sample_name) # qty 385
keep_subjects <- gsub(pattern = "-", replacement = "_", x = keep_subjects)

sel <- which( row.names(subjects) %in% keep_subjects)
subjects.ok <- subjects[sel, ]

# but add Run No to subjects data.frame & use Run as row.name
subjects.ok$Run <- NA
for (i in 1:dim(subjects.ok)[1]) {
  #i<-1
  this_subject <- subjects.ok$Sample_ID[i]
  #sel <- which(meta.ok$Sample_name == this_subject)
  sel <- which( gsub(pattern = "-", replacement = "_", x = meta.ok$Sample_name) == this_subject)
  subjects.ok$Run[i] <- meta.ok$Run[sel]
  print(paste0("completed ",i))
}

identical(subjects.ok$Run, meta.ok$Run) # FALSE
identical(subjects.ok$Sample_ID, meta.ok$Sample_name) # FALSE
length(subjects.ok$Sample_ID %in% meta.ok$Sample_name) # 385 - different order!

row.names(subjects.ok) <- subjects.ok$Run

identical( row.names(subjects.ok), sample_names(phy) ) # TRUE
head(row.names(subjects.ok)) # "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" "ERR2017415" "ERR2017416"
head(sample_names(phy)) # "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" "ERR2017415" "ERR2017416"
tail(row.names(subjects.ok)) # "ERR2017804" "ERR2017805" "ERR2017806" "ERR2017807" "ERR2017808" "ERR2017809"
tail(sample_names(phy)) # "ERR2017808" "ERR2017809" "ERR2017810" "ERR2017813" "ERR2017814" "ERR2017815"

length( row.names(subjects.ok) %in% sample_names(phy) ) # 385
# change order to match phy
subjects.ok2 <- subjects.ok[ sample_names(phy), ]


SAMP <- sample_data(subjects.ok2)



### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 21128 taxa and 385 samples ]
# sample_data() Sample Data:       [ 385 samples by 56 sample variables ]
# tax_table()   Taxonomy Table:    [ 21128 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#   subsys_L1                     subsys_L2 subsys_L3            
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn                                                                            
# fxn_1 "2-methylaconitate_cis-trans_isomerase"                                        
# fxn_2 "2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)"
# fxn_3 "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                      
# fxn_4 "Alanine_racemase_(EC_5.1.1.1)"                                                
# fxn_5 "Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic"                                
# fxn_6 "Alanine_racemase_(EC_5.1.1.1)_##_catabolic" 


getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-Jie-ACVD.RDS")

#phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")


#-------------------------

#### Jie ACVD - summarise functional data hierarchy
#-------------------------

#phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")

phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 21128 taxa and 385 samples ]
# sample_data() Sample Data:       [ 385 samples by 56 sample variables ]
# tax_table()   Taxonomy Table:    [ 21128 taxa by 4 taxonomic ranks ]

table(phy@sam_data$Gender)
# female   male     NA 
# 154    226      5
sel.na <- which(phy@sam_data$Gender=="NA") 
keep_samps <- sample_names(phy)[-sel.na]
phy.ok <- prune_samples(samples = keep_samps, phy)
min(taxa_sums(phy.ok))
# prune taxa with zero reads
phy.ok <- prune_taxa(taxa = taxa_sums(phy.ok) > 0, x = phy.ok)
phy.ok
# otu_table()   OTU Table:         [ 21117 taxa and 380 samples ]
# sample_data() Sample Data:       [ 380 samples by 56 sample variables ]
# tax_table()   Taxonomy Table:    [ 21117 taxa by 4 taxonomic ranks ]


rank_names(phy.ok) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"

length(unique( phy.ok@tax_table[ , "subsys_L1"] )) # 35

unique( phy.ok@tax_table[ , "subsys_L1"] )
# Taxonomy Table:     [35 taxa by 1 taxonomic ranks]:
#   subsys_L1                                                   
# fxn_1     "Amino Acids and Derivatives"                               
# fxn_1997  "Arabinose Sensor and transport module"                     
# fxn_1999  "Carbohydrates"                                             
# fxn_5105  "Cell Division and Cell Cycle"                              
# fxn_5231  "Cell Wall and Capsule"                                     
# fxn_6350  "Central metabolism"                                        
# fxn_6394  "Clustering-based subsystems"                               
# fxn_7567  "Cofactors, Vitamins, Prosthetic Groups, Pigments"          
# fxn_10199 "DNA Metabolism"                                            
# fxn_10732 "Dormancy and Sporulation"                                  
# fxn_10870 "Fatty Acids, Lipids, and Isoprenoids"                      
# fxn_11479 "Iron acquisition and metabolism"                           
# fxn_11820 "Membrane Transport"                                        
# fxn_12457 "Metabolism of Aromatic Compounds"                          
# fxn_12823 "Miscellaneous"                                             
# fxn_13694 "Motility and Chemotaxis"                                   
# fxn_13876 "Nitrogen Metabolism"                                       
# fxn_14132 "Nucleosides and Nucleotides"                               
# fxn_14531 "Phages, Prophages, Transposable elements"                  
# fxn_14628 "Phages, Prophages, Transposable elements, Plasmids"        
# fxn_14913 "Phosphorus Metabolism"                                     
# fxn_15082 "Photosynthesis"                                            
# fxn_15138 "Plant cell walls and outer surfaces"                       
# fxn_15161 "Potassium metabolism"                                      
# fxn_15301 "Predictions based on plant-prokaryote comparative analysis"
# fxn_15419 "Protein Metabolism"                                        
# fxn_16373 "Regulation and Cell signaling"                             
# fxn_16882 "Respiration"                                               
# fxn_18143 "RNA Metabolism"                                            
# fxn_19086 "Secondary Metabolism"                                      
# fxn_19126 "Stress Response"                                           
# fxn_19979 "Sulfur Metabolism"                                         
# fxn_20244 "Transcriptional regulation"                                
# fxn_20267 "Virulence"                                                 
# fxn_20876 "Virulence, Disease and Defense" 

length( unique( phy.ok@tax_table[ , "subsys_L2"] ) ) # 188
unique( phy.ok@tax_table[ , "subsys_L2"] )
# Taxonomy Table:     [188 taxa by 1 taxonomic ranks]:
#   subsys_L2                                                                                        
# fxn_1     "-"                                                                                              
# fxn_34    "Alanine, serine, and glycine"                                                                   
# fxn_188   "Arginine; urea cycle, polyamines"                                                               
# fxn_469   "Aromatic amino acids and derivatives"                                                           
# fxn_897   "Branched-chain amino acids"                                                                     
# fxn_1166  "Glutamine, glutamate, aspartate, asparagine; ammonia assimilation"                              
# fxn_1290  "Histidine Metabolism"                                                                           
# fxn_1365  "Lysine, threonine, methionine, and cysteine"                                                    
# fxn_1893  "Proline and 4-hydroxyproline"                                                                   
# fxn_1997  "Arabinose"                                                                                      
# fxn_2170  "Aminosugars"                                                                                    
# fxn_2315  "Central carbohydrate metabolism"                                                                
# fxn_3061  "CO2 fixation"                                                                                   
# fxn_3170  "Di- and oligosaccharides"                                                                       
# fxn_3574  "Fermentation"                                                                                   
# fxn_3817  "Monosaccharides"                                                                                
# fxn_4471  "Nucleotide sugars"                                                                              
# fxn_4478  "One-carbon Metabolism"                                                                          
# fxn_4693  "Organic acids"                                                                                  
# fxn_4835  "Polysaccharides"                                                                                
# fxn_4902  "Sugar alcohols"                                                                                 
# fxn_5105  " Bacterial checkpoint control"                                                                  
# fxn_5495  "Capsular and extracellular polysacchrides"                                                      
# fxn_5963  "Cell wall of Mycobacteria"                                                                      
# fxn_5964  "Gram-Negative cell wall components"                                                             
# fxn_6243  "Gram-Positive cell wall components"                                                             
# fxn_6393  "TCA"                                                                                            
# fxn_6394  " Tartronate-semialdehyde related area (links to pyridoxine and aldorate metabolism)"            
# fxn_6972  "A bicyclomycin resistance protein, a helicase, and a pseudouridine synthase"                    
# fxn_6979  "alpha-proteobacterial cluster of hypotheticals"                                                 
# fxn_6980  "Biosynthesis of galactoglycans and related lipopolysacharides"                                  
# fxn_6994  "Carotenoid biosynthesis"                                                                        
# fxn_7002  "Catabolism of an unknown compound"                                                              
# fxn_7003  "Cell Division"                                                                                  
# fxn_7085  "Chemotaxis, response regulators"                                                                
# fxn_7087  "Choline bitartrate degradation, putative"                                                       
# fxn_7097  "Chromosome Replication"                                                                         
# fxn_7102  "clustering of 2 heat shock proteins, phosphoenolpyruvate carboxykinase and a putative hydrolase"
# fxn_7108  "contains Thr-tRNA-syn, pyridoxine biosyn, lipid A biosyn, 3 hypos"                              
# fxn_7115  "CRISPRs and associated hypotheticals"                                                           
# fxn_7126  "Cytochrome biogenesis"                                                                          
# fxn_7152  "D-tyrosyl-tRNA(Tyr) deacylase (EC 3.1.-.-) cluster"                                             
# fxn_7158  "Degradation of Polyphenols (?)"                                                                 
# fxn_7164  "DNA metabolism"                                                                                 
# fxn_7166  "DNA polymerase III epsilon cluster"                                                             
# fxn_7179  "Fage-related, replication"                                                                      
# fxn_7192  "Fatty acid metabolic cluster"                                                                   
# fxn_7205  "Flagella protein?"                                                                              
# fxn_7208  "heat shock, cell division, proteases, and a methyltransferase"                                  
# fxn_7215  "Hypothetical associated with RecF"                                                              
# fxn_7219  "Hypothetical in Lysine biosynthetic cluster"                                                    
# fxn_7228  "Hypothetical lipase related to Phosphatidate metabolism"                                        
# fxn_7232  "Hypothetical protein possible functionally linked with Alanyl-tRNA synthetase"                  
# fxn_7239  "Hypothetical Related to Dihydroorate Dehydrogenase"                                             
# fxn_7240  "Isoprenoid/cell wall biosynthesis: PREDICTED UNDECAPRENYL DIPHOSPHATE PHOSPHATASE"              
# fxn_7261  "Lysine Biosynthesis"                                                                            
# fxn_7265  "Methylamine utilization"                                                                        
# fxn_7267  "Molybdopterin oxidoreductase"                                                                   
# fxn_7273  "Nucleotidyl-phosphate metabolic cluster"                                                        
# fxn_7285  "Pigment biosynthesis"                                                                           
# fxn_7289  "Probably GTP or GMP signaling related"                                                          
# fxn_7304  "Probably organic hydroperoxide resistance related hypothetical protein"                         
# fxn_7310  "Probably Pyrimidine biosynthesis-related"                                                       
# fxn_7312  "Probably Ybbk-related hypothetical membrane proteins"                                           
# fxn_7316  "Proteasome related clusters"                                                                    
# fxn_7324  "Protein export?"                                                                                
# fxn_7345  "proteosome related"                                                                             
# fxn_7354  "Putative asociate of RNA polymerase sigma-54 factor rpoN"                                       
# fxn_7365  "Putative GGDEF domain protein related to agglutinin secretion"                                  
# fxn_7370  "Putative Isoquinoline 1-oxidoreductase subunit"                                                 
# fxn_7373  "Putrescine/GABA utilization cluster-temporal,to add to SSs"                                     
# fxn_7389  "Recombination related cluster"                                                                  
# fxn_7392  "recX and regulatory cluster"                                                                    
# fxn_7397  "Related to Menaquinone-cytochrome C reductase "                                                 
# fxn_7402  "Ribosomal Protein L28P relates to a set of uncharacterized proteins"                            
# fxn_7417  "Ribosome-related cluster"                                                                       
# fxn_7431  "Shiga toxin cluster"                                                                            
# fxn_7455  "Streptococcus cluster: maybe related to transport"                                              
# fxn_7459  "Sulfatases and sulfatase modifying factor 1 (and a hypothetical)"                               
# fxn_7467  "Three hypotheticals linked to lipoprotein biosynthesis"                                         
# fxn_7468  "TldD cluster"                                                                                   
# fxn_7485  "Translation"                                                                                    
# fxn_7510  "Tricarboxylate transporter"                                                                     
# fxn_7528  "tRNA sulfuration"                                                                               
# fxn_7534  "Two related proteases"                                                                          
# fxn_7550  "Type III secretion system, extended"                                                            
# fxn_7557  "Urate degradation"                                                                              
# fxn_8253  "Biotin"                                                                                         
# fxn_8406  "Coenzyme A"                                                                                     
# fxn_8512  "Coenzyme B"                                                                                     
# fxn_8513  "Coenzyme F420"                                                                                  
# fxn_8530  "Coenzyme M"                                                                                     
# fxn_8543  "Fe-S clusters"                                                                                  
# fxn_8589  "Folate and pterines"                                                                            
# fxn_9078  "Lipoic acid"                                                                                    
# fxn_9175  "NAD and NADP"                                                                                   
# fxn_9282  "Pyridoxine"                                                                                     
# fxn_9364  "Quinone cofactors"                                                                              
# fxn_9628  "Riboflavin, FMN, FAD"                                                                           
# fxn_9838  "Tetrapyrroles"                                                                                  
# fxn_10174 "Thiamin"                                                                                        
# fxn_10263 "CRISPs"                                                                                         
# fxn_10313 "DNA recombination"                                                                              
# fxn_10328 "DNA repair"                                                                                     
# fxn_10528 "DNA replication"                                                                                
# fxn_10666 "DNA uptake, competence"                                                                         
# fxn_10861 "Spore DNA protection"                                                                           
# fxn_10934 "Fatty acids"                                                                                    
# fxn_11118 "Isoprenoids"                                                                                    
# fxn_11396 "Phospholipids"                                                                                  
# fxn_11468 "Triacylglycerols"                                                                               
# fxn_11646 "Siderophores"                                                                                   
# fxn_12029 "ABC transporters"                                                                               
# fxn_12168 "Protein and nucleoprotein secretion system, Type IV"                                            
# fxn_12249 "Protein secretion system, Chaperone-Usher pathway (CU)"                                         
# fxn_12294 "Protein secretion system, Type II"                                                              
# fxn_12320 "Protein secretion system, Type III"                                                             
# fxn_12324 "Protein secretion system, Type VI"                                                              
# fxn_12356 "Protein secretion system, Type VII"                                                             
# fxn_12376 "Protein secretion system, Type VIII (Extracellular nucleation/precipitation pathway, ENP)"      
# fxn_12385 "Protein translocation across cytoplasmic membrane"                                              
# fxn_12409 "Sugar Phosphotransferase Systems, PTS"                                                          
# fxn_12430 "TRAP transporters"                                                                              
# fxn_12433 "Uni- Sym- and Antiporters"                                                                      
# fxn_12525 "Anaerobic degradation of aromatic compounds"                                                    
# fxn_12548 "Metabolism of central aromatic intermediates"                                                   
# fxn_12685 "Peripheral pathways for catabolism of aromatic compounds"                                       
# fxn_13107 "Lactate racemization"                                                                           
# fxn_13116 "Plant-Prokaryote comparative genomics"                                                          
# fxn_13731 "Flagellar motility in Prokaryota"                                                               
# fxn_13839 "Social motility and nonflagellar swimming in bacteria"                                          
# fxn_14202 "Purines"                                                                                        
# fxn_14388 "Pyrimidines"                                                                                    
# fxn_14554 "Bacteriophage integration/excision/lysogeny"                                                    
# fxn_14571 "Bacteriophage structural proteins"                                                              
# fxn_14607 "Phage family-specific subsystems"                                                               
# fxn_14614 "Phage Host Interactions"                                                                        
# fxn_14625 "Superinfection Exclusion"                                                                       
# fxn_14634 "Pathogenicity islands"                                                                          
# fxn_14668 "Phages, Prophages"                                                                              
# fxn_14865 "Plasmid related functions"                                                                      
# fxn_14866 "Transposable elements"                                                                          
# fxn_15092 "Electron transport and photophosphorylation"                                                    
# fxn_15419 "Protein biosynthesis"                                                                           
# fxn_15980 "Protein degradation"                                                                            
# fxn_16127 "Protein folding"                                                                                
# fxn_16186 "Protein processing and modification"                                                            
# fxn_16312 "Secretion"                                                                                      
# fxn_16327 "Selenoproteins"                                                                                 
# fxn_16730 "Programmed Cell Death and Toxin-antitoxin Systems"                                              
# fxn_16823 "Proteolytic pathway"                                                                            
# fxn_16830 "Quorum sensing and biofilm formation"                                                           
# fxn_16856 "Regulation of virulence"                                                                        
# fxn_16873 "Signal transduction in Eukaryotes"                                                              
# fxn_17102 "ATP synthases"                                                                                  
# fxn_17142 "Electron accepting reactions"                                                                   
# fxn_17747 "Electron donating reactions"                                                                    
# fxn_17991 "General Stress Response and Stationary Phase Response"                                          
# fxn_18009 "Mitochondrial electron transport system in plants"                                              
# fxn_18055 "Plastidial (cyanobacterial) electron transport system"                                          
# fxn_18084 "Reverse electron transport"                                                                     
# fxn_18110 "Sodium Ion-Coupled Energetics"                                                                  
# fxn_18145 "RNA processing and modification"                                                                
# fxn_18932 "Transcription"                                                                                  
# fxn_19090 "Bacterial cytostatics, differentiation factors and antibiotics"                                 
# fxn_19096 "Biologically active compounds in metazoan cell defence and differentiation"                     
# fxn_19098 "Biosynthesis of phenylpropanoids"                                                               
# fxn_19118 "Plant Alkaloids"                                                                                
# fxn_19121 "Plant Glucosinolates"                                                                           
# fxn_19123 "Plant Hormones"                                                                                 
# fxn_19330 "Acid stress"                                                                                    
# fxn_19335 "Cold shock"                                                                                     
# fxn_19345 "Dessication stress"                                                                             
# fxn_19366 "Detoxification"                                                                                 
# fxn_19493 "Heat shock"                                                                                     
# fxn_19598 "Osmotic stress"                                                                                 
# fxn_19674 "Oxidative stress"                                                                               
# fxn_19959 "Periplasmic Stress"                                                                             
# fxn_20079 "Inorganic sulfur assimilation"                                                                  
# fxn_20122 "Organic sulfur assimilation"                                                                    
# fxn_20321 "Fimbriae of the Chaperone/Usher Assembly Pathway"                                               
# fxn_20336 "Resistance to antibiotics and toxic compounds"                                                  
# fxn_20796 "Type III, Type IV, Type VI, ESAT secretion systems"                                             
# fxn_20931 "Adhesion"                                                                                       
# fxn_21001 "Bacteriocins, ribosomally synthesized antibacterial peptides"                                   
# fxn_21089 "Detection"                                                                                      
# fxn_21101 "Invasion and intracellular resistance"                                                          
# fxn_21111 "Toxins and superantigens"

length( unique( phy.ok@tax_table[ , "subsys_L3"] ) ) # 1149 

length( unique( phy.ok@tax_table[ , "fxn"] ) ) #11138

unique( phy.ok@tax_table[ , "fxn"] )[1:50]
# Taxonomy Table:     [50 taxa by 1 taxonomic ranks]:
#   fxn                                                                                                                                           
# fxn_1  "2-methylaconitate_cis-trans_isomerase"                                                                                                       
# fxn_2  "2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)"                                                               
# fxn_3  "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                                                                                     
# fxn_4  "Alanine_racemase_(EC_5.1.1.1)"                                                                                                               
# fxn_5  "Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic"                                                                                               
# fxn_6  "Alanine_racemase_(EC_5.1.1.1)_##_catabolic"                                                                                                  
# fxn_7  "Amino_acid_racemase_RacX"                                                                                                                    
# fxn_8  "Arginine_racemase_(EC_5.1.1.9)_@_Lysine_racemase_(EC_5.1.1.5)_@_Ornithine_racemase_(EC_5.1.1.12)"                                            
# fxn_9  "Aspartate_racemase_(EC_5.1.1.13)"                                                                                                            
# fxn_10 "Diaminopimelate_epimerase_(EC_5.1.1.7)"                                                                                                      
# fxn_11 "Glutamate_racemase_(EC_5.1.1.3)"                                                                                                             
# fxn_12 "Glutamate_racemase_(EC_5.1.1.3)_/_Nucleoside_5-triphosphatase_RdgB_(dHAPTP,_dITP,_XTP-specific)_(EC_3.6.1.15)"                               
# fxn_13 "hypothetical_protein"                                                                                                                        
# fxn_14 "Not_a_Proline_racemase,_nor_4-hydroxyproline_epimerase_[missing_catalytic_residues]"                                                         
# fxn_15 "Proline_racemase_(EC_5.1.1.4)"                                                                                                               
# fxn_16 "UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-alanine_ligase_(EC_6.3.2.10)_/_Alanine_racemase_(EC_5.1.1.1)"                                    
# fxn_17 "UDP-N-acetylmuramoylalanyl-D-glutamyl-2,6-diaminopimelate--D-alanyl-D-alanine_ligase_(EC_6.3.2.10)_/_Alanine_racemase_(EC_5.1.1.1)"          
# fxn_18 "Creatinase_(EC_3.5.3.3)"                                                                                                                     
# fxn_19 "Creatinine_amidohydrolase_(EC_3.5.2.10)"                                                                                                     
# fxn_20 "Cytosine_deaminase_(EC_3.5.4.1)"                                                                                                             
# fxn_21 "Cytosine_permease"                                                                                                                           
# fxn_22 "Cytosine/purine/uracil/thiamine/allantoin_permease_family_protein"                                                                           
# fxn_23 "Hydantoinase/oxoprolinase_family_protein"                                                                                                    
# fxn_25 "N-carbamoylsarcosine_amidase_(EC_3.5.1.59)"                                                                                                  
# fxn_26 "N-methylhydantoinase_(ATP-hydrolyzing)_(EC_3.5.2.14)"                                                                                        
# fxn_27 "N-methylhydantoinase_A_(EC_3.5.2.14)"                                                                                                        
# fxn_28 "N-methylhydantoinase_B_(EC_3.5.2.14)"                                                                                                        
# fxn_29 "Sarcosine_dehydrogenase_(EC_1.5.99.1)"                                                                                                       
# fxn_30 "Allantoate_amidohydrolase_(EC_3.5.3.9)"                                                                                                      
# fxn_31 "Beta-ureidopropionase_(EC_3.5.1.6)"                                                                                                          
# fxn_32 "L-2-amino-thiazoline-4-carboxylic_acid_hydrolase_(EC_3.5.2.-)"                                                                               
# fxn_33 "N-carbamoyl-L-amino_acid_hydrolase_(EC_3.5.1.87);_Beta-ureidopropionase_(EC_3.5.1.6)"                                                        
# fxn_35 "Alanine_racemase_(EC_5.1.1.1)_#_present_in_exosporium,_involved_in_spore_germination"                                                        
# fxn_38 "Alanine_racemase,_biosynthetic_(EC_5.1.1.1)"                                                                                                 
# fxn_39 "Alanine_racemase,_catabolic_(EC_5.1.1.1)"                                                                                                    
# fxn_40 "Alanine_transaminase_(EC_2.6.1.2)"                                                                                                           
# fxn_41 "Alanine_transaminase_(EC_2.6.1.2)_/_Glutamate-glyoxylate_aminotransferase_(EC_2.6.1.4)_/_L-alanine:glyoxylate_aminotransferase_(EC_2.6.1.44)"
# fxn_42 "Alanine_transaminase_(EC_2.6.1.2)_##_AlaA"                                                                                                   
# fxn_43 "Alpha-aminoadipate_aminotransferase_(EC_2.6.1.39)_@_Leucine_transaminase_(EC_2.6.1.6)_@_Valine_transaminase"                                 
# fxn_44 "Aminodeoxychorismate_lyase_(EC_4.1.3.38)"                                                                                                    
# fxn_45 "aminotransferase,_class_IV"                                                                                                                  
# fxn_46 "Branched-chain_amino_acid_aminotransferase_(EC_2.6.1.42)"                                                                                    
# fxn_47 "Cysteine_desulfurase_(EC_2.8.1.7)"                                                                                                           
# fxn_48 "Cysteine_desulfurase_(EC_2.8.1.7),_IscS_subfamily"                                                                                           
# fxn_49 "Cysteine_desulfurase_(EC_2.8.1.7),_NifS_subfamily"                                                                                           
# fxn_50 "Cysteine_desulfurase_(EC_2.8.1.7),_SufS_subfamily"                                                                                           
# fxn_51 "Cysteine_desulfurase_(EC_2.8.1.7),_SufS_subfamily;_Selenocysteine_lyase(_EC:4.4.1.16_)_subunit"                                              
# fxn_52 "Cysteine_desulfurase_CsdA-CsdE_(EC_2.8.1.7),_main_protein_CsdA"                                                                              
# fxn_53 "D-amino_acid_dehydrogenase_small_subunit_(EC_1.4.99.1)_/_Alanine_racemase_(EC_5.1.1.1)"                                                      
# fxn_54 "flagellar_associated_protein"




#-------------------------

#### Jie ACVD - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 21128    4


# this function is used later

atomic_no
# function(form, atom) {
#   ##form=form
#   ##atom="O"
#   coefs <- list()
#   for (c in 1:length(form)) {
#     #c<-1
#     pos <- str_locate(string = form[c], pattern = atom)[1]
#     if (is.na(pos)) { # not present
#       ##print("zero")
#       coef<-0
#     } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A 
#       ##print("A")
#       coef<-1
#     } else {
#       # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
#       ##print("B")
#       checkstring <- substring(text = form[c], first = pos, last = pos+1)
#       
#       if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
#         ##print("C")
#         coef<-0
#       } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
#         ##print("D")
#         coef<-0
#       } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E 
#         ##print("E")
#         coef<-0
#       } else { # F
#         ##print("F")
#         coef<-1
#         coef.try <- coef
#         pos.end <- pos+1
#         
#         while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
#           # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
#           coef <- coef.try
#           coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
#           pos.end <- pos.end+1
#           ##print("G")
#         } # END while loop
#       } # END else
#       
#     } # END else
#     
#     coefs[[c]] <- coef
#     
#   } # END c loop
#   return( unlist(coefs) )
# }
# <bytecode: 0x7fcaee74aa70>





get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)

        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1

          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else

      } # END else

      coefs[[c]] <- coef

    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  
  
  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      

    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Jie-ACVD-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row

#time.finish <- Sys.time()


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()



time.start #  "2023-04-05 16:57:15 ACST"
time.finish # "2023-04-05 18:37:32 ACST"

# 1.5 hr

## assemble results


modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Jie-ACVD-R-working-files"



# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# total results written to disk
dim(df.tax) # 21128    4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 24851     9

table(df.out$matching_method)
# EC number      Exact hierachy match Matched Reactions aliases    Matched Reactions name     Matched Subsytem role 
# 9027                      3892                        48                       389                       611 
# No match found 
# 10884 

100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 43.79703% No match found

# 56% match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.617   0.810   0.859   1.078   2.500   10015

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.256   1.401   1.423   1.589   4.000   10015

table(df.out.distil$matching_method)
# EC number                       EC number;Exact hierachy match EC number;Exact hierachy match;Matched Subsytem role 
# 6516                                                  624                                                    2 
# EC number;Exact hierachy match;No match found                  EC number;Matched Reactions aliases                     EC number;Matched Reactions name 
# 17                                                    6                                                    7 
# EC number;Matched Subsytem role       EC number;Matched Subsytem role;No match found                             EC number;No match found 
# 71                                                    2                                                  535 
# Exact hierachy match       Exact hierachy match;Matched Reactions aliases          Exact hierachy match;Matched Reactions name 
# 2612                                                    2                                                    7 
# Exact hierachy match;Matched Subsytem role                  Exact hierachy match;No match found                            Matched Reactions aliases 
# 24                                                  222                                                   39 
# Matched Reactions aliases;No match found                               Matched Reactions name                Matched Reactions name;No match found 
# 1                                                  370                                                    5 
# Matched Subsytem role                 Matched Subsytem role;No match found                                       No match found 
# 446                                                   55                                                 9565 

100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 38.4894% No match found

100-38.5 # i.e. in 61.5% of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_Jie-ACVD.RDS")

df.out.distil <- readRDS(file = "df.out.distil_Jie-ACVD.RDS")



#-------------------------

#### Jie ACVD - inspect & clean data
#-------------------------

this_study <- "-Jie-ACVD-"
phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Jie-ACVD.RDS")

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 21128    4


phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE

df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 21128  385
df.OTU[1:5, 1:8]
# ERR2017411 ERR2017412 ERR2017413 ERR2017414 ERR2017415 ERR2017416 ERR2017417 ERR2017418
# fxn_1          0          0          0          0          0          0          0          0
# fxn_2          0          0          0          0          0          0          0          0
# fxn_3          0          0          0          0          0          0          0          0
# fxn_4          0          0          0          0          0          0          0          0
# fxn_5          0          0          0          0          0          0          0          0

mean( df.OTU[ , "ERR2017411"] ) # 0.004733056
sum( df.OTU[ , "ERR2017411"] ) # 100

sample_sums(phy) # all values of 100


identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
# tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.9285714     0.8571429
# fxn_3     0.6000000     1.8000000
# fxn_4     0.6666667     2.3333333
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333



dim(df.fxn_vk_coords) # 21128     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.617   0.810   0.859   1.078   2.500   10015
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.256   1.401   1.423   1.589   4.000   10015

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.9285714     0.8571429
# fxn_3     0.6000000     1.8000000
# fxn_4     0.6666667     2.3333333
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333
# fxn_8     0.4831893     2.2257060


## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 385
dim(df.fxn_vk_coords.noNAs) # 11113     2
dim(df.OTU.noNAs) # 11113   385

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# ERR2017433 ERR2017417 ERR2017431 ERR2017707 ERR2017427 ERR2017608 ERR2017523 ERR2017500 ERR2017710 ERR2017696 ERR2017586 ERR2017713 ERR2017706 
# 54.44444   55.35714   55.55556   56.41089   56.52174   57.12776   57.53800   57.62287   58.18820   58.21867   58.25494   58.34585   58.48193 
# ERR2017726 ERR2017674 ERR2017536 ERR2017626 ERR2017709 ERR2017526 ERR2017633 ERR2017776 ERR2017518 ERR2017711 ERR2017520 ERR2017545 ERR2017766 
# 58.48194   58.54518   58.64235   58.64576   58.64808   58.69266   58.69797   58.89809   58.90825   58.95530   58.96473   59.04691   59.10599 
# ERR2017780 ERR2017759 ERR2017794 ERR2017728 ERR2017743 ERR2017699 ERR2017810 ERR2017482 ERR2017693 ERR2017717 ERR2017461 ERR2017751 ERR2017429 
# 59.14649   59.18213   59.20153   59.25159   59.25462   59.35249   59.36907   59.38578   59.40595   59.42861   59.44459   59.45795   59.45946 
# ERR2017670 ERR2017466 ERR2017566 ERR2017469 ERR2017802 ERR2017724 ERR2017552 ERR2017755 ERR2017739 ERR2017676 ERR2017720 ERR2017597 ERR2017641 
# 59.47841   59.50664   59.51101   59.51859   59.55382   59.56562   59.59732   59.60270   59.60494   59.64332   59.65434   59.66166   59.66822 
# ERR2017472 ERR2017730 ERR2017623 ERR2017475 ERR2017551 ERR2017725 ERR2017687 ERR2017712 ERR2017731 ERR2017644 ERR2017792 ERR2017468 ERR2017549 
# 59.75004   59.79181   59.81088   59.82144   59.87969   59.98736   60.02969   60.06000   60.07356   60.10567   60.12360   60.14166   60.14453 
# ERR2017543 ERR2017548 ERR2017799 ERR2017464 ERR2017777 ERR2017621 ERR2017815 ERR2017701 ERR2017756 ERR2017797 ERR2017437 ERR2017702 ERR2017538 
# 60.14804   60.15355   60.17159   60.20941   60.21207   60.21701   60.24933   60.27630   60.28492   60.28604   60.31746   60.38469   60.39633 
# ERR2017692 ERR2017413 ERR2017753 ERR2017682 ERR2017723 ERR2017580 ERR2017553 ERR2017521 ERR2017714 ERR2017483 ERR2017735 ERR2017473 ERR2017774 
# 60.40669   60.41667   60.42307   60.42371   60.45103   60.45780   60.46653   60.55787   60.57753   60.59141   60.59640   60.60483   60.63049 
# ERR2017689 ERR2017804 ERR2017495 ERR2017603 ERR2017708 ERR2017773 ERR2017556 ERR2017691 ERR2017575 ERR2017525 ERR2017619 ERR2017524 ERR2017607 
# 60.64090   60.68850   60.69096   60.74578   60.78619   60.78848   60.81582   60.83670   60.84214   60.85028   60.85035   60.87998   60.91561 
# ERR2017721 ERR2017680 ERR2017741 ERR2017606 ERR2017627 ERR2017760 ERR2017600 ERR2017470 ERR2017685 ERR2017569 ERR2017795 ERR2017700 ERR2017571 
# 60.92110   60.95037   61.01286   61.04252   61.04693   61.05863   61.10616   61.19191   61.22488   61.24423   61.25111   61.25253   61.26556 
# ERR2017557 ERR2017772 ERR2017733 ERR2017533 ERR2017467 ERR2017476 ERR2017506 ERR2017465 ERR2017803 ERR2017649 ERR2017767 ERR2017738 ERR2017510 
# 61.27286   61.27564   61.30637   61.35663   61.35808   61.36096   61.36251   61.37931   61.39459   61.43110   61.45114   61.45597   61.45915 
# ERR2017737 ERR2017679 ERR2017628 ERR2017595 ERR2017471 ERR2017522 ERR2017778 ERR2017581 ERR2017474 ERR2017763 ERR2017489 ERR2017513 ERR2017515 
# 61.49438   61.52212   61.54035   61.62880   61.65961   61.71541   61.72367   61.73193   61.73306   61.73428   61.74449   61.74953   61.78011 
# ERR2017716 ERR2017715 ERR2017577 ERR2017499 ERR2017697 ERR2017578 ERR2017613 ERR2017530 ERR2017512 ERR2017599 ERR2017554 ERR2017550 ERR2017652 
# 61.78041   61.78282   61.79591   61.81088   61.84912   61.85115   61.85941   61.88926   61.90482   61.92690   61.93259   61.93370   61.95487 
# ERR2017769 ERR2017788 ERR2017529 ERR2017564 ERR2017462 ERR2017744 ERR2017690 ERR2017783 ERR2017463 ERR2017558 ERR2017747 ERR2017516 ERR2017532 
# 61.96992   62.00841   62.07819   62.13052   62.13538   62.15332   62.15669   62.17749   62.20828   62.21614   62.24870   62.25117   62.25171 
# ERR2017688 ERR2017646 ERR2017785 ERR2017698 ERR2017734 ERR2017660 ERR2017806 ERR2017793 ERR2017718 ERR2017781 ERR2017491 ERR2017562 ERR2017681 
# 62.26095   62.27610   62.28144   62.29009   62.33152   62.37172   62.51443   62.51597   62.53049   62.60932   62.63282   62.66416   62.71221 
# ERR2017750 ERR2017704 ERR2017542 ERR2017487 ERR2017418 ERR2017615 ERR2017609 ERR2017614 ERR2017694 ERR2017705 ERR2017555 ERR2017421 ERR2017610 
# 62.71655   62.75431   62.78014   62.78431   62.79070   62.79363   62.87046   62.87271   62.87582   62.89129   62.89864   62.92683   62.98401 
# ERR2017809 ERR2017477 ERR2017632 ERR2017504 ERR2017808 ERR2017629 ERR2017590 ERR2017664 ERR2017594 ERR2017784 ERR2017807 ERR2017546 ERR2017786 
# 62.98642   62.99444   63.02400   63.02616   63.03633   63.08369   63.12453   63.14468   63.15675   63.15865   63.16823   63.17354   63.17387 
# ERR2017563 ERR2017754 ERR2017591 ERR2017636 ERR2017654 ERR2017612 ERR2017742 ERR2017736 ERR2017650 ERR2017647 ERR2017517 ERR2017435 ERR2017686 
# 63.17851   63.17868   63.22652   63.23203   63.23430   63.24208   63.26280   63.26624   63.29894   63.33691   63.37190   63.37209   63.40597 
# ERR2017805 ERR2017653 ERR2017508 ERR2017478 ERR2017757 ERR2017587 ERR2017514 ERR2017663 ERR2017771 ERR2017568 ERR2017796 ERR2017616 ERR2017622 
# 63.43972   63.44605   63.55341   63.58742   63.60832   63.62658   63.67879   63.68864   63.70397   63.71585   63.75921   63.77642   63.79765 
# ERR2017559 ERR2017722 ERR2017762 ERR2017791 ERR2017695 ERR2017790 ERR2017640 ERR2017484 ERR2017801 ERR2017565 ERR2017531 ERR2017486 ERR2017503 
# 63.83755   63.86165   63.88558   63.97370   63.98670   64.01981   64.03921   64.05026   64.06466   64.06556   64.07068   64.11967   64.13207 
# ERR2017604 ERR2017659 ERR2017584 ERR2017618 ERR2017414 ERR2017481 ERR2017703 ERR2017748 ERR2017768 ERR2017535 ERR2017479 ERR2017527 ERR2017537 
# 64.14057   64.14517   64.15228   64.16763   64.18919   64.21279   64.22355   64.23399   64.27835   64.28256   64.33370   64.34265   64.37506 
# ERR2017576 ERR2017779 ERR2017746 ERR2017519 ERR2017719 ERR2017764 ERR2017684 ERR2017547 ERR2017637 ERR2017643 ERR2017749 ERR2017651 ERR2017789 
# 64.38792   64.45121   64.46949   64.47952   64.49619   64.53410   64.55267   64.55510   64.58002   64.63477   64.65503   64.66424   64.66599 
# ERR2017412 ERR2017656 ERR2017541 ERR2017666 ERR2017624 ERR2017494 ERR2017732 ERR2017761 ERR2017675 ERR2017798 ERR2017658 ERR2017534 ERR2017573 
# 64.70588   64.71583   64.72561   64.80969   64.82996   64.88860   64.92188   64.95196   65.11395   65.12451   65.12943   65.14540   65.14651 
# ERR2017570 ERR2017729 ERR2017507 ERR2017579 ERR2017765 ERR2017814 ERR2017485 ERR2017667 ERR2017605 ERR2017419 ERR2017592 ERR2017574 ERR2017775 
# 65.20329   65.20945   65.22235   65.24391   65.25877   65.28186   65.34222   65.36058   65.41491   65.41587   65.43034   65.46922   65.46979 
# ERR2017661 ERR2017572 ERR2017635 ERR2017617 ERR2017727 ERR2017648 ERR2017770 ERR2017593 ERR2017668 ERR2017588 ERR2017662 ERR2017560 ERR2017540 
# 65.48203   65.50652   65.51630   65.54620   65.57679   65.59026   65.59339   65.64942   65.74438   65.75019   65.76608   65.88621   65.91306 
# ERR2017420 ERR2017639 ERR2017752 ERR2017598 ERR2017634 ERR2017683 ERR2017544 ERR2017601 ERR2017800 ERR2017665 ERR2017561 ERR2017672 ERR2017611 
# 66.00000   66.00171   66.05772   66.07274   66.12527   66.17699   66.18036   66.19696   66.20153   66.20240   66.26626   66.30252   66.32638 
# ERR2017457 ERR2017490 ERR2017638 ERR2017456 ERR2017539 ERR2017450 ERR2017673 ERR2017567 ERR2017631 ERR2017453 ERR2017655 ERR2017441 ERR2017782 
# 66.57503   66.58184   66.61930   66.69014   66.86264   66.88954   66.89454   67.08636   67.17301   67.19705   67.22299   67.25161   67.26878 
# ERR2017528 ERR2017671 ERR2017620 ERR2017678 ERR2017439 ERR2017625 ERR2017422 ERR2017645 ERR2017745 ERR2017740 ERR2017589 ERR2017630 ERR2017459 
# 67.27282   67.34295   67.40276   67.41110   67.41342   67.42942   67.46135   67.46888   67.63643   67.68875   67.72385   67.72777   67.78404 
# ERR2017440 ERR2017452 ERR2017657 ERR2017669 ERR2017602 ERR2017677 ERR2017415 ERR2017442 ERR2017596 ERR2017444 ERR2017448 ERR2017813 ERR2017430 
# 67.80855   67.81319   67.90090   67.97113   67.98246   68.04172   68.10345   68.18450   68.22273   68.38549   68.53587   68.55593   68.55611 
# ERR2017642 ERR2017458 ERR2017455 ERR2017446 ERR2017411 ERR2017443 ERR2017449 ERR2017438 ERR2017451 ERR2017460 ERR2017434 ERR2017424 ERR2017432 
# 68.59999   68.71398   68.99694   69.06261   69.24735   69.71993   69.74667   69.83592   69.92553   70.36265   70.83333   71.11111   71.15385 
# ERR2017454 ERR2017423 ERR2017445 ERR2017428 ERR2017416 ERR2017447 ERR2017436 ERR2017426 
# 71.22081   71.62162   71.87580   72.39583   73.07692   74.35571   79.16667   84.09091

summary( colSums(df.OTU.noNAs) )
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54.44   60.79   62.90   63.25   65.22   84.09

# this reflects that 54-84% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

## PREVIOUSLY for env/soil sample
## this reflects the just >60% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

colSums(df.OTU)
# all values of 100


dim(df.tax) # 21128     4


# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE


dim(df.fxn_vk_coords.noNAs) # 11113     2
dim(df.in) # 11113  385

row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  group=NA, 
                                  samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"],
                         group= rep( phy@sam_data$`ACVD status`[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}



dim(vk.samp.noNAs.long) # 4278506       8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok)

str(vk.samp.noNAs.long)
unique(vk.samp.noNAs.long$group) # 0 1
class(vk.samp.noNAs.long$group) # "numeric"
vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c(1,0), labels = c("ACVD", "Normal"), ordered = TRUE) # CHECK GROUP VARIABLE !!
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	4278505 obs. of  8 variables:
# $ sample   : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ fxn_id   : Factor w/ 11113 levels "fxn_10","fxn_100",..: 5124 6258 7065 7834 8501 9568 10365 1 469 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.929 0.6 0.667 0.667 0.667 ...
# $ HC_y     : num  0.857 1.8 2.333 2.333 2.333 ...
# $ group    : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" ...


saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-object-Jie-ACVD.RDS")


#vk.samp.noNAs.long <- readRDS(file = "vk.samp.noNAs.long-object-Jie-ACVD.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.text = element_text(angle=25),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "ACVD & Normal human gut", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust =0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")




## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Jie-ACVD-R-working-files"

i<-2
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
names(temp)
names( temp[[1]] ) # "fxns" "rxns"
temp[[1]][["fxns"]]
# superfocus_fxn f                                                                         f__in matching_method min_adist_modelSEED min_amatch_modelSEED     rxns
# 1          fxn_2 1 2-methylcitrate dehydratase (2-methyl-trans-aconitate forming) (EC 4.2.1.117)       EC number                  NA                   NA rxn25279
# tot_mean_OC_x tot_mean_HC_y
# 1    0.9285714     0.8571429
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                                                                         f__in   rxn_id                                                       rxn_name
# 1          fxn_2 1 2-methylcitrate dehydratase (2-methyl-trans-aconitate forming) (EC 4.2.1.117) rxn25279 2-methylcitrate dehydratase .2-methyl-trans-aconitate forming.
# rxn_eqn                                                                                             rxn_defn
# 1 (1) cpd24620[0] <=> (1) cpd00001[0] + (1) cpd25681[0] (1) (2S,3S)-2-hydroxybutane-1,2,3-tricarboxylate[0] <=> (1) H2O[0] + (1) 2-methyl-trans-aconitate[0]
# compds compd_coef        chem_formx              OC_ratios              HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 cpd24620;cpd00001;cpd25681      1;1;1 C7H7O7;H2O;C7H5O6 1;NA;0.857142857142857 1;NA;0.714285714285714       0.9285714       0.8571429

i<-99
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
names(temp)
names( temp[[1]] ) # "fxns" "rxns"
temp[[1]][["fxns"]]
# superfocus_fxn f                                                        f__in      matching_method min_adist_modelSEED min_amatch_modelSEED              rxns tot_mean_OC_x tot_mean_HC_y
# 1         fxn_99 1             L-serine dehydratase, beta subunit (EC 4.3.1.17) Exact hierachy match                  NA                   NA          rxn00165     1.0000000      1.666667
# 2         fxn_99 2 L-serine dehydratase, alpha subunit (EC 4.3.1.17) # fragment            EC number                  NA                   NA rxn00426;rxn14245     0.8333333      1.666667

temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                                            f__in   rxn_id               rxn_name                                              rxn_eqn                                        rxn_defn
# 1         fxn_99 1 L-serine dehydratase, beta subunit .EC 4.3.1.17. rxn00165 L-serine ammonia-lyase (1) cpd00054[0] => (1) cpd00013[0] + (1) cpd00020[0] (1) L-Serine[0] => (1) NH3[0] + (1) Pyruvate[0]
# compds compd_coef         chem_formx OC_ratios             HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 cpd00054;cpd00013;cpd00020      1;1;1 C3H7NO3;H4N;C3H3O3   1;NaN;1 2.33333333333333;NA;1               1        1.666667
# 
# [[2]]
# superfocus_fxn f                                                        f__in   rxn_id                       rxn_name                                                                rxn_eqn
# 1         fxn_99 2 L-serine dehydratase, alpha subunit (EC 4.3.1.17) # fragment rxn00426           L-Serine hydro-lyase                  (1) cpd00054[0] <=> (1) cpd00001[0] + (1) cpd01495[0]
# 2         fxn_99 2 L-serine dehydratase, alpha subunit (EC 4.3.1.17) # fragment rxn14245 2-Aminoacrylate aminohydrolase (1) cpd00001[0] + (1) cpd01495[0] => (1) cpd00013[0] + (1) cpd00020[0]
# rxn_defn                              compds compd_coef             chem_formx                  OC_ratios                            HC_ratios
# 1             (1) L-Serine[0] <=> (1) H2O[0] + (1) Dehydroalanine[0]          cpd00054;cpd00001;cpd01495      1;1;1    C3H7NO3;H2O;C3H5NO2     1;NA;0.666666666666667 2.33333333333333;NA;1.66666666666667
# 2 (1) H2O[0] + (1) Dehydroalanine[0] => (1) NH3[0] + (1) Pyruvate[0] cpd00001;cpd01495;cpd00013;cpd00020    1;1;1;1 H2O;C3H5NO2;H4N;C3H3O3 NA;0.666666666666667;NaN;1             NA;1.66666666666667;NA;1
# coefwtmean_OC_x coefwtmean_HC_y
# 1       0.8333333        2.000000
# 2       0.8333333        1.333333



## check anomalous high value in Lignin zone in PCaN density plot
p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405)

length(sel) # 107415
dim(vk.samp.noNAs.long) # 4278505       8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 2.510573% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
# sample  fxn_id abun abun_type      OC_x     HC_y  group      samplabel
# 12  ERR2017411  fxn_13    0   relabun 0.5206427 1.400701 Normal ERR2017411 (0)
# 20  ERR2017411  fxn_24    0   relabun 0.5206427 1.400701 Normal ERR2017411 (0)
# 73  ERR2017411  fxn_92    0   relabun 0.5206427 1.400701 Normal ERR2017411 (0)
# 94  ERR2017411 fxn_121    0   relabun 0.5206427 1.400701 Normal ERR2017411 (0)
# 112 ERR2017411 fxn_142    0   relabun 0.5206427 1.400701 Normal ERR2017411 (0)
# 181 ERR2017411 fxn_236    0   relabun 0.5206427 1.400701 Normal ERR2017411 (0)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.04515317 = 0.045% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 61.75283 = 61.8% per sample

# remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702) # qty 9633


#fxn_13
i<-13
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED                                                                             rxns tot_mean_OC_x
# 1         fxn_22 1 hypothetical protein Matched Reactions name                  NA                   NA rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427
# tot_mean_HC_y
# 1      1.400701
temp[[1]][["rxns"]]
# [[1]]
#   superfocus_fxn f                f__in   rxn_id                                                      rxn_name
# 1         fxn_22 1 hypothetical protein rxn01869     CHLREDRAFT_140509 hypothetical protein  .SP:A8JHB7_CHLRE.
# 2         fxn_22 1 hypothetical protein rxn05992     CHLREDRAFT_142823 hypothetical protein  .SP:A8I7C1_CHLRE.
# 3         fxn_22 1 hypothetical protein rxn06889     CHLREDRAFT_121409 hypothetical protein  .SP:A8JB38_CHLRE.
# 4         fxn_22 1 hypothetical protein rxn11873     CHLREDRAFT_187700 hypothetical protein  .SP:A8I0K5_CHLRE.
# 5         fxn_22 1 hypothetical protein rxn11874     CHLREDRAFT_187700 hypothetical protein  .SP:A8I0K5_CHLRE.
# 6         fxn_22 1 hypothetical protein rxn31711 glycosyltransferase family protein 28 or hypothetical protein
# 7         fxn_22 1 hypothetical protein rxn34308     CHLREDRAFT_121409 hypothetical protein  .SP:A8JB38_CHLRE.
# 8         fxn_22 1 hypothetical protein rxn34312     CHLREDRAFT_142782 hypothetical protein  .SP:A8I707_CHLRE.
# 9         fxn_22 1 hypothetical protein rxn34322     CHLREDRAFT_186686 hypothetical protein  .SP:A8I9K8_CHLRE.
# rxn_eqn
# 1 (1) cpd00001[0] + (1) cpd00015[0] + (1) cpd00447[0] => (1) cpd00067[0] + (1) cpd00540[0] + (1) cpd00982[0]
# 2                                    (1) cpd00144[0] + (1) cpd11609[0] <=> (1) cpd00014[0] + (1) cpd12241[0]
# 3                                                       (1) cpd04016[0] => (2) cpd00067[0] + (1) cpd15064[0]
# 4                                                                         (1) cpd04740[0] => (1) cpd16416[0]
# 5                                                                         (1) cpd16412[0] => (1) cpd16416[0]
# 6                                     (1) cpd00037[0] + (1) cpd12559[0] <= (1) cpd00014[0] + (1) cpd29932[0]
# 7                                                       (1) cpd04015[0] <= (2) cpd00067[0] + (1) cpd30045[0]
# 8                                     (1) cpd12407[0] + (1) cpd30050[0] <= (1) cpd11619[0] + (1) cpd29780[0]
# 9                                     (1) cpd11456[0] + (1) cpd29781[0] <= (1) cpd11423[0] + (1) cpd30057[0]
# rxn_defn                                                compds  compd_coef
# 1                     (1) H2O[0] + (1) FAD[0] + (1) Betaine aldehyde[0] => (1) H+[0] + (1) BET[0] + (1) FADH2[0] cpd00001;cpd00015;cpd00447;cpd00067;cpd00540;cpd00982 1;1;1;1;1;1
# 2                                           (1) UDPglucuronate[0] + (1) A[0] <=> (1) UDP[0] + (1) Glucuronide[0]                   cpd00144;cpd11609;cpd00014;cpd12241     1;1;1;1
# 3                            (1) cis-3-Chloro-2-propene-1-ol[0] => (2) H+[0] + (1) cis-3-Chloroallyl aldehyde[0]                            cpd04016;cpd00067;cpd15064       1;2;1
# 4                                                                (1) Citalopram[0] => (1) Citalopram aldehyde[0]                                     cpd04740;cpd16416         1;1
# 5                                                        (1) Demethylcitalopram[0] => (1) Citalopram aldehyde[0]                                     cpd16412;cpd16416         1;1
# 6 (1) UDP-N-acetylglucosamine[0] + (1) N-Acetyl-D-glucosaminyldiphosphodolichol[0] <= (1) UDP[0] + (1) G00002[0]                   cpd00037;cpd12559;cpd00014;cpd29932     1;1;1;1
# 7                              (1) trans-3-Chloro-2-propene-1-ol[0] <= (2) H+[0] + (1) 3-Chloroallyl aldehyde[0]                            cpd04015;cpd00067;cpd30045       1;2;1
# 8               (1) Dolichyl phosphate D-mannose[0] + (1) G00147[0] <= (1) Dolichyl phosphate[0] + (1) G00148[0]                   cpd12407;cpd30050;cpd11619;cpd29780     1;1;1;1
# 9               (1) Phosphatidylethanolamine[0] + (1) G00149[0] <= (1) 1,2-Diacyl-sn-glycerol[0] + (1) G13044[0]                   cpd11456;cpd29781;cpd11423;cpd30057     1;1;1;1
# chem_formx                                          OC_ratios                                              HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 H2O;C27H31N9O15P2;C5H12NO;H;C5H11NO2;C27H33N9O15P2 NA;0.555555555555556;0.2;NaN;0.4;0.555555555555556        NA;1.14814814814815;2.4;NA;2.2;1.22222222222222      0.42777778       1.7425926
# 2            C15H19N2O18P2;H2NR;C9H12N2O12P2;C6H8O7R          1.2;NaN;1.33333333333333;1.16666666666667  1.26666666666667;NA;1.33333333333333;1.33333333333333      1.23333333       1.3111111
# 3                                  C3H5ClO;H;C3H3ClO            0.333333333333333;NaN;0.333333333333333                                  1.66666666666667;NA;1      0.33333333       1.3333333
# 4                              C20H22FN2O;C18H14FNO2                             0.05;0.111111111111111                                  1.1;0.777777777777778      0.08055556       0.9388889
# 5                              C19H20FN2O;C18H14FNO2               0.0526315789473684;0.111111111111111                     1.05263157894737;0.777777777777778      0.08187135       0.9152047
# 6       C17H25N3O17P2;C33H57NO12P2;C9H12N2O12P2;null           1;0.363636363636364;1.33333333333333;NaN 1.47058823529412;1.72727272727273;1.33333333333333;NaN      0.89898990       1.5103981
# 7                                     C3H5ClO;H;null                          0.333333333333333;NaN;NaN                                1.66666666666667;NA;NaN      0.33333333       1.6666667
# 8                      C31H54O9P;null;C25H43O4P;null                     0.290322580645161;NaN;0.16;NaN                          1.74193548387097;NaN;1.72;NaN      0.22516129       1.7309677
# 9                     C7H12NO8PR2;null;C5H6O5R2;null                         1.14285714285714;NaN;1;NaN                           1.71428571428571;NaN;1.2;NaN      1.07142857       1.4571429



### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 107415


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-Jie-ACVD.RDS")


#-------------------------

#### Jie ACVD - CPP-class: total rel abun in vk spatial regions
#-------------------------

this_study <- "-Jie-ACVD-"
phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Jie-ACVD.RDS")

df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-Jie-ACVD.RDS")

# reset using clean data
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

header <- "-vk-zones-"


# from earlier
str(vk.samp.noNAs.long)
# 'data.frame':	4171090 obs. of  8 variables:
# $ sample   : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ fxn_id   : Factor w/ 11113 levels "fxn_10","fxn_100",..: 5124 6258 7065 7834 8501 9568 10365 1 469 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.929 0.6 0.667 0.667 0.667 ...
# $ HC_y     : num  0.857 1.8 2.333 2.333 2.333 ...
# $ group    : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" ...


# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 


spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]
# add sex
spdf$sex <- NA
samps <- unique(spdf$sample)
length(samps) # 385
length(which(phy@sam_data$Run %in% samps)) # 385
for (i in 1:length(samps)) {
  #i<-1
  this_samp <- samps[i]
  sel.spdf <- which(spdf$sample == this_samp)
  sel.meta <- which(phy@sam_data$Run == this_samp)
  spdf$sex[sel.spdf] <- phy@sam_data$Gender[sel.meta]
  print(paste0("completed ",i," - no rows = ",length(sel.spdf)))
}

dim(spdf) #  1945785       6

# any samples with unknown sex?
table(spdf$sex)
# female    male      NA 
# 716223 1203146   26416 
sel.na <- which(spdf$sex=="NA") # 26834 rows
unique(spdf$sample[sel.na]) # "ERR2017473" "ERR2017778" "ERR2017779" "ERR2017780" "ERR2017808" - 5 samples without gender data
spdf <- spdf[-sel.na, ]

head(spdf)
# sample  group      OC_x     HC_y        abun  sex
# 41  ERR2017411 Normal 0.6722222 1.511111 0.004222260 male
# 65  ERR2017411 Normal 1.5000000 1.130952 0.023644655 male
# 74  ERR2017411 Normal 1.0000000 1.666667 0.031244722 male
# 78  ERR2017411 Normal 1.0000000 1.666667 0.060800540 male
# 95  ERR2017411 Normal 0.7083333 1.630952 0.042222598 male
# 113 ERR2017411 Normal 1.0000000 1.666667 0.001688904 male

dim(spdf) #   1919369       6


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 380

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54.44   60.78   62.89   63.22   65.19   84.09 

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"   

## now want to sum total rela abun within vK spatial zones

## to de-duplicate get total relabun within a sample, then mean across samples

sum_relabun_in_vk_zones <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  # lipids
  subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  subsel.all <- c(subsel.lipids)
  # proteins
  subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.proteins)
  # amino_sugars
  subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.amino_sugars)
  # carbs - Carbohydrates
  subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  subsel.all <- c(subsel.all, subsel.carbs)
  # cond_aromtx - Condensed aromatics
  subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  subsel.all <- c(subsel.all, subsel.cond_aromtx)
  # lignin
  subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.lignin)
  # tannins
  subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.tannins)
  # Other compounds ...
  
  # methylated
  subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
  # and not already used?
  sel.dup <- which(subsel.methylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.methylated)
  
  # hydrated
  subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
  # and not already used?
  sel.dup <- which(subsel.hydrated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.hydrated)
  
  # demethylated / Oxidised-dehydrogenated
  subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.demethylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.demethylated)
  
  # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
  subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
  
  # and not already used?
  sel.dup <- which(subsel.condensed %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.condensed)
  
  # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
  subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.k2_b12 %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
  
  df_out <- data.frame(
    sample = this_samp,
    group = unique(spdf_in$group),
    sex = unique(spdf_in$sex),
    
    lipids = sum(spdf_in$abun[ subsel.lipids ]),
    proteins = sum(spdf_in$abun[ subsel.proteins ]),
    amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
    carbs = sum(spdf_in$abun[ subsel.carbs ]),
    cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
    lignin = sum(spdf_in$abun[ subsel.lignin ]),
    tannins = sum(spdf_in$abun[ subsel.tannins ]),
    # Other compounds ...
    methylated = sum(spdf_in$abun[ subsel.methylated ]),
    hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
    demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
    condensed = sum(spdf_in$abun[ subsel.condensed ]),
    k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
    
  )
  return(df_out)
}



time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 380
out[[1]]

names(out[[1]])
# [1] "sample"       "group"        "sex"          "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx" 
# [9] "lignin"       "tannins"      "methylated"   "hydrated"     "demethylated" "condensed"    "k2_b12"

df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample  group    sex lipids  proteins amino_sugars    carbs cond_aromtx    lignin   tannins methylated hydrated demethylated condensed    k2_b12
# 1 ERR2017411 Normal   male      0  2.829638     3.035805 34.40339  0.05911164  3.682655 12.756445   1.659348 3.035805     6.552243   0.87823 0.3504476
# 2 ERR2017412 Normal   male      0 11.764706     0.000000 29.41176  0.00000000 11.764706  5.882353   0.000000 0.000000     5.882353   0.00000 0.0000000
# 3 ERR2017413 Normal   male      0  0.000000     0.000000 41.66667  0.00000000  6.250000 10.416667   0.000000 0.000000     2.083333   0.00000 0.0000000
# 4 ERR2017414 Normal female      0  2.702703     2.702703 25.67568  0.00000000  7.357357 10.885886   0.000000 5.405405     9.459459   0.00000 0.0000000
# 5 ERR2017415 Normal female      0  6.034483     4.310345 27.58621  0.00000000  5.172414 10.344828   0.000000 6.034483     8.620690   0.00000 0.0000000
# 6 ERR2017416 Normal   male      0  2.564103     0.000000 25.64103  0.00000000 11.538462 10.256410   5.128205 6.410256    10.256410   0.00000 1.2820513

df.zones$group_mf <- paste0(df.zones$group,"__",df.zones$sex)
unique(df.zones$group_mf)
# "Normal__male"   "Normal__female" "ACVD__male"     "ACVD__female"  
df.zones$group_mf <- factor(df.zones$group_mf, levels = c("ACVD__female","Normal__female","ACVD__male","Normal__male"), 
                                  labels = c("ACVD (F)","Normal (F)","ACVD (M)","Normal (M)"), 
                                  ordered = TRUE)

names(df.zones)
# [1] "sample"       "group"        "sex"          "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx" 
# [9] "lignin"       "tannins"      "methylated"   "hydrated"     "demethylated" "condensed"    "k2_b12"       "group_mf"    

names(df.zones[ ,c(4:15)])

df.zones$all <- apply(X = df.zones[ ,c(4:15)], MARGIN = 1, FUN = sum)

table(df.zones$group_mf)
# ACVD (F) Normal (F)   ACVD (M) Normal (M) 
# 53        101        157         69 

saveRDS(object = df.zones, file = "df.zones--Jie-ACVD.RDS")

str(df.zones)
# 'data.frame':	380 obs. of  17 variables:
# $ sample      : chr  "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" ...
# $ group       : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex         : chr  "male" "male" "male" "female" ...
# $ lipids      : num  0 0 0 0 0 ...
# $ proteins    : num  2.83 11.76 0 2.7 6.03 ...
# $ amino_sugars: num  3.04 0 0 2.7 4.31 ...
# $ carbs       : num  34.4 29.4 41.7 25.7 27.6 ...
# $ cond_aromtx : num  0.0591 0 0 0 0 ...
# $ lignin      : num  3.68 11.76 6.25 7.36 5.17 ...
# $ tannins     : num  12.76 5.88 10.42 10.89 10.34 ...
# $ methylated  : num  1.66 0 0 0 0 ...
# $ hydrated    : num  3.04 0 0 5.41 6.03 ...
# $ demethylated: num  6.55 5.88 2.08 9.46 8.62 ...
# $ condensed   : num  0.878 0 0 0 0 ...
# $ k2_b12      : num  0.35 0 0 0 0 ...
# $ group_mf    : Ord.factor w/ 4 levels "ACVD (F)"<"Normal (F)"<..: 4 4 4 2 2 4 4 2 2 2 ...
# $ all         : num  69.2 64.7 60.4 64.2 68.1 ...


df.zones.melt <- melt(df.zones, id.vars = c("sample","group","sex","group_mf"))
head(df.zones.melt)
# sample  group    sex   group_mf variable value
# 1 ERR2017411 Normal   male Normal (M)   lipids     0
# 2 ERR2017412 Normal   male Normal (M)   lipids     0
# 3 ERR2017413 Normal   male Normal (M)   lipids     0
# 4 ERR2017414 Normal female Normal (F)   lipids     0
# 5 ERR2017415 Normal female Normal (F)   lipids     0
# 6 ERR2017416 Normal   male Normal (M)   lipids     0

unique( df.zones.melt$group )
# [1] Normal ACVD  
# Levels: ACVD < Normal

unique( df.zones.melt$sex ) #  "male"   "female"
df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)


str(df.zones.melt)
# 'data.frame':	4940 obs. of  6 variables:
# $ sample  : chr  "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" ...
# $ group   : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 2 2 2 1 1 2 2 1 1 1 ...
# $ group_mf: Ord.factor w/ 4 levels "ACVD (F)"<"Normal (F)"<..: 4 4 4 2 2 4 4 2 2 2 ...
# $ variable: Factor w/ 13 levels "lipids","proteins",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0 0 0 0 0 ...

hist(df.zones.melt$value)
hist(log10( df.zones.melt$value))


p <- ggplot(data = df.zones.melt, aes(x = group_mf, y = value))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  facet_wrap(facets = vars(variable), scales = "free_y")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=600, compression="lzw",type="cairo")


names(df.zones.melt)
#  "sample"    "group"     "sex"       "group_mf"  "variable"  "value"     "log10abun"


dat <- df.zones.melt

dim(dat) #  4940    6




ok <- complete.cases(dat)
sel.ok <- which(ok==TRUE)

##NOT RUN
# add log10abun - perform 95% Winzorization within each category to avoid undue influence
# from outliers, while also performing zero-replacement, then calculate log10abun

## Winsorization 95%
win95 <- function(x) {
  y<-x
  y[ y < quantile(y, 0.025) ] <- quantile(y, 0.025)
  y[ y > quantile(y, 0.975) ] <- quantile(y, 0.975)
  return(y)
  ##names(x)[i] <- paste0(names(dat)[i],"_WIN95")  # do explicitly for dataframe variables==columns
}


# decide to win95 if excess peakedness - kurtosis > 4
# based on measure of kurtosis (peakedness)
# high kurtosis tend to have a distinct peak near the mean and tend to decline rapidly

dat$new_value <- NA
dat$which_val_used <- NA
#dat$zero_correct <- NA
#dat$log10abun <- NA
dat$win95abun <- NA

vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  kurt <- kurtosis(dat$value[sel])
  print(paste0("kurtosis: ", round(kurt, 2)))
  if (kurt > 4) {
    # perform Win95
    dat$win95abun[sel] <- win95(dat$value[sel])
    dat$new_value[sel] <- dat$win95abun[sel]
    dat$which_val_used[sel] <- "Win95"
    print("BUT Win95 values were used:")
    print(summary(dat$win95abun[sel]))
    hist(dat$win95abun[sel], main = paste0(this_var," (% rel abun)\n95% Winsorized"))
  } else {
    dat$new_value[sel] <- dat$value[sel]
    dat$which_val_used[sel] <- "Rel abun (un-transformed)"
    print("Rel abun (un-transformed) values were used")
    hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  }
  
  ##NOT PERFORM LOG10
  # # log10
  # # set zero-replacement value at 1/2 smallest non-zero value of that group
  # subsel.zero <- which(dat$win95abun[sel] == 0)
  # #zero_replace <- 0.5*min(dat$win95abun[sel][dat$win95abun[sel] > 0])
  # zero_replace <- 0.5*min(dat$win95abun[sel [-subsel.zero] ])
  # dat$win95abun[sel [subsel.zero] ] <- zero_replace
  # dat$win95log10abun[sel] <- log10(dat$win95abun[sel])
  # hist(dat$win95log10abun[sel], main = paste0(this_var," (log10 % rel abun; \n95% Winzorized)"))
}


# [1] "# # # # # # # # #"
# [1] "Data for -- lipids"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00000 0.06079 0.09903 0.14398 0.18386 0.81469 
# [1] "kurtosis: 7.7"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00000 0.06079 0.09903 0.14150 0.18386 0.53894 
# [1] "# # # # # # # # #"
# [1] "Data for -- proteins"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   2.234   2.420   2.561   2.696  12.500 
# [1] "kurtosis: 54.27"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.899   2.234   2.420   2.488   2.696   3.440 
# [1] "# # # # # # # # #"
# [1] "Data for -- amino_sugars"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   3.487   3.614   3.492   3.735   6.410 
# [1] "kurtosis: 16.91"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.852   3.487   3.614   3.510   3.735   4.068 
# [1] "# # # # # # # # #"
# [1] "Data for -- carbs"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 12.32   15.88   17.20   18.72   18.39   48.11 
# [1] "kurtosis: 8.33"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 13.47   15.88   17.20   18.62   18.39   34.97 
# [1] "# # # # # # # # #"
# [1] "Data for -- cond_aromtx"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00000 0.09714 0.12948 0.14141 0.16623 4.00000 
# [1] "kurtosis: 302.65"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00000 0.09714 0.12948 0.12974 0.16623 0.27388 
# [1] "# # # # # # # # #"
# [1] "Data for -- lignin"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   5.079   5.337   5.343   5.695  12.000 
# [1] "kurtosis: 20.03"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 3.153   5.079   5.337   5.337   5.695   6.721 
# [1] "# # # # # # # # #"
# [1] "Data for -- tannins"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 2.174  12.907  14.273  14.192  15.523  33.333 
# [1] "kurtosis: 15.57"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 10.30   12.91   14.27   14.21   15.52   17.87 
# [1] "# # # # # # # # #"
# [1] "Data for -- methylated"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.5091  0.5591  0.6407  0.6077  6.0000 
# [1] "kurtosis: 60.96"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.5091  0.5591  0.6130  0.6077  1.6843 
# [1] "# # # # # # # # #"
# [1] "Data for -- hydrated"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   4.891   5.393   5.233   5.897   8.333 
# [1] "kurtosis: 9.03"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.9486  4.8913  5.3925  5.2448  5.8974  6.9139 
# [1] "# # # # # # # # #"
# [1] "Data for -- demethylated"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   11.27   12.23   11.82   13.21   16.59 
# [1] "kurtosis: 8.85"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 5.92   11.27   12.23   11.90   13.21   14.93 
# [1] "# # # # # # # # #"
# [1] "Data for -- condensed"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.2870  0.5835  0.5991  0.8603  4.5455 
# [1] "kurtosis: 21.11"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.2870  0.5835  0.5842  0.8603  1.3451 
# [1] "# # # # # # # # #"
# [1] "Data for -- k2_b12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.2011  0.2784  0.3318  0.3848  2.2727 
# [1] "kurtosis: 19.71"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.2011  0.2784  0.3197  0.3848  0.9376 
# [1] "# # # # # # # # #"
# [1] "Data for -- all"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54.44   60.78   62.89   63.22   65.19   84.09 
# [1] "kurtosis: 6.99"
# [1] "BUT Win95 values were used:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.20   60.78   62.89   63.18   65.19   70.98 


unique(dat$which_val_used) # all are "Win95"




cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)

p <- ggplot(data = dat, aes(x = group, y = win95abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  scale_colour_manual(values = cols, name = "Sex")+
  scale_shape_manual(values = shapes, name = "Sex")+
  facet_wrap(facets = vars(variable), scales = "free_y")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-WIN95-",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=600, compression="lzw",type="cairo")



## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"


unique(dat$group_mf)
# [1] Normal (M) Normal (F) ACVD (M)   ACVD (F)  
# Levels: ACVD (F) < Normal (F) < ACVD (M) < Normal (M)


y_anotations <- list()
sig_anotations <- list()

# sig converter - Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
sig_symb <- function(x) {
  out <- character(length = length(x))
  for (i in seq_along(x)) {
    if (is.na(x[i])) {
      out[i] <- NA
    } else if (x[i] < 0.001) {
      out[i] <- "***"
    } else if (x[i] >= 0.001 & x[i] < 0.01) {
      out[i] <- "**"
    } else if (x[i] >= 0.01 & x[i] <= 0.05) {
      out[i] <- "*" 
    } else { 
      out[i] <- "ns"
    }
  } # END loop
  return(out)
}



## lipids
this_var <- "Lipids"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 2.5791, num df = 52, denom df = 100, p-value = 4.919e-05
var.test(w,z) # F = 1.8698, num df = 156, denom df = 68, p-value = 0.004055

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 4053, p-value = 8.234e-08

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 8280.5, p-value = 1.259e-10

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m)
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )



## "Proteins"
this_var <- "Proteins"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.73597, num df = 52, denom df = 100, p-value = 0.224
var.test(w,z) # F = 0.67656, num df = 156, denom df = 68, p-value = 0.04908

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.58394, df = 152, p-value = 0.2801

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 6520, p-value = 0.007413

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.074598, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.10281, num df = 156, denom df = 68, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2504, p-value = 0.7447

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 6092.5, p-value = 0.06782

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.081804, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.057818, num df = 156, denom df = 68, p-value < 2.2e-16
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1587, p-value = 1.725e-05

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 2502, p-value = 6.08e-11

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.61102, num df = 52, denom df = 100, p-value = 0.05173
var.test(w,z) # F = 0.81312, num df = 156, denom df = 68, p-value = 0.2964

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 6.2815, df = 152, p-value = 1.678e-09
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.04304486        Inf
# sample estimates:
#   mean of x  mean of y 
# 0.15470022 0.09625793 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 8.8025, df = 224, p-value < 2.2e-16
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.05402284        Inf
# sample estimates:
#   mean of x  mean of y 
# 0.15915025 0.09264929

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.28831, num df = 52, denom df = 100, p-value = 2.904e-06
var.test(w,z) # F = 0.39037, num df = 156, denom df = 68, p-value = 1.537e-06
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3437.5, p-value = 0.001912

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 7633, p-value = 4.908e-07

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Tannins"
this_var <- "Tannins"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.56567, num df = 52, denom df = 100, p-value = 0.02496
var.test(w,z) # F = 0.71006, num df = 156, denom df = 68, p-value = 0.08481
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2422, p-value = 0.167

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 3.0221, df = 224, p-value = 0.001401
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.3559389       Inf
# sample estimates:
#   mean of x mean of y 
# 14.60996  13.82502 

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other methylated\ncompounds"
this_var <- "Other methylated\ncompounds"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.052549, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.027651, num df = 156, denom df = 68, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3000, p-value = 0.2192

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 4791, p-value = 0.1674

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other hydrated\ncompounds"
this_var <- "Other hydrated\ncompounds"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.25614, num df = 52, denom df = 100, p-value = 3.868e-07
var.test(w,z) # F = 0.18391, num df = 156, denom df = 68, p-value < 2.2e-16
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3115, p-value = 0.04788

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 7980, p-value = 7.488e-09

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.22953, num df = 52, denom df = 100, p-value = 5.44e-08
var.test(w,z) # F = 0.22839, num df = 156, denom df = 68, p-value = 2.853e-14
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2942, p-value = 0.1568

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 8113, p-value = 1.294e-09

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.93417, num df = 52, denom df = 100, p-value = 0.7994
var.test(w,z) # F = 0.91665, num df = 156, denom df = 68, p-value = 0.6514

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.5246, df = 152, p-value = 0.1294

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -1.9823, df = 224, p-value = 0.02433
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.01698314
# sample estimates:
#   mean of x mean of y 
# 0.5366322 0.6384481 

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z



y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds" 

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.0135, num df = 52, denom df = 100, p-value = 0.9356
var.test(w,z) # F = 1.2376, num df = 156, denom df = 68, p-value = 0.3208

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 2.2958, df = 152, p-value = 0.01153
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.02097627        Inf
# sample estimates:
#   mean of x mean of y 
# 0.3523982 0.2772555 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.80596, df = 224, p-value = 0.2106

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "All compounds"
this_var <- "All compounds"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$win95abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$win95abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$win95abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = win95abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.55736, num df = 52, denom df = 100, p-value = 0.02153
var.test(w,z) # F = 0.43774, num df = 156, denom df = 68, p-value = 2.564e-05

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1675, p-value = 7.037e-05

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 5245, p-value = 0.3528

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



# Add significance annotation
# females
annotation_df.f <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  start=rep(0.8, times=13), end = rep(1.8, times=13),
  y = c( y_anotations[["Lipids"]][1],
         y_anotations[["Proteins"]][1],
         y_anotations[["Amino sugars"]][1],
         y_anotations[["Carbohydrates"]][1],
         y_anotations[["Condensed\naromatics"]][1],
         y_anotations[["Lignin"]][1],
         y_anotations[["Tannins"]][1],
         y_anotations[["Other methylated\ncompounds"]][1],
         y_anotations[["Other hydrated\ncompounds"]][1],
         y_anotations[["Other demethylated\ncompounds"]][1],
         y_anotations[["Other condensed\ncompounds"]][1],
         y_anotations[["Near VitK2-VitB12\ncompounds"]][1],
         y_anotations[["All compounds"]][1]
         
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Lipids"]][1],
            sig_anotations[["Proteins"]][1],
            sig_anotations[["Amino sugars"]][1],
            sig_anotations[["Carbohydrates"]][1],
            sig_anotations[["Condensed\naromatics"]][1],
            sig_anotations[["Lignin"]][1],
            sig_anotations[["Tannins"]][1],
            sig_anotations[["Other methylated\ncompounds"]][1],
            sig_anotations[["Other hydrated\ncompounds"]][1],
            sig_anotations[["Other demethylated\ncompounds"]][1],
            sig_anotations[["Other condensed\ncompounds"]][1],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]][1],
            sig_anotations[["All compounds"]][1]
            
  ), stringsAsFactors = FALSE)
str(annotation_df.f)

# ensure no conflict in factor levels
annotation_df.f$variable <- factor(annotation_df.f$variable, 
                                   levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                              "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                   ordered = TRUE)

# males
annotation_df.m <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  start=rep(1.2, times=13), end = rep(2.2, times=13),
  y = c( y_anotations[["Lipids"]][2],
         y_anotations[["Proteins"]][2],
         y_anotations[["Amino sugars"]][2],
         y_anotations[["Carbohydrates"]][2],
         y_anotations[["Condensed\naromatics"]][2],
         y_anotations[["Lignin"]][2],
         y_anotations[["Tannins"]][2],
         y_anotations[["Other methylated\ncompounds"]][2],
         y_anotations[["Other hydrated\ncompounds"]][2],
         y_anotations[["Other demethylated\ncompounds"]][2],
         y_anotations[["Other condensed\ncompounds"]][2],
         y_anotations[["Near VitK2-VitB12\ncompounds"]][2],
         y_anotations[["All compounds"]][2]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Lipids"]][2],
            sig_anotations[["Proteins"]][2],
            sig_anotations[["Amino sugars"]][2],
            sig_anotations[["Carbohydrates"]][2],
            sig_anotations[["Condensed\naromatics"]][2],
            sig_anotations[["Lignin"]][2],
            sig_anotations[["Tannins"]][2],
            sig_anotations[["Other methylated\ncompounds"]][2],
            sig_anotations[["Other hydrated\ncompounds"]][2],
            sig_anotations[["Other demethylated\ncompounds"]][2],
            sig_anotations[["Other condensed\ncompounds"]][2],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]][2],
            sig_anotations[["All compounds"]][2]
  ), stringsAsFactors = FALSE)
str(annotation_df.m)

# ensure no conflict in factor levels
annotation_df.m$variable <- factor(annotation_df.m$variable, 
                                   levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                              "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                   ordered = TRUE)


annotation_df.f$label # "***" "ns"  "ns"  "***" "***" "**"  "ns"  "ns"  "*"   "ns"  "ns"  "*"   "***"
annotation_df.m$label # "***" "**"  "ns"  "***" "***" "***" "**"  "ns"  "***" "***" "*"   "ns"  "ns" 

sel <- which(annotation_df.f$label == "ns")
annotation_df.f$label[sel] <- NA
annotation_df.f$start[sel] <- NA
annotation_df.f$end[sel] <- NA
annotation_df.f$y[sel] <- NA

sel <- which(annotation_df.m$label == "ns")
annotation_df.m$label[sel] <- NA
annotation_df.m$start[sel] <- NA
annotation_df.m$end[sel] <- NA
annotation_df.m$y[sel] <- NA


p <- ggplot(data = dat, aes(x = group, y = win95abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  
  # females
  geom_text(data=annotation_df.f, mapping = aes(x = 1.3, y = y, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=annotation_df.f, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#7570b3")+
  # males
  geom_text(data=annotation_df.m, mapping = aes(x = 1.7, y = y, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=annotation_df.m, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#d95f02")+
  
  scale_colour_manual(values = cols, name = NULL)+
  scale_shape_manual(values = shapes, name = NULL)+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function % rel. abun. in vK zones (Winsorized)") +
  theme(
    #legend.position="bottom",
    legend.position = c(0.95, 0.25), # as fraction of plot dimensions
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.7))
        )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-vkZones-Win95-boxplots-",this_study,header,"-v9.tiff"), width = 24, height = 14, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Jie ACVD - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25)
#-------------------------

this_study <- "-Jie-ACVD-"
phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Jie-ACVD.RDS")

df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-Jie-ACVD.RDS")

# reset using clean data
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

header <- "-vk-hac-buffers-"

str(vk.samp.noNAs.long)
# 'data.frame':	4171090 obs. of  8 variables:
# $ sample   : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ fxn_id   : Factor w/ 11113 levels "fxn_10","fxn_100",..: 5124 6258 7065 7834 8501 9568 10365 1 469 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.929 0.6 0.667 0.667 0.667 ...
# $ HC_y     : num  0.857 1.8 2.333 2.333 2.333 ...
# $ group    : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" ...



# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]
# add sex
spdf$sex <- NA
samps <- unique(spdf$sample)
length(samps) # 385
length(which(phy@sam_data$Run %in% samps)) # 385
for (i in 1:length(samps)) {
  #i<-1
  this_samp <- samps[i]
  sel.spdf <- which(spdf$sample == this_samp)
  sel.meta <- which(phy@sam_data$Run == this_samp)
  spdf$sex[sel.spdf] <- phy@sam_data$Gender[sel.meta]
  print(paste0("completed ",i," - no rows = ",length(sel.spdf)))
}

dim(spdf) #  1945785       6

# any samples with unknown sex?
table(spdf$sex)
# female    male      NA 
# 716223 1203146   26416 
sel.na <- which(spdf$sex=="NA") # 26854 rows
unique(spdf$sample[sel.na]) # "ERR2017473" "ERR2017778" "ERR2017779" "ERR2017780" "ERR2017808" - 5 samples without gender data
spdf <- spdf[-sel.na, ]

head(spdf)
# sample  group      OC_x     HC_y        abun  sex
# 41  ERR2017411 Normal 0.6722222 1.511111 0.004222260 male
# 65  ERR2017411 Normal 1.5000000 1.130952 0.023644655 male
# 74  ERR2017411 Normal 1.0000000 1.666667 0.031244722 male
# 78  ERR2017411 Normal 1.0000000 1.666667 0.060800540 male
# 95  ERR2017411 Normal 0.7083333 1.630952 0.042222598 male
# 113 ERR2017411 Normal 1.0000000 1.666667 0.001688904 male

dim(spdf) #  1950716       6


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 380

hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.000002  0.000190  0.001358  0.012517  0.008167 12.500000

12.5/(pi*0.05^2) # 1591.549

add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54.44   60.78   62.89   63.22   65.19   84.09 

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"  

## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin
df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)

sum_relabun_in_vk_hac_buffers <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"   
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  #df_out <- data.frame(sample=NA, group=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA )
  df_out <- data.frame(sample=NA, group=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      group = unique(spdf_in$group),
      sex = unique(spdf_in$sex),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
      
      )
    #names(temp_out) # "sample"     "group"      "sex"        "rad"        "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"     
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 380
out[[1]]
# sample  group  sex  rad  Acetate Propionate   Butyrate     VitB2   VitB12    VitB6    VitB9     VitK2 Glutamate Pyruvate
# 2 ERR2017411 Normal male 0.05 116.8373  16.737118  0.0000000 22.578982 2.150379 64.72641 0.000000 0.0000000  592.1786 31.89729
# 3 ERR2017411 Normal male 0.10 446.3381   7.123131  0.5375948  7.347129 7.744352 63.11811 0.000000 0.0000000  282.0491 68.81214
# 4 ERR2017411 Normal male 0.15 239.4806   3.763164 12.5438789  3.802985 4.081738 46.70504 1.592874 0.0000000  182.5672 37.20355
# 5 ERR2017411 Normal male 0.20 152.0677  25.338635 10.2815007  6.439938 6.865534 43.92262 2.508776 1.1423890  139.3916 39.11226
# 6 ERR2017411 Normal male 0.25 131.8441  55.901259  7.1177553 11.153300 9.775010 70.24937 3.024867 0.9963424  167.6049 54.95007


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample  group  sex  rad  Acetate Propionate   Butyrate      VitB2   VitB12    VitB6    VitB9     VitK2 Glutamate Pyruvate
# 2  ERR2017411 Normal male 0.05 116.8373  16.737118  0.0000000  22.578982 2.150379 64.72641 0.000000 0.0000000  592.1786 31.89729
# 3  ERR2017411 Normal male 0.10 446.3381   7.123131  0.5375948   7.347129 7.744352 63.11811 0.000000 0.0000000  282.0491 68.81214
# 4  ERR2017411 Normal male 0.15 239.4806   3.763164 12.5438789   3.802985 4.081738 46.70504 1.592874 0.0000000  182.5672 37.20355
# 5  ERR2017411 Normal male 0.20 152.0677  25.338635 10.2815007   6.439938 6.865534 43.92262 2.508776 1.1423890  139.3916 39.11226
# 6  ERR2017411 Normal male 0.25 131.8441  55.901259  7.1177553  11.153300 9.775010 70.24937 3.024867 0.9963424  167.6049 54.95007
# 21 ERR2017412 Normal male 0.05   0.0000   0.000000  0.0000000 748.964438 0.000000  0.00000 0.000000 0.0000000  624.1370  0.00000

df.buffers$group_mf <- paste0(df.buffers$group,"__",df.buffers$sex)
unique(df.buffers$group_mf)
# "Normal__male"   "Normal__female" "ACVD__male"     "ACVD__female"  
df.buffers$group_mf <- factor(df.buffers$group_mf, levels = c("ACVD__female","Normal__female","ACVD__male","Normal__male"), 
                            labels = c("ACVD (F)","Normal (F)","ACVD (M)","Normal (M)"), 
                            ordered = TRUE)
table(df.buffers$group_mf[ which(df.buffers$rad==0.05) ]) # count at single radius only
# ACVD (F) Normal (F)   ACVD (M) Normal (M) 
# 53        101        157         69 


saveRDS(object = df.buffers, file = "df.buffers--Jie-ACVD.RDS")

str(df.buffers)
# 'data.frame':	1900 obs. of  15 variables:
# $ sample    : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ group     : chr  "Normal" "Normal" "Normal" "Normal" ...
# $ sex       : chr  "male" "male" "male" "male" ...
# $ rad       : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate   : num  117 446 239 152 132 ...
# $ Propionate: num  16.74 7.12 3.76 25.34 55.9 ...
# $ Butyrate  : num  0 0.538 12.544 10.282 7.118 ...
# $ VitB2     : num  22.58 7.35 3.8 6.44 11.15 ...
# $ VitB12    : num  2.15 7.74 4.08 6.87 9.78 ...
# $ VitB6     : num  64.7 63.1 46.7 43.9 70.2 ...
# $ VitB9     : num  0 0 1.59 2.51 3.02 ...
# $ VitK2     : num  0 0 0 1.142 0.996 ...
# $ Glutamate : num  592 282 183 139 168 ...
# $ Pyruvate  : num  31.9 68.8 37.2 39.1 55 ...
# $ group_mf  : Ord.factor w/ 4 levels "ACVD (F)"<"Normal (F)"<..: 4 4 4 4 4 4 4 4 4 4 ...


df.buffers.melt <- melt(df.buffers, id.vars = c("sample","group","sex","group_mf","rad"))
head(df.buffers.melt)
# sample  group  sex   group_mf  rad variable    value
# 1 ERR2017411 Normal male Normal (M) 0.05  Acetate 116.8373
# 2 ERR2017411 Normal male Normal (M) 0.10  Acetate 446.3381
# 3 ERR2017411 Normal male Normal (M) 0.15  Acetate 239.4806
# 4 ERR2017411 Normal male Normal (M) 0.20  Acetate 152.0677
# 5 ERR2017411 Normal male Normal (M) 0.25  Acetate 131.8441
# 6 ERR2017412 Normal male Normal (M) 0.05  Acetate   0.0000

unique( df.buffers.melt$group ) # "Normal" "ACVD"  
df.buffers.melt$group <- factor(df.buffers.melt$group, levels = c("ACVD","Normal"), ordered = TRUE)

unique( df.buffers.melt$sex ) #  "male"   "female"
df.buffers.melt$sex <- factor(df.buffers.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)

unique( df.buffers.melt$group_mf )
# [1] Normal (M) Normal (F) ACVD (M)   ACVD (F)  
# Levels: ACVD (F) < Normal (F) < ACVD (M) < Normal (M)

unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	19000 obs. of  7 variables:
#   $ sample  : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ group   : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 2 2 2 2 2 2 2 2 2 2 ...
# $ group_mf: Ord.factor w/ 4 levels "ACVD (F)"<"Normal (F)"<..: 4 4 4 4 4 4 4 4 4 4 ...
# $ rad     : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ variable: Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  117 446 239 152 132 ...

hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))
#hist(sqrt( df.buffers.melt$value))

dat <- df.buffers.melt



dat$log10abun <- dat$value # then over-write later

vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}

# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   47.08   62.45   85.33   95.79 1364.19 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.083   1.673   1.796   1.843   1.981   3.135 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   12.61   16.03   17.65   22.79  163.24 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.09769 1.10061 1.20496 1.18288 1.35775 2.21282 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   2.486   8.512   7.165  11.061  48.783 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.2314  0.3954  0.9300  0.4109  1.0438  1.6883 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   15.30   16.99   26.86   24.41  748.96 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.1423  1.1847  1.2302  1.2743  1.3875  2.8745 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   3.111   5.973   7.155  10.966  72.343 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.8497  0.4930  0.7762  0.6904  1.0401  1.8594 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   72.34   83.23   88.05  104.55  688.24 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.9663  1.8594  1.9203  1.9169  2.0193  2.8377 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   1.919   5.151   5.223   7.996  62.414 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -2.2955  0.2831  0.7119  0.3714  0.9029  1.7953 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# 0.00000  0.05285  0.37435  0.99067  1.48558 13.39587 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -2.9069 -1.2770 -0.4267 -0.7215  0.1719  1.1270 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   66.35   79.64  111.96  121.91 1149.45 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.210   1.822   1.901   1.975   2.086   3.060 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   68.83   76.35   74.13   83.01  310.55 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.6594  1.8378  1.8828  1.8462  1.9191  2.4921



#cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
#shapes <- c("Female" = 2, "Male" = 4)

head(dat)
# sample  group  sex   group_mf  rad variable    value log10abun
# 1 ERR2017411 Normal Male Normal (M) 0.05  Acetate 116.8373  2.067581
# 2 ERR2017411 Normal Male Normal (M) 0.10  Acetate 446.3381  2.649664
# 3 ERR2017411 Normal Male Normal (M) 0.15  Acetate 239.4806  2.379270
# 4 ERR2017411 Normal Male Normal (M) 0.20  Acetate 152.0677  2.182037
# 5 ERR2017411 Normal Male Normal (M) 0.25  Acetate 131.8441  2.120061
# 6 ERR2017412 Normal Male Normal (M) 0.05  Acetate   0.0000  1.082941

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  scale_colour_manual(values = cols, name = "Sex")+
  scale_shape_manual(values = shapes, name = "Sex")+
  facet_wrap(facets = vars(variable), scales = "free_y")
p


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group_mf, color = group_mf))+
  facet_wrap(facets = vars(variable), scales = "free")
p


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group_mf, color = group_mf))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    legend.position="bottom",
    legend.title = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )
  
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v8.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


saveRDS(object = dat, file = paste0("dat--All-F-M-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--All-F-M-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))


## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]

cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)

head(dat)
# sample  group    sex   group_mf  rad variable    value log10abun
# 1  ERR2017411 Normal   Male Normal (M) 0.05  Acetate 116.8373  2.067581
# 6  ERR2017412 Normal   Male Normal (M) 0.05  Acetate   0.0000  1.082941
# 11 ERR2017413 Normal   Male Normal (M) 0.05  Acetate 265.2582  2.423669
# 16 ERR2017414 Normal Female Normal (F) 0.05  Acetate   0.0000  1.082941
# 21 ERR2017415 Normal Female Normal (F) 0.05  Acetate   0.0000  1.082941
# 26 ERR2017416 Normal   Male Normal (M) 0.05  Acetate   0.0000  1.082941

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  #geom_point(aes(color = sex, shape = sex), alpha=0.6)+
  #geom_point()+
  labs(x = NULL, y = "log10 fxn % rel. abun. in 0.05 radius buffer (per vK unit area)") +
  scale_colour_manual(values = cols, name = "Sex")+
  scale_shape_manual(values = shapes, name = "Sex")+
  facet_wrap(facets = vars(variable), scales = "free")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-rough-boxplots",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")




## perform sig tests for annotations !!

levels( dat$variable )
# "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"      "Glutamate"  "Pyruvate"  

unique(dat$group_mf)
# [1] Normal (M) Normal (F) ACVD (M)   ACVD (F)  
# Levels: ACVD (F) < Normal (F) < ACVD (M) < Normal (M)

names(dat) # "sample"    "group"     "sex"       "group_mf"  "rad"       "variable"  "value"     "log10abun"


y_anotations <- list()
sig_anotations <- list()

# sig converter - Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
sig_symb <- function(x) {
  out <- character(length = length(x))
  for (i in seq_along(x)) {
    if (is.na(x[i])) {
      out[i] <- NA
    } else if (x[i] < 0.001) { 
      out[i] <- "***"
    } else if (x[i] >= 0.001 & x[i] < 0.01) { 
      out[i] <- "**"
    } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
      out[i] <- "*" 
    } else { 
      out[i] <- "ns"
    }
  } # END loop
  return(out)
}




## Acetate
this_var <- "Acetate"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.089089, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.13036, num df = 156, denom df = 68, p-value < 2.2e-16
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2888, p-value = 0.79

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 3418, p-value = 5.081e-06

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Propionate"
this_var <- "Propionate"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.033511, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.03557, num df = 156, denom df = 68, p-value < 2.2e-16
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3228, p-value = 0.01804

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 6117, p-value = 0.06101

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Butyrate"
this_var <- "Butyrate"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = NaN, num df = 52, denom df = 100, p-value = NA
var.test(w,z) # F = NaN, num df = 156, denom df = 68, p-value = NA

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2676.5, p-value = NA

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 5416.5, p-value = NA

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- c(NA,"NA",NA)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB2"
this_var <- "VitB2"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.017362, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.030067, num df = 156, denom df = 68, p-value < 2.2e-16
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2823, p-value = 0.2893

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 5379, p-value = 0.4674

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB12"
this_var <- "VitB12"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.21572, num df = 52, denom df = 100, p-value = 1.728e-08
var.test(w,z) # F = 0.25022, num df = 156, denom df = 68, p-value = 1e-12
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3626, p-value = 0.0001513

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 6154, p-value = 0.05175

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB6"
this_var <- "VitB6"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.010178, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.015165, num df = 156, denom df = 68, p-value < 2.2e-16
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3047, p-value = 0.07967

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 7352, p-value = 9.578e-06

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB9"
this_var <- "VitB9"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.26444, num df = 52, denom df = 100, p-value = 6.73e-07
var.test(w,z) # F = 0.30859, num df = 156, denom df = 68, p-value = 1.582e-09
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3803, p-value = 9.031e-06

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 7416, p-value = 5.004e-06

# y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
# y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anot.f <- 0.10*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.025*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitK2"
this_var <- "VitK2"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.5249, num df = 52, denom df = 100, p-value = 0.07223
var.test(w,z) # F = 1.0235, num df = 156, denom df = 68, p-value = 0.9318
# But clearly not normally distributed

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 4209, p-value = 1.873e-10

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 5.4167, df = 224, p-value = 7.808e-08
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.4031282       Inf
# sample estimates:
#   mean of x mean of y 
# -1.770681 -2.350657 

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Glutamate"  
this_var <- "Glutamate"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.59507, num df = 52, denom df = 100, p-value = 0.04061
var.test(w,z) # F = 0.43803, num df = 156, denom df = 68, p-value = 2.604e-05
# But clearly not normally distributed

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1758, p-value = 0.0002405

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 2641, p-value = 4.392e-10

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

# y_anot.f <- 0.11*(max(x,y) - min(x,y)) + max(x,y)
# y_anot.m <- 0.005*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Pyruvate" 
this_var <- "Pyruvate"

x=filter(dat, group_mf == "ACVD (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "ACVD (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.015813, num df = 52, denom df = 100, p-value < 2.2e-16
var.test(w,z) # F = 0.016205, num df = 156, denom df = 68, p-value < 2.2e-16
# But clearly not normally distributed

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2817, p-value = 0.2972

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 7544, p-value = 1.309e-06

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



# Add significance annotation
# females
annotation_df.f <- data.frame(
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" ,"Glutamate", "Pyruvate"),
  start=rep(0.8, times=10), end = rep(1.8, times=10),
  y = c( y_anotations[["Acetate"]][1],
         y_anotations[["Propionate"]][1],
         y_anotations[["Butyrate"]][1],
         y_anotations[["VitB2"]][1],
         y_anotations[["VitB12"]][1],
         y_anotations[["VitB6"]][1],
         y_anotations[["VitB9"]][1],
         y_anotations[["VitK2"]][1] ,
         
         y_anotations[["Glutamate"]][1],
         y_anotations[["Pyruvate"]][1]
         ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Acetate"]][1],
            sig_anotations[["Propionate"]][1],
            sig_anotations[["Butyrate"]][1],
            sig_anotations[["VitB2"]][1],
            sig_anotations[["VitB12"]][1],
            sig_anotations[["VitB6"]][1],
            sig_anotations[["VitB9"]][1],
            sig_anotations[["VitK2"]][1] ,
            
            sig_anotations[["Glutamate"]][1],
            sig_anotations[["Pyruvate"]][1]
            ),
  stringsAsFactors = FALSE)
str(annotation_df.f)

# ensure no conflict in factor levels
annotation_df.f$variable <- factor(annotation_df.f$variable, 
                                   levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" , "Glutamate", "Pyruvate"),
                                   ordered = TRUE)

# males
annotation_df.m <- data.frame(
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" , "Glutamate", "Pyruvate"),
  start=rep(1.2, times=10), end = rep(2.2, times=10),
  y = c( y_anotations[["Acetate"]][2],
         y_anotations[["Propionate"]][2],
         y_anotations[["Butyrate"]][2],
         y_anotations[["VitB2"]][2],
         y_anotations[["VitB12"]][2],
         y_anotations[["VitB6"]][2],
         y_anotations[["VitB9"]][2],
         y_anotations[["VitK2"]][2] ,
         
         y_anotations[["Glutamate"]][2],
         y_anotations[["Pyruvate"]][2]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Acetate"]][2],
            sig_anotations[["Propionate"]][2],
            sig_anotations[["Butyrate"]][2],
            sig_anotations[["VitB2"]][2],
            sig_anotations[["VitB12"]][2],
            sig_anotations[["VitB6"]][2],
            sig_anotations[["VitB9"]][2],
            sig_anotations[["VitK2"]][2] ,
            
            sig_anotations[["Glutamate"]][2],
            sig_anotations[["Pyruvate"]][2]
  ),
  stringsAsFactors = FALSE)
str(annotation_df.m)

# ensure no conflict in factor levels
annotation_df.m$variable <- factor(annotation_df.m$variable, 
                                   levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" , "Glutamate", "Pyruvate"),
                                   ordered = TRUE)

sel <- which(annotation_df.f$label == "ns")
annotation_df.f$label[sel] <- NA
annotation_df.f$start[sel] <- NA
annotation_df.f$end[sel] <- NA
annotation_df.f$y[sel] <- NA

sel <- which(annotation_df.m$label == "ns")
annotation_df.m$label[sel] <- NA
annotation_df.m$start[sel] <- NA
annotation_df.m$end[sel] <- NA
annotation_df.m$y[sel] <- NA


p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
 
  # females
  geom_text(data=annotation_df.f, mapping = aes(x = 1.3, y = y, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=annotation_df.f, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#7570b3")+
  # males
  geom_text(data=annotation_df.m, mapping = aes(x = 1.7, y = y, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=annotation_df.m, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#d95f02")+
  
  scale_colour_manual(values = cols, name = NULL)+
  scale_shape_manual(values = shapes, name = NULL)+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function log10 % rel. abun. in 0.05 radius vK buffers") +
  theme(
    legend.position="bottom",
    #strip.background = element_rect(fill="white", linetype = "blank"),
    strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.8))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-vkBuffers-log1oabun-boxplots-",this_study,header,"-v9.tiff"), width = 22, height = 13, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Jie - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

# Determines weighted coordinates in each sample, then later assess stats by group_mf?

this_study <- "-Jie-ACVD-"
phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Jie-ACVD.RDS")

df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-Jie-ACVD.RDS")

# reset using clean data
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

header <- "-vk-coords-wtmean-"


str(vk.samp.noNAs.long)
# 'data.frame':	4171090 obs. of  8 variables:
# $ sample   : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ fxn_id   : Factor w/ 11113 levels "fxn_10","fxn_100",..: 5124 6258 7065 7834 8501 9568 10365 1 469 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.929 0.6 0.667 0.667 0.667 ...
# $ HC_y     : num  0.857 1.8 2.333 2.333 2.333 ...
# $ group    : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" "ERR2017411 (0)" ...



# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]
# add sex
spdf$sex <- NA
samps <- unique(spdf$sample)
length(samps) # 385
length(which(phy@sam_data$Run %in% samps)) # 385
for (i in 1:length(samps)) {
  #i<-1
  this_samp <- samps[i]
  sel.spdf <- which(spdf$sample == this_samp)
  sel.meta <- which(phy@sam_data$Run == this_samp)
  spdf$sex[sel.spdf] <- phy@sam_data$Gender[sel.meta]
  print(paste0("completed ",i," - no rows = ",length(sel.spdf)))
}

dim(spdf) #  1945785       6

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# female    male      NA 
# 716223 1203146   26416 
sel.na <- which(spdf$sex=="NA") # 26416 rows
unique(spdf$sample[sel.na]) # "ERR2017473" "ERR2017778" "ERR2017779" "ERR2017780" "ERR2017808" - 5 samples without gender data
spdf <- spdf[-sel.na, ]

head(spdf)
# sample  group      OC_x     HC_y        abun  sex
# 41  ERR2017411 Normal 0.6722222 1.511111 0.004222260 male
# 65  ERR2017411 Normal 1.5000000 1.130952 0.023644655 male
# 74  ERR2017411 Normal 1.0000000 1.666667 0.031244722 male
# 78  ERR2017411 Normal 1.0000000 1.666667 0.060800540 male
# 95  ERR2017411 Normal 0.7083333 1.630952 0.042222598 male
# 113 ERR2017411 Normal 1.0000000 1.666667 0.001688904 male

dim(spdf) #  1919369       6


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 380

hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.000002  0.000190  0.001358  0.012517  0.008167 12.500000 


add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54.44   60.78   62.89   63.22   65.19   84.09

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"




get_wtmean_vk_coords <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)

  spdf_in <- spdf_in[sel, ]

  subsel <- list()

  # lipids
  subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  # proteins
  subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # amino_sugars
  subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # carbs - Carbohydrates
  subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  # cond_aromtx - Condensed aromatics
  subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  # lignin
  subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  # tannins
  subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)

  vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")

  # establish blank template data frame
  df_out <- data.frame(sample = NA,group = NA,sex = NA, no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)

  for (v in 1:length(vars)) {
    #v<-1
    this_var <- vars[v]
    if (length(subsel[[ this_var ]]) > 0) {
      df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
    } else {
      df_out$wtmean_OC_x[v] <- NA
      df_out$wtmean_HC_y[v] <- NA
    }
    print(paste0("completed ",v))
  }

  df_out$sample <- this_samp
  df_out$group <- unique(spdf_in$group)
  df_out$sex <- unique(spdf_in$sex)

  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 380
out[[1]]
# sample  group  sex no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1 ERR2017411 Normal male      NA       lipids          NA          NA
# 2 ERR2017411 Normal male      58     proteins   0.4017742    1.725632
# 3 ERR2017411 Normal male      53 amino_sugars   0.6603348    1.704530
# 4 ERR2017411 Normal male     171        carbs   0.9112019    1.862786
# 5 ERR2017411 Normal male       4  cond_aromtx   0.1733224    1.014000
# 6 ERR2017411 Normal male      90       lignin   0.5287526    1.239192
# 7 ERR2017411 Normal male     166      tannins   0.7716333    1.404565

names(out[[1]])
# "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 2660    7

ok <- complete.cases(df.wtcoords)
sel.ok <- which(ok==TRUE) # qty 2601

df.wtcoords <- df.wtcoords[sel.ok, ]

dim(df.wtcoords) #  2601    7


str(df.wtcoords)
# 'data.frame':	2601 obs. of  7 variables:
# $ sample     : chr  "ERR2017411" "ERR2017411" "ERR2017411" "ERR2017411" ...
# $ group      : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex        : chr  "male" "male" "male" "male" ...
# $ no_fxns    : int  58 53 171 4 90 166 2 6 6 1 ...
# $ variable   : chr  "proteins" "amino_sugars" "carbs" "cond_aromtx" ...
# $ wtmean_OC_x: num  0.402 0.66 0.911 0.173 0.529 ...
# $ wtmean_HC_y: num  1.73 1.7 1.86 1.01 1.24 ...

saveRDS(object = df.wtcoords, file = "df.wtcoords-Jie-ACVD.RDS")

#df.wtcoords <- readRDS("df.wtcoords-Jie-ACVD.RDS")


dat <- df.wtcoords

unique(dat$variable)
# "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "lipids" 

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)

dat$group_mf <- paste0(dat$group,"__",dat$sex)
unique(dat$group_mf) # "Normal__male"   "Normal__female" "ACVD__male"     "ACVD__female"
dat$group_mf <- factor(dat$group_mf, levels = c("ACVD__female","Normal__female","ACVD__male","Normal__male"),
                       labels = c("ACVD (F)","Normal (F)","ACVD (M)","Normal (M)"),
                       ordered = TRUE)

names(dat) # "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"

dat.F <- filter(dat, sex == "female")
dat.M <- filter(dat, sex == "male")

cols.F <- c("ACVD (F)" = "#f46d43", "Normal (F)" = "#66c2a5")
#cols.M <- c("ACVD (M)" = "#e08214", "Normal (M)" = "#8073ac")

#library(ggforce)

p <- ggplot(data = dat.F, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group_mf) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  geom_density_2d(contour_var = "ndensity", bins = 5, alpha=.5) + #
  geom_point(alpha=0.5)+
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols.F) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Diagnosis"))

p

grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat.F, file = paste0("dat.F--Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9.RDS"))

#dat.F <- readRDS(file = paste0("dat.F--Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9.RDS"))



cols.M <- c("ACVD (M)" = "#e08214", "Normal (M)" = "#8073ac")

p <- ggplot(data = dat.M, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group_mf) )+

  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  geom_density_2d(contour_var = "ndensity", bins = 5, alpha=.5) + #
  geom_point(alpha=0.5)+

  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols.M) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Diagnosis"))

p

grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-MALE-",this_study,header,"-v9.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat.M, file = paste0("dat.M--Weighted-vkZone-coords-MALE-",this_study,header,"-v9.RDS"))

#dat.M <- readRDS(file = paste0("dat.M--Weighted-vkZone-coords-MALE-",this_study,header,"-v9.RDS"))






## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.F) # "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"


levels(dat.F$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"




## "Lipids"
this_var <- "Lipids"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)   
# samps.class   1  0.04916 0.05736 8.3359  0.005 **
# Residual    137  0.80793 0.94264                 
# Total       138  0.85709 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq   Mean Sq      F N.Perm Pr(>F)
# Groups      1 0.002551 0.0025514 1.1116    999  0.307
# Residuals 137 0.314450 0.0022952



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)
# samps.class   1  0.00328 0.01025 1.5631  0.199
# Residual    151  0.31695 0.98975              
# Total       152  0.32023 1.00000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq  Mean Sq      F N.Perm Pr(>F)   
# Groups      1 0.015269 0.015269 10.112    999  0.002 **
# Residuals 151 0.228013 0.001510                        
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)  
# samps.class   1 0.002775 0.02249 3.4504  0.052 .
# Residual    150 0.120648 0.97751                
# Total       151 0.123424 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq   Mean Sq      F N.Perm Pr(>F)  
# Groups      1 0.003806 0.0038057 6.3286    999  0.017 *
# Residuals 150 0.090202 0.0006013                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)   
# samps.class   1 0.009676 0.05896 9.5226  0.003 **
# Residual    152 0.154445 0.94104                 
# Total       153 0.164120 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq    Mean Sq      F N.Perm Pr(>F)  
# Groups      1 0.002112 0.00211156 5.5391    999  0.026 *
# Residuals 152 0.057944 0.00038121                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)  
# samps.class   1 0.006175 0.03034 4.3184  0.035 *
# Residual    138 0.197322 0.96966                
# Total       139 0.203497 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      1 0.000664 0.00066415 1.3313    999  0.264
# Residuals 138 0.068844 0.00049887 



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)  
# samps.class   1 0.003684 0.02097 3.2121  0.038 *
# Residual    150 0.172017 0.97903                
# Total       151 0.175701 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))

# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq   Mean Sq      F N.Perm Pr(>F)   
# Groups      1 0.007932 0.0079317 9.4535    999  0.004 **
# Residuals 150 0.125852 0.0008390                        
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)    
# samps.class   1 0.012583 0.08882 14.817  0.001 ***
# Residual    152 0.129081 0.91118                  
# Total       153 0.141663 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq   Mean Sq     F N.Perm Pr(>F)   
# Groups      1 0.006921 0.0069210 10.35    999  0.004 **
# Residuals 152 0.101647 0.0006687                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



annot_df.acvd.female.coord <- data.frame(
  variable = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
  p_value = c(0.005, 0.199, 0.052, 0.003, 0.035, 0.038, 0.001),
  r_squared = c(0.057, 0.010, 0.022, 0.059, 0.0300, 0.021, 0.089),
  npcx = c("left", "right", "left", "right", "right", "right", "right"),
  npcy = c("top", "bottom", "top", "bottom", "bottom", "top", "top")
)

annot_df.acvd.female.coord$variable <- factor(annot_df.acvd.female.coord$variable, 
                                              levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                                              ordered = TRUE)
annot_df.acvd.female.coord$displaylabel <- paste0("PERMANOVA\n","P = ",annot_df.acvd.female.coord$p_value,"\nR^2 = ",annot_df.acvd.female.coord$r_squared)


p +
  geom_text_npc(data = annot_df.acvd.female.coord, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 2.5, lineheight=0.9 )



grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9e-with-Permanova.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



# # # # # # # # # # # # #


# Males


## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.M) 
# "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"

levels(dat.M$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"




## "Lipids"
this_var <- "Lipids"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)
# samps.class   1  0.00519 0.00321 0.6914  0.447
# Residual    215  1.61379 0.99679              
# Total       216  1.61898 1.00000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
# (betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df  Sum Sq   Mean Sq      F N.Perm Pr(>F)   
# Groups      1 0.02370 0.0237044 8.2366    999  0.005 **
# Residuals 215 0.61875 0.0028779                        
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)    
# samps.class   1 0.013397 0.05734 13.444  0.001 ***
# Residual    221 0.220234 0.94266                  
# Total       222 0.233631 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df  Sum Sq   Mean Sq      F N.Perm Pr(>F)    
# Groups      1 0.01673 0.0167295 23.566    999  0.001 ***
# Residuals 221 0.15689 0.0007099                         
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)    
# samps.class   1 0.002021 0.04021 9.2584  0.001 ***
# Residual    221 0.048233 0.95979                  
# Total       222 0.050254 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq    Mean Sq      F N.Perm Pr(>F)    
# Groups      1 0.002408 0.00240800 18.472    999  0.001 ***
# Residuals 221 0.028809 0.00013036                         
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2  F Pr(>F)    
# samps.class   1 0.057259 0.22491 65  0.001 ***
# Residual    224 0.197324 0.77509              
# Total       225 0.254583 1.00000              
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq    Mean Sq      F N.Perm Pr(>F)  
# Groups      1 0.002117 0.00211741 7.0836    999  0.012 *
# Residuals 224 0.066957 0.00029892                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)    
# samps.class   1 0.031695 0.10425 25.021  0.001 ***
# Residual    215 0.272347 0.89575                  
# Total       216 0.304042 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      1 0.000970 0.00097038 2.4834    999  0.114
# Residuals 215 0.084011 0.00039075



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)    
# samps.class   1  0.00661 0.03682 8.5247  0.001 ***
# Residual    223  0.17291 0.96318                  
# Total       224  0.17952 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df  Sum Sq   Mean Sq      F N.Perm Pr(>F)    
# Groups      1 0.01404 0.0140396 27.095    999  0.001 ***
# Residuals 223 0.11555 0.0005182                         
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)    
# samps.class   1 0.031015 0.20279 56.981  0.001 ***
# Residual    224 0.121923 0.79721                  
# Total       225 0.152938 1.00000                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq   Mean Sq     F N.Perm Pr(>F)    
# Groups      1 0.021774 0.0217737 50.93    999  0.001 ***
# Residuals 224 0.095766 0.0004275                        
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## run Male plot

annot_df.acvd.male.coord <- data.frame(
  variable = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
  p_value = c(0.447, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001),
  r_squared = c(0.003, 0.057, 0.04, 0.225, 0.104, 0.037, 0.203),
  npcx = c("right",  "left", "left", "right",  "right",  "left",  "left"),
  npcy = c("bottom", "top",  "top",  "bottom", "bottom", "bottom", "bottom")
)

annot_df.acvd.male.coord$variable <- factor(annot_df.acvd.male.coord$variable, 
                                              levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                                              ordered = TRUE)
annot_df.acvd.male.coord$displaylabel <- paste0("PERMANOVA\n","P = ",annot_df.acvd.male.coord$p_value,"\nR^2 = ",annot_df.acvd.male.coord$r_squared)


p +
  geom_text_npc(data = annot_df.acvd.male.coord, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 2.5, lineheight=0.9 )


grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-MALE-",this_study,header,"-v9e-with-Permanova.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Jie ACVD - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-Jie-ACVD.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 21128 taxa and 385 samples ]
# sample_data() Sample Data:       [ 385 samples by 56 sample variables ]
# tax_table()   Taxonomy Table:    [ 21128 taxa by 4 taxonomic ranks ]

df.zones <- readRDS("df.zones--Jie-ACVD.RDS")

vk.samp.noNAs.long.clean <- readRDS(file = "vk.samp.noNAs.long.clean-Jie-ACVD.RDS")

samps <- unique(df.zones$sample) # qty 380
phy <- prune_samples( samples = samps, phy )
# prune taxa with zero reads
phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 21117 taxa and 380 samples ]
# sample_data() Sample Data:       [ 380 samples by 56 sample variables ]
# tax_table()   Taxonomy Table:    [ 21117 taxa by 4 taxonomic ranks ]

head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 380
NonZeroFxns


mean(NonZeroFxns) # 9300.55
sd(NonZeroFxns) # 3523.022
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 10829
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 2535
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 5050.971
sd(fxns_converted) # 1842.965

mean(fxns_relabun_converted) # 63.22233
sd(fxns_relabun_converted) # 3.468045

mean(vks) # 1509.374
sd(vks) # 502.0301


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Zeller et al 2014 - Colorectal cancer vs normal gut metagenomes
# # # # # # # # #
# # # # # # # # #

#### Zeller CRC - prepare for data download from SRA
#-------------------------
# Zeller, G. et al. Potential of fecal microbiota for early-stage detection of colorectal cancer. 
# Molecular Systems Biology 10, 766, doi:https://doi.org/10.15252/msb.20145645 (2014).

# Zeller et al 2014 - Colorectal cancer - (N = 156, but ignore  adenoma patients)
# N = 114: 53 patients with CRC, and 61 randomly chosen controls [[ignore , 42 adenoma patients]]

# Just use French population dataset !!!!

# "msb145645-sup-0002-datasets1.xlsx"
# A1:M157
# Sheet = "Pop. F Subject Metadata"

subjects <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Zeller-2014-Cancer/msb145645-sup-0002-datasets1.xlsx",
                       sheet="Pop. F Subject Metadata", range="A1:M157", col_names = TRUE)
subjects <- as.data.frame(subjects)
str(subjects)
length(unique( subjects$`Subject ID` )) # 156
table(subjects$Diagnosis)
# Cancer Large adenoma        Normal Small adenoma 
#     53            15            61            27 
sel <- which(subjects$Diagnosis %in% c("Cancer","Normal")) # qty 114
subjects <- subjects[sel, ]
table(subjects$Group)
# Control     CRC 
#      61      53 

sub.id <- subjects$`Subject ID`
length(unique(sub.id)) # 114

samp.id <- subjects$`Sample ID`
length(unique(samp.id)) # 114


# "Zeller-etal-2014-Colorectal-Cancer-SRA-Run-Selector--SraRunTable-METADATA.xlsx"
# Sheet = 1
# A1:BN1516
# 
# host_subject_id
# `Assay Type` = "WGS"
# Sample_Name # align to `Sample ID`

meta <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Zeller-2014-Cancer/Zeller-etal-2014-Colorectal-Cancer-SRA-Run-Selector--SraRunTable-METADATA.xlsx",
                   sheet=1, range="A1:BN1516", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)
# 'data.frame':	1515 obs. of  66 variables: 
sel <- which(meta$`Assay Type` == "WGS") # qty = 1290
sel <- which(meta$host_subject_id %in% sub.id & meta$`Assay Type` == "WGS") # qty = 702
sel <- which(meta$Sample_Name %in% samp.id & meta$LibraryLayout == "PAIRED" & meta$`Assay Type` == "WGS" & meta$LibrarySelection == "RANDOM") # qty = 702

sel <- which(meta$host_subject_id %in% sub.id & meta$`Assay Type` == "WGS" & meta$LibraryLayout == "PAIRED" & meta$`geographic_location_(countryand/orsea_region)` == "France") # qty = 351

meta <- meta[sel, ]

hist(meta$Bytes)
sum(meta$Bytes)/1e9 # 377 Gb

length(unique(meta$`Library Name`)) # 351
length(unique(meta$Bytes)) # 351
length(meta$Run) # 351

row.names(meta) <- meta$Run
names(meta)

head(sort(meta$Run)) # "ERR478958" "ERR478959" "ERR478960" "ERR478961" "ERR478962" "ERR478963"

write.csv(x = meta[ sort(meta$Run), c("Run","host_subject_id","Diagnosis") ], file = "zeller-crc-sra-run-list.txt", quote = FALSE, row.names = FALSE)

# # Download from command line using format:
# fastq-dump --outdir /scratch/user/lidd0026/gmrepo_1_meta_raw --skip-technical --readids --read-filter pass --dumpbase --clip --split-3 ERR719096

write.csv(x = paste0("fastq-dump --outdir /scratch/user/lidd0026/zeller-crc/zcrc_1_meta_raw --skip-technical --readids --read-filter pass --dumpbase --clip --split-3 ",sort(meta$Run)), file = "zeller-crc-sra-runs.txt", quote = FALSE, row.names = FALSE)

#-------------------------

#### Zeller CRC - Metadata
#-------------------------

# metadata as per above

str(meta)
# 'data.frame':	351 obs. of  66 variables:
# $ Run                                          : chr  "ERR478958" "ERR478959" "ERR478960" "ERR478961" ...
# $ Age                                          : num  72 72 72 72 53 53 35 35 35 35 ...
# $ Assay Type                                   : chr  "WGS" "WGS" "WGS" "WGS" ...
# $ BioProject                                   : chr  "PRJEB6070" "PRJEB6070" "PRJEB6070" "PRJEB6070" ...
# $ BioSample                                    : chr  "SAMEA2467039" "SAMEA2467039" "SAMEA2467039" "SAMEA2467039" ...
# $ Bytes                                        : num  1.03e+09 1.02e+09 1.03e+09 1.03e+09 1.13e+09 ...
# $ Center Name                                  : chr  "EMBL - GERMANY" "EMBL - GERMANY" "EMBL - GERMANY" "EMBL - GERMANY" ...
# $ Collection_Date                              : chr  "2004/2006" "2004/2006" "2004/2006" "2004/2006" ...
# $ Consent                                      : chr  "public" "public" "public" "public" ...
# $ DATASTORE filetype                           : chr  "sra,fastq" "sra,fastq" "fastq,sra" "fastq,sra" ...
# $ DATASTORE provider                           : chr  "s3,gs,ena" "s3,gs,ena" "ena,gs,s3" "ena,gs,s3" ...
# $ DATASTORE region                             : chr  "ena,s3.us-east-1,gs.US" "s3.us-east-1,ena,gs.US" "gs.US,s3.us-east-1,ena" "s3.us-east-1,ena,gs.US" ...
# $ ENA-FIRST-PUBLIC (run)                       : POSIXct, format: "2014-11-01" "2014-11-01" "2014-11-01" "2014-11-01" ...
# $ ENA-LAST-UPDATE (run)                        : POSIXct, format: "2018-11-16" "2018-11-16" "2018-11-16" "2018-11-16" ...
# $ environment_(biome)                          : chr  "-" "-" "-" "-" ...
# $ environment_(feature)                        : chr  "-" "-" "-" "-" ...
# $ environment_(material)                       : chr  "-" "-" "-" "-" ...
# $ Experiment                                   : chr  "ERX444790" "ERX444791" "ERX444792" "ERX444793" ...
# $ geographic_location_(latitude)               : num  0 0 0 0 0 0 0 0 0 0 ...
# $ geographic_location_(longitude)              : num  0 0 0 0 0 0 0 0 0 0 ...
# $ host_subject_id                              : chr  "FR-726" "FR-726" "FR-726" "FR-726" ...
# $ INSDC_center_alias                           : chr  "EMBL - Germany" "EMBL - Germany" "EMBL - Germany" "EMBL - Germany" ...
# $ INSDC_center_name                            : chr  "EMBL - Germany" "EMBL - Germany" "EMBL - Germany" "EMBL - Germany" ...
# $ INSDC_first_public                           : chr  "2014-11-01T17:01:02Z" "2014-11-01T17:01:02Z" "2014-11-01T17:01:02Z" "2014-11-01T17:01:02Z" ...
# $ INSDC_last_update                            : chr  "2016-10-21T09:35:52Z" "2016-10-21T09:35:52Z" "2016-10-21T09:35:52Z" "2016-10-21T09:35:52Z" ...
# $ INSDC_status                                 : chr  "public" "public" "public" "public" ...
# $ Instrument                                   : chr  "Illumina HiSeq 2000" "Illumina HiSeq 2000" "Illumina HiSeq 2000" "Illumina HiSeq 2000" ...
# $ Investigation_type                           : chr  "metagenome" "metagenome" "metagenome" "metagenome" ...
# $ Library Name                                 : chr  "CCIS00146684ST-4-0_13s007952-1-1_lane1" "CCIS00146684ST-4-0_13s007952-2-1_lane2" "CCIS00146684ST-4-0_13s007952-3-1_lane3" "CCIS00146684ST-4-0_13s007952-4-1_lane5" ...
# $ LibraryLayout                                : chr  "PAIRED" "PAIRED" "PAIRED" "PAIRED" ...
# $ LibrarySelection                             : chr  "RANDOM" "RANDOM" "RANDOM" "RANDOM" ...
# $ LibrarySource                                : chr  "METAGENOMIC" "METAGENOMIC" "METAGENOMIC" "METAGENOMIC" ...
# $ Organism                                     : chr  "Homo sapiens" "Homo sapiens" "Homo sapiens" "Homo sapiens" ...
# $ Platform                                     : chr  "ILLUMINA" "ILLUMINA" "ILLUMINA" "ILLUMINA" ...
# $ project_name                                 : chr  "Potential of fecal microbiota for early stage detection of colorectal cancer" "Potential of fecal microbiota for early stage detection of colorectal cancer" "Potential of fecal microbiota for early stage detection of colorectal cancer" "Potential of fecal microbiota for early stage detection of colorectal cancer" ...
# $ ReleaseDate                                  : chr  "2014-11-02T00:00:00Z" "2014-11-02T00:00:00Z" "2014-11-02T00:00:00Z" "2014-11-02T00:00:00Z" ...
# $ Sample Name                                  : chr  "SAMEA2467039" "SAMEA2467039" "SAMEA2467039" "SAMEA2467039" ...
# $ Sample_Name                                  : chr  "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" ...
# $ Sequencing_method                            : chr  "Illumina" "Illumina" "Illumina" "Illumina" ...
# $ SRA Study                                    : chr  "ERP005534" "ERP005534" "ERP005534" "ERP005534" ...
# $ body_mass_index                              : num  25 25 25 25 32 32 23 23 23 23 ...
# $ common_name                                  : chr  "human" "human" "human" "human" ...
# $ Diagnosis                                    : chr  "Normal" "Normal" "Normal" "Normal" ...
# $ ENA-FIRST-PUBLIC                             : POSIXct, format: "2014-11-01" "2014-11-01" "2014-11-01" "2014-11-01" ...
# $ ENA-LAST-UPDATE                              : POSIXct, format: "2016-10-21" "2016-10-21" "2016-10-21" "2016-10-21" ...
# $ External_Id                                  : chr  "SAMEA2467039" "SAMEA2467039" "SAMEA2467039" "SAMEA2467039" ...
# $ geographic_location_(countryand/orsea_region): chr  "France" "France" "France" "France" ...
# $ human_gut_environmental_package              : chr  "human-gut" "human-gut" "human-gut" "human-gut" ...
# $ sex                                          : chr  "female" "female" "female" "female" ...
# $ Submitter_Id                                 : chr  "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" ...
# $ AvgSpotLen                                   : num  173 172 172 172 141 141 179 180 180 180 ...
# $ Bases                                        : num  1.55e+09 1.52e+09 1.52e+09 1.54e+09 1.39e+09 ...
# $ body_product                                 : chr  "stool" "stool" "stool" "stool" ...
# $ AJCC_Stage                                   : chr  NA NA NA NA ...
# $ Alias                                        : chr  NA NA NA NA ...
# $ ENA_CHECKLIST                                : chr  NA NA NA NA ...
# $ Environmental_package                        : chr  NA NA NA NA ...
# $ gender                                       : chr  NA NA NA NA ...
# $ geographic_location_(country_and/or_sea      : chr  NA NA NA NA ...
#                        $ region)                                      : chr  NA NA NA NA ...
# $ localization                                 : chr  NA NA NA NA ...
# $ SRA_accession                                : chr  NA NA NA NA ...
# $ tissue_type                                  : chr  NA NA NA NA ...
# $ Title                                        : chr  NA NA NA NA ...
# $ TNM_Stage                                    : chr  NA NA NA NA ...
# $ BMI                                          : logi  NA NA NA NA NA NA ...

str(subjects)
# 'data.frame':	114 obs. of  13 variables:
# $ Subject ID                 : chr  "FR-001" "FR-003" "FR-024" "FR-025" ...
# $ Sample ID                  : chr  "CCIS27304052ST-3-0" "CCIS13047523ST-4-0" "CCIS15794887ST-4-0" "CCIS94603952ST-4-0" ...
# $ Age (years)                : num  52 70 37 80 74 54 65 57 53 62 ...
# $ Gender                     : chr  "F" "M" "F" "F" ...
# $ BMI (kg/m²)                : chr  "20" "22" "18" "21" ...
# $ Country of Residence       : chr  "France" "France" "France" "France" ...
# $ Diagnosis                  : chr  "Normal" "Normal" "Normal" "Normal" ...
# $ TNM Stage                  : chr  "NA" "NA" "NA" "NA" ...
# $ AJCC Stage                 : chr  "NA" "NA" "NA" "NA" ...
# $ Localization               : chr  "NA" "NA" "NA" "NA" ...
# $ FOBT                       : chr  "Negative" "Negative" "Negative" "Negative" ...
# $ WIF-1 Gene Methylation Test: chr  "Negative" "Negative" "Negative" "Negative" ...
# $ Group                      : chr  "Control" "Control" "Control" "Control" ...

# convert characters to numeric
subjects$`BMI (kg/m²)` <- as.numeric(subjects$`BMI (kg/m²)`)

table(meta$Diagnosis)
# Cancer Normal 
# 146    205

table(meta$`geographic_location_(countryand/orsea_region)`)
# France 
# 351 

length(unique(meta$host_subject_id)) # 114
head(meta$host_subject_id) # "FR-726" "FR-726" "FR-726" "FR-726" "FR-060" "FR-060"
length(unique(meta$Sample_Name)) # 114
head(meta$Sample_Name) # "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00146684ST-4-0" "CCIS00281083ST-3-0" "CCIS00281083ST-3-0"

table(subjects$Diagnosis)
# Cancer Normal 
# 53     61

names(subjects)
# [1] "Subject ID"                  "Sample ID"                   "Age (years)"                 "Gender"                      "BMI (kg/m²)"                
# [6] "Country of Residence"        "Diagnosis"                   "TNM Stage"                   "AJCC Stage"                  "Localization"               
# [11] "FOBT"                        "WIF-1 Gene Methylation Test" "Group"  

names(subjects)[1] # "Subject ID"
names(subjects)[1] <- "Subject_ID"

subjects$Subject_ID <- gsub(pattern = "-", replacement = "_", x = subjects$Subject_ID)

row.names(subjects) <- subjects$Subject_ID

( varnames <- names(subjects) )
# [1] "Subject_ID"                  "Sample ID"                   "Age (years)"                 "Gender"                      "BMI (kg/m²)"                
# [6] "Country of Residence"        "Diagnosis"                   "TNM Stage"                   "AJCC Stage"                  "Localization"               
# [11] "FOBT"                        "WIF-1 Gene Methylation Test" "Group" 



dim(subjects) # 114  13


sel <- which(varnames %in% c("Subject_ID",
                             "Age (years)",
                             "Gender",
                             "BMI (kg/m²)",
                             "Diagnosis",
                             "TNM Stage",
                             "AJCC Stage",
                             "Localization",
                             "FOBT",
                             "WIF-1 Gene Methylation Test"
))

plot_vars <- varnames[sel]
length(plot_vars) # 10

str(subjects[ ,plot_vars])

subjects.melt <- melt(subjects[ ,plot_vars], id.vars = c("Subject_ID", "Diagnosis"))

names(subjects.melt)
#  "Subject_ID" "Diagnosis" "variable"   "value" 

#sel.plot <- which(meta.melt$variable %in% plot#_vars)

p <- ggplot( subjects.melt, aes(x = Diagnosis, y = value )) +
  geom_jitter(size=1.5,width = 0.15, alpha=0.25) +
  facet_wrap(facets = vars(variable), ncol = 3, scales = "free_y")+
  theme_bw()+
  theme(
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    strip.text = element_text(size = rel(0.7)),
    #strip.text.y = element_text(size = rel(1.1)),
    strip.background = element_blank()
  )

p

#grid.text(label = "(a)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=15, fontface="bold") )

dev.print(tiff, file = paste0(workdir,"/plots/","Zeller-CRC-metadata-values.tiff"), width = 22, height = 20, units = "cm", res=600, compression="lzw", type = "cairo")


#-------------------------

#### Zeller CRC - read in superfocus - fxn potential outputs
#-------------------------

head( subjects$Subject_ID )
# "FR_001" "FR_003" "FR_024" "FR_025" "FR_027" "FR_030"

## but use the other metadata table to compile Superfocus data ...

head(meta)
#                 Run Age Assay Type BioProject    BioSample      Bytes    Center Name Collection_Date Consent DATASTORE filetype DATASTORE provider       DATASTORE region
# ERR478958 ERR478958  72        WGS  PRJEB6070 SAMEA2467039 1034393883 EMBL - GERMANY       2004/2006  public          sra,fastq          s3,gs,ena ena,s3.us-east-1,gs.US
# ERR478959 ERR478959  72        WGS  PRJEB6070 SAMEA2467039 1024956152 EMBL - GERMANY       2004/2006  public          sra,fastq          s3,gs,ena s3.us-east-1,ena,gs.US
# ERR478960 ERR478960  72        WGS  PRJEB6070 SAMEA2467039 1028014576 EMBL - GERMANY       2004/2006  public          fastq,sra          ena,gs,s3 gs.US,s3.us-east-1,ena
# ERR478961 ERR478961  72        WGS  PRJEB6070 SAMEA2467039 1034499020 EMBL - GERMANY       2004/2006  public          fastq,sra          ena,gs,s3 s3.us-east-1,ena,gs.US
# ERR478962 ERR478962  53        WGS  PRJEB6070 SAMEA2466896 1126047436 EMBL - GERMANY       2004/2006  public          sra,fastq          s3,ena,gs s3.us-east-1,gs.US,ena
# ERR478963 ERR478963  53        WGS  PRJEB6070 SAMEA2466896 1086759659 EMBL - GERMANY       2004/2006  public          fastq,sra          gs,ena,s3 ena,gs.US,s3.us-east-1
#           ENA-FIRST-PUBLIC (run) ENA-LAST-UPDATE (run) environment_(biome) environment_(feature) environment_(material) Experiment geographic_location_(latitude)
# ERR478958             2014-11-01            2018-11-16                   -                     -                      -  ERX444790                              0
# ERR478959             2014-11-01            2018-11-16                   -                     -                      -  ERX444791                              0
# ERR478960             2014-11-01            2018-11-16                   -                     -                      -  ERX444792                              0
# ERR478961             2014-11-01            2018-11-16                   -                     -                      -  ERX444793                              0
# ERR478962             2014-11-01            2018-11-16                   -                     -                      -  ERX444794                              0
# ERR478963             2014-11-01            2018-11-16                   -                     -                      -  ERX444795                              0
#           geographic_location_(longitude) host_subject_id INSDC_center_alias INSDC_center_name   INSDC_first_public    INSDC_last_update INSDC_status          Instrument
# ERR478958                               0          FR-726     EMBL - Germany    EMBL - Germany 2014-11-01T17:01:02Z 2016-10-21T09:35:52Z       public Illumina HiSeq 2000
# ERR478959                               0          FR-726     EMBL - Germany    EMBL - Germany 2014-11-01T17:01:02Z 2016-10-21T09:35:52Z       public Illumina HiSeq 2000
# ERR478960                               0          FR-726     EMBL - Germany    EMBL - Germany 2014-11-01T17:01:02Z 2016-10-21T09:35:52Z       public Illumina HiSeq 2000
# ERR478961                               0          FR-726     EMBL - Germany    EMBL - Germany 2014-11-01T17:01:02Z 2016-10-21T09:35:52Z       public Illumina HiSeq 2000
# ERR478962                               0          FR-060     EMBL - Germany    EMBL - Germany 2014-11-01T17:01:02Z 2016-10-21T09:35:52Z       public Illumina HiSeq 2000
# ERR478963                               0          FR-060     EMBL - Germany    EMBL - Germany 2014-11-01T17:01:02Z 2016-10-21T09:35:52Z       public Illumina HiSeq 2000
#           Investigation_type                           Library Name LibraryLayout LibrarySelection LibrarySource     Organism Platform
# ERR478958         metagenome CCIS00146684ST-4-0_13s007952-1-1_lane1        PAIRED           RANDOM   METAGENOMIC Homo sapiens ILLUMINA
# ERR478959         metagenome CCIS00146684ST-4-0_13s007952-2-1_lane2        PAIRED           RANDOM   METAGENOMIC Homo sapiens ILLUMINA
# ERR478960         metagenome CCIS00146684ST-4-0_13s007952-3-1_lane3        PAIRED           RANDOM   METAGENOMIC Homo sapiens ILLUMINA
# ERR478961         metagenome CCIS00146684ST-4-0_13s007952-4-1_lane5        PAIRED           RANDOM   METAGENOMIC Homo sapiens ILLUMINA
# ERR478962         metagenome CCIS00281083ST-3-0_11s002718-1-1_lane7        PAIRED           RANDOM   METAGENOMIC Homo sapiens ILLUMINA
# ERR478963         metagenome CCIS00281083ST-3-0_11s002718-1-1_lane8        PAIRED           RANDOM   METAGENOMIC Homo sapiens ILLUMINA
#                                                                           project_name          ReleaseDate  Sample Name        Sample_Name Sequencing_method SRA Study
# ERR478958 Potential of fecal microbiota for early stage detection of colorectal cancer 2014-11-02T00:00:00Z SAMEA2467039 CCIS00146684ST-4-0          Illumina ERP005534
# ERR478959 Potential of fecal microbiota for early stage detection of colorectal cancer 2014-11-02T00:00:00Z SAMEA2467039 CCIS00146684ST-4-0          Illumina ERP005534
# ERR478960 Potential of fecal microbiota for early stage detection of colorectal cancer 2014-11-02T00:00:00Z SAMEA2467039 CCIS00146684ST-4-0          Illumina ERP005534
# ERR478961 Potential of fecal microbiota for early stage detection of colorectal cancer 2014-11-02T00:00:00Z SAMEA2467039 CCIS00146684ST-4-0          Illumina ERP005534
# ERR478962 Potential of fecal microbiota for early stage detection of colorectal cancer 2014-11-02T00:00:00Z SAMEA2466896 CCIS00281083ST-3-0          Illumina ERP005534
# ERR478963 Potential of fecal microbiota for early stage detection of colorectal cancer 2014-11-02T00:00:00Z SAMEA2466896 CCIS00281083ST-3-0          Illumina ERP005534
#           body_mass_index common_name Diagnosis ENA-FIRST-PUBLIC ENA-LAST-UPDATE  External_Id geographic_location_(countryand/orsea_region) human_gut_environmental_package
# ERR478958              25       human    Normal       2014-11-01      2016-10-21 SAMEA2467039                                        France                       human-gut
# ERR478959              25       human    Normal       2014-11-01      2016-10-21 SAMEA2467039                                        France                       human-gut
# ERR478960              25       human    Normal       2014-11-01      2016-10-21 SAMEA2467039                                        France                       human-gut
# ERR478961              25       human    Normal       2014-11-01      2016-10-21 SAMEA2467039                                        France                       human-gut
# ERR478962              32       human    Normal       2014-11-01      2016-10-21 SAMEA2466896                                        France                       human-gut
# ERR478963              32       human    Normal       2014-11-01      2016-10-21 SAMEA2466896                                        France                       human-gut
#              sex       Submitter_Id AvgSpotLen      Bases body_product AJCC_Stage Alias ENA_CHECKLIST Environmental_package gender geographic_location_(country_and/or_sea
# ERR478958 female CCIS00146684ST-4-0        173 1546377290        stool       <NA>  <NA>          <NA>                  <NA>   <NA>                                    <NA>
# ERR478959 female CCIS00146684ST-4-0        172 1521749944        stool       <NA>  <NA>          <NA>                  <NA>   <NA>                                    <NA>
# ERR478960 female CCIS00146684ST-4-0        172 1524048617        stool       <NA>  <NA>          <NA>                  <NA>   <NA>                                    <NA>
# ERR478961 female CCIS00146684ST-4-0        172 1537171602        stool       <NA>  <NA>          <NA>                  <NA>   <NA>                                    <NA>
# ERR478962   male CCIS00281083ST-3-0        141 1392711576        stool       <NA>  <NA>          <NA>                  <NA>   <NA>                                    <NA>
# ERR478963   male CCIS00281083ST-3-0        141 1346817555        stool       <NA>  <NA>          <NA>                  <NA>   <NA>                                    <NA>
#           region) localization SRA_accession tissue_type Title TNM_Stage BMI
# ERR478958    <NA>         <NA>          <NA>        <NA>  <NA>      <NA>  NA
# ERR478959    <NA>         <NA>          <NA>        <NA>  <NA>      <NA>  NA
# ERR478960    <NA>         <NA>          <NA>        <NA>  <NA>      <NA>  NA
# ERR478961    <NA>         <NA>          <NA>        <NA>  <NA>      <NA>  NA
# ERR478962    <NA>         <NA>          <NA>        <NA>  <NA>      <NA>  NA
# ERR478963    <NA>         <NA>          <NA>        <NA>  <NA>      <NA>  NA


## NOTE THESE RUNS DID NOT PASS QC (FAILED TO PRODUCE GOOD READ 1's FROM SEQUENCE DATA)

failed_runs <- c("ERR479093","ERR479094","ERR479095","ERR479096","ERR479146","ERR479148","ERR479176","ERR479322")
sel <- which(row.names(meta) %in% failed_runs)
meta.ok <- meta[-sel, ]

sampid <- unique(meta.ok$host_subject_id) # USING meta.ok !!!!!!!!!!!!!!!!!!!!


superfocus_out_dir <- "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc"

list.dirs(superfocus_out_dir)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc"                         
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478958"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478959"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478960"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478961"
# etc ...


# # don't keep 1st directory
( results_dirs <- list.dirs(superfocus_out_dir)[-c(1)] )
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478958"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478959"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478960"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478961"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/zeller-crc/superfocus_out_ERR478962"
# etc ...

head(results_dirs)



## In this data there are multiple Runs for a single Sample_ID !!!


# collate results into a long-format table

#sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)
sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA, full_fxn_tax=NA)

for (i in 1:length(sampid)) {
  #i<-1
  this_samp <- sampid[i]
  # identify corresponding Runs?
  runs <- meta.ok$Run[ which(meta.ok$host_subject_id == this_samp) ] # 4 Runs    # USING meta.ok !!!!!!!!!!!!!!!!!!!!
  
  if (length(runs)>=1) {
    
    tab.temp <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)
    
    for (j in 1:length(runs)) {
      #j<-4
      this_run <- runs[j]
      sel.folder <- grep(pattern = this_run, x = results_dirs)
      this_folder <- results_dirs[sel.folder]
      
      #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)
      tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
      
      # names(tab)
      # [1] "Subsystem.Level.1"                                                           
      # [2] "Subsystem.Level.2"                                                           
      # [3] "Subsystem.Level.3"                                                           
      # [4] "Function"                                                                    
      # [5] "X.scratch.user.lidd0026.zeller.crc.zcrc_2_fastp_qc.ERR478958_R1.good.fastq"  
      # [6] "X.scratch.user.lidd0026.zeller.crc.zcrc_2_fastp_qc.ERR478958_R1.good.fastq.." # this is %
      
      tab$sampid <- this_samp
      #names(tab)
      
      # last column is sampid
      # take average of percentages
      
      sel.col.percent <- grep(pattern = "R1.good.fastq..$", x = names(tab))
      if (length(sel.col.percent)>1) {
        tab$percent_abun <- apply(X = tab[ ,sel.col.percent], MARGIN = 1, FUN = mean )
      } else {
        tab$percent_abun <- tab[ ,sel.col.percent]
      }
      
      # sum(tab$percent_abun) # 100
      # mean(tab$percent_abun) # 0.004338583
      
      #names(sfx.long) # "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"    "percent_abun"  "full_fxn_tax"
      #names(tab)
      # [1] "Subsystem.Level.1"                                                                                                 
      # [2] "Subsystem.Level.2"                                                                                                 
      # [3] "Subsystem.Level.3"                                                                                                 
      # [4] "Function" 
      # ...
      # [13] "sampid"                                                                                                            
      # [14] "percent_abun"   
      
      tab <- tab[ ,c("sampid","Subsystem.Level.1","Subsystem.Level.2","Subsystem.Level.3","Function","percent_abun")]
      names(tab) <- names(tab.temp)
      
      tab.temp <- rbind(tab.temp,tab)
      
      print(paste0("run # ",j))
      
    } # END j/runs loop
  } # END if length()
  
  #head(tab.temp)
  tab.temp <- tab.temp[-1, ]
  
  # tidy up repeated functions across multiple runs
  # normalise percent_abun from multiple runs back to 100%
  tab.temp$percent_abun <- tab.temp$percent_abun/length(runs)
  
  tab.temp$full_fxn_tax <- paste0(tab.temp$subsys_L1,"___", tab.temp$subsys_L2,"___", tab.temp$subsys_L3,"___", tab.temp$fxn)
  head(tab.temp)
  
  unique_fxn_rows <- unique(tab.temp$full_fxn_tax)
  #length(unique_fxn_rows) # 2768
  for (f in 1:length(unique_fxn_rows)) {
    #f<-3
    sel.rows <- which(tab.temp$full_fxn_tax == unique_fxn_rows[f])
    # if no of rows is greater than 1, then take average and delete remaining rows
    if (length(sel.rows)>1) {
      new_value <- sum(tab.temp$percent_abun[sel.rows])
      # replace with sum - as it has been previously normalised to 100%
      tab.temp$percent_abun[ sel.rows[1] ] <- new_value
      # identify rows to delete
      subsel.obsolete <- which(!sel.rows %in% sel.rows[1])
      
      tab.temp <- tab.temp[ -sel.rows[subsel.obsolete] ,]
    } # END if length > 1
    
  } # END f loop for repeated fxn rows
  
  sfx.long <- rbind(sfx.long,tab.temp)
  
  print(paste0("completed ",i," - sample ID: ",sampid[i]))
  
} # END for length sampid



head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
dim(sfx.long) # 317404      7
head(sfx.long)
# sampleID                   subsys_L1                    subsys_L2                           subsys_L3
# 2   FR-726 Amino Acids and Derivatives                            -                 Amino acid racemase
# 3   FR-726 Amino Acids and Derivatives                            - Creatine and Creatinine Degradation
# 4   FR-726 Amino Acids and Derivatives                            - Creatine and Creatinine Degradation
# 5   FR-726 Amino Acids and Derivatives                            - Creatine and Creatinine Degradation
# 6   FR-726 Amino Acids and Derivatives Alanine, serine, and glycine                Alanine biosynthesis
# 7   FR-726 Amino Acids and Derivatives Alanine, serine, and glycine                Alanine biosynthesis
# fxn percent_abun
# 2                            Alanine_racemase_(EC_5.1.1.1)  0.001550676
# 3                  Creatinine_amidohydrolase_(EC_3.5.2.10)  0.033339686
# 4                          Cytosine_deaminase_(EC_3.5.4.1)  0.019300564
# 5                                        Cytosine_permease  0.014300060
# 6                            Alanine_racemase_(EC_5.1.1.1)  0.003973626
# 7 Branched-chain_amino_acid_aminotransferase_(EC_2.6.1.42)  0.025061002
# full_fxn_tax
# 2                                                        Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)
# 3                              Amino Acids and Derivatives___-___Creatine and Creatinine Degradation___Creatinine_amidohydrolase_(EC_3.5.2.10)
# 4                                      Amino Acids and Derivatives___-___Creatine and Creatinine Degradation___Cytosine_deaminase_(EC_3.5.4.1)
# 5                                                    Amino Acids and Derivatives___-___Creatine and Creatinine Degradation___Cytosine_permease
# 6                            Amino Acids and Derivatives___Alanine, serine, and glycine___Alanine biosynthesis___Alanine_racemase_(EC_5.1.1.1)
# 7 Amino Acids and Derivatives___Alanine, serine, and glycine___Alanine biosynthesis___Branched-chain_amino_acid_aminotransferase_(EC_2.6.1.42)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 12896   114

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 12896
length(unique(full_fxn_names)) # 12896

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_##_catabolic" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___Aspartate_racemase_(EC_5.1.1.13)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___Diaminopimelate_epimerase_(EC_5.1.1.7)" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___Glutamate_racemase_(EC_5.1.1.3)" 


tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)
# full_fxn_tax FR-001 FR-003 FR-024 FR-025 FR-027 FR-030 FR-039 FR-040
# 1      Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase      0      0      0      0      0      0      0      0
# 2              Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)      0      0      0      0      0      0      0      0
# 3 Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_##_catabolic      0      0      0      0      0      0      0      0
# 4           Amino Acids and Derivatives___-___Amino acid racemase___Aspartate_racemase_(EC_5.1.1.13)      0      0      0      0      0      0      0      0
# 5     Amino Acids and Derivatives___-___Amino acid racemase___Diaminopimelate_epimerase_(EC_5.1.1.7)      0      0      0      0      0      0      0      0
# 6            Amino Acids and Derivatives___-___Amino acid racemase___Glutamate_racemase_(EC_5.1.1.3)      0      0      0      0      0      0      0      0
# FR-052       FR-053 FR-060       FR-105 FR-110 FR-116       FR-118      FR-121       FR-125 FR-129      FR-132      FR-139 FR-142      FR-152 FR-156
# 1      0 0.0000000000      0 0.0000000000      0      0 0.0000000000 0.000000000 0.0000000000      0 0.000000000 0.000000000      0 0.000000000      0
# 2      0 0.0000000000      0 0.0000000000      0      0 0.0000000000 0.000000000 0.0005796953      0 0.000000000 0.000000000      0 0.001519572      0
# 3      0 0.0000000000      0 0.0000000000      0      0 0.0000000000 0.000000000 0.0000000000      0 0.000000000 0.000000000      0 0.000000000      0
# 4      0 0.0000000000      0 0.0000000000      0      0 0.0000000000 0.000000000 0.0002962366      0 0.000000000 0.000000000      0 0.000000000      0
# 5      0 0.0008863991      0 0.0008175278      0      0 0.0008704398 0.003751928 0.0004251880      0 0.001931397 0.002589064      0 0.006793817      0
# 6      0 0.0000000000      0 0.0015818780      0      0 0.0053942384 0.000000000 0.0009491216      0 0.000000000 0.000000000      0 0.001111080      0
# etc ....

names(sfx.wide)
# [1] "full_fxn_tax" "FR-001"       "FR-003"       "FR-024"       "FR-025"       "FR-027"       "FR-030"       "FR-039"       "FR-040"       "FR-052"       "FR-053"      
# [12] "FR-060"       "FR-105"       "FR-110"       "FR-116"       "FR-118"       "FR-121"       "FR-125"       "FR-129"       "FR-132"       "FR-139"       "FR-142"      
# [23] "FR-152"       "FR-156"       "FR-166"       "FR-169"       "FR-170"       "FR-173"       "FR-187"       "FR-192"       "FR-193"       "FR-194"       "FR-195"      
# [34] "FR-198"       "FR-200"       "FR-202"       "FR-208"       "FR-212"       "FR-213"       "FR-221"       "FR-223"       "FR-268"       "FR-275"       "FR-276"      
# [45] "FR-281"       "FR-293"       "FR-294"       "FR-298"       "FR-302"       "FR-305"       "FR-310"       "FR-316"       "FR-328"       "FR-344"       "FR-349"      
# [56] "FR-376"       "FR-390"       "FR-393"       "FR-398"       "FR-399"       "FR-400"       "FR-404"       "FR-414"       "FR-430"       "FR-450"       "FR-451"      
# [67] "FR-459"       "FR-460"       "FR-503"       "FR-504"       "FR-505"       "FR-506"       "FR-542"       "FR-551"       "FR-552"       "FR-558"       "FR-568"      
# [78] "FR-617"       "FR-628"       "FR-652"       "FR-654"       "FR-667"       "FR-682"       "FR-684"       "FR-696"       "FR-716"       "FR-719"       "FR-721"      
# [89] "FR-722"       "FR-723"       "FR-726"       "FR-728"       "FR-730"       "FR-734"       "FR-751"       "FR-759"       "FR-767"       "FR-768"       "FR-770"      
# [100] "FR-772"       "FR-780"       "FR-783"       "FR-788"       "FR-812"       "FR-815"       "FR-817"       "FR-823"       "FR-824"       "FR-825"       "FR-827"      
# [111] "FR-828"       "FR-830"       "FR-901"       "FR-902" 

names(sfx.wide) <- gsub(pattern = "-", replacement = "_", x = names(sfx.wide))

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide)
# [1] "FR_001" "FR_003" "FR_024" "FR_025" "FR_027" "FR_030" "FR_039" "FR_040" "FR_052" "FR_053" "FR_060" "FR_105" "FR_110" "FR_116" "FR_118" "FR_121"
# [17] "FR_125" "FR_129" "FR_132" "FR_139" "FR_142" "FR_152" "FR_156" "FR_166" "FR_169" "FR_170" "FR_173" "FR_187" "FR_192" "FR_193" "FR_194" "FR_195"
# [33] "FR_198" "FR_200" "FR_202" "FR_208" "FR_212" "FR_213" "FR_221" "FR_223" "FR_268" "FR_275" "FR_276" "FR_281" "FR_293" "FR_294" "FR_298" "FR_302"
# [49] "FR_305" "FR_310" "FR_316" "FR_328" "FR_344" "FR_349" "FR_376" "FR_390" "FR_393" "FR_398" "FR_399" "FR_400" "FR_404" "FR_414" "FR_430" "FR_450"
# [65] "FR_451" "FR_459" "FR_460" "FR_503" "FR_504" "FR_505" "FR_506" "FR_542" "FR_551" "FR_552" "FR_558" "FR_568" "FR_617" "FR_628" "FR_652" "FR_654"
# [81] "FR_667" "FR_682" "FR_684" "FR_696" "FR_716" "FR_719" "FR_721" "FR_722" "FR_723" "FR_726" "FR_728" "FR_730" "FR_734" "FR_751" "FR_759" "FR_767"
# [97] "FR_768" "FR_770" "FR_772" "FR_780" "FR_783" "FR_788" "FR_812" "FR_815" "FR_817" "FR_823" "FR_824" "FR_825" "FR_827" "FR_828" "FR_830" "FR_901"
# [113] "FR_902"

head(sampid)
# "FR-726" "FR-060" "FR-568" "FR-828" "FR-027" "FR-192"
length(sampid) # 113

names(sampid) # NULL - in this case there is NOT an alternative sample name being used

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # FALSE
identical(names(sfx.wide) , as.character(gsub(pattern = "-",replacement = "_",x = sampid))) # FALSE
length( names(sfx.wide) %in% as.character(gsub(pattern = "-",replacement = "_",x = sampid)) ) # 113 - i.e. matching but order different

#NOT RUN THIS TIME
#names(sfx.wide) <- names(sampid)



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 12896     4

length(unique(tax.fxn$subsys_L1)) # 34
length(unique(tax.fxn$subsys_L2)) # 165
length(unique(tax.fxn$subsys_L3)) # 946
length(unique(tax.fxn$fxn)) # 7067

# sel.dup <- which(duplicated.default(tax.fxn$fxn)==TRUE)
# head(tax.fxn$fxn[sel.dup])
# 
# 
# 
# # examine last example 'tab'
# 
# head(tab)
# #   sampleID                   subsys_L1 subsys_L2           subsys_L3                                                                           fxn percent_abun
# # 1     9468 Amino Acids and Derivatives         - Amino acid racemase                                         2-methylaconitate_cis-trans_isomerase 3.819676e-04
# # 2     9468 Amino Acids and Derivatives         - Amino acid racemase                                                   2-methylaconitate_isomerase 1.766030e-05
# # 3     9468 Amino Acids and Derivatives         - Amino acid racemase 2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117) 2.587839e-05
# # 4     9468 Amino Acids and Derivatives         - Amino acid racemase                       2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79) 3.514521e-05
# # 5     9468 Amino Acids and Derivatives         - Amino acid racemase                                       4-hydroxyproline_epimerase_(EC_5.1.1.8) 1.287010e-03
# # 6     9468 Amino Acids and Derivatives         - Amino acid racemase                                    4-oxalomesaconate_tautomerase_(EC_5.3.2.8) 4.401921e-05
# 
# 
# length(unique(tab$subsys_L1)) # 35
# length(unique(tab$subsys_L2)) # 191
# length(unique(tab$subsys_L3)) # 1184
# length(unique(tab$fxn)) # 10975

#-------------------------

#### Zeller CRC - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 12896     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 12896   113

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 12896 taxa and 113 samples ]
# tax_table()   Taxonomy Table:    [ 12896 taxa by 4 taxonomic ranks ]

sample_names(phy)
# [1] "FR_001" "FR_003" "FR_024" "FR_025" "FR_027" "FR_030" "FR_039" "FR_040" "FR_052" "FR_053" "FR_060" "FR_105" "FR_110" "FR_116" "FR_118" "FR_121"
# [17] "FR_125" "FR_129" "FR_132" "FR_139" "FR_142" "FR_152" "FR_156" "FR_166" "FR_169" "FR_170" "FR_173" "FR_187" "FR_192" "FR_193" "FR_194" "FR_195"
# [33] "FR_198" "FR_200" "FR_202" "FR_208" "FR_212" "FR_213" "FR_221" "FR_223" "FR_268" "FR_275" "FR_276" "FR_281" "FR_293" "FR_294" "FR_298" "FR_302"
# [49] "FR_305" "FR_310" "FR_316" "FR_328" "FR_344" "FR_349" "FR_376" "FR_390" "FR_393" "FR_398" "FR_399" "FR_400" "FR_404" "FR_414" "FR_430" "FR_450"
# [65] "FR_451" "FR_459" "FR_460" "FR_503" "FR_504" "FR_505" "FR_506" "FR_542" "FR_551" "FR_552" "FR_558" "FR_568" "FR_617" "FR_628" "FR_652" "FR_654"
# [81] "FR_667" "FR_682" "FR_684" "FR_696" "FR_716" "FR_719" "FR_721" "FR_722" "FR_723" "FR_726" "FR_728" "FR_730" "FR_734" "FR_751" "FR_759" "FR_767"
# [97] "FR_768" "FR_770" "FR_772" "FR_780" "FR_783" "FR_788" "FR_812" "FR_815" "FR_817" "FR_823" "FR_824" "FR_825" "FR_827" "FR_828" "FR_830" "FR_901"
# [113] "FR_902"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta.ok) # 343 66
dim(subjects) # 114 13

keep_subjects <- unique(meta.ok$host_subject_id) # qty 113
keep_subjects <- gsub(pattern = "-", replacement = "_", x = keep_subjects)
sel <- which(row.names(subjects) %in% keep_subjects)
subjects.ok <- subjects[sel, ]

identical( row.names(subjects.ok), sample_names(phy) ) # TRUE
head(row.names(subjects.ok)) # "FR_001" "FR_003" "FR_024" "FR_025" "FR_027" "FR_030"
head(sample_names(phy)) # "FR_001" "FR_003" "FR_024" "FR_025" "FR_027" "FR_030"


SAMP <- sample_data(subjects.ok)



### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 12896 taxa and 113 samples ]
# sample_data() Sample Data:       [ 113 samples by 13 sample variables ]
# tax_table()   Taxonomy Table:    [ 12896 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#   subsys_L1                     subsys_L2 subsys_L3             fxn                                         
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylaconitate_cis-trans_isomerase"     
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Alanine_racemase_(EC_5.1.1.1)"             
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Alanine_racemase_(EC_5.1.1.1)_##_catabolic"
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Aspartate_racemase_(EC_5.1.1.13)"          
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Diaminopimelate_epimerase_(EC_5.1.1.7)"    
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Glutamate_racemase_(EC_5.1.1.3)"


getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-Zeller-CRC.RDS")


#phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")



#-------------------------

#### Zeller CRC - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")


## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 12896     4



get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)
        
        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1
          
          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else
        
      } # END else
      
      coefs[[c]] <- coef
      
    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  
  
  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      

    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Zeller-CRC-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%  #
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()



time.start #  "2023-04-07 10:58:05 ACST"
time.finish # "2023-04-07 11:53:54 ACST"

# ~1 hours

## assemble results

# results saved here

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Zeller-CRC-R-working-files"



# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# total results written to disk - 37335
dim(df.tax) # 12896     4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 14852     9

table(df.out$matching_method)
# EC number      Exact hierachy match Matched Reactions aliases    Matched Reactions name     Matched Subsytem role 
# 5207                      2892                        25                       183                       366 
# No match found 
# 6179 


100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 41.60382% No match found

# 58% match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.621   0.827   0.869   1.103   2.500    5828

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.257   1.401   1.436   1.618   4.000    5828

table(df.out.distil$matching_method)
# EC number                EC number;Exact hierachy match EC number;Exact hierachy match;No match found 
# 3913                                           326                                            10 
# EC number;Matched Subsytem role                      EC number;No match found                          Exact hierachy match 
# 43                                           288                                          2175 
# Exact hierachy match;Matched Subsytem role           Exact hierachy match;No match found                     Matched Reactions aliases 
# 15                                           114                                            24 
# Matched Reactions aliases;No match found                        Matched Reactions name         Matched Reactions name;No match found 
# 1                                           182                                             1 
# Matched Subsytem role          Matched Subsytem role;No match found                                No match found 
# 270                                            34                                          5500


100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 37.03205% No match found

100-37 # i.e. in 63% of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_Zeller-CRC.RDS")

#df.out.distil <- readRDS(file = "df.out.distil_Zeller-CRC.RDS")

#-------------------------

#### Zeller CRC - inspect & clean data
#-------------------------

this_study <- "-Zeller-CRC-"
phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Zeller-CRC.RDS")

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 12896     4

phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE

df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 12896   113
df.OTU[1:5, 1:8]
# FR_001 FR_003 FR_024 FR_025 FR_027 FR_030 FR_039 FR_040
# fxn_1      0      0      0      0      0      0      0      0
# fxn_2      0      0      0      0      0      0      0      0
# fxn_3      0      0      0      0      0      0      0      0
# fxn_4      0      0      0      0      0      0      0      0
# fxn_5      0      0      0      0      0      0      0      0

mean( df.OTU[ , "FR_001"] ) # 0.007754342
sum( df.OTU[ , "FR_001"] ) # 100

sample_sums(phy) # all values of 100

identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
# tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.6666667      2.333333
# fxn_3     0.6666667      2.333333
# fxn_4     1.0000000      1.500000
# fxn_5     0.5714286      2.000000
# fxn_6     0.8000000      1.600000


dim(df.fxn_vk_coords) # 12896     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.621   0.827   0.869   1.103   2.500    5828 
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.257   1.401   1.436   1.618   4.000    5828

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.6666667      2.333333
# fxn_3     0.6666667      2.333333
# fxn_4     1.0000000      1.500000
# fxn_5     0.5714286      2.000000
# fxn_6     0.8000000      1.600000
# fxn_7     0.5206427      1.400701


## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 113
dim(df.fxn_vk_coords.noNAs) # 7068    2
dim(df.OTU.noNAs) # 7068 113

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# FR_039   FR_430   FR_823   FR_116   FR_060   FR_040   FR_208   FR_824   FR_825   FR_780   FR_223   FR_400   FR_542   FR_504   FR_770   FR_682 
# 57.17807 60.17195 61.22960 62.05105 62.13598 62.17844 62.24284 62.42623 62.59714 62.74135 62.81389 62.94704 63.26847 63.36451 63.38726 63.38865 
# FR_376   FR_459   FR_652   FR_767   FR_506   FR_719   FR_118   FR_827   FR_450   FR_551   FR_142   FR_110   FR_275   FR_302   FR_202   FR_281 
# 63.40930 63.43315 63.92203 64.13896 64.38232 64.60029 64.62076 64.63143 64.73149 64.91355 64.91945 64.96957 65.21732 65.32030 65.43764 65.48675 
# FR_788   FR_132   FR_654   FR_052   FR_193   FR_195   FR_503   FR_817   FR_310   FR_125   FR_166   FR_734   FR_393   FR_726   FR_721   FR_460 
# 65.48980 65.51813 65.53113 65.74233 65.76730 65.82238 65.85735 65.90000 65.90270 65.96878 65.97332 65.97348 66.00251 66.19938 66.25132 66.25675 
# FR_344   FR_723   FR_667   FR_030   FR_768   FR_451   FR_772   FR_404   FR_505   FR_105   FR_053   FR_830   FR_902   FR_552   FR_192   FR_730 
# 66.32567 66.32990 66.47217 66.52933 66.53299 66.63698 66.81097 66.82991 66.85887 66.97861 66.98436 66.99983 67.08757 67.33988 67.35609 67.45113 
# FR_212   FR_213   FR_200   FR_812   FR_001   FR_684   FR_828   FR_696   FR_390   FR_276   FR_815   FR_027   FR_751   FR_268   FR_139   FR_398 
# 67.49434 67.54200 67.57569 67.66611 67.72740 67.81263 67.92575 67.95127 67.96181 68.04988 68.07635 68.16175 68.16853 68.20307 68.21383 68.22142 
# FR_194   FR_759   FR_783   FR_187   FR_024   FR_152   FR_628   FR_716   FR_414   FR_121   FR_170   FR_568   FR_173   FR_901   FR_328   FR_558 
# 68.27586 68.27612 68.29465 68.29569 68.49309 68.57795 68.59898 68.66388 68.66946 68.78727 68.80062 68.83945 68.92276 68.95072 69.07579 69.20615 
# FR_722   FR_169   FR_025   FR_728   FR_198   FR_293   FR_316   FR_003   FR_129   FR_298   FR_305   FR_221   FR_156   FR_617   FR_399   FR_349 
# 69.20732 69.40608 69.48166 69.50192 69.59747 69.65478 69.75207 69.76796 69.88595 70.08364 70.11154 70.22970 71.30883 71.42144 71.49478 73.27327 
# FR_294 
# 74.91881

summary( colSums(df.OTU.noNAs) )
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 57.18   65.22   66.86   66.69   68.49   74.92

# this reflects that 57-75% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates


colSums(df.OTU)
# all values of 100


dim(df.tax) # 12896     4


# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE


dim(df.fxn_vk_coords.noNAs) # 7068    2
dim(df.in) # 7068    113

row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  group=NA, 
                                  samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"],
                         group= rep( phy@sam_data$Diagnosis[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}



dim(vk.samp.noNAs.long) # 798685      8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok)

str(vk.samp.noNAs.long)
unique(vk.samp.noNAs.long$group) # "Normal" "Cancer"
class(vk.samp.noNAs.long$group) # "character"
vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c("Cancer", "Normal"), ordered = TRUE) # CHECK GROUP VARIABLE !!
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	798684 obs. of  8 variables:
# $ sample   : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ fxn_id   : Factor w/ 7068 levels "fxn_10","fxn_100",..: 2055 2868 3521 4059 4856 5501 6015 6613 1 512 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.667 0.667 1 0.571 0.8 ...
# $ HC_y     : num  2.33 2.33 1.5 2 1.6 ...
# $ group    : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" ...


saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-object-Zeller-CRC.RDS")


#vk.samp.noNAs.long <- readRDS(file = "vk.samp.noNAs.long-object-Zeller-CRC.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.text = element_text(angle=25),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "Colorectal cancer & Normal human gut", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust =0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")




## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Zeller-CRC-R-working-files"

i<-2
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
names(temp)
names( temp[[1]] ) # "fxns" "rxns"
temp[[1]][["fxns"]]
# superfocus_fxn f                         f__in matching_method min_adist_modelSEED min_amatch_modelSEED                       rxns tot_mean_OC_x tot_mean_HC_y
# 1          fxn_2 1 Alanine racemase (EC 5.1.1.1)       EC number                  NA                   NA rxn00283;rxn19085;rxn38030     0.6666667      2.333333
temp[[1]][["rxns"]]
# [[1]]
# superfocus_fxn f                         f__in   rxn_id           rxn_name                             rxn_eqn                              rxn_defn
# 1          fxn_2 1 Alanine racemase (EC 5.1.1.1) rxn00283   alanine racemase (1) cpd00035[0] <=> (1) cpd00117[0] (1) L-Alanine[0] <=> (1) D-Alanine[0]
# 2          fxn_2 1 Alanine racemase (EC 5.1.1.1) rxn19085 L-alanine racemase (1) cpd00035[0] <=> (1) cpd00117[0] (1) L-Alanine[0] <=> (1) D-Alanine[0]
# 3          fxn_2 1 Alanine racemase (EC 5.1.1.1) rxn38030   ALARACECAT-RXN.e (1) cpd00035[0] <=> (1) cpd00117[0] (1) L-Alanine[0] <=> (1) D-Alanine[0]
# compds compd_coef      chem_formx                           OC_ratios                         HC_ratios coefwtmean_OC_x coefwtmean_HC_y
# 1 cpd00035;cpd00117        1;1 C3H7NO2;C3H7NO2 0.666666666666667;0.666666666666667 2.33333333333333;2.33333333333333       0.6666667        2.333333
# 2 cpd00035;cpd00117        1;1 C3H7NO2;C3H7NO2 0.666666666666667;0.666666666666667 2.33333333333333;2.33333333333333       0.6666667        2.333333
# 3 cpd00035;cpd00117        1;1 C3H7NO2;C3H7NO2 0.666666666666667;0.666666666666667 2.33333333333333;2.33333333333333       0.6666667        2.333333


## check anomalous high value in Lignin zone in PCaN density plot
#p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405)

length(sel) # 15368
dim(vk.samp.noNAs.long) # 798684      8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 1.924% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
#     sample  fxn_id abun abun_type      OC_x     HC_y  group       samplabel
# 7   FR_001   fxn_7    0   relabun 0.5206427 1.400701 Normal FR_001 (Normal)
# 49  FR_001  fxn_57    0   relabun 0.5206427 1.400701 Normal FR_001 (Normal)
# 114 FR_001 fxn_136    0   relabun 0.5206427 1.400701 Normal FR_001 (Normal)
# 161 FR_001 fxn_190    0   relabun 0.5206427 1.400701 Normal FR_001 (Normal)
# 184 FR_001 fxn_222    0   relabun 0.5206427 1.400701 Normal FR_001 (Normal)
# 203 FR_001 fxn_259    0   relabun 0.5206427 1.400701 Normal FR_001 (Normal)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.02637615 = 0.026% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 66.69482% per sample


#fxn_7
i<-7
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED
# 1          fxn_7 1 hypothetical protein Matched Reactions name                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427      1.400701



### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 15368


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-Zeller-CRC.RDS")


#-------------------------

#### Zeller CRC - CPP-class: total rel abun in vk spatial regions
#-------------------------

this_study <- "-Zeller-CRC-"
header <- "-vk-zones-"

phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")

df.out.distil <- readRDS(file = "df.out.distil_Zeller-CRC.RDS")

df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Zeller-CRC.RDS")

# use clean dataset
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


str(vk.samp.noNAs.long)
# 'data.frame':	783316 obs. of  8 variables:
# $ sample   : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ fxn_id   : Factor w/ 7068 levels "fxn_10","fxn_100",..: 2055 2868 3521 4059 4856 6015 6613 1 512 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.667 0.667 1 0.571 0.8 ...
# $ HC_y     : num  2.33 2.33 1.5 2 1.6 ...
# $ group    : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" ...

# vK spatial data

df.temp <- vk.samp.noNAs.long # reminder this is long format data 


spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]
# add sex
spdf$sex <- NA
samps <- unique(spdf$sample)
length(samps) # 113
length(which(phy@sam_data$Subject_ID %in% samps)) # 113
for (i in 1:length(samps)) {
  #i<-1
  this_samp <- samps[i]
  sel.spdf <- which(spdf$sample == this_samp)
  sel.meta <- which(phy@sam_data$Subject_ID == this_samp)
  spdf$sex[sel.spdf] <- phy@sam_data$Gender[sel.meta]
  print(paste0("completed ",i," - no rows = ",length(sel.spdf)))
}

dim(spdf) #  197710      6

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# F      M 
# 97270 100440 
sel.na <- which(spdf$sex=="NA") # empty
# unique(spdf$sample[sel.na]) #
# spdf <- spdf[-sel.na, ]

head(spdf)
# sample  group      OC_x     HC_y       abun sex
# 25  FR_001 Normal 0.6722222 1.511111 0.10683761   F
# 83  FR_001 Normal 1.5000000 1.130952 0.32051282   F
# 92  FR_001 Normal 0.5777778 3.033333 0.19880716   F
# 117 FR_001 Normal 0.7333333 2.100000 0.04970179   F
# 124 FR_001 Normal 1.9666667 2.255556 0.04970179   F
# 145 FR_001 Normal 0.6444444 1.566667 0.19880716   F

#dim(spdf) # 


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 113

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 57.18   65.21   66.85   66.67   68.44   74.85

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"  

## now want to sum total rela abun within vK spatial zones

sum_relabun_in_vk_zones <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  # lipids
  subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  subsel.all <- c(subsel.lipids)
  # proteins
  subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.proteins)
  # amino_sugars
  subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.amino_sugars)
  # carbs - Carbohydrates
  subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  subsel.all <- c(subsel.all, subsel.carbs)
  # cond_aromtx - Condensed aromatics
  subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  subsel.all <- c(subsel.all, subsel.cond_aromtx)
  # lignin
  subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.lignin)
  # tannins
  subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.tannins)
  # Other compounds ...
  
  # methylated
  subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
  # and not already used?
  sel.dup <- which(subsel.methylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.methylated)
  
  # hydrated
  subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
  # and not already used?
  sel.dup <- which(subsel.hydrated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.hydrated)
  
  # demethylated / Oxidised-dehydrogenated
  subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.demethylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.demethylated)
  
  # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
  #subsel.condensed = which(                   spdf_in$OC_x <= 0.97                       & spdf_in$HC_y < 0.75)
  subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
  
  # and not already used?
  sel.dup <- which(subsel.condensed %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.condensed)
  
  # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
  subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.k2_b12 %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
  
  df_out <- data.frame(
    sample = this_samp,
    group = unique(spdf_in$group),
    sex = unique(spdf_in$sex),
    
    lipids = sum(spdf_in$abun[ subsel.lipids ]),
    proteins = sum(spdf_in$abun[ subsel.proteins ]),
    amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
    carbs = sum(spdf_in$abun[ subsel.carbs ]),
    cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
    lignin = sum(spdf_in$abun[ subsel.lignin ]),
    tannins = sum(spdf_in$abun[ subsel.tannins ]),
    # Other compounds ...
    methylated = sum(spdf_in$abun[ subsel.methylated ]),
    hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
    demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
    condensed = sum(spdf_in$abun[ subsel.condensed ]),
    k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
    
  )
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 113
out[[1]]
# sample  group sex lipids proteins amino_sugars    carbs cond_aromtx   lignin tannins methylated hydrated demethylated  condensed    k2_b12
# 1 FR_001 Normal   F      0 4.032854     2.054978 20.92095           0 6.704007 17.8138  0.1988072 5.137445     10.61605 0.09940358 0.1491054

names(out[[1]])
# [1] "sample"       "group"        "sex"          "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"      
# [10] "tannins"      "methylated"   "hydrated"     "demethylated" "condensed"    "k2_b12"    



df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample  group sex     lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated  condensed    k2_b12
# 1 FR_001 Normal   F 0.00000000 4.032854     2.054978 20.92095  0.00000000 6.704007 17.81380  0.1988072 5.137445    10.616047 0.09940358 0.1491054
# 2 FR_003 Normal   M 0.06537675 3.798994     2.093552 31.48553  0.05580846 4.266416 14.59883  0.9164036 3.548767     8.281397 0.47495822 0.1819197
# 3 FR_024 Normal   F 0.18374737 3.694617     2.733026 23.04244  0.08661294 5.466318 17.84014  0.6232216 4.742808     9.331765 0.47586047 0.2204881
# 4 FR_025 Normal   F 0.07305875 3.126871     1.869436 28.02567  0.12356634 3.783079 18.86076  0.9618951 4.409713     7.678094 0.31969712 0.2403182
# 5 FR_027 Normal   M 0.26198224 3.375054     2.151553 29.04179  0.04394986 4.094285 16.29724  1.1025614 3.863885     7.046685 0.41408541 0.4642839
# 6 FR_030 Normal   M 0.10158547 2.942853     2.340780 29.22596  0.05819979 3.673044 15.52753  1.0664224 2.965823     7.714871 0.54172548 0.3551452

df.zones$group_mf <- paste0(df.zones$group,"__",df.zones$sex)
unique(df.zones$group_mf)
# "Normal__F" "Normal__M" "Cancer__F" "Cancer__M"
df.zones$group_mf <- factor(df.zones$group_mf, levels = c("Cancer__F","Normal__F","Cancer__M","Normal__M"),
                            labels = c("Cancer (F)","Normal (F)","Cancer (M)","Normal (M)"), 
                            ordered = TRUE)

names(df.zones)
# [1] "sample"       "group"        "sex"          "lipids"       "proteins"     "amino_sugars" "carbs"       
# [8] "cond_aromtx"  "lignin"       "tannins"      "methylated"   "hydrated"     "demethylated" "condensed"   
# [15] "k2_b12"       "group_mf" 

names(df.zones[ ,c(4:15)])

df.zones$all <- apply(X = df.zones[ ,c(4:15)], MARGIN = 1, FUN = sum)

table(df.zones$group_mf)
# Cancer (F) Normal (F) Cancer (M) Normal (M) 
#         24         33         29         27

saveRDS(object = df.zones, file = "df.zones--Zeller-CRC.RDS")

str(df.zones)
# 'data.frame':	113 obs. of  17 variables:
# $ sample      : chr  "FR_001" "FR_003" "FR_024" "FR_025" ...
# $ group       : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex         : chr  "F" "M" "F" "F" ...
# $ lipids      : num  0 0.0654 0.1837 0.0731 0.262 ...
# $ proteins    : num  4.03 3.8 3.69 3.13 3.38 ...
# $ amino_sugars: num  2.05 2.09 2.73 1.87 2.15 ...
# $ carbs       : num  20.9 31.5 23 28 29 ...
# $ cond_aromtx : num  0 0.0558 0.0866 0.1236 0.0439 ...
# $ lignin      : num  6.7 4.27 5.47 3.78 4.09 ...
# $ tannins     : num  17.8 14.6 17.8 18.9 16.3 ...
# $ methylated  : num  0.199 0.916 0.623 0.962 1.103 ...
# $ hydrated    : num  5.14 3.55 4.74 4.41 3.86 ...
# $ demethylated: num  10.62 8.28 9.33 7.68 7.05 ...
# $ condensed   : num  0.0994 0.475 0.4759 0.3197 0.4141 ...
# $ k2_b12      : num  0.149 0.182 0.22 0.24 0.464 ...
# $ group_mf    : Ord.factor w/ 4 levels "Cancer (F)"<"Normal (F)"<..: 2 4 2 2 4 4 4 4 2 2 ...
# $ all         : num  67.7 69.8 68.4 69.5 68.2 ...


df.zones.melt <- melt(df.zones, id.vars = c("sample","group","sex","group_mf"))
head(df.zones.melt)
# sample  group sex   group_mf variable      value
# 1 FR_001 Normal   F Normal (F)   lipids 0.00000000
# 2 FR_003 Normal   M Normal (M)   lipids 0.06537675
# 3 FR_024 Normal   F Normal (F)   lipids 0.18374737
# 4 FR_025 Normal   F Normal (F)   lipids 0.07305875
# 5 FR_027 Normal   M Normal (M)   lipids 0.26198224
# 6 FR_030 Normal   M Normal (M)   lipids 0.10158547

unique( df.zones.melt$group )
# [1] Normal Cancer
# Levels: Cancer < Normal

unique( df.zones.melt$sex ) #  "F" "M"
df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("F","M"), labels = c("Female","Male"), ordered = TRUE)


str(df.zones.melt)
# 'data.frame':	1469 obs. of  6 variables:
# $ sample  : chr  "FR_001" "FR_003" "FR_024" "FR_025" ...
# $ group   : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 1 2 1 1 2 2 2 2 1 1 ...
# $ group_mf: Ord.factor w/ 4 levels "Cancer (F)"<"Normal (F)"<..: 2 4 2 2 4 4 4 4 2 2 ...
# $ variable: Factor w/ 13 levels "lipids","proteins",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0 0.0654 0.1837 0.0731 0.262 ...



dat <- df.zones.melt



cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)

p <- ggplot(data = dat, aes(x = group, y = value))+ # y = value   y = win95abun
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  
  scale_colour_manual(values = cols, name = "Sex")+
  scale_shape_manual(values = shapes, name = "Sex")+
  facet_wrap(facets = vars(variable), scales = "free_y")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")


## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"

unique(dat$group_mf)
# [1] Normal (F) Normal (M) Cancer (F) Cancer (M)
# Levels: Cancer (F) < Normal (F) < Cancer (M) < Normal (M)


y_anotations <- list()
sig_anotations <- list()

# sig converter - Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
sig_symb <- function(x) {
  out <- character(length = length(x))
  for (i in seq_along(x)) {
    if (is.na(x[i])) {
      out[i] <- NA
    } else if (x[i] < 0.001) { 
      out[i] <- "***"
    } else if (x[i] >= 0.001 & x[i] < 0.01) { 
      out[i] <- "**"
    } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
      out[i] <- "*" 
    } else { 
      out[i] <- "ns"
    }
  } # END loop
  return(out)
}


## lipids
this_var <- "Lipids"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.37263, num df = 23, denom df = 32, p-value = 0.01638
var.test(w,z) # F = 2.847, num df = 28, denom df = 26, p-value = 0.008962
# ...ALL unequal variances

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 399.5, p-value = 0.4807

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 533, p-value = 0.01037

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Proteins"
this_var <- "Proteins"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.90372, num df = 23, denom df = 32, p-value = 0.8123
var.test(w,z) # F = 0.59711, num df = 28, denom df = 26, p-value = 0.1841

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -1.7898, df = 55, p-value = 0.03949
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.02065322
# sample estimates:
#   mean of x mean of y 
# 3.052113  3.368638

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y
# # 

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# 
# data:  w and z
# t = -3.3564, df = 54, p-value = 0.0007259
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.2640482
# sample estimates:
#   mean of x mean of y 
# 2.931573  3.458218 

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value   # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.82501, num df = 23, denom df = 32, p-value = 0.6391
var.test(w,z) # F = 0.22491, num df = 28, denom df = 26, p-value = 0.0002022
# ...ALL unequal variances

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 2.5087, df = 55, p-value = 0.007546
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.1472294       Inf
# sample estimates:
#   mean of x mean of y 
# 2.824578  2.382607 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y
# # 

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 581, p-value = 0.0007872

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.99595, num df = 23, denom df = 32, p-value = 0.9914
var.test(w,z) # F = 1.1721, num df = 28, denom df = 26, p-value = 0.6868

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 0.32896, df = 55, p-value = 0.3717

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -0.27519, df = 54, p-value = 0.3921

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z
#

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.58581, num df = 23, denom df = 32, p-value = 0.186
var.test(w,z) # F = 1.729, num df = 28, denom df = 26, p-value = 0.1644
# ...ALL unequal variances

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.0198, df = 55, p-value = 0.1562

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 1.92, df = 54, p-value = 0.03008
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.005359016         Inf
# sample estimates:
#   mean of x  mean of y 
# 0.12442301 0.08266316 

#wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
#wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.50917, num df = 23, denom df = 32, p-value = 0.09628
var.test(w,z) # F = 1.5879, num df = 28, denom df = 26, p-value = 0.2395

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 0.58884, df = 55, p-value = 0.2792

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y


# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 2.8019, df = 54, p-value = 0.003519
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.2439694       Inf
# sample estimates:
#   mean of x mean of y 
# 4.841317  4.235487 

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z


y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Tannins"
this_var <- "Tannins"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value   # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.6421, num df = 23, denom df = 32, p-value = 0.1925
var.test(w,z) # F = 0.45349, num df = 28, denom df = 26, p-value = 0.04306
# ...ALL unequal variances

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -2.3048, df = 55, p-value = 0.01249
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.5552168
# sample estimates:
#   mean of x mean of y 
# 14.59932  16.62474 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 176, p-value = 0.0001423

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)


y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other methylated\ncompounds"
this_var <- "Other methylated\ncompounds"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value   # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.7485, num df = 23, denom df = 32, p-value = 0.1426
var.test(w,z) # F = 0.50903, num df = 28, denom df = 26, p-value = 0.08312
# 

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.3372, df = 55, p-value = 0.09332

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.89216, df = 54, p-value = 0.1881

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other hydrated\ncompounds"
this_var <- "Other hydrated\ncompounds"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.4776, num df = 23, denom df = 32, p-value = 0.3033
var.test(w,z) # F = 0.61095, num df = 28, denom df = 26, p-value = 0.2042
#

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.0894, df = 55, p-value = 0.1404

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -1.171, df = 54, p-value = 0.1234

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.28566, num df = 23, denom df = 32, p-value = 0.002684
var.test(w,z) # F = 0.46082, num df = 28, denom df = 26, p-value = 0.04737

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 455, p-value = 0.1736

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 479, p-value = 0.0774

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.1971, num df = 23, denom df = 32, p-value = 0.6282
var.test(w,z) # F = 1.7915, num df = 28, denom df = 26, p-value = 0.139

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.9452, df = 55, p-value = 0.02844
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.03022286        Inf
# sample estimates:
#   mean of x mean of y 
# 0.7712153 0.5551769 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 2.5591, df = 54, p-value = 0.006664
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.08205092        Inf
# sample estimates:
#   mean of x mean of y 
# 0.7242075 0.4870925

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds" 

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.45384, num df = 23, denom df = 32, p-value = 0.05257
var.test(w,z) # F = 2.9897, num df = 28, denom df = 26, p-value = 0.006337

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 0.19195, df = 55, p-value = 0.4242

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 587, p-value = 0.0006927

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "All compounds"
this_var <- "All compounds"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$value  # win95abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$value

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$value
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$value

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group_mf, y = value))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.2574, num df = 23, denom df = 32, p-value = 0.541
var.test(w,z) # F = 0.4203, num df = 28, denom df = 26, p-value = 0.02692

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -1.0719, df = 55, p-value = 0.1442

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 223, p-value = 0.00262

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



# Add significance annotation
# females
annotation_df.f <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  start=rep(0.8, times=13), end = rep(1.8, times=13),
  y = c( y_anotations[["Lipids"]][1],
         y_anotations[["Proteins"]][1],
         y_anotations[["Amino sugars"]][1],
         y_anotations[["Carbohydrates"]][1],
         y_anotations[["Condensed\naromatics"]][1],
         y_anotations[["Lignin"]][1],
         y_anotations[["Tannins"]][1],
         y_anotations[["Other methylated\ncompounds"]][1],
         y_anotations[["Other hydrated\ncompounds"]][1],
         y_anotations[["Other demethylated\ncompounds"]][1],
         y_anotations[["Other condensed\ncompounds"]][1],
         y_anotations[["Near VitK2-VitB12\ncompounds"]][1],
         y_anotations[["All compounds"]][1]
         
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Lipids"]][1],
            sig_anotations[["Proteins"]][1],
            sig_anotations[["Amino sugars"]][1],
            sig_anotations[["Carbohydrates"]][1],
            sig_anotations[["Condensed\naromatics"]][1],
            sig_anotations[["Lignin"]][1],
            sig_anotations[["Tannins"]][1],
            sig_anotations[["Other methylated\ncompounds"]][1],
            sig_anotations[["Other hydrated\ncompounds"]][1],
            sig_anotations[["Other demethylated\ncompounds"]][1],
            sig_anotations[["Other condensed\ncompounds"]][1],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]][1],
            sig_anotations[["All compounds"]][1]
            
  ), stringsAsFactors = FALSE)
str(annotation_df.f)

# ensure no conflict in factor levels
annotation_df.f$variable <- factor(annotation_df.f$variable, 
                                   levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                              "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                   ordered = TRUE)

# males
annotation_df.m <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  start=rep(1.2, times=13), end = rep(2.2, times=13),
  y = c( y_anotations[["Lipids"]][2],
         y_anotations[["Proteins"]][2],
         y_anotations[["Amino sugars"]][2],
         y_anotations[["Carbohydrates"]][2],
         y_anotations[["Condensed\naromatics"]][2],
         y_anotations[["Lignin"]][2],
         y_anotations[["Tannins"]][2],
         y_anotations[["Other methylated\ncompounds"]][2],
         y_anotations[["Other hydrated\ncompounds"]][2],
         y_anotations[["Other demethylated\ncompounds"]][2],
         y_anotations[["Other condensed\ncompounds"]][2],
         y_anotations[["Near VitK2-VitB12\ncompounds"]][2],
         y_anotations[["All compounds"]][2]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Lipids"]][2],
            sig_anotations[["Proteins"]][2],
            sig_anotations[["Amino sugars"]][2],
            sig_anotations[["Carbohydrates"]][2],
            sig_anotations[["Condensed\naromatics"]][2],
            sig_anotations[["Lignin"]][2],
            sig_anotations[["Tannins"]][2],
            sig_anotations[["Other methylated\ncompounds"]][2],
            sig_anotations[["Other hydrated\ncompounds"]][2],
            sig_anotations[["Other demethylated\ncompounds"]][2],
            sig_anotations[["Other condensed\ncompounds"]][2],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]][2],
            sig_anotations[["All compounds"]][2]
  ), stringsAsFactors = FALSE)
str(annotation_df.m)

# ensure no conflict in factor levels
annotation_df.m$variable <- factor(annotation_df.m$variable, 
                                   levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                              "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                   ordered = TRUE)


annotation_df.f$label # "ns" "*"  "**" "ns" "ns" "ns" "*"  "ns" "ns" "ns" "*"  "ns" "ns"

sel <- which(annotation_df.f$label == "ns")
annotation_df.f$label[sel] <- NA
annotation_df.f$start[sel] <- NA
annotation_df.f$end[sel] <- NA
annotation_df.f$y[sel] <- NA

annotation_df.m$label #  "*"   "***" "***" "ns"  "*"   "**"  "***" "ns"  "ns"  "ns"  "**"  "***" "**" 

sel <- which(annotation_df.m$label == "ns")
annotation_df.m$label[sel] <- NA
annotation_df.m$start[sel] <- NA
annotation_df.m$end[sel] <- NA
annotation_df.m$y[sel] <- NA




p <- ggplot(data = dat, aes(x = group, y = value))+ # y = value   y = win95abun
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  
  # females
  geom_text(data=annotation_df.f, mapping = aes(x = 1.3, y = y, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=annotation_df.f, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#7570b3")+
  # males
  geom_text(data=annotation_df.m, mapping = aes(x = 1.7, y = y, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=annotation_df.m, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#d95f02")+
  
  scale_colour_manual(values = cols, name = NULL)+
  scale_shape_manual(values = shapes, name = NULL)+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function % rel. abun. in vK zones") +  # "Function % rel. abun. in vK zones (Winsorized)"
  
  guides(color = guide_legend(direction = "vertical", vjust = 0.5, hjust = 1) )+
  
  theme(
    #legend.position="bottom",
    legend.position = c(0.95, 0.25), # as fraction of plot dimensions
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.7))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-vkZones-boxplots-",this_study,header,"-v9.tiff"), width = 24, height = 14, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Zeller CRC - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25)
#-------------------------

this_study <- "-Zeller-CRC-"
header <- "-vk-hac-buffers-"

phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")

df.out.distil <- readRDS(file = "df.out.distil_Zeller-CRC.RDS")

df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Zeller-CRC.RDS")

# use clean dataset
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


str(vk.samp.noNAs.long)
# 'data.frame':	783316 obs. of  8 variables:
# $ sample   : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ fxn_id   : Factor w/ 7068 levels "fxn_10","fxn_100",..: 2055 2868 3521 4059 4856 6015 6613 1 512 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.667 0.667 1 0.571 0.8 ...
# $ HC_y     : num  2.33 2.33 1.5 2 1.6 ...
# $ group    : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" ...

# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]
# add sex
spdf$sex <- NA
samps <- unique(spdf$sample)
length(samps) # 113
length(which(phy@sam_data$Subject_ID %in% samps)) # 113
for (i in 1:length(samps)) {
  #i<-1
  this_samp <- samps[i]
  sel.spdf <- which(spdf$sample == this_samp)
  sel.meta <- which(phy@sam_data$Subject_ID == this_samp)
  spdf$sex[sel.spdf] <- phy@sam_data$Gender[sel.meta]
  print(paste0("completed ",i," - no rows = ",length(sel.spdf)))
}

dim(spdf) #  197710      6

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# F      M 
# 97270 100440
sel.na <- which(spdf$sex=="NA") # empty
# unique(spdf$sample[sel.na]) #
# spdf <- spdf[-sel.na, ]

head(spdf)
# sample  group      OC_x     HC_y       abun sex
# 25  FR_001 Normal 0.6722222 1.511111 0.10683761   F
# 83  FR_001 Normal 1.5000000 1.130952 0.32051282   F
# 92  FR_001 Normal 0.5777778 3.033333 0.19880716   F
# 117 FR_001 Normal 0.7333333 2.100000 0.04970179   F
# 124 FR_001 Normal 1.9666667 2.255556 0.04970179   F
# 145 FR_001 Normal 0.6444444 1.566667 0.19880716   F

dim(spdf) #  197710      6


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 113

hist(spdf$abun);summary(spdf$abun)
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# 0.000011 0.002021 0.006873 0.038104 0.023719 5.882966

12.5/(pi*0.05^2) # 1591.549

add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 57.18   65.21   66.85   66.67   68.44   74.85

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"  

## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin
df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)

sum_relabun_in_vk_hac_buffers <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"   
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  df_out <- data.frame(sample=NA, group=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      group = unique(spdf_in$group),
      sex = unique(spdf_in$sex),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
      
    )
    #names(temp_out) # "sample"     "group"      "sex"        "rad"        "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"     
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 113
out[[1]]
# sample  group sex  rad   Acetate Propionate  Butyrate     VitB2   VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2 FR_001 Normal   F 0.05  87.94615  25.312913  0.000000 25.312913 0.000000  95.27113  0.000000 0.0000000 163.88489 25.31291
# 3 FR_001 Normal   F 0.10 131.89894  12.656457  3.164114  7.910285 0.000000  87.77519 10.202240 0.0000000 103.91884 47.11691
# 4 FR_001 Normal   F 0.15  71.20609   7.031365 35.328101  4.921955 9.068658  68.30340  4.534329 0.0000000  88.86113 61.90666
# 5 FR_001 Normal   F 0.20  43.79053  14.900240 22.245142 15.430127 7.474206 105.24026  4.132617 0.7910285  79.92177 64.27310
# 6 FR_001 Normal   F 0.25  42.43158  21.550051 20.539158 20.286028 8.947141 100.54938  5.344919 1.0125165  75.74078 99.12818


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample  group sex  rad   Acetate Propionate  Butyrate     VitB2   VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2  FR_001 Normal   F 0.05  87.94615  25.312913  0.000000 25.312913 0.000000  95.27113  0.000000 0.0000000 163.88489 25.31291
# 3  FR_001 Normal   F 0.10 131.89894  12.656457  3.164114  7.910285 0.000000  87.77519 10.202240 0.0000000 103.91884 47.11691
# 4  FR_001 Normal   F 0.15  71.20609   7.031365 35.328101  4.921955 9.068658  68.30340  4.534329 0.0000000  88.86113 61.90666
# 5  FR_001 Normal   F 0.20  43.79053  14.900240 22.245142 15.430127 7.474206 105.24026  4.132617 0.7910285  79.92177 64.27310
# 6  FR_001 Normal   F 0.25  42.43158  21.550051 20.539158 20.286028 8.947141 100.54938  5.344919 1.0125165  75.74078 99.12818
# 21 FR_003 Normal   M 0.05 138.85357  18.642400  0.000000 11.617501 1.556594 129.83956  1.505367 0.0000000 380.49396 18.36904

df.buffers$group_mf <- paste0(df.buffers$group,"__",df.buffers$sex)
unique(df.buffers$group_mf)
# "Normal__F" "Normal__M" "Cancer__F" "Cancer__M"
df.buffers$group_mf <- factor(df.buffers$group_mf, levels = c("Cancer__F","Normal__F","Cancer__M","Normal__M"), 
                              labels = c("Cancer (F)","Normal (F)","Cancer (M)","Normal (M)"), 
                              ordered = TRUE)
table(df.buffers$group_mf[ which(df.buffers$rad==0.05) ]) # count at single radius only
# Cancer (F) Normal (F) Cancer (M) Normal (M) 
# 24         33         29         27 


saveRDS(object = df.buffers, file = "df.buffers--Zeller-CRC.RDS")

str(df.buffers)
# 'data.frame':	565 obs. of  15 variables:
# $ sample    : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ group     : chr  "Normal" "Normal" "Normal" "Normal" ...
# $ sex       : chr  "F" "F" "F" "F" ...
# $ rad       : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate   : num  87.9 131.9 71.2 43.8 42.4 ...
# $ Propionate: num  25.31 12.66 7.03 14.9 21.55 ...
# $ Butyrate  : num  0 3.16 35.33 22.25 20.54 ...
# $ VitB2     : num  25.31 7.91 4.92 15.43 20.29 ...
# $ VitB12    : num  0 0 9.07 7.47 8.95 ...
# $ VitB6     : num  95.3 87.8 68.3 105.2 100.5 ...
# $ VitB9     : num  0 10.2 4.53 4.13 5.34 ...
# $ VitK2     : num  0 0 0 0.791 1.013 ...
# $ Glutamate : num  163.9 103.9 88.9 79.9 75.7 ...
# $ Pyruvate  : num  25.3 47.1 61.9 64.3 99.1 ...
# $ group_mf  : Ord.factor w/ 4 levels "Cancer (F)"<"Normal (F)"<..: 2 2 2 2 2 4 4 4 4 4 ...



df.buffers.melt <- melt(df.buffers, id.vars = c("sample","group","sex","group_mf","rad"))
head(df.buffers.melt)
# sample  group sex   group_mf  rad variable     value
# 1 FR_001 Normal   F Normal (F) 0.05  Acetate  87.94615
# 2 FR_001 Normal   F Normal (F) 0.10  Acetate 131.89894
# 3 FR_001 Normal   F Normal (F) 0.15  Acetate  71.20609
# 4 FR_001 Normal   F Normal (F) 0.20  Acetate  43.79053
# 5 FR_001 Normal   F Normal (F) 0.25  Acetate  42.43158
# 6 FR_003 Normal   M Normal (M) 0.05  Acetate 138.85357

unique( df.buffers.melt$group ) # "Normal" "Cancer"  
df.buffers.melt$group <- factor(df.buffers.melt$group, levels = c("Cancer","Normal"), ordered = TRUE)

unique( df.buffers.melt$sex ) #  "F"   "M"
df.buffers.melt$sex <- factor(df.buffers.melt$sex, levels = c("F","M"), labels = c("Female","Male"), ordered = TRUE)

unique( df.buffers.melt$group_mf )
# [1] Normal (F) Normal (M) Cancer (F) Cancer (M)
# Levels: Cancer (F) < Normal (F) < Cancer (M) < Normal (M)

unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	5650 obs. of  7 variables:
#   $ sample  : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ group   : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 1 1 1 1 1 2 2 2 2 2 ...
# $ group_mf: Ord.factor w/ 4 levels "Cancer (F)"<"Normal (F)"<..: 2 2 2 2 2 4 4 4 4 4 ...
# $ rad     : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ variable: Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  87.9 131.9 71.2 43.8 42.4 ...

hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))
#hist(sqrt( df.buffers.melt$value))

dat <- df.buffers.melt



dat$log10abun <- dat$value # then over-write later


vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}

# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   77.89  104.26  129.65  154.32  561.79 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7609  1.8915  2.0181  2.0433  2.1884  2.7496 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   8.026  15.288  16.718  22.194  58.260 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.1972  0.9045  1.1844  1.1178  1.3462  1.7654 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.3706  8.3904  7.6656 12.3307 43.3428 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.1670 -0.4311  0.9238  0.3825  1.0910  1.6369 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   6.550   9.638  13.125  13.257 375.786 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.4909  0.8162  0.9840  0.9542  1.1224  2.5749 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   1.994   4.116   5.691   9.138  30.254 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.4552  0.2997  0.6145  0.4740  0.9609  1.4808 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   62.85   76.70   82.97   97.50  231.98 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.164   1.798   1.885   1.895   1.989   2.365 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.8286  2.6539  2.9585  4.5577 20.0465 
# [1] "log10 values:"
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# -1.37084 -0.08165  0.42389  0.13722  0.65875  1.30204 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.0000  0.3465  0.9783  1.3895 13.8314 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.9986 -1.9986 -0.4603 -0.7174  0.1429  1.1409 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.56  102.03  124.92  160.40  176.82  616.68 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.745   2.009   2.097   2.149   2.248   2.790 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   48.33   60.51   59.96   73.42  175.35 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.1289  1.6843  1.7818  1.7392  1.8658  2.2439



#cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
#shapes <- c("Female" = 2, "Male" = 4)

head(dat)
# sample  group    sex   group_mf  rad variable     value log10abun
# 1 FR_001 Normal Female Normal (F) 0.05  Acetate  87.94615  1.944217
# 2 FR_001 Normal Female Normal (F) 0.10  Acetate 131.89894  2.120241
# 3 FR_001 Normal Female Normal (F) 0.15  Acetate  71.20609  1.852517
# 4 FR_001 Normal Female Normal (F) 0.20  Acetate  43.79053  1.641380
# 5 FR_001 Normal Female Normal (F) 0.25  Acetate  42.43158  1.627689
# 6 FR_003 Normal   Male Normal (M) 0.05  Acetate 138.85357  2.142557


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group_mf, color = group_mf))+
  
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    legend.position="bottom",
    legend.title = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


saveRDS(object = dat, file = paste0("dat--All-F-M-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--All-F-M-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))



## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]

cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)

head(dat)
# sample  group    sex   group_mf  rad variable     value log10abun
# 1  FR_001 Normal Female Normal (F) 0.05  Acetate  87.94615  1.944217
# 6  FR_003 Normal   Male Normal (M) 0.05  Acetate 138.85357  2.142557
# 11 FR_024 Normal Female Normal (F) 0.05  Acetate  45.34699  1.656548
# 16 FR_025 Normal Female Normal (F) 0.05  Acetate 100.92697  2.004007
# 21 FR_027 Normal   Male Normal (M) 0.05  Acetate 170.16576  2.230872
# 26 FR_030 Normal   Male Normal (M) 0.05  Acetate  94.08631  1.973526

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  
  labs(x = NULL, y = "log10 fxn % rel. abun. in 0.05 radius buffer (per vK unit area)") +
  scale_colour_manual(values = cols, name = "Sex")+
  scale_shape_manual(values = shapes, name = "Sex")+
  facet_wrap(facets = vars(variable), scales = "free")
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-HAC-vkBuffers-rough-boxplots",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")




## perform sig tests for annotations !!
unique( dat$variable )
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

unique(dat$group_mf)
# [1] Normal (F) Normal (M) Cancer (F) Cancer (M)
# Levels: Cancer (F) < Normal (F) < Cancer (M) < Normal (M)

names(dat) # "sample"    "group"     "sex"       "group_mf"  "rad"       "variable"  "value"     "log10abun"


y_anotations <- list()
sig_anotations <- list()

# sig converter - Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
sig_symb <- function(x) {
  out <- character(length = length(x))
  for (i in seq_along(x)) {
    if (is.na(x[i])) {
      out[i] <- NA
    } else if (x[i] < 0.001) { 
      out[i] <- "***"
    } else if (x[i] >= 0.001 & x[i] < 0.01) { 
      out[i] <- "**"
    } else if (x[i] >= 0.01 & x[i] <= 0.05) { 
      out[i] <- "*" 
    } else { 
      out[i] <- "ns"
    }
  } # END loop
  return(out)
}



## Acetate
this_var <- "Acetate"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun # USE LOG10abun!!!!!!!!!!
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.60585, num df = 23, denom df = 32, p-value = 0.215
var.test(w,z) # F = 0.17027, num df = 28, denom df = 26, p-value = 1.376e-05

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.33173, df = 55, p-value = 0.3707

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 403, p-value = 0.429

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m)
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Propionate"
this_var <- "Propionate"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.69207, num df = 23, denom df = 32, p-value = 0.3626
var.test(w,z) # F = 0.92021, num df = 28, denom df = 26, p-value = 0.8268

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -1.0079, df = 55, p-value = 0.159

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -0.6181, df = 54, p-value = 0.2696

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Butyrate"
this_var <- "Butyrate"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = NaN, num df = 23, denom df = 32, p-value = NA
var.test(w,z) # F = NaN, num df = 28, denom df = 26, p-value = NA

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 396, p-value = NA

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 391.5, p-value = NA

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB2"
this_var <- "VitB2"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.4918, num df = 23, denom df = 32, p-value = 0.2918
var.test(w,z) # F = 0.86479, num df = 28, denom df = 26, p-value = 0.705

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 0.53546, df = 55, p-value = 0.2972

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 2.1453, df = 54, p-value = 0.01822
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.07031468        Inf
# sample estimates:
#   mean of x mean of y 
# 1.397929  1.078134 

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB12"
this_var <- "VitB12"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.80334, num df = 23, denom df = 32, p-value = 0.5918
var.test(w,z) # F = 0.6418, num df = 28, denom df = 26, p-value = 0.2526

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 2.7282, df = 55, p-value = 0.004266
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.2129343       Inf
# sample estimates:
#   mean of x  mean of y 
# 0.2012741 -0.3492732 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 3.2034, df = 54, p-value = 0.00114
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.3216661       Inf
# sample estimates:
#   mean of x  mean of y 
# -0.1023845 -0.7759459

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB6"
this_var <- "VitB6"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.36442, num df = 23, denom df = 32, p-value = 0.01422
var.test(w,z) # F = 0.085731, num df = 28, denom df = 26, p-value = 6.858e-09

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# Two Sample t-test
# data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y
# W = 448, p-value = 0.2041

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# Two Sample t-test
# data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 395, p-value = 0.4805

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitB9"
this_var <- "VitB9"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 1.9325, num df = 23, denom df = 32, p-value = 0.08452
var.test(w,z) # F = 1.5025, num df = 28, denom df = 26, p-value = 0.3

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 2.1972, df = 55, p-value = 0.01612
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.09963927        Inf
# sample estimates:
#   mean of x  mean of y 
# -0.6198929 -1.0375807 

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 2.9123, df = 54, p-value = 0.002603
# alternative hypothesis: true difference in means is greater than 0
# 95 percent confidence interval:
#   0.2154443       Inf
# sample estimates:
#   mean of x  mean of y 
# -0.5380622 -1.0445674

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "VitK2"
this_var <- "VitK2"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = NaN, num df = 23, denom df = 32, p-value = NA
var.test(w,z) # F = 5.6783, num df = 28, denom df = 26, p-value = 2.847e-05

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 396, p-value = NA

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "two.sided") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 418.5, p-value = 0.4183

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Glutamate"
this_var <- "Glutamate"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.89847, num df = 23, denom df = 32, p-value = 0.8008
var.test(w,z) # F = 1.3872, num df = 28, denom df = 26, p-value = 0.4045

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.0106, df = 55, p-value = 0.1583

# wt.f <- wilcox.test(x, y, alternative = "two.sided") # "less" "two.sided"  "greater"
# wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# t = 0.89098, df = 54, p-value = 0.1884

# wt.m <- wilcox.test(w, z, alternative = "two.sided") # "less" "two.sided"  "greater"
# wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Pyruvate"
this_var <- "Pyruvate"

x=filter(dat, group_mf == "Cancer (F)" & variable == this_var)$log10abun
y=filter(dat, group_mf == "Normal (F)" & variable == this_var)$log10abun

w=filter(dat, group_mf == "Cancer (M)" & variable == this_var)$log10abun
z=filter(dat, group_mf == "Normal (M)" & variable == this_var)$log10abun

p <- ggplot(data = filter(dat, variable == this_var), aes(x = group, y = log10abun))+
  geom_boxplot()+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)
p

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data

# test for homogeneity of variances
var.test(x,y) # F = 0.24309, num df = 23, denom df = 32, p-value = 0.0007939
var.test(w,z) # F = 0.24243, num df = 28, denom df = 26, p-value = 0.0003975

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y
wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 490, p-value = 0.06536

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "two.sided") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 519, p-value = 0.01829

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



# Add significance annotation
# females
annotation_df.f <- data.frame(
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
  start=rep(0.8, times=10), end = rep(1.8, times=10),
  y = c( y_anotations[["Acetate"]][1],
         y_anotations[["Propionate"]][1],
         y_anotations[["Butyrate"]][1],
         y_anotations[["VitB2"]][1],
         y_anotations[["VitB12"]][1],
         y_anotations[["VitB6"]][1],
         y_anotations[["VitB9"]][1],
         y_anotations[["VitK2"]][1],
         y_anotations[["Glutamate"]][1],
         y_anotations[["Pyruvate"]][1]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Acetate"]][1],
            sig_anotations[["Propionate"]][1],
            sig_anotations[["Butyrate"]][1],
            sig_anotations[["VitB2"]][1],
            sig_anotations[["VitB12"]][1],
            sig_anotations[["VitB6"]][1],
            sig_anotations[["VitB9"]][1],
            sig_anotations[["VitK2"]][1],
            sig_anotations[["Glutamate"]][1],
            sig_anotations[["Pyruvate"]][1]
  ),
  stringsAsFactors = FALSE)
str(annotation_df.f)

# ensure no conflict in factor levels
annotation_df.f$variable <- factor(annotation_df.f$variable, 
                                   levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
                                   ordered = TRUE)

# males
annotation_df.m <- data.frame(
  variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
  start=rep(1.2, times=10), end = rep(2.2, times=10),
  y = c( y_anotations[["Acetate"]][2],
         y_anotations[["Propionate"]][2],
         y_anotations[["Butyrate"]][2],
         y_anotations[["VitB2"]][2],
         y_anotations[["VitB12"]][2],
         y_anotations[["VitB6"]][2],
         y_anotations[["VitB9"]][2],
         y_anotations[["VitK2"]][2],
         y_anotations[["Glutamate"]][2],
         y_anotations[["Pyruvate"]][2]
  ),
  # label order: Female, Male, All
  label = c(sig_anotations[["Acetate"]][2],
            sig_anotations[["Propionate"]][2],
            sig_anotations[["Butyrate"]][2],
            sig_anotations[["VitB2"]][2],
            sig_anotations[["VitB12"]][2],
            sig_anotations[["VitB6"]][2],
            sig_anotations[["VitB9"]][2],
            sig_anotations[["VitK2"]][2],
            sig_anotations[["Glutamate"]][2],
            sig_anotations[["Pyruvate"]][2]
  ),
  stringsAsFactors = FALSE)
str(annotation_df.m)

# ensure no conflict in factor levels
annotation_df.m$variable <- factor(annotation_df.m$variable, 
                                   levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2", "Glutamate", "Pyruvate"),
                                   ordered = TRUE)

sel <- which(annotation_df.f$label == "ns" | is.na(annotation_df.f$label))
annotation_df.f$label[sel] <- NA
annotation_df.f$start[sel] <- NA
annotation_df.f$end[sel] <- NA
annotation_df.f$y[sel] <- NA

sel <- which(annotation_df.m$label == "ns" | is.na(annotation_df.m$label))
annotation_df.m$label[sel] <- NA
annotation_df.m$start[sel] <- NA
annotation_df.m$end[sel] <- NA
annotation_df.m$y[sel] <- NA


p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_violin(width = 1.1)+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  
  # females
  geom_text(data=annotation_df.f, mapping = aes(x = 1.3, y = y, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=annotation_df.f, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#7570b3")+
  # males
  geom_text(data=annotation_df.m, mapping = aes(x = 1.7, y = y, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=annotation_df.m, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#d95f02")+
  
  scale_colour_manual(values = cols, name = NULL)+
  scale_shape_manual(values = shapes, name = NULL)+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function log10 % rel. abun. in 0.05 radius vK buffer zones") +
  theme(
    legend.position="bottom",
    #strip.background = element_rect(fill="white", linetype = "blank"),
    strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.8))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-vkBuffers-log1oabun-boxplots-",this_study,header,"-v8.tiff"), width = 22, height = 13, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Zeller CRC - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

# Determines weighted coordinates in each sample, then later assess stats by group_mf?

this_study <- "-Zeller-CRC-"
phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Zeller-CRC.RDS")

df.tax <- as.data.frame(phy@tax_table)

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Zeller-CRC.RDS")

# use clean dataset
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

header <- "-vk-coords-wtmean-"


str(vk.samp.noNAs.long)
# 'data.frame':	783316 obs. of  8 variables:
# $ sample   : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ fxn_id   : Factor w/ 7068 levels "fxn_10","fxn_100",..: 2055 2868 3521 4059 4856 6015 6613 1 512 1045 ...
# $ abun     : num  0 0 0 0 0 0 0 0 0 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.667 0.667 1 0.571 0.8 ...
# $ HC_y     : num  2.33 2.33 1.5 2 1.6 ...
# $ group    : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ samplabel: chr  "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" "FR_001 (Normal)" ...

# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group", "OC_x", "HC_y", "abun") ]
# add sex
spdf$sex <- NA
samps <- unique(spdf$sample)
length(samps) # 113
length(which(phy@sam_data$Subject_ID %in% samps)) # 113
for (i in 1:length(samps)) {
  #i<-1
  this_samp <- samps[i]
  sel.spdf <- which(spdf$sample == this_samp)
  sel.meta <- which(phy@sam_data$Subject_ID == this_samp)
  spdf$sex[sel.spdf] <- phy@sam_data$Gender[sel.meta]
  print(paste0("completed ",i," - no rows = ",length(sel.spdf)))
}

dim(spdf) #  197710      6

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# F      M 
# 97270 100440
sel.na <- which(spdf$sex=="NA") # empty
# unique(spdf$sample[sel.na]) #
# spdf <- spdf[-sel.na, ]

head(spdf)
#     sample  group      OC_x     HC_y       abun sex
# 25  FR_001 Normal 0.6722222 1.511111 0.10683761   F
# 83  FR_001 Normal 1.5000000 1.130952 0.32051282   F
# 92  FR_001 Normal 0.5777778 3.033333 0.19880716   F
# 117 FR_001 Normal 0.7333333 2.100000 0.04970179   F
# 124 FR_001 Normal 1.9666667 2.255556 0.04970179   F
# 145 FR_001 Normal 0.6444444 1.566667 0.19880716   F

dim(spdf) #  197710      6


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 113

hist(spdf$abun);summary(spdf$abun)
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# 0.000011 0.002021 0.006873 0.038104 0.023719 5.882966

12.5/(pi*0.05^2) # 1591.549

add <- NA
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 57.18   65.21   66.85   66.67   68.44   74.85

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"



get_wtmean_vk_coords <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  subsel <- list()
  
  # lipids
  subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  # proteins
  subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # amino_sugars
  subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # carbs - Carbohydrates
  subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  # cond_aromtx - Condensed aromatics
  subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  # lignin
  subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  # tannins
  subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  
  vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")
  
  # establish blank template data frame
  df_out <- data.frame(sample = NA,group = NA,sex = NA, no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)
  
  for (v in 1:length(vars)) {
    #v<-1
    this_var <- vars[v]
    if (length(subsel[[ this_var ]]) > 0) {
      df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
    } else {
      df_out$wtmean_OC_x[v] <- NA
      df_out$wtmean_HC_y[v] <- NA
    }
    print(paste0("completed ",v))
  }
  
  df_out$sample <- this_samp
  df_out$group <- unique(spdf_in$group)
  df_out$sex <- unique(spdf_in$sex)
  
  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 113
out[[1]]
#   sample  group sex no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1 FR_001 Normal   F      NA       lipids          NA          NA
# 2 FR_001 Normal   F      20     proteins   0.3988602    2.009520
# 3 FR_001 Normal   F      21 amino_sugars   0.6373043    1.709973
# 4 FR_001 Normal   F     108        carbs   0.8922460    1.739806
# 5 FR_001 Normal   F      NA  cond_aromtx          NA          NA
# 6 FR_001 Normal   F      52       lignin   0.5282177    1.236953
# 7 FR_001 Normal   F     101      tannins   0.7993664    1.385094

names(out[[1]])
# "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 791    7

ok <- complete.cases(df.wtcoords)
sel.ok <- which(ok==TRUE) # qty 769

df.wtcoords <- df.wtcoords[sel.ok, ]

dim(df.wtcoords) #  769    7


str(df.wtcoords)
# 'data.frame':	769 obs. of  7 variables:
# $ sample     : chr  "FR_001" "FR_001" "FR_001" "FR_001" ...
# $ group      : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex        : chr  "F" "F" "F" "F" ...
# $ no_fxns    : int  20 21 108 52 101 18 75 55 203 6 ...
# $ variable   : chr  "proteins" "amino_sugars" "carbs" "lignin" ...
# $ wtmean_OC_x: num  0.399 0.637 0.892 0.528 0.799 ...
# $ wtmean_HC_y: num  2.01 1.71 1.74 1.24 1.39 ...

saveRDS(object = df.wtcoords, file = "df.wtcoords-Zeller-CRC.RDS")


dat <- df.wtcoords

unique(dat$variable)
# "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "lipids" 

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)

dat$group_mf <- paste0(dat$group,"__",dat$sex)
unique(dat$group_mf) # "Normal__F" "Normal__M" "Cancer__F" "Cancer__M"
dat$group_mf <- factor(dat$group_mf, levels = c("Cancer__F","Normal__F","Cancer__M","Normal__M"),
                       labels = c("Cancer (F)","Normal (F)","Cancer (M)","Normal (M)"),
                       ordered = TRUE)

names(dat) # "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"

dat.F <- filter(dat, sex == "F")
dat.M <- filter(dat, sex == "M")

cols.F <- c("Cancer (F)" = "#f46d43", "Normal (F)" = "#66c2a5")
cols.M <- c("Cancer (M)" = "#e08214", "Normal (M)" = "#8073ac")

#library(ggforce)

p <- ggplot(data = dat.F, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group_mf) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  geom_density_2d(contour_var = "ndensity", bins = 5, alpha=.5) + #
  geom_point(alpha=0.5)+
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols.F) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Diagnosis"))

p

grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat.F, file = paste0("dat.F--Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9e.RDS"))

#dat.F <- readRDS(file = paste0("dat.F--Weighted-vkZone-coords-FEMALE-",this_study,header,"-v9.RDS"))



cols.M <- c("Cancer (M)" = "#e08214", "Normal (M)" = "#8073ac")

p <- ggplot(data = dat.M, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group_mf) )+
  
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  geom_density_2d(contour_var = "ndensity", bins = 5, alpha=.5) + #
  geom_point(alpha=0.5)+
  
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols.M) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "Diagnosis"))

p

grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-MALE-",this_study,header,"-v9e.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat.M, file = paste0("dat.M--Weighted-vkZone-coords-MALE-",this_study,header,"-v9e.RDS"))

#dat.M <- readRDS(file = paste0("dat.M--Weighted-vkZone-coords-MALE-",this_study,header,"-v9e.RDS"))





length(unique(dat$sample)) # 113


# Females

dat.F



## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.F) # "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"


levels(dat.F$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"




## "Lipids"
this_var <- "Lipids"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)  
# samps.class  1  0.07641 0.07493 3.9687  0.068 .
# Residual    49  0.94345 0.92507                
# Total       50  1.01986 1.00000                
# ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df  Sum Sq   Mean Sq    F N.Perm Pr(>F)
# Groups     1 0.00033 0.0003335 0.05    999  0.818
# Residuals 49 0.32711 0.0066757 



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.000716 0.00317 0.1751  0.717
# Residual    55 0.224856 0.99683              
# Total       56 0.225572 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df  Sum Sq   Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.01019 0.0101899 5.2275    999  0.018 *
#   Residuals 55 0.10721 0.0019493                       
# ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0008462 0.02682 1.5156  0.217
# Residual    55 0.0307065 0.97318              
# Total       56 0.0315527 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq     F N.Perm Pr(>F)
# Groups     1 0.0000731 7.3137e-05 0.366    999  0.563
# Residuals 55 0.0109920 1.9986e-04



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.001120 0.02325 1.3094  0.252
# Residual    55 0.047051 0.97675              
# Total       56 0.048171 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0004471 0.00044707 1.5881    999  0.228
# Residuals 55 0.0154837 0.00028152



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)  
# samps.class  1 0.013169 0.06212 3.3777   0.05 *
# Residual    51 0.198833 0.93788                
# Total       52 0.212002 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df   Sum Sq   Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.007146 0.0071463 6.7428    999  0.011 *
# Residuals 51 0.054052 0.0010598                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.001031 0.02733 1.5452  0.212
# Residual    55 0.036703 0.97267              
# Total       56 0.037734 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0000013 1.3220e-06 0.0045    999   0.94
# Residuals 55 0.0162034 2.9461e-04 



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0007793 0.03315 1.8859  0.156
# Residual    55 0.0227279 0.96685              
# Total       56 0.0235072 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0002134 0.00021342 1.6021    999  0.224
# Residuals 55 0.0073266 0.00013321





# # # # # # # # # # # # #


# Males


## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.M) 
# "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"

levels(dat.M$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)
# samps.class  1  0.02752 0.03066 1.5496  0.216
# Residual    49  0.87021 0.96934              
# Total       50  0.89773 1.00000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df  Sum Sq   Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.02215 0.0221477 3.1035    999  0.093 .
# Residuals 49 0.34968 0.0071364                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.000895 0.00558 0.3031  0.609
# Residual    54 0.159443 0.99442              
# Total       55 0.160338 1.00000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df   Sum Sq   Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.006122 0.0061217 4.3356    999   0.04 *
# Residuals 54 0.076247 0.0014120                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)  
# samps.class  1 0.003174 0.05073 2.8858  0.075 .
# Residual    54 0.059389 0.94927                
# Total       55 0.062563 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df   Sum Sq    Mean Sq     F N.Perm Pr(>F)  
# Groups     1 0.002046 0.00204562 3.154    999  0.066 .
# Residuals 54 0.035024 0.00064858                      
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0003641 0.01705 0.9365  0.377
# Residual    54 0.0209928 0.98295              
# Total       55 0.0213569 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0001624 0.00016243 1.1058    999  0.286
# Residuals 54 0.0079318 0.00014689



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.006683 0.04109 2.0141  0.165
# Residual    47 0.155956 0.95891              
# Total       48 0.162639 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df   Sum Sq   Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.008501 0.0085005 5.1275    999  0.018 *
# Residuals 47 0.077919 0.0016578                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)   
# samps.class  1 0.004741 0.10052 6.0347  0.005 **
# Residual    54 0.042427 0.89948                 
# Total       55 0.047169 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df   Sum Sq  Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.000119 0.000119 0.3191    999  0.572
# Residuals 54 0.020142 0.000373 



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df SumOfSqs      R2      F Pr(>F)   
# samps.class  1 0.002955 0.11106 6.7462  0.008 **
# Residual    54 0.023654 0.88894                 
# Total       55 0.026609 1.00000                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0002987 0.00029872 1.9267    999  0.185
# Residuals 54 0.0083726 0.00015505 


#-------------------------

#### Zeller CRC - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-Zeller-CRC.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 12896 taxa and 113 samples ]
# sample_data() Sample Data:       [ 113 samples by 13 sample variables ]
# tax_table()   Taxonomy Table:    [ 12896 taxa by 4 taxonomic ranks ]

df.zones <- readRDS("df.zones--Zeller-CRC.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Zeller-CRC.RDS")

samps <- unique(df.zones$sample) # qty 113
# phy <- prune_samples( samples = samps, phy )
# # prune taxa with zero reads
# phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
# phy

head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 113
NonZeroFxns

mean(NonZeroFxns) # 2808.885
sd(NonZeroFxns) # 1923.561
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 6932
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 1954
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 1749.646
sd(fxns_converted) # 1116.908

mean(fxns_relabun_converted) # 66.66845
sd(fxns_relabun_converted) # 2.729966

mean(vks) # 711.3009
sd(vks) # 361.7311


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Forslund et al 2015 - T2D -/+ Metformin disease vs normal gut metagenomes
# # # # # # # # #
# # # # # # # # #

#### Forslund T2D - prepare for data download from SRA
#-------------------------

# Type 2 diabetes? Forslund et al 2015
# 
# Data source: Forslund et al., Disentangling type 2 diabetes and metformin treatment signatures in the human gut microbiota. 
# Nature 528, 262-266 (2015). DOI: 10.1038/nature15766

## Using population:
# SWE: 93 ND Ctrl, 33 Met-, 20 Met+
# European Nucleotide Archive (accession numbers: ERP002469, SWE samples




# List of samples and diagnosis (Forslund et al SItableS1)
# "Samples-and-diagnosis-SItableS1.xlsx"
# "A1:C785" $`Country subset` == "SWE"
subjects <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Forslund-2015-Type2-diabetes/Samples-and-diagnosis-SItableS1.xlsx",
                       sheet=1, range="A1:C785", col_names = TRUE)
subjects <- as.data.frame(subjects)
str(subjects)

sel <- which(subjects$`Country subset` == "SWE")
subjects <- subjects[sel, ]
str(subjects)
# 'data.frame':	145 obs. of  3 variables:
# $ Sample        : chr  "NG-5425_137" "NG-5425_249" "NG-5425_342" "NG-5425_485" ...
# $ Country subset: chr  "SWE" "SWE" "SWE" "SWE" ...
# $ Status        : chr  "ND CTRL" "ND CTRL" "ND CTRL" "ND CTRL" ...

length(unique( subjects$Sample )) # 145
table(subjects$Status) # , useNA = TRUE
# ND CTRL T2D metformin- T2D metformin+ 
#   92             33             20 

92 + 33 + 20 # 145


run_details <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Forslund-2015-Type2-diabetes/Samples-and-diagnosis-SItableS1.xlsx",
                          sheet="Sequencing statistics SWE sampl", range="A2:K149", col_names = TRUE)
run_details <- as.data.frame(run_details)
str(run_details)
# 'data.frame':	147 obs. of  11 variables:
# $ Sample           : chr  "Sum" "Average" "NG-5425_137" "NG-5425_249" ...
# $ Raw bases        : num  4.53e+11 3.13e+09 7.98e+08 9.67e+08 1.06e+09 ...
# $ Raw reads        : num  4.49e+09 3.10e+07 8.31e+06 1.01e+07 1.11e+07 ...
# $ Trimmed bases    : num  2.78e+11 1.92e+09 3.87e+08 4.74e+08 5.31e+08 ...
# $ Trimmed reads    : num  3.31e+09 2.28e+07 5.04e+06 6.14e+06 6.78e+06 ...
# $ Trimmed inserts  : num  1.89e+09 1.30e+07 3.03e+06 3.68e+06 4.06e+06 ...
# $ Unassembled bases: num  2.78e+11 1.92e+09 3.87e+08 4.74e+08 5.31e+08 ...
# $ Unassembled reads: num  3.31e+09 2.28e+07 5.04e+06 6.14e+06 6.78e+06 ...
# $ Inserts          : num  1.89e+09 1.30e+07 3.03e+06 3.68e+06 4.06e+06 ...
# $ Max read length  : chr  "-" "97.855172413793099" "94" "92" ...
# $ Avg read length  : chr  "-" "83.303448275862067" "77" "77" ...

head(run_details)
# Sample    Raw bases  Raw reads Trimmed bases Trimmed reads Trimmed inserts Unassembled bases
# 1         Sum 453465327668 4494103168  277759781846    3312028283      1889248942      277759781846
# 2     Average   3127347087   30993815    1915584702      22841574        13029303        1915584702
# 3 NG-5425_137    797799936    8310416     387280402       5036087         3027036         387280402
# 4 NG-5425_249    966913152   10072012     474310006       6141684         3679946         474310006
# 5 NG-5425_286   1061488704   11057174     530656068       6778413         4062863         530656068
# 6 NG-5425_297    761805504    7935474     379222624       4895014         2914433         379222624
# Unassembled reads    Inserts    Max read length    Avg read length
# 1        3312028283 1889248942                  -                  -
#   2          22841574   13029303 97.855172413793099 83.303448275862067
# 3           5036087    3027036                 94                 77
# 4           6141684    3679946                 92                 77
# 5           6778413    4062863                 96                 78
# 6           4895014    2914433                 93                 77

# delete first 2 rows:
run_details <- run_details[-c(1,2), ]
head(run_details)
#        Sample  Raw bases Raw reads Trimmed bases Trimmed reads Trimmed inserts Unassembled bases Unassembled reads Inserts Max read length Avg read length
# 3 NG-5425_137  797799936   8310416     387280402       5036087         3027036         387280402           5036087 3027036              94              77
# 4 NG-5425_249  966913152  10072012     474310006       6141684         3679946         474310006           6141684 3679946              92              77
# 5 NG-5425_286 1061488704  11057174     530656068       6778413         4062863         530656068           6778413 4062863              96              78
# 6 NG-5425_297  761805504   7935474     379222624       4895014         2914433         379222624           4895014 2914433              93              77
# 7 NG-5425_342  789010752   8218862     378848377       4968844         2988398         378848377           4968844 2988398              92              76
# 8 NG-5425_470  820824384   8550254     402254631       5219247         3151422         402254631           5219247 3151422              93              77

length(unique(run_details$`Raw bases`)) # 145



# SRA Runs Metadata: 
# "Metadata-Forslund-ERP002469-SWE-samples--SraRunTable"  

meta <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Forslund-2015-Type2-diabetes/Metadata-Forslund-ERP002469-SWE-samples--SraRunTable.xlsx",
                   sheet=1, range="A1:AI146", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)
# 'data.frame':	145 obs. of  35 variables:

names(meta)
# [1] "Run"                    "Assay Type"             "AvgSpotLen"             "Bases"                  "BioProject"             "BioSample"             
# [7] "Bytes"                  "Center Name"            "Consent"                "DATASTORE filetype"     "DATASTORE provider"     "DATASTORE region"      
# [13] "ENA-FIRST-PUBLIC (run)" "ENA-FIRST-PUBLIC"       "ENA-LAST-UPDATE (run)"  "ENA-LAST-UPDATE"        "Experiment"             "External_Id"           
# [19] "INSDC_center_alias"     "INSDC_center_name"      "INSDC_first_public"     "INSDC_last_update"      "INSDC_status"           "Instrument"            
# [25] "Library Name"           "LibraryLayout"          "LibrarySelection"       "LibrarySource"          "Organism"               "Platform"              
# [31] "ReleaseDate"            "Sample Name"            "Sample_Name"            "SRA Study"              "Submitter_Id" 

length(unique(meta$Bases)) # 145

length(meta$Bases %in% run_details$`Raw bases`) # 145 
# i.e. meta$Run - has meta$Bases 
# >> match to run_details$`Raw bases` and run_details$Sample
# link to subjects$Sample

dim(subjects) # 145   5
dim(meta)     # 145  35
dim(run_details) # 145  11


subjects$Bases <- NA
subjects$Run <- NA

for (i in 1:length(subjects$Sample)) {
  #i<-1
  this_samp <- subjects$Sample[i]
  sel1 <- which(run_details$Sample == this_samp)
  subjects$Bases[i] <- run_details$`Raw bases`[sel1]
  
  sel2 <- which(meta$Bases == run_details$`Raw bases`[sel1])
  
  subjects$Run[i] <- meta$Run[sel2]
  
  print(paste0("completed ",i))
}

length(meta$Run) # 145

head(subjects)
# Sample Country subset  Status      Bases       Run
# 640 NG-5425_137            SWE ND CTRL  797799936 ERR260268
# 641 NG-5425_249            SWE ND CTRL  966913152 ERR260269
# 642 NG-5425_342            SWE ND CTRL  789010752 ERR260272
# 643 NG-5425_485            SWE ND CTRL 1090824000 ERR260275
# 644 NG-5636_112            SWE ND CTRL 1954234860 ERR260249
# 645 NG-5636_118            SWE ND CTRL 3307295096 ERR260250
tail(subjects)
#         Sample Country subset         Status      Bases       Run
# 779 NG-5636_530            SWE T2D metformin+ 3860516334 ERR260192
# 780 NG-5636_536            SWE T2D metformin+ 4292398798 ERR260194
# 781 NG-5636_543            SWE T2D metformin+ 4253796800 ERR260196
# 782 NG-5636_563            SWE T2D metformin+ 3464254550 ERR260202
# 783 NG-5636_591            SWE T2D metformin+ 3833173412 ERR260208
# 784 NG-5636_604            SWE T2D metformin+ 3923443172 ERR260214

names(subjects)
# "Sample"         "Country subset" "Status"         "Bases"          "Run"  


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"


write.csv(x = subjects, file = "Forslund-T2D-SWE-subjects-status-sra-run-list.txt", quote = FALSE, row.names = FALSE)

# # Download from command line using format:
# fastq-dump --outdir /scratch/user/lidd0026/_1_meta_raw --skip-technical --readids --read-filter pass --dumpbase --clip --split-3 ERR719096


identical(subjects$Run,sort(subjects$Run)) # FALSE

write.csv(x = paste0("fastq-dump --outdir /scratch/user/lidd0026/forslund-t2d/ft2d_1_meta_raw --skip-technical --readids --read-filter pass --dumpbase --clip --split-3 ",sort(subjects$Run)), file = "forslund-t2d-SWE-sra-runs-download.txt", quote = FALSE, row.names = FALSE)



#-------------------------

#### Forslund T2D - Metadata
#-------------------------

# metadata as per above

str(subjects)
# 'data.frame':	145 obs. of  5 variables:
# $ Sample        : chr  "NG-5425_137" "NG-5425_249" "NG-5425_342" "NG-5425_485" ...
# $ Country subset: chr  "SWE" "SWE" "SWE" "SWE" ...
# $ Status        : chr  "ND CTRL" "ND CTRL" "ND CTRL" "ND CTRL" ...
# $ Bases         : num  7.98e+08 9.67e+08 7.89e+08 1.09e+09 1.95e+09 ...
# $ Run           : chr  "ERR260268" "ERR260269" "ERR260272" "ERR260275" ...

table(subjects$Status)
# ND CTRL T2D metformin- T2D metformin+ 
#      92             33             20 

length(unique(subjects$Run)) # 145

sampid <- sort( subjects$Run )


#-------------------------

#### Forslund T2D - read in superfocus - fxn potential outputs
#-------------------------

# sampid <- subjects$Run


superfocus_out_dir <- "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus"



list.dirs(superfocus_out_dir)
head( list.dirs(superfocus_out_dir) )

# # don't keep 1st two 
# ( results_dirs <- list.dirs(superfocus_out_dir)[-c(1,2)] )

# # don't keep 1st directory
( results_dirs <- list.dirs(superfocus_out_dir)[-c(1)] )

head(results_dirs)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260132"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260133"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260134"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260135"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260136"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260137"

names(results_dirs) <- gsub(pattern = "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_", replacement = "", x = results_dirs)
head(results_dirs)
# ERR260132 
# "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260132" 
# ERR260133 
# "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260133" 
# ERR260134 
# "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260134" 
# ERR260135 
# "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260135" 
# ERR260136 
# "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260136" 
# ERR260137 
# "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/forslund-t2d/superfocus/superfocus_out_ERR260137"

length(results_dirs) # 145

sel <- which(names(results_dirs) %in% sampid) # qty 145
#results_dirs <- results_dirs[sel]

length( which(names(results_dirs) %in% sampid)) # 145

# check identical order
identical(sampid, names(results_dirs)) # TRUE
length(results_dirs) # 145



# In this data one Run corresponds to a single Sample_ID !!!

# collate results into a long-format table

sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)

for (i in 1:length(sampid)) {
  #i<-1
  this_samp <- sampid[i]
  sel.folder <- grep(pattern = this_samp, x = results_dirs)
  this_folder <- results_dirs[sel.folder]
  
  #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)
  
  tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
  # names(tab)
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # [5] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq"
  # [6] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq.."
  
  
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # [5] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"
  # [6] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"
  # [7] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"
  # [8] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"
  # [9] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [10] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  # [11] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [12] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  
  
  tab$sampid <- this_samp
  names(tab)
  
  #tab <- tab[,c(7,1,2,3,4,6)]
  
  # last column is sampid
  # take average of percentages
  
  sel.col.percent <- grep(pattern = "R1.good.fastq..$", x = names(tab))
  if (length(sel.col.percent)>1) {
    tab$percent_abun <- apply(X = tab[ ,sel.col.percent], MARGIN = 1, FUN = mean )
  } else {
    tab$percent_abun <- tab[ ,sel.col.percent]
  }
  
  # sum(tab$percent_abun) # 100
  # mean(tab$percent_abun) # 0.004338583
  
  names(sfx.long) # "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"    "percent_abun"
  # names(tab)
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # ...
  # [13] "sampid"
  # [14] "percent_abun"
  
  tab <- tab[ ,c("sampid","Subsystem.Level.1","Subsystem.Level.2","Subsystem.Level.3","Function","percent_abun")]
  names(tab) <- names(sfx.long)
  
  sfx.long <- rbind(sfx.long,tab)
  
  print(paste0("completed ",i," - sample ID: ",sampid[i]))
}






head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
dim(sfx.long) # 1131150       6
head(sfx.long)
# sampleID                   subsys_L1 subsys_L2           subsys_L3
# 2 ERR260132 Amino Acids and Derivatives         - Amino acid racemase
# 3 ERR260132 Amino Acids and Derivatives         - Amino acid racemase
# 4 ERR260132 Amino Acids and Derivatives         - Amino acid racemase
# 5 ERR260132 Amino Acids and Derivatives         - Amino acid racemase
# 6 ERR260132 Amino Acids and Derivatives         - Amino acid racemase
# 7 ERR260132 Amino Acids and Derivatives         - Amino acid racemase
# fxn percent_abun
# 2                                                                                                      Alanine_racemase_(EC_5.1.1.1) 1.602096e-02
# 3                                                                                                   Aspartate_racemase_(EC_5.1.1.13) 1.287399e-03
# 4                                                                                             Diaminopimelate_epimerase_(EC_5.1.1.7) 1.227797e-02
# 5                                                                                                    Glutamate_racemase_(EC_5.1.1.3) 1.077601e-02
# 6                           UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-alanine_ligase_(EC_6.3.2.10)_/_Alanine_racemase_(EC_5.1.1.1) 1.304738e-04
# 7 UDP-N-acetylmuramoylalanyl-D-glutamyl-2,6-diaminopimelate--D-alanyl-D-alanine_ligase_(EC_6.3.2.10)_/_Alanine_racemase_(EC_5.1.1.1) 7.629031e-05


sfx.long$full_fxn_tax <- paste0(sfx.long$subsys_L1,"___", sfx.long$subsys_L2,"___", sfx.long$subsys_L3,"___", sfx.long$fxn)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 19099   146

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 19099
length(unique(full_fxn_names)) # 19099

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic" 


tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)


names(sfx.wide)
# [1] "full_fxn_tax" "ERR260132"    "ERR260133"    "ERR260134"    "ERR260135"    "ERR260136"    "ERR260137"    "ERR260138"    "ERR260139"    "ERR260140"    "ERR260141"    "ERR260142"   
# [13] "ERR260143"    "ERR260144"    "ERR260145"    "ERR260146"    "ERR260147"    "ERR260148"    "ERR260149"    "ERR260150"    "ERR260151"    "ERR260152"    "ERR260153"    "ERR260154"   
# [25] "ERR260155"    "ERR260156"    "ERR260157"    "ERR260158"    "ERR260159"    "ERR260160"    "ERR260161"    "ERR260162"    "ERR260163"    "ERR260164"    "ERR260165"    "ERR260166"   
# [37] "ERR260167"    "ERR260168"    "ERR260169"    "ERR260170"    "ERR260171"    "ERR260172"    "ERR260173"    "ERR260174"    "ERR260175"    "ERR260176"    "ERR260177"    "ERR260178"   
# [49] "ERR260179"    "ERR260180"    "ERR260181"    "ERR260182"    "ERR260183"    "ERR260184"    "ERR260185"    "ERR260186"    "ERR260187"    "ERR260188"    "ERR260189"    "ERR260190"   
# [61] "ERR260191"    "ERR260192"    "ERR260193"    "ERR260194"    "ERR260195"    "ERR260196"    "ERR260197"    "ERR260198"    "ERR260199"    "ERR260200"    "ERR260201"    "ERR260202"   
# [73] "ERR260203"    "ERR260204"    "ERR260205"    "ERR260206"    "ERR260207"    "ERR260208"    "ERR260209"    "ERR260210"    "ERR260211"    "ERR260214"    "ERR260215"    "ERR260216"   
# [85] "ERR260217"    "ERR260218"    "ERR260219"    "ERR260220"    "ERR260221"    "ERR260222"    "ERR260223"    "ERR260224"    "ERR260225"    "ERR260226"    "ERR260227"    "ERR260228"   
# [97] "ERR260229"    "ERR260230"    "ERR260231"    "ERR260232"    "ERR260233"    "ERR260234"    "ERR260235"    "ERR260236"    "ERR260237"    "ERR260238"    "ERR260239"    "ERR260240"   
# [109] "ERR260241"    "ERR260242"    "ERR260243"    "ERR260244"    "ERR260245"    "ERR260246"    "ERR260247"    "ERR260248"    "ERR260249"    "ERR260250"    "ERR260251"    "ERR260252"   
# [121] "ERR260253"    "ERR260254"    "ERR260255"    "ERR260256"    "ERR260257"    "ERR260258"    "ERR260259"    "ERR260260"    "ERR260261"    "ERR260262"    "ERR260263"    "ERR260264"   
# [133] "ERR260265"    "ERR260266"    "ERR260267"    "ERR260268"    "ERR260269"    "ERR260270"    "ERR260271"    "ERR260272"    "ERR260273"    "ERR260274"    "ERR260275"    "ERR260276"   
# [145] "ERR275251"    "ERR275252"   

#names(sfx.wide) <- gsub(pattern = "-", replacement = "_", x = names(sfx.wide))

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide)


head(sampid)
# "ERR260132" "ERR260133" "ERR260134" "ERR260135" "ERR260136" "ERR260137"

length(sampid) # 145

names(sampid) # NULL - in this case there is NOT an alternative sample name being used

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # TRUE
# identical(names(sfx.wide) , as.character(gsub(pattern = "-",replacement = "_",x = sampid))) # FALSE
# length( names(sfx.wide) %in% as.character(gsub(pattern = "-",replacement = "_",x = sampid)) ) # 113 - i.e. matching but order different

#NOT RUN THIS TIME
#names(sfx.wide) <- names(sampid)



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 19099     4

length(unique(tax.fxn$subsys_L1)) # 35
length(unique(tax.fxn$subsys_L2)) # 188
length(unique(tax.fxn$subsys_L3)) # 1104
length(unique(tax.fxn$fxn)) # 10161


#-------------------------

#### Forslund T2D - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 19099     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 19099   145

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 19099 taxa and 145 samples ]
# tax_table()   Taxonomy Table:    [ 19099 taxa by 4 taxonomic ranks ]

sample_names(phy)
# [1] "ERR260132" "ERR260133" "ERR260134" "ERR260135" "ERR260136" "ERR260137" "ERR260138" "ERR260139" "ERR260140" "ERR260141" "ERR260142" "ERR260143" "ERR260144" "ERR260145" "ERR260146"
# [16] "ERR260147" "ERR260148" "ERR260149" "ERR260150" "ERR260151" "ERR260152" "ERR260153" "ERR260154" "ERR260155" "ERR260156" "ERR260157" "ERR260158" "ERR260159" "ERR260160" "ERR260161"
# [31] "ERR260162" "ERR260163" "ERR260164" "ERR260165" "ERR260166" "ERR260167" "ERR260168" "ERR260169" "ERR260170" "ERR260171" "ERR260172" "ERR260173" "ERR260174" "ERR260175" "ERR260176"
# [46] "ERR260177" "ERR260178" "ERR260179" "ERR260180" "ERR260181" "ERR260182" "ERR260183" "ERR260184" "ERR260185" "ERR260186" "ERR260187" "ERR260188" "ERR260189" "ERR260190" "ERR260191"
# [61] "ERR260192" "ERR260193" "ERR260194" "ERR260195" "ERR260196" "ERR260197" "ERR260198" "ERR260199" "ERR260200" "ERR260201" "ERR260202" "ERR260203" "ERR260204" "ERR260205" "ERR260206"
# [76] "ERR260207" "ERR260208" "ERR260209" "ERR260210" "ERR260211" "ERR260214" "ERR260215" "ERR260216" "ERR260217" "ERR260218" "ERR260219" "ERR260220" "ERR260221" "ERR260222" "ERR260223"
# [91] "ERR260224" "ERR260225" "ERR260226" "ERR260227" "ERR260228" "ERR260229" "ERR260230" "ERR260231" "ERR260232" "ERR260233" "ERR260234" "ERR260235" "ERR260236" "ERR260237" "ERR260238"
# [106] "ERR260239" "ERR260240" "ERR260241" "ERR260242" "ERR260243" "ERR260244" "ERR260245" "ERR260246" "ERR260247" "ERR260248" "ERR260249" "ERR260250" "ERR260251" "ERR260252" "ERR260253"
# [121] "ERR260254" "ERR260255" "ERR260256" "ERR260257" "ERR260258" "ERR260259" "ERR260260" "ERR260261" "ERR260262" "ERR260263" "ERR260264" "ERR260265" "ERR260266" "ERR260267" "ERR260268"
# [136] "ERR260269" "ERR260270" "ERR260271" "ERR260272" "ERR260273" "ERR260274" "ERR260275" "ERR260276" "ERR275251" "ERR275252"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta.ok) # 106  4
dim(subjects) # 145  5

head(subjects)
#          Sample Country subset  Status      Bases       Run
# 640 NG-5425_137            SWE ND CTRL  797799936 ERR260268
# 641 NG-5425_249            SWE ND CTRL  966913152 ERR260269
# 642 NG-5425_342            SWE ND CTRL  789010752 ERR260272
# 643 NG-5425_485            SWE ND CTRL 1090824000 ERR260275
# 644 NG-5636_112            SWE ND CTRL 1954234860 ERR260249
# 645 NG-5636_118            SWE ND CTRL 3307295096 ERR260250



head(row.names(subjects)) # "640" "641" "642" "643" "644" "645"
row.names(subjects) <- subjects$Run

identical(row.names(subjects), sample_names(phy)) # FALSE
identical(sort( row.names(subjects)), sample_names(phy)) # TRUE

subjects2 <- subjects[ sort(row.names(subjects)), ]

SAMP <- sample_data(subjects2)



### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 19099 taxa and 145 samples ]
# sample_data() Sample Data:       [ 145 samples by 5 sample variables ]
# tax_table()   Taxonomy Table:    [ 19099 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#   subsys_L1                     subsys_L2 subsys_L3             fxn                                                                            
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylaconitate_cis-trans_isomerase"                                        
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylcitrate_dehydratase_(2-methyl-trans-aconitate_forming)_(EC_4.2.1.117)"
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "2-methylcitrate_dehydratase_FeS_dependent_(EC_4.2.1.79)"                      
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                      
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Alanine_racemase_(EC_5.1.1.1)"                                                
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase" "Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic" 


getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-Forslund-SWE-T2D.RDS")

#phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")

#-------------------------

#### Forslund T2D - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 19099    4



get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)
        
        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1
          
          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else
        
      } # END else
      
      coefs[[c]] <- coef
      
    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  
  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      

    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Forslund-T2D-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%  #
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()






time.start #  "2023-04-07 13:53:55 ACST"
time.finish # "2023-04-07 15:21:34 ACST"

# 1.5 hr

## assemble results


modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Forslund-T2D-R-working-files"



# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

dim(df.tax) # 19099     4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 22383     9

table(df.out$matching_method)
# EC number      Exact hierachy match Matched Reactions aliases    Matched Reactions name     Matched Subsytem role 
# 8054                      3706                        37                       332                       566 
# No match found 
# 9688 

100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 43.28285% No match found

# 57% match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.618   0.814   0.861   1.083   2.500    8941

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.257   1.401   1.427   1.600   4.000    8941 

table(df.out.distil$matching_method)
# EC number                       EC number;Exact hierachy match 
# 5851                                                  547 
# EC number;Exact hierachy match;Matched Subsytem role        EC number;Exact hierachy match;No match found 
# 2                                                   22 
# EC number;Matched Reactions aliases                     EC number;Matched Reactions name 
# 3                                                    2 
# EC number;Matched Subsytem role       EC number;Matched Subsytem role;No match found 
# 70                                                    1 
# EC number;No match found                                 Exact hierachy match 
# 477                                                 2556 
# Exact hierachy match;Matched Reactions name           Exact hierachy match;Matched Subsytem role 
# 2                                                   22 
# Exact hierachy match;No match found                            Matched Reactions aliases 
# 197                                                   33 
# Matched Reactions aliases;No match found                               Matched Reactions name 
# 1                                                  322 
# Matched Reactions name;No match found                                Matched Subsytem role 
# 6                                                  409 
# Matched Subsytem role;No match found                                       No match found 
# 54                                                 8522 

100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 38.07354% No match found

100-38 # i.e. in 62% of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_Forslund-SWE-T2D.RDS")

#df.out.distil <- readRDS(file = "df.out.distil_Forslund-SWE-T2D.RDS")

#-------------------------

#### Forslund T2D - inspect & clean data
#-------------------------

this_study <- "-Forslund-T2D-"

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")

df.out.distil <- readRDS("df.out.distil_Forslund-SWE-T2D.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"


df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 19099     4


phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE

df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 19099   145
df.OTU[1:5, 1:8]
# ERR260132   ERR260133   ERR260134  ERR260135    ERR260136   ERR260137   ERR260138   ERR260139
# fxn_1 0.00000000 0.000000000 0.000000000 0.00000000 0.0000000000 0.000000000 0.000000000 0.000000000
# fxn_2 0.00000000 0.000000000 0.000000000 0.00000000 0.0000000000 0.000000000 0.000000000 0.000000000
# fxn_3 0.00000000 0.000000000 0.000000000 0.00000000 0.0000000000 0.000000000 0.000000000 0.000000000
# fxn_4 0.00000000 0.000000000 0.000000000 0.00000000 0.0000000000 0.000000000 0.000000000 0.000000000
# fxn_5 0.01602096 0.004141166 0.004530646 0.00113941 0.0007990692 0.001384355 0.001487265 0.009191508

mean( df.OTU[ , "ERR260132"] ) # 0.005235876
sum( df.OTU[ , "ERR260132"] ) # 100

sample_sums(phy) # all values of 100


identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
# tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.9285714     0.8571429
# fxn_3     0.9285714     0.8571429
# fxn_4     0.6000000     1.8000000
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333



dim(df.fxn_vk_coords) # 19099     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.618   0.814   0.861   1.083   2.500    8941 
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.257   1.401   1.427   1.600   4.000    8941

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.9285714     0.8571429
# fxn_3     0.9285714     0.8571429
# fxn_4     0.6000000     1.8000000
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333
# fxn_7     0.6666667     2.3333333

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(df.fxn_vk_coords.noNAs, "df.fxn_vk_coords.noNAs.Forslund-SWE-T2D.RDS")


## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 145
dim(df.fxn_vk_coords.noNAs) # 10158     2
dim(df.OTU.noNAs) # 10158   145

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# ERR260147 ERR260265 ERR260221 ERR260234 ERR260229 ERR260227 ERR260219 ERR260222 ERR260146 ERR260158 ERR260216 ERR260238 ERR260140 
# 55.91464  59.63855  59.77112  59.78290  59.86268  60.01362  60.18155  60.28752  60.43768  60.44684  60.49991  60.68396  60.96656 
# ERR260167 ERR260230 ERR260160 ERR260218 ERR260250 ERR260148 ERR260135 ERR260159 ERR260133 ERR260261 ERR260203 ERR260170 ERR260149 
# 61.03736  61.18269  61.31864  61.87692  61.95002  62.03169  62.04640  62.05591  62.14737  62.21644  62.23910  62.47037  62.55147 
# ERR260264 ERR260204 ERR260166 ERR260246 ERR260137 ERR260208 ERR260251 ERR260151 ERR260255 ERR260181 ERR260231 ERR260136 ERR260196 
# 62.60450  62.97600  63.03807  63.04356  63.09180  63.13979  63.18380  63.29131  63.29902  63.33474  63.37211  63.41600  63.48867 
# ERR260256 ERR260153 ERR260242 ERR260182 ERR260252 ERR260171 ERR260184 ERR260157 ERR260258 ERR260178 ERR260245 ERR260155 ERR260224 
# 63.49094  63.50023  63.51057  63.52311  63.61396  63.63355  63.70962  63.75970  63.77400  63.79285  63.85328  63.87730  63.88186 
# ERR260169 ERR260190 ERR260235 ERR260257 ERR260232 ERR260225 ERR260193 ERR260154 ERR260248 ERR260228 ERR260164 ERR260199 ERR260239 
# 63.92691  63.99716  64.02831  64.07238  64.17629  64.18522  64.19479  64.24187  64.24877  64.27820  64.38420  64.50581  64.53791 
# ERR260220 ERR260141 ERR260237 ERR260217 ERR260263 ERR275251 ERR260202 ERR260174 ERR260268 ERR260206 ERR260276 ERR260259 ERR260223 
# 64.60550  64.71059  64.73111  64.82549  64.84582  64.85305  64.90269  64.90426  64.94464  64.94585  64.95081  64.99790  65.08516 
# ERR260260 ERR260165 ERR260150 ERR260262 ERR260163 ERR260247 ERR260233 ERR260195 ERR260162 ERR260186 ERR260240 ERR275252 ERR260241 
# 65.13422  65.15312  65.16160  65.16511  65.17193  65.19236  65.22076  65.25660  65.34498  65.34591  65.42876  65.47869  65.63490 
# ERR260143 ERR260198 ERR260183 ERR260201 ERR260273 ERR260266 ERR260189 ERR260236 ERR260179 ERR260209 ERR260142 ERR260244 ERR260156 
# 65.63544  65.66953  65.69418  65.71388  65.72939  65.73940  65.75344  65.78273  65.81990  65.83547  65.84652  65.86180  65.87199 
# ERR260210 ERR260188 ERR260253 ERR260270 ERR260205 ERR260145 ERR260200 ERR260267 ERR260249 ERR260152 ERR260194 ERR260168 ERR260254 
# 65.87932  65.97516  66.03382  66.06111  66.07392  66.11058  66.12779  66.17928  66.19216  66.20201  66.21375  66.24348  66.26219 
# ERR260187 ERR260172 ERR260272 ERR260132 ERR260185 ERR260138 ERR260243 ERR260215 ERR260134 ERR260177 ERR260226 ERR260275 ERR260191 
# 66.28110  66.31984  66.32602  66.38121  66.47844  66.54898  66.55011  66.61164  66.62745  66.63138  66.64181  66.69803  66.71656 
# ERR260207 ERR260197 ERR260173 ERR260161 ERR260175 ERR260176 ERR260214 ERR260144 ERR260269 ERR260192 ERR260180 ERR260211 ERR260139 
# 66.79327  66.81645  66.93860  66.97165  67.18395  67.28306  67.30930  67.33121  67.42090  67.54428  67.81182  67.82954  68.07543 
# ERR260271 ERR260274 
# 68.54857  69.48576

summary( colSums(df.OTU.noNAs) )
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.91   63.37   64.90   64.50   66.07   69.49 

# this reflects that 56-69% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

## PREVIOUSLY for env/soil sample
## this reflects the just >60% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

colSums(df.OTU)
# all values of 100


dim(df.tax) # 19099     4


# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE


dim(df.fxn_vk_coords.noNAs) # 10158     2
dim(df.in) # 10158   145

row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  group=NA, 
                                  samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"],
                         group= rep( phy@sam_data$Status[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}



dim(vk.samp.noNAs.long) # 1472911       8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok)

str(vk.samp.noNAs.long)
unique(vk.samp.noNAs.long$group) # "ND CTRL"        "T2D metformin-" "T2D metformin+"
class(vk.samp.noNAs.long$group) # "character"
vk.samp.noNAs.long$group <- factor(vk.samp.noNAs.long$group, levels = c("T2D metformin-", "T2D metformin+", "ND CTRL"), labels = c("T2D metformin-", "T2D metformin+", "Normal"), ordered = TRUE) # CHECK GROUP VARIABLE !!
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	1472910 obs. of  8 variables:
# $ sample   : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ fxn_id   : Factor w/ 10158 levels "fxn_10","fxn_100",..: 4627 5509 6263 7042 7621 8099 8854 9640 1 514 ...
# $ abun     : num  0 0 0 0.016 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.929 0.929 0.6 0.667 0.667 ...
# $ HC_y     : num  0.857 0.857 1.8 2.333 2.333 ...
# $ group    : Ord.factor w/ 3 levels "T2D metformin-"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel: chr  "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" ...

this_study # "Forslund-T2D"

saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-object-Forslund-T2D.RDS")


#vk.samp.noNAs.long <- readRDS(file = "vk.samp.noNAs.long-object-Forslund-T2D.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    #legend.text = element_text(angle=25),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "T2D & Normal human gut", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust =0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")




## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Forslund-T2D-R-working-files"



## check anomalous high value in Lignin zone in PCaN density plot
#p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405)

length(sel) # 35090
dim(vk.samp.noNAs.long) # 1472910       8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 2.382359% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
# sample  fxn_id         abun abun_type      OC_x     HC_y  group           samplabel
# 13  ERR260132  fxn_13 7.152216e-05   relabun 0.5206427 1.400701 Normal ERR260132 (ND CTRL)
# 21  ERR260132  fxn_24 0.000000e+00   relabun 0.5206427 1.400701 Normal ERR260132 (ND CTRL)
# 62  ERR260132  fxn_80 0.000000e+00   relabun 0.5206427 1.400701 Normal ERR260132 (ND CTRL)
# 164 ERR260132 fxn_213 2.145665e-04   relabun 0.5206427 1.400701 Normal ERR260132 (ND CTRL)
# 223 ERR260132 fxn_285 2.145665e-04   relabun 0.5206427 1.400701 Normal ERR260132 (ND CTRL)
# 254 ERR260132 fxn_331 6.436995e-04   relabun 0.5206427 1.400701 Normal ERR260132 (ND CTRL)

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.06962988% per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 64.49825% per sample

# remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702) # qty 9633


#fxn_13
i<-13
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED
# 1         fxn_13 1 hypothetical protein Matched Reactions name                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427      1.400701


### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 35090


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-Forslund-T2D.RDS")

this_study # "-Forslund-T2D-"

#-------------------------

#### Forslund T2D - CPP-class: total rel abun in vk spatial regions
#-------------------------


this_study <- "-Forslund-T2D-"
header <- "-vk-zones-"

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Forslund-SWE-T2D.RDS")

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 19099     4

# use clean dataset
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Forslund-T2D.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

str(vk.samp.noNAs.long)
# 'data.frame':	1437820 obs. of  8 variables:
# $ sample   : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ fxn_id   : Factor w/ 10158 levels "fxn_10","fxn_100",..: 4627 5509 6263 7042 7621 8099 8854 9640 1 514 ...
# $ abun     : num  0 0 0 0.016 0 ...
# $ abun_type: chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x     : num  0.929 0.929 0.6 0.667 0.667 ...
# $ HC_y     : num  0.857 0.857 1.8 2.333 2.333 ...
# $ group    : Ord.factor w/ 3 levels "T2D metformin-"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel: chr  "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" ...



# NOTE THIS SWE T2D DATASET IS FEMALE ONLY!!
# Karlsson, F. H. et al. Gut metagenome in European women with normal,
# impaired and diabetic glucose control. Nature 498, 99–103 (2013).

# T2D (n = 53), impaired glucose tolerance (IGT; n = 49) or normal glucose tolerance (NGT; n = 43)

table(phy@sam_data$Status)
# ND CTRL T2D metformin- T2D metformin+ 
#   92             33             20 


df.temp <- vk.samp.noNAs.long

# but additional data from Karlsson et al 2013 includes impaired glucose tolerance (IGT; n = 49) or normal glucose tolerance (NGT; n = 43)

# this lookup has IGT / NGT classification
t2d.class <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Forslund-2015-Type2-diabetes/Karlsson et al 2013--41586_2013_BFnature12198_MOESM507_ESM.xlsx",
                   sheet="Supplementary Table 3", range="A2:AD147", col_names = TRUE)
t2d.class <- as.data.frame(t2d.class)
str(t2d.class)
# 'data.frame':	145 obs. of  30 variables:
# $ Sample ID                                                                             : num  51 53 54 58 59 60 77 80 88 92 ...
# $ Age (years)                                                                           : num  69.1 70.3 69.9 70.2 69.4 ...
# $ Classification                                                                        : chr  "IGT" "NGT" "IGT" "NGT" ...
# etc ...

# this lookup has no of reads / bases
t2d.nbases <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/data/Forslund-2015-Type2-diabetes/Karlsson et al 2013--41586_2013_BFnature12198_MOESM507_ESM.xlsx",
                        sheet="Supplementary Table 4", range="A2:C147", col_names = TRUE)
t2d.nbases <- as.data.frame(t2d.nbases)
str(t2d.nbases)
# 'data.frame':	145 obs. of  3 variables:
# $ Sample ID                 : num  51 53 54 58 59 60 77 80 88 92 ...
# $ Number of reads           : num  10653221 13971868 13561331 13213300 12460015 ...
# $ Total number of bases (bp): num  2.15e+09 2.82e+09 2.74e+09 2.67e+09 2.52e+09 ...

identical(t2d.class$`Sample ID`,t2d.nbases$`Sample ID`) # TRUE

str(meta)
# 'data.frame':	145 obs. of  35 variables:
# $ Run                   : chr  "ERR260132" "ERR260133" "ERR260134" "ERR260135" ...
# $ Assay Type            : chr  "WGS" "WGS" "WGS" "WGS" ...
# $ AvgSpotLen            : num  202 202 202 202 202 202 202 202 202 202 ...
# $ Bases                 : num  2.76e+09 2.66e+09 1.49e+10 1.40e+10 2.59e+09 ...

unique_samps <- unique(df.temp$sample) # qty 145

df.temp$group_new <- NA

for (i in 1:length(unique_samps)) {
  #i<-12
  this_run <- unique_samps[i]
  sel.rows <- which(df.temp$sample == this_run)
  
  sel.meta <- which(meta$Run == this_run)
  no_bases <- meta$Bases[sel.meta]
  
  sel.bases <- which(t2d.nbases$`Total number of bases (bp)` == no_bases)
  
  if (length(sel.bases)==1) {
    sampleID <- t2d.nbases$`Sample ID`[sel.bases]
    
    sel.class <- which(t2d.class$`Sample ID` == sampleID)
    new_class <- t2d.class$Classification[sel.class]
    
  }
  
  df.temp$group_new[sel.rows] <- new_class
  
  print(paste0("completed ", i))
  
}


unique(df.temp$group_new) # "IGT" "T2D" "NGT"

df.temp$group_new2 <- paste0(df.temp$group,"__",df.temp$group_new)

unique(df.temp$group_new2)
# "Normal__IGT"         "T2D metformin-__T2D" "Normal__NGT"         "T2D metformin+__T2D"

# T2D (n = 53), impaired glucose tolerance (IGT; n = 49) or normal glucose tolerance (NGT; n = 43)

df.temp$group_new2 <- gsub(pattern = "Normal__NGT", replacement = "Normal", x = df.temp$group_new2)
df.temp$group_new2 <- gsub(pattern = "Normal__IGT", replacement = "IGT", x = df.temp$group_new2)
df.temp$group_new2 <- gsub(pattern = "T2D metformin-__T2D", replacement = "T2D met neg", x = df.temp$group_new2)
#df.temp$group_new2 <- gsub(pattern = "T2D metformin+__T2D", replacement = "T2D met pos", x = df.temp$group_new2)
sel <- which(df.temp$group_new2 == "T2D metformin+__T2D")
df.temp$group_new2[sel] <- "T2D met pos"

unique(df.temp$group_new2)
# "IGT"         "T2D met neg" "Normal"      "T2D met pos"

df.temp$group_new2 <- factor(df.temp$group_new2, levels = c("T2D met neg", "T2D met pos", "IGT", "Normal"), ordered=TRUE)

vk.samp.noNAs.long <- df.temp

saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long.cleaned.addIGT--Forslund-T2D.RDS")



names(df.temp)
#  "sample"     "fxn_id"     "abun"       "abun_type"  "OC_x"       "HC_y"       "group"      "samplabel"  "group_new"  "group_new2"


# plot as spatial data in vK space

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group_new2", "OC_x", "HC_y", "abun") ]

names(spdf)[2] # "group_new2"
names(spdf)[2] <- "group"

# add sex ... NO this study is females only


dim(spdf) #  651009      5

head(spdf)
# sample group      OC_x     HC_y         abun
# 5  ERR260132   IGT 0.6666667 2.333333 1.602096e-02
# 9  ERR260132   IGT 1.0000000 1.500000 1.287399e-03
# 10 ERR260132   IGT 0.5714286 2.000000 1.227797e-02
# 11 ERR260132   IGT 0.8000000 1.600000 1.077601e-02
# 15 ERR260132   IGT 0.7489216 1.934608 1.304738e-04
# 16 ERR260132   IGT 0.7489216 1.934608 7.629031e-05


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 145

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.83   63.33   64.86   64.43   66.02   69.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"  

## now want to sum total rela abun within vK spatial zones

## to de-duplicate get total relabun within a sample, then mean across samples

# this case study can use 'sum_relabun_in_vk_zones_env'
# because it is female-only, and there is no sex variable in the data frame

sum_relabun_in_vk_zones_env

# function(spdf_in, samps) {
#   ###spdf_in = spdf
#   ###samps = unique_samps
#   ###i = 1
#   this_samp <- samps[i]
#   sel <- which(spdf_in$sample == this_samp)
#   
#   spdf_in <- spdf_in[sel, ]
#   
#   # lipids
#   subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
#   subsel.all <- c(subsel.lipids)
#   # proteins
#   subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
#   subsel.all <- c(subsel.all, subsel.proteins)
#   # amino_sugars
#   subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
#   subsel.all <- c(subsel.all, subsel.amino_sugars)
#   # carbs - Carbohydrates
#   subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
#   subsel.all <- c(subsel.all, subsel.carbs)
#   # cond_aromtx - Condensed aromatics
#   subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
#   subsel.all <- c(subsel.all, subsel.cond_aromtx)
#   # lignin
#   subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
#   subsel.all <- c(subsel.all, subsel.lignin)
#   # tannins
#   subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
#   subsel.all <- c(subsel.all, subsel.tannins)
#   # Other compounds ...
#   
#   # methylated
#   subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
#   # and not already used?
#   sel.dup <- which(subsel.methylated %in% subsel.all)
#   if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
#   subsel.all <- c(subsel.all, subsel.methylated)
#   
#   # hydrated
#   subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
#   # and not already used?
#   sel.dup <- which(subsel.hydrated %in% subsel.all)
#   if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
#   subsel.all <- c(subsel.all, subsel.hydrated)
#   
#   # demethylated / Oxidised-dehydrogenated
#   subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
#   # and not already used?
#   sel.dup <- which(subsel.demethylated %in% subsel.all)
#   if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
#   subsel.all <- c(subsel.all, subsel.demethylated)
#   
#   # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
#   #subsel.condensed = which(                   spdf_in$OC_x <= 0.97                       & spdf_in$HC_y < 0.75)
#   subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
#   
#   # and not already used?
#   sel.dup <- which(subsel.condensed %in% subsel.all)
#   if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
#   subsel.all <- c(subsel.all, subsel.condensed)
#   
#   # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
#   subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
#   # and not already used?
#   sel.dup <- which(subsel.k2_b12 %in% subsel.all)
#   if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
#   
#   df_out <- data.frame(
#     sample = this_samp,
#     group = unique(spdf_in$group),
#     
#     #sex = unique(spdf_in$sex),
#     
#     lipids = sum(spdf_in$abun[ subsel.lipids ]),
#     proteins = sum(spdf_in$abun[ subsel.proteins ]),
#     amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
#     carbs = sum(spdf_in$abun[ subsel.carbs ]),
#     cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
#     lignin = sum(spdf_in$abun[ subsel.lignin ]),
#     tannins = sum(spdf_in$abun[ subsel.tannins ]),
#     # Other compounds ...
#     methylated = sum(spdf_in$abun[ subsel.methylated ]),
#     hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
#     demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
#     condensed = sum(spdf_in$abun[ subsel.condensed ]),
#     k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
#   )
#   return(df_out)
# }



time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones_env( spdf_in = spdf, samps = unique_samps)
  #sum_relabun_in_vk_zones( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()




length(out) # 145
out[[1]]

names(out[[1]])
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"     
# [10] "methylated"   "hydrated"     "demethylated" "condensed"    "k2_b12"

df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample group     lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed     k2_b12
# 1 ERR260132   IGT 0.03438786 2.892572     3.783381 17.20662  0.12039087 4.932285 15.54704  0.4585763 6.940524     14.15708 0.2010488 0.04980327
# 2 ERR260133   IGT 0.03948898 2.662378     3.304312 17.11556  0.12228273 4.824248 14.28514  0.4897915 5.792350     12.75896 0.5859011 0.10315941
# 3 ERR260134   IGT 0.16687174 3.085027     3.847989 15.13572  0.20701100 5.858046 16.41319  0.4670384 6.518448     14.60175 0.1800135 0.06892373
# 4 ERR260135   IGT 0.19577625 2.326600     3.540476 16.86854  0.10419167 5.307887 14.29670  0.4777078 5.433785     12.83646 0.3948496 0.15542030
# 5 ERR260136   IGT 0.08068202 2.523069     3.339084 15.96910  0.08893641 4.856440 15.39480  0.4575790 6.041655     14.15942 0.3477629 0.09053455
# 6 ERR260137   IGT 0.10847807 2.587656     3.483784 15.09904  0.10507255 5.201798 15.33871  0.5172366 5.768099     14.35317 0.3536335 0.07870059


names(df.zones)
# [1] "sample"       "group"        "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"      "methylated"  
# [11] "hydrated"     "demethylated" "condensed"    "k2_b12"  

names(df.zones[ ,c(3:14)])

df.zones$all <- apply(X = df.zones[ ,c(3:14)], MARGIN = 1, FUN = sum)

table(df.zones$group)
# T2D met neg T2D met pos         IGT      Normal 
# 33          20          49          43 

saveRDS(object = df.zones, file = "df.zones--Forslund-T2D.RDS")

#df.zones <- readRDS("df.zones--Forslund-T2D.RDS")

str(df.zones)
# 'data.frame':	145 obs. of  15 variables:
# $ sample      : chr  "ERR260132" "ERR260133" "ERR260134" "ERR260135" ...
# $ group       : Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 1 1 3 ...
# $ lipids      : num  0.0344 0.0395 0.1669 0.1958 0.0807 ...
# $ proteins    : num  2.89 2.66 3.09 2.33 2.52 ...
# $ amino_sugars: num  3.78 3.3 3.85 3.54 3.34 ...
# $ carbs       : num  17.2 17.1 15.1 16.9 16 ...
# $ cond_aromtx : num  0.1204 0.1223 0.207 0.1042 0.0889 ...
# $ lignin      : num  4.93 4.82 5.86 5.31 4.86 ...
# $ tannins     : num  15.5 14.3 16.4 14.3 15.4 ...
# $ methylated  : num  0.459 0.49 0.467 0.478 0.458 ...
# $ hydrated    : num  6.94 5.79 6.52 5.43 6.04 ...
# $ demethylated: num  14.2 12.8 14.6 12.8 14.2 ...
# $ condensed   : num  0.201 0.586 0.18 0.395 0.348 ...
# $ k2_b12      : num  0.0498 0.1032 0.0689 0.1554 0.0905 ...
# $ all         : num  66.3 62.1 66.6 61.9 63.3 ...


df.zones.melt <- melt(df.zones, id.vars = c("sample","group"))
head(df.zones.melt)
# sample group variable      value
# 1 ERR260132   IGT   lipids 0.03438786
# 2 ERR260133   IGT   lipids 0.03948898
# 3 ERR260134   IGT   lipids 0.16687174
# 4 ERR260135   IGT   lipids 0.19577625
# 5 ERR260136   IGT   lipids 0.08068202
# 6 ERR260137   IGT   lipids 0.10847807

unique( df.zones.melt$group )
# [1] IGT         T2D met neg Normal      T2D met pos
# Levels: T2D met neg < T2D met pos < IGT < Normal


str(df.zones.melt)
# 'data.frame':	1885 obs. of  4 variables:
#   $ sample  : chr  "ERR260132" "ERR260133" "ERR260134" "ERR260135" ...
# $ group   : Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 1 1 3 ...
# $ variable: Factor w/ 13 levels "lipids","proteins",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  0.0344 0.0395 0.1669 0.1958 0.0807 ...

hist(df.zones.melt$value)
hist(log10( df.zones.melt$value))

#df.zones.melt$log10abun <- log10( df.zones.melt$value)

p <- ggplot(data = df.zones.melt, aes(x = group, y = value))+
  geom_boxplot(outlier.shape = NA)+
  #geom_point(aes(color = sex, shape = sex), position = position_jitterdodge(0.2), alpha=0.6)+
  geom_point( alpha=0.6)+
  facet_wrap(facets = vars(variable), scales = "free_y")
p


dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=600, compression="lzw",type="cairo")


names(df.zones.melt)
#  "sample"   "group"    "variable" "value" 

dat <- df.zones.melt

dim(dat) #  1885    4


ok <- complete.cases(dat)
sel.ok <- which(ok==TRUE)


## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                      "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                        "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds" "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"

unique(dat$group)
# [1] IGT         T2D met neg Normal      T2D met pos
# Levels: T2D met neg < T2D met pos < IGT < Normal

dat$group_short <- factor(dat$group, levels = c("T2D met neg", "T2D met pos", "IGT", "Normal"),
                          labels = c("T2D met-", "T2D met+", "IGT", "Normal"),
                          ordered = TRUE)

y_anotations <- list()
sig_anotations <- list()



## lipids
this_var <- "Lipids"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value  Pr(>F)  
# group   3  2.3171 0.07823 .
#       141                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.0439 0.014634   1.891  0.134
# Residuals   141 1.0909 0.007737

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 2.2586, df = 3, p-value = 0.5205

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Proteins"
this_var <- "Proteins"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  2.0807 0.1055
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3  0.559 0.18620   2.072  0.107
# Residuals   141 12.673 0.08988 

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 5.0426, df = 3, p-value = 0.1687

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Amino sugars"
this_var <- "Amino sugars"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  1.0082 0.3911
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3  0.109 0.03648   0.927   0.43
# Residuals   141  5.552 0.03938

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 7.8467, df = 3, p-value = 0.04929
# 
# 
# ## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector
# levels( dat$group ) # "T2D met neg" "T2D met pos" "IGT"         "Normal"
# 
# ## adjusts the p-value for multiple comparisons using the Bonferroni
# pt <- dunnTest( x = dat$value[which(dat$variable==this_var)], g = dat$group[which(dat$variable==this_var)], method = "bonferroni" ) # 
# pt
# # Dunn (1964) Kruskal-Wallis multiple comparison
# # p-values adjusted with the Bonferroni method.
# # 
# #                  Comparison          Z   P.unadj     P.adj
# # 1              IGT - Normal -1.2465584 0.2125595 1.0000000
# # 2         IGT - T2D met neg -2.7986106 0.0051323 0.0307938
# # 3      Normal - T2D met neg -1.5976611 0.1101184 0.6607105
# # 4         IGT - T2D met pos -0.8606379 0.3894375 1.0000000
# # 5      Normal - T2D met pos  0.1186503 0.9055524 1.0000000
# # 6 T2D met neg - T2D met pos  1.4180955 0.1561629 0.9369773
# pt$dtres
# pt <- pt$res
# pt
# cldList(comparison = pt$Comparison,
#         p.value    = pt$P.adj,
#         threshold  = 0.05)
# #       Group Letter MonoLetter
# # 1       IGT      a         a 
# # 2    Normal     ab         ab
# # 3 T2Dmetneg      b          b
# # 4 T2Dmetpos     ab         ab
# 
# ## Include significance letters in plot
# levels(dat$group) # "T2D met neg" "T2D met pos" "IGT"         "Normal"
# 
# y_anot <- c(3.8, 3.8, 3.8, 3.8)
# sig_anot <- c("b","ab","a","ab")

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Carbohydrates"
this_var <- "Carbohydrates"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3     0.6 0.6161
#       141

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3   16.5   5.513   1.989  0.118
# Residuals   141  390.9   2.772    

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 2.463, df = 3, p-value = 0.482

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value  Pr(>F)  
# group   3  2.4962 0.06231 .
#       141                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df  Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.00422 0.001408   0.861  0.463
# Residuals   141 0.23046 0.001635

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 1.9297, df = 3, p-value = 0.5871

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Lignin"
this_var <- "Lignin"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value  Pr(>F)  
# group   3  2.6146 0.05358 .
#       141                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3  0.473  0.1577   1.545  0.206
# Residuals   141 14.392  0.1021

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 2.5804, df = 3, p-value = 0.4609

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Tannins"
this_var <- "Tannins"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.2493 0.8617
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3   3.51   1.169   0.851  0.468
# Residuals   141 193.75   1.374

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 4.6977, df = 3, p-value = 0.1953

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other methylated\ncompounds"
this_var <- "Other methylated\ncompounds"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.0369 0.9905
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq  Mean Sq F value Pr(>F)  
# group         3 0.0365 0.012165    2.77 0.0439 *
# Residuals   141 0.6192 0.004392                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Tukey test with compact letter display
tukey <- TukeyHSD(res.aov)
# https://www.rdocumentation.org/packages/multcompView/versions/0.1-9/topics/multcompLetters
tukey.cld <- multcompLetters4(res.aov, tukey) # multcompLetters4(): create a compact letters display using a aov object
tukey.cld
# $group
# Normal T2D met neg T2D met pos         IGT 
# "a"        "ab"        "ab"         "b" 

temp <- filter(dat, variable == this_var)
tukey.cld$group$Letters[ levels(temp$group) ]
# T2D met neg T2D met pos         IGT      Normal 
#        "ab"        "ab"         "b"         "a"

summary(temp$value)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.3523  0.4434  0.4794  0.4863  0.5172  0.8728

y_anot <- c(0.68, 0.68, 0.68, 0.68)
sig_anot <- c("ab","ab","b","a")

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 8.2587, df = 3, p-value = 0.04096
# 
# ## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector
# levels( dat$group ) # "T2D met neg" "T2D met pos" "IGT"         "Normal"
# 
# ## adjusts the p-value for multiple comparisons using the Bonferroni
# pt <- dunnTest( x = dat$value[which(dat$variable==this_var)], g = dat$group[which(dat$variable==this_var)], method = "bonferroni" ) # 
# pt
# # Dunn (1964) Kruskal-Wallis multiple comparison
# # p-values adjusted with the Bonferroni method.
# # 
# # Comparison          Z     P.unadj      P.adj
# # 1              IGT - Normal -2.8400139 0.004511157 0.02706694
# # 2         IGT - T2D met neg -1.6227534 0.104642124 0.62785274
# # 3      Normal - T2D met neg  0.9852618 0.324495478 1.00000000
# # 4         IGT - T2D met pos -1.0704870 0.284400135 1.00000000
# # 5      Normal - T2D met pos  1.1431339 0.252983020 1.00000000
# # 6 T2D met neg - T2D met pos  0.2871834 0.773971869 1.00000000
# pt$dtres
# pt <- pt$res
# pt
# cldList(comparison = pt$Comparison,
#         p.value    = pt$P.adj,
#         threshold  = 0.05)
# # Group Letter MonoLetter
# # 1       IGT      a         a 
# # 2    Normal      b          b
# # 3 T2Dmetneg     ab         ab
# # 4 T2Dmetpos     ab         ab
# 
# ## Include significance letters in plot
# levels(dat$group) # "T2D met neg" "T2D met pos" "IGT"         "Normal"
# 
# y_anot <- c(0.68, 0.68, 0.68, 0.68)
# sig_anot <- c("ab","ab","a","b")

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other hydrated\ncompounds"
this_var <- "Other hydrated\ncompounds"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.1589 0.9238
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3   1.33  0.4425   1.636  0.184
# Residuals   141  38.13  0.2704 

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 4.4549, df = 3, p-value = 0.2164

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.5353 0.6588
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3   4.82   1.607   1.078   0.36
# Residuals   141 210.10   1.490 

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 3.6749, df = 3, p-value = 0.2988

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.7033 0.5516
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3  0.044 0.01464   0.403  0.751
# Residuals   141  5.124 0.03634

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 0.85208, df = 3, p-value = 0.837

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds" 

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  1.9108 0.1306
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)  
# group         3 0.1561 0.05204   3.553 0.0161 *
# Residuals   141 2.0648 0.01464                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Tukey test with compact letter display
tukey <- TukeyHSD(res.aov)
# https://www.rdocumentation.org/packages/multcompView/versions/0.1-9/topics/multcompLetters
tukey.cld <- multcompLetters4(res.aov, tukey) # multcompLetters4(): create a compact letters display using a aov object
tukey.cld
# $group
# Normal T2D met pos T2D met neg         IGT 
# "a"        "ab"        "ab"         "b"

temp <- filter(dat, variable == this_var)
tukey.cld$group$Letters[ levels(temp$group) ]
# T2D met neg T2D met pos         IGT      Normal 
#        "ab"        "ab"         "b"         "a" 

summary(temp$value)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.01254 0.07616 0.10286 0.13628 0.15180 0.88442 

y_anot <- c(0.45, 0.45, 0.45, 0.45)
sig_anot <- c("ab","ab","b","a")

# # Kruskal-Wallis test
# kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  value by group
# # Kruskal-Wallis chi-squared = 12.669, df = 3, p-value = 0.005409
# 
# 
# ## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector
# levels( dat$group ) # "T2D met neg" "T2D met pos" "IGT"         "Normal"
# 
# ## adjusts the p-value for multiple comparisons using the Bonferroni
# pt <- dunnTest( x = dat$value[which(dat$variable==this_var)], g = dat$group[which(dat$variable==this_var)], method = "bonferroni" )
# pt
# # Dunn (1964) Kruskal-Wallis multiple comparison
# # p-values adjusted with the Bonferroni method.
# # 
# # Comparison          Z     P.unadj      P.adj
# # 1              IGT - Normal -2.9530326 0.003146688 0.01888013
# # 2         IGT - T2D met neg -1.6528952 0.098352187 0.59011312
# # 3      Normal - T2D met neg  1.0579784 0.290065311 1.00000000
# # 4         IGT - T2D met pos -2.9345005 0.003340850 0.02004510
# # 5      Normal - T2D met pos -0.5970358 0.550483467 1.00000000
# # 6 T2D met neg - T2D met pos -1.4342623 0.151497422 0.90898453
# pt$dtres
# pt <- pt$res
# pt
# cldList(comparison = pt$Comparison,
#         p.value    = pt$P.adj,
#         threshold  = 0.05)
# # Group Letter MonoLetter
# # 1       IGT      a         a 
# # 2    Normal      b          b
# # 3 T2Dmetneg     ab         ab
# # 4 T2Dmetpos      b          b
# 
# ## Include significance letters in plot
# 
# levels(dat$group) # "T2D met neg" "T2D met pos" "IGT"         "Normal"
# 
# y_anot <- c(0.45, 0.45, 0.45, 0.45)
# sig_anot <- c("ab","b","a","b")

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "All compounds"
this_var <- "All compounds"

# levene's test for common variance?
leveneTest(value ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  1.1851 0.3177
#       141 

# Compute the analysis of variance
res.aov <- aov(value ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)  
# group         3   38.9   12.96   2.836 0.0404 *
# Residuals   141  644.4    4.57                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# Tukey test with compact letter display
tukey <- TukeyHSD(res.aov)
# https://www.rdocumentation.org/packages/multcompView/versions/0.1-9/topics/multcompLetters
tukey.cld <- multcompLetters4(res.aov, tukey) # multcompLetters4(): create a compact letters display using a aov object
tukey.cld
# $group
# T2D met neg         IGT T2D met pos      Normal 
# "a"        "ab"        "ab"         "b" 

temp <- filter(dat, variable == this_var)
tukey.cld$group$Letters[ levels(temp$group) ]
# T2D met neg T2D met pos         IGT      Normal 
#          "a"        "ab"        "ab"         "b" 

summary(temp$value)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.83   63.33   64.86   64.43   66.02   69.46 

y_anot <- c(58, 58, 58, 58)
sig_anot <- c("a","ab","ab","b")

# # # Kruskal-Wallis test
# # kt <- kruskal.test( value ~ group, filter(dat, variable == this_var))
# # kt
# # # Kruskal-Wallis rank sum test
# # # 
# # # data:  value by group
# # # Kruskal-Wallis chi-squared = 7.6444, df = 3, p-value = 0.05396
# 
# y_anot <- c(NA,NA,NA,NA)
# sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



# Add significance annotation
# females
annotation_df <- data.frame(
  variable=rep(c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"), each=4),
  x_vals=rep(c(1.3, 2.3, 3.3, 4.3), times=13),
  y_vals = c( y_anotations[["Lipids"]],
         y_anotations[["Proteins"]],
         y_anotations[["Amino sugars"]],
         y_anotations[["Carbohydrates"]],
         y_anotations[["Condensed\naromatics"]],
         y_anotations[["Lignin"]],
         y_anotations[["Tannins"]],
         y_anotations[["Other methylated\ncompounds"]],
         y_anotations[["Other hydrated\ncompounds"]],
         y_anotations[["Other demethylated\ncompounds"]],
         y_anotations[["Other condensed\ncompounds"]],
         y_anotations[["Near VitK2-VitB12\ncompounds"]],
         y_anotations[["All compounds"]]

  ),
  label = c(sig_anotations[["Lipids"]],
            sig_anotations[["Proteins"]],
            sig_anotations[["Amino sugars"]],
            sig_anotations[["Carbohydrates"]],
            sig_anotations[["Condensed\naromatics"]],
            sig_anotations[["Lignin"]],
            sig_anotations[["Tannins"]],
            sig_anotations[["Other methylated\ncompounds"]],
            sig_anotations[["Other hydrated\ncompounds"]],
            sig_anotations[["Other demethylated\ncompounds"]],
            sig_anotations[["Other condensed\ncompounds"]],
            sig_anotations[["Near VitK2-VitB12\ncompounds"]],
            sig_anotations[["All compounds"]]

  ), stringsAsFactors = FALSE)
str(annotation_df)

# ensure no conflict in factor levels
annotation_df$variable <- factor(annotation_df$variable,
                                   levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                              "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                   ordered = TRUE)

annotation_df$label # 




p <- ggplot(data = dat, aes(x = group_short, y = value))+ # y = value win95abun
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(width=0.2, alpha=0.4)+
  
  # sig annotation
  geom_text(data=annotation_df, mapping = aes(x = x_vals, y = y_vals, label = label), size = 3.5)+
  
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function % rel. abun. in vK zones") +
  theme(
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.7))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","Female-ONLY-vkZones-boxplots-",this_study,header,"-v9e.tiff"), width = 24, height = 14, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Forslund T2D - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25)
#-------------------------

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Forslund-SWE-T2D.RDS")
##vk.samp.noNAs.long <- readRDS("vk.samp.noNAs.long--Forslund-T2D.RDS")

#saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long.cleaned.addIGT--Forslund-T2D.RDS")
vk.samp.noNAs.long <- readRDS("vk.samp.noNAs.long.cleaned.addIGT--Forslund-T2D.RDS")

this_study <- "-Forslund-T2D-"
header <- "-vk-hac-buffers-"

# from earlier

str(vk.samp.noNAs.long)
# 'data.frame':	1437820 obs. of  10 variables:
# $ sample    : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ fxn_id    : Factor w/ 10158 levels "fxn_10","fxn_100",..: 4627 5509 6263 7042 7621 8099 8854 9640 1 514 ...
# $ abun      : num  0 0 0 0.016 0 ...
# $ abun_type : chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x      : num  0.929 0.929 0.6 0.667 0.667 ...
# $ HC_y      : num  0.857 0.857 1.8 2.333 2.333 ...
# $ group     : Ord.factor w/ 3 levels "T2D metformin-"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel : chr  "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" ...
# $ group_new : chr  "IGT" "IGT" "IGT" "IGT" ...
# $ group_new2: Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 3 3 3 ...


# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group_new2", "OC_x", "HC_y", "abun") ]

names(spdf)[2] # "group_new2"
names(spdf)[2] <- "group"

# add sex ... NO this study is females only

dim(spdf) #  651009      5

head(spdf)
# sample group      OC_x     HC_y         abun
# 5  ERR260132   IGT 0.6666667 2.333333 1.602096e-02
# 9  ERR260132   IGT 1.0000000 1.500000 1.287399e-03
# 10 ERR260132   IGT 0.5714286 2.000000 1.227797e-02
# 11 ERR260132   IGT 0.8000000 1.600000 1.077601e-02
# 15 ERR260132   IGT 0.7489216 1.934608 1.304738e-04
# 16 ERR260132   IGT 0.7489216 1.934608 7.629031e-05


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 145

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.83   63.33   64.86   64.43   66.02   69.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"

temp <- spdf


hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000026 0.0004320 0.0024221 0.0143503 0.0111937 2.4547910


12.5/(pi*0.05^2) # 1591.549


## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin
df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)

# use _env because no sex variable in this dataset

sum_relabun_in_vk_hac_buffers_env <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  #df_out <- data.frame(sample=NA, group=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA )
  df_out <- data.frame(sample=NA, group=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      group = unique(spdf_in$group),
      #sex = unique(spdf_in$sex),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
    )
    #names(temp_out) # "sample"     "group"      "sex"        "rad"        "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"     
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}



time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers_env( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 145
out[[1]]
# sample group  rad   Acetate Propionate  Butyrate    VitB2    VitB12     VitB6     VitB9     VitK2 Glutamate Pyruvate
# 2 ERR260132   IGT 0.05  74.36659   23.78432  0.000000 54.82711  1.120098 103.75142 0.1092778 0.0000000 130.42819 66.26547
# 3 ERR260132   IGT 0.10 106.15121   15.31498  3.788905 20.68159  1.113680 113.99703 3.7291055 0.0000000  75.01866 78.78180
# 4 ERR260132   IGT 0.15  63.82896   11.74507 16.223352 13.29904  4.168409  85.27629 3.9768356 0.0000000  81.08569 70.22744
# 5 ERR260132   IGT 0.20  45.38571   15.27533 13.273819 14.16630  5.815325  77.74144 7.3316687 0.2487209  70.13778 87.58063
# 6 ERR260132   IGT 0.25  44.00247   19.88987 11.847815 12.58155 11.022876  79.96601 6.5438713 0.6399309  67.89171 88.89323


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample group  rad   Acetate Propionate  Butyrate    VitB2    VitB12     VitB6      VitB9     VitK2 Glutamate Pyruvate
# 2  ERR260132   IGT 0.05  74.36659   23.78432  0.000000 54.82711  1.120098 103.75142 0.10927782 0.0000000 130.42819 66.26547
# 3  ERR260132   IGT 0.10 106.15121   15.31498  3.788905 20.68159  1.113680 113.99703 3.72910547 0.0000000  75.01866 78.78180
# 4  ERR260132   IGT 0.15  63.82896   11.74507 16.223352 13.29904  4.168409  85.27629 3.97683558 0.0000000  81.08569 70.22744
# 5  ERR260132   IGT 0.20  45.38571   15.27533 13.273819 14.16630  5.815325  77.74144 7.33166875 0.2487209  70.13778 87.58063
# 6  ERR260132   IGT 0.25  44.00247   19.88987 11.847815 12.58155 11.022876  79.96601 6.54387131 0.6399309  67.89171 88.89323
# 21 ERR260133   IGT 0.05  69.95111   21.66011  0.000000 59.59717  2.225204 117.76631 0.03766212 0.0000000 194.97742 71.00114


table(df.buffers$group[ which(df.buffers$rad==0.05) ]) # count at single radius only
# IGT      Normal T2D met neg T2D met pos 
# 49          43          33          20 


saveRDS(object = df.buffers, file = "df.buffers--Forslund-T2D.RDS")

str(df.buffers)
# 'data.frame':	725 obs. of  13 variables:
# $ sample    : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ group     : chr  "IGT" "IGT" "IGT" "IGT" ...
# $ rad       : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate   : num  74.4 106.2 63.8 45.4 44 ...
# $ Propionate: num  23.8 15.3 11.7 15.3 19.9 ...
# $ Butyrate  : num  0 3.79 16.22 13.27 11.85 ...
# $ VitB2     : num  54.8 20.7 13.3 14.2 12.6 ...
# $ VitB12    : num  1.12 1.11 4.17 5.82 11.02 ...
# $ VitB6     : num  103.8 114 85.3 77.7 80 ...
# $ VitB9     : num  0.109 3.729 3.977 7.332 6.544 ...
# $ VitK2     : num  0 0 0 0.249 0.64 ...
# $ Glutamate : num  130.4 75 81.1 70.1 67.9 ...
# $ Pyruvate  : num  66.3 78.8 70.2 87.6 88.9 ...


df.buffers.melt <- melt(df.buffers, id.vars = c("sample","group","rad"))
head(df.buffers.melt)
# sample group  rad variable     value
# 1 ERR260132   IGT 0.05  Acetate  74.36659
# 2 ERR260132   IGT 0.10  Acetate 106.15121
# 3 ERR260132   IGT 0.15  Acetate  63.82896
# 4 ERR260132   IGT 0.20  Acetate  45.38571
# 5 ERR260132   IGT 0.25  Acetate  44.00247
# 6 ERR260133   IGT 0.05  Acetate  69.95111

unique( df.buffers.melt$group ) # "IGT"         "T2D met neg" "Normal"      "T2D met pos"
df.buffers.melt$group <- factor(df.buffers.melt$group, 
                                levels = c("T2D met neg", "T2D met pos", "IGT", "Normal"),
                                labels = c("T2D met-", "T2D met+", "IGT", "Normal"),
                                ordered = TRUE)

unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	7250 obs. of  5 variables:
#   $ sample  : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ group   : Ord.factor w/ 4 levels "T2D met-"<"T2D met+"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ rad     : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ variable: Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  74.4 106.2 63.8 45.4 44 ...

hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))

dat <- df.buffers.melt


dat$log10abun <- dat$value # then over-write later


vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}

# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 21.54   39.56   49.34   56.33   66.35  195.74 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.333   1.597   1.693   1.718   1.822   2.292 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 7.212  12.525  15.315  15.888  18.825  27.926 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.858   1.098   1.185   1.187   1.275   1.446 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   3.383  10.433   8.820  13.361  28.021 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.1571  0.5293  1.0184  0.7528  1.1259  1.4475 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 7.696  14.063  15.489  23.212  20.787  98.261 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.8862  1.1481  1.1900  1.2938  1.3178  1.9924 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   1.830   5.631   5.682   9.203  18.877 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.4069  0.2625  0.7506  0.6156  0.9639  1.2759 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 63.46   81.55   87.54   95.65  112.17  151.11 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.802   1.911   1.942   1.973   2.050   2.179 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   3.907   5.143   5.076   7.468  10.950 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.7521  0.5918  0.7112  0.5001  0.8732  1.0394 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00000 0.02118 0.15336 0.45139 0.66568 6.31864 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -2.9670 -1.6740 -0.8143 -1.0797 -0.1767  0.8006 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 42.32   59.26   66.91   81.22   86.16  324.47 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.627   1.773   1.825   1.878   1.935   2.511 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 49.68   75.51   84.04   83.56   92.27  108.44 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.696   1.878   1.924   1.918   1.965   2.035


head(dat)
#      sample group  rad variable     value log10abun
# 1 ERR260132   IGT 0.05  Acetate  74.36659  1.871378
# 2 ERR260132   IGT 0.10  Acetate 106.15121  2.025925
# 3 ERR260132   IGT 0.15  Acetate  63.82896  1.805018
# 4 ERR260132   IGT 0.20  Acetate  45.38571  1.656919
# 5 ERR260132   IGT 0.25  Acetate  44.00247  1.643477
# 6 ERR260133   IGT 0.05  Acetate  69.95111  1.844795


p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free")
p



p <- ggplot(data = dat, aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = group, color = group))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    legend.position="bottom",
    legend.title = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

dev.print(tiff, file = paste0(workdir,"/plots/","Female-ONLY-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.tiff"), width = 22, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


saveRDS(object = dat, file = paste0("dat--Female-ONLY-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--Female-ONLY-HAC-vkBuffers-Loess-smooth-",this_study,header,"-v9.RDS"))


## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]


head(dat)
# sample group  rad variable    value log10abun
# 1  ERR260132   IGT 0.05  Acetate 74.36659  1.871378
# 6  ERR260133   IGT 0.05  Acetate 69.95111  1.844795
# 11 ERR260134   IGT 0.05  Acetate 44.53940  1.648744
# 16 ERR260135   IGT 0.05  Acetate 57.19550  1.757362
# 21 ERR260136   IGT 0.05  Acetate 61.73623  1.790540
# 26 ERR260137   IGT 0.05  Acetate 55.92780  1.747628

p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(width = 0.2, height=0, alpha=0.6)+
  labs(x = NULL, y = "log10 fxn % rel. abun. in 0.05 radius buffer (per vK unit area)") +
  facet_wrap(facets = vars(variable), nrow = 2, scales = "free")+
  theme(
    legend.position = "bottom"
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","Female-ONLY-HAC-vkBuffers-rough-boxplots",this_study,header,"-v9.tiff"), width = 20, height = 20, units = "cm", res=450, compression="lzw",type="cairo")




levels( dat$variable )
# "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"      "Glutamate"  "Pyruvate" 

unique(dat$group)
# [1] IGT      T2D met- Normal   T2D met+
#   Levels: T2D met- < T2D met+ < IGT < Normal

names(dat) # "sample"    "group"     "rad"       "variable"  "value"     "log10abun"


y_anotations <- list()
sig_anotations <- list()



## Acetate
this_var <- "Acetate"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.5004 0.6826
#       141 

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.0157 0.005246   0.695  0.557
# Residuals   141 1.0649 0.007552 

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 1.4979, df = 3, p-value = 0.6828

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Propionate"
this_var <- "Propionate"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3   0.285 0.8362
#       141 

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.0112 0.003723   0.683  0.564
# Residuals   141 0.7684 0.005449

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 1.7001, df = 3, p-value = 0.6369

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Butyrate"
this_var <- "Butyrate"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3     NaN    NaN
#       141

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = NaN, df = 3, p-value = NA

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "VitB2"
this_var <- "VitB2"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3   1.278 0.2843
#       141  

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.0235 0.007829   0.787  0.503
# Residuals   141 1.4032 0.009952

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 2.3839, df = 3, p-value = 0.4966

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "VitB12"
this_var <- "VitB12"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  1.6146 0.1887
#       141  

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3  0.579 0.19290   1.989  0.118
# Residuals   141 13.671 0.09696  

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 4.0588, df = 3, p-value = 0.2552

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "VitB6"
this_var <- "VitB6"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.8959  0.445
#       141

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df  Sum Sq   Mean Sq F value Pr(>F)  
# group         3 0.00616 0.0020535   2.142 0.0976 .
# Residuals   141 0.13516 0.0009586                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 5.7289, df = 3, p-value = 0.1256

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "VitB9"
this_var <- "VitB9"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.3584 0.7831
#       141

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq Mean Sq F value Pr(>F)
# group         3   0.32  0.1081   0.426  0.734
# Residuals   141  35.75  0.2535 

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 1.581, df = 3, p-value = 0.6637

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "VitK2"
this_var <- "VitK2"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value  Pr(>F)  
# group   3  2.9413 0.03529 *
#       141                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# # Compute the analysis of variance
# res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# # Summary of the analysis
# summary(res.aov)

# Kruskal-Wallis test
kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
kt
# Kruskal-Wallis rank sum test
#
# data:  log10abun by group
# Kruskal-Wallis chi-squared = 7.1495, df = 3, p-value = 0.06728

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Glutamate"
this_var <- "Glutamate"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.3247 0.8075
#       141 

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.0179 0.005959   0.376  0.771
# Residuals   141 2.2370 0.015865 

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 0.28247, df = 3, p-value = 0.9633

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Pyruvate" 
this_var <- "Pyruvate"

# levene's test for common variance?
leveneTest(log10abun ~ group, data = filter(dat, variable == this_var))
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.2087 0.8903
#       141

# Compute the analysis of variance
res.aov <- aov(log10abun ~ group, data = filter(dat, variable == this_var))
# Summary of the analysis
summary(res.aov)
#              Df  Sum Sq  Mean Sq F value Pr(>F)
# group         3 0.00873 0.002910   1.306  0.275
# Residuals   141 0.31407 0.002227

# # Kruskal-Wallis test
# kt <- kruskal.test( log10abun ~ group, filter(dat, variable == this_var))
# kt
# # Kruskal-Wallis rank sum test
# # 
# # data:  log10abun by group
# # Kruskal-Wallis chi-squared = 6.6711, df = 3, p-value = 0.08315

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## No significant differences !!
## To update IF/WHEN sig differences - need to add Glutamate, Pyruvate

# # Add significance annotation
# # females
# annotation_df.f <- data.frame(
#   variable=c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2"),
#   start=rep(0.8, times=8), end = rep(1.8, times=8),
#   y = c( y_anotations[["Acetate"]][1],
#          y_anotations[["Propionate"]][1],
#          y_anotations[["Butyrate"]][1],
#          y_anotations[["VitB2"]][1],
#          y_anotations[["VitB12"]][1],
#          y_anotations[["VitB6"]][1],
#          y_anotations[["VitB9"]][1],
#          y_anotations[["VitK2"]][1]
#   ),
#   # label order: Female, Male, All
#   label = c(sig_anotations[["Acetate"]][1],
#             sig_anotations[["Propionate"]][1],
#             sig_anotations[["Butyrate"]][1],
#             sig_anotations[["VitB2"]][1],
#             sig_anotations[["VitB12"]][1],
#             sig_anotations[["VitB6"]][1],
#             sig_anotations[["VitB9"]][1],
#             sig_anotations[["VitK2"]][1]
#   ),
#   stringsAsFactors = FALSE)
# str(annotation_df.f)
# 
# # ensure no conflict in factor levels
# annotation_df.f$variable <- factor(annotation_df.f$variable, 
#                                    levels = c("Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2"),
#                                    ordered = TRUE)
# 
# sel <- which(annotation_df.f$label == "ns")
# annotation_df.f$label[sel] <- NA
# annotation_df.f$start[sel] <- NA
# annotation_df.f$end[sel] <- NA
# annotation_df.f$y[sel] <- NA




p <- ggplot(data = dat, aes(x = group, y = log10abun))+ # y = value
  geom_boxplot(outlier.shape=NA, alpha = 0.5)+
  geom_jitter(width=0.2, height=0, alpha=0.5)+
  
  # # females
  # geom_text(data=annotation_df.f, mapping = aes(x = 1.3, y = y, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  # geom_segment(data=annotation_df.f, mapping = aes(x = start, y = y, xend = end, yend = y), color = "#7570b3")+
  
  # scale_colour_manual(values = cols, name = NULL)+
  # scale_shape_manual(values = shapes, name = NULL)+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+ # scales = "free_y",
  theme_classic() +
  labs(x = NULL, y = "Function log10 % rel. abun. in 0.05 radius vK buffer zones") +
  theme(
    #legend.position="bottom",
    #strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.background = element_rect(fill=NULL, linetype = "blank"), # fill=NULL
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 5,l = 0,"pt"), size = rel(0.8))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","Female-ONLY-vkBuffers-log1oabun-boxplots-",this_study,header,"-v9.tiff"), width = 22, height = 13, units = "cm", res=450, compression="lzw",type="cairo")



#-------------------------

#### Forslund T2D  - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Forslund-SWE-T2D.RDS")
vk.samp.noNAs.long <- readRDS("vk.samp.noNAs.long.cleaned.addIGT--Forslund-T2D.RDS")

this_study <- "-Forslund-T2D-"
header <- "-vk-coords-wtmean-"

str(vk.samp.noNAs.long)
# 'data.frame':	1437820 obs. of  10 variables:
# $ sample    : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ fxn_id    : Factor w/ 10158 levels "fxn_10","fxn_100",..: 4627 5509 6263 7042 7621 8099 8854 9640 1 514 ...
# $ abun      : num  0 0 0 0.016 0 ...
# $ abun_type : chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x      : num  0.929 0.929 0.6 0.667 0.667 ...
# $ HC_y      : num  0.857 0.857 1.8 2.333 2.333 ...
# $ group     : Ord.factor w/ 3 levels "T2D metformin-"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel : chr  "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" ...
# $ group_new : chr  "IGT" "IGT" "IGT" "IGT" ...
# $ group_new2: Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 3 3 3 ...


# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group_new2", "OC_x", "HC_y", "abun") ]

names(spdf)[2] # "group_new2"
names(spdf)[2] <- "group"

# add sex ... NO this study is females only

dim(spdf) #  651009      5

head(spdf)
# sample group      OC_x     HC_y         abun
# 5  ERR260132   IGT 0.6666667 2.333333 1.602096e-02
# 9  ERR260132   IGT 1.0000000 1.500000 1.287399e-03
# 10 ERR260132   IGT 0.5714286 2.000000 1.227797e-02
# 11 ERR260132   IGT 0.8000000 1.600000 1.077601e-02
# 15 ERR260132   IGT 0.7489216 1.934608 1.304738e-04
# 16 ERR260132   IGT 0.7489216 1.934608 7.629031e-05


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 145

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.83   63.33   64.86   64.43   66.02   69.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"

temp <- spdf


hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000026 0.0004320 0.0024221 0.0143503 0.0111937 2.4547910


diags <- character(length = length(unique_samps))
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  diags[i] <- unique(as.character(spdf$group[sel]))
  print(paste0("completed ",i))
}
table(diags)
# IGT      Normal T2D met neg T2D met pos 
# 49          43          33          20 


# use _env version as no sex variable in this dataset

get_wtmean_vk_coords_env
# function(spdf_in, samps) {
#   ###spdf_in = spdf
#   ###samps = unique_samps
#   ###i = 1
#   this_samp <- samps[i]
#   sel <- which(spdf_in$sample == this_samp)
#   
#   spdf_in <- spdf_in[sel, ]
#   
#   subsel <- list()
#   
#   # lipids
#   subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
#   # proteins
#   subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
#   # amino_sugars
#   subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
#   # carbs - Carbohydrates
#   subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
#   # cond_aromtx - Condensed aromatics
#   subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
#   # lignin
#   subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
#   # tannins
#   subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
#   
#   vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")
#   
#   # establish blank template data frame
#   #df_out <- data.frame(sample = NA,group = NA,sex = NA, no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)
#   df_out <- data.frame(sample = NA,group = NA,no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)
#   
#   for (v in 1:length(vars)) {
#     #v<-1
#     this_var <- vars[v]
#     if (length(subsel[[ this_var ]]) > 0) {
#       df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
#       df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
#       df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
#     } else {
#       df_out$wtmean_OC_x[v] <- NA
#       df_out$wtmean_HC_y[v] <- NA
#     }
#     print(paste0("completed ",v))
#   }
#   
#   df_out$sample <- this_samp
#   df_out$group <- unique(spdf_in$group)
#   #df_out$sex <- unique(spdf_in$sex)
#   
#   return(df_out)
# }




time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords_env( spdf_in = spdf, samps = unique_samps)
  #get_wtmean_vk_coords( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 145
out[[1]]
# sample group no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1 ERR260132   IGT      44       lipids   0.1154373   1.6842041
# 2 ERR260132   IGT     165     proteins   0.3787323   1.7876106
# 3 ERR260132   IGT     221 amino_sugars   0.6296017   1.7301139
# 4 ERR260132   IGT     675        carbs   0.9011834   1.7676387
# 5 ERR260132   IGT      17  cond_aromtx   0.1820605   0.8907624
# 6 ERR260132   IGT     530       lignin   0.5034537   1.2551779
# 7 ERR260132   IGT     901      tannins   0.7965590   1.3515118

names(out[[1]])
# "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 1015    6

ok <- complete.cases(df.wtcoords)
sel.ok <- which(ok==TRUE) # qty 1015

#df.wtcoords <- df.wtcoords[sel.ok, ]
#dim(df.wtcoords) # 


str(df.wtcoords)
# 'data.frame':	1015 obs. of  6 variables:
# $ sample     : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ group      : Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ no_fxns    : int  44 165 221 675 17 530 901 28 179 233 ...
# $ variable   : chr  "lipids" "proteins" "amino_sugars" "carbs" ...
# $ wtmean_OC_x: num  0.115 0.379 0.63 0.901 0.182 ...
# $ wtmean_HC_y: num  1.684 1.788 1.73 1.768 0.891 ...

saveRDS(object = df.wtcoords, file = "df.wtcoords-Forslund-T2D.RDS")


dat <- df.wtcoords

unique(dat$variable)
# "proteins"     "amino_sugars" "carbs"        "lignin"       "tannins"      "lipids"       "cond_aromtx" 

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)


names(dat) # "sample"      "group"       "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_mf"  

unique(dat$group)
# [1] IGT         T2D met neg Normal      T2D met pos
# Levels: T2D met neg < T2D met pos < IGT < Normal

dat$group_short <- factor(dat$group, 
                          levels = c("Normal", "IGT", "T2D met neg", "T2D met pos"),
                          labels = c("Normal", "IGT", "T2D Met-", "T2D Met+"),
                          ordered = TRUE)


cols <- c("Normal" = "#8dd3c7","IGT" = "#8c6bb1","T2D Met-" = "#fb8072","T2D Met+" = "#80b1d3")


#library(ggforce)

p <- ggplot(data = dat, aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group_short) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  geom_density_2d(contour_var = "ndensity", bins = 5, alpha=.5) + # 
  geom_point(alpha=0.5)+
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols, name = "Diagnosis") +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) 

p

dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-FEMALE-ONLY",this_study,header,"-v9d.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")

saveRDS(object = dat, file = paste0("dat--Weighted-vkZone-coords-FEMALE-ONLY",this_study,header,"-v9d.RDS"))

#dat <- readRDS(file = paste0("dat--Weighted-vkZone-coords-FEMALE-ONLY",this_study,header,"-v9d.RDS"))



## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat) # "sample"      "group"       "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group_short"


levels(dat$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)
# samps.class   3  0.03442 0.03925 1.9203  0.128
# Residual    141  0.84240 0.96075              
# Total       144  0.87682 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df  Sum Sq   Mean Sq      F N.Perm Pr(>F)
# Groups      3 0.01470 0.0049014 1.7124    999  0.158
# Residuals 141 0.40357 0.0028622 



## "Proteins"
this_var <- "Proteins"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)  
# samps.class   3 0.004543 0.06975 3.5239  0.013 *
# Residual    141 0.060587 0.93025                
# Total       144 0.065130 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      3 0.0005024 0.00016747 1.0888    999  0.349
# Residuals 141 0.0216867 0.00015381 



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df  SumOfSqs      R2      F Pr(>F)
# samps.class   3 0.0001966 0.02968 1.4377  0.215
# Residual    141 0.0064256 0.97032              
# Total       144 0.0066221 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      3 0.0000228 7.5985e-06 0.5638    999   0.63
# Residuals 141 0.0019003 1.3477e-05   



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs     R2      F Pr(>F)
# samps.class   3 0.000430 0.0089 0.4223  0.716
# Residual    141 0.047825 0.9911              
# Total       144 0.048255 1.0000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df   Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      3 0.000306 0.00010199 0.8359    999   0.46
# Residuals 141 0.017204 0.00012201 



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df SumOfSqs      R2      F Pr(>F)
# samps.class   3 0.000811 0.00765 0.3625  0.841
# Residual    141 0.105193 0.99235              
# Total       144 0.106004 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      3 0.0003645 0.00012151 0.5607    999  0.646
# Residuals 141 0.0305587 0.00021673 



## "Lignin"
this_var <- "Lignin"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df  SumOfSqs      R2      F Pr(>F)  
# samps.class   3 0.0003334 0.03873 1.8936   0.07 .
# Residual    141 0.0082741 0.96127                
# Total       144 0.0086074 1.00000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df     Sum Sq    Mean Sq     F N.Perm Pr(>F)
# Groups      3 0.00004238 1.4128e-05 0.899    999  0.431
# Residuals 141 0.00221585 1.5715e-05



## "Tannins"
this_var <- "Tannins"

sel <- which(dat$variable == this_var)
x <- dat[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#              Df  SumOfSqs     R2      F Pr(>F)
# samps.class   3 0.0002587 0.0359 1.7503  0.146
# Residual    141 0.0069456 0.9641              
# Total       144 0.0072043 1.0000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#            Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups      3 0.0000076 2.5310e-06 0.0782    999  0.974
# Residuals 141 0.0045657 3.2381e-05


#-------------------------

#### Forslund T2D - diff's sum rel abun at unique vK-coordinates, networks? (with vK zones) in health-disease?
#-------------------------

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Forslund-SWE-T2D.RDS")
##vk.samp.noNAs.long <- readRDS("vk.samp.noNAs.long--Forslund-T2D.RDS")

# use clean dataset

#saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long.cleaned.addIGT--Forslund-T2D.RDS")
vk.samp.noNAs.long <- readRDS("vk.samp.noNAs.long.cleaned.addIGT--Forslund-T2D.RDS")

this_study <- "-Forslund-T2D-"
header <- "-sumrelabun_at-vk-coords-"

# from earlier

str(vk.samp.noNAs.long)
# 'data.frame':	1437820 obs. of  10 variables:
# $ sample    : chr  "ERR260132" "ERR260132" "ERR260132" "ERR260132" ...
# $ fxn_id    : Factor w/ 10158 levels "fxn_10","fxn_100",..: 4627 5509 6263 7042 7621 8099 8854 9640 1 514 ...
# $ abun      : num  0 0 0 0.016 0 ...
# $ abun_type : chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x      : num  0.929 0.929 0.6 0.667 0.667 ...
# $ HC_y      : num  0.857 0.857 1.8 2.333 2.333 ...
# $ group     : Ord.factor w/ 3 levels "T2D metformin-"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ samplabel : chr  "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" "ERR260132 (ND CTRL)" ...
# $ group_new : chr  "IGT" "IGT" "IGT" "IGT" ...
# $ group_new2: Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 3 3 3 ...




# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "group_new2", "OC_x", "HC_y", "abun") ]

names(spdf)[2] # "group_new2"
names(spdf)[2] <- "group"

# add sex ... NO this study is females only

dim(spdf) #  651009      5

head(spdf)
# sample group      OC_x     HC_y         abun
# 5  ERR260132   IGT 0.6666667 2.333333 1.602096e-02
# 9  ERR260132   IGT 1.0000000 1.500000 1.287399e-03
# 10 ERR260132   IGT 0.5714286 2.000000 1.227797e-02
# 11 ERR260132   IGT 0.8000000 1.600000 1.077601e-02
# 15 ERR260132   IGT 0.7489216 1.934608 1.304738e-04
# 16 ERR260132   IGT 0.7489216 1.934608 7.629031e-05


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 145

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.83   63.33   64.86   64.43   66.02   69.46

names(spdf) # "sample" "group"  "OC_x"   "HC_y"   "abun"

temp <- spdf



table(spdf$group)
# T2D met neg T2D met pos         IGT      Normal 
#      144380       96232      210557      199840


spdf$coords <- paste0(spdf$OC_x,"__",spdf$HC_y)

unique_coords <- unique(spdf$coords) # qty 2393

temp <- spdf

head(spdf)
#       sample group      OC_x     HC_y         abun                              coords
# 5  ERR260132   IGT 0.6666667 2.333333 1.602096e-02 0.666666666666667__2.33333333333333
# 9  ERR260132   IGT 1.0000000 1.500000 1.287399e-03                              1__1.5
# 10 ERR260132   IGT 0.5714286 2.000000 1.227797e-02                0.571428571428571__2
# 11 ERR260132   IGT 0.8000000 1.600000 1.077601e-02                            0.8__1.6
# 15 ERR260132   IGT 0.7489216 1.934608 1.304738e-04 0.748921568627451__1.93460784313725
# 16 ERR260132   IGT 0.7489216 1.934608 7.629031e-05 0.748921568627451__1.93460784313725


## within a sample (i)
## at every unique coordinate
## identify duplicated coordinates within a sample and obtain total rel abun
## assign a vK compound zone
## output data table with columns: "sample" "group" "OC_x" "HC_y" "tot_relabun" "coords" "compound"

get_tot_relabun_and_compound_at_coords <- function(spdf_in, coords) {
  ###spdf_in = spdf
  ###coords = unique_coords
  ###i = 1
  
  all_samples <- unique(spdf_in$sample)
  this_samp <- all_samples[i]
  
  # cut down to sample of interest
  sel <- which(spdf$sample == this_samp)
  spdf_in <- spdf_in[sel, ]
  
  df.out <- data.frame(coords = coords, sample=this_samp, group=unique(spdf_in$group), OC_x=NA, HC_y=NA, tot_relabun=NA, compound=NA)
  
  # loop through each unique coord to sum rel abun and assign compound
  
  for (c in 1:length(coords)) {
    #c<-1
    this_coord <- coords[c] # "0.666666666666667__2.33333333333333"
    # split out x & y
    end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
    start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
    df.out$OC_x[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
    df.out$HC_y[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
    
    sel <- which(spdf_in$coords == this_coord)
    if (length(sel)>0) {
      df.out$tot_relabun[c] <- sum(spdf_in$abun[sel], na.rm = TRUE)
    } else {
      df.out$tot_relabun[c] <- 0
    }
    
    # assign compound based on coordinates
    
    OC_x <- df.out$OC_x[c]
    HC_y <- df.out$HC_y[c]
    
    # lipids
    if (OC_x >=0 & OC_x <= 0.2 & HC_y >= 1.5 & HC_y <= 2.3) {
      compd <- "lipids"
    } else if (OC_x > 0.2 & OC_x <= 0.52 & HC_y >= 1.5 & HC_y <= 2.2) {
      # proteins
      compd <- "proteins"
    } else if (OC_x > 0.52 & OC_x <= 0.7 & HC_y >= 1.5 & HC_y <= 2.2) {
      # amino_sugars
      compd <- "amino_sugars"
    } else if (OC_x > 0.7 & OC_x <= 1.1 & HC_y >= 1.5 & HC_y <= 2.4) {
      # carbs - Carbohydrates
      compd <- "carbs"
    } else if (OC_x >=0 & OC_x <= 0.25 & HC_y >= 0.5 & HC_y <= 1.25) {
      # cond_aromtx - Condensed aromatics
      compd <- "cond_aromtx"
    } else if (OC_x > 0.25 & OC_x <= 0.67 & HC_y >= 0.75 & HC_y < 1.5) {
      # lignin
      compd <- "lignin"
    } else if (OC_x > 0.67 & OC_x <= 0.97 & HC_y >= 0.53 & HC_y < 1.5) {
      # tannins
      compd <- "tannins"
    } else if ( OC_x <= 1.1 & HC_y > 2.2 ) {
      # other methylated
      compd <- "methylated"
    } else if ( OC_x > 1.1 & HC_y >= 1.5 ) {
      # other hydrated
      compd <- "hydrated"
    } else if ( OC_x > 0.97 & HC_y < 1.5 ) {
      # other demethylated
      compd <- "demethylated"
    } else if ( OC_x <= 0.97  & HC_y < 0.75 & !( OC_x ==0 & HC_y == 0 )) {
      # condensed - but exclude compounds that simultaneously contain no carbon or hydrogen
      compd <- "condensed"
    } else if ( OC_x >= 0 & OC_x <= 0.25 & HC_y > 1.25 & HC_y < 1.5) {
      # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
      compd <- "k2_b12"
    } else {
      compd <- NA
    }
    
    df.out$compound[c] <- compd
    
  } # END loop for all unique coordinates

  return(df.out)
}


# get_tot_relabun_and_compound_at_coords <- function(spdf_in, coords)

# one output table per sample

time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps), .packages = "stringr" ) %dopar%
  get_tot_relabun_and_compound_at_coords( spdf_in = spdf, coords = unique_coords)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 145
length(unique_coords) # 2393

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = out, file = "out.list.get_tot_relabun_and_compound_at_coords.T2D.RDS")
# out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.T2D.RDS")


temp <- out[[1]]
head(temp)
#                                coords    sample group      OC_x     HC_y tot_relabun     compound
# 1 0.666666666666667__2.33333333333333 ERR260132   IGT 0.6666667 2.333333  0.10385018   methylated
# 2                              1__1.5 ERR260132   IGT 1.0000000 1.500000  0.11500764        carbs
# 3                0.571428571428571__2 ERR260132   IGT 0.5714286 2.000000  0.10924295 amino_sugars
# 4                            0.8__1.6 ERR260132   IGT 0.8000000 1.600000  0.05435684        carbs
# 5 0.748921568627451__1.93460784313725 ERR260132   IGT 0.7489216 1.934608  0.02546189        carbs
# 6                            0.375__2 ERR260132   IGT 0.3750000 2.000000  0.00858266     proteins

temp <- as.data.frame( out[[1]][ ,"tot_relabun"] )
names(temp) <- unique_samps[1]  

for (i in 2:length(out)) {
  #i<-2
  new_temp <- as.data.frame( out[[i]][ ,"tot_relabun"] )
  names(new_temp) <- unique_samps[i]  
  
  temp <- cbind(temp, new_temp)
  print(paste0("comleted ",i))
}

dat <- temp
row.names(dat) <- unique_coords

rm(temp, new_temp)

str(dat)
dim(dat) # 2393  145

ok <- complete.cases(dat)
sel.ok <- which(ok==TRUE) # qty 2393


saveRDS(dat, "dat--rows-are-uniquevKcoords-cols-are-samples-T2D.RDS")
#dat <- readRDS("dat--rows-are-uniquevKcoords-cols-are-samples-T2D.RDS")



# separate data into groups

identical( names(dat) , sample_names(phy) ) # TRUE
levels(spdf$group) # "T2D met neg" "T2D met pos" "IGT" "Normal"  

groups <- character(length = length(out))
names(groups) <- paste0("name_",c(1:length(out)))
for (i in 1:length(out)) {
  #i<-1
  groups[i] <- unique( as.character( out[[i]]$group ) )
  names(groups)[i] <- unique( as.character( out[[i]]$sample ) )
  print(paste0("completed ",i))
}

table(groups)
# IGT      Normal T2D met neg T2D met pos 
# 49          43          33          20

sel <- which(groups == "IGT")
dat.IGT <- dat[ ,sel]

sel <- which(groups == "Normal")
dat.Normal <- dat[ ,sel]

sel <- which(groups == "T2D met neg")
dat.T2D_met_neg <- dat[ ,sel]

sel <- which(groups == "T2D met pos")
dat.T2D_met_pos <- dat[ ,sel]
 

# x: n x p data matrix (row/column is sample/variable) - rows are samples. columns are variables
#    n samples & p compositional variables


### Establish Networks with n=20 randomly selected subjects

# Randomly choose 20 cases from each - as n = 20 is the minimum across groups in this dataset
# Set common P-value threshold, P ≤ 0.05
# use set.seed() for sample() & cclasso!!!!




## dat.Normal

x.frac <- t(dat.Normal)
dim(x.frac) # 43 2393

set.seed(123)
subsel <- sample(x=1:dim(x.frac)[1], size = 20, replace=FALSE)
x.frac <- x.frac[subsel, ]
saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"Normal.RDS"))


ok <- complete.cases(x.frac)
sel.ok <- which(ok==TRUE)
identical(dim(x.frac)[1],length(sel.ok)) # TRUE
# only retain functional coords (analagous to taxa) that have non-zero counts in >= 60% of samples
checkNumZerosCol <- apply(x.frac,2,function(x) sum(x==0))
coordColSums <- colSums(x.frac) #

hist( coordColSums )
hist( log10( coordColSums ) )
log10( 0.1*dim(x.frac)[1] ) # 0.30103
log10( 0.01*dim(x.frac)[1] ) # -0.69897
log10( 0.001*dim(x.frac)[1] ) # -1.69897

sel1 <- which(checkNumZerosCol < 0.4*nrow(x.frac) & coordColSums >= 2)
# functional vK coords (~taxa) with non-zero counts in at least 60% of samples
# AND minimum total functional relative abundance of 2% summed across samples

keep_coords <- colnames(x.frac)[sel1] # qty 170
x.frac.nozero <- x.frac[ ,keep_coords]
hist( x.frac.nozero[] )
hist( log10(x.frac.nozero[]) )
# any zeros?
sel.zero <- which(x.frac.nozero[]==0, arr.ind = TRUE)
sel.zero # NONE for 0.1% threshold



dim(x.frac.nozero) # 20 170
getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = x.frac.nozero, file = paste0("x.frac.nozero-ColSum0.1percent-upto40percentzerosallowed-",this_study,header,"Normal.RDS"))
#x.frac.nozero <- readRDS(paste0("x.frac.nozero-ColSum0.1percent-upto40percentzerosallowed-",this_study,header,"Normal.RDS"))


# CCLasso using fraction data
time.start <- Sys.time()
set.seed(123)
res_cclasso_frac <- cclasso(x = x.frac.nozero, counts = F)
time.finish <- Sys.time()

saveRDS(object = res_cclasso_frac, file = "res_cclasso_frac-T2D-Normal.RDS")
#res_cclasso_frac <- readRDS("res_cclasso_frac-T2D-Normal.RDS")


# cclasso correlation matrix
cor_w <- res_cclasso_frac$cor_w
colnames(cor_w) <- keep_coords
rownames(cor_w) <- keep_coords
# only keep sig correlations
sel.notsig <- which(res_cclasso_frac$p_vals[] > 0.05, arr.ind = TRUE)
#sel.notsig <- which(res_cclasso_frac$p_vals[] >= 0.001, arr.ind = TRUE)
cor_w[sel.notsig] <- NA
# https://stackoverflow.com/questions/28035001/transform-correlation-matrix-into-dataframe-with-records-for-each-row-column-pai
links <- data.frame(node1=rownames(cor_w)[row(cor_w)[upper.tri(cor_w)]], 
                    node2=colnames(cor_w)[col(cor_w)[upper.tri(cor_w)]], 
                    Corr=cor_w[upper.tri(cor_w)])
links <- na.omit(links)
sig_coords_only <- unique(c(links$node1, links$node2))
length(sig_coords_only) # 48
# nodes / vertices
temp <- out[[1]] # from earlier
names(temp) # "coords"      "sample"      "group"       "OC_x"        "HC_y"        "tot_relabun" "compound"   
sel <- which(temp$coords %in% sig_coords_only)
nodes <- temp[sel, c("coords", "group", "OC_x", "HC_y", "compound")]
nodes
# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 
summary(net)
# IGRAPH 861f285 UN-- 48 96 -- 

nodes$degree <- degree(net, mode="in")

links$OC_x1<-NA
links$HC_y1<-NA
links$OC_x2<-NA
links$HC_y2<-NA
for (c in 1:length(links$node1)) {
  #c<-1
  this_coord <- links$node1[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x1[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y1[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
for (c in 1:length(links$node2)) {
  #c<-1
  this_coord <- links$node2[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x2[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y2[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
links$pos_or_neg <- NA
links$pos_or_neg[ links$Corr <0 ] <- "Negative"
links$pos_or_neg[ links$Corr >0 ] <- "Positive"


saveRDS(object = nodes , file = "nodes.Normal.T2D.n20.minColSums0.1.p0.05.RDS")
saveRDS(object = links , file = "links.Normal.T2D.n20.minColSums0.1.p0.05.RDS")
#nodes <- readRDS("nodes.Normal.T2D.n20.minColSums0.1.p0.05.RDS")
#links <- readRDS("links.Normal.T2D.n20.minColSums0.1.p0.05.RDS")

# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 
summary(net)
# IGRAPH 861f285 UN-- 48 96 -- 


# RUN ONCE
net.dat <- data.frame(group=c("Normal", "IGT", "T2D_met_neg", "T2D_met_pos"),
                      n_vertices=NA,
                      n_edges=NA,
                      edge_density=NA, # proportion of present edges from all possible edges in the network
                      frac_neg_edges=NA, # of edges present what is fraction that are negative
                      centr_degree=NA, # Centrality functions (vertex level) and $centralization functions (graph level)
                      centr_closeness=NA,
                      centr_betweenness=NA,
                      mean_distance=NA,
                      modularity=NA
                      )
net.dat$group # "Normal"  "IGT"    "Normal"      "T2D_met_neg" "T2D_met_pos"

net.dat$n_vertices[1] <- length(V(net))
net.dat$n_edges[1] <- length(E(net))
net.dat$edge_density[1] <- edge_density(net, loops=F) # proportion of present edges from all possible edges in the network
net.dat$frac_neg_edges[1] <- length(which(links$pos_or_neg == "Negative"))/length(E(net))  # of edges present what is fraction that are negative
net.dat$centr_degree[1] <- centr_degree(net, mode="in", normalized=T)$centralization # Centrality functions (vertex level) and $centralization functions (graph level)
net.dat$centr_closeness[1] <- centr_clo(net, mode="all", normalized=T)$centralization
net.dat$centr_betweenness[1] <- centr_betw(net, directed=F, normalized=T)$centralization
# Average path length
net.dat$mean_distance[1] <- mean_distance(net, directed=F)
# Community detection
ceb <- cluster_edge_betweenness(net) 
dendPlot(ceb, mode="hclust")
# Modularity
net.dat$modularity[1] <- modularity(ceb) 






## dat.IGT

x.frac <- t(dat.IGT)
dim(x.frac) # 49 2393

set.seed(123)
subsel <- sample(x=1:dim(x.frac)[1], size = 20, replace=FALSE)
x.frac <- x.frac[subsel, ]
saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"IGT.RDS"))


ok <- complete.cases(x.frac)
sel.ok <- which(ok==TRUE)
identical(dim(x.frac)[1],length(sel.ok)) # TRUE
# only retain functional coords (analagous to taxa) that have non-zero counts in >= 60% of samples
checkNumZerosCol <- apply(x.frac,2,function(x) sum(x==0))
coordColSums <- colSums(x.frac) #

hist( coordColSums )
hist( log10( coordColSums ) )
log10( 0.1*dim(x.frac)[1] ) # 0.30103
# log10( 0.01*dim(x.frac)[1] ) # -0.69897
# log10( 0.001*dim(x.frac)[1] ) # -1.69897

sel1 <- which(checkNumZerosCol < 0.4*nrow(x.frac) & coordColSums >= 2)
# functional vK coords (~taxa) with non-zero counts in at least 60% of samples
# AND minimum total functional relative abundance of 2% summed across samples

keep_coords <- colnames(x.frac)[sel1] # qty 167
x.frac.nozero <- x.frac[ ,keep_coords]
hist( x.frac.nozero[] )
hist( log10(x.frac.nozero[]) )
# any zeros?
sel.zero <- which(x.frac.nozero[]==0, arr.ind = TRUE)
sel.zero # NONE for 0.1% threshold


dim(x.frac.nozero) # 20 167
getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = x.frac.nozero, file = paste0("x.frac.nozero-ColSum0.1percent-upto40percentzerosallowed-",this_study,header,"IGT.RDS"))
#x.frac.nozero <- readRDS(paste0("x.frac.nozero-ColSum0.1percent-upto40percentzerosallowed-",this_study,header,"IGT.RDS"))


# CCLasso using fraction data
time.start <- Sys.time()
set.seed(123)
res_cclasso_frac <- cclasso(x = x.frac.nozero, counts = F)
time.finish <- Sys.time()

saveRDS(object = res_cclasso_frac, file = "res_cclasso_frac-T2D-IGT.RDS")
#res_cclasso_frac <- readRDS("res_cclasso_frac-T2D-IGT.RDS")


# cclasso correlation matrix
cor_w <- res_cclasso_frac$cor_w
colnames(cor_w) <- keep_coords
rownames(cor_w) <- keep_coords
# only keep sig correlations
sel.notsig <- which(res_cclasso_frac$p_vals[] > 0.05, arr.ind = TRUE)
#sel.notsig <- which(res_cclasso_frac$p_vals[] >= 0.001, arr.ind = TRUE)
cor_w[sel.notsig] <- NA
# https://stackoverflow.com/questions/28035001/transform-correlation-matrix-into-dataframe-with-records-for-each-row-column-pai
links <- data.frame(node1=rownames(cor_w)[row(cor_w)[upper.tri(cor_w)]], 
                    node2=colnames(cor_w)[col(cor_w)[upper.tri(cor_w)]], 
                    Corr=cor_w[upper.tri(cor_w)])
links <- na.omit(links)
sig_coords_only <- unique(c(links$node1, links$node2))
length(sig_coords_only) # 49
# nodes / vertices
temp <- out[[1]] # from earlier
names(temp) # "coords"      "sample"      "group"       "OC_x"        "HC_y"        "tot_relabun" "compound"   
sel <- which(temp$coords %in% sig_coords_only)
nodes <- temp[sel, c("coords", "group", "OC_x", "HC_y", "compound")]
nodes
# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 

nodes$degree <- degree(net, mode="in")

links$OC_x1<-NA
links$HC_y1<-NA
links$OC_x2<-NA
links$HC_y2<-NA
for (c in 1:length(links$node1)) {
  #c<-1
  this_coord <- links$node1[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x1[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y1[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
for (c in 1:length(links$node2)) {
  #c<-1
  this_coord <- links$node2[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x2[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y2[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
links$pos_or_neg <- NA
links$pos_or_neg[ links$Corr <0 ] <- "Negative"
links$pos_or_neg[ links$Corr >0 ] <- "Positive"


## UPDATE to IGT [2]
saveRDS(object = nodes , file = "nodes.IGT.T2D.n20.minColSums0.1.p0.05.RDS")
saveRDS(object = links , file = "links.IGT.T2D.n20.minColSums0.1.p0.05.RDS")
#nodes <- readRDS("nodes.IGT.T2D.n20.minColSums0.1.p0.05.RDS")
#links <- readRDS("links.IGT.T2D.n20.minColSums0.1.p0.05.RDS")

# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 
summary(net)
# IGRAPH 593dde8 UN-- 49 67 -- 


net.dat$group # "Normal"  "IGT"    "Normal"      "T2D_met_neg" "T2D_met_pos"

net.dat$n_vertices[2] <- length(V(net))
net.dat$n_edges[2] <- length(E(net))
net.dat$edge_density[2] <- edge_density(net, loops=F) # proportion of present edges from all possible edges in the network
net.dat$frac_neg_edges[2] <- length(which(links$pos_or_neg == "Negative"))/length(E(net))  # of edges present what is fraction that are negative
net.dat$centr_degree[2] <- centr_degree(net, mode="in", normalized=T)$centralization # Centrality functions (vertex level) and $centralization functions (graph level)
net.dat$centr_closeness[2] <- centr_clo(net, mode="all", normalized=T)$centralization
net.dat$centr_betweenness[2] <- centr_betw(net, directed=F, normalized=T)$centralization
# Average path length
net.dat$mean_distance[2] <- mean_distance(net, directed=F)
# Community detection
ceb <- cluster_edge_betweenness(net) 
dendPlot(ceb, mode="hclust")
# Modularity
net.dat$modularity[2] <- modularity(ceb) 






## dat.T2D_met_neg

x.frac <- t(dat.T2D_met_neg)
dim(x.frac) # 33 2393

set.seed(123)
subsel <- sample(x=1:dim(x.frac)[1], size = 20, replace=FALSE)
x.frac <- x.frac[subsel, ]
saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"T2D_met_neg.RDS"))


ok <- complete.cases(x.frac)
sel.ok <- which(ok==TRUE)
identical(dim(x.frac)[1],length(sel.ok)) # TRUE
# only retain functional coords (analagous to taxa) that have non-zero counts in >= 60% of samples
checkNumZerosCol <- apply(x.frac,2,function(x) sum(x==0))
coordColSums <- colSums(x.frac) #

hist( coordColSums )
hist( log10( coordColSums ) )
log10( 0.1*dim(x.frac)[1] ) # 0.30103
log10( 0.01*dim(x.frac)[1] ) # -0.69897
log10( 0.001*dim(x.frac)[1] ) # -1.69897

sel1 <- which(checkNumZerosCol < 0.4*nrow(x.frac) & coordColSums >= 2)
# functional vK coords (~taxa) with non-zero counts in at least 60% of samples
# AND minimum total functional relative abundance of 2% summed across samples

keep_coords <- colnames(x.frac)[sel1] # qty 173
x.frac.nozero <- x.frac[ ,keep_coords]
hist( x.frac.nozero[] )
hist( log10(x.frac.nozero[]) )
# any zeros?
sel.zero <- which(x.frac.nozero[]==0, arr.ind = TRUE)
sel.zero # NONE for 0.1% threshold



dim(x.frac.nozero) # 20 173
getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = x.frac.nozero, file = paste0("x.frac.nozero-ColSum0.1percent-upto40percentzerosallowed-",this_study,header,"T2D_met_neg.RDS"))

# CCLasso using fraction data
time.start <- Sys.time()
set.seed(123)
res_cclasso_frac <- cclasso(x = x.frac.nozero, counts = F)
time.finish <- Sys.time()

saveRDS(object = res_cclasso_frac, file = "res_cclasso_frac-T2D-T2D_met_neg.RDS")
#res_cclasso_frac <- readRDS("res_cclasso_frac-T2D-T2D_met_neg.RDS")

# cclasso correlation matrix
cor_w <- res_cclasso_frac$cor_w
colnames(cor_w) <- keep_coords
rownames(cor_w) <- keep_coords
# only keep sig correlations
sel.notsig <- which(res_cclasso_frac$p_vals[] > 0.05, arr.ind = TRUE)
#sel.notsig <- which(res_cclasso_frac$p_vals[] >= 0.001, arr.ind = TRUE)
cor_w[sel.notsig] <- NA
# https://stackoverflow.com/questions/28035001/transform-correlation-matrix-into-dataframe-with-records-for-each-row-column-pai
links <- data.frame(node1=rownames(cor_w)[row(cor_w)[upper.tri(cor_w)]], 
                    node2=colnames(cor_w)[col(cor_w)[upper.tri(cor_w)]], 
                    Corr=cor_w[upper.tri(cor_w)])
links <- na.omit(links)
sig_coords_only <- unique(c(links$node1, links$node2))
length(sig_coords_only) # 173
# nodes / vertices
temp <- out[[1]] # from earlier
names(temp) # "coords"      "sample"      "group"       "OC_x"        "HC_y"        "tot_relabun" "compound"   
sel <- which(temp$coords %in% sig_coords_only)
nodes <- temp[sel, c("coords", "group", "OC_x", "HC_y", "compound")]
nodes
# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 

nodes$degree <- degree(net, mode="in")

links$OC_x1<-NA
links$HC_y1<-NA
links$OC_x2<-NA
links$HC_y2<-NA
for (c in 1:length(links$node1)) {
  #c<-1
  this_coord <- links$node1[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x1[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y1[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
for (c in 1:length(links$node2)) {
  #c<-1
  this_coord <- links$node2[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x2[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y2[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
links$pos_or_neg <- NA
links$pos_or_neg[ links$Corr <0 ] <- "Negative"
links$pos_or_neg[ links$Corr >0 ] <- "Positive"


## UPDATE to T2D_met_neg [3]
saveRDS(object = nodes , file = "nodes.T2D_met_neg.T2D.n20.minColSums0.1.p0.05.RDS")
saveRDS(object = links , file = "links.T2D_met_neg.T2D.n20.minColSums0.1.p0.05.RDS")
#nodes <- readRDS("nodes.T2D_met_neg.T2D.n20.minColSums0.1.p0.05.RDS")
#links <- readRDS("links.T2D_met_neg.T2D.n20.minColSums0.1.p0.05.RDS")

# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 
summary(net)
# IGRAPH 91129d9 UN-- 173 4028 -- 




net.dat$group # "Normal"  "IGT"    "Normal"      "T2D_met_neg" "T2D_met_pos"

net.dat$n_vertices[3] <- length(V(net))
net.dat$n_edges[3] <- length(E(net))
net.dat$edge_density[3] <- edge_density(net, loops=F) # proportion of present edges from all possible edges in the network
net.dat$frac_neg_edges[3] <- length(which(links$pos_or_neg == "Negative"))/length(E(net))  # of edges present what is fraction that are negative
net.dat$centr_degree[3] <- centr_degree(net, mode="in", normalized=T)$centralization # Centrality functions (vertex level) and $centralization functions (graph level)
net.dat$centr_closeness[3] <- centr_clo(net, mode="all", normalized=T)$centralization
net.dat$centr_betweenness[3] <- centr_betw(net, directed=F, normalized=T)$centralization
# Average path length
net.dat$mean_distance[3] <- mean_distance(net, directed=F)
# Community detection
ceb <- cluster_edge_betweenness(net) 
dendPlot(ceb, mode="hclust")
# Modularity
net.dat$modularity[3] <- modularity(ceb) 





## dat.T2D_met_pos


x.frac <- t(dat.T2D_met_pos)
dim(x.frac) # 20 2393
saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"T2D_met_pos.RDS"))

# set.seed(123)
# subsel <- sample(x=1:dim(x.frac)[1], size = 20, replace=FALSE)
# x.frac <- x.frac[subsel, ]

ok <- complete.cases(x.frac)
sel.ok <- which(ok==TRUE)
identical(dim(x.frac)[1],length(sel.ok)) # TRUE
# only retain functional coords (analagous to taxa) that have non-zero counts in >= 60% of samples
checkNumZerosCol <- apply(x.frac,2,function(x) sum(x==0))
coordColSums <- colSums(x.frac) #

hist( coordColSums )
hist( log10( coordColSums ) )
log10( 0.1*dim(x.frac)[1] ) # 0.30103
log10( 0.01*dim(x.frac)[1] ) # -0.69897
log10( 0.001*dim(x.frac)[1] ) # -1.69897

sel1 <- which(checkNumZerosCol < 0.4*nrow(x.frac) & coordColSums >= 2)
# functional vK coords (~taxa) with non-zero counts in at least 60% of samples
# AND minimum total functional relative abundance of 2% summed across samples

keep_coords <- colnames(x.frac)[sel1] # qty 170
x.frac.nozero <- x.frac[ ,keep_coords]
hist( x.frac.nozero[] )
hist( log10(x.frac.nozero[]) )
# any zeros?
sel.zero <- which(x.frac.nozero[]==0, arr.ind = TRUE)
sel.zero # NONE for 0.1% threshold



dim(x.frac.nozero) # 20 170
getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = x.frac.nozero, file = paste0("x.frac.nozero-ColSum0.1percent-upto40percentzerosallowed-",this_study,header,"T2D_met_pos.RDS"))

# CCLasso using fraction data
time.start <- Sys.time()
set.seed(123)
res_cclasso_frac <- cclasso(x = x.frac.nozero, counts = F)
time.finish <- Sys.time()

saveRDS(object = res_cclasso_frac, file = "res_cclasso_frac-T2D-T2D_met_pos.RDS")
#res_cclasso_frac <- readRDS("res_cclasso_frac-T2D-T2D_met_pos.RDS")

# cclasso correlation matrix
cor_w <- res_cclasso_frac$cor_w
colnames(cor_w) <- keep_coords
rownames(cor_w) <- keep_coords
# only keep sig correlations
sel.notsig <- which(res_cclasso_frac$p_vals[] > 0.05, arr.ind = TRUE)
#sel.notsig <- which(res_cclasso_frac$p_vals[] >= 0.001, arr.ind = TRUE)
cor_w[sel.notsig] <- NA
# https://stackoverflow.com/questions/28035001/transform-correlation-matrix-into-dataframe-with-records-for-each-row-column-pai
links <- data.frame(node1=rownames(cor_w)[row(cor_w)[upper.tri(cor_w)]], 
                    node2=colnames(cor_w)[col(cor_w)[upper.tri(cor_w)]], 
                    Corr=cor_w[upper.tri(cor_w)])
links <- na.omit(links)
sig_coords_only <- unique(c(links$node1, links$node2))
length(sig_coords_only) # 103
# nodes / vertices
temp <- out[[1]] # from earlier
names(temp) # "coords"      "sample"      "group"       "OC_x"        "HC_y"        "tot_relabun" "compound"   
sel <- which(temp$coords %in% sig_coords_only)
nodes <- temp[sel, c("coords", "group", "OC_x", "HC_y", "compound")]
nodes
# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 

nodes$degree <- degree(net, mode="in")

links$OC_x1<-NA
links$HC_y1<-NA
links$OC_x2<-NA
links$HC_y2<-NA
for (c in 1:length(links$node1)) {
  #c<-1
  this_coord <- links$node1[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x1[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y1[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
for (c in 1:length(links$node2)) {
  #c<-1
  this_coord <- links$node2[c]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  links$OC_x2[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  links$HC_y2[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",c))
}
links$pos_or_neg <- NA
links$pos_or_neg[ links$Corr <0 ] <- "Negative"
links$pos_or_neg[ links$Corr >0 ] <- "Positive"



## UPDATE to T2D_met_pos [4]
saveRDS(object = nodes , file = "nodes.T2D_met_pos.T2D.n20.minColSums0.1.p0.05.RDS")
saveRDS(object = links , file = "links.T2D_met_pos.T2D.n20.minColSums0.1.p0.05.RDS")
#nodes <- readRDS("nodes.T2D_met_pos.T2D.n20.minColSums0.1.p0.05.RDS")
#links <- readRDS("links.T2D_met_pos.T2D.n20.minColSums0.1.p0.05.RDS")

# https://kateto.net/networks-r-igraph
net <- graph_from_data_frame(d=links, vertices=nodes, directed=FALSE) 
summary(net)
# IGRAPH 55705e7 UN-- 103 352 -- 



net.dat$group # "Normal"  "IGT"    "Normal"      "T2D_met_neg" "T2D_met_pos"

net.dat$n_vertices[4] <- length(V(net))
net.dat$n_edges[4] <- length(E(net))
net.dat$edge_density[4] <- edge_density(net, loops=F) # proportion of present edges from all possible edges in the network
net.dat$frac_neg_edges[4] <- length(which(links$pos_or_neg == "Negative"))/length(E(net))  # of edges present what is fraction that are negative
net.dat$centr_degree[4] <- centr_degree(net, mode="in", normalized=T)$centralization # Centrality functions (vertex level) and $centralization functions (graph level)
net.dat$centr_closeness[4] <- centr_clo(net, mode="all", normalized=T)$centralization
net.dat$centr_betweenness[4] <- centr_betw(net, directed=F, normalized=T)$centralization
# Average path length
net.dat$mean_distance[4] <- mean_distance(net, directed=F)
# Community detection
ceb <- cluster_edge_betweenness(net) 
dendPlot(ceb, mode="hclust")
# Modularity
net.dat$modularity[4] <- modularity(ceb) 


net.dat
#         group n_vertices n_edges edge_density frac_neg_edges centr_degree centr_closeness centr_betweenness mean_distance modularity
# 1      Normal         48      96   0.08510638      0.1041667    0.5106383       1.1098652        0.38026442      2.480706  0.2176107
# 2         IGT         49      67   0.05697279      0.3432836    0.2346939       1.3492611        0.38832445      3.621976  0.5938962
# 3 T2D_met_neg        173    4028   0.27073531      0.4813803    0.1943810       0.1647462        0.01008612      1.766568  0.2609761
# 4 T2D_met_pos        103     352   0.06700933      0.3295455    0.1878926       1.3975630        0.38688620      3.442091  0.4624508




## run each group's respective network (links, nodes) to plot below ...

# overlay vK ref data ...
vkgrouprect

p <- ggplot()+

  geom_segment(data = links, aes(x = OC_x1, y = HC_y1, xend = OC_x2,  yend = HC_y2, color = pos_or_neg, size = abs(Corr)) )+
  geom_point(data = nodes, aes(x = OC_x, y = HC_y, size = degree))+ # , color = compound

  #xlim(0,1.5)+ ylim(0,3.1)+
  #xlim(0,2.1)+ ylim(0,3.1)+

  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #https://ggplot2-book.org/scale-colour.html#:~:text=11.3.&text=scale_colour_brewer()%20is%20a%20discrete,http%3A%2F%2Fcolorbrewer2.org%2F.

  theme_bw()+

  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#525252", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  #geom_point(data = df.hac ,aes(x=OC_x, y=HC_y), color = "#253494")+ # , pch=8  "#80b1d3"
  #geom_text(data = df.hac ,aes(x=OC_x, y=HC_y, label=sample), color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#525252") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#525252") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#525252") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#525252") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#525252") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#525252" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#525252") + # bottom-left

  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#525252", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#525252", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#525252", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#525252", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#525252", lineheight = 0.8) +

  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#525252")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#525252")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#525252")+

  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank() ,
    
    legend.position = "none"

  )
p

# All figs are: (n=20; CCLasso P≤0.05)

grid.text(label = "(a)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
grid.text(label = "Normal (F)", x = unit(0.095, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=12), hjust=0 )
#dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-Normal--v9.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-Normal--v9.tiff"), width = 12, height = 9, units = "cm", res=450, compression="lzw",type="cairo")

grid.text(label = "(b)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
grid.text(label = "IGT (F)", x = unit(0.095, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=12), hjust=0 )
#dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-IGT--v9.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-IGT--v9.tiff"), width = 12, height = 9, units = "cm", res=450, compression="lzw",type="cairo")

grid.text(label = "(c)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
grid.text(label = "T2D Met- (F)", x = unit(0.095, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=12), hjust=0 )
#dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-T2D_met_neg--v9.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-T2D_met_neg--v9.tiff"), width = 12, height = 9, units = "cm", res=450, compression="lzw",type="cairo")

grid.text(label = "(d)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
grid.text(label = "T2D Met+ (F)", x = unit(0.095, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=12), hjust=0 )
#dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-T2D_met_pos--v9.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","vK-Network-based-on",header,this_study,"-T2D_met_pos--v9.tiff"), width = 12, height = 9, units = "cm", res=450, compression="lzw",type="cairo")



## compare each group to bootstrap (B = 5000) randomized distribution of network properties?
## choose 20 from 80 samples

## representative subject data
net.dat
#         group n_vertices n_edges edge_density frac_neg_edges centr_degree centr_closeness centr_betweenness mean_distance modularity
# 1      Normal         48      96   0.08510638      0.1041667    0.5106383       1.1098652        0.38026442      2.480706  0.2176107
# 2         IGT         49      67   0.05697279      0.3432836    0.2346939       1.3492611        0.38832445      3.621976  0.5938962
# 3 T2D_met_neg        173    4028   0.27073531      0.4813803    0.1943810       0.1647462        0.01008612      1.766568  0.2609761
# 4 T2D_met_pos        103     352   0.06700933      0.3295455    0.1878926       1.3975630        0.38688620      3.442091  0.4624508

write.csv(net.dat, "net.dat.T2D.csv")


## pool the 4 groups 

# saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"Normal.RDS"))
# saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"IGT.RDS"))
# saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"T2D_met_neg.RDS"))
# saveRDS(object = x.frac, file = paste0("x.frac.subsel20-",this_study,header,"T2D_met_pos.RDS")) # # T2D Met pos - only had 20 subjects to start with

temp1 <- readRDS(paste0("x.frac.subsel20-",this_study,header,"Normal.RDS"))
dim(temp1) # 20 2393
temp2 <- readRDS(paste0("x.frac.subsel20-",this_study,header,"IGT.RDS"))
dim(temp2) # 20 2393
temp3 <- readRDS(paste0("x.frac.subsel20-",this_study,header,"T2D_met_neg.RDS"))
dim(temp3) # 20 2393
temp4 <- readRDS(paste0("x.frac.subsel20-",this_study,header,"T2D_met_pos.RDS")) # # T2D Met pos - only had 20 subjects to start with
dim(temp4) # 20 2393

pool <- rbind(temp1,temp2,temp3,temp4)
dim(pool) # 80 2393
rm(temp1,temp2,temp3,temp4)

pool[1:5,1:5]
# 0.666666666666667__2.33333333333333    1__1.5 0.571428571428571__2   0.8__1.6
# ERR260252                          0.12144369 0.1793630           0.08445007 0.03783438
# ERR260218                          0.03602052 0.2106417           0.07008340 0.02270859
# ERR260217                          0.08596632 0.1693407           0.10475796 0.05140700
# ERR260163                          0.12290141 0.1373085           0.09491047 0.03322455
# ERR260260                          0.07974206 0.1662958           0.09439474 0.04522955
# 0.748921568627451__1.93460784313725
# ERR260252                          0.04670911
# ERR260218                          0.05794605
# ERR260217                          0.02246356
# ERR260163                          0.02940225
# ERR260260                          0.03669225


boot_net_statistics <- function(pool_in) { # needs: zCompositions
  ###pool_in = pool
  ###i = 1
  
  
  #### cclasso.R function
  #-------------------------
  # https://github.com/huayingfang/CCLasso/blob/master/R/cclasso.R
  # File: cclasso.R
  # Aim : Correlation inference for compositional data through lasso
  # - - - - - - - - - - - - - - - - - - - - - 
  # Author : Fang Huaying (Peking University)
  # Email  : hyfang@pku.edu.cn
  # Date   : 2016-01-08
  # Version: 2.0
  # - - - - - - - - - - - - - - - - - - - - - 
  # Main function: cclasso(x, counts = FALSE, pseudo = 0.5, k_cv = 3, 
  #                        lam_int = c(1e-4, 1), k_max = 20, n_boot = 20) 
  #
  #  Input:
  #           x ------ n x p data matrix (row/column is sample/variable)
  #                    n samples & p compositional variables
  #      counts ------ Is the compositional data matrix a count matrix? 
  #                    Default: FALSE
  #      pseudo ------ pseudo count if counts = TRUE
  #                    Default: 0.5
  #        k_cv ------ folds of cross validation
  #                    Default: 3     
  #     lam_int ------ tuning parameter interval
  #                    Default: [1e-4, 1]
  #       k_max ------ maximum iterations for golden section method
  #                    Default: 20
  #      n_boot ------ Bootstrap times
  #                    Default: 20
  #  Output: 
  #      A list structure contains:
  #       var_w ------ variance estimation
  #       cor_w ------ correlation estimation
  #      p_vals ------ p-values for elements of cor_w equal 0 or not
  #      lambda ------ final tuning parameter
  #     info_cv ------ information for cross validation
  # - - - - - - - - - - - - - - - - - - - - - 
  cclasso <- function(x, counts = FALSE, pseudo = 0.5, k_cv = 3, 
                      lam_int = c(1e-4, 1), k_max = 20, n_boot = 20) {
    n <- nrow(x);
    p <- ncol(x);
    
    if(counts) {
      x <- x + pseudo;
      x <- x / rowSums(x);
    }
    x <- log(x);
    vx2 <- stats::var(x);
    
    # Diagonal weight for loss function
    rmean_vx2 <- rowMeans(vx2);
    wd <- 1/diag(vx2 - rmean_vx2 - rep(rmean_vx2, each = p) + mean(rmean_vx2));
    wd2 <- sqrt(wd);
    
    # Some global parameters for optimization with single lambda
    rho <- 1;
    u_f <- eigen(diag(p) - 1/p)$vectors;
    wd_u <- (t(u_f) %*% (wd * u_f))[-p, -p];
    wd_u_eig <- eigen(wd_u);
    d0_wd <- 1 / ( (rep(wd_u_eig$values, each = p-1) + wd_u_eig$values) / 
                     (2 * rho) + 1 );
    u0_wd <- wd_u_eig$vectors;
    
    # Golden section method for the selection of lambda (log10 scale)
    sigma <- vx2;
    lam_int2 <- log10(range(lam_int));
    a1 <- lam_int2[1]; 
    b1 <- lam_int2[2];
    # Store lambda and corresponding cross validation's loss
    lams <- NULL; 
    fvals <- NULL;
    # Two trial points in first 
    a2 <- a1 + 0.382 * (b1 - a1); 
    b2 <- a1 + 0.618 * (b1 - a1);
    fb2 <- cv_loss_cclasso(lambda2 = 10^b2 / rho, x = x, k_cv = k_cv, 
                           sigma = sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                           wd2 = wd2);
    lams <- c(lams, b2); 
    fvals <- c(fvals, fb2$cv_loss);
    fa2 <- cv_loss_cclasso(lambda2 = 10^a2 / rho, x = x, k_cv = k_cv, 
                           sigma = fb2$sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                           wd2 = wd2);
    lams <- c(lams, a2); 
    fvals <- c(fvals, fa2$cv_loss);
    # Error tolerance for convergence
    err_lam2 <- 1e-1 * max(1, lam_int2);
    err_fval <- 1e-4;
    
    err <- b1 - a1;
    k <- 0;
    while(err > err_lam2 && k < k_max) {
      fval_max <- max(fa2$cv_loss, fb2$cv_loss);
      
      if(fa2$cv_loss > fb2$cv_loss) {
        a1 <- a2;      
        a2 <- b2;
        fa2 <- fb2;
        b2 <- a1 + 0.618 * (b1 - a1);
        fb2 <- cv_loss_cclasso(lambda2 = 10^b2 / rho, x = x, k_cv = k_cv, 
                               sigma = fa2$sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                               wd2 = wd2);
        
        lams <- c(lams, b2); 
        fvals <- c(fvals, fb2$cv_loss);
      } else {
        b1 <- b2;
        b2 <- a2;
        fb2 <- fa2;
        a2 <- a1 + 0.382 * (b1 - a1);
        fa2 <- cv_loss_cclasso(lambda2 = 10^a2 / rho, x = x, k_cv = k_cv, 
                               sigma = fb2$sigma, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd, 
                               wd2 = wd2);
        
        lams <- c(lams, a2); 
        fvals <- c(fvals, fa2$cv_loss);
      }
      fval_min <- min(fa2$cv_loss, fb2$cv_loss);      
      
      k <- k + 1;
      err <- b1 - a1;
      if(abs(fval_max - fval_min) / (1 + fval_min) <= err_fval) {
        break;
      }
    }
    info_cv <- list(lams = lams, fvals = fvals, k = k + 2, 
                    lam_int = 10^c(a1, b1)); 
    if(a1 == lam_int2[1] || b1 == lam_int2[2]) {
      cat("WARNING:\n", "\tOptimal lambda is near boundary! ([", 10^a1, ",", 
          10^b1, "])\n", sep = "");
    }
    
    lambda <- 10^((a2 + b2)/2);
    # Bootstrap for cclasso
    lambda2 <- lambda / rho;
    info_boot <- boot_cclasso(x = x, sigma = fb2$sigma, lambda2 = lambda2, 
                              n_boot = n_boot, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
    
    return(list(var_w = info_boot$var_w, cor_w = info_boot$cor_w, 
                p_vals = info_boot$p_vals, lambda = lambda, info_cv = info_cv));
  }
  # - - - - - - - - - - - - - - - - - - - - - 
  # Bootstrap for cclasso
  boot_cclasso <- function(x, sigma, lambda2, n_boot = 20, 
                           wd, u_f, u0_wd, d0_wd) {
    n <- nrow(x);
    p <- ncol(x);
    
    # Store the result of bootstrap
    cors_boot <- matrix(0, nrow = p * (p - 1)/2, ncol = n_boot + 1);
    vars_boot <- matrix(0, nrow = p, ncol = n_boot + 1);
    cors_mat <- matrix(0, p, p);
    ind_low <- lower.tri(cors_mat);
    
    # Bootstrap procedure
    sam_boot <- matrix(sample(1:n, size = n * n_boot, replace = T), 
                       ncol = n_boot);
    for(k in 1:n_boot) {
      ind_samp <- sam_boot[, k];
      sigma2 <- cclasso_sub(sigma = sigma, vx = var(x[ind_samp, ]), 
                            lambda2 = lambda2, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
      
      vars_boot[, k] <- diag(sigma2);
      Is <- 1 / sqrt(vars_boot[, k]);
      cors_mat <- Is * sigma2 * rep(Is, each = p);
      cors_boot[, k] <- cors_mat[ind_low];
    }
    Is <- 1 / sqrt(diag(sigma));
    cors_mat <- sigma * Is * rep(Is, each = p);
    cors_boot[, n_boot + 1] <- cors_mat[ind_low];
    vars_boot[, n_boot + 1] <- diag(sigma);
    
    # - - - - - - - - - - - - - - - - - - - - - 
    # Variance estimation via bootstrap
    vars2 <- rowMeans(vars_boot);
    # - - - - - - - - - - - - - - - - - - - - - 
    # Correlations' relationship for artificial null sample
    tol_cor <- 1e-3;
    sam_art0 <- matrix(rnorm(n * p), nrow = n) * rep(sqrt(vars2), each = n);
    cors_art0 <- cor(sam_art0)[ind_low];
    sam_art <- sam_art0 - log(rowSums(exp(sam_art0)));
    sigma_art <- cclasso_sub(sigma = sigma, vx = var(sam_art), 
                             lambda2 = lambda2, wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
    Is <- 1 / sqrt(diag(sigma_art));
    cors_mat <- Is * sigma_art * rep(Is, each = p);
    cors_art2 <- cors_mat[ind_low];
    # Bias of estimation between absolute data and cclasso of compositional data
    cors0m2 <- log( ((1 + cors_art0) * (1 - cors_art2)) / ((1 + cors_art2) * 
                                                             (1 - cors_art0)) );
    tmp <- abs(cors_art2) >= tol_cor;
    bias02 <- ifelse(sum(tmp), median(abs(cors0m2)[tmp]), 0);
    # Modification of estimation for cclasso
    cors2 <- log( (1 + cors_boot) / (1 - cors_boot) );    
    cors2mod <- (cors_boot >= tol_cor) * (cors2 + bias02) + 
      (cors_boot <= -tol_cor) * (cors2 - bias02);
    cors2mod <- 1 - rowMeans(2 / (exp(cors2mod) + 1));
    cors2_mat <- diag(p);
    cors2_mat[ind_low] <- cors2mod;
    cors2_mat <- t(cors2_mat);
    cors2_mat[ind_low] <- cors2mod;
    # P-values with null distribution of correlation estimations of absolute data
    p_vals <- pt(cors2mod * sqrt((n - 2) / (1 - cors2mod^2)), df = n - 2);
    p_vals <- ifelse(p_vals <= 0.5, p_vals, 1 - p_vals);
    pval_mat <- diag(p);
    pval_mat[ind_low] <- p_vals;
    pval_mat <- t(pval_mat);
    pval_mat[ind_low] <- p_vals;
    # - - - - - - - - - - - - - - - - - - - - - 
    
    return(list(var_w = vars2, cor_w = cors2_mat, p_vals = pval_mat));    
  }
  # - - - - - - - - - - - - - - - - - - - - - 
  # cross validation's loss of cclasso for single lambda
  cv_loss_cclasso <- function(lambda2, x, k_cv, sigma, 
                              wd, u_f, u0_wd, d0_wd, wd2) {
    n <- nrow(x);
    p <- ncol(x);
    
    n_b <- floor(n / k_cv);
    cv_loss <- 0;
    for(k in 1:k_cv) {
      itest <- (n_b * (k - 1) + 1):(n_b * k);
      vxk <- stats::var(x[itest, ]);
      vx2k <- stats::var(x[-itest, ]);
      
      sigma <- cclasso_sub(sigma = sigma, vx = vx2k, lambda2 = lambda2, 
                           wd = wd, u_f = u_f, u0_wd = u0_wd, d0_wd = d0_wd);
      
      dsig <- sigma - vxk;
      tmp <- rowMeans(dsig);
      dsig <- dsig - tmp - rep(tmp, each = p) + mean(tmp);
      cv_loss <- cv_loss + base::norm(wd2 * dsig, "F")^2;
    }
    
    return(list(cv_loss = cv_loss, sigma = sigma));
  }
  # - - - - - - - - - - - - - - - - - - - - - 
  # cclasso for single lambda
  cclasso_sub <- function(sigma, vx, lambda2, 
                          wd, u_f, u0_wd, d0_wd, 
                          k_max = 200, x_tol = 1e-4) {
    p <- ncol(sigma);
    sigma2 <- sigma;
    LAMBDA <- matrix(0, p, p);
    lambda2 <- matrix(lambda2, p, p);
    diag(lambda2) <- 0;
    
    k <- 0;
    err <- 1;
    while(err > x_tol && k < k_max) {
      # Update sigma
      x_sigma <- t(u_f) %*% ((sigma2 - vx) - LAMBDA) %*% u_f;
      x_sigma[-p,-p] <- u0_wd %*% ((t(u0_wd) %*% x_sigma[-p, -p] %*% u0_wd) * 
                                     d0_wd) %*% t(u0_wd);
      sigma_new <- vx + u_f %*% x_sigma %*% t(u_f);
      # Update sigma2
      A <- LAMBDA + sigma_new;
      sigma2_new <- (A > lambda2) * (A - lambda2) + (A < -lambda2) * 
        (A + lambda2);
      # Update Lambda
      LAMBDA <- LAMBDA + (sigma_new - sigma2_new);
      
      err <- max( abs(sigma_new - sigma)/(abs(sigma) + 1), 
                  abs(sigma2_new - sigma2)/(abs(sigma2) + 1) ); 
      k <- k + 1;
      sigma <- sigma_new;
      sigma2 <- sigma2_new;
    }
    
    if(k >= k_max) {
      cat("WARNING of cclasso_sub:\n", "\tMaximum Iteration:", k_max, 
          "&& Relative error:", err, "!\n");
    }
    
    return(sigma);
  }
  # - - - - - - - - - - - - - - - - - - - - - 
  
  #-------------------------
  
  
  set.seed(123+i)
  subsel <- sample(x=1:dim(pool_in)[1], size = 20, replace=FALSE)
  x.frac <- pool[subsel, ]
  ok <- complete.cases(x.frac)
  sel.ok <- which(ok==TRUE)
  x.frac <- x.frac[sel.ok, ]
  checkNumZerosCol <- apply(x.frac,2,function(x) sum(x==0))
  coordColSums <- colSums(x.frac)
  
  sel1 <- which(checkNumZerosCol < 0.4*nrow(x.frac) & coordColSums >= 2)
  # functional vK coords (~taxa) with non-zero counts in at least 60% of samples
  # AND minimum total functional relative abundance of 2% summed across samples
  
  keep_coords <- colnames(x.frac)[sel1]
  x.frac.nozero <- x.frac[ ,keep_coords]
  
  # any zeros?
  sel.zero <- which(x.frac.nozero[]==0, arr.ind = TRUE)
  if (length(sel.zero)>0) {
    # replace zeros with a small value - using package zCompositions
    # 'prop' imputed proportions
    set.seed(123)
    x.frac.nozero <- cmultRepl(x.frac.nozero, output = "prop")
  }
  
  # CCLasso using fraction data
  set.seed(123)
  res_cclasso_frac <- cclasso(x = x.frac.nozero, counts = F)
  
  # cclasso correlation matrix
  cor_w <- res_cclasso_frac$cor_w
  colnames(cor_w) <- keep_coords
  rownames(cor_w) <- keep_coords
  # only keep sig correlations
  sel.notsig <- which(res_cclasso_frac$p_vals[] > 0.05, arr.ind = TRUE)
  cor_w[sel.notsig] <- NA
  # https://stackoverflow.com/questions/28035001/transform-correlation-matrix-into-dataframe-with-records-for-each-row-column-pai
  links <- data.frame(node1=rownames(cor_w)[row(cor_w)[upper.tri(cor_w)]], 
                      node2=colnames(cor_w)[col(cor_w)[upper.tri(cor_w)]], 
                      Corr=cor_w[upper.tri(cor_w)])
  links <- na.omit(links)
  sig_coords_only <- unique(c(links$node1, links$node2))
  length(sig_coords_only) # 175
  # nodes / vertices
  
  # https://kateto.net/networks-r-igraph
  net <- graph_from_data_frame(d=links, vertices=NULL, directed=FALSE) # don't include vertex metadata
  
  links$pos_or_neg <- NA
  links$pos_or_neg[ links$Corr <0 ] <- "Negative"
  links$pos_or_neg[ links$Corr >0 ] <- "Positive"
  
  net.boot.out <- data.frame(boot_i = i,
                        n_vertices=NA,
                        n_edges=NA,
                        edge_density=NA, # proportion of present edges from all possible edges in the network
                        frac_neg_edges=NA, # of edges present what is fraction that are negative
                        centr_degree=NA, # Centrality functions (vertex level) and $centralization functions (graph level)
                        centr_closeness=NA,
                        centr_betweenness=NA,
                        mean_distance=NA,
                        modularity=NA
  )
  
  net.boot.out$n_vertices[1] <- length(V(net))
  net.boot.out$n_edges[1] <- length(E(net))
  net.boot.out$edge_density[1] <- edge_density(net, loops=F) # proportion of present edges from all possible edges in the network
  net.boot.out$frac_neg_edges[1] <- length(which(links$pos_or_neg == "Negative"))/length(E(net))  # of edges present what is fraction that are negative
  net.boot.out$centr_degree[1] <- centr_degree(net, mode="in", normalized=T)$centralization # Centrality functions (vertex level) and $centralization functions (graph level)
  net.boot.out$centr_closeness[1] <- centr_clo(net, mode="all", normalized=T)$centralization
  net.boot.out$centr_betweenness[1] <- centr_betw(net, directed=F, normalized=T)$centralization
  # Average path length
  net.boot.out$mean_distance[1] <- mean_distance(net, directed=F)
  # Community detection
  ceb <- cluster_edge_betweenness(net) 
  # Modularity
  net.boot.out$modularity[1] <- modularity(ceb) 
  
  #return(net.boot.out)
  saveRDS(net.boot.out, paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/boot-t2d-net/b_out_",i,".RDS"))
  
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

#b_out <- foreach(i=1:5000, .packages = c("igraph","zCompositions") ) %dopar%

#foreach(i=1:100, .packages = c("igraph","zCompositions") ) %dopar%
foreach(i=101:1000, .packages = c("igraph","zCompositions") ) %dopar%
  boot_net_statistics( pool_in = pool )
  
stopCluster(cl)
time.finish <- Sys.time()




time.start #  "2023-08-01 09:31:08 ACST"
time.finish # "2023-08-01 10:38:54 ACST"
# 1 hr per 100

time.start #  "2023-08-01 18:20:29 ACST"
time.finish # "2023-08-02 04:09:28 ACST"
# 10 hr per 900


## assemble results


boot_t2d_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/boot-t2d-net"


# read first output
i<-1
temp <- readRDS(paste0(boot_t2d_dir,"/b_out_",i,".RDS"))
df.out <- temp

num_results_files <- 1000

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(boot_t2d_dir,"/b_out_",i,".RDS"))
  df.out <- rbind(df.out, temp)
  print(paste0("added results ",i," of ",num_results_files ))
}

dim(df.out) # 1000 10

names(df.out)
# [1] "boot_i"            "n_vertices"        "n_edges"           "edge_density"      "frac_neg_edges"    "centr_degree"      "centr_closeness"  
# [8] "centr_betweenness" "mean_distance"     "modularity"

df.out.melt <- melt(df.out, id.vars = "boot_i")

names(net.dat)
# [1] "group"             "n_vertices"        "n_edges"           "edge_density"      "frac_neg_edges"    "centr_degree"      "centr_closeness"  
# [8] "centr_betweenness" "mean_distance"     "modularity"

net.dat.melt <- melt(net.dat, id.vars = "group")



p <- ggplot()+
  geom_density(data= df.out.melt,aes(x = value))+
  geom_vline(data=net.dat.melt, aes(xintercept = value, color = group))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)
p

p <- ggplot()+
  geom_density(data= df.out.melt,aes(x = value))+
  geom_vline(data=net.dat.melt, aes(xintercept = value, color = group))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)
p


# https://stackoverflow.com/questions/41971150/add-vline-to-geom-density-and-shade-confidence-interval-of-mean-r


## build dataframes for 2.5th and 97.5th percentile of density distributions

( unique_variables <- unique(df.out.melt$variable) )
# [1] n_vertices        n_edges           edge_density      frac_neg_edges    centr_degree      centr_closeness   centr_betweenness
# [8] mean_distance     modularity       
# Levels: n_vertices n_edges edge_density frac_neg_edges centr_degree centr_closeness centr_betweenness mean_distance modularity

dat.2.5.perc <- data.frame(variable=NA, xmin=NA, xmax=NA, ymin=NA, ymax=NA)
for (i in 1:length(unique_variables)) {
  #i<-6
  this_var <- unique_variables[i]
  values = na.omit( df.out.melt$value[ which(df.out.melt$variable == this_var) ] )
  values_range = range(values, na.rm=TRUE)
  density_data = StatDensity$compute_group(
    data.frame(x=values),
    scales=list(
      x=scale_x_continuous(limits = values_range)
    )
  )
  min_x = min(values)
  low_2_5_perc = quantile(values, probs = 0.025)
  x_values = seq(min_x, low_2_5_perc, length.out=resolution)
  temp <- data.frame(
    variable = this_var,
    xmin=x_values,
    xmax=x_values,
    ymin=rep(0, resolution),
    ymax=approx(density_data$x, density_data$density, xout=x_values)$y
  )
  dat.2.5.perc <- rbind(dat.2.5.perc, temp)
  print(paste0(this_var,": min_value = ",round(min_x,4),"; 2.5th-percentile = ", round(low_2_5_perc,4)))
}
head(dat.2.5.perc)
# remove NA first row
dat.2.5.perc <- dat.2.5.perc[-1, ]

# [1] "n_vertices: min_value = 2; 2.5th-percentile = 17"
# [1] "n_edges: min_value = 1; 2.5th-percentile = 13"
# [1] "edge_density: min_value = 0.0311; 2.5th-percentile = 0.0404"
# [1] "frac_neg_edges: min_value = 0; 2.5th-percentile = 0.1291"
# [1] "centr_degree: min_value = 0; 2.5th-percentile = 0.1075"
# [1] "centr_closeness: min_value = 0; 2.5th-percentile = 0.1867"
# [1] "centr_betweenness: min_value = 0; 2.5th-percentile = 0.0098"
# [1] "mean_distance: min_value = 1; 2.5th-percentile = 1.6674"
# [1] "modularity: min_value = 0; 2.5th-percentile = 0.0645"




dat.97.5.perc <- data.frame(variable=NA, xmin=NA, xmax=NA, ymin=NA, ymax=NA)
for (i in 1:length(unique_variables)) {
  #i<-1
  this_var <- unique_variables[i]
  values = na.omit( df.out.melt$value[ which(df.out.melt$variable == this_var) ] )
  values_range = range(values, na.rm=TRUE)
  density_data = StatDensity$compute_group(
    data.frame(x=values),
    scales=list(
      x=scale_x_continuous(limits = values_range)
    )
  )
  #min_x = min(values)
  hi_97_5_perc = quantile(values, probs = 0.975)
  max_x = max(values)
  x_values = seq(hi_97_5_perc, max_x, length.out=resolution)
  temp <- data.frame(
    variable = this_var,
    xmin=x_values,
    xmax=x_values,
    ymin=rep(0, resolution),
    ymax=approx(density_data$x, density_data$density, xout=x_values)$y
  )
  dat.97.5.perc <- rbind(dat.97.5.perc, temp)
  print(paste0(this_var,": 97.5th-percentile = ", round(hi_97_5_perc,4),"; max_value = ",round(max_x,4)))
}
head(dat.97.5.perc)
# remove NA first row
dat.97.5.perc <- dat.97.5.perc[-1, ]

# [1] "n_vertices: 97.5th-percentile = 174; max_value = 177"
# [1] "n_edges: 97.5th-percentile = 4524.4; max_value = 6157"
# [1] "edge_density: 97.5th-percentile = 0.3194; max_value = 1"
# [1] "frac_neg_edges: 97.5th-percentile = 0.4971; max_value = 1"
# [1] "centr_degree: 97.5th-percentile = 0.4656; max_value = 0.6916"
# [1] "centr_closeness: 97.5th-percentile = 1.4243; max_value = 1.6071"
# [1] "centr_betweenness: 97.5th-percentile = 0.5352; max_value = 0.8992"
# [1] "mean_distance: 97.5th-percentile = 4.1135; max_value = 5.6503"
# [1] "modularity: 97.5th-percentile = 0.7189; max_value = 0.7853"
  
  

# ROUGH
p <- ggplot(data= df.out.melt,aes(x = value))+
  #geom_rect(data = dat.2.5.perc, mapping = aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax), fill = "yellow")+
  geom_segment(data = dat.2.5.perc, mapping = aes(x=xmin, xend=xmax, y=ymin, yend=ymax), colour = "#ffff99", linewidth=1)+
  geom_segment(data = dat.97.5.perc, mapping = aes(x=xmin, xend=xmax, y=ymin, yend=ymax), colour = "#ffff99", linewidth=1)+
  geom_density()+
  geom_vline(data=net.dat.melt, aes(xintercept = value, color = group))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)
p


df.out.melt$variable <- factor(df.out.melt$variable,
                               levels = c("n_vertices", "n_edges", "edge_density", "frac_neg_edges",
                                          "centr_degree", "centr_closeness", "centr_betweenness",
                                          "mean_distance", "modularity"),
                               labels = c("No.of vertices", "No. of edges", "Edge density", "Fraction of\nnegative edges",
                                          "Degree\ncentralization", "Closeness\ncentralization", "Betweenness\ncentralization",
                                          "Mean distance", "Modularity"),
                               ordered=TRUE)

net.dat.melt$variable <- factor(net.dat.melt$variable,
                                levels = c("n_vertices", "n_edges", "edge_density", "frac_neg_edges",
                                           "centr_degree", "centr_closeness", "centr_betweenness",
                                           "mean_distance", "modularity"),
                                labels = c("No.of vertices", "No. of edges", "Edge density", "Fraction of\nnegative edges",
                                           "Degree\ncentralization", "Closeness\ncentralization", "Betweenness\ncentralization",
                                           "Mean distance", "Modularity"),
                                ordered=TRUE)

unique(net.dat.melt$group)
# "Normal"      "IGT"         "T2D_met_neg" "T2D_met_pos"
net.dat.melt$group <- factor(net.dat.melt$group,
                             levels = c("Normal", "IGT", "T2D_met_neg", "T2D_met_pos"),
                             labels = c("Normal","IGT","T2D Met-","T2D Met+"),
                             ordered=TRUE)
net.dat.melt$group <- factor(net.dat.melt$group,
                             #levels = c("Normal", "IGT", "T2D_met_neg", "T2D_met_pos"),
                             levels = c("Normal","IGT","T2D Met-","T2D Met+"),
                             labels = c("Normal (F)","IGT (F)","T2D Met- (F)","T2D Met+ (F)"),
                             ordered=TRUE)

# group colours
#cols <- c("Normal" = "#8dd3c7","IGT" = "#8c6bb1","T2D Met-" = "#fb8072","T2D Met+" = "#80b1d3")
cols <- c("Normal (F)" = "#8dd3c7","IGT (F)" = "#8c6bb1","T2D Met- (F)" = "#fb8072","T2D Met+ (F)" = "#80b1d3")

dat.2.5.perc$variable <- factor(dat.2.5.perc$variable,
                                levels = c("n_vertices", "n_edges", "edge_density", "frac_neg_edges",
                                           "centr_degree", "centr_closeness", "centr_betweenness",
                                           "mean_distance", "modularity"),
                                labels = c("No.of vertices", "No. of edges", "Edge density", "Fraction of\nnegative edges",
                                           "Degree\ncentralization", "Closeness\ncentralization", "Betweenness\ncentralization",
                                           "Mean distance", "Modularity"),
                                ordered=TRUE)

dat.97.5.perc$variable <- factor(dat.97.5.perc$variable,
                                levels = c("n_vertices", "n_edges", "edge_density", "frac_neg_edges",
                                           "centr_degree", "centr_closeness", "centr_betweenness",
                                           "mean_distance", "modularity"),
                                labels = c("No.of vertices", "No. of edges", "Edge density", "Fraction of\nnegative edges",
                                           "Degree\ncentralization", "Closeness\ncentralization", "Betweenness\ncentralization",
                                           "Mean distance", "Modularity"),
                                ordered=TRUE)



p <- ggplot(data= df.out.melt,aes(x = value))+
  #geom_rect(data = dat.2.5.perc, mapping = aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax), fill = "yellow")+
  geom_segment(data = dat.2.5.perc, mapping = aes(x=xmin, xend=xmax, y=ymin, yend=ymax), colour = "#ffff99", linewidth=1)+
  geom_segment(data = dat.97.5.perc, mapping = aes(x=xmin, xend=xmax, y=ymin, yend=ymax), colour = "#ffff99", linewidth=1)+
  geom_density()+
  geom_vline(data=net.dat.melt, aes(xintercept = value, color = group), linewidth = 1, linetype="twodash")+
  scale_color_manual(values = cols, name = "Group")+
  theme_classic()+
  xlab("Network statistic value")+ ylab("Density")+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)+
  theme(
    #legend.position = "bottom",
    legend.margin=margin(0,0,0,0), # trbl
    legend.box.margin=margin(-5,0,0,0), 
    legend.title = element_text(size = rel(0.9)),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 1,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) #,
    #panel.grid.major = element_blank(),
    #panel.grid.minor = element_blank(), 
    #axis.text.x = element_text(size = rel(1.1))
    )
p

#grid.text(label = "(a)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","T2D-Network-cf-Bootstrap-Random-Graphs-",this_study,header,"-v9d.tiff"), width = 20, height = 14, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Forslund T2D - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-Forslund-SWE-T2D.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 19099 taxa and 145 samples ]
# sample_data() Sample Data:       [ 145 samples by 5 sample variables ]
# tax_table()   Taxonomy Table:    [ 19099 taxa by 4 taxonomic ranks ]

df.zones <- readRDS("df.zones--Forslund-T2D.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Forslund-T2D.RDS")

samps <- unique(df.zones$sample) # qty 145
# phy <- prune_samples( samples = samps, phy )
# # prune taxa with zero reads
# phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
# phy


head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 145
NonZeroFxns


mean(NonZeroFxns) # 7801.034
sd(NonZeroFxns) # 1787.25
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 9916
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 2393
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 4489.717
sd(fxns_converted) # 874.0584

mean(fxns_relabun_converted) # 64.42862
sd(fxns_relabun_converted) # 2.178273

mean(vks) # 1430.924
sd(vks) # 202.3919


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Flannery - Anxiety/behaviour in children gut metagenomes
# # # # # # # # #
# # # # # # # # #

#### Flannery behaviour - prepare for data download from SRA
#-------------------------

# Flannery JE, Stagaman K, Burns AR, Hickey RJ, Roos LE, Giuliano RJ, et al. Gut Feelings Begin in Childhood: 
# the Gut Metagenome Correlates with Early Environment, Caregiving, and Behavior. mBio. 2020;11(1):e02780-19.

# https://journals.asm.org/doi/10.1128/mBio.02780-19
# Gut Feelings Begin in Childhood: the Gut Metagenome Correlates with Early Environment, Caregiving, and Behavior
# PRJNA496479

meta <- read_excel(path= "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/METADATA-Flannery-behav-SraRunTable.xlsx",
                   sheet=1, range="A1:BY41", col_names = TRUE)
meta <- as.data.frame(meta)
str(meta)

length(unique(meta$Run)) # 40

sampid <- sort( meta$Run )

head(meta)
# Run Assay Type AvgSpotLen      Bases  BioProject    BioSample                      BioSampleModel      Bytes             Center Name child_behav_CBCL_aggbehav_t1 child_behav_CBCL_anx_depress_t1
# 1 SRR8204551        WGS        302 1945397930 PRJNA496479 SAMN10438918 MIMS.me,MIGS/MIMS/MIMARKS.human-gut  868417703 OREGON STATE UNIVERSITY                           50                              50
# 2 SRR8204552        WGS        302 2760964332 PRJNA496479 SAMN10438919 MIMS.me,MIGS/MIMS/MIMARKS.human-gut 1241684057 OREGON STATE UNIVERSITY                           51                              56
# 3 SRR8204553        WGS        302 2645349672 PRJNA496479 SAMN10438920 MIMS.me,MIGS/MIMS/MIMARKS.human-gut 1205444629 OREGON STATE UNIVERSITY                           50                              52
# 4 SRR8204554        WGS        302 3768903828 PRJNA496479 SAMN10438921 MIMS.me,MIGS/MIMS/MIMARKS.human-gut 1695818578 OREGON STATE UNIVERSITY                           52                              51
# 5 SRR8204555        WGS        302 3057781106 PRJNA496479 SAMN10438914 MIMS.me,MIGS/MIMS/MIMARKS.human-gut 1462637336 OREGON STATE UNIVERSITY                           50                              52
# 6 SRR8204556        WGS        302 1585851528 PRJNA496479 SAMN10438915 MIMS.me,MIGS/MIMS/MIMARKS.human-gut  715823405 OREGON STATE UNIVERSITY                           62                              66
# child_behav_CBCL_anxproblems_t1 child_behav_CBCL_depression_t1 child_behav_CBCL_emot_react_t1 child_behav_CBCL_extern_t1 child_behav_CBCL_intern_t1 child_behav_CBCL_total_t1
# 1                              50                             50                             50                         43                         37                        39
# 2                              57                             50                             55                         48                         58                        50
# 3                              51                             51                             59                         40                         59                        52
# 4                              57                             60                             55                         51                         58                        54
# 5                              55                             56                             55                         46                         56                        51
# 6                              57                             60                             69                         63                         68                        69
# child_behav_CBQ_angfrust_t1 child_behav_CBQ_fear_t1 child_behav_CBQ_IC_t1 child_behav_CBQ_impuls_t1 child_behav_CBQ_Sadness_t1 Collection_Date Consent DATASTORE filetype DATASTORE provider
# 1          4.3333333329999997                       2                   5.5                         5         3.1428571430000001         2016-04  public          sra,fastq                 s3
# 2                           3                     3.5    5.6666666670000003                       3.8                        3.8         2016-05  public          fastq,sra                 s3
# 3          5.6666666670000003      4.1666666670000003    3.6666666669999999                       3.4         4.4285714289999998         2016-05  public          sra,fastq                 s3
# 4          3.6666666669999999      4.6666666670000003                   4.5                         4         5.4285714289999998         2016-05  public          sra,fastq                 s3
# 5          2.8333333330000001      3.1666666669999999    4.8333333329999997                       4.2         4.8571428570000004         2016-04  public          sra,fastq                 s3
# 6                         4.5      3.3333333330000001    2.8333333330000001                         5                          5         2016-08  public          fastq,sra                 s3
# DATASTORE region    env_biome           env_feature env_material esa_income_to_needs esa_LEC_Familyillness_Combined esa_LEC_NeighViolence_Combined esa_LEC_Pov_Combined esa_LEC_Sep_Combined
# 1     s3.us-east-1 Homo sapiens lower digestive tract        feces  1.0162601630000001                              0                              0                    0                    0
# 2     s3.us-east-1 Homo sapiens lower digestive tract        feces  1.2195121950000001                              2                              1                    1                    0
# 3     s3.us-east-1 Homo sapiens lower digestive tract        feces  2.4390243900000002                              0                              0                    0                    0
# 4     s3.us-east-1 Homo sapiens lower digestive tract        feces  3.2520325200000002                              1                              1                    0                    1
# 5     s3.us-east-1 Homo sapiens lower digestive tract        feces  3.2520325200000002                              0                              1                    0                    0
# 6     s3.us-east-1 Homo sapiens lower digestive tract        feces  1.6260162600000001                              1                              2                    2                    1
# esa_LEC_TOT_Combined esa_LEC_Turmoil_Combined Ethnicity Experiment geo_loc_name_country geo_loc_name_country_continent geo_loc_name gut_antibiotic_frequency gut_antibiotic_recent gut_antibiotic
# 1                    0                        0         1 SRX5023916                  USA                  North America   USA:Oregon                        1                     1              1
# 2                    4                        1         1 SRX5023915                  USA                  North America   USA:Oregon                        2                     2              1
# 3                    1                        1         4 SRX5023914                  USA                  North America   USA:Oregon                        0                     0              0
# 4                    9                        6         1 SRX5023913                  USA                  North America   USA:Oregon                        2                     1              1
# 5                    2                        0         1 SRX5023912                  USA                  North America   USA:Oregon                        0                     1              1
# 6                   10                        6         1 SRX5023911                  USA                  North America   USA:Oregon                        1                     1              1
# gut_birthmethod gut_GutHistory gut_Locations gut_milk gut_probiotic_history gut_probiotics7days gut_season_allergies_asthma      Host_Age         Host host_sex          Instrument       Lat_Lon
# 1               2              0             0        1                     0                   0                           0             5 Homo sapiens   female Illumina HiSeq 4000 not collected
# 2               1              1             2        1                     2                   1                           0             6 Homo sapiens   female Illumina HiSeq 4000 not collected
# 3               1              0             0        3                     0                   0                           0             5 Homo sapiens     male Illumina HiSeq 4000 not collected
# 4               1              1             0        3                     2                   0                           0             5 Homo sapiens     male Illumina HiSeq 4000 not collected
# 5               1              1             0        1                     0                   0                           0             6 Homo sapiens   female Illumina HiSeq 4000 not collected
# 6               1              3             0        3                     1                   0                           0 not collected Homo sapiens     male Illumina HiSeq 4000 not collected
# Library Name LibraryLayout LibrarySelection LibrarySource       Organism parent_APCA_TOT_t2 parent_AWATOT_t2 parent_NJ_TOT_t2 parent_NJDTOT_t2 parent_NR_TOT_t2 parent_NRCTOT_t2 parent_PSI_DiffChild_t1
# 1        SP006        PAIRED           RANDOM   METAGENOMIC gut metagenome                 14               30               11               28               10               22                      24
# 2        SP008        PAIRED           RANDOM   METAGENOMIC gut metagenome                 11               32               10               29                6               23                      27
# 3        SP009        PAIRED           RANDOM   METAGENOMIC gut metagenome                 15               29               12               32               10               25                      25
# 4        SP010        PAIRED           RANDOM   METAGENOMIC gut metagenome                 14               19               13               21               10               17                      32
# 5      SP001_2        PAIRED           RANDOM   METAGENOMIC gut metagenome                 14               16               13               31               12               19                      28
# 6        SP002        PAIRED           RANDOM   METAGENOMIC gut metagenome                 15               20               10               22               11               16           not collected
# parent_PSI_parentdistress_t1 parent_PSI_PC_dysfunc_t1 parent_PSI_total_t1 parent_PSOC_TOT_t1 Platform          ReleaseDate Sample Name SRA Study
# 1                           19                       14                  57                 65 ILLUMINA 2018-11-18T00:00:00Z       SP006 SRP169523
# 2                           22                       16                  65                 44 ILLUMINA 2018-11-18T00:00:00Z       SP008 SRP169523
# 3                           19                       26                  70                 53 ILLUMINA 2018-11-18T00:00:00Z       SP009 SRP169523
# 4                           31                       15                  78                 49 ILLUMINA 2018-11-18T00:00:00Z       SP010 SRP169523
# 5                           27                       19                  74                 48 ILLUMINA 2018-11-18T00:00:00Z     SP001_2 SRP169523
# 6                not collected            not collected       not collected                 49 ILLUMINA 2018-11-18T00:00:00Z       SP002 SRP169523

names(meta)

vars <- c( 
           "child_behav_CBCL_aggbehav_t1"   ,
           "child_behav_CBCL_anx_depress_t1", "child_behav_CBCL_anxproblems_t1", "child_behav_CBCL_depression_t1",  "child_behav_CBCL_emot_react_t1" , "child_behav_CBCL_extern_t1"     ,
           "child_behav_CBCL_intern_t1"     , "child_behav_CBCL_total_t1"     ,  "child_behav_CBQ_angfrust_t1"  ,   "child_behav_CBQ_fear_t1"       ,  "child_behav_CBQ_IC_t1"        ,  
           "child_behav_CBQ_impuls_t1"      , "child_behav_CBQ_Sadness_t1"    ,  
           
           "esa_income_to_needs"          ,   "esa_LEC_Familyillness_Combined",  "esa_LEC_NeighViolence_Combined",  "esa_LEC_Pov_Combined"  ,          "esa_LEC_Sep_Combined"   ,        
           "esa_LEC_TOT_Combined"         ,   "esa_LEC_Turmoil_Combined"      ,  # "Ethnicity"     
           
           "gut_antibiotic_frequency"     ,   "gut_antibiotic_recent"         ,  "gut_antibiotic"           ,      
           "gut_birthmethod"              ,   "gut_GutHistory"                ,  "gut_Locations"            ,       "gut_milk"              ,          "gut_probiotic_history"     ,     
           "gut_probiotics7days"          ,   "gut_season_allergies_asthma"   , 
           
           "Host_Age"                 ,      #  "host_sex"              ,         
           
           "parent_APCA_TOT_t2"           ,   "parent_AWATOT_t2"         ,       "parent_NJ_TOT_t2"        ,       
           "parent_NJDTOT_t2"             ,   "parent_NR_TOT_t2"        ,        "parent_NRCTOT_t2"       ,         "parent_PSI_DiffChild_t1"     ,    "parent_PSI_parentdistress_t1"   ,
           "parent_PSI_PC_dysfunc_t1"     ,   "parent_PSI_total_t1"     ,        "parent_PSOC_TOT_t1"
           )

head(meta[ ,vars])

meta[ ,vars] <- lapply(X = meta[ ,vars], FUN = as.numeric)

head(meta[ ,vars])


## child_behav
vars.child_behav <- c("child_behav_CBCL_aggbehav_t1"   ,
                      "child_behav_CBCL_anx_depress_t1", "child_behav_CBCL_anxproblems_t1", "child_behav_CBCL_depression_t1",  "child_behav_CBCL_emot_react_t1" , "child_behav_CBCL_extern_t1"     ,
                      "child_behav_CBCL_intern_t1"     , "child_behav_CBCL_total_t1"     ,  "child_behav_CBQ_angfrust_t1"  ,   "child_behav_CBQ_fear_t1"       ,  "child_behav_CBQ_IC_t1"        ,  
                      "child_behav_CBQ_impuls_t1"      , "child_behav_CBQ_Sadness_t1"   )

cor.mat.child_behav <- cor( meta[ , vars.child_behav ] , use="na.or.complete" )
cor.mat.child_behav
#                                 child_behav_CBCL_aggbehav_t1 child_behav_CBCL_anx_depress_t1 child_behav_CBCL_anxproblems_t1 child_behav_CBCL_depression_t1 child_behav_CBCL_emot_react_t1 child_behav_CBCL_extern_t1
# child_behav_CBCL_aggbehav_t1                       1.0000000                       0.7462398                      0.74871674                     0.55952938                      0.7473657                  0.8768208
# child_behav_CBCL_anx_depress_t1                    0.7462398                       1.0000000                      0.80017452                     0.68780029                      0.8135678                  0.6832696
# child_behav_CBCL_anxproblems_t1                    0.7487167                       0.8001745                      1.00000000                     0.55486340                      0.8250154                  0.6835384
# child_behav_CBCL_depression_t1                     0.5595294                       0.6878003                      0.55486340                     1.00000000                      0.6443387                  0.6147402
# child_behav_CBCL_emot_react_t1                     0.7473657                       0.8135678                      0.82501541                     0.64433871                      1.0000000                  0.7216364
# child_behav_CBCL_extern_t1                         0.8768208                       0.6832696                      0.68353837                     0.61474021                      0.7216364                  1.0000000
# child_behav_CBCL_intern_t1                         0.5847958                       0.7454060                      0.68116349                     0.64873185                      0.8284262                  0.6707455
# child_behav_CBCL_total_t1                          0.8419234                       0.7904263                      0.74802937                     0.70803445                      0.8653398                  0.9289448
# child_behav_CBQ_angfrust_t1                        0.6130701                       0.4494178                      0.45453315                     0.49623779                      0.6603061                  0.6456014
# child_behav_CBQ_fear_t1                            0.4350419                       0.2784503                      0.55241577                     0.34033890                      0.4710383                  0.4246622
# child_behav_CBQ_IC_t1                             -0.5833670                      -0.3783920                     -0.35421922                    -0.39618798                     -0.4619387                 -0.6557286
# child_behav_CBQ_impuls_t1                          0.3849632                       0.0141279                      0.09931743                     0.08444878                      0.1624068                  0.5091331
# child_behav_CBQ_Sadness_t1                         0.3729416                       0.2701522                      0.39771565                     0.46599410                      0.5162160                  0.4175054
#                                 child_behav_CBCL_intern_t1 child_behav_CBCL_total_t1 child_behav_CBQ_angfrust_t1 child_behav_CBQ_fear_t1 child_behav_CBQ_IC_t1 child_behav_CBQ_impuls_t1 child_behav_CBQ_Sadness_t1
# child_behav_CBCL_aggbehav_t1                    0.58479578                 0.8419234                   0.6130701               0.4350419            -0.5833670                0.38496323                  0.3729416
# child_behav_CBCL_anx_depress_t1                 0.74540595                 0.7904263                   0.4494178               0.2784503            -0.3783920                0.01412790                  0.2701522
# child_behav_CBCL_anxproblems_t1                 0.68116349                 0.7480294                   0.4545332               0.5524158            -0.3542192                0.09931743                  0.3977156
# child_behav_CBCL_depression_t1                  0.64873185                 0.7080345                   0.4962378               0.3403389            -0.3961880                0.08444878                  0.4659941
# child_behav_CBCL_emot_react_t1                  0.82842617                 0.8653398                   0.6603061               0.4710383            -0.4619387                0.16240677                  0.5162160
# child_behav_CBCL_extern_t1                      0.67074546                 0.9289448                   0.6456014               0.4246622            -0.6557286                0.50913306                  0.4175054
# child_behav_CBCL_intern_t1                      1.00000000                 0.8701599                   0.5953177               0.4096183            -0.3766156                0.03276004                  0.4467575
# child_behav_CBCL_total_t1                       0.87015994                 1.0000000                   0.7205136               0.4750208            -0.6297083                0.30500839                  0.5091451
# child_behav_CBQ_angfrust_t1                     0.59531767                 0.7205136                   1.0000000               0.4499350            -0.5084791                0.28285612                  0.5204611
# child_behav_CBQ_fear_t1                         0.40961828                 0.4750208                   0.4499350               1.0000000            -0.4131182                0.14721125                  0.4253271
# child_behav_CBQ_IC_t1                          -0.37661565                -0.6297083                  -0.5084791              -0.4131182             1.0000000               -0.54481759                 -0.4093993
# child_behav_CBQ_impuls_t1                       0.03276004                 0.3050084                   0.2828561               0.1472113            -0.5448176                1.00000000                  0.1250348
# child_behav_CBQ_Sadness_t1                      0.44675753                 0.5091451                   0.5204611               0.4253271            -0.4093993                0.12503481                  1.0000000

length(vars.child_behav) # 13
highCorr <- findCorrelation(cor.mat.child_behav, cutoff = .75)
length(highCorr) # 4
highCorr #  8 6 5 3 - are considered highly correlated
vars.child_behav[highCorr]
# "child_behav_CBCL_total_t1"       "child_behav_CBCL_extern_t1"      "child_behav_CBCL_emot_react_t1"  "child_behav_CBCL_anxproblems_t1"

for (i in 1:length(highCorr)) {
  #i<-5
  this_var <- vars.child_behav[highCorr][i]
  sel.corr <- which( abs(cor.mat.child_behav[this_var, ]) > 0.75)
  sel.kept <- sel.corr[ which(names(sel.corr) %in% vars.child_behav[-highCorr] ) ]
  print(paste0(this_var," is correlated with: ",paste0(names(sel.corr[-which(names(sel.corr)==this_var)]),collapse = ", "),"; ... we kept: ",names(sel.kept)))
}
# [1] "child_behav_CBCL_total_t1 is correlated with: child_behav_CBCL_aggbehav_t1, child_behav_CBCL_anx_depress_t1, child_behav_CBCL_emot_react_t1, child_behav_CBCL_extern_t1, child_behav_CBCL_intern_t1; ... we kept: child_behav_CBCL_aggbehav_t1"   
# [2] "child_behav_CBCL_total_t1 is correlated with: child_behav_CBCL_aggbehav_t1, child_behav_CBCL_anx_depress_t1, child_behav_CBCL_emot_react_t1, child_behav_CBCL_extern_t1, child_behav_CBCL_intern_t1; ... we kept: child_behav_CBCL_anx_depress_t1"
# [3] "child_behav_CBCL_total_t1 is correlated with: child_behav_CBCL_aggbehav_t1, child_behav_CBCL_anx_depress_t1, child_behav_CBCL_emot_react_t1, child_behav_CBCL_extern_t1, child_behav_CBCL_intern_t1; ... we kept: child_behav_CBCL_intern_t1"     
# [1] "child_behav_CBCL_extern_t1 is correlated with: child_behav_CBCL_aggbehav_t1, child_behav_CBCL_total_t1; ... we kept: child_behav_CBCL_aggbehav_t1"
# [1] "child_behav_CBCL_emot_react_t1 is correlated with: child_behav_CBCL_anx_depress_t1, child_behav_CBCL_anxproblems_t1, child_behav_CBCL_intern_t1, child_behav_CBCL_total_t1; ... we kept: child_behav_CBCL_anx_depress_t1"
# [2] "child_behav_CBCL_emot_react_t1 is correlated with: child_behav_CBCL_anx_depress_t1, child_behav_CBCL_anxproblems_t1, child_behav_CBCL_intern_t1, child_behav_CBCL_total_t1; ... we kept: child_behav_CBCL_intern_t1"     
# [1] "child_behav_CBCL_anxproblems_t1 is correlated with: child_behav_CBCL_anx_depress_t1, child_behav_CBCL_emot_react_t1; ... we kept: child_behav_CBCL_anx_depress_t1"

( vars.child_behav.shortlist <- vars.child_behav[-highCorr] ) 
# [1] "child_behav_CBCL_aggbehav_t1"    "child_behav_CBCL_anx_depress_t1" "child_behav_CBCL_depression_t1"  "child_behav_CBCL_intern_t1"      "child_behav_CBQ_angfrust_t1"     "child_behav_CBQ_fear_t1"        
# [7] "child_behav_CBQ_IC_t1"           "child_behav_CBQ_impuls_t1"       "child_behav_CBQ_Sadness_t1"



names(phy@sam_data)
# [1] "Run"                             "Assay Type"                      "AvgSpotLen"                      "Bases"                          
# [5] "BioProject"                      "BioSample"                       "BioSampleModel"                  "Bytes"                          
# [9] "Center Name"                     "child_behav_CBCL_aggbehav_t1"    "child_behav_CBCL_anx_depress_t1" "child_behav_CBCL_anxproblems_t1"
# [13] "child_behav_CBCL_depression_t1"  "child_behav_CBCL_emot_react_t1"  "child_behav_CBCL_extern_t1"      "child_behav_CBCL_intern_t1"     
# [17] "child_behav_CBCL_total_t1"       "child_behav_CBQ_angfrust_t1"     "child_behav_CBQ_fear_t1"         "child_behav_CBQ_IC_t1"          
# [21] "child_behav_CBQ_impuls_t1"       "child_behav_CBQ_Sadness_t1"      "Collection_Date"                 "Consent"                        
# [25] "DATASTORE filetype"              "DATASTORE provider"              "DATASTORE region"                "env_biome"                      
# [29] "env_feature"                     "env_material"                    "esa_income_to_needs"             "esa_LEC_Familyillness_Combined" 
# [33] "esa_LEC_NeighViolence_Combined"  "esa_LEC_Pov_Combined"            "esa_LEC_Sep_Combined"            "esa_LEC_TOT_Combined"           
# [37] "esa_LEC_Turmoil_Combined"        "Ethnicity"                       "Experiment"                      "geo_loc_name_country"           
# [41] "geo_loc_name_country_continent"  "geo_loc_name"                    "gut_antibiotic_frequency"        "gut_antibiotic_recent"          
# [45] "gut_antibiotic"                  "gut_birthmethod"                 "gut_GutHistory"                  "gut_Locations"                  
# [49] "gut_milk"                        "gut_probiotic_history"           "gut_probiotics7days"             "gut_season_allergies_asthma"    
# [53] "Host_Age"                        "Host"                            "host_sex"                        "Instrument"                     
# [57] "Lat_Lon"                         "Library Name"                    "LibraryLayout"                   "LibrarySelection"               
# [61] "LibrarySource"                   "Organism"                        "parent_APCA_TOT_t2"              "parent_AWATOT_t2"               
# [65] "parent_NJ_TOT_t2"                "parent_NJDTOT_t2"                "parent_NR_TOT_t2"                "parent_NRCTOT_t2"               
# [69] "parent_PSI_DiffChild_t1"         "parent_PSI_parentdistress_t1"    "parent_PSI_PC_dysfunc_t1"        "parent_PSI_total_t1"            
# [73] "parent_PSOC_TOT_t1"              "Platform"                        "ReleaseDate"                     "Sample Name"                    
# [77] "SRA Study"

# Identified key covariates in paper: "child_behav_CBQ_impuls_t1" "child_behav_CBQ_IC_t1" "child_behav_CBCL_depression_t1" 
# ?? "child_behav_CBCL_anx_depress_t1" "child_behav_CBCL_anxproblems_t1" "child_behav_CBQ_Sadness_t1"

# ( vars.child_behav.shortlist <- vars.child_behav[-highCorr] ) 
# [1] "child_behav_CBCL_aggbehav_t1"    "child_behav_CBCL_anx_depress_t1" "child_behav_CBCL_depression_t1"  "child_behav_CBCL_intern_t1"      "child_behav_CBQ_angfrust_t1"     "child_behav_CBQ_fear_t1"        
# [7] "child_behav_CBQ_IC_t1"           "child_behav_CBQ_impuls_t1"       "child_behav_CBQ_Sadness_t1"

# 1 CBCL Aggressive Behavior
# 2 CBCL Anxious Depressed
# 3 CBCL Depressive Problems
# 4 CBCL Internalizing Behavior
# 5 CBQ  Anger Frustration
# 6 CBQ Fear
# 7 CBQ Inhibitory Control
# 8 CBQ Impulsivity
# 9 CBQ Sadness

dat <- as(phy@sam_data[ ,c("child_behav_CBCL_aggbehav_t1"  ,  "child_behav_CBCL_anx_depress_t1",
                           "child_behav_CBCL_depression_t1",  "child_behav_CBCL_intern_t1"   ,
                           "child_behav_CBQ_angfrust_t1"   ,  "child_behav_CBQ_fear_t1"     ,   
                           "child_behav_CBQ_IC_t1"         ,  "child_behav_CBQ_impuls_t1"   ,
                           "child_behav_CBQ_Sadness_t1") ], "data.frame")
str(dat)
# 'data.frame':	40 obs. of  9 variables:
# $ child_behav_CBCL_aggbehav_t1   : num  50 51 50 52 50 62 50 50 50 69 ...
# $ child_behav_CBCL_anx_depress_t1: num  50 56 52 51 52 66 50 51 50 56 ...
# $ child_behav_CBCL_depression_t1 : num  50 50 51 60 56 60 50 51 51 56 ...
# $ child_behav_CBCL_intern_t1     : num  37 58 59 58 56 68 51 53 53 58 ...
# $ child_behav_CBQ_angfrust_t1    : num  4.33 3 5.67 3.67 2.83 ...
# $ child_behav_CBQ_fear_t1        : num  2 3.5 4.17 4.67 3.17 ...
# $ child_behav_CBQ_IC_t1          : num  5.5 5.67 3.67 4.5 4.83 ...
# $ child_behav_CBQ_impuls_t1      : num  5 3.8 3.4 4 4.2 ...
# $ child_behav_CBQ_Sadness_t1     : num  3.14 3.8 4.43 5.43 4.86 ...

dim(dat) # 40  9
sel <- which(phy@sam_data$host_sex == "female")
fem <- row.names(phy@sam_data)[sel]
dat.F <- dat[fem, ]
ok <- complete.cases(dat.F)
sel.ok <- which(ok==TRUE) # qty 20
dat.F <- dat.F[sel.ok, ]
dim(dat.F) # 20  9

sel <- which(phy@sam_data$host_sex == "male")
male <- row.names(phy@sam_data)[sel]
dat.M <- dat[male, ]
ok <- complete.cases(dat.M)
sel.ok <- which(ok==TRUE)
dat.M <- dat.M[sel.ok, ]
dim( dat.M ) # 


pca.dat.F <- prcomp(dat.F, scale = TRUE)
p <- ggbiplot(pca.dat.F, labels = row.names(dat.F), obs.scale = 1, var.scale = 1,
              #groups = shortlistdata.class, ellipse = TRUE, 
              labels.size = 0,
              circle = TRUE)
p

p <- ggbiplot(pca.dat.F, labels = NULL, obs.scale = 1, var.scale = 1,
              #groups = samps.class.F, ellipse = TRUE,
              labels.size = 0,
              circle = TRUE)+
  geom_point()+
  geom_text(aes(label = row.names(dat.F)), vjust = 0, hjust = 0, nudge_x = 0.2, nudge_y = 0.15, size = 3, alpha = 0.3)+
  #geom_text(aes(label = row.names(shortdat), color = samps.class), position = position_dodge(width = ), size = 2.5)+
  #guides(color = legend( ))+
  theme_bw()+
  ggtitle("(a) Females")+
  xlim(-4,6.5)+ ylim(-3,5)+
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","PCA-ggbiplot-Flannery-behav-Females-key-behaviours-v9.tiff"), width = 18, height = 20, units = "cm", res=400, compression="lzw", type = "cairo")


pca.dat.M <- prcomp(dat.M, scale = TRUE)
p <- ggbiplot(pca.dat.M, labels = row.names(dat.M), obs.scale = 1, var.scale = 1,
              labels.size = 0,
              circle = TRUE)
p

p <- ggbiplot(pca.dat.M, labels = NULL, obs.scale = 1, var.scale = 1,
              labels.size = 0,
              circle = TRUE)+
  geom_point()+
  geom_text(aes(label = row.names(dat.M)), vjust = 0, hjust = 0, nudge_x = 0.2, nudge_y = 0.15, size = 3, alpha = 0.3)+
  theme_bw()+
  ggtitle("(b) Males")+
  xlim(-4,6)+ ylim(-3,3.9)+
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","PCA-ggbiplot-Flannery-behav-Males-key-behaviours-v9.tiff"), width = 18, height = 20, units = "cm", res=400, compression="lzw", type = "cairo")


## test these variables: $PC1 and $child_behav_CBCL_depression_t1

cor( pca.dat.F$x[ ,"PC1"], dat.F$child_behav_CBCL_depression_t1) # 0.8586378

cor( pca.dat.M$x[ , "PC1"], dat.M$child_behav_CBCL_depression_t1) # 0.7201324


#-------------------------

#### Flannery behaviour - read in superfocus - fxn potential outputs
#-------------------------

sampid <- meta$Run
sampid <- sort(as.character(sampid))
length(unique(sampid)) # 40

head(sampid) # "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" "SRR8204555" "SRR8204556"

superfocus_out_dir <- "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus"

list.dirs(superfocus_out_dir)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus"                          
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204551"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204552"
# etc ...

# # don't keep 1st two 
# ( results_dirs <- list.dirs(superfocus_out_dir)[-c(1,2)] )

# # don't keep 1st directory
( results_dirs <- list.dirs(superfocus_out_dir)[-c(1)] )

head(results_dirs)
# [1] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204551"
# [2] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204552"
# [3] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204553"
# [4] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204554"
# [5] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204555"
# [6] "/Users/lidd0026/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_SRR8204556"

# check identical order
identical(sort(as.character(sampid)), gsub(pattern = "~/WORKSPACE/PROJ/Gut-and-soil/modelling/DT/flannery-behav/superfocus/superfocus_out_", replacement = "", x = results_dirs) )
# TRUE

# In this data one Run corresponds to a single Sample_ID !!!

# collate results into a long-format table

sfx.long <- data.frame(sampleID=NA, subsys_L1=NA, subsys_L2=NA, subsys_L3=NA,fxn=NA,percent_abun=NA)

for (i in 1:length(sampid)) {
  #i<-1
  this_samp <- sampid[i]
  sel.folder <- grep(pattern = this_samp, x = results_dirs)
  this_folder <- results_dirs[sel.folder]
  
  #tab1 <- read_excel(path = paste0(this_folder,"/output_all_levels_and_function.xlsx"), skip = 4, col_names = TRUE)
  
  tab <- read.csv(file = paste0(this_folder,"/output_all_levels_and_function.xls"), sep = "\t", skip = 4 )
  # names(tab)
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # [5] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq"
  # [6] "X.scratch.user.lidd0026.nz.pilot.nz_2_fastp_qc.10923_R1.good.fastq.."
  
  
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # [5] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"
  # [6] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"
  # [7] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq"
  # [8] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq"
  # [9] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [10] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H2THFBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  # [11] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L001_R1.good.fastq.." # this is %
  # [12] "X.scratch.user.lidd0026.ami_2_fastp_qc.12465_1_PE_550bp_BASE_UNSW_H3WYJBCXX_TAATGCGC.TAATCTTA_L002_R1.good.fastq.." # this is %
  
  
  tab$sampid <- this_samp
  names(tab)
  
  #tab <- tab[,c(7,1,2,3,4,6)]
  
  # last column is sampid
  # take average of percentages
  
  sel.col.percent <- grep(pattern = "R1.good.fastq..$", x = names(tab))
  if (length(sel.col.percent)>1) {
    tab$percent_abun <- apply(X = tab[ ,sel.col.percent], MARGIN = 1, FUN = mean )
  } else {
    tab$percent_abun <- tab[ ,sel.col.percent]
  }
  
  # sum(tab$percent_abun) # 100
  # mean(tab$percent_abun) # 0.004338583
  
  names(sfx.long) # "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"    "percent_abun"
  # names(tab)
  # [1] "Subsystem.Level.1"
  # [2] "Subsystem.Level.2"
  # [3] "Subsystem.Level.3"
  # [4] "Function"
  # ...
  # [13] "sampid"
  # [14] "percent_abun"
  
  tab <- tab[ ,c("sampid","Subsystem.Level.1","Subsystem.Level.2","Subsystem.Level.3","Function","percent_abun")]
  names(tab) <- names(sfx.long)
  
  sfx.long <- rbind(sfx.long,tab)
  
  print(paste0("completed ",i," - sample ID: ",sampid[i]))
}





head(sfx.long)
# remove empty 1st row
sfx.long <- sfx.long[-1, ]
dim(sfx.long) # 413407      6
head(sfx.long)
# sampleID                   subsys_L1 subsys_L2           subsys_L3
# 2 SRR8204551 Amino Acids and Derivatives         - Amino acid racemase
# 3 SRR8204551 Amino Acids and Derivatives         - Amino acid racemase
# 4 SRR8204551 Amino Acids and Derivatives         - Amino acid racemase
# 5 SRR8204551 Amino Acids and Derivatives         - Amino acid racemase
# 6 SRR8204551 Amino Acids and Derivatives         - Amino acid racemase
# 7 SRR8204551 Amino Acids and Derivatives         - Amino acid racemase
# fxn percent_abun
# 2                                                                                                      Alanine_racemase_(EC_5.1.1.1) 0.0015655056
# 3                                                                                                   Aspartate_racemase_(EC_5.1.1.13) 0.0001533557
# 4                                                                                             Diaminopimelate_epimerase_(EC_5.1.1.7) 0.0017635900
# 5                                                                                                    Glutamate_racemase_(EC_5.1.1.3) 0.0036294171
# 6                           UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-alanine_ligase_(EC_6.3.2.10)_/_Alanine_racemase_(EC_5.1.1.1) 0.0007791264
# 7 UDP-N-acetylmuramoylalanyl-D-glutamyl-2,6-diaminopimelate--D-alanyl-D-alanine_ligase_(EC_6.3.2.10)_/_Alanine_racemase_(EC_5.1.1.1) 0.0009456932


sfx.long$full_fxn_tax <- paste0(sfx.long$subsys_L1,"___", sfx.long$subsys_L2,"___", sfx.long$subsys_L3,"___", sfx.long$fxn)


## translate from long to wide format

names(sfx.long)
# "sampleID"     "subsys_L1"    "subsys_L2"    "subsys_L3"    "fxn"          "percent_abun" "full_fxn_tax"

sfx.wide <- dcast(sfx.long, formula = full_fxn_tax ~ sampleID, value.var = "percent_abun")
dim(sfx.wide) # 20787    41

sel.na <- which(is.na(sfx.wide),arr.ind = TRUE)
sfx.wide[sel.na] <- 0

temp <- sfx.wide


names(sfx.wide)

names(sfx.wide)[-1]
meta$Run

identical(sort(names(sfx.wide)[-1]), sort(meta$Run)) # TRUE

# function taxonomy
full_fxn_names <- sfx.wide$full_fxn_tax

length(full_fxn_names) # 20787
length(unique(full_fxn_names)) # 20787

names(full_fxn_names) <- paste0("fxn_",c(1:length(full_fxn_names)))
head(full_fxn_names)
# fxn_1 
# "Amino Acids and Derivatives___-___Amino acid racemase___2-methylaconitate_cis-trans_isomerase" 
# fxn_2 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-hydroxyproline_epimerase_(EC_5.1.1.8)" 
# fxn_3 
# "Amino Acids and Derivatives___-___Amino acid racemase___4-oxalomesaconate_tautomerase_(EC_5.3.2.8)" 
# fxn_4 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)" 
# fxn_5 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_/_TsaE_protein,_required_for_threonylcarbamoyladenosine_t(6)A37_formation_in_tRNA" 
# fxn_6 
# "Amino Acids and Derivatives___-___Amino acid racemase___Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic"

tax.fxn <- separate(sfx.wide, full_fxn_tax, c("subsys_L1", "subsys_L2", "subsys_L3", "fxn"), sep= "___", remove=TRUE)
# remove sample ids
#tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% sampid)]
tax.fxn <- tax.fxn[ ,-which(names(tax.fxn) %in% meta$Run)]

row.names(tax.fxn) <- names(full_fxn_names)



head(sfx.wide)


names(sfx.wide)
# [1] "full_fxn_tax" "SRR8204551"   "SRR8204552"   "SRR8204553"   "SRR8204554"   "SRR8204555"   "SRR8204556"   "SRR8204557"   "SRR8204558"   "SRR8204559"   "SRR8204560"  
# [12] "SRR8204561"   "SRR8204562"   "SRR8204563"   "SRR8204564"   "SRR8204565"   "SRR8204566"   "SRR8204567"   "SRR8204568"   "SRR8204569"   "SRR8204570"   "SRR8204571"  
# [23] "SRR8204572"   "SRR8204573"   "SRR8204574"   "SRR8204575"   "SRR8204576"   "SRR8204577"   "SRR8204578"   "SRR8204579"   "SRR8204580"   "SRR8204581"   "SRR8204582"  
# [34] "SRR8204583"   "SRR8204584"   "SRR8204585"   "SRR8204586"   "SRR8204587"   "SRR8204588"   "SRR8204589"   "SRR8204590" 

#names(sfx.wide) <- gsub(pattern = "-", replacement = "_", x = names(sfx.wide))

identical(as.character(full_fxn_names), sfx.wide$full_fxn_tax) # TRUE

row.names(sfx.wide) <- names(full_fxn_names)
sfx.wide <- sfx.wide[ ,-1]

names(sfx.wide)

head(sampid)

# "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" "SRR8204555" "SRR8204556"

length(sampid) # 40

dim(sfx.wide) #  20787    40

names(sampid) # NULL - in this case there is NOT an alternative sample name being used

# check alignment of sample IDs and sample names
identical(names(sfx.wide) , as.character(sampid)) # TRUE
# identical(names(sfx.wide) , as.character(gsub(pattern = "-",replacement = "_",x = sampid))) # FALSE
# length( names(sfx.wide) %in% as.character(gsub(pattern = "-",replacement = "_",x = sampid)) ) # 113 - i.e. matching but order different

#NOT RUN THIS TIME
#names(sfx.wide) <- names(sampid)



names(tax.fxn) # "subsys_L1" "subsys_L2" "subsys_L3" "fxn"
dim(tax.fxn) # 20787     4

length(unique(tax.fxn$subsys_L1)) # 35
length(unique(tax.fxn$subsys_L2)) # 189
length(unique(tax.fxn$subsys_L3)) # 1150
length(unique(tax.fxn$fxn)) # 10868

#-------------------------

#### Flannery behaviour - get into Phyloseq object
#-------------------------

# sfx.wide - is equiv to OTU table

# tax.fxn - is equiv to TAX table

# meta - is equiv to sample table

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 20787     4

TAX <- tax_table( tax.m )


## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( sfx.wide )
dim(otu.m)
# 20787    40

OTU <- otu_table(otu.m, taxa_are_rows = TRUE)


## Create a phyloseq object, merging OTU & TAX tables
phy = phyloseq(OTU, TAX)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20787 taxa and 40 samples ]
# tax_table()   Taxonomy Table:    [ 20787 taxa by 4 taxonomic ranks ]

sample_names(phy)
# [1] "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" "SRR8204555" "SRR8204556" "SRR8204557" "SRR8204558" "SRR8204559" "SRR8204560" "SRR8204561" "SRR8204562"
# [13] "SRR8204563" "SRR8204564" "SRR8204565" "SRR8204566" "SRR8204567" "SRR8204568" "SRR8204569" "SRR8204570" "SRR8204571" "SRR8204572" "SRR8204573" "SRR8204574"
# [25] "SRR8204575" "SRR8204576" "SRR8204577" "SRR8204578" "SRR8204579" "SRR8204580" "SRR8204581" "SRR8204582" "SRR8204583" "SRR8204584" "SRR8204585" "SRR8204586"
# [37] "SRR8204587" "SRR8204588" "SRR8204589" "SRR8204590"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(meta) # 40 77
head(row.names(meta))

identical(row.names(meta), sample_names(phy)) # TRUE

SAMP <- sample_data(meta)



### Combine SAMPDATA into phyloseq object
phy <- merge_phyloseq(phy, SAMP)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20787 taxa and 40 samples ]
# sample_data() Sample Data:       [ 40 samples by 77 sample variables ]
# tax_table()   Taxonomy Table:    [ 20787 taxa by 4 taxonomic ranks ]

head(taxa_names(phy))
# "fxn_1" "fxn_2" "fxn_3" "fxn_4" "fxn_5" "fxn_6"

head(phy@tax_table)
# Taxonomy Table:     [6 taxa by 4 taxonomic ranks]:
#   subsys_L1                     subsys_L2 subsys_L3            
# fxn_1 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_2 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_3 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_4 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_5 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn_6 "Amino Acids and Derivatives" "-"       "Amino acid racemase"
# fxn                                                                                                              
# fxn_1 "2-methylaconitate_cis-trans_isomerase"                                                                          
# fxn_2 "4-hydroxyproline_epimerase_(EC_5.1.1.8)"                                                                        
# fxn_3 "4-oxalomesaconate_tautomerase_(EC_5.3.2.8)"                                                                     
# fxn_4 "Alanine_racemase_(EC_5.1.1.1)"                                                                                  
# fxn_5 "Alanine_racemase_(EC_5.1.1.1)_/_TsaE_protein,_required_for_threonylcarbamoyladenosine_t(6)A37_formation_in_tRNA"
# fxn_6 "Alanine_racemase_(EC_5.1.1.1)_##_biosynthetic" 

getwd()  # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = phy, file = "phy-phyloseq-object-Flannery-child-behav.RDS")

#phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")

#-------------------------

#### Flannery behaviour - build reaction search in parallel - get_reactions & compounds
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")

## convert each row in functional tax_table to "mean van Krevelen distance to health-associated compounds"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 20787     4


get_rxns_and_compounds <- function( df.tax, subsys.lut, rxns.lut, rxn_pathways.lut ) {
  
  #sub-function
  atomic_no <- function(form, atom) {
    ##form=form
    ##atom="O"
    coefs <- list()
    for (c in 1:length(form)) {
      #c<-1
      pos <- str_locate(string = form[c], pattern = atom)[1]
      if (is.na(pos)) { # not present
        ##print("zero")
        coef<-0
      } else if (pos==nchar(form[c])) { # at the end, i.e. coef = 1.    # A
        ##print("A")
        coef<-1
      } else {
        # check for off-targets ... otherwise decide coef = 1, 2, etc?       # B
        ##print("B")
        checkstring <- substring(text = form[c], first = pos, last = pos+1)
        
        if (atom == "O" & checkstring %in% c("Os", "Og")) {   # C
          ##print("C")
          coef<-0
        } else if (atom == "C" & checkstring %in% c("Cs", "Ca", "Ce", "Cr", "Co", "Cm", "Cu", "Cd", "Cn", "Cf", "Cl")) { # D
          ##print("D")
          coef<-0
        } else if (atom == "H" & checkstring %in% c("Hf", "Hs", "Hg", "Ho", "He")) { # E
          ##print("E")
          coef<-0
        } else { # F
          ##print("F")
          coef<-1
          coef.try <- coef
          pos.end <- pos+1
          
          while (is.numeric(coef.try) & !is.na(coef.try) & pos.end <= nchar(form[c])+1 ) {
            # substring allows last index to be too long & need to allow 'pos.end <= nchar(form[c])+1' for atomic numbers at end of formula
            coef <- coef.try
            coef.try <- as.numeric( substring(text = form[c], first = pos+1, last = pos.end) )
            pos.end <- pos.end+1
            ##print("G")
          } # END while loop
        } # END else
        
      } # END else
      
      coefs[[c]] <- coef
      
    } # END c loop
    return( unlist(coefs) )
  } # END sub-function
  
  
  sub1 <- df.tax$subsys_L1[i]
  sub2 <- df.tax$subsys_L2[i]
  sub3 <- df.tax$subsys_L3[i]
  
  fxn.temp <- df.tax$fxn[i]
  fxn.superfocus.rowlabel <- row.names(df.tax)[i]
  
  # store results corresponding to each Superfocus row
  fxn.list <- list()
  fxn.list[[ fxn.superfocus.rowlabel  ]] <- list()
  
  # check for multiple functions/reactions?
  flag1 <- grepl(pattern = "_/_|/", x = fxn.temp)
  flag2 <- grepl(pattern = "_@_", x = fxn.temp)
  if (!any(flag1,flag2)==TRUE) {
    # no multiples
    fxns <- fxn.temp
  } else if (flag1==TRUE) {
    fxns <- unlist( strsplit(fxn.temp, split = "_/_") )  ###### WHAT ABOUT SPLIT FOR "/" WITHOUT UNDERSCORES ??
  } else {
    fxns <- unlist( strsplit(fxn.temp, split = "_@_") )
  }
  # remove underscores
  ( fxns <- gsub(pattern = "_", replacement = " ", x = fxns) )
  
  # process each fxn & store attributes
  df.fxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel,f=1:length(fxns),`f__in`=fxns, matching_method=NA, min_adist_modelSEED=NA, min_amatch_modelSEED=NA, rxns=NA, tot_mean_OC_x=NA, tot_mean_HC_y=NA )
  
  # # do round brackets interfere with search? - YES
  # lookfor <- "option 4 (this one)"
  # lookuplist <- c("option 1", "option 2", "option 3 (this one)", "option 4 (this one)")
  # grep(pattern = lookfor, x = lookuplist)
  
  # Identify '/' separators with no '_'  ??
  
  for (f in 1:length(fxns)) {  # this accounts for multiple functions/reactions reported in Superfocus outputs
    #f<-1
    f.in <- fxns[f]
    
    # these concatenated expressions will be used to look for exact match using hierarchy in ModelSEED Subsystem table
    full_hier_target <- paste0(sub1,"__",sub2,"__",sub3,"__",f.in)
    full_hier_list <- paste0(subsys.lut$Class,"__",subsys.lut$Subclass,"__",gsub("_"," ",subsys.lut$Name),"__",subsys.lut$Role)
    
    ## data cleaning
    
    # trim off '_#' and '_##' tags
    trim_nchar <- str_locate(string = f.in, pattern = " # | ## ")[1]
    if (!is.na(trim_nchar) & length(trim_nchar)==1) {
      f.in <- substring(text = f.in , first = 1, last = trim_nchar-1)
    }
    
    # Eliminate unwanted parsing of regular expressions: '[', ']','***', '(', ')'
    f.in <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\} ", replacement ="." , x = f.in) # used later
    rxns.lut$name <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$name) # used later
    rxns.lut$aliases <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = rxns.lut$aliases) # used later
    
    full_hier_target <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_target)
    full_hier_list <- gsub(pattern = "\\[|\\]|\\*+|\\(|\\)|\\{|\\}", replacement ="." , x = full_hier_list)
    
    sel.rx <- grep(pattern = full_hier_target, x = full_hier_list)
    
    ## ALTERNATIVE #1 == FULL HIERACHICAL MATCH
    if (length(sel.rx)>=1) {
      df.fxns$matching_method[f] <- "Exact hierachy match"
      df.fxns$min_adist_modelSEED[f] <- NA
      df.fxns$min_amatch_modelSEED[f] <- NA
      df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
      
    } else if (str_detect(string = fxns[f], pattern = " \\(EC ")) {  ## ALTERNATIVE #2 == MATCHING ECs
      # search by EC id if present
      
      f.in <- fxns[f] # this goes back to string with brackets for EC
      ## LOOK FOR MULTIPLE ECs ??????????
      
      how_many_ECs <- str_count(string = f.in, pattern = "\\(EC.*?\\)")
      
      ECs <- as.character( str_extract_all(string = f.in, pattern = "\\(EC.*?\\)", simplify = TRUE) )
      #class(ECs)
      ECs <- gsub(pattern = "\\(EC |\\)", replacement = "", x = ECs)
      ECs.collapse <- paste0(ECs, collapse = "|")
      
      sel.rx <- which(rxns.lut$ec_numbers == ECs.collapse)
      
      if (length(how_many_ECs)==0 | length(ECs)==0) {
        # there was a glitch, database typo, or some error in identifying the EC number
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      } else if (length(sel.rx)>=1) {
        # combined EC hits identified
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(which(rxns.lut$ec_numbers %in% ECs)) >=1) {
        # treat EC hits individually
        sel.rx <- which(rxns.lut$ec_numbers %in% ECs) # look 1st where ECs are exact matches for EC numbers in Reactions lookup table
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if (length(grep(pattern = ECs, x = rxns.lut$ec_numbers)) >=1) {
        # this allows EC to be part of a combination of EC numbers that are listed in Reactions lookup table
        sel.rx <- grep(pattern = ECs, x = rxns.lut$ec_numbers)
        
        df.fxns$matching_method[f] <- "EC number"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else {
        # it had an EC number but couldn't find a match in the EC numbers listed in Reaction lookup table
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      # END EC matching
      
      
    } else {  ## ALTERNATIVE 3 == FXN NAME MATCHING
      ## otherwise attempt to match function name - first look for exact matches   
      # 1. 'reactions' table by name: rxns.lut$name
      # 2. 'reactions' table by aliases: rxns.lut$aliases
      # 3. 'Model SEED Subsystems' table by Role: subsys.lut$Role
      # 4. 'Unique_ModelSEED_Reaction_Pathways' table by External ID: rxn_pathways.lut$External_rxn_name
      
      if ( length( grep(pattern = f.in, x = rxns.lut$name) )>=1 ) {
        # 1a - exact match - rxns.lut$name
        sel.rx <- grep(pattern = f.in, x = rxns.lut$name)
        #rxns.lut$name[sel.rx]
        df.fxns$matching_method[f] <- "Matched Reactions name"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxns.lut$aliases) )>=1 ) {
        # 2a - exact match - rxns.lut$aliases
        sel.rx <- grep(pattern = f.in, x = rxns.lut$aliases)
        #rxns.lut$aliases[sel.rx]
        #rxns.lut$name[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Reactions aliases"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxns.lut$id[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = subsys.lut$Role) )>=1 ) {
        # 3a - exact match - subsys.lut$Role
        sel.rx <- grep(pattern = f.in, x = subsys.lut$Role)
        #subsys.lut$Role[sel.rx]
        #subsys.lut$Reaction[sel.rx]
        
        df.fxns$matching_method[f] <- "Matched Subsytem role"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(subsys.lut$Reaction[sel.rx]), collapse = ";")
        
      } else if ( length( grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name) )>=1 ) {
        # 4a - exact match - rxn_pathways.lut$External_rxn_name
        sel.rx <- grep(pattern = f.in, x = rxn_pathways.lut$External_rxn_name)
        
        df.fxns$matching_method[f] <- "Matched ModelSEED Reaction pathways"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- paste0( unique(rxn_pathways.lut$rxn_id[sel.rx]), collapse = ";")
        
        
      } else {
        df.fxns$matching_method[f] <- "No match found"
        df.fxns$min_adist_modelSEED[f] <- NA
        df.fxns$min_amatch_modelSEED[f] <- NA
        df.fxns$rxns[f] <- NA
        
      }
      

    } # END function - reaction search
    
    #fxn.list[[ fxn.superfocus.rowlabel  ]][[ f ]][[ "fxns" ]] <- df.fxns
    
    print(paste0("completed fxn ", f))
    
    
    ## now investigate these reactions ...
    # Reactions lookup table: 
    # - "equation": Definition of reaction expressed using compound IDs and after protonation
    # Compounds lookup table:
    # - "formula": Standard chemical format (using Hill system) in protonated form to match reported charge
    #df.fxns
    
    
    #if (df.fxns$matching_method == "No match found") {
    if (df.fxns$rxns[f] == "" | is.na(df.fxns$rxns[f])) {
      df.Rxns <- NA
      
    } else { # reaction(s) were identified
      
      # consider reactions for this f.in only (possibly > 1 f.in per Superfocus row)
      f.in.rxns <- unique(unlist(str_split(string = df.fxns$rxns[f], pattern = ";")))
      
      df.Rxns <- data.frame(superfocus_fxn=fxn.superfocus.rowlabel, f=f, f__in=f.in,rxn_id= f.in.rxns,
                            rxn_name=NA, rxn_eqn=NA, rxn_defn=NA,compds=NA,compd_coef=NA, chem_formx=NA , OC_ratios=NA, HC_ratios=NA, coefwtmean_OC_x=NA, coefwtmean_HC_y=NA)
      
      for (r in 1:dim(df.Rxns)[1]) {
        #r<-1
        this_rxn <- df.Rxns$rxn_id[r]
        sel <- which(rxns.lut$id == this_rxn)
        ( df.Rxns$rxn_name[r] <- rxns.lut$name[sel] )
        ( df.Rxns$rxn_eqn[r] <- rxns.lut$equation[sel] )
        ( df.Rxns$rxn_defn[r] <- rxns.lut$definition[sel] )
        
        # extract compound info
        
        #df.Rxns$rxn_eqn[r]
        #[1] "(1) cpd00010[0] + (1) cpd29672[0] <=> (1) cpd00045[0] + (1) cpd11493[0]"
        #[1] "(45) cpd00144[0] + (45) cpd00175[0] <=> (45) cpd00014[0] + (45) cpd00091[0] + (1) cpd15634[0]"
        
        ( compds.idx <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd")[[1]][,"start"] )
        # 5 23 43 61
        # 6 25 46 65 83
        
        ( compds <- as.character( str_extract_all(string = df.Rxns$rxn_eqn[r], pattern = "cpd.....", simplify = TRUE) ) )
        # "cpd00010" "cpd29672" "cpd00045" "cpd11493"
        
        if (length(compds)>=1) {
          
          df.Rxns$compds[r] <- paste0(compds, collapse = ";")
          
          ## get compound coefficients?
          #( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = compds.idx - 3, last = compds.idx - 3)) )
          # 1 1 1 1
          
          start_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\(")[[1]][,"start"]
          end_brackets <- str_locate_all(string = df.Rxns$rxn_eqn[r], pattern = "\\)")[[1]][,"start"]
          ( compd.coeff <- as.numeric( substring(text = df.Rxns$rxn_eqn[r], first = start_brackets+1, last = end_brackets-1)) )
          
          df.Rxns$compd_coef[r] <- paste0(compd.coeff, collapse = ";")
          
          # get formulas of compounds
          #sel.compd <- grep(pattern = compds, x = compounds.lut$id)
          
          formx <-filter(compounds.lut, id %in% compds )
          row.names(formx) <- formx$id
          ( formx.char <- formx[compds, ]$formula )
          # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS" 
          # "C15H19N2O18P2"      "C17H25N3O17P2"      "C9H12N2O12P2"       "C9H11N2O9P"         "C630H945N45O630P45"
          # "C7H7O7" "H2O"    "C7H5O6"
          df.Rxns$chem_formx[r] <- paste0(formx.char, collapse = ";")
          
          # atomic_no(form = "H2O", atom = "C")
          # 
          # formx.char # "C21H32N7O16P3S" "HOR"            "C10H11N5O10P2"  "C11H22N2O7PRS"
          # atomic_no(form = formx.char, atom = "C") # 21  0 10 11
          # atomic_no(form = formx.char, atom = "H") # 32  1 11 22
          # atomic_no(form = formx.char, atom = "O") # 16  1 10  7
          
          
          # O:C ratio (replace Inf with NA when C not present)
          OC_ratio <- atomic_no(form = formx.char, atom = "O")/atomic_no(form = formx.char, atom = "C")
          OC_ratio[is.infinite(OC_ratio)] <- NA
          df.Rxns$OC_ratios[r] <- paste0(OC_ratio, collapse = ";")
          # H:C ratio (replace Inf with NA when C not present)
          HC_ratio <- atomic_no(form = formx.char, atom = "H")/atomic_no(form = formx.char, atom = "C")
          HC_ratio[is.infinite(HC_ratio)] <- NA
          df.Rxns$HC_ratios[r] <- paste0(HC_ratio, collapse = ";")
          
          # don't forget compound coefficients !! 
          # compd.coeff # 1 1 1 1
          # compd.coeff <- c(1, 1, 1, 2)
          # OC_ratio # 0.7619048       NA 1.0000000 0.6363636
          # OC_ratio*compd.coeff
          
          # weighted mean vK coordinates for compounds with O,C and H,C present. Reactants and products with NA ratios not included
          sel.ok <- which(complete.cases( data.frame(OC_ratio=OC_ratio, HC_ratio=HC_ratio)) == TRUE) # same indices as 
          if (length(sel.ok) >= 1) {
            df.Rxns$coefwtmean_OC_x[r] <- sum( OC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
            df.Rxns$coefwtmean_HC_y[r] <- sum( HC_ratio[sel.ok]*compd.coeff[sel.ok]/sum(compd.coeff[sel.ok]) )
          }
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- mean( df.Rxns$coefwtmean_OC_x , na.rm = TRUE)
          df.fxns$tot_mean_HC_y[f] <- mean( df.Rxns$coefwtmean_HC_y , na.rm = TRUE)
          
        } else {
          # No specified reaction equation or chemical formula info
          df.Rxns$compds[r] <- NA
          df.Rxns$compd_coef[r] <- NA
          df.Rxns$chem_formx[r] <- NA
          df.Rxns$OC_ratios[r] <- NA
          df.Rxns$HC_ratios[r] <- NA
          
          df.Rxns$coefwtmean_OC_x[r] <- NA
          df.Rxns$coefwtmean_HC_y[r] <- NA
          
          ## return values for van Krevelen coordinates in df.fxns[ f,  ]
          df.fxns$tot_mean_OC_x[f] <- NA
          df.fxns$tot_mean_HC_y[f] <- NA
          
        }
        
        
      } # END loop for r - rxn_id's per f/f.in
      
    } # END else loop when reactions identified
    
    # store results corresponding to each sub-reaction of each Superfocus row
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "fxns" ]] <- df.fxns
    if (f==1) { fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]] <- list() } # set this only once
    fxn.list[[ fxn.superfocus.rowlabel  ]][[ "rxns" ]][[ f ]] <- df.Rxns
    
  } # END loop - f in 1:length(fxns)) - to account for multiple functions/reactions reported in each row of Superfocus outputs
  
  
  #return(fxn.list)
  saveRDS(object = fxn.list, file = paste0("~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Flannery-behav-R-working-files/fxn-list-",fxn.superfocus.rowlabel,".rds") ) # use readRDS()
  
  
  print(paste0("COMPLETED ROW ",i," OF SUPERFOCUS FUNCTIONAL TAXA  # # # # # # # # # # # # # # # # # # # # #"))
  
} # END function to be run in parallel for each superfocus row


# # # # # # # # # # # # # # # # # #


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

foreach(i=1:dim(df.tax)[1] , .packages=c('stringr', 'dplyr')) %dopar%  #
  get_rxns_and_compounds( df.tax=df.tax, subsys.lut=subsys.lut, rxns.lut=rxns.lut, rxn_pathways.lut=rxn_pathways.lut )

stopCluster(cl)
time.finish <- Sys.time()



time.start #  "2023-04-07 18:46:03 ACST"
time.finish # "2023-04-07 20:22:23 ACST"

# 1hr 45min


## assemble results


modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Flannery-behav-R-working-files"



# read first output
i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
print( length(temp) )
print( names(temp) )
print( paste0( class(temp[[1]])," ",names(temp)," of length ", length(temp[[1]])," named ",paste0(names(temp[[1]]), collapse = " & ") ))
# "list fxn_1 of length 2 named fxns & rxns"
class ( temp[[1]][["fxns"]] ) # data.frame
df.out <- temp[[1]][["fxns"]]

names(df.out) #
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 


dim(df.tax) # 20787     4
num_results_files <- dim(df.tax)[1]

i<-1
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
df.out <- temp[[1]][["fxns"]]

for (i in 2: num_results_files ) {
  temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
  df.temp <- temp[[1]][["fxns"]]
  df.out <- rbind(df.out,df.temp)
  print(paste0("added df ",i," of ",num_results_files ))
}

dim(df.out) # 24462     9

table(df.out$matching_method)
# EC number      Exact hierachy match Matched Reactions aliases    Matched Reactions name     Matched Subsytem role 
# 8516                      3835                        46                       397                       628 
# No match found 
# 11040

100*length(which(df.out$matching_method=="No match found"))/length(df.out$matching_method)
# 45.13122% No match found

# 55% match found


## distil back to result per Superfocus row ...

names(df.out)
# [1] "superfocus_fxn"       "f"                    "f__in"                "matching_method"      "min_adist_modelSEED"  "min_amatch_modelSEED"
# [7] "rxns"                 "tot_mean_OC_x"        "tot_mean_HC_y" 

# distil results
df.out.distil <- data.frame( matrix(nrow = dim(df.tax), ncol = length( names(df.out) ))  )
# same names in df.conc as df.out
# num rows is same as df.tax

names(df.out.distil) <- names(df.out)
head( row.names(df.tax) )
df.out.distil$superfocus_fxn <- row.names(df.tax)

# loop through fxn_i to collate mean values (na.rm=TRUE) for each Superfocus row: "tot_mean_OC_x"        "tot_mean_HC_y"  

for (i in 1:dim(df.out.distil)[1]) {
  #i<-1
  this_fxn <- df.out.distil$superfocus_fxn[i]
  sel.row <- which(df.out$superfocus_fxn == this_fxn)
  
  df.out.distil$f[i] <- paste0(df.out$f[sel.row], collapse = ";")
  df.out.distil$matching_method[i] <- paste0(  sort(unique(df.out$matching_method[sel.row])), collapse = ";")
  df.out.distil$tot_mean_OC_x[i] <- mean(df.out$tot_mean_OC_x[sel.row], na.rm = TRUE)
  df.out.distil$tot_mean_HC_y[i] <- mean(df.out$tot_mean_HC_y[sel.row], na.rm = TRUE)
  
  print(paste0("completed i = ",i))
}

hist(df.out.distil$tot_mean_OC_x);summary(df.out.distil$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.638   0.832   0.876   1.103   2.500   10110

hist(df.out.distil$tot_mean_HC_y);summary(df.out.distil$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.257   1.401   1.426   1.601   4.000   10110

table(df.out.distil$matching_method)
# EC number                                  EC number;Exact hierachy match 
# 6119                                                             596 
# EC number;Exact hierachy match;Matched Reactions name            EC number;Exact hierachy match;Matched Subsytem role 
# 2                                                               2 
# EC number;Exact hierachy match;No match found                             EC number;Matched Reactions aliases 
# 18                                                               2 
# EC number;Matched Reactions name                                 EC number;Matched Subsytem role 
# 9                                                              67 
# EC number;Matched Subsytem role;No match found                                        EC number;No match found 
# 3                                                             528 
# Exact hierachy match                  Exact hierachy match;Matched Reactions aliases 
# 2562                                                               2 
# Exact hierachy match;Matched Reactions name                      Exact hierachy match;Matched Subsytem role 
# 6                                                              24 
# Exact hierachy match;No match found                                       Matched Reactions aliases 
# 232                                                              38 
# Matched Reactions aliases;Matched Reactions name;No match found                        Matched Reactions aliases;No match found 
# 3                                                               1 
# Matched Reactions name                           Matched Reactions name;No match found 
# 373                                                               4 
# Matched Subsytem role                            Matched Subsytem role;No match found 
# 464                                                              62 
# No match found 
# 9670 

100*length(which(df.out.distil$matching_method=="No match found"))/length(df.out$matching_method)
# 39.5307% No match found

100-39.5307 # i.e. in 60.5% of Superfocus function output rows, representative corresponding reaction, compound 
# van Krevelen coordinates were found


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"

saveRDS(object = df.out.distil, file = "df.out.distil_Flannery-behav.RDS")

#df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

#-------------------------

#### Flannery behaviour - inspect & clean data
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

this_study <- "-Flannery-behav-"

df.tax <- as.data.frame(phy@tax_table)
head(row.names(df.tax))
dim(df.tax) # 20787    4


phy@sam_data
sample_names(phy)
identical( sample_names(phy), colnames( as.matrix( phy@otu_table)) ) # TRUE

df.OTU <- as.data.frame( phy@otu_table ) # this is Superfocus functional relative abundance data represented in phyloseq OTU abundance table
dim(df.OTU) # 20787 40
df.OTU[1:5, 1:8]
# SRR8204551   SRR8204552  SRR8204553  SRR8204554  SRR8204555   SRR8204556   SRR8204557 SRR8204558
# fxn_1 0.000000000 0.000000e+00 0.000000000 0.000000000 0.000000000 3.118331e-05 0.0000000000          0
# fxn_2 0.000000000 4.608337e-05 0.000000000 0.000211155 0.000000000 1.870998e-04 0.0001831194          0
# fxn_3 0.000000000 0.000000e+00 0.000000000 0.000000000 0.000000000 0.000000e+00 0.0000000000          0
# fxn_4 0.001565506 7.680562e-04 0.005571716 0.001677509 0.002119524 1.559165e-03 0.0003440425          0
# fxn_5 0.000000000 0.000000e+00 0.000000000 0.000000000 0.000000000 0.000000e+00 0.0000000000          0

mean( df.OTU[ , "SRR8204551"] ) # 0.004810699
sum( df.OTU[ , "SRR8204551"] ) # 100

sample_sums(phy) # all values of 100


identical(row.names(df.tax), df.out.distil$superfocus_fxn) # TRUE

df.fxn_vk_coords <- df.out.distil[ ,c("tot_mean_OC_x", "tot_mean_HC_y")]
row.names(df.fxn_vk_coords) <- row.names(df.tax)
# replace all NaNs with NA
sel.na <- which(is.na( df.fxn_vk_coords), arr.ind = TRUE)
head(df.fxn_vk_coords[sel.na])
df.fxn_vk_coords[sel.na] <- NA


head( df.fxn_vk_coords )
# tot_mean_OC_x tot_mean_HC_y
# fxn_1            NA            NA
# fxn_2     0.6000000     1.8000000
# fxn_3     1.0000000     0.4285714
# fxn_4     0.6666667     2.3333333
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333



dim(df.fxn_vk_coords) # 20787     2
hist(df.fxn_vk_coords$tot_mean_OC_x);summary(df.fxn_vk_coords$tot_mean_OC_x)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   0.638   0.832   0.876   1.103   2.500   10110
hist(df.fxn_vk_coords$tot_mean_HC_y);summary(df.fxn_vk_coords$tot_mean_HC_y)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#   0.000   1.257   1.401   1.426   1.601   4.000   10110

sel.na1 <- which(is.na(df.fxn_vk_coords$tot_mean_OC_x))
sel.na2 <- which(is.na(df.fxn_vk_coords$tot_mean_HC_y))
identical(sel.na1, sel.na2) # TRUE
rm(sel.na2)

## get version with NAs removed (i.e. rows with no functional data available are stripped out)
df.fxn_vk_coords.noNAs <- df.fxn_vk_coords[-sel.na1, ]

head(df.fxn_vk_coords.noNAs)
# tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.6000000     1.8000000
# fxn_3     1.0000000     0.4285714
# fxn_4     0.6666667     2.3333333
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333
# fxn_7     0.6666667     2.3333333



saveRDS(df.fxn_vk_coords.noNAs, "df.fxn_vk_coords.noNAs.Flannery-child-behav.RDS")
#df.fxn_vk_coords.noNAs <- readRDS("df.fxn_vk_coords.noNAs.Flannery-child-behav.RDS")




## get version of fxn abundance tables corresponding to available vK coordinates
# this is relative abundance weighted version
df.OTU.noNAs <- df.OTU[-sel.na1, ]

length(sample_names(phy)) # 40
dim(df.fxn_vk_coords.noNAs) # 10677     2
dim(df.OTU.noNAs) # 10677    40

head(df.OTU.noNAs)

row.names(df.OTU.noNAs)

sort( colSums(df.OTU.noNAs) )
# SRR8204559 SRR8204585 SRR8204583 SRR8204558 SRR8204571 SRR8204577 SRR8204553 SRR8204557 SRR8204590 SRR8204566 SRR8204588 SRR8204560 
# 51.71926   52.51303   52.67207   53.20477   53.39564   53.54017   53.73302   53.80235   53.99742   54.10980   54.25424   54.47646 
# SRR8204580 SRR8204569 SRR8204586 SRR8204579 SRR8204589 SRR8204555 SRR8204584 SRR8204587 SRR8204564 SRR8204574 SRR8204554 SRR8204578 
# 54.50866   54.55780   54.65279   54.72824   54.74625   54.83567   54.94812   54.97382   55.05808   55.10364   55.21364   55.27003 
# SRR8204573 SRR8204563 SRR8204551 SRR8204576 SRR8204575 SRR8204567 SRR8204561 SRR8204552 SRR8204582 SRR8204581 SRR8204562 SRR8204565 
# 55.48825   55.70367   55.77974   55.92723   55.99199   55.99501   56.09938   56.29625   56.31064   56.41210   56.42714   56.84283 
# SRR8204568 SRR8204570 SRR8204572 SRR8204556 
# 56.88322   56.95883   57.02737   57.04438

summary( colSums(df.OTU.noNAs) )
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.72   54.22   55.02   55.03   56.02   57.04 

# this reflects that 52-57% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

## PREVIOUSLY for env/soil sample
## this reflects the just >60% of functional abundances that were mapped from Superfocus function outputs to vK mean-reaction-compound coordinates

colSums(df.OTU)
# all values of 100


dim(df.tax) # 20787     4


# keep all fxns in for now
df.in <- df.OTU.noNAs
identical(row.names(df.fxn_vk_coords.noNAs), row.names(df.OTU.noNAs)) # TRUE


dim(df.fxn_vk_coords.noNAs) # 10677     2
dim(df.in) # 10677     40

row.names(df.fxn_vk_coords.noNAs)[1:40]
row.names(df.in)[1:40]


## Add health response variable later - $PC1 and $child_behav_CBCL_depression_t1


vk.samp.noNAs.long <- data.frame( sample= NA,fxn_id=NA, 
                                  
                                  abun=NA,
                                  abun_type=NA, # filled in later in df.temp: "relabun",
                                  OC_x= NA, 
                                  HC_y= NA, 
                                  #group=NA, 
                                  sex = NA,
                                  
                                  depress_probs = NA #, # Depressive problems
                                  
                                  #behav_PC1 = NA # Later add - PC1 of behavioural problems
                                  
                                  #samplabel=NA#,
)


for (i in 1:length(sample_names(phy))) {
  #i<-1
  this_samp <- sample_names(phy)[i]
  df.temp <- data.frame( sample = rep( this_samp, times = length( row.names(df.in) ) ),
                         fxn_id = row.names(df.in),
                         
                         abun=df.in[  , this_samp ],
                         abun_type="relabun",
                         OC_x= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_OC_x"],
                         HC_y= df.fxn_vk_coords.noNAs[ row.names(df.in) , "tot_mean_HC_y"] ,
                         
                         sex = rep( phy@sam_data$host_sex[i], times = length( row.names(df.in) ) )  ,
                         
                         depress_probs = rep( phy@sam_data$child_behav_CBCL_depression_t1[i], times = length( row.names(df.in) ) ) # , # Depressive problems
                         
                         #behav_PC1 = rep( NA, times = length( row.names(df.in) ) ) # Add later
                         
                         #group= rep( phy@sam_data$`ACVD status`[i], times = length( row.names(df.in) ) ) # CHECK GROUP VARIABLE !!
  )
  # sample-label
  #df.temp$samplabel <- paste0(df.temp$sample," (", df.temp$group,")")
  
  vk.samp.noNAs.long <- rbind(vk.samp.noNAs.long, df.temp)
  print(paste0("completed adding sample ", i, " - i.e., ",sample_names(phy)[i] ))
  print(summary(vk.samp.noNAs.long))
}


dim(vk.samp.noNAs.long) # 427081      8
head(vk.samp.noNAs.long)

# remove 1st NA row
vk.samp.noNAs.long <- vk.samp.noNAs.long[-1, ]

ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok) # length = 416403 
# Note some missing values...don't remove NA rows yet

length( unique(vk.samp.noNAs.long$sample[sel.ok]) ) # 39
dim(pca.dat.F$x) # 20 9
dim(pca.dat.M$x) # 17 9

# Add $behav_PC1 variable

vk.samp.noNAs.long$behav_PC1 <- NA

# females
unique_samps <- rownames(pca.dat.F$x)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel.rows <- which(vk.samp.noNAs.long$sample == this_samp)
  vk.samp.noNAs.long$behav_PC1[sel.rows] <- pca.dat.F$x[ this_samp , "PC1"]
  print(paste0("sample #: ",i,"; # rows: ",length(sel.rows)))
}
# 20 samples

# males
unique_samps <- rownames(pca.dat.M$x)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel.rows <- which(vk.samp.noNAs.long$sample == this_samp)
  vk.samp.noNAs.long$behav_PC1[sel.rows] <- pca.dat.M$x[ this_samp , "PC1"]
  print(paste0("sample #: ",i,"; # rows: ",length(sel.rows)))
}
# 17 samples


str(vk.samp.noNAs.long)
vk.samp.noNAs.long$fxn_id <- factor(vk.samp.noNAs.long$fxn_id)
str(vk.samp.noNAs.long)
# 'data.frame':	427080 obs. of  9 variables:
# $ sample       : chr  "SRR8204551" "SRR8204551" "SRR8204551" "SRR8204551" ...
# $ fxn_id       : Factor w/ 10677 levels "fxn_10","fxn_100",..: 5020 5983 6773 7510 8200 8699 9206 9941 1 472 ...
# $ abun         : num  0 0 0.00157 0 0 ...
# $ abun_type    : chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x         : num  0.6 1 0.667 0.667 0.667 ...
# $ HC_y         : num  1.8 0.429 2.333 2.333 2.333 ...
# $ sex          : chr  "female" "female" "female" "female" ...
# $ depress_probs: num  50 50 50 50 50 50 50 50 50 50 ...
# $ behav_PC1    : num  -1.87 -1.87 -1.87 -1.87 -1.87 ...


saveRDS(object = vk.samp.noNAs.long, file = "vk.samp.noNAs.long-Flannery-behav.RDS")


#vk.samp.noNAs.long <- readRDS(file = "vk.samp.noNAs.long-Flannery-behav.RDS")


## plot density of mean reaction coordinates in vK space
p <- ggplot()+
  
  geom_bin2d(data = vk.samp.noNAs.long[ which(vk.samp.noNAs.long$abun > 0), ], aes(x = OC_x, y = HC_y), bins = 100) + # , bins = 70
  #scale_fill_viridis(option = "B", direction = -1) + # viridis inferno magma
  scale_fill_viridis(option = "viridis", direction = -1) + # viridis inferno magma option = "B"
  xlim(0,3.4)+ ylim(0,4.1)+
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  #guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "Number of\ncoordinates"))+
  guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5, title = "No. of\nCvKCs"))+
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    # #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.7, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.text = element_text(angle=25),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "Behavioural study human gut", x = unit(0.12, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=11), hjust =0 )
dev.print(tiff, file = paste0(workdir,"/plots/","Data-distribution-vK-space-scatterheatmap",this_study,"--v9.tiff"), width = 10, height = 10, units = "cm", res=450, compression="lzw",type="cairo")




## inspect examples: Superfocus fxn > reaction > compound > vK coords 

modelSEED_rxn_result_dir <- "~/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R/R-working-files/Flannery-behav-R-working-files"



## check anomalous high value in Lignin zone in PCaN density plot
#p + xlim(0.45, 0.55) + ylim(1.3,1.5)
p + xlim(0.52, 0.522) + ylim(1.39,1.41)
sel <- which(vk.samp.noNAs.long$OC_x > 0.5205 & vk.samp.noNAs.long$OC_x < 0.521 & vk.samp.noNAs.long$HC_y > 1.4 & vk.samp.noNAs.long$HC_y < 1.405)

length(sel) # 11360
dim(vk.samp.noNAs.long) # 427080      8
100*length(sel)/dim(vk.samp.noNAs.long)[1] # 2.659923% of all vK-coordinates

head( vk.samp.noNAs.long[ sel, ] )
# sample  fxn_id         abun abun_type      OC_x     HC_y    sex depress_probs
# 12  SRR8204551  fxn_12 0.0001469658   relabun 0.5206427 1.400701 female            50
# 20  SRR8204551  fxn_23 0.0000000000   relabun 0.5206427 1.400701 female            50
# 44  SRR8204551  fxn_49 0.0000000000   relabun 0.5206427 1.400701 female            50
# 69  SRR8204551  fxn_84 0.0000876318   relabun 0.5206427 1.400701 female            50
# 110 SRR8204551 fxn_136 0.0000000000   relabun 0.5206427 1.400701 female            50
# 170 SRR8204551 fxn_228 0.0002492029   relabun 0.5206427 1.400701 female            50

sum(vk.samp.noNAs.long$abun[sel])/length(unique(vk.samp.noNAs.long$sample)) # 0.04205094 % per sample
sum(vk.samp.noNAs.long$abun)/length(unique(vk.samp.noNAs.long$sample)) # 55.03008 % per sample


#fxn_12
i<-12
temp <- readRDS(paste0(modelSEED_rxn_result_dir,"/fxn-list-fxn_",i,".rds"))
temp[[1]][["fxns"]]
# superfocus_fxn f                f__in        matching_method min_adist_modelSEED min_amatch_modelSEED
# 1         fxn_12 1 hypothetical protein Matched Reactions name                  NA                   NA
# rxns tot_mean_OC_x tot_mean_HC_y
# 1 rxn01869;rxn05992;rxn06889;rxn11873;rxn11874;rxn31711;rxn34308;rxn34312;rxn34322     0.5206427      1.400701


### clean

# only remove 'hypothetical protein'
sel.rem1 <- which(vk.samp.noNAs.long$OC_x > 0.52064 & vk.samp.noNAs.long$OC_x < 0.520643 & vk.samp.noNAs.long$HC_y > 1.4007 & vk.samp.noNAs.long$HC_y < 1.400702)
length(sel.rem1) # qty 11360


vk.samp.noNAs.long.clean <- vk.samp.noNAs.long[-sel.rem1, ]

saveRDS(object = vk.samp.noNAs.long.clean, file = "vk.samp.noNAs.long.clean-Flannery-behav.RDS")


#-------------------------

#### Flannery behaviour - CPP-class: total rel abun in vk spatial regions - Pearson r
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

this_study <- "-Flannery-behav-"
header <- "-vk-zones-"

# use clean data
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


df.temp <- vk.samp.noNAs.long

names(df.temp)
# [1] "sample"        "fxn_id"        "abun"          "abun_type"     "OC_x"          "HC_y"          "sex"          
# [8] "depress_probs" "behav_PC1" 



# plot as spatial data in vK space

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "depress_probs", "behav_PC1", "sex", "OC_x", "HC_y", "abun") ]


dim(spdf) #  221111      7

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# female   male 
# 126388  94723

head(spdf)
# sample depress_probs behav_PC1    sex      OC_x     HC_y         abun
# 4  SRR8204551            50 -1.874486 female 0.6666667 2.333333 0.0015655056
# 8  SRR8204551            50 -1.874486 female 1.0000000 1.500000 0.0001533557
# 10 SRR8204551            50 -1.874486 female 0.5714286 2.000000 0.0017635900
# 11 SRR8204551            50 -1.874486 female 0.8000000 1.600000 0.0036294171
# 15 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0007791264
# 16 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0009456932


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 40

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.69   54.17   54.96   54.99   55.96   57.00

names(spdf)
# [1] "sample"        "depress_probs" "behav_PC1"     "sex"           "OC_x"          "HC_y"          "abun"      

## now want to sum total rela abun within vK spatial zones

## to de-duplicate get total relabun within a sample, then mean across samples

sum_relabun_in_vk_zones__child_behav <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  
  # lipids
  subsel.lipids <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  subsel.all <- c(subsel.lipids)
  # proteins
  subsel.proteins = which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.proteins)
  # amino_sugars
  subsel.amino_sugars = which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  subsel.all <- c(subsel.all, subsel.amino_sugars)
  # carbs - Carbohydrates
  subsel.carbs = which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  subsel.all <- c(subsel.all, subsel.carbs)
  # cond_aromtx - Condensed aromatics
  subsel.cond_aromtx = which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  subsel.all <- c(subsel.all, subsel.cond_aromtx)
  # lignin
  subsel.lignin = which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.lignin)
  # tannins
  subsel.tannins = which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)
  subsel.all <- c(subsel.all, subsel.tannins)
  # Other compounds ...
  
  # methylated
  subsel.methylated = which(                   spdf_in$OC_x <= 1.1 & spdf_in$HC_y > 2.2                        ) 
  # and not already used?
  sel.dup <- which(subsel.methylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.methylated <- subsel.methylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.methylated)
  
  # hydrated
  subsel.hydrated = which(spdf_in$OC_x > 1.1                       & spdf_in$HC_y >= 1.5                       )
  # and not already used?
  sel.dup <- which(subsel.hydrated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.hydrated <- subsel.hydrated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.hydrated)
  
  # demethylated / Oxidised-dehydrogenated
  subsel.demethylated = which(spdf_in$OC_x > 0.97                                            & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.demethylated %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.demethylated <- subsel.demethylated[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.demethylated)
  
  # condensed / Reduced-dehydrogenated - but exclude compounds that simultaneously contain no carbon or hydrogen
  #subsel.condensed = which(                   spdf_in$OC_x <= 0.97                       & spdf_in$HC_y < 0.75)
  subsel.condensed = which(  spdf_in$OC_x <= 0.97  & spdf_in$HC_y < 0.75 & !(spdf_in$OC_x ==0 & spdf_in$HC_y == 0 ) )
  
  # and not already used?
  sel.dup <- which(subsel.condensed %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.condensed <- subsel.condensed[-sel.dup] }
  subsel.all <- c(subsel.all, subsel.condensed)
  
  # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
  subsel.k2_b12 = which(spdf_in$OC_x >= 0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y > 1.25 & spdf_in$HC_y < 1.5)
  # and not already used?
  sel.dup <- which(subsel.k2_b12 %in% subsel.all)
  if (length(sel.dup) > 0) { subsel.k2_b12 <- subsel.k2_b12[-sel.dup] }
  
  df_out <- data.frame(
    sample = this_samp,
    #group = unique(spdf_in$group),
    
    depress_probs = unique(spdf_in$depress_probs),
    behav_PC1 = unique(spdf_in$behav_PC1),
    
    sex = unique(spdf_in$sex),
    
    lipids = sum(spdf_in$abun[ subsel.lipids ]),
    proteins = sum(spdf_in$abun[ subsel.proteins ]),
    amino_sugars = sum(spdf_in$abun[ subsel.amino_sugars ]),
    carbs = sum(spdf_in$abun[ subsel.carbs ]),
    cond_aromtx = sum(spdf_in$abun[ subsel.cond_aromtx ]),
    lignin = sum(spdf_in$abun[ subsel.lignin ]),
    tannins = sum(spdf_in$abun[ subsel.tannins ]),
    # Other compounds ...
    methylated = sum(spdf_in$abun[ subsel.methylated ]),
    hydrated = sum(spdf_in$abun[ subsel.hydrated ]),
    demethylated = sum(spdf_in$abun[ subsel.demethylated ]),
    condensed = sum(spdf_in$abun[ subsel.condensed ]),
    k2_b12 = sum(spdf_in$abun[ subsel.k2_b12 ])
    
  )
  return(df_out)
}



time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_zones__child_behav( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 40
out[[1]]

names(out[[1]])
# [1] "sample"        "depress_probs" "behav_PC1"     "sex"           "lipids"        "proteins"     
# [7] "amino_sugars"  "carbs"         "cond_aromtx"   "lignin"        "tannins"       "methylated"   
# [13] "hydrated"      "demethylated"  "condensed"     "k2_b12" 

df.zones <- out[[1]]
for (i in 2:length(out)) {
  df.zones <- rbind(df.zones, out[[i]])
  print(paste0("completed ",i))
}

head(df.zones)
# sample depress_probs   behav_PC1    sex    lipids proteins amino_sugars    carbs cond_aromtx   lignin  tannins methylated hydrated demethylated condensed    k2_b12
# 1 SRR8204551            50 -1.87448650 female 0.1507128 1.897631     3.239260 13.17554   0.1126499 5.264359 12.91243  0.3526106 5.137315     13.02428 0.3024634 0.1539537
# 2 SRR8204552            50 -0.43433382 female 0.1309083 2.000757     3.201692 13.74382   0.1151142 5.496099 12.63870  0.3460869 5.283503     12.72766 0.3976934 0.1660868
# 3 SRR8204553            51 -0.03233704   male 0.2211104 2.064452     2.962748 13.31103   0.1610775 4.767847 11.81230  0.3955733 4.883315     12.47729 0.5462004 0.1015234
# 4 SRR8204554            60  0.53364338   male 0.1917405 1.974012     3.182798 13.87624   0.1288090 4.950525 12.04284  0.3448731 5.152524     12.82348 0.3565881 0.1476455
# 5 SRR8204555            56  0.03186055 female 0.1552000 1.886734     3.246889 13.37759   0.1142747 5.001541 12.23241  0.4067835 4.982174     12.78444 0.4333244 0.1568141
# 6 SRR8204556            60  3.04880557   male 0.1997340 1.942059     3.219995 12.57751   0.1430784 5.247037 13.48899  0.3415040 5.418448     14.05777 0.2296276 0.1285064

names(df.zones)
# [1] "sample"        "depress_probs" "behav_PC1"     "sex"           "lipids"        "proteins"     
# [7] "amino_sugars"  "carbs"         "cond_aromtx"   "lignin"        "tannins"       "methylated"   
# [13] "hydrated"      "demethylated"  "condensed"     "k2_b12" 

names(df.zones[ ,c(5:16)])

df.zones$all <- apply(X = df.zones[ ,c(5:16)], MARGIN = 1, FUN = sum)


saveRDS(object = df.zones, file = "df.zones--Flannery-child-behav.RDS")
#df.zones <- readRDS("df.zones--Flannery-child-behav.RDS")


str(df.zones)
# 'data.frame':	40 obs. of  17 variables:
# $ sample       : chr  "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" ...
# $ depress_probs: num  50 50 51 60 56 60 50 51 51 56 ...
# $ behav_PC1    : num  -1.8745 -0.4343 -0.0323 0.5336 0.0319 ...
# $ sex          : chr  "female" "female" "male" "male" ...
# $ lipids       : num  0.151 0.131 0.221 0.192 0.155 ...
# $ proteins     : num  1.9 2 2.06 1.97 1.89 ...
# $ amino_sugars : num  3.24 3.2 2.96 3.18 3.25 ...
# $ carbs        : num  13.2 13.7 13.3 13.9 13.4 ...
# $ cond_aromtx  : num  0.113 0.115 0.161 0.129 0.114 ...
# $ lignin       : num  5.26 5.5 4.77 4.95 5 ...
# $ tannins      : num  12.9 12.6 11.8 12 12.2 ...
# $ methylated   : num  0.353 0.346 0.396 0.345 0.407 ...
# $ hydrated     : num  5.14 5.28 4.88 5.15 4.98 ...
# $ demethylated : num  13 12.7 12.5 12.8 12.8 ...
# $ condensed    : num  0.302 0.398 0.546 0.357 0.433 ...
# $ k2_b12       : num  0.154 0.166 0.102 0.148 0.157 ...
# $ all          : num  55.7 56.2 53.7 55.2 54.8 ...


sel.notNA <- which(!is.na(df.zones$behav_PC1))
table(df.zones$sex[sel.notNA])
# female   male 
# 20     17 


df.zones.melt <- melt(df.zones, id.vars = c("sample","depress_probs","behav_PC1","sex"))
head(df.zones.melt)
#       sample depress_probs   behav_PC1    sex variable     value
# 1 SRR8204551            50 -1.87448650 female   lipids 0.1507128
# 2 SRR8204552            50 -0.43433382 female   lipids 0.1309083
# 3 SRR8204553            51 -0.03233704   male   lipids 0.2208065
# 4 SRR8204554            60  0.53364338   male   lipids 0.1917405
# 5 SRR8204555            56  0.03186055 female   lipids 0.1548496
# 6 SRR8204556            60  3.04880557   male   lipids 0.1996404

str(df.zones.melt)
# sample depress_probs   behav_PC1    sex variable     value
# 1 SRR8204551            50 -1.87448650 female   lipids 0.1507128
# 2 SRR8204552            50 -0.43433382 female   lipids 0.1309083
# 3 SRR8204553            51 -0.03233704   male   lipids 0.2211104
# 4 SRR8204554            60  0.53364338   male   lipids 0.1917405
# 5 SRR8204555            56  0.03186055 female   lipids 0.1552000
# 6 SRR8204556            60  3.04880557   male   lipids 0.1997340

unique( df.zones.melt$sex ) #  "male"   "female"
df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)


hist(df.zones.melt$value)
hist(log10( df.zones.melt$value))


hist(df.zones.melt$depress_probs)
hist(log10(df.zones.melt$depress_probs))
hist(df.zones.melt$behav_PC1)



## child_behav_PC1
p <- ggplot(data = df.zones.melt, aes(x = behav_PC1, y = value))+
  geom_point(aes(color = sex, shape = sex), alpha=0.6)+
  geom_smooth(aes(color = sex), method="loess")+
  geom_smooth(color = "blue",linetype = "dashed", method="lm")+
  facet_wrap(facets = vars(variable), nrow=2, scales = "free_y")
p
dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-rough-boxplots-vsBehav_PC1",this_study,header,"-v9.tiff"), width = 24, height = 12, units = "cm", res=600, compression="lzw",type="cairo")



dat <- df.zones.melt

dim(dat) #  520    6



names(dat)
# [1] "sample"        "depress_probs" "behav_PC1"     "sex"           "variable"     
# [6] "value"  


cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)


## perform sig tests for annotations !!
unique(dat$variable)
# [1] lipids       proteins     amino_sugars carbs        cond_aromtx  lignin       tannins      methylated   hydrated     demethylated condensed   
# [12] k2_b12       all         
# Levels: lipids proteins amino_sugars carbs cond_aromtx lignin tannins methylated hydrated demethylated condensed k2_b12 all

dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins", "methylated", "hydrated", "demethylated", "condensed", "k2_b12", "all"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds", "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds","Near VitK2-VitB12\ncompounds", "All compounds"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                        "Proteins"                     
# [3] "Amino sugars"                  "Carbohydrates"                
# [5] "Condensed\naromatics"          "Lignin"                       
# [7] "Tannins"                       "Other methylated\ncompounds"  
# [9] "Other hydrated\ncompounds"     "Other demethylated\ncompounds"
# [11] "Other condensed\ncompounds"    "Near VitK2-VitB12\ncompounds" 
# [13] "All compounds"    





### Testing for relationships b/w vK zones & PC1 behavioural problems ...


pos_anotations <- list()
sig_anotations <- list()
correlation <- list()



## lipids
this_var <- "Lipids"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.90773, df = 18, p-value = 0.188
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.1844463  1.0000000
# sample estimates:
#       cor 
# 0.2092185

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 1.8943, df = 15, p-value = 0.03881
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.03183745 1.00000000
# sample estimates:
#       cor 
# 0.4393732 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Proteins"
this_var <- "Proteins"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.24694, df = 18, p-value = 0.4039
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3281592  1.0000000
# sample estimates:
#        cor 
# 0.05810613 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.96999, df = 15, p-value = 0.1737
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.1893873  1.0000000
# sample estimates:
#       cor 
# 0.2429474 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Amino sugars"
this_var <- "Amino sugars"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 1.7968, df = 18, p-value = 0.04458
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.01284844 1.00000000
# sample estimates:
#      cor 
# 0.389987 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 1.7183, df = 15, p-value = 0.05315
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.009340901  1.000000000
# sample estimates:
#       cor 
# 0.4055423

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Carbohydrates"
this_var <- "Carbohydrates"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -0.83634, df = 18, p-value = 0.207
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.2003165
# sample estimates:
#        cor 
# -0.1934062

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -3.0579, df = 15, p-value = 0.003987
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000 -0.2774154
# sample estimates:
#      cor 
# -0.61968

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.90324, df = 18, p-value = 0.1892
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.1854462  1.0000000
# sample estimates:
#       cor 
# 0.2082283 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.58498, df = 15, p-value = 0.2836
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.2813355  1.0000000
# sample estimates:
#       cor 
# 0.1493483

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Lignin"
this_var <- "Lignin"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.14444, df = 18, p-value = 0.4434
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3495201  1.0000000
# sample estimates:
#        cor 
# 0.03402549 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.77514, df = 15, p-value = 0.2252
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.2362301  1.0000000
# sample estimates:
#       cor 
# 0.1962487 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Tannins"
this_var <- "Tannins"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "two.sided") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.024216, df = 18, p-value = 0.9809
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.4379192  0.4470992
# sample estimates:
#         cor 
# 0.005707629

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 1.3763, df = 15, p-value = 0.09447
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.09108143  1.00000000
# sample estimates:
#       cor 
# 0.3348413 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Other methylated\ncompounds"
this_var <- "Other methylated\ncompounds"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 3.1203, df = 18, p-value = 0.002956
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.275248 1.000000
# sample estimates:
#       cor 
# 0.5924734

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.80569, df = 15, p-value = 0.2165
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.2289192  1.0000000
# sample estimates:
#       cor 
# 0.2036686 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Other hydrated\ncompounds"
this_var <- "Other hydrated\ncompounds"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -0.7259, df = 18, p-value = 0.2386
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.2247587
# sample estimates:
#        cor 
# -0.1686465

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 1.1504, df = 15, p-value = 0.134
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.1457369  1.0000000
# sample estimates:
#       cor 
# 0.2847316 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Other demethylated\ncompounds"
this_var <- "Other demethylated\ncompounds"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -0.20832, df = 18, p-value = 0.4187
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.3362467
# sample estimates:
#         cor 
# -0.04904154 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 2.3116, df = 15, p-value = 0.01771
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.1258582 1.0000000
# sample estimates:
#      cor 
# 0.512515 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Other condensed\ncompounds"
this_var <- "Other condensed\ncompounds"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.057461, df = 18, p-value = 0.4774
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3673813  1.0000000
# sample estimates:
#        cor 
# 0.01354234 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -1.0417, df = 15, p-value = 0.157
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.1720411
# sample estimates:
#        cor 
# -0.2597463 

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Near VitK2-VitB12\ncompounds" 
this_var <- "Near VitK2-VitB12\ncompounds" 

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 4.1305, df = 18, p-value = 0.0003141
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.4330301 1.0000000
# sample estimates:
#       cor 
# 0.6975721 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -0.56423, df = 15, p-value = 0.2905
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.2862089
# sample estimates:
#        cor 
# -0.1441623 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "All compounds"
this_var <- "All compounds"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = -0.33588, df = 18, p-value = 0.3704
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.3093714
# sample estimates:
#         cor 
# -0.07892099

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"value"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = 0.48541, df = 15, p-value = 0.3172
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3046158  1.0000000
# sample estimates:
#       cor 
# 0.1243588 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



annots.Flannery.behav_PC1.pearson <- list()
annots.Flannery.behav_PC1.pearson[["pos"]] <- pos_anotations
annots.Flannery.behav_PC1.pearson[["sig"]] <- sig_anotations
annots.Flannery.behav_PC1.pearson[["pearson"]] <- correlation
saveRDS(object = annots.Flannery.behav_PC1.pearson, file = "annots.Flannery.behav_PC1.pearson--list.RDS")


# combined female & male
annotation_df.combined <- data.frame(
  variable=c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
             "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
  npcx = c( pos_anotations[["Lipids"]][1],
            pos_anotations[["Proteins"]][1],
            pos_anotations[["Amino sugars"]][1],
            pos_anotations[["Carbohydrates"]][1],
            pos_anotations[["Condensed\naromatics"]][1],
            pos_anotations[["Lignin"]][1],
            pos_anotations[["Tannins"]][1],
            pos_anotations[["Other methylated\ncompounds"]][1],
            pos_anotations[["Other hydrated\ncompounds"]][1],
            pos_anotations[["Other demethylated\ncompounds"]][1],
            pos_anotations[["Other condensed\ncompounds"]][1],
            pos_anotations[["Near VitK2-VitB12\ncompounds"]][1],
            pos_anotations[["All compounds"]][1]
  ),
  npcy = c( pos_anotations[["Lipids"]][2],
            pos_anotations[["Proteins"]][2],
            pos_anotations[["Amino sugars"]][2],
            pos_anotations[["Carbohydrates"]][2],
            pos_anotations[["Condensed\naromatics"]][2],
            pos_anotations[["Lignin"]][2],
            pos_anotations[["Tannins"]][2],
            pos_anotations[["Other methylated\ncompounds"]][2],
            pos_anotations[["Other hydrated\ncompounds"]][2],
            pos_anotations[["Other demethylated\ncompounds"]][2],
            pos_anotations[["Other condensed\ncompounds"]][2],
            pos_anotations[["Near VitK2-VitB12\ncompounds"]][2],
            pos_anotations[["All compounds"]][2]
  ),
  
  test_results = c(
    # model on this template: 
    # test_result <- paste0("Kendall's tau",
    #                       "\nF: ",round(ktcor.f$estimate,3),sig_symb(ktcor.f$p.value),
    #                       "\nM: ",round(ktcor.m$estimate,3),sig_symb(ktcor.m$p.value))
    paste0("Pearson r",
           "\nF: ",round(correlation[["Lipids"]][1],3),sig_symb(sig_anotations[["Lipids"]][1]),
           "\nM: ",round(correlation[["Lipids"]][2],3),sig_symb(sig_anotations[["Lipids"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Proteins"]][1],3),sig_symb(sig_anotations[["Proteins"]][1]),
           "\nM: ",round(correlation[["Proteins"]][2],3),sig_symb(sig_anotations[["Proteins"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Amino sugars"]][1],3),sig_symb(sig_anotations[["Amino sugars"]][1]),
           "\nM: ",round(correlation[["Amino sugars"]][2],3),sig_symb(sig_anotations[["Amino sugars"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Carbohydrates"]][1],3),sig_symb(sig_anotations[["Carbohydrates"]][1]),
           "\nM: ",round(correlation[["Carbohydrates"]][2],3),sig_symb(sig_anotations[["Carbohydrates"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Condensed\naromatics"]][1],3),sig_symb(sig_anotations[["Condensed\naromatics"]][1]),
           "\nM: ",round(correlation[["Condensed\naromatics"]][2],3),sig_symb(sig_anotations[["Condensed\naromatics"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Lignin"]][1],3),sig_symb(sig_anotations[["Lignin"]][1]),
           "\nM: ",round(correlation[["Lignin"]][2],3),sig_symb(sig_anotations[["Lignin"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Tannins"]][1],3),sig_symb(sig_anotations[["Tannins"]][1]),
           "\nM: ",round(correlation[["Tannins"]][2],3),sig_symb(sig_anotations[["Tannins"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Other methylated\ncompounds"]][1],3),sig_symb(sig_anotations[["Other methylated\ncompounds"]][1]),
           "\nM: ",round(correlation[["Other methylated\ncompounds"]][2],3),sig_symb(sig_anotations[["Other methylated\ncompounds"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Other hydrated\ncompounds"]][1],3),sig_symb(sig_anotations[["Other hydrated\ncompounds"]][1]),
           "\nM: ",round(correlation[["Other hydrated\ncompounds"]][2],3),sig_symb(sig_anotations[["Other hydrated\ncompounds"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Other demethylated\ncompounds"]][1],3),sig_symb(sig_anotations[["Other demethylated\ncompounds"]][1]),
           "\nM: ",round(correlation[["Other demethylated\ncompounds"]][2],3),sig_symb(sig_anotations[["Other demethylated\ncompounds"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Other condensed\ncompounds"]][1],3),sig_symb(sig_anotations[["Other condensed\ncompounds"]][1]),
           "\nM: ",round(correlation[["Other condensed\ncompounds"]][2],3),sig_symb(sig_anotations[["Other condensed\ncompounds"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Near VitK2-VitB12\ncompounds"]][1],3),sig_symb(sig_anotations[["Near VitK2-VitB12\ncompounds"]][1]),
           "\nM: ",round(correlation[["Near VitK2-VitB12\ncompounds"]][2],3),sig_symb(sig_anotations[["Near VitK2-VitB12\ncompounds"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["All compounds"]][1],3),sig_symb(sig_anotations[["All compounds"]][1]),
           "\nM: ",round(correlation[["All compounds"]][2],3),sig_symb(sig_anotations[["All compounds"]][2]))
  ),
  stringsAsFactors = FALSE)
str(annotation_df.combined)
# 'data.frame':	13 obs. of  4 variables:
#   $ variable    : chr  "Lipids" "Proteins" "Amino sugars" "Carbohydrates" ...
# $ npcx        : chr  "left" "left" "right" "right" ...
# $ npcy        : chr  "top" "top" "bottom" "top" ...
# $ test_results: chr  "Pearson r\nF: 0.209ns\nM: 0.439*" "Pearson r\nF: 0.058ns\nM: 0.243ns" "Pearson r\nF: 0.39*\nM: 0.406ns" "Pearson r\nF: -0.193ns\nM: -0.62**" ...

# ensure no conflict in factor levels
annotation_df.combined$variable <- factor(annotation_df.combined$variable, 
                                          levels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins", "Other methylated\ncompounds",
                                                     "Other hydrated\ncompounds", "Other demethylated\ncompounds", "Other condensed\ncompounds", "Near VitK2-VitB12\ncompounds", "All compounds"),
                                          ordered = TRUE)

annotation_df.combined$test_results
# [1] "Pearson r\nF: 0.209ns\nM: 0.439*"    "Pearson r\nF: 0.058ns\nM: 0.243ns"   "Pearson r\nF: 0.39*\nM: 0.406ns"    
# [4] "Pearson r\nF: -0.193ns\nM: -0.62**"  "Pearson r\nF: 0.208ns\nM: 0.149ns"   "Pearson r\nF: 0.034ns\nM: 0.196ns"  
# [7] "Pearson r\nF: 0.006ns\nM: 0.335ns"   "Pearson r\nF: 0.592**\nM: 0.204ns"   "Pearson r\nF: -0.169ns\nM: 0.285ns" 
# [10] "Pearson r\nF: -0.049ns\nM: 0.513*"   "Pearson r\nF: 0.014ns\nM: -0.26ns"   "Pearson r\nF: 0.698***\nM: -0.144ns"
# [13] "Pearson r\nF: -0.079ns\nM: 0.124ns" 

annotation_df.combined$test_results <- gsub(pattern = "ns", replacement = "", x = annotation_df.combined$test_results)

str(dat)
# 'data.frame':	520 obs. of  6 variables:
# $ sample       : chr  "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" ...
# $ depress_probs: num  50 50 51 60 56 60 50 51 51 56 ...
# $ behav_PC1    : num  -1.8745 -0.4343 -0.0323 0.5336 0.0319 ...
# $ sex          : Ord.factor w/ 2 levels "Female"<"Male": 1 1 2 2 1 2 2 2 1 2 ...
# $ variable     : Ord.factor w/ 13 levels "Lipids"<"Proteins"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value        : num  0.151 0.131 0.221 0.192 0.155 ...


cols
# Female      Male 
# "#7570b3" "#d95f02" 

p <- ggplot(data = dat, aes(x=behav_PC1, y = value))+
  geom_point(aes(color=sex),shape = 1)+
  #geom_smooth(aes(color=sex),method="loess")+
  geom_smooth(aes(color=sex),method="lm")+
  
  scale_color_manual(values = cols, name = NULL)+
  
  #theme_bw()+
  theme_classic()+
  xlab("PC1 of problem behaviors")+ ylab("Function % rel. abun. in vK zones")+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)+
  geom_text_npc(data = annotation_df.combined, aes(npcx = npcx, npcy = npcy, label = test_results), size = 3, lineheight = .8)+
  theme(
    legend.position = c(0.715, 0.15),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 1,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(), 
    axis.text.x = element_text(size = rel(1.1))
  )
p

#dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-Trendplots-FM-PC1-BehavProbs",this_study,header,"-v9.tiff"), width = 24, height = 20, units = "cm", res=450, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","All-vkZones-Trendplots-FM-PC1-BehavProbs",this_study,header,"-v9e.tiff"), width = 24, height = 18, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------

#### Flannery behaviour - CPP-density: total rel abun in HAC buffers (radius: 0.05, 0.1, 0.15, 0.2, 0.25) - Pearson r
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

this_study <- "-Flannery-behav-"
header <- "-vk-hac-buffers-"

# use clean data
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

str(vk.samp.noNAs.long)
# 'data.frame':	415720 obs. of  9 variables:
# $ sample       : chr  "SRR8204551" "SRR8204551" "SRR8204551" "SRR8204551" ...
# $ fxn_id       : Factor w/ 10677 levels "fxn_10","fxn_100",..: 5020 5983 6773 7510 8200 8699 9206 9941 1 472 ...
# $ abun         : num  0 0 0.00157 0 0 ...
# $ abun_type    : chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x         : num  0.6 1 0.667 0.667 0.667 ...
# $ HC_y         : num  1.8 0.429 2.333 2.333 2.333 ...
# $ sex          : chr  "female" "female" "female" "female" ...
# $ depress_probs: num  50 50 50 50 50 50 50 50 50 50 ...
# $ behav_PC1    : num  -1.87 -1.87 -1.87 -1.87 -1.87 ...


# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data 

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "depress_probs", "behav_PC1", "sex", "OC_x", "HC_y", "abun") ]

dim(spdf) #  221111      7

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# female   male 
# 126388  94723 
sel.na <- which(spdf$sex=="NA") # empty
sel.na <- which(is.na(spdf$sex)) # empty

head(spdf)
# sample depress_probs behav_PC1    sex      OC_x     HC_y         abun
# 4  SRR8204551            50 -1.874486 female 0.6666667 2.333333 0.0015655056
# 8  SRR8204551            50 -1.874486 female 1.0000000 1.500000 0.0001533557
# 10 SRR8204551            50 -1.874486 female 0.5714286 2.000000 0.0017635900
# 11 SRR8204551            50 -1.874486 female 0.8000000 1.600000 0.0036294171
# 15 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0007791264
# 16 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0009456932


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 40

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.69   54.17   54.96   54.99   55.96   57.00

names(spdf) # "sample"        "depress_probs" "behav_PC1"     "sex"           "OC_x"          "HC_y"          "abun"


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 40

hist(spdf$abun);summary(spdf$abun)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000013 0.0000895 0.0008235 0.0099476 0.0073967 1.3698288 

12.5/(pi*0.05^2) # 1591.549


## now want to sum total rel abun within vK buffer radii of health-associated compounds (HAC)

df.hac
#                            sample       OC_x     HC_y                  group
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin
df.hac$short_name <- c( "Acetate", "Propionate", "Butyrate", "VitB2", "VitB12", "VitB6", "VitB9", "VitK2" )
df.hac$short_name # "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"  

## built in calculation is for these HAC above

rad <- c(0.05,0.1,0.15,0.2,0.25)


sum_relabun_in_vk_hac_buffers_childBehav <- function(spdf_in, radii, samps) {
  ###spdf_in = spdf
  ###radii = rad
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)
  
  spdf_in <- spdf_in[sel, ]
  names(spdf_in) # "sample" "group"  "OC_x"   "HC_y"   "abun"   "sex"   
  
  ## pre-calculate distances for moving window
  spdf_in$vkdist_acetate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 2)^2)^0.5           # vK coords: OC_x = 1, HC_y = 2
  spdf_in$vkdist_propionate <- ((spdf_in$OC_x - 0.6667)^2 + (spdf_in$HC_y - 2)^2)^0.5   # vK coords: OC_x = 0.6667, HC_y = 2
  spdf_in$vkdist_butyrate <- ((spdf_in$OC_x - 0.5)^2 + (spdf_in$HC_y - 2)^2)^0.5        # vK coords: OC_x = 0.5, HC_y = 2
  spdf_in$vkdist_vitB2 <- ((spdf_in$OC_x - 0.3529)^2 + (spdf_in$HC_y - 1.1765)^2)^0.5   # vK coords: OC_x = 0.3529, HC_y = 1.1765
  spdf_in$vkdist_vitB12 <- ((spdf_in$OC_x - 0.2258)^2 + (spdf_in$HC_y - 1.4194)^2)^0.5  # vK coords: OC_x = 0.2258, HC_y = 1.4194
  spdf_in$vkdist_vitB6 <- ((spdf_in$OC_x - 0.75)^2 + (spdf_in$HC_y - 1.25)^2)^0.5       # vK coords: OC_x = 0.75, HC_y = 1.25
  spdf_in$vkdist_vitB9 <- ((spdf_in$OC_x - 0.3158)^2 + (spdf_in$HC_y - 1)^2)^0.5        # vK coords: OC_x = 0.3158, HC_y = 1
  spdf_in$vkdist_vitK2 <- ((spdf_in$OC_x - 0.0645)^2 + (spdf_in$HC_y - 1.2903)^2)^0.5   # vK coords: OC_x = 0.0645, HC_y = 1.2903
  
  spdf_in$vkdist_glutamate <- ((spdf_in$OC_x - 0.8)^2 + (spdf_in$HC_y - 1.8)^2)^0.5   # vK coords: OC_x = 0.8, HC_y = 1.8
  spdf_in$vkdist_pyruvate <- ((spdf_in$OC_x - 1)^2 + (spdf_in$HC_y - 1.3333)^2)^0.5   # vK coords: OC_x = 1, HC_y = 1.3333
  
  #df_out <- data.frame(sample=NA, group=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA )
  df_out <- data.frame(sample=NA, depress_probs=NA, behav_PC1=NA, sex=NA, rad=NA, Acetate=NA, Propionate=NA, Butyrate=NA, VitB2=NA, VitB12=NA, VitB6=NA, VitB9=NA, VitK2=NA, Glutamate=NA, Pyruvate=NA )
  
  for (r in seq_along(radii) ) {
    #r<-1
    this_rad <- radii[r]
    subsel.acetate <- which(spdf_in$vkdist_acetate <= this_rad)
    subsel.propionate <- which(spdf_in$vkdist_propionate <= this_rad)
    subsel.butyrate <- which(spdf_in$vkdist_butyrate <= this_rad)
    subsel.vitB2 <- which(spdf_in$vkdist_vitB2 <= this_rad)
    subsel.vitB12 <- which(spdf_in$vkdist_vitB12 <= this_rad)
    subsel.vitB6 <- which(spdf_in$vkdist_vitB6 <= this_rad)
    subsel.vitB9 <- which(spdf_in$vkdist_vitB9 <= this_rad)
    subsel.vitK2 <- which(spdf_in$vkdist_vitK2 <= this_rad)
    
    subsel.glutamate <- which(spdf_in$vkdist_glutamate <= this_rad)
    subsel.pyruvate <- which(spdf_in$vkdist_pyruvate <= this_rad)
    
    temp_out <- data.frame( 
      sample = this_samp,
      #group = unique(spdf_in$group),
      
      depress_probs = unique(spdf_in$depress_probs), 
      behav_PC1 = unique(spdf_in$behav_PC1),
      
      sex = unique(spdf_in$sex),
      rad = this_rad,
      # sum fxn rel abun per unit area
      Acetate = sum( spdf_in$abun[subsel.acetate] )/(pi*this_rad^2),
      Propionate = sum( spdf_in$abun[subsel.propionate] )/(pi*this_rad^2),
      Butyrate = sum( spdf_in$abun[subsel.butyrate] )/(pi*this_rad^2),
      VitB2 = sum( spdf_in$abun[subsel.vitB2] )/(pi*this_rad^2),
      VitB12 = sum( spdf_in$abun[subsel.vitB12] )/(pi*this_rad^2), 
      VitB6 = sum( spdf_in$abun[subsel.vitB6] )/(pi*this_rad^2),
      VitB9 = sum( spdf_in$abun[subsel.vitB9] )/(pi*this_rad^2),
      VitK2 = sum( spdf_in$abun[subsel.vitK2] )/(pi*this_rad^2),
      
      Glutamate = sum( spdf_in$abun[subsel.glutamate] )/(pi*this_rad^2),
      Pyruvate = sum( spdf_in$abun[subsel.pyruvate] )/(pi*this_rad^2)
      
    )
    #names(temp_out) # "sample"     "group"      "sex"        "rad"        "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"     
    
    df_out <- rbind(df_out, temp_out)
    
  } # END seq_along(radii)
  
  #remove NA 1st row
  df_out <- df_out[-1, ]
  
  return(df_out)
}



time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  sum_relabun_in_vk_hac_buffers_childBehav( spdf_in = spdf, radii = rad, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 40
out[[1]]
# sample depress_probs behav_PC1    sex  rad  Acetate Propionate Butyrate     VitB2    VitB12     VitB6     VitB9      VitK2 Glutamate Pyruvate
# 2 SRR8204551            50 -1.874486 female 0.05 48.86425  19.005484 0.000000 101.66116  3.815351 117.87525  1.025107 0.00000000 104.77604 67.32615
# 3 SRR8204551            50 -1.874486 female 0.10 68.46169  10.239639 2.264196  31.68773  1.779603 111.43360  4.818247 0.06020470  65.33095 77.49673
# 4 SRR8204551            50 -1.874486 female 0.15 46.03689   9.506177 7.773733  19.21363  4.499171  82.80684  4.888621 0.03109672  56.24766 65.22680
# 5 SRR8204551            50 -1.874486 female 0.20 31.51873  13.056066 7.770143  16.51712 10.342801  69.83372 11.134727 0.79484439  46.93038 85.24401
# 6 SRR8204551            50 -1.874486 female 0.25 29.18636  17.157773 6.387077  15.62635 10.744058  67.43207  8.794826 1.47369385  49.07835 81.62921


df.buffers <- out[[1]]
for (i in 2:length(out)) {
  df.buffers <- rbind(df.buffers, out[[i]])
  print(paste0("completed ",i))
}

head(df.buffers)
# sample depress_probs  behav_PC1    sex  rad  Acetate Propionate Butyrate     VitB2    VitB12     VitB6     VitB9      VitK2 Glutamate Pyruvate
# 2  SRR8204551            50 -1.8744865 female 0.05 48.86425  19.005484 0.000000 101.66116  3.815351 117.87525  1.025107 0.00000000 104.77604 67.32615
# 3  SRR8204551            50 -1.8744865 female 0.10 68.46169  10.239639 2.264196  31.68773  1.779603 111.43360  4.818247 0.06020470  65.33095 77.49673
# 4  SRR8204551            50 -1.8744865 female 0.15 46.03689   9.506177 7.773733  19.21363  4.499171  82.80684  4.888621 0.03109672  56.24766 65.22680
# 5  SRR8204551            50 -1.8744865 female 0.20 31.51873  13.056066 7.770143  16.51712 10.342801  69.83372 11.134727 0.79484439  46.93038 85.24401
# 6  SRR8204551            50 -1.8744865 female 0.25 29.18636  17.157773 6.387077  15.62635 10.744058  67.43207  8.794826 1.47369385  49.07835 81.62921
# 21 SRR8204552            50 -0.4343338 female 0.05 55.04660  19.904282 0.000000 109.89371  4.303139 112.02454  0.170158 0.00000000 117.18805 74.13134


table(df.buffers$sex[ which(df.buffers$rad==0.05) ]) # count at single radius only
# female   male 
# 23     17 


saveRDS(object = df.buffers, file = "df.buffers--Flannery-behav.RDS")
#df.buffers <- readRDS("df.buffers--Flannery-behav.RDS")

str(df.buffers)
# 'data.frame':	200 obs. of  15 variables:
# $ sample       : chr  "SRR8204551" "SRR8204551" "SRR8204551" "SRR8204551" ...
# $ depress_probs: num  50 50 50 50 50 50 50 50 50 50 ...
# $ behav_PC1    : num  -1.87 -1.87 -1.87 -1.87 -1.87 ...
# $ sex          : chr  "female" "female" "female" "female" ...
# $ rad          : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ Acetate      : num  48.9 68.5 46 31.5 29.2 ...
# $ Propionate   : num  19.01 10.24 9.51 13.06 17.16 ...
# $ Butyrate     : num  0 2.26 7.77 7.77 6.39 ...
# $ VitB2        : num  101.7 31.7 19.2 16.5 15.6 ...
# $ VitB12       : num  3.82 1.78 4.5 10.34 10.74 ...
# $ VitB6        : num  117.9 111.4 82.8 69.8 67.4 ...
# $ VitB9        : num  1.03 4.82 4.89 11.13 8.79 ...
# $ VitK2        : num  0 0.0602 0.0311 0.7948 1.4737 ...
# $ Glutamate    : num  104.8 65.3 56.2 46.9 49.1 ...
# $ Pyruvate     : num  67.3 77.5 65.2 85.2 81.6 ...


df.buffers.melt <- melt(df.buffers, id.vars = c("sample","depress_probs","behav_PC1","sex","rad"))
head(df.buffers.melt)
# sample depress_probs  behav_PC1    sex  rad variable    value
# 1 SRR8204551            50 -1.8744865 female 0.05  Acetate 48.86425
# 2 SRR8204551            50 -1.8744865 female 0.10  Acetate 68.46169
# 3 SRR8204551            50 -1.8744865 female 0.15  Acetate 46.03689
# 4 SRR8204551            50 -1.8744865 female 0.20  Acetate 31.51873
# 5 SRR8204551            50 -1.8744865 female 0.25  Acetate 29.18636
# 6 SRR8204552            50 -0.4343338 female 0.05  Acetate 55.04660


unique( df.buffers.melt$sex ) #  "male"   "female"
df.buffers.melt$sex <- factor(df.buffers.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)


unique(df.buffers.melt$variable)
# [1] Acetate    Propionate Butyrate   VitB2      VitB12     VitB6      VitB9      VitK2      Glutamate  Pyruvate  
# Levels: Acetate Propionate Butyrate VitB2 VitB12 VitB6 VitB9 VitK2 Glutamate Pyruvate

str(df.buffers.melt)
# 'data.frame':	2000 obs. of  7 variables:
#   $ sample       : chr  "SRR8204551" "SRR8204551" "SRR8204551" "SRR8204551" ...
# $ depress_probs: num  50 50 50 50 50 50 50 50 50 50 ...
# $ behav_PC1    : num  -1.87 -1.87 -1.87 -1.87 -1.87 ...
# $ sex          : Ord.factor w/ 2 levels "Female"<"Male": 1 1 1 1 1 1 1 1 1 1 ...
# $ rad          : num  0.05 0.1 0.15 0.2 0.25 0.05 0.1 0.15 0.2 0.25 ...
# $ variable     : Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value        : num  48.9 68.5 46 31.5 29.2 ...

hist(df.buffers.melt$value)
hist(log10( df.buffers.melt$value))


dat <- df.buffers.melt


dat$log10abun <- dat$value # then over-write later


vars <- levels(dat$variable)

for (i in 1:length(vars)) {
  #i<-1
  this_var <- vars[i]
  sel <- which(dat$variable == this_var)
  print("# # # # # # # # #")
  print(paste0("Data for -- ",this_var))
  print("Raw values:")
  print(summary(dat$value[sel]))
  hist(dat$value[sel], main = paste0(this_var," (% rel abun)"))
  
  # log10
  # set zero-replacement value at 1/2 smallest non-zero value of that group
  subsel.zero <- which(dat$log10abun[sel] == 0)
  if (length(subsel.zero) > 0) {
    zero_replace <- 0.5*min(dat$log10abun[sel [-subsel.zero] ])
    dat$log10abun[sel [subsel.zero] ] <- zero_replace
  }
  dat$log10abun[sel] <- log10(dat$log10abun[sel])
  hist(dat$log10abun[sel], main = paste0(this_var," (log10 % rel abun)"))
  print("log10 values:")
  print(summary(dat$log10abun[sel]))
}

# [1] "# # # # # # # # #"
# [1] "Data for -- Acetate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 25.67   35.01   48.94   51.69   62.98  123.66 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.409   1.544   1.690   1.683   1.799   2.092 
# [1] "# # # # # # # # #"
# [1] "Data for -- Propionate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 6.847   9.831  13.178  13.875  17.752  25.674 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.8355  0.9926  1.1199  1.1209  1.2493  1.4095 
# [1] "# # # # # # # # #"
# [1] "Data for -- Butyrate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   1.905   6.348   4.711   7.481   8.795 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.2487  0.2799  0.8027  0.5231  0.8740  0.9443 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 11.91   16.56   18.95   36.40   33.13  132.96 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.076   1.219   1.278   1.434   1.520   2.124 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB12"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.5478  2.8268  5.1355  6.4268 10.4088 12.5905 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.2614  0.4513  0.7106  0.7158  1.0174  1.1000 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB6"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.87   66.82   78.77   84.19  104.37  118.88 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.770   1.825   1.896   1.915   2.019   2.075 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitB9"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   3.696   5.391   5.926   8.967  13.177 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -2.4822  0.5677  0.7317  0.5400  0.9526  1.1198 
# [1] "# # # # # # # # #"
# [1] "Data for -- VitK2"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00000 0.02697 0.14568 0.53962 0.98892 2.13869 
# [1] "log10 values:"
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# -2.887184 -1.570515 -0.836631 -1.015856 -0.004842  0.330148 
# [1] "# # # # # # # # #"
# [1] "Data for -- Glutamate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 44.29   51.07   59.56   71.07   75.93  188.76 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.646   1.708   1.775   1.823   1.880   2.276 
# [1] "# # # # # # # # #"
# [1] "Data for -- Pyruvate"
# [1] "Raw values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.23   66.94   74.63   73.54   80.68   90.03 
# [1] "log10 values:"
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.742   1.826   1.873   1.864   1.907   1.954


#cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
#shapes <- c("Female" = 2, "Male" = 4)

head(dat)
# sample depress_probs  behav_PC1    sex  rad variable    value log10abun
# 1 SRR8204551            50 -1.8744865 Female 0.05  Acetate 48.86425  1.688991
# 2 SRR8204551            50 -1.8744865 Female 0.10  Acetate 68.46169  1.835448
# 3 SRR8204551            50 -1.8744865 Female 0.15  Acetate 46.03689  1.663106
# 4 SRR8204551            50 -1.8744865 Female 0.20  Acetate 31.51873  1.498569
# 5 SRR8204551            50 -1.8744865 Female 0.25  Acetate 29.18636  1.465180
# 6 SRR8204552            50 -0.4343338 Female 0.05  Acetate 55.04660  1.740731





## PC1 of behavioural issues

##Females
p <- ggplot(data = filter(dat, sex == "Female"), aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = behav_PC1, color = behav_PC1))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)+
  guides(color = guide_colorbar(direction = "horizontal", barheight = 0.8, title = "PC1 of\nproblem\nbehaviors\n(Females)"))+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    #legend.position="bottom",
    legend.position=c(0.75, 0.12),
    legend.title = element_text(size = rel(0.9)),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

grid.text(label = "(a)", x = unit(0.02, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )

dev.print(tiff, file = paste0(workdir,"/plots/","HAC-vkBuffers-Females-PC1_behav-Loess-smooth-",this_study,header,"-v9d.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



##Males
p <- ggplot(data = filter(dat, sex == "Male"), aes(x = rad, y = log10abun))+ # y = value
  geom_smooth(aes(group = behav_PC1, color = behav_PC1))+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 3)+
  guides(color = guide_colorbar(direction = "horizontal", barheight = 0.8, title = "PC1 of\nproblem\nbehaviors\n(Males)"))+
  theme_classic() +
  labs(x = "Buffer radius (vK units)", y = "log10 fxn % rel. abun. in buffer (per vK unit area)") +
  theme(
    #legend.position="bottom",
    legend.position=c(0.75, 0.12),
    legend.title = element_text(size = rel(0.9)),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )

p

grid.text(label = "(b)", x = unit(0.02, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )

dev.print(tiff, file = paste0(workdir,"/plots/","HAC-vkBuffers-Males-PC1_behav-Loess-smooth-",this_study,header,"-v9d.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



saveRDS(object = dat, file = paste0("dat--F-M-HAC-vkBuffers-Depress_probs-PC1_behav-Loess-smooth-",this_study,header,"-v9.RDS"))

#dat <- readRDS(file = paste0("dat--F-M-HAC-vkBuffers-Depress_probs-PC1_behav-Loess-smooth-",this_study,header,"-v9.RDS"))




## now create NEW dat below, saving only rad = 0.05

sel <- which( dat$rad == 0.05 )

dat <- dat[sel, ]

cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)

head(dat)
#        sample depress_probs   behav_PC1    sex  rad variable    value log10abun
# 1  SRR8204551            50 -1.87448650 Female 0.05  Acetate 48.86425  1.688991
# 6  SRR8204552            50 -0.43433382 Female 0.05  Acetate 55.04660  1.740731
# 11 SRR8204553            51 -0.03233704   Male 0.05  Acetate 58.64373  1.768222
# 16 SRR8204554            60  0.53364338   Male 0.05  Acetate 49.80016  1.697231
# 21 SRR8204555            56  0.03186055 Female 0.05  Acetate 61.73077  1.790502
# 26 SRR8204556            60  3.04880557   Male 0.05  Acetate 39.26311  1.593985



## child_behav_PC1
p <- ggplot(data = dat, aes(x = behav_PC1, y = log10abun))+
  geom_point(aes(color = sex, shape = sex), alpha=0.6)+
  geom_smooth(aes(color = sex), method="loess")+
  #geom_smooth(color = "blue",linetype = "dashed", method="lm")+
  facet_wrap(facets = vars(variable), nrow=3, scales = "free_y")+
  theme(
    legend.position = c(0.715, 0.15)
  )
p
dev.print(tiff, file = paste0(workdir,"/plots/","F-M-HAC-vkBuffers-rough-trends-vsBehav_PC1",this_study,header,"-v9.tiff"), width = 24, height = 12, units = "cm", res=600, compression="lzw",type="cairo")



### Test for trends with PC1_behaviour


pos_anotations <- list()
sig_anotations <- list()
correlation <- list()



## Acetate
this_var <- "Acetate"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 0.14226, df = 18, p-value = 0.4442
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3499712  1.0000000
# sample estimates:
#        cor 
# 0.03351207 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -3.1463, df = 15, p-value = 0.003328
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000 -0.2937734
# sample estimates:
#        cor 
# -0.6305308 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Propionate"
this_var <- "Propionate"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 1.4029, df = 18, p-value = 0.08883
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.07388766  1.00000000
# sample estimates:
#       cor 
# 0.3139425 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 0.90131, df = 15, p-value = 0.1908
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.2059499  1.0000000
# sample estimates:
#       cor 
# 0.2266616 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Butyrate"
this_var <- "Butyrate"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = NA, df = 18, p-value = NA
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  NA  1
# sample estimates:
# cor 
#  NA 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "value"]
# t = NA, df = 15, p-value = NA
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  NA  1
# sample estimates:
# cor 
#  NA 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "VitB2"
this_var <- "VitB2"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 0.66358, df = 18, p-value = 0.2577
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.2384771  1.0000000
# sample estimates:
#       cor 
# 0.1545289

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -0.41116, df = 15, p-value = 0.3434
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.3217906
# sample estimates:
#        cor 
# -0.1055676 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "VitB12"
this_var <- "VitB12"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 1.3157, df = 18, p-value = 0.1024
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.09332213  1.00000000
# sample estimates:
#       cor 
# 0.2961928 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -2.3136, df = 15, p-value = 0.01764
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000 -0.1262733
# sample estimates:
#        cor 
# -0.5128259 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "VitB6"
this_var <- "VitB6"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 0.34271, df = 18, p-value = 0.3679
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3079197  1.0000000
# sample estimates:
#        cor 
# 0.08051535 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 1.7825, df = 15, p-value = 0.04746
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.005759188 1.000000000
# sample estimates:
#       cor 
# 0.4180815 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "VitB9"
this_var <- "VitB9"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -0.68998, df = 18, p-value = 0.2495
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.2326738
# sample estimates:
#        cor 
# -0.1605207 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 2.202, df = 15, p-value = 0.02187
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  0.1017101 1.0000000
# sample estimates:
#       cor 
# 0.4942501

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "VitK2"
this_var <- "VitK2"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 0.2202, df = 18, p-value = 0.4141
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3337642  1.0000000
# sample estimates:
#        cor 
# 0.05183073

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test

#exclude outlier
subsel <- which(dat$log10abun[sel] > -2.5)

#cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.m<- cor.test(x = dat$behav_PC1[sel[-subsel]], y = dat[sel[-subsel] ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
# Warning message:
#   In cor(x, y) : the standard deviation is zero
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel[-subsel]] and dat[sel[-subsel], "log10abun"]
# t = NA, df = 14, p-value = NA
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  NA  1
# sample estimates:
# cor 
#  NA 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Glutamate"
this_var <- "Glutamate"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "greater") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = 0.22547, df = 18, p-value = 0.4121
# alternative hypothesis: true correlation is greater than 0
# 95 percent confidence interval:
#  -0.3326614  1.0000000
# sample estimates:
#        cor 
# 0.05306788

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -2.771, df = 15, p-value = 0.007135
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.000000 -0.221931
# sample estimates:
#        cor 
# -0.5818744 

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



## "Pyruvate"
this_var <- "Pyruvate"

## Female:
sel <- which(dat$variable == this_var & dat$sex == "Female")
# correlation test
cor.f<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.f
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -0.36577, df = 18, p-value = 0.3594
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.3030087
# sample estimates:
#         cor 
# -0.08589422 

## Male:
sel <- which(dat$variable == this_var & dat$sex == "Male")
# correlation test
cor.m<- cor.test(x = dat$behav_PC1[sel], y = dat[sel ,"log10abun"], method = "pearson", alternative = "less") # default getOption("na.action") = "na.omit"
cor.m
# Pearson's product-moment correlation
# 
# data:  dat$behav_PC1[sel] and dat[sel, "log10abun"]
# t = -0.68423, df = 15, p-value = 0.2521
# alternative hypothesis: true correlation is less than 0
# 95 percent confidence interval:
#  -1.0000000  0.2578891
# sample estimates:
#        cor 
# -0.1739721 

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c(cor.f$p.value, cor.m$p.value )
correlation[[this_var]] <- c(cor.f$estimate, cor.m$estimate )



annots.Flannery.behav_PC1.buff.pearson <- list()
annots.Flannery.behav_PC1.buff.pearson[["pos"]] <- pos_anotations
annots.Flannery.behav_PC1.buff.pearson[["sig"]] <- sig_anotations
annots.Flannery.behav_PC1.buff.pearson[["correlation"]] <- correlation
saveRDS(object = annots.Flannery.behav_PC1.buff.pearson, file = "annots.Flannery.vkBuffers.behav_PC1.pearson--list-v9.RDS")

levels(dat$variable)
# "Acetate"    "Propionate" "Butyrate"   "VitB2"      "VitB12"     "VitB6"      "VitB9"      "VitK2"      "Glutamate"  "Pyruvate"  


# combined female & male
annotation_df.combined <- data.frame(
  variable=c("Acetate" , "Propionate" , "Butyrate" ,  "VitB2" , "VitB12"  , "VitB6"  ,  "VitB9" ,   "VitK2"   ,  "Glutamate", "Pyruvate"),
  npcx = c( pos_anotations[["Acetate"]][1],
            pos_anotations[["Propionate"]][1],
            pos_anotations[["Butyrate"]][1],
            pos_anotations[["VitB2"]][1],
            pos_anotations[["VitB12"]][1],
            pos_anotations[["VitB6"]][1],
            pos_anotations[["VitB9"]][1],
            pos_anotations[["VitK2"]][1],
            pos_anotations[["Glutamate"]][1],
            pos_anotations[["Pyruvate"]][1]
  ),
  npcy = c( pos_anotations[["Acetate"]][2],
            pos_anotations[["Propionate"]][2],
            pos_anotations[["Butyrate"]][2],
            pos_anotations[["VitB2"]][2],
            pos_anotations[["VitB12"]][2],
            pos_anotations[["VitB6"]][2],
            pos_anotations[["VitB9"]][2],
            pos_anotations[["VitK2"]][2],
            pos_anotations[["Glutamate"]][2],
            pos_anotations[["Pyruvate"]][2]
  ),
  
  test_results = c(
    # model on this template: 
    # test_result <- paste0("Kendall's tau",
    #                       "\nF: ",round(ktcor.f$estimate,3),sig_symb(ktcor.f$p.value),
    #                       "\nM: ",round(ktcor.m$estimate,3),sig_symb(ktcor.m$p.value))
    paste0("Pearson r",
           "\nF: ",round(correlation[["Acetate"]][1],3),sig_symb(sig_anotations[["Acetate"]][1]),
           "\nM: ",round(correlation[["Acetate"]][2],3),sig_symb(sig_anotations[["Acetate"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Propionate"]][1],3),sig_symb(sig_anotations[["Propionate"]][1]),
           "\nM: ",round(correlation[["Propionate"]][2],3),sig_symb(sig_anotations[["Propionate"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Butyrate"]][1],3),sig_symb(sig_anotations[["Butyrate"]][1]),
           "\nM: ",round(correlation[["Butyrate"]][2],3),sig_symb(sig_anotations[["Butyrate"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["VitB2"]][1],3),sig_symb(sig_anotations[["VitB2"]][1]),
           "\nM: ",round(correlation[["VitB2"]][2],3),sig_symb(sig_anotations[["VitB2"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["VitB12"]][1],3),sig_symb(sig_anotations[["VitB12"]][1]),
           "\nM: ",round(correlation[["VitB12"]][2],3),sig_symb(sig_anotations[["VitB12"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["VitB6"]][1],3),sig_symb(sig_anotations[["VitB6"]][1]),
           "\nM: ",round(correlation[["VitB6"]][2],3),sig_symb(sig_anotations[["VitB6"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["VitB9"]][1],3),sig_symb(sig_anotations[["VitB9"]][1]),
           "\nM: ",round(correlation[["VitB9"]][2],3),sig_symb(sig_anotations[["VitB9"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["VitK2"]][1],3),sig_symb(sig_anotations[["VitK2"]][1]),
           "\nM: ",round(correlation[["VitK2"]][2],3),sig_symb(sig_anotations[["VitK2"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Glutamate"]][1],3),sig_symb(sig_anotations[["Glutamate"]][1]),
           "\nM: ",round(correlation[["Glutamate"]][2],3),sig_symb(sig_anotations[["Glutamate"]][2])),
    paste0("Pearson r",
           "\nF: ",round(correlation[["Pyruvate"]][1],3),sig_symb(sig_anotations[["Pyruvate"]][1]),
           "\nM: ",round(correlation[["Pyruvate"]][2],3),sig_symb(sig_anotations[["Pyruvate"]][2]))
  ),
  stringsAsFactors = FALSE)
str(annotation_df.combined)
# 'data.frame':	10 obs. of  4 variables:
#   $ variable    : chr  "Acetate" "Propionate" "Butyrate" "VitB2" ...
# $ npcx        : chr  "left" "right" "left" "right" ...
# $ npcy        : chr  "bottom" "bottom" "bottom" "bottom" ...
# $ test_results: chr  "Pearson r\nF: 0.034ns\nM: -0.631**" "Pearson r\nF: 0.314ns\nM: 0.227ns" "Pearson r\nF: NANA\nM: NANA" "Pearson r\nF: 0.155ns\nM: -0.106ns" ...

# ensure no conflict in factor levels
annotation_df.combined$variable <- factor(annotation_df.combined$variable, 
                                          levels = c("Acetate" , "Propionate" , "Butyrate" ,  "VitB2" , "VitB12"  , "VitB6"  ,  "VitB9" ,   "VitK2"   ,  "Glutamate", "Pyruvate"),
                                          ordered = TRUE)

annotation_df.combined$test_results
# [1] "Pearson r\nF: 0.034ns\nM: -0.631**"  "Pearson r\nF: 0.314ns\nM: 0.227ns"   "Pearson r\nF: NANA\nM: NANA"        
# [4] "Pearson r\nF: 0.155ns\nM: -0.106ns"  "Pearson r\nF: 0.296ns\nM: -0.513*"   "Pearson r\nF: 0.081ns\nM: 0.418*"   
# [7] "Pearson r\nF: -0.161ns\nM: 0.494*"   "Pearson r\nF: 0.052ns\nM: NANA"      "Pearson r\nF: 0.053ns\nM: -0.582**" 
# [10] "Pearson r\nF: -0.086ns\nM: -0.174ns"

annotation_df.combined$test_results <- gsub(pattern = "ns", replacement = "", x = annotation_df.combined$test_results)
annotation_df.combined$test_results <- gsub(pattern = "NANA", replacement = "N/A", x = annotation_df.combined$test_results)

str(dat)
# 'data.frame':	400 obs. of  8 variables:
# $ sample       : chr  "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" ...
# $ depress_probs: num  50 50 51 60 56 60 50 51 51 56 ...
# $ behav_PC1    : num  -1.8745 -0.4343 -0.0323 0.5336 0.0319 ...
# $ sex          : Ord.factor w/ 2 levels "Female"<"Male": 1 1 2 2 1 2 2 2 1 2 ...
# $ rad          : num  0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 ...
# $ variable     : Factor w/ 10 levels "Acetate","Propionate",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value        : num  48.9 55 58.6 49.8 61.7 ...
# $ log10abun    : num  1.69 1.74 1.77 1.7 1.79 ...

cols
# Female      Male 
# "#7570b3" "#d95f02" 

p <- ggplot(data = dat, aes(x=behav_PC1, y = log10abun))+
  geom_point(aes(color=sex),shape = 1)+
  #geom_smooth(aes(color=sex),method="loess")+
  geom_smooth(aes(color=sex),method="lm")+
  scale_color_manual(values = cols, name = NULL)+
  
  #theme_bw()+
  theme_classic()+
  xlab("PC1 of problem behaviors")+ ylab("Function log10 % rel. abun. in 0.05 radius vK buffer zones")+
  facet_wrap(facets = vars(variable), scales = "free",  nrow = 2)+
  geom_text_npc(data = annotation_df.combined, aes(npcx = npcx, npcy = npcy, label = test_results), size = 3, lineheight = .8)+
  theme(
    #legend.position = c(0.715, 0.15),
    legend.position = "bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    strip.text = element_text(margin=margin(t = 1,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(), 
    axis.text.x = element_text(size = rel(1.1))
  )
p

#grid.text(label = "(b)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )

dev.print(tiff, file = paste0(workdir,"/plots/","All-F-M-vkBuffers-log1oabun-Trendplots-FM-behav_PC1",this_study,header,"-ved.tiff"), width = 22, height = 13, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Flannery behaviour  - do weighted-mean vK coordinates vary within vK zones - by groups?
#-------------------------

# Determines weighted coordinates in each sample, then later assess stats by group_mf?

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

this_study <- "-Flannery-behav-"
header <- "-vk-coords-wtmean-"

# use clean data
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


# from earlier
str(vk.samp.noNAs.long)
# 'data.frame':	415720 obs. of  9 variables:
# $ sample       : chr  "SRR8204551" "SRR8204551" "SRR8204551" "SRR8204551" ...
# $ fxn_id       : Factor w/ 10677 levels "fxn_10","fxn_100",..: 5020 5983 6773 7510 8200 8699 9206 9941 1 472 ...
# $ abun         : num  0 0 0.00157 0 0 ...
# $ abun_type    : chr  "relabun" "relabun" "relabun" "relabun" ...
# $ OC_x         : num  0.6 1 0.667 0.667 0.667 ...
# $ HC_y         : num  1.8 0.429 2.333 2.333 2.333 ...
# $ sex          : chr  "female" "female" "female" "female" ...
# $ depress_probs: num  50 50 50 50 50 50 50 50 50 50 ...
# $ behav_PC1    : num  -1.87 -1.87 -1.87 -1.87 -1.87 ...


ok <- complete.cases(vk.samp.noNAs.long)
sel.ok <- which(ok==TRUE) # qty 384541


# vK spatial data
df.temp <- vk.samp.noNAs.long # reminder this is long format data

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "depress_probs", "behav_PC1", "sex", "OC_x", "HC_y", "abun") ]


dim(spdf) #  221111      7

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# female   male 
# 126388  94723

sel.na <- which(spdf$sex=="NA") # empty
sel.na <- which(is.na(spdf$sex)) # empty

ok <- complete.cases(spdf)
sel.ok <- which(ok==TRUE) # qty 203753

sel.notNA <- which(!is.na(spdf$behav_PC1)) # qty 203753


head(spdf)
#        sample depress_probs behav_PC1    sex      OC_x     HC_y         abun
# 4  SRR8204551            50 -1.874486 female 0.6666667 2.333333 0.0015655056
# 8  SRR8204551            50 -1.874486 female 1.0000000 1.500000 0.0001533557
# 10 SRR8204551            50 -1.874486 female 0.5714286 2.000000 0.0017635900
# 11 SRR8204551            50 -1.874486 female 0.8000000 1.600000 0.0036294171
# 15 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0007791264
# 16 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0009456932

temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 40

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.69   54.17   54.96   54.99   55.96   57.00 

names(spdf) # "sample"        "depress_probs" "behav_PC1"     "sex"           "OC_x"          "HC_y"          "abun"



get_wtmean_vk_coords_Behav <- function(spdf_in, samps) {
  ###spdf_in = spdf
  ###samps = unique_samps
  ###i = 1
  this_samp <- samps[i]
  sel <- which(spdf_in$sample == this_samp)

  spdf_in <- spdf_in[sel, ]

  subsel <- list()

  # lipids
  subsel[["lipids"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.2 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.3)
  # proteins
  subsel[["proteins"]] <- which(spdf_in$OC_x > 0.2 & spdf_in$OC_x <= 0.52 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # amino_sugars
  subsel[["amino_sugars"]] <- which(spdf_in$OC_x > 0.52 & spdf_in$OC_x <= 0.7 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.2)
  # carbs - Carbohydrates
  subsel[["carbs"]] <- which(spdf_in$OC_x > 0.7 & spdf_in$OC_x <= 1.1 & spdf_in$HC_y >= 1.5 & spdf_in$HC_y <= 2.4)
  # cond_aromtx - Condensed aromatics
  subsel[["cond_aromtx"]] <- which(spdf_in$OC_x >=0 & spdf_in$OC_x <= 0.25 & spdf_in$HC_y >= 0.5 & spdf_in$HC_y <= 1.25)
  # lignin
  subsel[["lignin"]] <- which(spdf_in$OC_x > 0.25 & spdf_in$OC_x <= 0.67 & spdf_in$HC_y >= 0.75 & spdf_in$HC_y < 1.5)
  # tannins
  subsel[["tannins"]] <- which(spdf_in$OC_x > 0.67 & spdf_in$OC_x <= 0.97 & spdf_in$HC_y >= 0.53 & spdf_in$HC_y < 1.5)

  vars <- c("lipids", "proteins", "amino_sugars", "carbs","cond_aromtx", "lignin", "tannins")

  # establish blank template data frame
  #df_out <- data.frame(sample = NA,group = NA,sex = NA, no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA)
  df_out <- data.frame(sample = NA,behav_PC1 = NA,sex = NA, no_fxns = NA, variable = vars, wtmean_OC_x = NA, wtmean_HC_y = NA) # depress_probs = NA, 

  for (v in 1:length(vars)) {
    #v<-1
    this_var <- vars[v]
    if (length(subsel[[ this_var ]]) > 0) {
      df_out$wtmean_OC_x[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$OC_x[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$wtmean_HC_y[v] <- sum( spdf_in$abun[ subsel[[ this_var ]] ]*spdf_in$HC_y[ subsel[[ this_var ]] ] )/sum(spdf_in$abun[ subsel[[ this_var ]] ])
      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )
    } else {
      df_out$wtmean_OC_x[v] <- NA
      df_out$wtmean_HC_y[v] <- NA

      df_out$no_fxns[v] <- length( (subsel[[ this_var ]]) )

    }
    print(paste0("completed ",v))
  }

  df_out$sample <- this_samp
  #df_out$group <- unique(spdf_in$group)
  #df_out$depress_probs  <- unique(spdf_in$depress_probs)
  df_out$behav_PC1  <- unique(spdf_in$behav_PC1)

  df_out$sex <- unique(spdf_in$sex)

  return(df_out)
}


time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps) ) %dopar%
  get_wtmean_vk_coords_Behav( spdf_in = spdf, samps = unique_samps)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 40
out[[1]]
#       sample behav_PC1    sex no_fxns     variable wtmean_OC_x wtmean_HC_y
# 1 SRR8204551 -1.874486 female      54       lipids  0.09324483   1.5887172
# 2 SRR8204551 -1.874486 female     247     proteins  0.38298282   1.7413661
# 3 SRR8204551 -1.874486 female     317 amino_sugars  0.63490186   1.7086169
# 4 SRR8204551 -1.874486 female     929        carbs  0.90463552   1.7715580
# 5 SRR8204551 -1.874486 female      19  cond_aromtx  0.16344934   0.8748231
# 6 SRR8204551 -1.874486 female     703       lignin  0.47734192   1.2442966
# 7 SRR8204551 -1.874486 female    1253      tannins  0.80496115   1.3338338

names(out[[1]])
# "sample"       "behav_PC1"     "sex"           "no_fxns"       "variable"      "wtmean_OC_x"   "wtmean_HC_y"

df.wtcoords <- out[[1]]
for (i in 2:length(out)) {
  df.wtcoords <- rbind(df.wtcoords, out[[i]])
  print(paste0("completed ",i))
}

dim(df.wtcoords) # 280 7

ok <- complete.cases(df.wtcoords)
sel.ok <- which(ok==TRUE) # qty 259

df.wtcoords <- df.wtcoords[sel.ok, ]



str(df.wtcoords)
# 'data.frame':	259 obs. of  7 variables:
# $ sample     : chr  "SRR8204551" "SRR8204551" "SRR8204551" "SRR8204551" ...
# $ behav_PC1  : num  -1.87 -1.87 -1.87 -1.87 -1.87 ...
# $ sex        : chr  "female" "female" "female" "female" ...
# $ no_fxns    : int  54 247 317 929 19 703 1253 68 254 303 ...
# $ variable   : chr  "lipids" "proteins" "amino_sugars" "carbs" ...
# $ wtmean_OC_x: num  0.0932 0.383 0.6349 0.9046 0.1634 ...
# $ wtmean_HC_y: num  1.589 1.741 1.709 1.772 0.875 ...

saveRDS(object = df.wtcoords, file = "df.wtcoords-Flannery-behav.RDS")


dat <- df.wtcoords

unique(dat$variable)
# "lipids"       "proteins"     "amino_sugars" "carbs"        "cond_aromtx"  "lignin"       "tannins"


dat$variable <- factor(dat$variable, levels = c("lipids", "proteins", "amino_sugars", "carbs", "cond_aromtx", "lignin", "tannins"),
                       labels = c("Lipids", "Proteins", "Amino sugars", "Carbohydrates", "Condensed\naromatics", "Lignin", "Tannins"),
                       ordered = TRUE)

names(dat)
# "sample"       "behav_PC1"     "sex"           "no_fxns"       "variable"      "wtmean_OC_x"   "wtmean_HC_y"


dat$group <- NA

sel <- which(dat$sex == "female")
hist(dat$behav_PC1[sel])
median_val <- median(dat$behav_PC1[sel], na.rm = TRUE)
subsel <- which(dat$behav_PC1[sel] <= median_val)
dat$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(dat$behav_PC1[sel] > median_val)
dat$group[sel[subsel]] <- "high_PC1_behav"

sel <- which(dat$sex == "male")
hist(dat$behav_PC1[sel])
median_val <- median(dat$behav_PC1[sel], na.rm = TRUE)
subsel <- which(dat$behav_PC1[sel] <= median_val)
dat$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(dat$behav_PC1[sel] > median_val)
dat$group[sel[subsel]] <- "high_PC1_behav"


length(unique(dat$sample)) # 37


dat$group_mf <- paste0(dat$group,"__",dat$sex)


unique( dat$group_mf)
# "low_PC1_behav__female"  "low_PC1_behav__male"    "high_PC1_behav__male"   "high_PC1_behav__female"
dat$group_mf <- factor(dat$group_mf,
                                 levels = c("high_PC1_behav__female", "high_PC1_behav__male", "low_PC1_behav__female", "low_PC1_behav__male"),
                                 labels = c("High PC1 behav (F)", "High PC1 behav (M)", "Low PC1 behav (F)", "Low PC1 behav (M)"),
                                 ordered = TRUE)

unique(dat$group)
# "low_PC1_behav"  "high_PC1_behav"
dat$group <- factor(dat$group, 
                    levels = c("low_PC1_behav" , "high_PC1_behav"),
                    labels = c("Low PC1 behav" , "High PC1 behav"),
                    ordered= TRUE)


cols = c("Low PC1 behav" = "#7fbc41",  "High PC1 behav" = "#de77ae") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9


# Females

p <- ggplot(data = filter(dat, sex == "female"), aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  
  geom_density_2d(contour_var = "ndensity", bins = 3, alpha=.5) + # , alpha=.5
  
  geom_point()+ # alpha=0.5
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 2,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "PC1 of problem\nbehaviors (F)"))

p


this_study # "-Flannery-behav-"
header # "-vk-coords-wtmean-"

grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9d-a-Females.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")


cols = c("Low PC1 behav" = "#5aae61",  "High PC1 behav" = "#9970ab") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9


# Males

p <- ggplot(data = filter(dat, sex == "male"), aes(x = wtmean_OC_x, y = wtmean_HC_y, color = group) )+
  #https://ggplot2.tidyverse.org/reference/geom_density_2d.html
  # peak intensity is the same in each facet, use `contour_var = "ndensity"
  # after_stat(ndensity) - Density estimate, scaled to a maximum of 1
  
  geom_density_2d(contour_var = "ndensity", bins = 3, alpha=.5) + # , alpha=.5
  
  geom_point()+ # alpha=0.5
  labs(x="O:C molar ratio",
       y="H:C molar ratio") +
  scale_color_manual(values = cols) +
  theme_classic() +
  facet_wrap(facets = vars(variable), ncol = 4, scales = "free")+
  
  theme(
    legend.position = c(0.9, 0.25), # as fraction of plot dimensions
    #legend.position="bottom",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 2,r = 0,b = 2,l = 0,"pt"), size = rel(0.9)) # size = rel(0.7)
  ) +
  guides(colour = guide_legend(title = "PC1 of problem\nbehaviors (M)"))

p


this_study # "-Flannery-behav-"
header # "-vk-coords-wtmean-"

grid.text(label = "(b)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Weighted-vkZone-coords-",this_study,header,"-v9d-b-Males.tiff"), width = 20, height = 12, units = "cm", res=450, compression="lzw",type="cairo")




saveRDS(object = dat, file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9d.RDS"))

#dat <- readRDS(file = paste0("dat--Weighted-vkZone-coords-",this_study,header,"-v9d.RDS"))



table(dat$group_mf)/length(unique(dat$variable))
# High PC1 behav (F) High PC1 behav (M)  Low PC1 behav (F)  Low PC1 behav (M) 
#                 10                  8                 10                  9


# Females

dat.F <- filter(dat, sex == "female")



## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.F) # "sample"      "behav_PC1"   "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group"       "group_mf"   


levels(dat.F$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0001554 0.00899 0.1632  0.726
# Residual    18 0.0171445 0.99101              
# Total       19 0.0172999 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0000503 0.00005035 0.0979    999  0.755
# Residuals 18 0.0092580 0.00051434



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00008322 0.04844 0.9163  0.377
# Residual    18 0.00163467 0.95156              
# Total       19 0.00171789 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.00007221 7.2212e-05 1.9431    999  0.183
# Residuals 18 0.00066893 3.7163e-05 



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00002121 0.03536 0.6598  0.429
# Residual    18 0.00057856 0.96464              
# Total       19 0.00059977 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.00000114 1.1403e-06 0.0988    999  0.777
# Residuals 18 0.00020765 1.1536e-05 



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0000027 0.00075 0.0135  0.956
# Residual    18 0.0036392 0.99925              
# Total       19 0.0036419 1.00000  

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.0002606 2.6060e-04 4.0993    999  0.057 .
# Residuals 18 0.0011443 6.3572e-05                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0000774 0.00677 0.1227  0.812
# Residual    18 0.0113635 0.99323              
# Total       19 0.0114410 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq     F N.Perm Pr(>F)  
# Groups     1 0.0008013 0.00080133 4.212    999  0.046 *
# Residuals 18 0.0034245 0.00019025                      
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00003098 0.02531 0.4674  0.631
# Residual    18 0.00119294 0.97469              
# Total       19 0.00122392 1.00000  

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.00000725 7.2530e-06 0.1936    999  0.718
# Residuals 18 0.00067422 3.7457e-05    



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.F$variable == this_var)
x <- dat.F[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.F$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 1.2848e-05 0.06397 1.2302  0.298
# Residual    18 1.8798e-04 0.93603              
# Total       19 2.0083e-04 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 1.2920e-06 1.2920e-06 0.4876    999  0.491
# Residuals 18 4.7692e-05 2.6495e-06



# # # # # # # # # # # # #


# Males

dat.M <- filter(dat, sex == "male")


## PERMANOVA SIG DIFFERENCES ???

# PERMANOVA for different clusters??
# https://rstudio-pubs-static.s3.amazonaws.com/288625_b0cf85c5be974eb795cef778535289d8.html
# adonis(Data.3$values~ Data.3$grupo,method = "euclidian")

# # https://archetypalecology.wordpress.com/2018/02/21/permutational-multivariate-analysis-of-variance-permanova-in-r-preliminary/
# Permutational multivariate analysis of variance (PERMANOVA)
# is a non-parametric multivariate statistical test. It is
# used to compare groups of objects and test the null hypothesis
# that the centroids and dispersion of the groups as defined by
# measure space are equivalent for all groups. A rejection of the
# null hypothesis means that either the centroid and/or the spread
# of the objects is different between the groups. Hence the test
# is based on the prior calculation of the distance between any
# two objects included in your experiment.

names(dat.M) 
# [1] "sample"      "behav_PC1"   "sex"         "no_fxns"     "variable"    "wtmean_OC_x" "wtmean_HC_y" "group"      
# [9] "group_mf"

levels(dat.M$variable)
# [1] "Lipids"               "Proteins"             "Amino sugars"         "Carbohydrates"        "Condensed\naromatics"
# [6] "Lignin"               "Tannins"



## "Lipids"
this_var <- "Lipids"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.0003793 0.04493 0.7057  0.431
# Residual    15 0.0080626 0.95507              
# Total       16 0.0084419 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0004789 0.00047889 1.8172    999  0.213
# Residuals 15 0.0039529 0.00026352     



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs     R2      F Pr(>F)  
# samps.class  1 0.00032512 0.1577 2.8083  0.094 .
# Residual    15 0.00173659 0.8423                
# Total       16 0.00206171 1.0000                
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.00000707 7.0660e-06 0.1509    999  0.724
# Residuals 15 0.00070234 4.6822e-05 



## "Amino sugars"
this_var <- "Amino sugars"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00005518 0.07484 1.2135  0.253
# Residual    15 0.00068214 0.92516              
# Total       16 0.00073732 1.00000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq     F N.Perm Pr(>F)
# Groups     1 1.2000e-08 1.2500e-08 8e-04    999  0.984
# Residuals 15 2.3342e-04 1.5562e-05  



## "Carbohydrates"
this_var <- "Carbohydrates"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs    R2      F Pr(>F)
# samps.class  1 0.0004151 0.109 1.8351  0.205
# Residual    15 0.0033927 0.891              
# Total       16 0.0038078 1.000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0000241 2.4104e-05 0.2305    999  0.641
# Residuals 15 0.0015685 1.0457e-04 



## "Condensed\naromatics"
this_var <- "Condensed\naromatics"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs     R2      F Pr(>F)
# samps.class  1 0.0003758 0.0181 0.2765  0.649
# Residual    15 0.0203909 0.9819              
# Total       16 0.0207668 1.0000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df    Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 0.0004135 0.00041345 0.8346    999   0.37
# Residuals 15 0.0074305 0.00049537 



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df  SumOfSqs      R2     F Pr(>F)
# samps.class  1 0.0000089 0.00518 0.078  0.887
# Residual    15 0.0017102 0.99482             
# Total       16 0.0017191 1.00000

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)  
# Groups     1 0.00015023 1.5023e-04 3.5965    999  0.067 .
# Residuals 15 0.00062654 4.1769e-05                       
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.M$variable == this_var)
x <- dat.M[sel, c("wtmean_OC_x","wtmean_HC_y") ]
samps.class <- dat.M$group[sel]

set.seed(123)
(adon <- adonis2(formula = x ~ samps.class,method = "euclidian"))
# Permutation test for adonis under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# adonis2(formula = x ~ samps.class, method = "euclidian")
#             Df   SumOfSqs      R2      F Pr(>F)
# samps.class  1 0.00000959 0.05155 0.8153  0.432
# Residual    15 0.00017645 0.94845              
# Total       16 0.00018604 1.00000 

beta <- betadisper(d = vegdist(x = x, method = "euclidean") , group = samps.class)
set.seed(123)
(betadisp <- permutest(beta))
# Permutation test for homogeneity of multivariate dispersions
# Permutation: free
# Number of permutations: 999
# 
# Response: Distances
#           Df     Sum Sq    Mean Sq      F N.Perm Pr(>F)
# Groups     1 6.6200e-07 6.6233e-07 0.2988    999  0.602
# Residuals 15 3.3255e-05 2.2170e-06 

#-------------------------

#### Flannery behaviour - ANCOMBC - Function-level
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
#df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")


this_study <- "-Flannery-behav-"
header <- "-ANCOMBC-fxns-"

# also use: df.fxn_vk_coords.noNAs
df.fxn_vk_coords.noNAs <- readRDS("df.fxn_vk_coords.noNAs.Flannery-child-behav.RDS")

# PC1 behav dat ais in here:
# use clean data
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean

phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20787 taxa and 40 samples ]
# sample_data() Sample Data:       [ 40 samples by 77 sample variables ]
# tax_table()   Taxonomy Table:    [ 20787 taxa by 4 taxonomic ranks ]


head(phy@sam_data) # host_sex

dim(phy@sam_data) # 40 77

phy@sam_data$sex <- phy@sam_data$host_sex
phy@sam_data$behav_PC1 <- NA

for (i in 1:dim(phy@sam_data)[1]) {
  #i<-1
  this_samp <- phy@sam_data$Run[i]
  sel.vk <- which(vk.samp.noNAs.long.clean$sample == this_samp)
  phy@sam_data$behav_PC1[i] <- unique(vk.samp.noNAs.long.clean$behav_PC1[sel.vk])
  print(paste0("completed ",i))
}

phy@sam_data$group <- NA

sel <- which(phy@sam_data$sex == "female") # qty 23
hist(phy@sam_data$behav_PC1[sel])
median_val <- median(phy@sam_data$behav_PC1[sel], na.rm = TRUE)
subsel <- which(phy@sam_data$behav_PC1[sel] <= median_val) # 10
phy@sam_data$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(phy@sam_data$behav_PC1[sel] > median_val) # 10
phy@sam_data$group[sel[subsel]] <- "high_PC1_behav"

sel <- which(phy@sam_data$sex == "male")
hist(phy@sam_data$behav_PC1[sel])
median_val <- median(phy@sam_data$behav_PC1[sel], na.rm = TRUE)
subsel <- which(phy@sam_data$behav_PC1[sel] <= median_val) # 9
phy@sam_data$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(phy@sam_data$behav_PC1[sel] > median_val) # 8
phy@sam_data$group[sel[subsel]] <- "high_PC1_behav"


length(unique(phy@sam_data$Run)) # 40

sel.ok <- which(!is.na(phy@sam_data$behav_PC1)) # qty 37

phy.ok <- prune_samples(samples = row.names(phy@sam_data)[sel.ok], phy)

# prune taxa with zero reads
phy.ok <- prune_taxa(taxa = taxa_sums(phy.ok) > 0, x = phy.ok)
phy.ok
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20599 taxa and 37 samples ]
# sample_data() Sample Data:       [ 37 samples by 80 sample variables ]
# tax_table()   Taxonomy Table:    [ 20599 taxa by 4 taxonomic ranks ]


# add group2
phy.ok@sam_data$group2 <- phy.ok@sam_data$group
table(phy.ok@sam_data$group2, useNA="ifany")
# high_PC1_behav  low_PC1_behav 
# 18             19 

phy.ok@sam_data$group2 <- factor(phy.ok@sam_data$group2,
                              levels = c("low_PC1_behav", "high_PC1_behav"),
                              labels = c("A_low_PC1_behav", "B_high_PC1_behav"),
                              ordered = TRUE)
unique(phy.ok@sam_data$group2)
# [1] A_low_PC1_behav  B_high_PC1_behav
# Levels: A_low_PC1_behav < B_high_PC1_behav




## Females

sel.ok.f <- which(phy.ok@sam_data$sex == "female") # qty 20

phy.ok.f <- prune_samples(samples = row.names(phy.ok@sam_data)[sel.ok.f], phy.ok)

# prune taxa with zero reads
phy.ok.f <- prune_taxa(taxa = taxa_sums(phy.ok.f) > 0, x = phy.ok.f)
phy.ok.f
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 18775 taxa and 20 samples ]
# sample_data() Sample Data:       [ 20 samples by 81 sample variables ]
# tax_table()   Taxonomy Table:    [ 18775 taxa by 4 taxonomic ranks ]


## ANCOMBC - differential abundance testing??
set.seed(123)
out.ancombc.fxns = ancombc2(data = phy.ok.f,
                            tax_level = NULL,
                            fix_formula = "group2",
                            p_adj_method = "holm", prv_cut = 0.1, lib_cut = 0,
                            group = "group2", struc_zero = TRUE, neg_lb = TRUE,
                            alpha = 0.05, global = TRUE,  # change alpha level to reduce number !!!
                            n_cl = 6, verbose = TRUE)

# `tax_level` is not speficified 
# No agglomeration will be performed
# Otherwise, please speficy `tax_level` by one of the following: 
#   
#   Obtaining initial estimates ...
# Estimating sample-specific biases ...
# ANCOM-BC2 primary results ...
# The sensitivity analysis for the pseudo-count addition ...
# Warning message:
#   The group variable has < 3 categories 
# The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated

str(out.ancombc.fxns)
# List of 11
# $ feature_table  : num [1:12032, 1:20] 0 0.001566 0 0.000153 0 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:12032] "fxn_2" "fxn_4" "fxn_7" "fxn_8" ...
# .. ..$ : chr [1:20] "SRR8204551" "SRR8204552" "SRR8204555" "SRR8204559" ...
# $ zero_ind       :'data.frame':	18775 obs. of  3 variables:
#   ..$ taxon                                      : chr [1:18775] "fxn_2" "fxn_3" "fxn_4" "fxn_5" ...
# ..$ structural_zero (group2 = A_low_PC1_behav) : logi [1:18775] FALSE TRUE FALSE TRUE TRUE FALSE ...
# ..$ structural_zero (group2 = B_high_PC1_behav): logi [1:18775] FALSE TRUE FALSE TRUE TRUE FALSE ...
# $ samp_frac      : num [1:20] 0.2014 0.0424 -0.2971 -0.9124 -1.6936 ...
# $ delta_em       : num [1:2] 5.99e-18 3.38e-02
# $ delta_wls      : num [1:2] 1.46e-18 3.38e-02
# $ pseudo_sens_tab:'data.frame':	12032 obs. of  3 variables:
#   ..$ taxon      : chr [1:12032] "fxn_2" "fxn_4" "fxn_7" "fxn_8" ...
# ..$ (Intercept): num [1:12032] 1.48e-16 1.65e-16 1.28e-16 1.79e-16 1.42e-16 ...
# ..$ group2.L   : num [1:12032] 0.0162 0.0904 0.0496 0.0393 0.0742 ...
# $ res            :'data.frame':	12032 obs. of  13 variables:
#   ..$ taxon           : chr [1:12032] "fxn_2" "fxn_4" "fxn_7" "fxn_8" ...
# ..$ lfc_(Intercept) : num [1:12032] -2.16e-16 3.31e-16 -1.45e-16 -5.31e-16 3.77e-17 ...
# ..$ lfc_group2.L    : num [1:12032] 0.359 0.341 0.914 0.702 1.24 ...
# ..$ se_(Intercept)  : num [1:12032] 1.045 0.389 1.163 1.026 1.112 ...
# ..$ se_group2.L     : num [1:12032] 1.478 0.551 1.645 1.451 1.573 ...
# ..$ W_(Intercept)   : num [1:12032] -2.07e-16 8.49e-16 -1.24e-16 -5.18e-16 3.39e-17 ...
# ..$ W_group2.L      : num [1:12032] 0.243 0.619 0.555 0.484 0.789 ...
# ..$ p_(Intercept)   : num [1:12032] 1 1 1 1 1 ...
# ..$ p_group2.L      : num [1:12032] 0.808 0.536 0.579 0.629 0.43 ...
# ..$ q_(Intercept)   : num [1:12032] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ q_group2.L      : num [1:12032] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ diff_(Intercept): logi [1:12032] FALSE FALSE FALSE FALSE FALSE FALSE ...
# ..$ diff_group2.L   : logi [1:12032] FALSE FALSE FALSE FALSE FALSE FALSE ...
# $ res_global     : NULL
# $ res_pair       : NULL
# $ res_dunn       : NULL
# $ res_trend      : NULL

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = out.ancombc.fxns, file = "out.ancombc.fxns--Child_behav.Females.RDS")

#out.ancombc.fxns <- readRDS("out.ancombc.fxns--Child_behav.Females.RDS")

out <- out.ancombc.fxns


out$feature_table[1:5, 1:5]
#         SRR8204551   SRR8204552   SRR8204555   SRR8204559   SRR8204563
# fxn_2 0.0000000000 4.608337e-05 0.000000e+00 5.938193e-05 0.0002315611
# fxn_4 0.0015655056 7.680562e-04 2.119524e-03 9.896988e-04 0.0077658735
# fxn_7 0.0000000000 0.000000e+00 0.000000e+00 1.979398e-05 0.0000000000
# fxn_8 0.0001533557 4.608337e-05 2.627509e-05 0.000000e+00 0.0000000000
# fxn_9 0.0000000000 6.912506e-05 0.000000e+00 0.000000e+00 0.0000000000

fxns_out <- row.names(out$feature_table)

dim(out$feature_table) # 12032    20
phy.ok.f@otu_table[1:5, 1:5]
# OTU Table:          [5 taxa and 5 samples]
# taxa are rows
# SRR8204551   SRR8204552  SRR8204555   SRR8204559   SRR8204563
# fxn_2 0.000000000 4.608337e-05 0.000000000 5.938193e-05 2.315611e-04
# fxn_3 0.000000000 0.000000e+00 0.000000000 0.000000e+00 0.000000e+00
# fxn_4 0.001565506 7.680562e-04 0.002119524 9.896988e-04 7.765874e-03
# fxn_5 0.000000000 0.000000e+00 0.000000000 0.000000e+00 3.216127e-06
# fxn_6 0.000000000 0.000000e+00 0.000000000 0.000000e+00 0.000000e+00 


#  ANCOM-BC all results
res = out$res

class(res) # "data.frame"
names(res)
# [1] "taxon"            "lfc_(Intercept)"  "lfc_group2.L"     "se_(Intercept)"   "se_group2.L"     
# [6] "W_(Intercept)"    "W_group2.L"       "p_(Intercept)"    "p_group2.L"       "q_(Intercept)"   
# [11] "q_group2.L"       "diff_(Intercept)" "diff_group2.L"

tab_qval <- res[ ,c("taxon", "q_(Intercept)", "q_group2.L")]
dim(tab_qval) # 12032     3

summary( tab_qval )
# taxon           q_(Intercept)   q_group2.L   
# Length:12032       Min.   :1     Min.   :0.000  
# Class :character   1st Qu.:1     1st Qu.:1.000  
# Mode  :character   Median :1     Median :1.000  
#                    Mean   :1     Mean   :0.997  
#                    3rd Qu.:1     3rd Qu.:1.000  
#                    Max.   :1     Max.   :1.000

names(tab_qval) # "taxon"         "q_(Intercept)" "q_group2.L" 
# Levels: A_low_PC1_behav < B_high_PC1_behav
names(tab_qval) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_qval) <- tab_qval$taxon

head(tab_qval)
#         taxon Intercept High PC1 cf. Low PC1
# fxn_2   fxn_2         1                    1
# fxn_4   fxn_4         1                    1
# fxn_7   fxn_7         1                    1
# fxn_8   fxn_8         1                    1
# fxn_9   fxn_9         1                    1
# fxn_10 fxn_10         1                    1

# sig diff abun Fxns in High PC1 cf. Low PC1
sel.sig.HighPC1_cf_LowPC1 <- which(tab_qval$`High PC1 cf. Low PC1` <= 0.05) # qty 22
fxn.sig.HighPC1_cf_LowPC1 <- tab_qval$taxon[sel.sig.HighPC1_cf_LowPC1]
length(unique(fxn.sig.HighPC1_cf_LowPC1)) # 22


sel.sig <- unique(c(sel.sig.HighPC1_cf_LowPC1)) # qty 22
tab_qval[sel.sig, ]
#               taxon Intercept High PC1 cf. Low PC1
# fxn_118     fxn_118         1         2.243352e-03
# fxn_341     fxn_341         1         1.334659e-02
# fxn_532     fxn_532         1         3.115911e-09
# fxn_680     fxn_680         1         1.591804e-03
# fxn_2426   fxn_2426         1         4.153060e-02
# fxn_4797   fxn_4797         1         3.097754e-06
# fxn_5232   fxn_5232         1         4.067948e-03
# fxn_6310   fxn_6310         1         1.305863e-02
# fxn_7446   fxn_7446         1         2.071414e-03
# fxn_7724   fxn_7724         1         7.585022e-04
# fxn_8150   fxn_8150         1         1.304616e-06
# fxn_8356   fxn_8356         1         3.992178e-10
# fxn_9046   fxn_9046         1         1.358024e-04
# fxn_9072   fxn_9072         1         1.423373e-02
# fxn_10516 fxn_10516         1         2.062671e-02
# fxn_11346 fxn_11346         1         1.441931e-04
# fxn_11530 fxn_11530         1         6.820381e-05
# fxn_11660 fxn_11660         1         9.716419e-03
# fxn_16658 fxn_16658         1         2.133237e-05
# fxn_18172 fxn_18172         1         1.719516e-09
# fxn_18378 fxn_18378         1         5.349147e-03
# fxn_18462 fxn_18462         1         1.885600e-02



## log-fold change (LFC)
tab_lfc = res[ ,c("taxon", "lfc_(Intercept)",  "lfc_group2.L" )]
class(tab_lfc) # "data.frame"
names(tab_lfc) # "taxon"           "lfc_(Intercept)" "lfc_group2.L"
names(tab_lfc) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_lfc) <- tab_lfc$taxon
tab_lfc[sel.sig, ]


## get significant diff abun taxa across any pairs
tab_lfc.sig_all <- tab_lfc[ sel.sig, ]

dim(tab_lfc.sig_all) # 22  3

head(tab_lfc.sig_all)
#             taxon     Intercept High PC1 cf. Low PC1
# fxn_118   fxn_118  4.598028e-16           -0.5107555
# fxn_341   fxn_341  2.436250e-16            0.5932198
# fxn_532   fxn_532 -3.839909e-16            0.7338498
# fxn_680   fxn_680 -3.355944e-16            0.7591087
# fxn_2426 fxn_2426 -2.225111e-16            0.4815305
# fxn_4797 fxn_4797  1.433359e-16            0.6131127

hist(tab_lfc.sig_all$Intercept); summary(tab_lfc.sig_all$Intercept)
# Min.    1st Qu.     Median       Mean    3rd Qu.       Max. 
# -4.286e-16 -2.834e-16 -9.541e-17 -4.461e-17  2.186e-16  4.598e-16 
hist(tab_lfc.sig_all$`High PC1 cf. Low PC1`); summary(tab_lfc.sig_all$`High PC1 cf. Low PC1`)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.0632 -0.2494  0.5759  0.2734  0.6687  0.9616



names(tab_lfc.sig_all)[2] # "Intercept"

tab_lfc.sig_all.melt <- melt(tab_lfc.sig_all[ ,-2])
names(tab_lfc.sig_all.melt) # "taxon"    "variable" "value" 
names(tab_lfc.sig_all.melt) <- c("taxon", "Comparison groups", "Log-fold-change" )

p <- ggplot(tab_lfc.sig_all.melt, aes(`Log-fold-change`))+ #, color = `Comparison groups`)) +
  geom_density()+
  labs(x = "Log-fold-change", y = "Density") +
  # scale_colour_manual( values = c("High PC1 cf. Low PC1" = "#66c2a5"),
  #                      labels = c("High PC1 cf. Low PC1")) +
  theme_classic() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
p
#grid.text(label = "(a)", x = unit(0.04, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=13, fontface="bold") )
grid.text(label = "ANCOMBC: Function-level | Females", x = unit(0.17, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=9), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","LFC-density-plots",header,this_study,"Females-v9d.tiff"), width = 9, height = 6, units = "cm", res=450, compression="lzw",type="cairo")


# assign "Positive LFC", "Negative LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- NA
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` < 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Negative LFC"
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` > 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Positive LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- factor(tab_lfc.sig_all.melt$pos_or_neg_LFC, levels = c("Positive LFC", "Negative LFC"), ordered = TRUE)


head(tab_lfc.sig_all.melt)
#      taxon    Comparison groups Log-fold-change pos_or_neg_LFC
# 1  fxn_118 High PC1 cf. Low PC1      -0.5107555   Negative LFC
# 2  fxn_341 High PC1 cf. Low PC1       0.5932198   Positive LFC
# 3  fxn_532 High PC1 cf. Low PC1       0.7338498   Positive LFC
# 4  fxn_680 High PC1 cf. Low PC1       0.7591087   Positive LFC
# 5 fxn_2426 High PC1 cf. Low PC1       0.4815305   Positive LFC
# 6 fxn_4797 High PC1 cf. Low PC1       0.6131127   Positive LFC


table(tab_lfc.sig_all.melt$pos_or_neg_LFC, useNA = "ifany")
# Positive LFC Negative LFC 
#           16           6 



## join vK coords to data table

# from earlier ...
head(df.fxn_vk_coords.noNAs)
#       tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.6000000     1.8000000
# fxn_3     1.0000000     0.4285714
# fxn_4     0.6666667     2.3333333
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333
# fxn_7     0.6666667     2.3333333
row.names(df.fxn_vk_coords.noNAs)

res.sig <- tab_lfc.sig_all.melt

res.sig$OC_x <- NA
res.sig$HC_y <- NA

length(res.sig$taxon) # 22
length( which(res.sig$taxon %in% row.names(df.fxn_vk_coords.noNAs)) ) # 15
100*(15/22)
# 68.18182 %

dim(res.sig) # 22  6

for (i in 1:dim(res.sig)[1]) {
  #i<-1
  this_fxn <- res.sig$taxon[i]
  sel.fxn_vk <- which( row.names(df.fxn_vk_coords.noNAs) == this_fxn ) #
  
  if (length(sel.fxn_vk) ==1) {
    res.sig$OC_x[i] <- df.fxn_vk_coords.noNAs$tot_mean_OC_x[sel.fxn_vk]
    res.sig$HC_y[i] <- df.fxn_vk_coords.noNAs$tot_mean_HC_y[sel.fxn_vk]
  } else {
    res.sig$OC_x[i] <- NA
    res.sig$HC_y[i] <- NA
  }
  print(paste0("completed ",i))
}

unique( res.sig$`Comparison groups` )
# [1] High PC1 cf. Low PC1
# Levels: High PC1 cf. Low PC1

# res.sig$`Comparison groups` <- factor(res.sig$`Comparison groups`, 
#                                       levels = c("IGT cf. Normal", "T2D Met neg cf. Normal", "T2D Met pos cf. Normal"),
#                                       labels = c("IGT cf. Normal", "T2D Met- cf. Normal", "T2D Met+ cf. Normal"),
#                                       ordered=TRUE)

ok <- complete.cases(res.sig)
sel.ok <- which(ok==TRUE) # qty 15

res.sig.ok <- res.sig[sel.ok, ]

dim(res.sig)   #  22  6
dim(res.sig.ok)#  15  6

names(res.sig.ok)
# "taxon"             "Comparison groups" "Log-fold-change"   "pos_or_neg_LFC"    "OC_x"              "HC_y"             


summary(res.sig.ok$`Log-fold-change`[which(res.sig.ok$`Comparison groups`=="High PC1 cf. Low PC1")]);hist(res.sig.ok$`Log-fold-change`[which(res.sig.ok$`Comparison groups`=="High PC1 cf. Low PC1")])
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# -0.79665 -0.01461  0.61311  0.30775  0.67529  0.96163 



p <- ggplot(res.sig.ok, aes(x = OC_x, y = HC_y, color = `Log-fold-change`))+
  geom_point(alpha = 0.2)

p


# overlay vK ref data ...

# sel <- which(df.vk.ref$sample %in% c("Acetate", "Propionate",  "Butyrate", "Riboflavin (Vit B2)" ,
#                                      "Cobalamin (Vit B12)", "Pyridoxal 5'-phosphate (Vit B6)", "Folate (Vit B9)", "Menaquinone (Vit K2)" ) )
# df.hac <- df.vk.ref[sel, ]
df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2

vkgrouprect

#library(colorspace)

p <- ggplot()+
  
  geom_point(data = filter(res.sig.ok, `Comparison groups`=="High PC1 cf. Low PC1"), aes(x = OC_x, y = HC_y, color = `Log-fold-change`), size = 3.5)+ # , alpha = 0.6
  
  ## ADJUST SCALE !!!!!!!!!!
  scale_color_continuous_sequential(palette = "BluYl", rev=FALSE, name="Log-fold\n-change", limits = c(-1, 1) )+ # palette = "YlGnBu" "BluYl"
  
  guides(color = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5))+
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  #geom_point(data = df.hac ,aes(x=OC_x, y=HC_y), color = "#253494")+ # , pch=8  "#80b1d3"
  #geom_text(data = df.hac ,aes(x=OC_x, y=HC_y, label=sample), color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  geom_segment(aes(x = 1.1, y = 1.5, xend = 1.8,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  #geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  #geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.75, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "High cf. Low PC1 problem behaviors: Functions | Females", x = unit(0.07, "npc") , y = unit(0.96,"npc"), gp=gpar(fontsize=10), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","vK-fxn-logfoldchange-scatterpoints-High-vs-Low-PC1-Problem-Behav-Females--v9d.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")




## plot mean rel abun of each group alongside log-fold-change - for tab_lfc.sig_global


head(tab_lfc.sig_all)
#             taxon     Intercept High PC1 cf. Low PC1
# fxn_118   fxn_118  4.598028e-16           -0.5107555
# fxn_341   fxn_341  2.436250e-16            0.5932198
# fxn_532   fxn_532 -3.839909e-16            0.7338498
# fxn_680   fxn_680 -3.355944e-16            0.7591087
# fxn_2426 fxn_2426 -2.225111e-16            0.4815305
# fxn_4797 fxn_4797  1.433359e-16            0.6131127

temp <- tab_lfc.sig_all

# add vK coordinates
temp$OC_x <- NA
temp$HC_y <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_fxn <- temp$taxon[i]
  sel.fxn_vk <- which( row.names(df.fxn_vk_coords.noNAs) == this_fxn )
  
  if (length(sel.fxn_vk) ==1) {
    temp$OC_x[i] <- df.fxn_vk_coords.noNAs$tot_mean_OC_x[sel.fxn_vk]
    temp$HC_y[i] <- df.fxn_vk_coords.noNAs$tot_mean_HC_y[sel.fxn_vk]
  } else {
    temp$OC_x[i] <- NA
    temp$HC_y[i] <- NA
  }
  print(paste0("completed ",i))
}

# add matching compound?
temp$coords <- paste0(temp$OC_x,"__",temp$HC_y)

out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")
head( out[[1]] )
out.temp <- out[[1]]
row.names(out.temp) <- out.temp$coords

temp$compound <- NA

for (i in 1:dim(temp)[1]) {
  #i<-2
  this_coord <- temp$coords[i]
  if (is.na(temp$OC_x[i])){
    temp$compound[i] <- NA
  } else {
    sel <- which(out.temp$coords == this_coord)
    temp$compound[i] <- out.temp$compound[sel]
  }
  print(paste0("completed ",i))
}

head(temp)

temp <- cbind( temp, as.data.frame( phy.ok.f@tax_table[ temp$taxon , ]))

write.csv(x = temp, file = "ANCOMBC-Function-level-Sig-diff-abun-fxns--Child-Behav-Females-LFC-and-Function-details.csv")



## plot heatmap of LFC for sig diff abun Functions - ANCOMBC Function-level

m2 <- tab_lfc.sig_all[ ,-1] # exclude 'taxon' column, it is in row.name anyway
names(m2) #  "Intercept"            "High PC1 cf. Low PC1"
names(m2) <- c("Intercept", "High PC1\ncf. Low PC1" )

m2 <- as.matrix(m2)
colnames(m2) # "Intercept"            "High PC1\ncf. Low PC1"
m2
class(m2) # "matrix" "array" 

row.names(m2)
identical(row.names(m2), temp$taxon)
sel.ok <- which(!is.na(temp$OC_x))
row.names(m2)[sel.ok] <- paste0(row.names(m2)[sel.ok], "#")


pheatmap(m2, cluster_cols = FALSE, #cluster_row = FALSE, 
         #color = color,
         color = hcl.colors(50, "BluYl"), # palette = "BrBG", direction = 1
         border_color = FALSE, 
         fontsize_col = 12,
         main = "Differentially abundant functions\n(log-fold-changes): Females"
)

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-LogFoldChange-ANCOMBC-Function-level-Child-Behav-Females--v9d.tiff"), width = 9, height = 14, units = "cm", res=450, compression="lzw",type="cairo")



ph <- pheatmap(m2, cluster_cols = FALSE)
ph_taxa_rownames <- ph$tree_row$labels[ ph$tree_row$order ]

ph_taxa_rownames <- gsub(pattern = "#", replacement = "", x = ph_taxa_rownames)



## compare with relative abundance of functions in these groups

phy.relabun <- phy.ok.f

m <- as.matrix( as.data.frame(phy.relabun@otu_table[ ph_taxa_rownames , ]))
identical(colnames(m), phy.relabun@sam_data$Run) # TRUE

names( phy.relabun@sam_data )
# "Sample"         "Country subset" "Status"         "Bases"          "Run"            "group"          "group2"

levels( phy.relabun@sam_data$group2 ) 
# "A_low_PC1_behav"  "B_high_PC1_behav"
table( phy.relabun@sam_data$group ) 
# high_PC1_behav  low_PC1_behav 
#             10             10 

phy.relabun@sam_data$label <- phy.relabun@sam_data$group
phy.relabun@sam_data$label <- factor(phy.relabun@sam_data$label,
                                     levels = c("low_PC1_behav","high_PC1_behav"),
                                     labels = c("Low PC1","High PC1"),
                                     ordered=TRUE)
sample_order <- order(phy.relabun@sam_data$group2)
phy.relabun@sam_data$group_label <- paste0(phy.relabun@sam_data$label," (",phy.relabun@sam_data$Run,")")
phy.relabun@sam_data$group_label[sample_order]
phy.relabun@sam_data$group_label <- factor(phy.relabun@sam_data$group_label,
                                           levels = phy.relabun@sam_data$group_label[sample_order],
                                           ordered=TRUE)
colnames(m) <- phy.relabun@sam_data$group_label
# reorder columns
m <- m[ , sample_order]

# No dendrogram nor reordering for neither column or row
heatmap(m, Colv = NA, Rowv = NA, scale="column")
heatmap(m, Colv = NA, Rowv = NA)

# https://www.reneshbedre.com/blog/heatmap-with-pheatmap-package-r.html
#library(pheatmap)

row.names(m)
identical(row.names(m), ph_taxa_rownames) # TRUE
identical( row.names(m), gsub(pattern="#",replacement="", x= ph$tree_row$labels[ ph$tree_row$order ]) ) # TRUE
row.names(m) <- ph$tree_row$labels[ ph$tree_row$order ]
# use same '#' notation as above for functions that map to vK space


pheatmap(m, cluster_cols = FALSE,  cluster_row = FALSE, # gaps_col = c(6,12,18), 
         border_color = FALSE, main = "Function relative abundance (%)")


group_labels <- data.frame( Group = phy.relabun@sam_data$label[sample_order] )
row.names(group_labels) <- colnames(m)


cols = c("Low PC1" = "#7fbc41",  "High PC1" = "#de77ae") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9

mycolors <- c("#7fbc41","#de77ae")
names(mycolors) <- c("Low PC1","High PC1")


mycolors <- list(Group = mycolors)

color <- colorRampPalette((c("#fff5f0","#fc9272", "#cb181d")))(50)
pheatmap(m, gaps_col = c(10), cluster_cols = FALSE, cluster_row = FALSE, # 49          43          33          20
         annotation_col = group_labels,
         annotation_colors = mycolors,
         color = color,
         fontsize_col = 8,
         border_color = FALSE, 
         main = "Rel. abun. of functions (%): Females")

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-RelAbun-for-ANCOMBC-Function-level-Child-Behav-Females--v9.tiff"), width = 14, height = 14, units = "cm", res=450, compression="lzw",type="cairo")






## Males


sel.ok.m <- which(phy.ok@sam_data$sex == "male") # qty 17

phy.ok.m <- prune_samples(samples = row.names(phy.ok@sam_data)[sel.ok.m], phy.ok)

# prune taxa with zero reads
phy.ok.m <- prune_taxa(taxa = taxa_sums(phy.ok.m) > 0, x = phy.ok.m)
phy.ok.m
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 18851 taxa and 17 samples ]
# sample_data() Sample Data:       [ 17 samples by 81 sample variables ]
# tax_table()   Taxonomy Table:    [ 18851 taxa by 4 taxonomic ranks ]


## ANCOMBC - differential abundance testing??
set.seed(123)
out.ancombc.fxns = ancombc2(data = phy.ok.m,
                            tax_level = NULL,
                            fix_formula = "group2",
                            p_adj_method = "holm", prv_cut = 0.1, lib_cut = 0,
                            group = "group2", struc_zero = TRUE, neg_lb = TRUE,
                            alpha = 0.05, global = TRUE,  # change alpha level to reduce number !!!
                            n_cl = 6, verbose = TRUE)

# `tax_level` is not speficified 
# No agglomeration will be performed
# Otherwise, please speficy `tax_level` by one of the following: 
#   
#   Obtaining initial estimates ...
# Estimating sample-specific biases ...
# ANCOM-BC2 primary results ...
# The sensitivity analysis for the pseudo-count addition ...
# Warning message:
#   The group variable has < 3 categories 
# The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated

str(out.ancombc.fxns)
# List of 11
# $ feature_table  : num [1:10804, 1:17] 0 0.00557 0 0 0.00426 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:10804] "fxn_2" "fxn_4" "fxn_8" "fxn_9" ...
# .. ..$ : chr [1:17] "SRR8204553" "SRR8204554" "SRR8204556" "SRR8204557" ...
# $ zero_ind       :'data.frame':	18851 obs. of  3 variables:
#   ..$ taxon                                      : chr [1:18851] "fxn_1" "fxn_2" "fxn_3" "fxn_4" ...
# ..$ structural_zero (group2 = A_low_PC1_behav) : logi [1:18851] TRUE FALSE TRUE FALSE TRUE TRUE ...
# ..$ structural_zero (group2 = B_high_PC1_behav): logi [1:18851] TRUE FALSE TRUE FALSE TRUE FALSE ...
# $ samp_frac      : num [1:17] 0.337 -0.666 0.524 -0.715 3.654 ...
# $ delta_em       : num [1:2] -0.0126 -0.2746
# $ delta_wls      : num [1:2] -0.0126 -0.2774
# $ pseudo_sens_tab:'data.frame':	10804 obs. of  3 variables:
#   ..$ taxon      : chr [1:10804] "fxn_2" "fxn_4" "fxn_8" "fxn_9" ...
# ..$ (Intercept): num [1:10804] 0.004557 0.00299 0.002543 0.001227 0.000462 ...
# ..$ group2.L   : num [1:10804] 0.1567 0.0653 0.0878 0.02 0.0296 ...
# $ res            :'data.frame':	10804 obs. of  13 variables:
#   ..$ taxon           : chr [1:10804] "fxn_2" "fxn_4" "fxn_8" "fxn_9" ...
# ..$ lfc_(Intercept) : num [1:10804] -0.0957 0.00302 -0.0788 -0.03788 -0.01366 ...
# ..$ lfc_group2.L    : num [1:10804] -2.3286 0.0448 -1.9224 -0.9386 -0.3564 ...
# ..$ se_(Intercept)  : num [1:10804] 0.944 0.391 1.064 1.249 0.334 ...
# ..$ se_group2.L     : num [1:10804] 1.335 0.553 1.505 1.767 0.472 ...
# ..$ W_(Intercept)   : num [1:10804] -0.10139 0.00773 -0.07406 -0.03032 -0.04095 ...
# ..$ W_group2.L      : num [1:10804] -1.7445 0.0809 -1.2776 -0.5313 -0.7552 ...
# ..$ p_(Intercept)   : num [1:10804] 0.919 0.994 0.941 0.976 0.967 ...
# ..$ p_group2.L      : num [1:10804] 0.0811 0.9355 0.2014 0.5952 0.4501 ...
# ..$ q_(Intercept)   : num [1:10804] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ q_group2.L      : num [1:10804] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ diff_(Intercept): logi [1:10804] FALSE FALSE FALSE FALSE FALSE FALSE ...
# ..$ diff_group2.L   : logi [1:10804] FALSE FALSE FALSE FALSE FALSE FALSE ...
# $ res_global     : NULL
# $ res_pair       : NULL
# $ res_dunn       : NULL
# $ res_trend      : NULL

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = out.ancombc.fxns, file = "out.ancombc.fxns--Child_behav.Males.RDS")

#out.ancombc.fxns <- readRDS("out.ancombc.fxns--Child_behav.Males.RDS")

out <- out.ancombc.fxns



out$feature_table[1:5, 1:5]
# SRR8204553   SRR8204554   SRR8204556   SRR8204557 SRR8204558
# fxn_2  0.000000000 2.111550e-04 1.870998e-04 0.0001831194          0
# fxn_4  0.005571716 1.677509e-03 1.559165e-03 0.0003440425          0
# fxn_8  0.000000000 1.231738e-04 3.741997e-04 0.0000000000          0
# fxn_9  0.000000000 5.278876e-05 4.677496e-05 0.0000000000          0
# fxn_10 0.004258986 1.888665e-03 1.083620e-03 0.0026496821          0

fxns_out <- row.names(out$feature_table)

dim(out$feature_table) # 10804    17
phy.ok.m@otu_table[1:5, 1:5]
# OTU Table:          [5 taxa and 5 samples]
# taxa are rows
# SRR8204553  SRR8204554   SRR8204556   SRR8204557 SRR8204558
# fxn_1 0.000000000 0.000000000 3.118331e-05 0.0000000000          0
# fxn_2 0.000000000 0.000211155 1.870998e-04 0.0001831194          0
# fxn_3 0.000000000 0.000000000 0.000000e+00 0.0000000000          0
# fxn_4 0.005571716 0.001677509 1.559165e-03 0.0003440425          0
# fxn_6 0.000000000 0.000000000 0.000000e+00 0.0000000000          0


#  ANCOM-BC all results
res = out$res

class(res) # "data.frame"
names(res)
# [1] "taxon"            "lfc_(Intercept)"  "lfc_group2.L"     "se_(Intercept)"   "se_group2.L"      "W_(Intercept)"    "W_group2.L"       "p_(Intercept)"   
# [9] "p_group2.L"       "q_(Intercept)"    "q_group2.L"       "diff_(Intercept)" "diff_group2.L"

tab_qval <- res[ ,c("taxon", "q_(Intercept)", "q_group2.L")]
dim(tab_qval) # 10804     3

summary( tab_qval )
# taxon           q_(Intercept)   q_group2.L    
# Length:10804       Min.   :1     Min.   :0.0000  
# Class :character   1st Qu.:1     1st Qu.:1.0000  
# Mode  :character   Median :1     Median :1.0000  
#                    Mean   :1     Mean   :0.9993  
#                    3rd Qu.:1     3rd Qu.:1.0000  
#                    Max.   :1     Max.   :1.0000

names(tab_qval) # "taxon"         "q_(Intercept)" "q_group2.L" 
# Levels: A_low_PC1_behav < B_high_PC1_behav
names(tab_qval) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_qval) <- tab_qval$taxon

head(tab_qval)
# taxon Intercept High PC1 cf. Low PC1
# fxn_2   fxn_2         1                    1
# fxn_4   fxn_4         1                    1
# fxn_8   fxn_8         1                    1
# fxn_9   fxn_9         1                    1
# fxn_10 fxn_10         1                    1
# fxn_11 fxn_11         1                    1

# sig diff abun Fxns in High PC1 cf. Low PC1
sel.sig.HighPC1_cf_LowPC1 <- which(tab_qval$`High PC1 cf. Low PC1` <= 0.05) # qty 6
fxn.sig.HighPC1_cf_LowPC1 <- tab_qval$taxon[sel.sig.HighPC1_cf_LowPC1]
length(unique(fxn.sig.HighPC1_cf_LowPC1)) # 6


sel.sig <- unique(c(sel.sig.HighPC1_cf_LowPC1)) # qty 6
tab_qval[sel.sig, ]
# taxon Intercept High PC1 cf. Low PC1
# fxn_303     fxn_303         1         2.859978e-04
# fxn_3075   fxn_3075         1         1.406783e-02
# fxn_9200   fxn_9200         1         1.530325e-06
# fxn_10191 fxn_10191         1         2.916860e-08
# fxn_11908 fxn_11908         1         2.916860e-08
# fxn_14887 fxn_14887         1         3.763993e-03



## log-fold change (LFC)
tab_lfc = res[ ,c("taxon", "lfc_(Intercept)",  "lfc_group2.L" )]
class(tab_lfc) # "data.frame"
names(tab_lfc) # "taxon"           "lfc_(Intercept)" "lfc_group2.L"
names(tab_lfc) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_lfc) <- tab_lfc$taxon
tab_lfc[sel.sig, ]


## get significant diff abun taxa across any pairs
tab_lfc.sig_all <- tab_lfc[ sel.sig, ]

dim(tab_lfc.sig_all) # 6  3

head(tab_lfc.sig_all)
# taxon  Intercept High PC1 cf. Low PC1
# fxn_303     fxn_303 0.03341965            0.7755507
# fxn_3075   fxn_3075 0.03954231            0.9227495
# fxn_9200   fxn_9200 0.03610942            0.8402172
# fxn_10191 fxn_10191 0.03842993            0.8960060
# fxn_11908 fxn_11908 0.03842993            0.8960060
# fxn_14887 fxn_14887 0.03134828            0.7257516

hist(tab_lfc.sig_all$Intercept); summary(tab_lfc.sig_all$Intercept)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.03135 0.03409 0.03727 0.03621 0.03843 0.03954 
hist(tab_lfc.sig_all$`High PC1 cf. Low PC1`); summary(tab_lfc.sig_all$`High PC1 cf. Low PC1`)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7258  0.7917  0.8681  0.8427  0.8960  0.9227 



names(tab_lfc.sig_all)[2] # "Intercept"

tab_lfc.sig_all.melt <- melt(tab_lfc.sig_all[ ,-2])
names(tab_lfc.sig_all.melt) # "taxon"    "variable" "value" 
names(tab_lfc.sig_all.melt) <- c("taxon", "Comparison groups", "Log-fold-change" )

p <- ggplot(tab_lfc.sig_all.melt, aes(`Log-fold-change`))+ #, color = `Comparison groups`)) +
  geom_density(aes())+ # ..density.. | ..scaled.. | ..count..
  labs(x = "Log-fold-change", y = "Density") +
  # scale_colour_manual( values = c("High PC1 cf. Low PC1" = "#66c2a5"),
  #                      labels = c("High PC1 cf. Low PC1")) +
  xlim(-1,1)+
  theme_classic() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
p

grid.text(label = "ANCOMBC: Function-level | Males", x = unit(0.17, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=9), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","LFC-density-plots",header,this_study,"Males-v9d.tiff"), width = 9, height = 6, units = "cm", res=450, compression="lzw",type="cairo")


# assign "Positive LFC", "Negative LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- NA
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` < 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Negative LFC"
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` > 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Positive LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- factor(tab_lfc.sig_all.melt$pos_or_neg_LFC, levels = c("Positive LFC", "Negative LFC"), ordered = TRUE)


head(tab_lfc.sig_all.melt)
# taxon    Comparison groups Log-fold-change pos_or_neg_LFC
# 1   fxn_303 High PC1 cf. Low PC1       0.7755507   Positive LFC
# 2  fxn_3075 High PC1 cf. Low PC1       0.9227495   Positive LFC
# 3  fxn_9200 High PC1 cf. Low PC1       0.8402172   Positive LFC
# 4 fxn_10191 High PC1 cf. Low PC1       0.8960060   Positive LFC
# 5 fxn_11908 High PC1 cf. Low PC1       0.8960060   Positive LFC
# 6 fxn_14887 High PC1 cf. Low PC1       0.7257516   Positive LFC


table(tab_lfc.sig_all.melt$pos_or_neg_LFC, useNA = "ifany")
# Positive LFC Negative LFC 
#           6           0 



## join vK coords to data table

# from earlier ...
head(df.fxn_vk_coords.noNAs)
#       tot_mean_OC_x tot_mean_HC_y
# fxn_2     0.6000000     1.8000000
# fxn_3     1.0000000     0.4285714
# fxn_4     0.6666667     2.3333333
# fxn_5     0.6666667     2.3333333
# fxn_6     0.6666667     2.3333333
# fxn_7     0.6666667     2.3333333
row.names(df.fxn_vk_coords.noNAs)

res.sig <- tab_lfc.sig_all.melt

res.sig$OC_x <- NA
res.sig$HC_y <- NA

length(res.sig$taxon) # 6
length( which(res.sig$taxon %in% row.names(df.fxn_vk_coords.noNAs)) ) # 6
100*(3/6)
# 50 %

dim(res.sig) # 6  6

for (i in 1:dim(res.sig)[1]) {
  #i<-1
  this_fxn <- res.sig$taxon[i]
  sel.fxn_vk <- which( row.names(df.fxn_vk_coords.noNAs) == this_fxn ) #
  
  if (length(sel.fxn_vk) ==1) {
    res.sig$OC_x[i] <- df.fxn_vk_coords.noNAs$tot_mean_OC_x[sel.fxn_vk]
    res.sig$HC_y[i] <- df.fxn_vk_coords.noNAs$tot_mean_HC_y[sel.fxn_vk]
  } else {
    res.sig$OC_x[i] <- NA
    res.sig$HC_y[i] <- NA
  }
  print(paste0("completed ",i))
}

unique( res.sig$`Comparison groups` )
# [1] High PC1 cf. Low PC1
# Levels: High PC1 cf. Low PC1

# res.sig$`Comparison groups` <- factor(res.sig$`Comparison groups`, 
#                                       levels = c("IGT cf. Normal", "T2D Met neg cf. Normal", "T2D Met pos cf. Normal"),
#                                       labels = c("IGT cf. Normal", "T2D Met- cf. Normal", "T2D Met+ cf. Normal"),
#                                       ordered=TRUE)

ok <- complete.cases(res.sig)
sel.ok <- which(ok==TRUE) # qty 15

res.sig.ok <- res.sig[sel.ok, ]

dim(res.sig)   #  6  6
dim(res.sig.ok)#  3  6

names(res.sig.ok)
# "taxon"             "Comparison groups" "Log-fold-change"   "pos_or_neg_LFC"    "OC_x"              "HC_y"             


summary(res.sig.ok$`Log-fold-change`[which(res.sig.ok$`Comparison groups`=="High PC1 cf. Low PC1")]);hist(res.sig.ok$`Log-fold-change`[which(res.sig.ok$`Comparison groups`=="High PC1 cf. Low PC1")])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.7756  0.8079  0.8402  0.8462  0.8815  0.9227



p <- ggplot(res.sig.ok, aes(x = OC_x, y = HC_y, color = `Log-fold-change`))+
  geom_point(alpha = 0.2)

p


# overlay vK ref data ...

# sel <- which(df.vk.ref$sample %in% c("Acetate", "Propionate",  "Butyrate", "Riboflavin (Vit B2)" ,
#                                      "Cobalamin (Vit B12)", "Pyridoxal 5'-phosphate (Vit B6)", "Folate (Vit B9)", "Menaquinone (Vit K2)" ) )
# df.hac <- df.vk.ref[sel, ]
df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2

vkgrouprect

#library(colorspace)

p <- ggplot()+
  
  geom_point(data = filter(res.sig.ok, `Comparison groups`=="High PC1 cf. Low PC1"), aes(x = OC_x, y = HC_y, color = `Log-fold-change`), size = 3.5)+ # , alpha = 0.6
  
  ## ADJUST SCALE !!!!!!!!!!
  scale_color_continuous_sequential(palette = "BluYl", rev=FALSE, name="Log-fold\n-change", limits = c(-1, 1) )+ # palette = "YlGnBu" "BluYl"
  
  guides(color = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5))+
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  #geom_point(data = df.hac ,aes(x=OC_x, y=HC_y), color = "#253494")+ # , pch=8  "#80b1d3"
  #geom_text(data = df.hac ,aes(x=OC_x, y=HC_y, label=sample), color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  geom_segment(aes(x = 1.1, y = 1.5, xend = 1.8,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  #geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  #geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0.2), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.75, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "High cf. Low PC1 problem behaviors: Functions | Males", x = unit(0.07, "npc") , y = unit(0.96,"npc"), gp=gpar(fontsize=10), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","vK-fxn-logfoldchange-scatterpoints-High-vs-Low-PC1-Problem-Behav-Males--v9d.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")




## plot mean rel abun of each group alongside log-fold-change - for tab_lfc.sig_global


head(tab_lfc.sig_all)
#               taxon  Intercept High PC1 cf. Low PC1
# fxn_303     fxn_303 0.03341965            0.7755507
# fxn_3075   fxn_3075 0.03954231            0.9227495
# fxn_9200   fxn_9200 0.03610942            0.8402172
# fxn_10191 fxn_10191 0.03842993            0.8960060
# fxn_11908 fxn_11908 0.03842993            0.8960060
# fxn_14887 fxn_14887 0.03134828            0.7257516

temp <- tab_lfc.sig_all

# add vK coordinates
temp$OC_x <- NA
temp$HC_y <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_fxn <- temp$taxon[i]
  sel.fxn_vk <- which( row.names(df.fxn_vk_coords.noNAs) == this_fxn )
  
  if (length(sel.fxn_vk) ==1) {
    temp$OC_x[i] <- df.fxn_vk_coords.noNAs$tot_mean_OC_x[sel.fxn_vk]
    temp$HC_y[i] <- df.fxn_vk_coords.noNAs$tot_mean_HC_y[sel.fxn_vk]
  } else {
    temp$OC_x[i] <- NA
    temp$HC_y[i] <- NA
  }
  print(paste0("completed ",i))
}

# add matching compound?
temp$coords <- paste0(temp$OC_x,"__",temp$HC_y)

out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")
head( out[[1]] )
out.temp <- out[[1]]
row.names(out.temp) <- out.temp$coords

temp$compound <- NA

for (i in 1:dim(temp)[1]) {
  #i<-2
  this_coord <- temp$coords[i]
  if (is.na(temp$OC_x[i])){
    temp$compound[i] <- NA
  } else {
    sel <- which(out.temp$coords == this_coord)
    temp$compound[i] <- out.temp$compound[sel]
  }
  print(paste0("completed ",i))
}

head(temp)

temp <- cbind( temp, as.data.frame( phy.ok.m@tax_table[ temp$taxon , ]))

write.csv(x = temp, file = "ANCOMBC-Function-level-Sig-diff-abun-fxns--Child-Behav-Males-LFC-and-Function-details.csv")



## plot heatmap of LFC for sig diff abun Functions - ANCOMBC Function-level

m2 <- tab_lfc.sig_all[ ,-1] # exclude 'taxon' column, it is in row.name anyway
names(m2) #  "Intercept"            "High PC1 cf. Low PC1"
names(m2) <- c("Intercept", "High PC1\ncf. Low PC1" )

m2 <- as.matrix(m2)
colnames(m2) # "Intercept"            "High PC1\ncf. Low PC1"
m2
class(m2) # "matrix" "array" 

row.names(m2)
identical(row.names(m2), temp$taxon)
sel.ok <- which(!is.na(temp$OC_x))
row.names(m2)[sel.ok] <- paste0(row.names(m2)[sel.ok], "#")


pheatmap(m2, cluster_cols = FALSE, #cluster_row = FALSE, 
         #color = color,
         color = hcl.colors(50, "BluYl"), # palette = "BrBG", direction = 1
         border_color = FALSE, 
         fontsize_col = 12,
         main = "Differentially abundant functions\n(log-fold-changes): Males"
)

#dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-LogFoldChange-ANCOMBC-Function-level-Child-Behav-Males--v9d.tiff"), width = 9, height = 14, units = "cm", res=450, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-LogFoldChange-ANCOMBC-Function-level-Child-Behav-Males--v9d.tiff"), width = 9, height = 8, units = "cm", res=450, compression="lzw",type="cairo")


ph <- pheatmap(m2, cluster_cols = FALSE)
ph_taxa_rownames <- ph$tree_row$labels[ ph$tree_row$order ]

ph_taxa_rownames <- gsub(pattern = "#", replacement = "", x = ph_taxa_rownames)


## compare with relative abundance of functions in these groups

phy.relabun <- phy.ok.m

m <- as.matrix( as.data.frame(phy.relabun@otu_table[ ph_taxa_rownames , ]))
identical(colnames(m), phy.relabun@sam_data$Run) # TRUE

names( phy.relabun@sam_data )


levels( phy.relabun@sam_data$group2 ) 
# "A_low_PC1_behav"  "B_high_PC1_behav"
table( phy.relabun@sam_data$group ) 
# high_PC1_behav  low_PC1_behav 
#             8             9 

phy.relabun@sam_data$label <- phy.relabun@sam_data$group
phy.relabun@sam_data$label <- factor(phy.relabun@sam_data$label,
                                     levels = c("low_PC1_behav","high_PC1_behav"),
                                     labels = c("Low PC1","High PC1"),
                                     ordered=TRUE)
sample_order <- order(phy.relabun@sam_data$group2)
phy.relabun@sam_data$group_label <- paste0(phy.relabun@sam_data$label," (",phy.relabun@sam_data$Run,")")
phy.relabun@sam_data$group_label[sample_order]
phy.relabun@sam_data$group_label <- factor(phy.relabun@sam_data$group_label,
                                           levels = phy.relabun@sam_data$group_label[sample_order],
                                           ordered=TRUE)
colnames(m) <- phy.relabun@sam_data$group_label
# reorder columns
m <- m[ , sample_order]

# No dendrogram nor reordering for neither column or row
heatmap(m, Colv = NA, Rowv = NA, scale="column")
heatmap(m, Colv = NA, Rowv = NA)

# https://www.reneshbedre.com/blog/heatmap-with-pheatmap-package-r.html
#library(pheatmap)

row.names(m)
identical(row.names(m), ph_taxa_rownames) # TRUE
identical( row.names(m), gsub(pattern="#",replacement="", x= ph$tree_row$labels[ ph$tree_row$order ]) ) # TRUE
row.names(m) <- ph$tree_row$labels[ ph$tree_row$order ]
# use same '#' notation as above for functions that map to vK space

pheatmap(m, cluster_cols = FALSE,  cluster_row = FALSE, # gaps_col = c(6,12,18), 
         border_color = FALSE, main = "Function relative abundance (%)")


group_labels <- data.frame( Group = phy.relabun@sam_data$label[sample_order] )
row.names(group_labels) <- colnames(m)


cols = c("Low PC1" = "#7fbc41",  "High PC1" = "#de77ae") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9

mycolors <- c("#7fbc41","#de77ae")
names(mycolors) <- c("Low PC1","High PC1")


mycolors <- list(Group = mycolors)

color <- colorRampPalette((c("#fff5f0","#fc9272", "#cb181d")))(50)
pheatmap(m, gaps_col = c(9), cluster_cols = FALSE, cluster_row = FALSE, # 49          43          33          20
         annotation_col = group_labels,
         annotation_colors = mycolors,
         color = color,
         fontsize_col = 8,
         border_color = FALSE, 
         main = "Rel. abun. of functions (%): Males")

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-RelAbun-for-ANCOMBC-Function-level-Child-Behav-Males--v9.tiff"), width = 14, height = 8, units = "cm", res=450, compression="lzw",type="cairo")



#-------------------------

#### Flannery behaviour - differences in sum rel abun at unique vK-coordinates - needed for next step
#-------------------------

header <- "-sumrelabun_at-vk-coords-"

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

this_study <- "-Flannery-behav-"

# use clean data
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")

vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


df.temp <- vk.samp.noNAs.long

names(df.temp)
# [1] "sample"        "fxn_id"        "abun"          "abun_type"     "OC_x"          "HC_y"          "sex"          
# [8] "depress_probs" "behav_PC1" 




# plot as spatial data in vK space

spdf <- df.temp[which(df.temp$abun > 0), c("sample", "depress_probs", "behav_PC1", "sex", "OC_x", "HC_y", "abun") ]


dim(spdf) #  221111      7

# any samples with unknown sex?
table(spdf$sex, useNA = "ifany")
# female   male 
# 126388  94723

head(spdf)
# sample depress_probs behav_PC1    sex      OC_x     HC_y         abun
# 4  SRR8204551            50 -1.874486 female 0.6666667 2.333333 0.0015655056
# 8  SRR8204551            50 -1.874486 female 1.0000000 1.500000 0.0001533557
# 10 SRR8204551            50 -1.874486 female 0.5714286 2.000000 0.0017635900
# 11 SRR8204551            50 -1.874486 female 0.8000000 1.600000 0.0036294171
# 15 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0007791264
# 16 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0009456932


temp <- spdf

unique_samps <- unique(spdf$sample)
length(unique_samps) # 40

add <- NA
hist(spdf$abun)
for (i in 1:length(unique_samps)) {
  #i<-1
  this_samp <- unique_samps[i]
  sel <- which(spdf$sample == this_samp)
  add <- c(add, sum(spdf$abun[sel]))
  print(paste0("completed ",i))
}
add <- add[-1]
hist(add); summary(add)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.69   54.17   54.96   54.99   55.96   57.00

names(spdf)

temp <- spdf



spdf$coords <- paste0(spdf$OC_x,"__",spdf$HC_y)

unique_coords <- unique(spdf$coords) # qty 2466

temp <- spdf

head(spdf)
# sample depress_probs behav_PC1    sex      OC_x     HC_y         abun                              coords
# 4  SRR8204551            50 -1.874486 female 0.6666667 2.333333 0.0015655056 0.666666666666667__2.33333333333333
# 8  SRR8204551            50 -1.874486 female 1.0000000 1.500000 0.0001533557                              1__1.5
# 10 SRR8204551            50 -1.874486 female 0.5714286 2.000000 0.0017635900                0.571428571428571__2
# 11 SRR8204551            50 -1.874486 female 0.8000000 1.600000 0.0036294171                            0.8__1.6
# 15 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0007791264 0.748921568627451__1.93460784313725
# 16 SRR8204551            50 -1.874486 female 0.7489216 1.934608 0.0009456932 0.748921568627451__1.93460784313725


## within a sample (i)
## at every unique coordinate
## identify duplicated coordinates within a sample and obtain total rel abun
## assign a vK compound zone
## output data table with columns: "sample" "behav_PC1" "OC_x" "HC_y" "tot_relabun" "coords" "compound"

get_tot_relabun_and_compound_at_coords <- function(spdf_in, coords) {
  ###spdf_in = spdf
  ###coords = unique_coords
  ###i = 1
  
  all_samples <- unique(spdf_in$sample)
  this_samp <- all_samples[i]
  
  # cut down to sample of interest
  sel <- which(spdf$sample == this_samp)
  spdf_in <- spdf_in[sel, ]
  
  #df.out <- data.frame(coords = coords, sample=this_samp, group=unique(spdf_in$group), OC_x=NA, HC_y=NA, tot_relabun=NA, compound=NA)
  df.out <- data.frame(coords = coords, sample=this_samp, behav_PC1=unique(spdf_in$behav_PC1), OC_x=NA, HC_y=NA, tot_relabun=NA, compound=NA)
  
  # loop through each unique coord to sum rel abun and assign compound
  
  for (c in 1:length(coords)) {
    #c<-1
    this_coord <- coords[c] # "0.666666666666667__2.33333333333333"
    # split out x & y
    end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
    start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
    df.out$OC_x[c]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
    df.out$HC_y[c]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
    
    sel <- which(spdf_in$coords == this_coord)
    if (length(sel)>0) {
      df.out$tot_relabun[c] <- sum(spdf_in$abun[sel], na.rm = TRUE)
    } else {
      df.out$tot_relabun[c] <- 0
    }
    
    # assign compound based on coordinates
    
    OC_x <- df.out$OC_x[c]
    HC_y <- df.out$HC_y[c]
    
    # lipids
    if (OC_x >=0 & OC_x <= 0.2 & HC_y >= 1.5 & HC_y <= 2.3) {
      compd <- "lipids"
    } else if (OC_x > 0.2 & OC_x <= 0.52 & HC_y >= 1.5 & HC_y <= 2.2) {
      # proteins
      compd <- "proteins"
    } else if (OC_x > 0.52 & OC_x <= 0.7 & HC_y >= 1.5 & HC_y <= 2.2) {
      # amino_sugars
      compd <- "amino_sugars"
    } else if (OC_x > 0.7 & OC_x <= 1.1 & HC_y >= 1.5 & HC_y <= 2.4) {
      # carbs - Carbohydrates
      compd <- "carbs"
    } else if (OC_x >=0 & OC_x <= 0.25 & HC_y >= 0.5 & HC_y <= 1.25) {
      # cond_aromtx - Condensed aromatics
      compd <- "cond_aromtx"
    } else if (OC_x > 0.25 & OC_x <= 0.67 & HC_y >= 0.75 & HC_y < 1.5) {
      # lignin
      compd <- "lignin"
    } else if (OC_x > 0.67 & OC_x <= 0.97 & HC_y >= 0.53 & HC_y < 1.5) {
      # tannins
      compd <- "tannins"
    } else if ( OC_x <= 1.1 & HC_y > 2.2 ) {
      # other methylated
      compd <- "methylated"
    } else if ( OC_x > 1.1 & HC_y >= 1.5 ) {
      # other hydrated
      compd <- "hydrated"
    } else if ( OC_x > 0.97 & HC_y < 1.5 ) {
      # other demethylated
      compd <- "demethylated"
    } else if ( OC_x <= 0.97  & HC_y < 0.75 & !( OC_x ==0 & HC_y == 0 )) {
      # condensed - but exclude compounds that simultaneously contain no carbon or hydrogen
      compd <- "condensed"
    } else if ( OC_x >= 0 & OC_x <= 0.25 & HC_y > 1.25 & HC_y < 1.5) {
      # k2_b12 - gap in vK space spanning approx VitK2 to VitB12
      compd <- "k2_b12"
    } else {
      compd <- NA
    }
    
    df.out$compound[c] <- compd
    
  } # END loop for all unique coordinates
  
  return(df.out)
}


# get_tot_relabun_and_compound_at_coords <- function(spdf_in, coords)

# one output table per sample

time.start <- Sys.time()
cl<-makeCluster( detectCores()-1 )
registerDoParallel(cl)

out <- foreach(i=1:length(unique_samps), .packages = "stringr" ) %dopar%
  get_tot_relabun_and_compound_at_coords( spdf_in = spdf, coords = unique_coords)

stopCluster(cl)
time.finish <- Sys.time()


length(out) # 40
length(unique_coords) # 2466

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = out, file = "out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")
# out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")


temp <- out[[1]]
head(temp)
#                                coords     sample behav_PC1      OC_x     HC_y tot_relabun     compound
# 1 0.666666666666667__2.33333333333333 SRR8204551 -1.874486 0.6666667 2.333333  0.03458170   methylated
# 2                              1__1.5 SRR8204551 -1.874486 1.0000000 1.500000  0.13877409        carbs
# 3                0.571428571428571__2 SRR8204551 -1.874486 0.5714286 2.000000  0.05894608 amino_sugars
# 4                            0.8__1.6 SRR8204551 -1.874486 0.8000000 1.600000  0.05770773        carbs
# 5 0.748921568627451__1.93460784313725 SRR8204551 -1.874486 0.7489216 1.934608  0.11497840        carbs
# 6                            0.375__2 SRR8204551 -1.874486 0.3750000 2.000000  0.01939949     proteins

temp <- as.data.frame( out[[1]][ ,"tot_relabun"] )
names(temp) <- unique_samps[1]  

for (i in 2:length(out)) {
  #i<-2
  new_temp <- as.data.frame( out[[i]][ ,"tot_relabun"] )
  names(new_temp) <- unique_samps[i]  
  
  temp <- cbind(temp, new_temp)
  print(paste0("comleted ",i))
}

dat <- temp
row.names(dat) <- unique_coords

rm(temp, new_temp)

str(dat)
dim(dat) # 2466   40

ok <- complete.cases(dat)
sel.ok <- which(ok==TRUE) # qty 2466


saveRDS(dat, "dat--rows-are-uniquevKcoords-cols-are-samples.Child_behav.RDS")


#-------------------------

#### Flannery Behaviour - ANCOMBC - unique vK coord-level
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
df.out.distil <- readRDS(file = "df.out.distil_Flannery-behav.RDS")

this_study <- "-Flannery-behav-"
header <- "-ANCOMBC-vKcoords-"

# also use: df.fxn_vk_coords.noNAs
df.fxn_vk_coords.noNAs <- readRDS("df.fxn_vk_coords.noNAs.Flannery-child-behav.RDS")

# use clean data
vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")
vk.samp.noNAs.long <- vk.samp.noNAs.long.clean


phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20787 taxa and 40 samples ]
# sample_data() Sample Data:       [ 40 samples by 77 sample variables ]
# tax_table()   Taxonomy Table:    [ 20787 taxa by 4 taxonomic ranks ]


head(phy@sam_data) # host_sex

dim(phy@sam_data) # 40 77

phy@sam_data$sex <- phy@sam_data$host_sex
phy@sam_data$behav_PC1 <- NA

for (i in 1:dim(phy@sam_data)[1]) {
  #i<-1
  this_samp <- phy@sam_data$Run[i]
  sel.vk <- which(vk.samp.noNAs.long.clean$sample == this_samp)
  phy@sam_data$behav_PC1[i] <- unique(vk.samp.noNAs.long.clean$behav_PC1[sel.vk])
  print(paste0("completed ",i))
}

phy@sam_data$group <- NA

sel <- which(phy@sam_data$sex == "female") # qty 23
hist(phy@sam_data$behav_PC1[sel])
median_val <- median(phy@sam_data$behav_PC1[sel], na.rm = TRUE)
subsel <- which(phy@sam_data$behav_PC1[sel] <= median_val) # 10
phy@sam_data$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(phy@sam_data$behav_PC1[sel] > median_val) # 10
phy@sam_data$group[sel[subsel]] <- "high_PC1_behav"

sel <- which(phy@sam_data$sex == "male")
hist(phy@sam_data$behav_PC1[sel])
median_val <- median(phy@sam_data$behav_PC1[sel], na.rm = TRUE)
subsel <- which(phy@sam_data$behav_PC1[sel] <= median_val) # 9
phy@sam_data$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(phy@sam_data$behav_PC1[sel] > median_val) # 8
phy@sam_data$group[sel[subsel]] <- "high_PC1_behav"


# add group2
phy@sam_data$group2 <- phy@sam_data$group
table(phy@sam_data$group2, useNA="ifany")
# high_PC1_behav  low_PC1_behav           <NA> 
#             18             19              3

phy@sam_data$group2 <- factor(phy@sam_data$group2,
                              levels = c("low_PC1_behav", "high_PC1_behav"),
                              labels = c("A_low_PC1_behav", "B_high_PC1_behav"),
                              ordered = TRUE)
unique(phy@sam_data$group2)
# [1] A_low_PC1_behav  B_high_PC1_behav
# Levels: A_low_PC1_behav < B_high_PC1_behav




length(unique(phy@sam_data$Run)) # 40

sel.ok <- which(!is.na(phy@sam_data$behav_PC1)) # qty 37

phy.ok <- prune_samples(samples = row.names(phy@sam_data)[sel.ok], phy)

# prune taxa with zero reads
phy.ok <- prune_taxa(taxa = taxa_sums(phy.ok) > 0, x = phy.ok)
phy.ok
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20599 taxa and 37 samples ]
# sample_data() Sample Data:       [ 37 samples by 80 sample variables ]
# tax_table()   Taxonomy Table:    [ 20599 taxa by 4 taxonomic ranks ]


# # add group2
# phy.ok@sam_data$group2 <- phy.ok@sam_data$group
# table(phy.ok@sam_data$group2, useNA="ifany")
# # high_PC1_behav  low_PC1_behav 
# # 18             19 
# 
# phy.ok@sam_data$group2 <- factor(phy.ok@sam_data$group2,
#                                  levels = c("low_PC1_behav", "high_PC1_behav"),
#                                  labels = c("A_low_PC1_behav", "B_high_PC1_behav"),
#                                  ordered = TRUE)
# unique(phy.ok@sam_data$group2)
# # [1] A_low_PC1_behav  B_high_PC1_behav
# # Levels: A_low_PC1_behav < B_high_PC1_behav




## but need to build phyloseq object using rel abundances at unique vK coords (aggregated functions)

#saveRDS(dat, "dat--rows-are-uniquevKcoords-cols-are-samples.Child_behav.RDS")
dat <- readRDS("dat--rows-are-uniquevKcoords-cols-are-samples.Child_behav.RDS")

dat[1:5, 1:5]
#                                     SRR8204551 SRR8204552 SRR8204553 SRR8204554 SRR8204555
# 0.666666666666667__2.33333333333333 0.03458170 0.02433202 0.03727985 0.03962676 0.03163521
# 1__1.5                              0.13877409 0.14316568 0.14494060 0.14935700 0.14547642
# 0.571428571428571__2                0.05894608 0.05423322 0.05120914 0.05541412 0.05258609
# 0.8__1.6                            0.05770773 0.05952820 0.04351004 0.05039567 0.04792577
# 0.748921568627451__1.93460784313725 0.11497840 0.12520853 0.06311235 0.07893679 0.07654372

identical(names(dat),sample_names(phy)) # TRUE

## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( dat )
dim(otu.m)
# 2466   40


OTU <- otu_table(otu.m, taxa_are_rows = TRUE)

## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.fxn <- data.frame(vk_coord = row.names(dat))
row.names(tax.fxn) <- tax.fxn$vk_coord

# assign compound classes
out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")
head( out[[1]] )
#                                coords     sample behav_PC1      OC_x     HC_y tot_relabun     compound
# 1 0.666666666666667__2.33333333333333 SRR8204551 -1.874486 0.6666667 2.333333  0.03458170   methylated
# 2                              1__1.5 SRR8204551 -1.874486 1.0000000 1.500000  0.13877409        carbs
# 3                0.571428571428571__2 SRR8204551 -1.874486 0.5714286 2.000000  0.05894608 amino_sugars
# 4                            0.8__1.6 SRR8204551 -1.874486 0.8000000 1.600000  0.05770773        carbs
# 5 0.748921568627451__1.93460784313725 SRR8204551 -1.874486 0.7489216 1.934608  0.11497840        carbs
# 6                            0.375__2 SRR8204551 -1.874486 0.3750000 2.000000  0.01939949     proteins
identical(out[[1]]$coords, tax.fxn$vk_coord) # TRUE
tax.fxn <- cbind(tax.fxn, out[[1]][ ,c("OC_x", "HC_y", "compound") ])
names(tax.fxn) # "vk_coord" "OC_x"     "HC_y"     "compound"

tax.m <- as.matrix( tax.fxn )
dim(tax.m) # 2466    4

TAX <- tax_table( tax.m )





## Create a phyloseq object, merging OTU & TAX tables
phy.vk = phyloseq(OTU, TAX)
phy.vk
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2466 taxa and 40 samples ]
# tax_table()   Taxonomy Table:    [ 2466 taxa by 4 taxonomic ranks ]

identical(sample_names(phy), sample_names(phy.vk)) # TRUE

phy@sam_data[1:5, ]


SAMP <- sample_data( as( phy@sam_data ,"data.frame") )



### Combine SAMPDATA into phyloseq object
phy.vk <- merge_phyloseq(phy.vk, SAMP)
phy.vk
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2466 taxa and 40 samples ]
# sample_data() Sample Data:       [ 40 samples by 81 sample variables ]
# tax_table()   Taxonomy Table:    [ 2466 taxa by 4 taxonomic ranks ]


## select only those with non-NA behav_PC1 data
phy.vk.ok <- prune_samples(samples = sample_names(phy.ok), phy.vk)

# prune taxa with zero reads
phy.vk.ok <- prune_taxa(taxa = taxa_sums(phy.vk.ok) > 0, x = phy.vk.ok)
phy.vk.ok
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2459 taxa and 37 samples ]
# sample_data() Sample Data:       [ 37 samples by 81 sample variables ]
# tax_table()   Taxonomy Table:    [ 2459 taxa by 4 taxonomic ranks ]


str(phy.vk.ok@sam_data)

# use this for 
unique(phy.vk.ok@sam_data$group2)
# [1] A_low_PC1_behav  B_high_PC1_behav
# Levels: A_low_PC1_behav < B_high_PC1_behav



## Females

sel.f <- which(phy.vk.ok@sam_data$sex == "female") # qty 20
phy.vk.ok.f <- prune_samples(samples = row.names(phy.vk.ok@sam_data)[sel.f], phy.vk.ok)

# prune taxa with zero reads
phy.vk.ok.f <- prune_taxa(taxa = taxa_sums(phy.vk.ok.f) > 0, x = phy.vk.ok.f)
phy.vk.ok.f
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2317 taxa and 20 samples ]
# sample_data() Sample Data:       [ 20 samples by 81 sample variables ]
# tax_table()   Taxonomy Table:    [ 2317 taxa by 4 taxonomic ranks ]


## ANCOMBC - differential abundance testing??
set.seed(123)
out.ancombc.vk = ancombc2(data = phy.vk.ok.f,
                          tax_level = NULL,
                          fix_formula = "group2",
                          p_adj_method = "holm", prv_cut = 0.1, lib_cut = 0,
                          group = "group2", struc_zero = TRUE, neg_lb = TRUE,
                          alpha = 0.05, global = TRUE,  # change alpha level to reduce number !!!
                          n_cl = 6, verbose = TRUE)

# `tax_level` is not speficified 
# No agglomeration will be performed
# Otherwise, please speficy `tax_level` by one of the following: 
#   
#   Obtaining initial estimates ...
# Estimating sample-specific biases ...
# ANCOM-BC2 primary results ...
# The sensitivity analysis for the pseudo-count addition ...
# Warning message:
#   The group variable has < 3 categories 
# The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated 

str(out.ancombc.vk)
# List of 11
# $ feature_table  : num [1:1845, 1:20] 0.0346 0.1388 0.0589 0.0577 0.115 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:1845] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# .. ..$ : chr [1:20] "SRR8204551" "SRR8204552" "SRR8204555" "SRR8204559" ...
# $ zero_ind       :'data.frame':	2317 obs. of  3 variables:
#   ..$ taxon                                      : chr [1:2317] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# ..$ structural_zero (group2 = A_low_PC1_behav) : logi [1:2317] FALSE FALSE FALSE FALSE FALSE FALSE ...
# ..$ structural_zero (group2 = B_high_PC1_behav): logi [1:2317] FALSE FALSE FALSE FALSE FALSE FALSE ...
# $ samp_frac      : num [1:20] 0.144 -0.122 -0.269 -0.463 -1.002 ...
# $ delta_em       : num [1:2] -1.11e-17 4.12e-02
# $ delta_wls      : num [1:2] -1.90e-17 4.12e-02
# $ pseudo_sens_tab:'data.frame':	1845 obs. of  3 variables:
#   ..$ taxon      : chr [1:1845] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# ..$ (Intercept): num [1:1845] 1.24e-16 1.08e-16 1.39e-16 1.30e-16 1.14e-16 ...
# ..$ group2.L   : num [1:1845] 0.0423 0.0207 0.0172 0.0023 0.0501 ...
# $ res            :'data.frame':	1845 obs. of  13 variables:
#   ..$ taxon           : chr [1:1845] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# ..$ lfc_(Intercept) : num [1:1845] 8.60e-17 6.06e-18 3.69e-17 -9.46e-17 -1.06e-16 ...
# ..$ lfc_group2.L    : num [1:1845] 0.14307 0.00301 0.00679 -0.03286 -0.14481 ...
# ..$ se_(Intercept)  : num [1:1845] 0.0993 0.0514 0.0634 0.0632 0.0728 ...
# ..$ se_group2.L     : num [1:1845] 0.1405 0.0727 0.0896 0.0894 0.103 ...
# ..$ W_(Intercept)   : num [1:1845] 8.66e-16 1.18e-16 5.82e-16 -1.50e-15 -1.46e-15 ...
# ..$ W_group2.L      : num [1:1845] 1.0184 0.0414 0.0757 -0.3676 -1.4063 ...
# ..$ p_(Intercept)   : num [1:1845] 1 1 1 1 1 ...
# ..$ p_group2.L      : num [1:1845] 0.309 0.967 0.94 0.713 0.16 ...
# ..$ q_(Intercept)   : num [1:1845] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ q_group2.L      : num [1:1845] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ diff_(Intercept): logi [1:1845] FALSE FALSE FALSE FALSE FALSE FALSE ...
# ..$ diff_group2.L   : logi [1:1845] FALSE FALSE FALSE FALSE FALSE FALSE ...
# $ res_global     : NULL
# $ res_pair       : NULL
# $ res_dunn       : NULL
# $ res_trend      : NULL

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = out.ancombc.vk, file = "out.ancombc.vk.Child_behav.Females.RDS")

#out.ancombc.vk <- readRDS("out.ancombc.vk.Child_behav.Females.RDS")




out <- out.ancombc.vk


out$feature_table[1:5, 1:5]
#                                     SRR8204551 SRR8204552 SRR8204555 SRR8204559 SRR8204563
# 0.666666666666667__2.33333333333333 0.03458170 0.02433202 0.03163521 0.02541547 0.06570868
# 1__1.5                              0.13877409 0.14316568 0.14547642 0.13986424 0.14175485
# 0.571428571428571__2                0.05894608 0.05423322 0.05258609 0.04144859 0.04923890
# 0.8__1.6                            0.05770773 0.05952820 0.04792577 0.04566470 0.04597774
# 0.748921568627451__1.93460784313725 0.11497840 0.12520853 0.07654372 0.07274286 0.08436865

fxns_out <- row.names(out$feature_table)

dim(out$feature_table) #  1845   20
phy.vk.ok.f@otu_table[1:5, 1:5]
# OTU Table:          [5 taxa and 5 samples]
# taxa are rows
#                                     SRR8204551 SRR8204552 SRR8204555 SRR8204559 SRR8204563
# 0.666666666666667__2.33333333333333 0.03458170 0.02433202 0.03163521 0.02541547 0.06570868
# 1__1.5                              0.13877409 0.14316568 0.14547642 0.13986424 0.14175485
# 0.571428571428571__2                0.05894608 0.05423322 0.05258609 0.04144859 0.04923890
# 0.8__1.6                            0.05770773 0.05952820 0.04792577 0.04566470 0.04597774
# 0.748921568627451__1.93460784313725 0.11497840 0.12520853 0.07654372 0.07274286 0.08436865



#  ANCOM-BC all results
res = out$res

class(res) # "data.frame"
names(res)
# [1] "taxon"            "lfc_(Intercept)"  "lfc_group2.L"     "se_(Intercept)"   "se_group2.L"      "W_(Intercept)"   
# [7] "W_group2.L"       "p_(Intercept)"    "p_group2.L"       "q_(Intercept)"    "q_group2.L"       "diff_(Intercept)"
# [13] "diff_group2.L"


tab_qval <- res[ ,c("taxon", "q_(Intercept)", "q_group2.L")]
dim(tab_qval) # 1845    3

summary( tab_qval )
# taxon           q_(Intercept)   q_group2.L       
# Length:1845        Min.   :1     Min.   :0.0000414  
# Class :character   1st Qu.:1     1st Qu.:1.0000000  
# Mode  :character   Median :1     Median :1.0000000  
#                    Mean   :1     Mean   :0.9974203  
#                    3rd Qu.:1     3rd Qu.:1.0000000  
#                    Max.   :1     Max.   :1.0000000 

names(tab_qval) # "taxon"         "q_(Intercept)" "q_group2.L"  
# Levels: A_low_PC1_behav < B_high_PC1_behav
names(tab_qval) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_qval) <- tab_qval$taxon

head(tab_qval)
#                                                                   taxon Intercept High PC1 cf. Low PC1
# 0.666666666666667__2.33333333333333 0.666666666666667__2.33333333333333         1                    1
# 1__1.5                                                           1__1.5         1                    1
# 0.571428571428571__2                               0.571428571428571__2         1                    1
# 0.8__1.6                                                       0.8__1.6         1                    1
# 0.748921568627451__1.93460784313725 0.748921568627451__1.93460784313725         1                    1
# 0.375__2                                                       0.375__2         1                    1

# sig diff abun vk-coords in IGT cf. Normal
sel.sig.HighPC1_cf_LowPC1 <- which(tab_qval$`High PC1 cf. Low PC1` <= 0.05) # qty 2
fxn.sig.HighPC1_cf_LowPC1 <- tab_qval$taxon[sel.sig.HighPC1_cf_LowPC1]
length(unique(fxn.sig.HighPC1_cf_LowPC1)) # 2


sel.sig <- unique(c(sel.sig.HighPC1_cf_LowPC1)) # qty 2
tab_qval[sel.sig, ]
#                                                                   taxon Intercept High PC1 cf. Low PC1
# 0.593253968253968__1.59325396825397 0.593253968253968__1.59325396825397         1         1.656226e-04
# 0.922222222222222__1.37777777777778 0.922222222222222__1.37777777777778         1         4.144386e-05


## log-fold change (LFC)
tab_lfc = res[ ,c("taxon", "lfc_(Intercept)",  "lfc_group2.L" )]
class(tab_lfc) # "data.frame"
names(tab_lfc) # "taxon"           "lfc_(Intercept)" "lfc_group2.L"
names(tab_lfc) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_lfc) <- tab_lfc$taxon
tab_lfc[sel.sig, ]
#                                                                   taxon     Intercept High PC1 cf. Low PC1
# 0.593253968253968__1.59325396825397 0.593253968253968__1.59325396825397 -3.210779e-16           -0.4082637
# 0.922222222222222__1.37777777777778 0.922222222222222__1.37777777777778 -1.200219e-16            0.4173254


## get significant diff abun taxa across any pairs (note this is not global test result)
tab_lfc.sig_all <- tab_lfc[ sel.sig, ]

dim(tab_lfc.sig_all) # 2 3

head(tab_lfc.sig_all)
#                                                                   taxon     Intercept High PC1 cf. Low PC1
# 0.593253968253968__1.59325396825397 0.593253968253968__1.59325396825397 -3.210779e-16           -0.4082637
# 0.922222222222222__1.37777777777778 0.922222222222222__1.37777777777778 -1.200219e-16            0.4173254

hist(tab_lfc.sig_all$Intercept); summary(tab_lfc.sig_all$Intercept)
# Min.    1st Qu.     Median       Mean    3rd Qu.       Max. 
# -3.211e-16 -2.708e-16 -2.205e-16 -2.205e-16 -1.703e-16 -1.200e-16 
hist(tab_lfc.sig_all$`High PC1 cf. Low PC1`); summary(tab_lfc.sig_all$`High PC1 cf. Low PC1`)
#      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# -0.408264 -0.201867  0.004531  0.004531  0.210928  0.417325 


names(tab_lfc.sig_all)[2] # "Intercept"

tab_lfc.sig_all.melt <- melt(tab_lfc.sig_all[ ,-2])
names(tab_lfc.sig_all.melt) # "taxon"    "variable" "value" 
names(tab_lfc.sig_all.melt) <- c("taxon", "Comparison groups", "Log-fold-change" )

p <- ggplot(tab_lfc.sig_all.melt, aes(`Log-fold-change`))+ #, color = `Comparison groups`)) +
  geom_density()+
  labs(x = "Log-fold-change", y = "Density") +
  # scale_colour_manual( values = c("High PC1 cf. Low PC1" = "#66c2a5"),
  #                      labels = c("High PC1 cf. Low PC1")) +
  theme_classic() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
p

grid.text(label = "ANCOMBC: vK coord-level | Females", x = unit(0.17, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=9), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","LFC-density-plots",header,this_study,"Females-v9d.tiff"), width = 9, height = 6, units = "cm", res=450, compression="lzw",type="cairo")


# assign "Positive LFC", "Negative LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- NA
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` < 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Negative LFC"
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` > 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Positive LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- factor(tab_lfc.sig_all.melt$pos_or_neg_LFC, levels = c("Positive LFC", "Negative LFC"), ordered = TRUE)


head(tab_lfc.sig_all.melt)
# taxon    Comparison groups Log-fold-change pos_or_neg_LFC
# 1 0.593253968253968__1.59325396825397 High PC1 cf. Low PC1      -0.4082637   Negative LFC
# 2 0.922222222222222__1.37777777777778 High PC1 cf. Low PC1       0.4173254   Positive LFC


table(tab_lfc.sig_all.melt$pos_or_neg_LFC, useNA = "ifany")
# Positive LFC Negative LFC 
#            1            1


## join vK coords to data table - derive directly from vK coords

res.sig <- tab_lfc.sig_all.melt
res.sig$OC_x <- NA
res.sig$HC_y <- NA

dim(res.sig) # 2 6 

for (i in 1:dim(res.sig)[1]) {
  #i<-1
  this_coord <- res.sig$taxon[i]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  
  res.sig$OC_x[i]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  res.sig$HC_y[i]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",i))
}


unique( res.sig$`Comparison groups` )
# [1] High PC1 cf. Low PC1
# Levels: High PC1 cf. Low PC1

# res.sig$`Comparison groups` <- factor(res.sig$`Comparison groups`, 
#                                       levels = c("IGT cf. Normal", "T2D Met neg cf. Normal", "T2D Met pos cf. Normal"),
#                                       labels = c("IGT cf. Normal", "T2D Met- cf. Normal", "T2D Met+ cf. Normal"),
#                                       ordered=TRUE)

ok <- complete.cases(res.sig)
sel.ok <- which(ok==TRUE)

res.sig.ok <- res.sig[sel.ok, ]

dim(res.sig)   #  2  6
dim(res.sig.ok)#  2  6

names(res.sig.ok)
# "taxon"             "Comparison groups" "Log-fold-change"   "pos_or_neg_LFC"    "OC_x"              "HC_y"             


summary(res.sig.ok$`Log-fold-change`);hist(res.sig.ok$`Log-fold-change`)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# -0.408264 -0.201867  0.004531  0.004531  0.210928  0.417325


p <- ggplot(res.sig.ok, aes(x = OC_x, y = HC_y, color = `Log-fold-change`))+
  geom_point(alpha = 0.2)

p


# overlay vK ref data ...

# sel <- which(df.vk.ref$sample %in% c("Acetate", "Propionate",  "Butyrate", "Riboflavin (Vit B2)" ,
#                                      "Cobalamin (Vit B12)", "Pyridoxal 5'-phosphate (Vit B6)", "Folate (Vit B9)", "Menaquinone (Vit K2)" ) )
# df.hac <- df.vk.ref[sel, ]
df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2

vkgrouprect

#library(colorspace)

p <- ggplot()+
  
  geom_point(data = filter(res.sig.ok, `Comparison groups`=="High PC1 cf. Low PC1"), aes(x = OC_x, y = HC_y, color = `Log-fold-change`), size = 3.5)+ # , alpha = 0.6
  
  ## ADJUST SCALE !!!!!!!!!!
  scale_color_continuous_sequential(palette = "BluYl", rev=FALSE, name="Log-fold\n-change", limits = c(-1, 1) )+ # palette = "YlGnBu" "BluYl"
  guides(color = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5))+
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  #geom_point(data = df.hac ,aes(x=OC_x, y=HC_y), color = "#253494")+ # , pch=8  "#80b1d3"
  #geom_text(data = df.hac ,aes(x=OC_x, y=HC_y, label=sample), color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.75, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "High cf. Low PC1 problem behaviors: vK coordinates | Females", x = unit(0.07, "npc") , y = unit(0.96,"npc"), gp=gpar(fontsize=10), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","vK-Coords-logfoldchange-scatterpoints-High-vs-Low-PC1-Problem-Behav-Females--v9d.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



## plot mean rel abun of each group alongside log-fold-change - for tab_lfc.sig_all


head(tab_lfc.sig_all)
#                                                                   taxon     Intercept High PC1 cf. Low PC1
# 0.593253968253968__1.59325396825397 0.593253968253968__1.59325396825397 -3.210779e-16           -0.4082637
# 0.922222222222222__1.37777777777778 0.922222222222222__1.37777777777778 -1.200219e-16            0.4173254

temp <- tab_lfc.sig_all

# add vK coordinates
temp$OC_x <- NA
temp$HC_y <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_coord <- temp$taxon[i]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  
  temp$OC_x[i]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  temp$HC_y[i]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",i))
}

# add compound
out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")
head( out[[1]] )
out.temp <- out[[1]]
row.names(out.temp) <- out.temp$coords

temp$compound <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_coord <- temp$taxon[i]
  if (is.na(temp$OC_x[i])){
    temp$compound[i] <- NA
  } else {
    sel <- which(out.temp$coords == this_coord)
    temp$compound[i] <- out.temp$compound[sel]
  }
  print(paste0("completed ",i))
}

head(temp)
#                                                                   taxon     Intercept High PC1 cf. Low PC1      OC_x     HC_y
# 0.593253968253968__1.59325396825397 0.593253968253968__1.59325396825397 -3.210779e-16           -0.4082637 0.5932540 1.593254
# 0.922222222222222__1.37777777777778 0.922222222222222__1.37777777777778 -1.200219e-16            0.4173254 0.9222222 1.377778
#                                         compound
# 0.593253968253968__1.59325396825397 amino_sugars
# 0.922222222222222__1.37777777777778      tannins


# join functions?
head(df.out.distil)

fxn_match <- df.out.distil
fxn_match$coords <- paste0(fxn_match$tot_mean_OC_x,"__",fxn_match$tot_mean_HC_y)
dim(fxn_match)
# 20787    10

# match fxns to vk-coords
temp$fxns <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_coord <- temp$taxon[i]
  if (is.na(temp$OC_x[i])){
    temp$fxns[i] <- NA
  } else {
    sel <- which(fxn_match$coords == this_coord)
    temp$fxns[i] <- paste0(fxn_match$superfocus_fxn[sel],collapse=";")
  }
  print(paste0("completed ",i))
}

list_fxns <- paste0(temp$fxns, collapse=";")
list_fxns <- str_split_1(string=list_fxns, pattern=";")
list_fxns <- sort(list_fxns)

head(phy@tax_table)

#test
#phy@tax_table[ c("fxn_6","fxn_2") , ]

write.csv(x = temp, file = "ANCOMBC-vKcoord-level-Sig-diff-abun-fxns--Females-Child_behav-LFC-and-Function-list-Part1.csv")
write.csv(x = as.data.frame( phy@tax_table[ list_fxns , ]), file = "ANCOMBC-vKcoord-level-Sig-diff-abun-fxns--Females-Child_behav-LFC-and-Function-Details-Part2.csv")

temp
#                                                                   taxon     Intercept High PC1 cf. Low PC1      OC_x     HC_y
# 0.593253968253968__1.59325396825397 0.593253968253968__1.59325396825397 -3.210779e-16           -0.4082637 0.5932540 1.593254
# 0.922222222222222__1.37777777777778 0.922222222222222__1.37777777777778 -1.200219e-16            0.4173254 0.9222222 1.377778
# compound                fxns
# 0.593253968253968__1.59325396825397 amino_sugars fxn_12938;fxn_12942
# 0.922222222222222__1.37777777777778      tannins           fxn_14491

as.data.frame( phy@tax_table[ list_fxns , ])
# subsys_L1   subsys_L2                  subsys_L3
# fxn_12938               Miscellaneous           - EC 4.1.2.- Aldehyde-lyases
# fxn_12942               Miscellaneous           - EC 4.1.2.- Aldehyde-lyases
# fxn_14491 Nucleosides and Nucleotides Pyrimidines     pyrimidine conversions
# fxn
# fxn_12938                                          Dihydroneopterin_aldolase_(EC_4.1.2.25)
# fxn_12942 Dihydroneopterin_phosphate_phosphatase_/_Dihydroneopterin_aldolase_(EC_4.1.2.25)
# fxn_14491                                               Uridine_phosphorylase_(EC_2.4.2.3)


## plot heatmap of LFC for sig diff abun Functions - ANCOMBC Function-level

m2 <- tab_lfc.sig_all[ ,-1] # exclude 'taxon' column, it is in row.name anyway


row.names(m2) <- paste0(round(temp$OC_x,4),"__",round(temp$HC_y,4))

names(m2) #  "Intercept"            "High PC1 cf. Low PC1"
names(m2) <- c("Intercept", "High PC1\ncf. Low PC1" )

m2 <- as.matrix(m2)
colnames(m2) # "Intercept"             "High PC1\ncf. Low PC1"
m2
class(m2) # "matrix" "array" 



pheatmap(m2, cluster_cols = FALSE, #cluster_row = FALSE, 
         #color = color,
         color = hcl.colors(50, "BluYl"), # palette = "BrBG", direction = 1
         border_color = FALSE, 
         fontsize_col = 12,
         fontsize_main = 13,
         #fontsize_row = 8,
         main = "Differentially abundant vK-coords\n(log-fold-changes): Females"
)

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-LogFoldChange-ANCOMBC-vKcoord-level-Child-Behav-Females--v9d.tiff"), width = 10, height = 14, units = "cm", res=450, compression="lzw",type="cairo")



ph <- pheatmap(m2, cluster_cols = FALSE)
#ph_taxa_rownames <- ph$tree_row$labels[ ph$tree_row$order ]
ph_taxa_rownames <- row.names(tab_lfc.sig_all)[ ph$tree_row$order ]



## compare with relative abundance of functions in these groups

phy.relabun <- phy.vk.ok.f

m <- as.matrix( as.data.frame(phy.relabun@otu_table[ ph_taxa_rownames , ]))
identical(colnames(m), phy.relabun@sam_data$Run) # TRUE

names( phy.relabun@sam_data )

row.names(m)
# "0.593253968253968__1.59325396825397" "0.922222222222222__1.37777777777778"

paste0(round(temp$OC_x,4),"__",round(temp$HC_y,4))
# "0.5933__1.5933" "0.9222__1.3778"

row.names(m) <- paste0(round(temp$OC_x,4),"__",round(temp$HC_y,4))


levels( phy.relabun@sam_data$group2 ) 
# "A_low_PC1_behav"  "B_high_PC1_behav"
table( phy.relabun@sam_data$group ) 
# high_PC1_behav  low_PC1_behav 
# 10             10 

phy.relabun@sam_data$label <- phy.relabun@sam_data$group
phy.relabun@sam_data$label <- factor(phy.relabun@sam_data$label,
                                     levels = c("low_PC1_behav","high_PC1_behav"),
                                     labels = c("Low PC1","High PC1"),
                                     ordered=TRUE)
sample_order <- order(phy.relabun@sam_data$group2)
phy.relabun@sam_data$group_label <- paste0(phy.relabun@sam_data$label," (",phy.relabun@sam_data$Run,")")
phy.relabun@sam_data$group_label[sample_order]
phy.relabun@sam_data$group_label <- factor(phy.relabun@sam_data$group_label,
                                           levels = phy.relabun@sam_data$group_label[sample_order],
                                           ordered=TRUE)
colnames(m) <- phy.relabun@sam_data$group_label
# reorder columns
m <- m[ , sample_order]

# No dendrogram nor reordering for neither column or row
heatmap(m, Colv = NA, Rowv = NA, scale="column")
heatmap(m, Colv = NA, Rowv = NA)

# https://www.reneshbedre.com/blog/heatmap-with-pheatmap-package-r.html
#library(pheatmap)

pheatmap(m, cluster_cols = FALSE,  cluster_row = FALSE, # gaps_col = c(6,12,18), 
         border_color = FALSE, main = "vK-coordinate relative abundance (%)")


group_labels <- data.frame( Group = phy.relabun@sam_data$label[sample_order] )
row.names(group_labels) <- colnames(m)
dim(group_labels) # 20  1

dim(m) # 2 20

cols = c("Low PC1" = "#7fbc41",  "High PC1" = "#de77ae") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9

mycolors <- c("#7fbc41","#de77ae")
names(mycolors) <- c("Low PC1","High PC1")

mycolors <- list(Group = mycolors)

color <- colorRampPalette((c("#fff5f0","#fc9272", "#cb181d")))(50)
pheatmap(m, gaps_col = c(10), cluster_cols = FALSE, cluster_row = FALSE, # 49          43          33          20
         annotation_col = group_labels,
         annotation_colors = mycolors,
         color = color,
         fontsize_col = 8,
         border_color = FALSE, 
         main = "Rel. abun. vK coords (%): Females")

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-RelAbun-for-ANCOMBC-vKCoord-level-Child-Behav-Females--v9d.tiff"), width = 14, height = 14, units = "cm", res=450, compression="lzw",type="cairo")





## Males

sel.m <- which(phy.vk.ok@sam_data$sex == "male") # qty 17
phy.vk.ok.m <- prune_samples(samples = row.names(phy.vk.ok@sam_data)[sel.m], phy.vk.ok)

# prune taxa with zero reads
phy.vk.ok.m <- prune_taxa(taxa = taxa_sums(phy.vk.ok.m) > 0, x = phy.vk.ok.m)
phy.vk.ok.m
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2350 taxa and 17 samples ]
# sample_data() Sample Data:       [ 17 samples by 81 sample variables ]
# tax_table()   Taxonomy Table:    [ 2350 taxa by 4 taxonomic ranks ]


## ANCOMBC - differential abundance testing??
set.seed(123)
out.ancombc.vk = ancombc2(data = phy.vk.ok.m,
                          tax_level = NULL,
                          fix_formula = "group2",
                          p_adj_method = "holm", prv_cut = 0.1, lib_cut = 0,
                          group = "group2", struc_zero = TRUE, neg_lb = TRUE,
                          alpha = 0.05, global = TRUE,  # change alpha level to reduce number !!!
                          n_cl = 6, verbose = TRUE)

# `tax_level` is not speficified
# No agglomeration will be performed
# Otherwise, please speficy `tax_level` by one of the following:
#
#   Obtaining initial estimates ...
# Estimating sample-specific biases ...
# ANCOM-BC2 primary results ...
# The sensitivity analysis for the pseudo-count addition ...
# Warning message:
#   The group variable has < 3 categories
# The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated

str(out.ancombc.vk)
# List of 11
# $ feature_table  : num [1:1741, 1:17] 0.0373 0.1449 0.0512 0.0435 0.0631 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:1741] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# .. ..$ : chr [1:17] "SRR8204553" "SRR8204554" "SRR8204556" "SRR8204557" ...
# $ zero_ind       :'data.frame':	2350 obs. of  3 variables:
#   ..$ taxon                                      : chr [1:2350] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# ..$ structural_zero (group2 = A_low_PC1_behav) : logi [1:2350] FALSE FALSE FALSE FALSE FALSE FALSE ...
# ..$ structural_zero (group2 = B_high_PC1_behav): logi [1:2350] FALSE FALSE FALSE FALSE FALSE FALSE ...
# $ samp_frac      : num [1:17] 0.226 -0.442 0.398 -0.503 2.956 ...
# $ delta_em       : num [1:2] -0.00672 -0.15438
# $ delta_wls      : num [1:2] -0.00672 -0.15494
# $ pseudo_sens_tab:'data.frame':	1741 obs. of  3 variables:
#   ..$ taxon      : chr [1:1741] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# ..$ (Intercept): num [1:1741] 0.00687 0.00524 0.00161 0.00108 0.00146 ...
# ..$ group2.L   : num [1:1741] 0.1914 0.1052 0.0397 0.025 0.0324 ...
# $ res            :'data.frame':	1741 obs. of  13 variables:
#   ..$ taxon           : chr [1:1741] "0.666666666666667__2.33333333333333" "1__1.5" "0.571428571428571__2" "0.8__1.6" ...
# ..$ lfc_(Intercept) : num [1:1741] 0.021453 -0.000179 0.009533 0.00504 0.005811 ...
# ..$ lfc_group2.L    : num [1:1741] 0.5085 -0.0115 0.2219 0.1139 0.1325 ...
# ..$ se_(Intercept)  : num [1:1741] 0.1483 0.0759 0.0754 0.0788 0.1107 ...
# ..$ se_group2.L     : num [1:1741] 0.21 0.107 0.107 0.111 0.157 ...
# ..$ W_(Intercept)   : num [1:1741] 0.14468 -0.00236 0.12637 0.06394 0.05249 ...
# ..$ W_group2.L      : num [1:1741] 2.425 -0.108 2.081 1.022 0.846 ...
# ..$ p_(Intercept)   : num [1:1741] 0.885 0.998 0.899 0.949 0.958 ...
# ..$ p_group2.L      : num [1:1741] 0.0153 0.9143 0.0375 0.3067 0.3975 ...
# ..$ q_(Intercept)   : num [1:1741] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ q_group2.L      : num [1:1741] 1 1 1 1 1 1 1 1 1 1 ...
# ..$ diff_(Intercept): logi [1:1741] FALSE FALSE FALSE FALSE FALSE FALSE ...
# ..$ diff_group2.L   : logi [1:1741] FALSE FALSE FALSE FALSE FALSE FALSE ...
# $ res_global     : NULL
# $ res_pair       : NULL
# $ res_dunn       : NULL
# $ res_trend      : NULL

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/PCaN-NZ/nz-city-resto/modelling/R"
saveRDS(object = out.ancombc.vk, file = "out.ancombc.vk.Child_behav.Males.RDS")

#out.ancombc.vk <- readRDS("out.ancombc.vk.Child_behav.Males.RDS")


# # # # #



out <- out.ancombc.vk


out$feature_table[1:5, 1:5]
#                                     SRR8204553 SRR8204554 SRR8204556 SRR8204557 SRR8204558
# 0.666666666666667__2.33333333333333 0.03727985 0.03962676 0.05004921 0.01877806 0.02145462
# 1__1.5                              0.14494060 0.14935700 0.12002455 0.17640501 0.21454623
# 0.571428571428571__2                0.05120914 0.05541412 0.05517886 0.04038060 0.04290925
# 0.8__1.6                            0.04351004 0.05039567 0.05173311 0.04188440 0.06436387
# 0.748921568627451__1.93460784313725 0.06311235 0.07893679 0.09172570 0.08779742 0.06436387

fxns_out <- row.names(out$feature_table)

dim(out$feature_table) #  1741   17
phy.vk.ok.m@otu_table[1:5, 1:5]
# OTU Table:          [5 taxa and 5 samples]
# taxa are rows
#                                     SRR8204553 SRR8204554 SRR8204556 SRR8204557 SRR8204558
# 0.666666666666667__2.33333333333333 0.03727985 0.03962676 0.05004921 0.01877806 0.02145462
# 1__1.5                              0.14494060 0.14935700 0.12002455 0.17640501 0.21454623
# 0.571428571428571__2                0.05120914 0.05541412 0.05517886 0.04038060 0.04290925
# 0.8__1.6                            0.04351004 0.05039567 0.05173311 0.04188440 0.06436387
# 0.748921568627451__1.93460784313725 0.06311235 0.07893679 0.09172570 0.08779742 0.06436387



#  ANCOM-BC all results
res = out$res

class(res) # "data.frame"
names(res)
# [1] "taxon"            "lfc_(Intercept)"  "lfc_group2.L"     "se_(Intercept)"   "se_group2.L"      "W_(Intercept)"   
# [7] "W_group2.L"       "p_(Intercept)"    "p_group2.L"       "q_(Intercept)"    "q_group2.L"       "diff_(Intercept)"
# [13] "diff_group2.L"


tab_qval <- res[ ,c("taxon", "q_(Intercept)", "q_group2.L")]
dim(tab_qval) # 1741   3

summary( tab_qval )
# taxon           q_(Intercept)   q_group2.L      
# Length:1741        Min.   :1     Min.   :0.001034  
# Class :character   1st Qu.:1     1st Qu.:1.000000  
# Mode  :character   Median :1     Median :1.000000  
#                    Mean   :1     Mean   :0.998550  
#                    3rd Qu.:1     3rd Qu.:1.000000  
#                    Max.   :1     Max.   :1.000000

names(tab_qval) # "taxon"         "q_(Intercept)" "q_group2.L"  
# Levels: A_low_PC1_behav < B_high_PC1_behav
names(tab_qval) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_qval) <- tab_qval$taxon

head(tab_qval)
#                                                                   taxon Intercept High PC1 cf. Low PC1
# 0.666666666666667__2.33333333333333 0.666666666666667__2.33333333333333         1                    1
# 1__1.5                                                           1__1.5         1                    1
# 0.571428571428571__2                               0.571428571428571__2         1                    1
# 0.8__1.6                                                       0.8__1.6         1                    1
# 0.748921568627451__1.93460784313725 0.748921568627451__1.93460784313725         1                    1
# 0.375__2                                                       0.375__2         1                    1

# sig diff abun vk-coords in IGT cf. Normal
sel.sig.HighPC1_cf_LowPC1 <- which(tab_qval$`High PC1 cf. Low PC1` <= 0.05) # qty 2
fxn.sig.HighPC1_cf_LowPC1 <- tab_qval$taxon[sel.sig.HighPC1_cf_LowPC1]
length(unique(fxn.sig.HighPC1_cf_LowPC1)) # 2


sel.sig <- unique(c(sel.sig.HighPC1_cf_LowPC1)) # qty 2
tab_qval[sel.sig, ]
#                                                                   taxon Intercept High PC1 cf. Low PC1
# 1.55__0.753333333333333                         1.55__0.753333333333333         1           0.02643157
# 0.583308133090742__1.44842868819318 0.583308133090742__1.44842868819318         1           0.00103441


## log-fold change (LFC)
tab_lfc = res[ ,c("taxon", "lfc_(Intercept)",  "lfc_group2.L" )]
class(tab_lfc) # "data.frame"
names(tab_lfc) # "taxon"           "lfc_(Intercept)" "lfc_group2.L"
names(tab_lfc) <- c("taxon", "Intercept", "High PC1 cf. Low PC1")
row.names(tab_lfc) <- tab_lfc$taxon
tab_lfc[sel.sig, ]
#                                                                   taxon  Intercept High PC1 cf. Low PC1
# 1.55__0.753333333333333                         1.55__0.753333333333333 0.03729669            0.8894350
# 0.583308133090742__1.44842868819318 0.583308133090742__1.44842868819318 0.02002620            0.4742243


## get significant diff abun taxa across any pairs (note this is not global test result)
tab_lfc.sig_all <- tab_lfc[ sel.sig, ]

dim(tab_lfc.sig_all) # 2 3

head(tab_lfc.sig_all)
#                                                                   taxon  Intercept High PC1 cf. Low PC1
# 1.55__0.753333333333333                         1.55__0.753333333333333 0.03729669            0.8894350
# 0.583308133090742__1.44842868819318 0.583308133090742__1.44842868819318 0.02002620            0.4742243

hist(tab_lfc.sig_all$Intercept); summary(tab_lfc.sig_all$Intercept)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.02003 0.02434 0.02866 0.02866 0.03298 0.03730 
hist(tab_lfc.sig_all$`High PC1 cf. Low PC1`); summary(tab_lfc.sig_all$`High PC1 cf. Low PC1`)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.4742  0.5780  0.6818  0.6818  0.7856  0.8894


names(tab_lfc.sig_all)[2] # "Intercept"

tab_lfc.sig_all.melt <- melt(tab_lfc.sig_all[ ,-2])
names(tab_lfc.sig_all.melt) # "taxon"    "variable" "value" 
names(tab_lfc.sig_all.melt) <- c("taxon", "Comparison groups", "Log-fold-change" )

p <- ggplot(tab_lfc.sig_all.melt, aes(`Log-fold-change`))+ #, color = `Comparison groups`)) +
  geom_density()+
  labs(x = "Log-fold-change", y = "Density") +
  # scale_colour_manual( values = c("High PC1 cf. Low PC1" = "#66c2a5"),
  #                      labels = c("High PC1 cf. Low PC1")) +
  theme_classic() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
p

grid.text(label = "ANCOMBC: vK coord-level | Males", x = unit(0.17, "npc") , y = unit(0.95,"npc"), gp=gpar(fontsize=9), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","LFC-density-plots",header,this_study,"Males-v9d.tiff"), width = 9, height = 6, units = "cm", res=450, compression="lzw",type="cairo")


# assign "Positive LFC", "Negative LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- NA
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` < 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Negative LFC"
sel <- which(tab_lfc.sig_all.melt$`Log-fold-change` > 0)
tab_lfc.sig_all.melt$pos_or_neg_LFC[sel] <- "Positive LFC"
tab_lfc.sig_all.melt$pos_or_neg_LFC <- factor(tab_lfc.sig_all.melt$pos_or_neg_LFC, levels = c("Positive LFC", "Negative LFC"), ordered = TRUE)


head(tab_lfc.sig_all.melt)
#                                 taxon    Comparison groups Log-fold-change pos_or_neg_LFC
# 1             1.55__0.753333333333333 High PC1 cf. Low PC1       0.8894350   Positive LFC
# 2 0.583308133090742__1.44842868819318 High PC1 cf. Low PC1       0.4742243   Positive LFC


table(tab_lfc.sig_all.melt$pos_or_neg_LFC, useNA = "ifany")
# Positive LFC Negative LFC 
#            2            0


## join vK coords to data table - derive directly from vK coords

res.sig <- tab_lfc.sig_all.melt
res.sig$OC_x <- NA
res.sig$HC_y <- NA

dim(res.sig) # 2 6 

for (i in 1:dim(res.sig)[1]) {
  #i<-1
  this_coord <- res.sig$taxon[i]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  
  res.sig$OC_x[i]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  res.sig$HC_y[i]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",i))
}


unique( res.sig$`Comparison groups` )
# [1] High PC1 cf. Low PC1
# Levels: High PC1 cf. Low PC1

# res.sig$`Comparison groups` <- factor(res.sig$`Comparison groups`, 
#                                       levels = c("IGT cf. Normal", "T2D Met neg cf. Normal", "T2D Met pos cf. Normal"),
#                                       labels = c("IGT cf. Normal", "T2D Met- cf. Normal", "T2D Met+ cf. Normal"),
#                                       ordered=TRUE)

ok <- complete.cases(res.sig)
sel.ok <- which(ok==TRUE)

res.sig.ok <- res.sig[sel.ok, ]

dim(res.sig)   #  2  6
dim(res.sig.ok)#  2  6

names(res.sig.ok)
# "taxon"             "Comparison groups" "Log-fold-change"   "pos_or_neg_LFC"    "OC_x"              "HC_y"             


summary(res.sig.ok$`Log-fold-change`);hist(res.sig.ok$`Log-fold-change`)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.4742  0.5780  0.6818  0.6818  0.7856  0.8894


p <- ggplot(res.sig.ok, aes(x = OC_x, y = HC_y, color = `Log-fold-change`))+
  geom_point(alpha = 0.2)

p


# overlay vK ref data ...

# sel <- which(df.vk.ref$sample %in% c("Acetate", "Propionate",  "Butyrate", "Riboflavin (Vit B2)" ,
#                                      "Cobalamin (Vit B12)", "Pyridoxal 5'-phosphate (Vit B6)", "Folate (Vit B9)", "Menaquinone (Vit K2)" ) )
# df.hac <- df.vk.ref[sel, ]
df.hac
#                            sample       OC_x     HC_y                  group short_name
# 1                         Acetate 1.00000000 2.000000 Short chain fatty acid    Acetate
# 2                      Propionate 0.66666667 2.000000 Short chain fatty acid Propionate
# 3                        Butyrate 0.50000000 2.000000 Short chain fatty acid   Butyrate
# 4             Riboflavin (Vit B2) 0.35294118 1.176471                Vitamin      VitB2
# 5             Cobalamin (Vit B12) 0.22580645 1.419355                Vitamin     VitB12
# 6 Pyridoxal 5'-phosphate (Vit B6) 0.75000000 1.250000                Vitamin      VitB6
# 7                 Folate (Vit B9) 0.31578947 1.000000                Vitamin      VitB9
# 8            Menaquinone (Vit K2) 0.06451613 1.290323                Vitamin      VitK2

vkgrouprect

#library(colorspace)

p <- ggplot()+
  
  geom_point(data = filter(res.sig.ok, `Comparison groups`=="High PC1 cf. Low PC1"), aes(x = OC_x, y = HC_y, color = `Log-fold-change`), size = 3.5)+ # , alpha = 0.6
  
  ## ADJUST SCALE !!!!!!!!!!
  scale_color_continuous_sequential(palette = "BluYl", rev=FALSE, name="Log-fold\n-change", limits = c(-1, 1) )+ # palette = "YlGnBu" "BluYl"
  guides(color = guide_colorbar(direction = "horizontal", barheight = 0.7, vjust = 0, hjust = 0.5))+
  
  xlab("O:C molar ratio")+ ylab("H:C molar ratio")+
  
  theme_bw()+
  
  geom_mark_rect(data=vkgrouprect, aes(x = OC_x, y = HC_y, group = label), color="#737373", expand = unit(0, "mm"),radius = unit(0, "mm"), position = "identity", inherit.aes=FALSE )+
  #geom_point(data = df.hac ,aes(x=OC_x, y=HC_y), color = "#253494")+ # , pch=8  "#80b1d3"
  #geom_text(data = df.hac ,aes(x=OC_x, y=HC_y, label=sample), color = "#253494", hjust=0, vjust=1,size = 3, nudge_x = 0.01)+ # , pch=8.  color = "#e6550d",
  annotate(geom="text", x= 0+0.01, y= 2.3+0.02, label = "Lipid", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.2+0.01, y= 2.2+0.02, label = "Protein", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.52+0.01, y= 2.2+0.02, label = "Amino sugar", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0.7+0.01, y= 2.4+0.02, label = "Carbohydrate", hjust=0, vjust=0, size = 3 , col="#737373") + # top-left
  annotate(geom="text", x= 0+0.01, y= 0.5-0.02, label = "Condensed aromatics", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  annotate(geom="text", x= 0.25+0.01, y= 0.75-0.02, label = "Lignin", hjust=0, vjust=1, size = 3 , col="#737373" ) + # bottom-left
  annotate(geom="text", x= 0.67+0.01, y= 0.53-0.02, label = "Tannin", hjust=0, vjust=1, size = 3 , col="#737373") + # bottom-left
  
  annotate(geom="text", x= 0.35, y= 2.65, label = "Other methylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 0.8, label = "Other demethylated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0.35, y= 0.2, label = "Other condensed\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 1.3, y= 2.3, label = "Other hydrated\ncompounds", hjust=0.5, vjust=0.5, size = 3 , col="#737373", lineheight = 0.8) +
  annotate(geom="text", x= 0+0.01, y= 1.5-0.02, label = "Near VitK2-VitB12\ncompounds", hjust=0, vjust=1, size = 3 , col="#737373", lineheight = 0.8) +
  
  geom_segment(aes(x = 1.1, y = 2.4, xend = 1.1,  yend = 3), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 1.1, y = 1.5, xend = 2.1,  yend = 1.5), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  geom_segment(aes(x = 0.97, y = 0.53, xend = 0.97,  yend = 0), inherit.aes = FALSE , linetype = "dashed", color = "#737373")+
  
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    #legend.position = c(0.5, 0.05), # as fraction of plot dimensions
    legend.position = c(0.75, 0.05), # as fraction of plot dimensions
    legend.background = element_rect(fill = "transparent"),
    legend.title = element_text(size = rel(0.9))
  )
p

grid.text(label = "High cf. Low PC1 problem behaviors: vK coordinates | Males", x = unit(0.07, "npc") , y = unit(0.96,"npc"), gp=gpar(fontsize=10), hjust=0 )
dev.print(tiff, file = paste0(workdir,"/plots/","vK-Coords-logfoldchange-scatterpoints-High-vs-Low-PC1-Problem-Behav-Males--v9d.tiff"), width = 16, height = 12, units = "cm", res=450, compression="lzw",type="cairo")



## plot mean rel abun of each group alongside log-fold-change - for tab_lfc.sig_all


head(tab_lfc.sig_all)
# taxon  Intercept High PC1 cf. Low PC1
# 1.55__0.753333333333333                         1.55__0.753333333333333 0.03729669            0.8894350
# 0.583308133090742__1.44842868819318 0.583308133090742__1.44842868819318 0.02002620            0.4742243

temp <- tab_lfc.sig_all

# add vK coordinates
temp$OC_x <- NA
temp$HC_y <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_coord <- temp$taxon[i]
  end_x <- str_locate(string = this_coord, pattern = "__")[1]-1
  start_y <- str_locate(string = this_coord, pattern = "__")[2]+1
  
  temp$OC_x[i]  <- as.numeric( substring(text = this_coord, first =1, last = end_x ))
  temp$HC_y[i]  <- as.numeric( substring(text = this_coord, first =start_y, last = nchar(this_coord) ) )
  print(paste0("completed ",i))
}

# add compound
out <- readRDS("out.list.get_tot_relabun_and_compound_at_coords.Child_behav.RDS")
head( out[[1]] )
out.temp <- out[[1]]
row.names(out.temp) <- out.temp$coords

temp$compound <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_coord <- temp$taxon[i]
  if (is.na(temp$OC_x[i])){
    temp$compound[i] <- NA
  } else {
    sel <- which(out.temp$coords == this_coord)
    temp$compound[i] <- out.temp$compound[sel]
  }
  print(paste0("completed ",i))
}

head(temp)
# taxon  Intercept High PC1 cf. Low PC1      OC_x
# 1.55__0.753333333333333                         1.55__0.753333333333333 0.03729669            0.8894350 1.5500000
# 0.583308133090742__1.44842868819318 0.583308133090742__1.44842868819318 0.02002620            0.4742243 0.5833081
# HC_y     compound
# 1.55__0.753333333333333             0.7533333 demethylated
# 0.583308133090742__1.44842868819318 1.4484287       lignin

# join functions?
head(df.out.distil)

fxn_match <- df.out.distil
fxn_match$coords <- paste0(fxn_match$tot_mean_OC_x,"__",fxn_match$tot_mean_HC_y)
dim(fxn_match)
# 20787    10

# match fxns to vk-coords
temp$fxns <- NA

for (i in 1:dim(temp)[1]) {
  #i<-1
  this_coord <- temp$taxon[i]
  if (is.na(temp$OC_x[i])){
    temp$fxns[i] <- NA
  } else {
    sel <- which(fxn_match$coords == this_coord)
    temp$fxns[i] <- paste0(fxn_match$superfocus_fxn[sel],collapse=";")
  }
  print(paste0("completed ",i))
}

list_fxns <- paste0(temp$fxns, collapse=";")
list_fxns <- str_split_1(string=list_fxns, pattern=";")
list_fxns <- sort(list_fxns)

head(phy@tax_table)

#test
#phy@tax_table[ c("fxn_6","fxn_2") , ]

write.csv(x = temp, file = "ANCOMBC-vKcoord-level-Sig-diff-abun-fxns--Males-Child_behav-LFC-and-Function-list-Part1.csv")
write.csv(x = as.data.frame( phy@tax_table[ list_fxns , ]), file = "ANCOMBC-vKcoord-level-Sig-diff-abun-fxns--Males-Child_behav-LFC-and-Function-Details-Part2.csv")

temp
# taxon  Intercept High PC1 cf. Low PC1      OC_x
# 1.55__0.753333333333333                         1.55__0.753333333333333 0.03729669            0.8894350 1.5500000
# 0.583308133090742__1.44842868819318 0.583308133090742__1.44842868819318 0.02002620            0.4742243 0.5833081
# HC_y     compound
# 1.55__0.753333333333333             0.7533333 demethylated
# 0.583308133090742__1.44842868819318 1.4484287       lignin
# fxns
# 1.55__0.753333333333333                                                                      fxn_2926
# 0.583308133090742__1.44842868819318 fxn_821;fxn_2958;fxn_2973;fxn_12703;fxn_12705;fxn_12786;fxn_12788

as.data.frame( phy@tax_table[ list_fxns , ])
# subsys_L1                                                subsys_L2
# fxn_12703 Metabolism of Aromatic Compounds             Metabolism of central aromatic intermediates
# fxn_12705 Metabolism of Aromatic Compounds             Metabolism of central aromatic intermediates
# fxn_12786 Metabolism of Aromatic Compounds Peripheral pathways for catabolism of aromatic compounds
# fxn_12788 Metabolism of Aromatic Compounds Peripheral pathways for catabolism of aromatic compounds
# fxn_2926                     Carbohydrates                          Central carbohydrate metabolism
# fxn_2958                     Carbohydrates                          Central carbohydrate metabolism
# fxn_2973                     Carbohydrates                          Central carbohydrate metabolism
# fxn_821        Amino Acids and Derivatives                     Aromatic amino acids and derivatives
# subsys_L3
# fxn_12703 Central meta-cleavage pathway of aromatic compound degradation
# fxn_12705 Central meta-cleavage pathway of aromatic compound degradation
# fxn_12786                                           Biphenyl Degradation
# fxn_12788                                           Biphenyl Degradation
# fxn_2926               Pyruvate metabolism I: anaplerotic reactions, PEP
# fxn_2958  Pyruvate metabolism II: acetyl-CoA, acetogenesis from pyruvate
# fxn_2973  Pyruvate metabolism II: acetyl-CoA, acetogenesis from pyruvate
# fxn_821                                            Tryptophan catabolism
# fxn
# fxn_12703                                     Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_/_Alcohol_dehydrogenase_(EC_1.1.1.1)
# fxn_12705 Alcohol_dehydrogenase_(EC_1.1.1.1)_@_Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_@_Pyruvate-formate-lyase_deactivase
# fxn_12786                                     Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_/_Alcohol_dehydrogenase_(EC_1.1.1.1)
# fxn_12788 Alcohol_dehydrogenase_(EC_1.1.1.1)_@_Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_@_Pyruvate-formate-lyase_deactivase
# fxn_2926                                                              Phosphoenolpyruvate_carboxykinase_[GTP]_(EC_4.1.1.32)
# fxn_2958                                      Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_/_Alcohol_dehydrogenase_(EC_1.1.1.1)
# fxn_2973  Alcohol_dehydrogenase_(EC_1.1.1.1)_@_Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_@_Pyruvate-formate-lyase_deactivase
# fxn_821                                       Acetaldehyde_dehydrogenase_(EC_1.2.1.10)_/_Alcohol_dehydrogenase_(EC_1.1.1.1)


## plot heatmap of LFC for sig diff abun Functions - ANCOMBC Function-level

m2 <- tab_lfc.sig_all[ ,-1] # exclude 'taxon' column, it is in row.name anyway


row.names(m2) <- paste0(round(temp$OC_x,4),"__",round(temp$HC_y,4))

names(m2) #  "Intercept"            "High PC1 cf. Low PC1"
names(m2) <- c("Intercept", "High PC1\ncf. Low PC1" )

m2 <- as.matrix(m2)
colnames(m2) # "Intercept"             "High PC1\ncf. Low PC1"
m2
class(m2) # "matrix" "array" 



pheatmap(m2, cluster_cols = FALSE, #cluster_row = FALSE, 
         #color = color,
         color = hcl.colors(50, "BluYl"), # palette = "BrBG", direction = 1
         border_color = FALSE, 
         fontsize_col = 12,
         fontsize_main = 13,
         #fontsize_row = 8,
         main = "Differentially abundant vK-coords\n(log-fold-changes): Males"
)

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-LogFoldChange-ANCOMBC-vKcoord-level-Child-Behav-Males--v9d.tiff"), width = 10, height = 14, units = "cm", res=450, compression="lzw",type="cairo")



ph <- pheatmap(m2, cluster_cols = FALSE)
#ph_taxa_rownames <- ph$tree_row$labels[ ph$tree_row$order ]
ph_taxa_rownames <- row.names(tab_lfc.sig_all)[ ph$tree_row$order ]



## compare with relative abundance of functions in these groups

phy.relabun <- phy.vk.ok.m

m <- as.matrix( as.data.frame(phy.relabun@otu_table[ ph_taxa_rownames , ]))
identical(colnames(m), phy.relabun@sam_data$Run) # TRUE

names( phy.relabun@sam_data )

row.names(m)
# "1.55__0.753333333333333"             "0.583308133090742__1.44842868819318"

paste0(round(temp$OC_x,4),"__",round(temp$HC_y,4))
# "1.55__0.7533"   "0.5833__1.4484"

row.names(m) <- paste0(round(temp$OC_x,4),"__",round(temp$HC_y,4))


levels( phy.relabun@sam_data$group2 ) 
# "A_low_PC1_behav"  "B_high_PC1_behav"
table( phy.relabun@sam_data$group ) 
# high_PC1_behav  low_PC1_behav 
# 8             9 

phy.relabun@sam_data$label <- phy.relabun@sam_data$group
phy.relabun@sam_data$label <- factor(phy.relabun@sam_data$label,
                                     levels = c("low_PC1_behav","high_PC1_behav"),
                                     labels = c("Low PC1","High PC1"),
                                     ordered=TRUE)
sample_order <- order(phy.relabun@sam_data$group2)
phy.relabun@sam_data$group_label <- paste0(phy.relabun@sam_data$label," (",phy.relabun@sam_data$Run,")")
phy.relabun@sam_data$group_label[sample_order]
phy.relabun@sam_data$group_label <- factor(phy.relabun@sam_data$group_label,
                                           levels = phy.relabun@sam_data$group_label[sample_order],
                                           ordered=TRUE)
colnames(m) <- phy.relabun@sam_data$group_label
# reorder columns
m <- m[ , sample_order]

# No dendrogram nor reordering for neither column or row
heatmap(m, Colv = NA, Rowv = NA, scale="column")
heatmap(m, Colv = NA, Rowv = NA)

# https://www.reneshbedre.com/blog/heatmap-with-pheatmap-package-r.html
#library(pheatmap)

pheatmap(m, cluster_cols = FALSE,  cluster_row = FALSE, # gaps_col = c(6,12,18), 
         border_color = FALSE, main = "vK-coordinate relative abundance (%)")


group_labels <- data.frame( Group = phy.relabun@sam_data$label[sample_order] )
row.names(group_labels) <- colnames(m)
dim(group_labels) # 17  1

dim(m) # 2 17

cols = c("Low PC1" = "#7fbc41",  "High PC1" = "#de77ae") # https://colorbrewer2.org/#type=diverging&scheme=Spectral&n=9

mycolors <- c("#7fbc41","#de77ae")
names(mycolors) <- c("Low PC1","High PC1")

mycolors <- list(Group = mycolors)

color <- colorRampPalette((c("#fff5f0","#fc9272", "#cb181d")))(50)
pheatmap(m, gaps_col = c(9), cluster_cols = FALSE, cluster_row = FALSE, # 49          43          33          20
         annotation_col = group_labels,
         annotation_colors = mycolors,
         color = color,
         fontsize_col = 8,
         border_color = FALSE, 
         main = "Rel. abun. vK coords (%): Males")

dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-RelAbun-for-ANCOMBC-vKCoord-level-Child-Behav-Males--v9d.tiff"), width = 14, height = 14, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

#### Flannery behaviour - cleaned sample summary stats
#-------------------------

phy <- readRDS("phy-phyloseq-object-Flannery-child-behav.RDS")
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20787 taxa and 40 samples ]
# sample_data() Sample Data:       [ 40 samples by 77 sample variables ]
# tax_table()   Taxonomy Table:    [ 20787 taxa by 4 taxonomic ranks ]

df.zones <- readRDS("df.zones--Flannery-child-behav.RDS")

vk.samp.noNAs.long.clean <- readRDS("vk.samp.noNAs.long.clean-Flannery-behav.RDS")

samps <- unique(df.zones$sample) # qty 40
samps <- unique(df.zones$sample[ which(!is.na(df.zones$behav_PC1)) ]) # qty 37

phy <- prune_samples( samples = samps, phy )
# prune taxa with zero reads
phy <- prune_taxa(taxa = taxa_sums(phy) > 0, x = phy)
phy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 20599 taxa and 37 samples ]
# sample_data() Sample Data:       [ 37 samples by 77 sample variables ]
# tax_table()   Taxonomy Table:    [ 20599 taxa by 4 taxonomic ranks ]


head(phy@otu_table)

fxns <- as.data.frame( phy@otu_table )

NonZeroFxns <- apply( fxns , 2,function(x) length(which(x > 0)) )
length(NonZeroFxns) # 37
NonZeroFxns


mean(NonZeroFxns) # 10281.86
sd(NonZeroFxns) # 2481.15
rm(NonZeroFxns)

vkcoords <- vk.samp.noNAs.long.clean
vkcoords <- filter(vkcoords, sample %in% samps)
vkcoords$vk_xy <- paste0(vkcoords$OC_x,"__",vkcoords$HC_y)
summary( vkcoords$abun )
# remove zero abundances
sel.zero <- which(vkcoords$abun == 0)
vkcoords <- vkcoords[-sel.zero, ]

# After translation to vK coordinates
# Total no unique functions?
length(unique(vkcoords$fxn_id)) # 10322
# Total no unique vK coords?
length(unique(vkcoords$vk_xy)) # 2459
identical(samps, unique(vkcoords$sample)) # TRUE
vks <- numeric(length = length(samps))
fxns_converted <- numeric(length = length(samps))
fxns_relabun_converted <- numeric(length = length(samps))
for (i in 1:length(samps)) {
  #i<-1
  sel <- which(vkcoords$sample == samps[i])
  vks[i] <- length(unique(vkcoords$vk_xy[sel]))
  fxns_converted[i] <- length(unique(vkcoords$fxn_id[sel]))  
  fxns_relabun_converted[i] <- sum(vkcoords$abun[sel])  
  print(paste0("completed sample: ", samps[i],"; # fxns: ",length(unique(vkcoords$fxn_id[sel]))," fxn %: ",sum(vkcoords$abun[sel]) ,"; # vk coords: ",length(unique(vkcoords$vk_xy[sel])) ))
}

mean(fxns_converted) # 5506.838
sd(fxns_converted) # 1224.467

mean(fxns_relabun_converted) # 54.85995
sd(fxns_relabun_converted) # 1.29186

mean(vks) # 1644.595
sd(vks) # 280.5938


#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Comparison b/w datasets - Human Gut - CPP-ASALR
# # # # # # # # #
# # # # # # # # #

#### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - PLOT SETUP - for All Gut datasets
#-------------------------

## Zeller-CRC

#saveRDS(object = df.zones, file = "df.zones--Zeller-CRC.RDS")

df.zones <- readRDS("df.zones--Zeller-CRC.RDS")

str(df.zones)
# 'data.frame':	113 obs. of  17 variables:
# $ sample      : chr  "FR_001" "FR_003" "FR_024" "FR_025" ...
# $ group       : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex         : chr  "F" "M" "F" "F" ...
# $ lipids      : num  0 0.0654 0.1837 0.0731 0.262 ...
# $ proteins    : num  4.03 3.8 3.69 3.13 3.38 ...
# $ amino_sugars: num  2.05 2.09 2.73 1.87 2.15 ...
# $ carbs       : num  20.9 31.5 23 28 29 ...
# $ cond_aromtx : num  0 0.0558 0.0866 0.1236 0.0439 ...
# $ lignin      : num  6.7 4.27 5.47 3.78 4.09 ...
# $ tannins     : num  17.8 14.6 17.8 18.9 16.3 ...
# $ methylated  : num  0.199 0.916 0.623 0.962 1.103 ...
# $ hydrated    : num  5.14 3.55 4.74 4.41 3.86 ...
# $ demethylated: num  10.62 8.28 9.33 7.68 7.05 ...
# $ condensed   : num  0.0994 0.475 0.4759 0.3197 0.4141 ...
# $ k2_b12      : num  0.149 0.182 0.22 0.24 0.464 ...
# $ group_mf    : Ord.factor w/ 4 levels "Cancer (F)"<"Normal (F)"<..: 2 4 2 2 4 4 4 4 2 2 ...
# $ all         : num  67.7 69.8 68.4 69.5 68.2 ...

length(unique(df.zones$sample)) # 113

sel.nan <- which(df.zones$amino_sugars == 0) # none

# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)



# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)



# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)


get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"     "log10_carbs_to_amino_sugars"       
# [4] "log10_cond_aromtx_to_amino_sugars"  "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"     "log10_demethylated_to_amino_sugars"
# [10] "log10_condensed_to_amino_sugars"    "log10_k2_b12_to_amino_sugars" 


df.zones.melt <- melt(df.zones[ ,c("sample","group","sex","group_mf", names(df.zones)[get_vars] )], id.vars = c("sample","group","sex","group_mf"))
head(df.zones.melt)
# sample  group sex   group_mf                     variable      value
# 1 FR_001 Normal   F Normal (F) log10_lipids_to_amino_sugars -2.4854122
# 2 FR_003 Normal   M Normal (M) log10_lipids_to_amino_sugars -1.5054605
# 3 FR_024 Normal   F Normal (F) log10_lipids_to_amino_sugars -1.1724227
# 4 FR_025 Normal   F Normal (F) log10_lipids_to_amino_sugars -1.4080383
# 5 FR_027 Normal   M Normal (M) log10_lipids_to_amino_sugars -0.9144802
# 6 FR_030 Normal   M Normal (M) log10_lipids_to_amino_sugars -1.3625290

unique( df.zones.melt$group )
# [1] Normal Cancer
# Levels: Cancer < Normal

unique( df.zones.melt$group_mf)
# [1] Normal (F) Normal (M) Cancer (F) Cancer (M)
# Levels: Cancer (F) < Normal (F) < Cancer (M) < Normal (M)

unique( df.zones.melt$sex ) #  "F" "M"
df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("F","M"), labels = c("Female","Male"), ordered = TRUE)

str(df.zones.melt)
# 'data.frame':	1243 obs. of  6 variables:
#   $ sample  : chr  "FR_001" "FR_003" "FR_024" "FR_025" ...
# $ group   : Ord.factor w/ 2 levels "Cancer"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 1 2 1 1 2 2 2 2 1 1 ...
# $ group_mf: Ord.factor w/ 4 levels "Cancer (F)"<"Normal (F)"<..: 2 4 2 2 4 4 4 4 2 2 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -2.485 -1.505 -1.172 -1.408 -0.914 ...

df.zones.melt$study <- "Cancer"

dat <- df.zones.melt

saveRDS(object = dat, file = "dat.ASALR--Zeller-CRC.RDS")


temp1 <- dat





## Jie-ACVD

#saveRDS(object = df.zones, file = "df.zones--Jie-ACVD.RDS")

df.zones <- readRDS("df.zones--Jie-ACVD.RDS")

str(df.zones)
# 'data.frame':	380 obs. of  17 variables:
#   $ sample      : chr  "ERR2017411" "ERR2017412" "ERR2017413" "ERR2017414" ...
# $ group       : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex         : chr  "male" "male" "male" "female" ...
# $ lipids      : num  0 0 0 0 0 ...
# $ proteins    : num  2.83 11.76 0 2.7 6.03 ...
# $ amino_sugars: num  3.04 0 0 2.7 4.31 ...
# $ carbs       : num  34.4 29.4 41.7 25.7 27.6 ...
# $ cond_aromtx : num  0.0591 0 0 0 0 ...
# $ lignin      : num  3.68 11.76 6.25 7.36 5.17 ...
# $ tannins     : num  12.76 5.88 10.42 10.89 10.34 ...
# $ methylated  : num  1.66 0 0 0 0 ...
# $ hydrated    : num  3.04 0 0 5.41 6.03 ...
# $ demethylated: num  6.55 5.88 2.08 9.46 8.62 ...
# $ condensed   : num  0.878 0 0 0 0 ...
# $ k2_b12      : num  0.35 0 0 0 0 ...
# $ group_mf    : Ord.factor w/ 4 levels "ACVD (F)"<"Normal (F)"<..: 4 4 4 2 2 4 4 2 2 2 ...
# $ all         : num  69.2 64.7 60.4 64.2 68.1 ...


length(unique(df.zones$sample)) # 380

sel.nan <- which(df.zones$amino_sugars == 0) # none
sort(df.zones$amino_sugars)[1:20] # 
# [1] 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.7407407 0.7575758 1.0869565 1.2195122 1.8485397 1.8551302 2.1377661 2.2301668 2.2522523
# [15] 2.2809144 2.3063830 2.3070378 2.3255814 2.3255814 2.3258879

# remove cases with ZERO df.zones$amino_sugars
df.zones <- df.zones[-sel.nan, ]


# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)


# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)


# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)


get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"     "log10_carbs_to_amino_sugars"       
# [4] "log10_cond_aromtx_to_amino_sugars"  "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"     "log10_demethylated_to_amino_sugars"
# [10] "log10_condensed_to_amino_sugars"    "log10_k2_b12_to_amino_sugars"   


df.zones.melt <- melt(df.zones[ ,c("sample","group","sex","group_mf", names(df.zones)[get_vars] )], id.vars = c("sample","group","sex","group_mf"))
head(df.zones.melt)
# sample  group    sex   group_mf                     variable     value
# 1 ERR2017411 Normal   male Normal (M) log10_lipids_to_amino_sugars -2.705094
# 2 ERR2017414 Normal female Normal (F) log10_lipids_to_amino_sugars -2.705094
# 3 ERR2017415 Normal female Normal (F) log10_lipids_to_amino_sugars -2.705094
# 4 ERR2017417 Normal   male Normal (M) log10_lipids_to_amino_sugars -2.705094
# 5 ERR2017418 Normal female Normal (F) log10_lipids_to_amino_sugars -2.705094
# 6 ERR2017419 Normal female Normal (F) log10_lipids_to_amino_sugars -1.454813

unique( df.zones.melt$group )
# [1] Normal ACVD  
# Levels: ACVD < Normal

unique( df.zones.melt$group_mf)
# [1] Normal (M) Normal (F) ACVD (M)   ACVD (F)  
# Levels: ACVD (F) < Normal (F) < ACVD (M) < Normal (M)

unique( df.zones.melt$sex ) #  "male" "female"
df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)

str(df.zones.melt)
# 'data.frame':	4125 obs. of  6 variables:
# $ sample  : chr  "ERR2017411" "ERR2017414" "ERR2017415" "ERR2017417" ...
# $ group   : Ord.factor w/ 2 levels "ACVD"<"Normal": 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 2 1 1 2 1 1 1 1 1 1 ...
# $ group_mf: Ord.factor w/ 4 levels "ACVD (F)"<"Normal (F)"<..: 4 2 2 4 2 2 2 2 2 2 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -2.71 -2.71 -2.71 -2.71 -2.71 ...


df.zones.melt$study <- "ACVD"

saveRDS(object = df.zones.melt, file = "dat.ASALR--Jie-ACVD.RDS")


dat <- rbind(dat, df.zones.melt)

temp2 <- dat






## Forslund T2D


#saveRDS(object = df.zones, file = "df.zones--Forslund-T2D.RDS")

df.zones <- readRDS("df.zones--Forslund-T2D.RDS")


str(df.zones)
# 'data.frame':	145 obs. of  15 variables:
# $ sample      : chr  "ERR260132" "ERR260133" "ERR260134" "ERR260135" ...
# $ group       : Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 1 1 3 ...
# $ lipids      : num  0.0344 0.0395 0.1669 0.1958 0.0807 ...
# $ proteins    : num  2.89 2.66 3.09 2.33 2.52 ...
# $ amino_sugars: num  3.78 3.3 3.85 3.54 3.34 ...
# $ carbs       : num  17.2 17.1 15.1 16.9 16 ...
# $ cond_aromtx : num  0.1204 0.1223 0.207 0.1042 0.0889 ...
# $ lignin      : num  4.93 4.82 5.86 5.31 4.86 ...
# $ tannins     : num  15.5 14.3 16.4 14.3 15.4 ...
# $ methylated  : num  0.459 0.49 0.467 0.478 0.458 ...
# $ hydrated    : num  6.94 5.79 6.52 5.43 6.04 ...
# $ demethylated: num  14.2 12.8 14.6 12.8 14.2 ...
# $ condensed   : num  0.201 0.586 0.18 0.395 0.348 ...
# $ k2_b12      : num  0.0498 0.1032 0.0689 0.1554 0.0905 ...
# $ all         : num  66.3 62.1 66.6 61.9 63.3 ...

df.zones$sex <- "Female"

unique(df.zones$group)
# [1] IGT         T2D met neg Normal      T2D met pos
# Levels: T2D met neg < T2D met pos < IGT < Normal

df.zones$group_mf <- paste0(df.zones$group," (F)")
unique(df.zones$group_mf)
df.zones$group_mf <- factor(df.zones$group_mf, 
                            levels = c("T2D met neg (F)", "T2D met pos (F)", "IGT (F)", "Normal (F)"),
                            labels = c("T2D met- (F)", "T2D met+ (F)", "IGT (F)", "Normal (F)"),
                            ordered = TRUE)

str(df.zones)
# 'data.frame':	145 obs. of  17 variables:
# $ sample      : chr  "ERR260132" "ERR260133" "ERR260134" "ERR260135" ...
# $ group       : Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 1 1 3 ...
# $ lipids      : num  0.0344 0.0395 0.1669 0.1958 0.0807 ...
# $ proteins    : num  2.89 2.66 3.09 2.33 2.52 ...
# $ amino_sugars: num  3.78 3.3 3.85 3.54 3.34 ...
# $ carbs       : num  17.2 17.1 15.1 16.9 16 ...
# $ cond_aromtx : num  0.1204 0.1223 0.207 0.1042 0.0889 ...
# $ lignin      : num  4.93 4.82 5.86 5.31 4.86 ...
# $ tannins     : num  15.5 14.3 16.4 14.3 15.4 ...
# $ methylated  : num  0.459 0.49 0.467 0.478 0.458 ...
# $ hydrated    : num  6.94 5.79 6.52 5.43 6.04 ...
# $ demethylated: num  14.2 12.8 14.6 12.8 14.2 ...
# $ condensed   : num  0.201 0.586 0.18 0.395 0.348 ...
# $ k2_b12      : num  0.0498 0.1032 0.0689 0.1554 0.0905 ...
# $ all         : num  66.3 62.1 66.6 61.9 63.3 ...
# $ sex         : chr  "Female" "Female" "Female" "Female" ...
# $ group_mf    : Ord.factor w/ 4 levels "T2D met- (F)"<..: 3 3 3 3 3 3 3 1 1 3 ...


length(unique(df.zones$sample)) # 145

sel.nan <- which(df.zones$amino_sugars == 0) # none

sort(df.zones$amino_sugars)[1:20]
# [1] 2.491837 2.761183 2.824586 2.840318 2.967005 3.019464 3.030784 3.055788 3.099551 3.113873 3.114606 3.116613 3.135579 3.245942 3.258200 3.264923
# [17] 3.267722 3.268699 3.270320 3.273735


# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)


# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)


# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)


get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"     "log10_carbs_to_amino_sugars"       
# [4] "log10_cond_aromtx_to_amino_sugars"  "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"     "log10_demethylated_to_amino_sugars"
# [10] "log10_condensed_to_amino_sugars"    "log10_k2_b12_to_amino_sugars" 


df.zones.melt <- melt(df.zones[ ,c("sample","group","sex","group_mf", names(df.zones)[get_vars] )], id.vars = c("sample","group","sex","group_mf"))
head(df.zones.melt)
# sample group    sex group_mf                     variable     value
# 1 ERR260132   IGT Female  IGT (F) log10_lipids_to_amino_sugars -2.041475
# 2 ERR260133   IGT Female  IGT (F) log10_lipids_to_amino_sugars -1.922605
# 3 ERR260134   IGT Female  IGT (F) log10_lipids_to_amino_sugars -1.362851
# 4 ERR260135   IGT Female  IGT (F) log10_lipids_to_amino_sugars -1.257302
# 5 ERR260136   IGT Female  IGT (F) log10_lipids_to_amino_sugars -1.616851
# 6 ERR260137   IGT Female  IGT (F) log10_lipids_to_amino_sugars -1.506709

unique( df.zones.melt$group )
# [1] IGT         T2D met neg Normal      T2D met pos
# Levels: T2D met neg < T2D met pos < IGT < Normal

unique( df.zones.melt$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# Levels: T2D met- (F) < T2D met+ (F) < IGT (F) < Normal (F)

unique( df.zones.melt$sex ) #  "Female"
#df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)

str(df.zones.melt)
# 'data.frame':	1595 obs. of  6 variables:
# $ sample  : chr  "ERR260132" "ERR260133" "ERR260134" "ERR260135" ...
# $ group   : Ord.factor w/ 4 levels "T2D met neg"<..: 3 3 3 3 3 3 3 1 1 3 ...
# $ sex     : chr  "Female" "Female" "Female" "Female" ...
# $ group_mf: Ord.factor w/ 4 levels "T2D met- (F)"<..: 3 3 3 3 3 3 3 1 1 3 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -2.04 -1.92 -1.36 -1.26 -1.62 ...


df.zones.melt$study <- "T2D"

saveRDS(object = df.zones.melt, file = "dat.ASALR--Forslund-T2D.RDS")


dat <- rbind(dat, df.zones.melt)

temp4 <- dat


unique(dat$study) # "Cancer" "ACVD"   "T2D"  




## Flannery behav

#saveRDS(object = df.zones, file = "df.zones--Flannery-child-behav.RDS")

df.zones <- readRDS("df.zones--Flannery-child-behav.RDS")

str(df.zones)
# 'data.frame':	40 obs. of  17 variables:
# $ sample       : chr  "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" ...
# $ depress_probs: num  50 50 51 60 56 60 50 51 51 56 ...
# $ behav_PC1    : num  -1.8745 -0.4343 -0.0323 0.5336 0.0319 ...
# $ sex          : chr  "female" "female" "male" "male" ...
# $ lipids       : num  0.151 0.131 0.221 0.192 0.155 ...
# $ proteins     : num  1.9 2 2.06 1.97 1.89 ...
# $ amino_sugars : num  3.24 3.2 2.96 3.18 3.25 ...
# $ carbs        : num  13.2 13.7 13.3 13.9 13.4 ...
# $ cond_aromtx  : num  0.113 0.115 0.161 0.129 0.114 ...
# $ lignin       : num  5.26 5.5 4.77 4.95 5 ...
# $ tannins      : num  12.9 12.6 11.8 12 12.2 ...
# $ methylated   : num  0.353 0.346 0.396 0.345 0.407 ...
# $ hydrated     : num  5.14 5.28 4.88 5.15 4.98 ...
# $ demethylated : num  13 12.7 12.5 12.8 12.8 ...
# $ condensed    : num  0.302 0.398 0.546 0.357 0.433 ...
# $ k2_b12       : num  0.154 0.166 0.102 0.148 0.157 ...
# $ all          : num  55.7 56.2 53.7 55.2 54.8 ...

sel.na <- which(is.na(df.zones$behav_PC1)) # qty 3
df.zones <- df.zones[-sel.na, ]

df.zones$group <- NA

sel <- which(df.zones$sex == "female")
hist(df.zones$behav_PC1[sel])
median_val <- median(df.zones$behav_PC1[sel], na.rm = TRUE)
subsel <- which(df.zones$behav_PC1[sel] <= median_val)
df.zones$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(df.zones$behav_PC1[sel] > median_val)
df.zones$group[sel[subsel]] <- "high_PC1_behav"

sel <- which(df.zones$sex == "male")
hist(df.zones$behav_PC1[sel])
median_val <- median(df.zones$behav_PC1[sel], na.rm = TRUE)
subsel <- which(df.zones$behav_PC1[sel] <= median_val)
df.zones$group[sel[subsel]] <- "low_PC1_behav"
subsel <- which(df.zones$behav_PC1[sel] > median_val)
df.zones$group[sel[subsel]] <- "high_PC1_behav"


length(unique(df.zones$sample)) # 37

sel.nan <- which(df.zones$amino_sugars == 0) # none

sort(df.zones$amino_sugars)[1:20]
# [1] 2.962748 2.994312 3.065136 3.078098 3.082083 3.085439 3.089291 3.099678 3.107669 3.119441 3.119914 3.119984 3.129313 3.139716 3.158951 3.160921
# [17] 3.164199 3.169065 3.182798 3.186722


# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)


# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)


# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)


get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"     "log10_carbs_to_amino_sugars"       
# [4] "log10_cond_aromtx_to_amino_sugars"  "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"     "log10_demethylated_to_amino_sugars"
# [10] "log10_condensed_to_amino_sugars"    "log10_k2_b12_to_amino_sugars"


names(df.zones)

df.zones$group_mf <- paste0(df.zones$group,"__",df.zones$sex)


df.zones.melt <- melt(df.zones[ ,c("sample","group","sex","group_mf", names(df.zones)[get_vars] )], id.vars = c("sample","group","sex","group_mf"))
head(df.zones.melt)
# sample          group    sex               group_mf                     variable     value
# 1 SRR8204551  low_PC1_behav female  low_PC1_behav__female log10_lipids_to_amino_sugars -1.332296
# 2 SRR8204552  low_PC1_behav female  low_PC1_behav__female log10_lipids_to_amino_sugars -1.388412
# 3 SRR8204553  low_PC1_behav   male    low_PC1_behav__male log10_lipids_to_amino_sugars -1.127085
# 4 SRR8204554 high_PC1_behav   male   high_PC1_behav__male log10_lipids_to_amino_sugars -1.220095
# 5 SRR8204555 high_PC1_behav female high_PC1_behav__female log10_lipids_to_amino_sugars -1.320576
# 6 SRR8204556 high_PC1_behav   male   high_PC1_behav__male log10_lipids_to_amino_sugars -1.207403

unique( df.zones.melt$group )
# "low_PC1_behav"  "high_PC1_behav"

unique( df.zones.melt$group_mf)
# "low_PC1_behav__female"  "low_PC1_behav__male"    "high_PC1_behav__male"   "high_PC1_behav__female"
df.zones.melt$group_mf <- factor(df.zones.melt$group_mf,
                                 levels = c("high_PC1_behav__female", "high_PC1_behav__male", "low_PC1_behav__female", "low_PC1_behav__male"),
                                 labels = c("High PC1 behav (F)", "High PC1 behav (M)", "Low PC1 behav (F)", "Low PC1 behav (M)"),
                                 ordered = TRUE)
unique( df.zones.melt$group_mf)
# [1] Low PC1 behav (F)  Low PC1 behav (M)  High PC1 behav (M) High PC1 behav (F)
# Levels: High PC1 behav (F) < High PC1 behav (M) < Low PC1 behav (F) < Low PC1 behav (M)

unique( df.zones.melt$sex ) #  "female" "male"
df.zones.melt$sex <- factor(df.zones.melt$sex, levels = c("female","male"), labels = c("Female","Male"), ordered = TRUE)

str(df.zones.melt)
# 'data.frame':	407 obs. of  6 variables:
#   $ sample  : chr  "SRR8204551" "SRR8204552" "SRR8204553" "SRR8204554" ...
# $ group   : chr  "low_PC1_behav" "low_PC1_behav" "low_PC1_behav" "high_PC1_behav" ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 1 1 2 2 1 2 2 2 1 2 ...
# $ group_mf: Ord.factor w/ 4 levels "High PC1 behav (F)"<..: 3 3 4 2 1 2 4 4 3 2 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -1.33 -1.39 -1.13 -1.22 -1.32 ...



df.zones.melt$study <- "Behaviour"

saveRDS(object = df.zones.melt, file = "dat.ASALR--Flannery-child-behav.RDS")


dat <- rbind(dat, df.zones.melt)

temp5 <- dat

# clean up
rm(temp, temp1, temp2, temp3, temp4, temp5)

#dat <- temp5
saveRDS(object = dat, file = "dat--all-gut-datasets-vk-zone-AminoSugars-ratios.RDS")

#dat <- readRDS("dat--all-gut-datasets-vk-zone-AminoSugars-ratios.RDS")


unique(dat$variable)
# [1] log10_lipids_to_amino_sugars       log10_proteins_to_amino_sugars     log10_carbs_to_amino_sugars        log10_cond_aromtx_to_amino_sugars 
# [5] log10_lignin_to_amino_sugars       log10_tannins_to_amino_sugars      log10_methylated_to_amino_sugars   log10_hydrated_to_amino_sugars    
# [9] log10_demethylated_to_amino_sugars log10_condensed_to_amino_sugars    log10_k2_b12_to_amino_sugars      
# 11 Levels: log10_lipids_to_amino_sugars log10_proteins_to_amino_sugars log10_carbs_to_amino_sugars ... log10_k2_b12_to_amino_sugars

dat$variable <- factor(dat$variable, 
                       levels = c("log10_lipids_to_amino_sugars", "log10_proteins_to_amino_sugars", "log10_carbs_to_amino_sugars",
                                  "log10_cond_aromtx_to_amino_sugars", "log10_lignin_to_amino_sugars", "log10_tannins_to_amino_sugars",
                                  "log10_methylated_to_amino_sugars", "log10_hydrated_to_amino_sugars", "log10_demethylated_to_amino_sugars",
                                  "log10_condensed_to_amino_sugars", "log10_k2_b12_to_amino_sugars" ),
                       # labels = c("Lipids\n:Amino sugars", "Proteins\n:Amino sugars", "Carbohydrates\n:Amino sugars",
                       #            "Condensed\nAromatics\n:Amino sugars", "Lignin\n:Amino sugars", "Tannins\n:Amino sugars",
                       #            "Other\nMethylated\n:Amino sugars", "Other\nHydrated\n:Amino sugars", "Other\nDemethylated\n:Amino sugars",
                       #            "Other\nCondensed\n:Amino sugars","Near\nVitK2-VitB12\n:Amino sugars"),
                       labels = c("Lipids", "Proteins", "Carbo-\nhydrates",
                                  "Condensed\nAromatics", "Lignin", "Tannins",
                                  "Other\nMethyl\n-ated", "Other\nHydrated", "Other\nDemethyl\n-ated",
                                  "Other\nCondensed","Near\nVitK2\n-VitB12"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                 "Proteins"               "Carbo-\nhydrates"       "Condensed\nAromatics"   "Lignin"                
# [6] "Tannins"                "Other\nMethyl\n-ated"   "Other\nHydrated"        "Other\nDemethyl\n-ated" "Other\nCondensed"      
# [11] "Near\nVitK2\n-VitB12"


cols <- c("Female" = "#7570b3", "Male" = "#d95f02")
shapes <- c("Female" = 2, "Male" = 4)

unique(dat$group_mf)
# [1] Normal (F)         Normal (M)         Cancer (F)         Cancer (M)         ACVD (M)           ACVD (F)           IGT (F)            T2D met- (F)      
# [9] T2D met+ (F)       Low PC1 behav (F)  Low PC1 behav (M)  High PC1 behav (M) High PC1 behav (F)
# 13 Levels: Cancer (F) < Normal (F) < Cancer (M) < Normal (M) < ACVD (F) < ACVD (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < ... < Low PC1 behav (M)

levels(dat$group_mf)
# [1] "Cancer (F)"         "Normal (F)"         "Cancer (M)"         "Normal (M)"         "ACVD (F)"           "ACVD (M)"           "T2D met- (F)"      
# [8] "T2D met+ (F)"       "IGT (F)"            "High PC1 behav (F)" "High PC1 behav (M)" "Low PC1 behav (F)"  "Low PC1 behav (M)"

dat$group_mf <- factor(dat$group_mf,
                       levels = c(
                         "ACVD (F)", "ACVD (M)" ,
                         "Cancer (F)", "Cancer (M)",
                         "T2D met- (F)", "T2D met+ (F)", "IGT (F)",
                         #"UC (F)", "Crohns (F)", 
                         #"UC (M)", "Crohns (M)" ,
                         "High PC1 behav (F)", "High PC1 behav (M)",
                         "Low PC1 behav (F)", "Low PC1 behav (M)",
                         "Normal (F)", "Normal (M)" ),
                       ordered = TRUE)

unique(dat$study) #  "Cancer"    "ACVD"      "T2D"       "Behaviour"

dat$study <- factor(dat$study, 
                    levels = c( "ACVD", "Cancer", "T2D" , "IBD", "Behaviour"),
                    labels = c( "ACVD", "Colorectal\nCancer", "Type 2\nDiabetes" , "IBD", "Problem\nBehaviors"),
                    ordered = TRUE)


str(dat)
# 'data.frame':	7370 obs. of  7 variables:
# $ sample  : chr  "FR_001" "FR_003" "FR_024" "FR_025" ...
# $ group   : Ord.factor w/ 8 levels "Cancer"<"Normal"<..: 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 1 2 1 1 2 2 2 2 1 1 ...
# $ group_mf: Ord.factor w/ 17 levels "ACVD (F)"<"ACVD (M)"<..: 16 17 16 16 17 17 17 17 16 16 ...
# $ variable: Ord.factor w/ 11 levels "Lipids"<"Proteins"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -2.485 -1.505 -1.172 -1.408 -0.914 ...
# $ study   : Ord.factor w/ 5 levels "ACVD"<"Colorectal\nCancer"<..: 2 2 2 2 2 2 2 2 2 2 ...




p <- ggplot(data = dat, aes(x = group_mf, y = value))+ # y = value   y = win95abun
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  scale_colour_manual(values = cols, name = "Sex")+
  #scale_shape_manual(values = shapes, name = "Sex")+
  facet_grid(rows = vars(variable), cols = vars(study), scales = "free", space = "free_x")+
  theme_classic() +
  labs(x = NULL, y = "Log10 ratio of CPP (cf. Amino sugars)") +
  theme(
    legend.position = "bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-gut-log10-CPP-ratio-cf-Amino-sugars-ROUGH--v9.tiff"), width = 20, height = 24, units = "cm", res=450, compression="lzw",type="cairo")



# remove outliers from plot??
# https://stackoverflow.com/questions/59140960/remove-outliers-and-reduce-ylim-appropriately-for-each-facet-in-ggplot2

# filtering function - turns outliers into NAs to be removed
filter_lims <- function(x){
  l <- boxplot.stats(x)$stats[1] # extreme of the lower whisker
  u <- boxplot.stats(x)$stats[5] # extreme of the upper whisker
  
  for (i in 1:length(x)){
    x[i] <- ifelse(x[i]>l & x[i]<u, x[i], NA)
  }
  return(x)
}

dat_lim <- dat
dat_lim$value2 <- NA

## across variables set any outliers to NA

unique_vars <- unique(dat_lim$variable) # 11 levels

for (i in 1:length(unique_vars)) {
  #i<-1
  this_var <- unique_vars[i]
  sel <- which(dat_lim$variable == this_var)
  dat_lim$value2[sel] <- filter_lims(dat_lim$value[sel])
  print(paste0("completed ",i))
}


p <- ggplot(data = dat_lim, aes(x = group_mf, y = value2))+ # y = value   y = win95abun
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  scale_colour_manual(values = cols, name = "Sex")+
  #scale_shape_manual(values = shapes, name = "Sex")+
  facet_grid(rows = vars(variable), cols = vars(study), scales = "free", space = "free_x")+
  theme_classic() +
  labs(x = NULL, y = "Log10 ratio of CPP (cf. Amino sugars)") +
  theme(
    legend.position = "bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-gut-log10-CPP-ratio-cf-Amino-sugars-OUTLIERS-REMOVED--v9.tiff"), width = 20, height = 24, units = "cm", res=450, compression="lzw",type="cairo")



## assign significance annotation
# use order for annotations per x-axis of figure: "All-gut-log10-CPP-ratio-cf-Amino-sugars-OUTLIERS-REMOVED--v9.tiff"
# allow up to 6 x-axis entries


str(dat_lim)
# 'data.frame':	7370 obs. of  8 variables:
# $ sample  : chr  "FR_001" "FR_003" "FR_024" "FR_025" ...
# $ group   : Ord.factor w/ 8 levels "Cancer"<"Normal"<..: 2 2 2 2 2 2 2 2 2 2 ...
# $ sex     : Ord.factor w/ 2 levels "Female"<"Male": 1 2 1 1 2 2 2 2 1 1 ...
# $ group_mf: Ord.factor w/ 17 levels "ACVD (F)"<"ACVD (M)"<..: 16 17 16 16 17 17 17 17 16 16 ...
# $ variable: Ord.factor w/ 11 levels "Lipids"<"Proteins"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -2.485 -1.505 -1.172 -1.408 -0.914 ...
# $ study   : Ord.factor w/ 5 levels "ACVD"<"Colorectal\nCancer"<..: 2 2 2 2 2 2 2 2 2 2 ...
# $ value2  : num  NA -1.505 -1.172 -1.408 -0.914 ...


saveRDS(dat_lim, "dat_lim--All-gut-log10-CPP-ratio-cf-Amino-sugars-OUTLIERS-REMOVED-v9e.RDS")


y_anotations <- list()
x_anotations <- list()
sig_anotations <- list()


levels(dat_lim$study)
# "ACVD"               "Colorectal\nCancer" "Type 2\nDiabetes"   "IBD"                "Problem\nBehaviors"

levels(dat_lim$variable)
# [1] "Lipids"                 "Proteins"               "Carbo-\nhydrates"       "Condensed\nAromatics"   "Lignin"                
# [6] "Tannins"                "Other\nMethyl\n-ated"   "Other\nHydrated"        "Other\nDemethyl\n-ated" "Other\nCondensed"      
# [11] "Near\nVitK2\n-VitB12"

#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "ACVD"
#-------------------------
this_study <- "ACVD"


## "Lipids"     
this_var <- "Lipids"

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.56745, num df = 52, denom df = 98, p-value = 0.02612
var.test(w,z) # F = 0.51649, num df = 156, denom df = 65, p-value = 0.0009415

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3910, p-value = 3.287e-07

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 7589, p-value = 2.198e-08

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Proteins"        
this_var <- "Proteins"

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.10945, num df = 52, denom df = 98, p-value = 1.65e-14
var.test(w,z) # F = 0.10998, num df = 156, denom df = 65, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # # Two Sample t-test
# # # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2546, p-value = 0.383

# # males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 5805, p-value = 0.07814

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Carbo-\nhydrates"  
this_var <- "Carbo-\nhydrates"  

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66


# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.038244, num df = 52, denom df = 98, p-value < 2.2e-16
var.test(w,z) # F = 0.045295, num df = 156, denom df = 65, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # # Two Sample t-test
# # # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1880, p-value = 0.002035

# # males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 2866, p-value = 7.101e-08

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Condensed\nAromatics"  
this_var <- "Condensed\nAromatics"  

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.13211, num df = 52, denom df = 98, p-value = 9.756e-13
var.test(w,z) # F = 0.26304, num df = 156, denom df = 65, p-value = 1.072e-11

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 4036, p-value = 2.374e-08

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 7989, p-value = 8.646e-11

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Lignin"      
this_var <- "Lignin" 

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.097556, num df = 52, denom df = 98, p-value = 1.275e-15
var.test(w,z) # F = 0.075457, num df = 156, denom df = 65, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3061, p-value = 0.04556

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 6152, p-value = 0.01367

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Tannins"       
this_var <- "Tannins"  

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.19406, num df = 52, denom df = 98, p-value = 2.446e-09
var.test(w,z) # F = 0.10573, num df = 156, denom df = 65, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2300, p-value = 0.1059

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 5416, p-value = 0.2969

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nMethyl\n-ated"  
this_var <- "Other\nMethyl\n-ated"  

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.054677, num df = 52, denom df = 98, p-value < 2.2e-16
var.test(w,z) # F = 0.051999, num df = 156, denom df = 65, p-value < 2.2e-16

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3093, p-value = 0.03489

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 4671, p-value = 0.1233

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nHydrated"   
this_var <- "Other\nHydrated"

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.13684, num df = 52, denom df = 98, p-value = 2.06e-12
var.test(w,z) # F = 0.11946, num df = 156, denom df = 65, p-value < 2.2e-16

#females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3084, p-value = 0.03766

#males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 7445, p-value = 1.326e-07

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nDemethyl\n-ated" 
this_var <- "Other\nDemethyl\n-ated"

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.085675, num df = 52, denom df = 98, p-value < 2.2e-16
var.test(w,z) # F = 0.073922, num df = 156, denom df = 65, p-value < 2.2e-16

#females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2825, p-value = 0.2185

#males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 7404, p-value = 2.17e-07

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nCondensed"     
this_var <- "Other\nCondensed"

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.37885, num df = 52, denom df = 98, p-value = 0.0002014
var.test(w,z) # F = 0.65069, num df = 156, denom df = 65, p-value = 0.03261

#females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 2890, p-value = 0.1519

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 3835, p-value = 0.001109

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

x=na.omit( filter(dat, group_mf == "ACVD (F)" & variable == this_var & study == this_study)$value )  # 53
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value )# 99
w=na.omit( filter(dat, group_mf == "ACVD (M)" & variable == this_var & study == this_study)$value )  # 157
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value )# 66

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.40101, num df = 52, denom df = 98, p-value = 0.0004425
var.test(w,z) # F = 0.43515, num df = 156, denom df = 65, p-value = 2.821e-05

#females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 3199, p-value = 0.0131

#males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 4778, p-value = 0.18

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )


annot_df.acvd.female <- data.frame(
  study = rep("ACVD", times = 11),
  sex = rep("Female", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]][1],
         y_anotations[["Proteins"]][1],
         y_anotations[["Carbo-\nhydrates"]][1],
         y_anotations[["Condensed\nAromatics"]][1],
         y_anotations[["Lignin"]][1],
         y_anotations[["Tannins"]][1],
         y_anotations[["Other\nMethyl\n-ated"]][1],
         y_anotations[["Other\nHydrated"]][1],
         y_anotations[["Other\nDemethyl\n-ated"]][1],
         y_anotations[["Other\nCondensed"]][1],
         y_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  x_label = rep(1.5, times = 11),
  x_seg_start = rep(1, times = 11),
  x_seg_end = rep(3, times = 11),
  label = c(sig_anotations[["Lipids"]][1],
            sig_anotations[["Proteins"]][1],
            sig_anotations[["Carbo-\nhydrates"]][1],
            sig_anotations[["Condensed\nAromatics"]][1],
            sig_anotations[["Lignin"]][1],
            sig_anotations[["Tannins"]][1],
            sig_anotations[["Other\nMethyl\n-ated"]][1],
            sig_anotations[["Other\nHydrated"]][1],
            sig_anotations[["Other\nDemethyl\n-ated"]][1],
            sig_anotations[["Other\nCondensed"]][1],
            sig_anotations[["Near\nVitK2\n-VitB12"]][1]
            ), stringsAsFactors = FALSE)
str(annot_df.acvd.female)

# ensure no conflict in factor levels
annot_df.acvd.female$variable <- factor(annot_df.acvd.female$variable, 
                                   levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                              "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.acvd.female$study <- factor(annot_df.acvd.female$study, 
                                     levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "IBD", "Problem\nBehaviours"),
                                     ordered = TRUE)
str( annot_df.acvd.female )
saveRDS(object = annot_df.acvd.female, file = "annot_df.acvd.female.RDS" )


annot_df.acvd.male <- data.frame(
  study = rep("ACVD", times = 11),
  sex = rep("Male", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]][2],
             y_anotations[["Proteins"]][2],
             y_anotations[["Carbo-\nhydrates"]][2],
             y_anotations[["Condensed\nAromatics"]][2],
             y_anotations[["Lignin"]][2],
             y_anotations[["Tannins"]][2],
             y_anotations[["Other\nMethyl\n-ated"]][2],
             y_anotations[["Other\nHydrated"]][2],
             y_anotations[["Other\nDemethyl\n-ated"]][2],
             y_anotations[["Other\nCondensed"]][2],
             y_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  x_label = rep(3.5, times = 11),
  x_seg_start = rep(2, times = 11),
  x_seg_end = rep(4, times = 11),
  label = c(sig_anotations[["Lipids"]][2],
            sig_anotations[["Proteins"]][2],
            sig_anotations[["Carbo-\nhydrates"]][2],
            sig_anotations[["Condensed\nAromatics"]][2],
            sig_anotations[["Lignin"]][2],
            sig_anotations[["Tannins"]][2],
            sig_anotations[["Other\nMethyl\n-ated"]][2],
            sig_anotations[["Other\nHydrated"]][2],
            sig_anotations[["Other\nDemethyl\n-ated"]][2],
            sig_anotations[["Other\nCondensed"]][2],
            sig_anotations[["Near\nVitK2\n-VitB12"]][2]
  ), stringsAsFactors = FALSE)
str(annot_df.acvd.male)

# ensure no conflict in factor levels
annot_df.acvd.male$variable <- factor(annot_df.acvd.male$variable, 
                                        levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                   "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.acvd.male$study <- factor(annot_df.acvd.male$study, 
                                     levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "IBD", "Problem\nBehaviours"),
                                     ordered = TRUE)
str( annot_df.acvd.male )
saveRDS(object = annot_df.acvd.male, file = "annot_df.acvd.male.RDS" )


round(annot_df.acvd.female$y_val, 3) # -0.483  1.041  1.854  0.432  0.843  1.420  0.542  0.459  1.163  0.525 -0.141
annot_df.acvd.female$y_val2 <- c(-0.75, 0.05, 1.1, -0.9, 0.35, 0.87,   -0.3, 0.45, 0.80, -0.025, -0.25)
#annot_df.acvd.female$y_val2 <- c(-0.75, 0.05, 1.1, -0.9, 0.35, 0.87,   -0.3, 0.45, 0.80, -0.025, -0.25)



round(annot_df.acvd.male$y_val, 3) # -0.580  0.312  1.815 -0.749  0.490  1.329 -0.087  0.508  1.094  0.025  0.505
#annot_df.acvd.male$y_val2 <-   c(-0.8,  0.04, 1.05, -1.0, 0.32, 0.84, -0.33, 0.44, 0.75, -0.05, -0.3)
annot_df.acvd.male$y_val2 <-   c(-0.95,  0.04, 1.0, -1.0, 0.32, 0.84, -0.33, 0.39, 0.75, -0.05, -0.3)


#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "Colorectal\nCancer" 
#-------------------------
this_study <- "Colorectal\nCancer"


## "Lipids"     
this_var <- "Lipids"

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 24
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 33
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 29
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 27

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.37662, num df = 28, denom df = 26, p-value = 0.01299
var.test(w,z) # F = 0.37662, num df = 28, denom df = 26, p-value = 0.01299

# females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # # Two Sample t-test
# # # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 372.5, p-value = 0.355

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 499, p-value = 0.03962 

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Proteins"        
this_var <- "Proteins"

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 24
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 33
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 29
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 27

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.1081, num df = 23, denom df = 32, p-value = 0.7756
var.test(w,z) # F = 0.52507, num df = 28, denom df = 26, p-value = 0.09803

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# # Two Sample t-test
# # data:  x and y
# t = -2.4265, df = 55, p-value = 0.009276
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.03702955
# sample estimates:
#   mean of x  mean of y 
# 0.03998147 0.15923688 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -4.09, df = 54, p-value = 7.23e-05
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.1092951
# sample estimates:
#   mean of x  mean of y 
# 0.01277536 0.19776463 

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Carbo-\nhydrates"  
this_var <- "Carbo-\nhydrates"  

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 24
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 33
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 29
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 27

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.76958, num df = 23, denom df = 32, p-value = 0.5193
var.test(w,z) # F = 0.31864, num df = 28, denom df = 26, p-value = 0.003831

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -1.5487, df = 55, p-value = 0.0636

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 225, p-value = 0.002916

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Condensed\nAromatics"  
this_var <- "Condensed\nAromatics"  

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 24
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 33
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 29
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 27

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.57924, num df = 23, denom df = 32, p-value = 0.177
var.test(w,z) # F = 0.69266, num df = 28, denom df = 26, p-value = 0.3427

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# # Two Sample t-test
# # data:  x and y
# t = 0.80569, df = 55, p-value = 0.2119

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.59421, df = 54, p-value = 0.2774

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Lignin"      
this_var <- "Lignin" 

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 24
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 33
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 29
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 27

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.4356, num df = 23, denom df = 32, p-value = 0.3398
var.test(w,z) # F = 0.23202, num df = 28, denom df = 26, p-value = 0.0002684

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# # Two Sample t-test
# # data:  x and y
# t = -1.9964, df = 55, p-value = 0.02543
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.01010372
# sample estimates:
#   mean of x mean of y 
# 0.2163699 0.2787521 

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 283, p-value = 0.03823

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Tannins"       
this_var <- "Tannins"  

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 24
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 33
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 29
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 27

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.3475, num df = 23, denom df = 32, p-value = 0.4292
var.test(w,z) # F = 0.26797, num df = 28, denom df = 26, p-value = 0.000946

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -2.8604, df = 55, p-value = 0.002985
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.05771286
# sample estimates:
#   mean of x mean of y 
# 0.7145771 0.8536086 

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 170, p-value = 9.233e-05

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nMethyl\n-ated"  
this_var <- "Other\nMethyl\n-ated"  

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.51102, num df = 23, denom df = 32, p-value = 0.09804
var.test(w,z) # F = 0.301, num df = 28, denom df = 26, p-value = 0.002455

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.33137, df = 55, p-value = 0.3708

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 260, p-value = 0.01547

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nHydrated"   
this_var <- "Other\nHydrated"

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) #
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) #
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) #
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) #

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 2.0511, num df = 23, denom df = 32, p-value = 0.06024
var.test(w,z) # F = 0.25582, num df = 28, denom df = 26, p-value = 0.0006364

#females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -1.1328, df = 55, p-value = 0.1311

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

#males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 210, p-value = 0.001267

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nDemethyl\n-ated" 
this_var <- "Other\nDemethyl\n-ated"

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.1926, num df = 23, denom df = 32, p-value = 0.6351
var.test(w,z) # F = 0.15443, num df = 28, denom df = 26, p-value = 5.038e-06

#females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -1.8615, df = 55, p-value = 0.03401
# alternative hypothesis: true difference in means is less than 0
# 95 percent confidence interval:
#   -Inf -0.006575761
# sample estimates:
#   mean of x mean of y 
# 0.5107703 0.5757178 

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

#males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# # data:  w and z
# W = 216, p-value = 0.001784

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nCondensed"     
this_var <- "Other\nCondensed"

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.98033, num df = 23, denom df = 32, p-value = 0.9763
var.test(w,z) # F = 0.84054, num df = 28, denom df = 26, p-value = 0.6514

#females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 0.8308, df = 55, p-value = 0.2048

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

#males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 1.0314, df = 54, p-value = 0.1535

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

x=na.omit( filter(dat, group_mf == "Cancer (F)" & variable == this_var & study == this_study)$value ) # 
y=na.omit( filter(dat, group_mf == "Normal (F)" & variable == this_var & study == this_study)$value ) # 
w=na.omit( filter(dat, group_mf == "Cancer (M)" & variable == this_var & study == this_study)$value ) # 
z=na.omit( filter(dat, group_mf == "Normal (M)" & variable == this_var & study == this_study)$value ) # 

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.33192, num df = 23, denom df = 32, p-value = 0.007696
var.test(w,z) # F = 0.29567, num df = 28, denom df = 26, p-value = 0.002128

#females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 383, p-value = 0.4206

#males
# tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.m
# # Two Sample t-test
# # data:  w and z

wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
wt.m
# Wilcoxon rank sum test with continuity correction
# data:  w and z
# W = 486, p-value = 0.0616

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )


annot_df.crc.female <- data.frame(
  study = rep("Colorectal\nCancer", times = 11),
  sex = rep("Female", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]][1],
             y_anotations[["Proteins"]][1],
             y_anotations[["Carbo-\nhydrates"]][1],
             y_anotations[["Condensed\nAromatics"]][1],
             y_anotations[["Lignin"]][1],
             y_anotations[["Tannins"]][1],
             y_anotations[["Other\nMethyl\n-ated"]][1],
             y_anotations[["Other\nHydrated"]][1],
             y_anotations[["Other\nDemethyl\n-ated"]][1],
             y_anotations[["Other\nCondensed"]][1],
             y_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  x_label = rep(1.5, times = 11),
  x_seg_start = rep(1, times = 11),
  x_seg_end = rep(3, times = 11),
  label = c(sig_anotations[["Lipids"]][1],
            sig_anotations[["Proteins"]][1],
            sig_anotations[["Carbo-\nhydrates"]][1],
            sig_anotations[["Condensed\nAromatics"]][1],
            sig_anotations[["Lignin"]][1],
            sig_anotations[["Tannins"]][1],
            sig_anotations[["Other\nMethyl\n-ated"]][1],
            sig_anotations[["Other\nHydrated"]][1],
            sig_anotations[["Other\nDemethyl\n-ated"]][1],
            sig_anotations[["Other\nCondensed"]][1],
            sig_anotations[["Near\nVitK2\n-VitB12"]][1]
  ), stringsAsFactors = FALSE)
str(annot_df.crc.female)

# ensure no conflict in factor levels
annot_df.crc.female$variable <- factor(annot_df.crc.female$variable, 
                                        levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                   "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.crc.female$study <- factor(annot_df.crc.female$study, 
                                     levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "IBD", "Problem\nBehaviours"),
                                     ordered = TRUE)
str( annot_df.crc.female )
saveRDS(object = annot_df.crc.female, file = "annot_df.crc.female.RDS" )


annot_df.crc.male <- data.frame(
  study = rep("Colorectal\nCancer", times = 11),
  sex = rep("Male", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]][2],
             y_anotations[["Proteins"]][2],
             y_anotations[["Carbo-\nhydrates"]][2],
             y_anotations[["Condensed\nAromatics"]][2],
             y_anotations[["Lignin"]][2],
             y_anotations[["Tannins"]][2],
             y_anotations[["Other\nMethyl\n-ated"]][2],
             y_anotations[["Other\nHydrated"]][2],
             y_anotations[["Other\nDemethyl\n-ated"]][2],
             y_anotations[["Other\nCondensed"]][2],
             y_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  x_label = rep(3.5, times = 11),
  x_seg_start = rep(2, times = 11),
  x_seg_end = rep(4, times = 11),
  label = c(sig_anotations[["Lipids"]][2],
            sig_anotations[["Proteins"]][2],
            sig_anotations[["Carbo-\nhydrates"]][2],
            sig_anotations[["Condensed\nAromatics"]][2],
            sig_anotations[["Lignin"]][2],
            sig_anotations[["Tannins"]][2],
            sig_anotations[["Other\nMethyl\n-ated"]][2],
            sig_anotations[["Other\nHydrated"]][2],
            sig_anotations[["Other\nDemethyl\n-ated"]][2],
            sig_anotations[["Other\nCondensed"]][2],
            sig_anotations[["Near\nVitK2\n-VitB12"]][2]
  ), stringsAsFactors = FALSE)
str(annot_df.crc.male)

# ensure no conflict in factor levels
annot_df.crc.male$variable <- factor(annot_df.crc.male$variable, 
                                      levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                 "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.crc.male$study <- factor(annot_df.crc.male$study, 
                                   levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "IBD", "Problem\nBehaviours"),
                                   ordered = TRUE)
str( annot_df.crc.male )
saveRDS(object = annot_df.crc.male, file = "annot_df.crc.male.RDS" )


## adjust heights of significance indicator & segment for display

round(annot_df.crc.female$y_val, 3)
#annot_df.crc.female$y_val2 <- c(-0.75, 0.13, 1.32, -0.9, 0.35, 1.1,  0.05, 0.45, 0.80, 0.05, -0.25)
annot_df.crc.female$y_val2 <- c(-0.75, 0.135, 1.32, -0.9, 0.35, 1.07,  0.05, 0.45, 0.80, 0.05, -0.25)


round(annot_df.crc.male$y_val, 3)
#annot_df.crc.male$y_val2 <-   c(-0.8,  0.11, 1.28, -1.0, 0.32, 1.0,  0.00, 0.44, 0.75, 0.025, -0.3)
annot_df.crc.male$y_val2 <-   c(-0.8,  0.11, 1.28, -1.0, 0.32, 1.0,  0.00, 0.44, 0.75, 0.025, -0.3)


#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "Type 2\nDiabetes"  
#-------------------------
this_study <- "Type 2\nDiabetes"  


y_anotations <- list()
sig_anotations <- list()


## "Lipids"     
this_var <- "Lipids"

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))


# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  1.8773 0.1362
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                  Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3   0.20 0.06653   1.148  0.332
# Residuals       141   8.17 0.05794 

unique(temp$group_mf_NODASH) # "IGT (F)"     "T2D metMINUS (F)" "Normal (F)"   "T2D met+ (F)"
temp$group_mf_NODASH <- factor(temp$group_mf_NODASH, 
                               levels = c(
                                 "T2D metMINUS (F)", "T2D met+ (F)", "IGT (F)", "Normal (F)" 
                               ))
# 

# Tukey test with compact letter display
tukey <- TukeyHSD(res.aov)
# https://www.rdocumentation.org/packages/multcompView/versions/0.1-9/topics/multcompLetters
tukey.cld <- multcompLetters4(res.aov, tukey) # multcompLetters4(): create a compact letters display using a aov object
tukey.cld
# $group_mf_NODASH
# $group_mf_NODASH$Letters
# Normal (F)     T2D met+ (F) T2D metMINUS (F)          IGT (F) 
# "a"              "a"              "a"              "a" 
# 
# $group_mf_NODASH$LetterMatrix
# a
# Normal (F)       TRUE
# T2D met+ (F)     TRUE
# T2D metMINUS (F) TRUE
# IGT (F)          TRUE

tukey.cld$group_mf_NODASH$Letters[ levels(temp$group_mf_NODASH) ]
# T2D metMINUS (F)     T2D met+ (F)          IGT (F)       Normal (F) 
# "a"              "a"              "a"              "a" 

summary(temp$value)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -2.0415 -1.7431 -1.6167 -1.5870 -1.4945 -0.7106 

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Proteins"        
this_var <- "Proteins"

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))


# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#       Df F value Pr(>F)
# group   3   1.368 0.2551
#       141 

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                   Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3 0.0202 0.006737   1.805  0.149
# Residuals       141 0.5264 0.003733

unique(temp$group_mf_NODASH) # "IGT (F)"     "T2D metMINUS (F)" "Normal (F)"   "T2D met+ (F)"
temp$group_mf_NODASH <- factor(temp$group_mf_NODASH, 
                               levels = c(
                                 "T2D metMINUS (F)", "T2D met+ (F)", "IGT (F)", "Normal (F)" 
                               ))
# 

# Tukey test with compact letter display
tukey <- TukeyHSD(res.aov)
# https://www.rdocumentation.org/packages/multcompView/versions/0.1-9/topics/multcompLetters
tukey.cld <- multcompLetters4(res.aov, tukey) # multcompLetters4(): create a compact letters display using a aov object

tukey.cld$group_mf_NODASH$Letters[ levels(temp$group_mf_NODASH) ]
# T2D metMINUS (F)     T2D met+ (F)          IGT (F)       Normal (F) 
# "a"              "a"              "a"              "a" 

summary(temp$value)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.20397 -0.14916 -0.11513 -0.10609 -0.07915  0.15170 

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Carbo-\nhydrates"  
this_var <- "Carbo-\nhydrates"  

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))


# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  1.3412 0.2635
# 141 

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                   Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3 0.0149 0.004974   1.452   0.23
# Residuals       141 0.4831 0.003426

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Condensed\nAromatics"  
this_var <- "Condensed\nAromatics"  

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)  
# group   3  2.3368 0.0763 .
#       141                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                  Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3 0.0363 0.01211    0.79  0.501
# Residuals       141 2.1603 0.01532 

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Lignin"      
this_var <- "Lignin" 

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.5133 0.6737
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                     Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3 0.00366 0.0012209   1.331  0.267
# Residuals       141 0.12934 0.0009173

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Tannins"       
this_var <- "Tannins"  

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3    1.52  0.212
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                    Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3 0.0022 0.0007348   0.436  0.728
# Residuals       141 0.2378 0.0016864

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other\nMethyl\n-ated"  
this_var <- "Other\nMethyl\n-ated"  

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.0502 0.9851
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                   Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3 0.0184 0.006130   1.409  0.243
# Residuals       141 0.6136 0.004352

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other\nHydrated"   
this_var <- "Other\nHydrated"

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.1537 0.9271
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                  Df  Sum Sq  Mean Sq F value Pr(>F)  
# group_mf_NODASH   3 0.01163 0.003878   2.308 0.0791 .
# Residuals       141 0.23687 0.001680                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other\nDemethyl\n-ated" 
this_var <- "Other\nDemethyl\n-ated"

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value  Pr(>F)  
# group   3  2.6526 0.05104 .
# 141                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                  Df  Sum Sq  Mean Sq F value Pr(>F)  
# group_mf_NODASH   3 0.00872 0.002906   2.395 0.0709 .
# Residuals       141 0.17111 0.001213                 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Other\nCondensed"     
this_var <- "Other\nCondensed"

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.2237 0.8798
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                  Df Sum Sq Mean Sq F value Pr(>F)
# group_mf_NODASH   3  0.068  0.0226   0.191  0.902
# Residuals       141 16.685  0.1183 

y_anot <- c(NA,NA,NA,NA)
sig_anot <- c(NA,NA,NA,NA)

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

temp <- filter(dat, variable == this_var & study == this_study)
unique(temp$group_mf)
# [1] IGT (F)      T2D met- (F) Normal (F)   T2D met+ (F)
# 17 Levels: ACVD (F) < ACVD (M) < Cancer (F) < Cancer (M) < T2D met- (F) < T2D met+ (F) < IGT (F) < UC (F) < Crohns (F) < ... < Normal (M)
temp$group_mf_NODASH <- gsub(pattern = "-", replacement = "MINUS", x = as.character(temp$group_mf))

# http://www.sthda.com/english/wiki/one-way-anova-test-in-r

# levene's test for common variance?
leveneTest(value ~ group_mf_NODASH, data = temp)
# Levene's Test for Homogeneity of Variance (center = median)
#        Df F value Pr(>F)
# group   3  0.1354 0.9388
#       141

# p-value is not less than the significance level of 0.05. This means that there is 
# no evidence to suggest that the variance across groups is statistically significantly different. 
# Therefore, we can assume the homogeneity of variances in the different treatment groups.

# Compute the analysis of variance
res.aov <- aov(value ~ group_mf_NODASH, data = temp)
# Summary of the analysis
summary(res.aov)
#                  Df Sum Sq Mean Sq F value  Pr(>F)   
# group_mf_NODASH   3  1.146  0.3819   4.342 0.00585 **
# Residuals       141 12.402  0.0880                   
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

unique(temp$group_mf_NODASH) # "IGT (F)"     "T2D metMINUS (F)" "Normal (F)"   "T2D met+ (F)"
temp$group_mf_NODASH <- factor(temp$group_mf_NODASH, 
                               levels = c(
                                 "T2D metMINUS (F)", "T2D met+ (F)", "IGT (F)", "Normal (F)" 
                               ))
# 

# Tukey test with compact letter display
tukey <- TukeyHSD(res.aov)
# https://www.rdocumentation.org/packages/multcompView/versions/0.1-9/topics/multcompLetters
tukey.cld <- multcompLetters4(res.aov, tukey) # multcompLetters4(): create a compact letters display using a aov object
tukey.cld
# $group_mf_NODASH
# T2D met+ (F)       Normal (F) T2D metMINUS (F)          IGT (F) 
# "a"              "a"             "ab"              "b"

tukey.cld$group_mf_NODASH$Letters[ levels(temp$group_mf_NODASH) ]
# T2D metMINUS (F)     T2D met+ (F)          IGT (F)       Normal (F) 
# "ab"              "a"              "b"              "a" 

summary(temp$value)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -2.4146 -1.6555 -1.5242 -1.5120 -1.3420 -0.5676 

y_anot <- c(-0.6,-0.6,-0.6,-0.6)
sig_anot <- c("ab", "a", "b", "a")

y_anotations[[this_var]] <- c(y_anot)
sig_anotations[[this_var]] <- c(sig_anot)


annot_df.t2d.female <- data.frame(
  study = rep("Type 2\nDiabetes", times = 11),
  sex = rep("Female", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y1 = c( y_anotations[["Lipids"]][1],
          y_anotations[["Proteins"]][1],
          y_anotations[["Carbo-\nhydrates"]][1],
          y_anotations[["Condensed\nAromatics"]][1],
          y_anotations[["Lignin"]][1],
          y_anotations[["Tannins"]][1],
          y_anotations[["Other\nMethyl\n-ated"]][1],
          y_anotations[["Other\nHydrated"]][1],
          y_anotations[["Other\nDemethyl\n-ated"]][1],
          y_anotations[["Other\nCondensed"]][1],
          y_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  y2 = c( y_anotations[["Lipids"]][2],
          y_anotations[["Proteins"]][2],
          y_anotations[["Carbo-\nhydrates"]][2],
          y_anotations[["Condensed\nAromatics"]][2],
          y_anotations[["Lignin"]][2],
          y_anotations[["Tannins"]][2],
          y_anotations[["Other\nMethyl\n-ated"]][2],
          y_anotations[["Other\nHydrated"]][2],
          y_anotations[["Other\nDemethyl\n-ated"]][2],
          y_anotations[["Other\nCondensed"]][2],
          y_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  y3 = c( y_anotations[["Lipids"]][3],
          y_anotations[["Proteins"]][3],
          y_anotations[["Carbo-\nhydrates"]][3],
          y_anotations[["Condensed\nAromatics"]][3],
          y_anotations[["Lignin"]][3],
          y_anotations[["Tannins"]][3],
          y_anotations[["Other\nMethyl\n-ated"]][3],
          y_anotations[["Other\nHydrated"]][3],
          y_anotations[["Other\nDemethyl\n-ated"]][3],
          y_anotations[["Other\nCondensed"]][3],
          y_anotations[["Near\nVitK2\n-VitB12"]][3]
  ),
  y4 = c( y_anotations[["Lipids"]][4],
          y_anotations[["Proteins"]][4],
          y_anotations[["Carbo-\nhydrates"]][4],
          y_anotations[["Condensed\nAromatics"]][4],
          y_anotations[["Lignin"]][4],
          y_anotations[["Tannins"]][4],
          y_anotations[["Other\nMethyl\n-ated"]][4],
          y_anotations[["Other\nHydrated"]][4],
          y_anotations[["Other\nDemethyl\n-ated"]][4],
          y_anotations[["Other\nCondensed"]][4],
          y_anotations[["Near\nVitK2\n-VitB12"]][4]
  ),
  label1 = c(sig_anotations[["Lipids"]][1],
             sig_anotations[["Proteins"]][1],
             sig_anotations[["Carbo-\nhydrates"]][1],
             sig_anotations[["Condensed\nAromatics"]][1],
             sig_anotations[["Lignin"]][1],
             sig_anotations[["Tannins"]][1],
             sig_anotations[["Other\nMethyl\n-ated"]][1],
             sig_anotations[["Other\nHydrated"]][1],
             sig_anotations[["Other\nDemethyl\n-ated"]][1],
             sig_anotations[["Other\nCondensed"]][1],
             sig_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  label2 = c(sig_anotations[["Lipids"]][2],
             sig_anotations[["Proteins"]][2],
             sig_anotations[["Carbo-\nhydrates"]][2],
             sig_anotations[["Condensed\nAromatics"]][2],
             sig_anotations[["Lignin"]][2],
             sig_anotations[["Tannins"]][2],
             sig_anotations[["Other\nMethyl\n-ated"]][2],
             sig_anotations[["Other\nHydrated"]][2],
             sig_anotations[["Other\nDemethyl\n-ated"]][2],
             sig_anotations[["Other\nCondensed"]][2],
             sig_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  label3 = c(sig_anotations[["Lipids"]][3],
             sig_anotations[["Proteins"]][3],
             sig_anotations[["Carbo-\nhydrates"]][3],
             sig_anotations[["Condensed\nAromatics"]][3],
             sig_anotations[["Lignin"]][3],
             sig_anotations[["Tannins"]][3],
             sig_anotations[["Other\nMethyl\n-ated"]][3],
             sig_anotations[["Other\nHydrated"]][3],
             sig_anotations[["Other\nDemethyl\n-ated"]][3],
             sig_anotations[["Other\nCondensed"]][3],
             sig_anotations[["Near\nVitK2\n-VitB12"]][3]
  ),
  label4 = c(sig_anotations[["Lipids"]][4],
             sig_anotations[["Proteins"]][4],
             sig_anotations[["Carbo-\nhydrates"]][4],
             sig_anotations[["Condensed\nAromatics"]][4],
             sig_anotations[["Lignin"]][4],
             sig_anotations[["Tannins"]][4],
             sig_anotations[["Other\nMethyl\n-ated"]][4],
             sig_anotations[["Other\nHydrated"]][4],
             sig_anotations[["Other\nDemethyl\n-ated"]][4],
             sig_anotations[["Other\nCondensed"]][4],
             sig_anotations[["Near\nVitK2\n-VitB12"]][4]
  ),
  stringsAsFactors = FALSE)
str(annot_df.t2d.female)

# ensure no conflict in factor levels
annot_df.t2d.female$variable <- factor(annot_df.t2d.female$variable, 
                                       levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                  "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.t2d.female$study <- factor(annot_df.t2d.female$study, 
                                    levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "IBD", "Problem\nBehaviours"),
                                    ordered = TRUE)
str( annot_df.t2d.female )
saveRDS(object = annot_df.t2d.female, file = "annot_df.t2d.female.RDS" )


## adjust heights of significance indicator & segment for display?


#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "Problem\nBehaviors" - High PC1 cf. Low PC1 of problem behavior
#-------------------------

this_study <- "Problem\nBehaviors"

# dat object is setup earlier ~ line 49059



## "Lipids"
this_var <- "Lipids"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9


# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.69965, num df = 9, denom df = 9, p-value = 0.6032
var.test(w,z) # F = 1.009, num df = 7, denom df = 8, p-value = 0.9773

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 0.1513, df = 18, p-value = 0.4407

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 1.3589, df = 15, p-value = 0.09713

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Proteins"
this_var <- "Proteins"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.1278, num df = 9, denom df = 9, p-value = 0.8608
var.test(w,z) # F = 0.27178, num df = 7, denom df = 8, p-value = 0.1031

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# # Two Sample t-test
# # data:  x and y
# t = 0.048783, df = 18, p-value = 0.4808

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -0.39845, df = 15, p-value = 0.348

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Carbo-\nhydrates"
this_var <- "Carbo-\nhydrates"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.0861, num df = 9, denom df = 9, p-value = 0.9041
var.test(w,z) # F = 0.49282, num df = 7, denom df = 8, p-value = 0.3668

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.73907, df = 18, p-value = 0.2347

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -1.8496, df = 15, p-value = 0.04209

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Condensed\nAromatics"
this_var <- "Condensed\nAromatics"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.83019, num df = 9, denom df = 9, p-value = 0.7861
var.test(w,z) # F = 0.30031, num df = 7, denom df = 8, p-value = 0.1306

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# # Two Sample t-test
# # data:  x and y
# t = -0.48313, df = 18, p-value = 0.3174

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.3084, df = 15, p-value = 0.381

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )



## "Lignin"
this_var <- "Lignin"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.48779, num df = 9, denom df = 9, p-value = 0.2998
var.test(w,z) # F = 0.74836, num df = 7, denom df = 8, p-value = 0.7153

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# # Two Sample t-test
# # data:  x and y
# t = -0.64084, df = 18, p-value = 0.2649

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.35417, df = 15, p-value = 0.3641

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Tannins"
this_var <- "Tannins"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.27592, num df = 9, denom df = 9, p-value = 0.06862
var.test(w,z) # F = 0.84273, num df = 7, denom df = 8, p-value = 0.8344

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.38014, df = 18, p-value = 0.3541

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.43559, df = 15, p-value = 0.3347

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nMethyl\n-ated"
this_var <- "Other\nMethyl\n-ated"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.2045, num df = 9, denom df = 9, p-value = 0.7862
var.test(w,z) # F = 0.40528, num df = 7, denom df = 8, p-value = 0.2512

# females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.8125, df = 18, p-value = 0.04331

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

# males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.2132, df = 15, p-value = 0.417

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nHydrated"
this_var <- "Other\nHydrated"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.59662, num df = 9, denom df = 9, p-value = 0.4535
var.test(w,z) # F = 0.54645, num df = 7, denom df = 8, p-value = 0.4405

#females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.68223, df = 18, p-value = 0.2519

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

#males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 0.36756, df = 15, p-value = 0.3592

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nDemethyl\n-ated"
this_var <- "Other\nDemethyl\n-ated"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.28737, num df = 9, denom df = 9, p-value = 0.07729
var.test(w,z) # F = 0.96962, num df = 7, denom df = 8, p-value = 0.9807

#females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = -0.36602, df = 18, p-value = 0.3593

# wt.f <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

#males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = 1.1824, df = 15, p-value = 0.1277

# wt.m <- wilcox.test(w, z, alternative = "less") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



## "Other\nCondensed"
this_var <- "Other\nCondensed"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.10312, num df = 9, denom df = 9, p-value = 0.002343
var.test(w,z) # F = 0.36479, num df = 7, denom df = 8, p-value = 0.2016

#females
# tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt.f
# # Two Sample t-test
# # data:  x and y

wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt.f
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 41, p-value = 0.2644

#males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -1.2479, df = 15, p-value = 0.1156

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )
sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, tt.m$p.value) )



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

x=na.omit( filter(dat, group_mf == "High PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
y=na.omit( filter(dat, group_mf == "Low PC1 behav (F)" & variable == this_var & study == this_study)$value ) # 10
w=na.omit( filter(dat, group_mf == "High PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 8
z=na.omit( filter(dat, group_mf == "Low PC1 behav (M)" & variable == this_var & study == this_study)$value ) # 9

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 3.0461, num df = 9, denom df = 9, p-value = 0.1125
var.test(w,z) # F = 0.55679, num df = 7, denom df = 8, p-value = 0.4549

#females
tt.f <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.f
# Two Sample t-test
# data:  x and y
# t = 1.3605, df = 18, p-value = 0.09523

# wt.f <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
# wt.f
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

#males
tt.m <- t.test(w,z,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt.m
# Two Sample t-test
# data:  w and z
# t = -0.28504, df = 15, p-value = 0.3898

# wt.m <- wilcox.test(w, z, alternative = "greater") # "less" "two.sided"  "greater"
# wt.m
# # Wilcoxon rank sum test with continuity correction
# # data:  w and z

y_anot.f <- 0.08*(max(x,y) - min(x,y)) + max(x,y)
y_anot.m <- 0.01*(max(w,z) - min(w,z)) + max(w,z)

x_anotations[[this_var]] <- c(1.5, 2.5)
y_anotations[[this_var]] <- c(y_anot.f, y_anot.m )
sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, tt.m$p.value) )
#sig_anotations[[this_var]] <- sig_symb( c(wt.f$p.value, wt.m$p.value) )
##sig_anotations[[this_var]] <- sig_symb( c(tt.f$p.value, wt.m$p.value) )



annot_df.behav.female <- data.frame(
  study = rep("Problem\nBehaviors", times = 11),
  sex = rep("Female", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]][1],
             y_anotations[["Proteins"]][1],
             y_anotations[["Carbo-\nhydrates"]][1],
             y_anotations[["Condensed\nAromatics"]][1],
             y_anotations[["Lignin"]][1],
             y_anotations[["Tannins"]][1],
             y_anotations[["Other\nMethyl\n-ated"]][1],
             y_anotations[["Other\nHydrated"]][1],
             y_anotations[["Other\nDemethyl\n-ated"]][1],
             y_anotations[["Other\nCondensed"]][1],
             y_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  x_label = rep(1.5, times = 11),
  x_seg_start = rep(1, times = 11),
  x_seg_end = rep(3, times = 11),
  label = c(sig_anotations[["Lipids"]][1],
            sig_anotations[["Proteins"]][1],
            sig_anotations[["Carbo-\nhydrates"]][1],
            sig_anotations[["Condensed\nAromatics"]][1],
            sig_anotations[["Lignin"]][1],
            sig_anotations[["Tannins"]][1],
            sig_anotations[["Other\nMethyl\n-ated"]][1],
            sig_anotations[["Other\nHydrated"]][1],
            sig_anotations[["Other\nDemethyl\n-ated"]][1],
            sig_anotations[["Other\nCondensed"]][1],
            sig_anotations[["Near\nVitK2\n-VitB12"]][1]
  ), stringsAsFactors = FALSE)
str(annot_df.behav.female)

# ensure no conflict in factor levels
annot_df.behav.female$variable <- factor(annot_df.behav.female$variable,
                                       levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                  "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
# #fix
# annot_df.behav.female$study <- factor(annot_df.behav.female$study,
#                                       levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "Problem\nBehaviours"),
#                                       labels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "Problem\nBehaviors"),
#                                       ordered = TRUE)

annot_df.behav.female$study <- factor(annot_df.behav.female$study,
                                    levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "Problem\nBehaviors"),
                                    ordered = TRUE)
str( annot_df.behav.female )
saveRDS(object = annot_df.behav.female, file = "annot_df.behav.female.RDS" )


annot_df.behav.male <- data.frame(
  study = rep("Problem\nBehaviors", times = 11),
  sex = rep("Male", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]][2],
             y_anotations[["Proteins"]][2],
             y_anotations[["Carbo-\nhydrates"]][2],
             y_anotations[["Condensed\nAromatics"]][2],
             y_anotations[["Lignin"]][2],
             y_anotations[["Tannins"]][2],
             y_anotations[["Other\nMethyl\n-ated"]][2],
             y_anotations[["Other\nHydrated"]][2],
             y_anotations[["Other\nDemethyl\n-ated"]][2],
             y_anotations[["Other\nCondensed"]][2],
             y_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  x_label = rep(3.5, times = 11),
  x_seg_start = rep(2, times = 11),
  x_seg_end = rep(4, times = 11),
  label = c(sig_anotations[["Lipids"]][2],
            sig_anotations[["Proteins"]][2],
            sig_anotations[["Carbo-\nhydrates"]][2],
            sig_anotations[["Condensed\nAromatics"]][2],
            sig_anotations[["Lignin"]][2],
            sig_anotations[["Tannins"]][2],
            sig_anotations[["Other\nMethyl\n-ated"]][2],
            sig_anotations[["Other\nHydrated"]][2],
            sig_anotations[["Other\nDemethyl\n-ated"]][2],
            sig_anotations[["Other\nCondensed"]][2],
            sig_anotations[["Near\nVitK2\n-VitB12"]][2]
  ), stringsAsFactors = FALSE)
str(annot_df.behav.male)

# ensure no conflict in factor levels
annot_df.behav.male$variable <- factor(annot_df.behav.male$variable,
                                     levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
# #fix
# annot_df.behav.male$study <- factor(annot_df.behav.male$study,
#                                       levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "Problem\nBehaviours"),
#                                       labels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "Problem\nBehaviors"),
#                                       ordered = TRUE)

annot_df.behav.male$study <- factor(annot_df.behav.male$study,
                                  levels = c("ACVD", "Colorectal\nCancer", "Type 2\nDiabetes", "Problem\nBehaviors"),
                                  ordered = TRUE)
str( annot_df.behav.male )
saveRDS(object = annot_df.behav.male, file = "annot_df.behav.male.RDS" )


## adjust heights of significance indicator & segment for display

round(annot_df.behav.female$y_val, 3) # -1.061 -0.194  0.701 -1.098  0.255  0.623 -0.706  0.235  0.654 -0.644 -1.118
annot_df.behav.female$y_val2 <-          c(-0.75, 0.03, 1.12, -1.1, 0.30, 0.9,  -0.3, 0.4, 0.78, 0.0, -0.5)
round(annot_df.behav.male$y_val, 3) # -1.103 -0.156  0.725 -1.259  0.257  0.633 -0.872  0.244  0.641 -0.721 -1.144
annot_df.behav.male$y_val2 <-            c(-0.8,  0.01, 1.08, -1.2, 0.27, 0.8,  -0.4, 0.36, 0.74, -0.1, -0.7)


#-------------------------

#### Gut - log10 ratio CPP-ASALR (normalized wrt Amino sugars) - FINAL PLOT
#-------------------------

#saveRDS(object = dat, file = "dat--all-gut-datasets-vk-zone-AminoSugars-ratios.RDS")
#saveRDS(dat_lim, "dat_lim--All-gut-log10-CPP-ratio-cf-Amino-sugars-OUTLIERS-REMOVED-v9e.RDS")

dat_lim <- readRDS("dat_lim--All-gut-log10-CPP-ratio-cf-Amino-sugars-OUTLIERS-REMOVED-v9e.RDS")

#dat_lim.noIBD <- filter(dat_lim, study != "IBD")

cols <- c("Female" = "#7570b3", "Male" = "#d95f02")

#saveRDS(object = annot_df.acvd.female, file = "annot_df.acvd.female.RDS" )
#saveRDS(object = annot_df.acvd.male, file = "annot_df.acvd.male.RDS" )

#saveRDS(object = annot_df.crc.female, file = "annot_df.crc.female.RDS" )
#saveRDS(object = annot_df.crc.male, file = "annot_df.crc.male.RDS" )


p <- ggplot(data = dat_lim, aes(x = group_mf, y = value2))+
  geom_boxplot(aes(color = sex), outlier.shape=NA)+
  scale_colour_manual(values = cols, name = "Sex")+
  facet_grid(rows = vars(variable), cols = vars(study), scales = "free", space = "free_x")+
  theme_classic() +
  #labs(x = NULL, y = "Log10 ratio of CPP (cf. Amino sugars)") +
  labs(x = NULL, y = "Log10 ( CPP[class] / CPP[amino sugars] )") +
  
  # names(annot_df.acvd.female) # "study"  "sex"   "variable"  "y_val"   "x_label"   "x_seg_start" "x_seg_end"   "label"   
  
  # females-ACVD
  geom_text(data= filter(annot_df.acvd.female, label != "ns"), mapping = aes(x = x_label, y = y_val2, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=filter(annot_df.acvd.female, label != "ns"), mapping = aes(x = x_seg_start, y = y_val2, xend = x_seg_end, yend = y_val2), color = "#7570b3")+
  # males-ACVD
  geom_text(data=filter(annot_df.acvd.male, label != "ns"), mapping = aes(x = x_label, y = y_val2, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=filter(annot_df.acvd.male, label != "ns"), mapping = aes(x = x_seg_start, y = y_val2, xend = x_seg_end, yend = y_val2), color = "#d95f02")+
  
  # females-Cancer
  geom_text(data= filter(annot_df.crc.female, label != "ns"), mapping = aes(x = x_label, y = y_val2, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=filter(annot_df.crc.female, label != "ns"), mapping = aes(x = x_seg_start, y = y_val2, xend = x_seg_end, yend = y_val2), color = "#7570b3")+
  # males-Cancer
  geom_text(data=filter(annot_df.crc.male, label != "ns"), mapping = aes(x = x_label, y = y_val2, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=filter(annot_df.crc.male, label != "ns"), mapping = aes(x = x_seg_start, y = y_val2, xend = x_seg_end, yend = y_val2), color = "#d95f02")+
  
  # females-T2d
  geom_text(data= filter(annot_df.t2d.female, label1 != "ns"), mapping = aes(x = 1.1, y = y1, label = label1), size = 3, color = "#7570b3", vjust = 0.4)+
  geom_text(data= filter(annot_df.t2d.female, label2 != "ns"), mapping = aes(x = 2.1, y = y2, label = label2), size = 3, color = "#7570b3", vjust = 0.4)+
  geom_text(data= filter(annot_df.t2d.female, label3 != "ns"), mapping = aes(x = 3.1, y = y3, label = label3), size = 3, color = "#7570b3", vjust = 0.4)+
  geom_text(data= filter(annot_df.t2d.female, label4 != "ns"), mapping = aes(x = 4.1, y = y4, label = label4), size = 3, color = "#7570b3", vjust = 0.4)+
  
  # females-Behav
  geom_text(data= filter(annot_df.behav.female, label != "ns"), mapping = aes(x = x_label, y = y_val2, label = label), size = 5, color = "#7570b3", vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=filter(annot_df.behav.female, label != "ns"), mapping = aes(x = x_seg_start, y = y_val2, xend = x_seg_end, yend = y_val2), color = "#7570b3")+
  # males-Behav
  geom_text(data=filter(annot_df.behav.male, label != "ns"), mapping = aes(x = x_label, y = y_val2, label = label), size = 5, color = "#d95f02", vjust = 0.4)+
  geom_segment(data=filter(annot_df.behav.male, label != "ns"), mapping = aes(x = x_seg_start, y = y_val2, xend = x_seg_end, yend = y_val2), color = "#d95f02")+
  
  scale_y_continuous(expand = expansion(mult = c(0.02, 0.14)))+
  
  theme(
    legend.position = "bottom",
    legend.margin=margin(0,0,0,0),
    legend.box.margin=margin(-10,0,-5,0), 
    legend.title = element_text(size = rel(0.9)),
    strip.background = element_rect(fill="white", linetype = "blank"),
    axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )
p

dev.print(tiff, file = paste0(workdir,"/plots/","All-gut-log10-CPP-ratio-cf-Amino-sugars-OUTLIERS-REMOVED-FROM-DISPLAY-WITH-SIG-INCL-OUTLIERS--v9e.tiff"), width = 17, height = 24, units = "cm", res=450, compression="lzw",type="cairo")

#-------------------------





# # # # # # # # #
# # # # # # # # #
# # # Comparison b/w datasets - Soils - CPP-ASALR
# # # # # # # # #
# # # # # # # # #

#### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - PLOT SETUP - for All Soil datasets
#-------------------------

## PCaN

#saveRDS(object = df.zones, file = "df.zones--PCaN-pilot.RDS")

df.zones <- readRDS("df.zones--PCaN-pilot.RDS")

str(df.zones)
# 'data.frame':	19 obs. of  15 variables:
# $ sample      : chr  "ChrMrs" "ChrRic" "DunIsp" "DunSiH" ...
# $ group       : num  14 41 11 31 14 40 23 NA 13 31 ...
# $ lipids      : num  0.237 0.309 0.295 0.32 0.302 ...
# $ proteins    : num  2.67 2.61 2.83 2.56 2.62 ...
# $ amino_sugars: num  4.66 4.55 4.89 4.78 4.71 ...
# $ carbs       : num  9.69 9.22 9.71 9.87 9.47 ...
# $ cond_aromtx : num  0.944 1.028 0.859 1.062 1.037 ...
# $ lignin      : num  8.33 8.27 8.28 8.13 8.32 ...
# $ tannins     : num  15.9 16.1 15.7 15.3 16.1 ...
# $ methylated  : num  0.479 0.486 0.454 0.473 0.462 ...
# $ hydrated    : num  4.71 4.55 4.75 4.65 4.7 ...
# $ demethylated: num  13 13.2 12.9 12.8 13.3 ...
# $ condensed   : num  0.476 0.721 0.41 0.533 0.499 ...
# $ k2_b12      : num  0.445 0.497 0.485 0.465 0.503 ...
# $ all         : num  61.6 61.6 61.5 60.9 62 ...


length(unique(df.zones$sample)) # 19

sel.nan <- which(df.zones$amino_sugars == 0) # none


# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)


# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)


# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)


get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"     "log10_carbs_to_amino_sugars"       
# [4] "log10_cond_aromtx_to_amino_sugars"  "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"     "log10_demethylated_to_amino_sugars"
# [10] "log10_condensed_to_amino_sugars"    "log10_k2_b12_to_amino_sugars" 


df.zones.melt <- melt(df.zones[ ,c("sample","group", names(df.zones)[get_vars] )], id.vars = c("sample","group"))
head(df.zones.melt)
# sample group                     variable     value
# 1 ChrMrs    14 log10_lipids_to_amino_sugars -1.293208
# 2 ChrRic    41 log10_lipids_to_amino_sugars -1.167185
# 3 DunIsp    11 log10_lipids_to_amino_sugars -1.219466
# 4 DunSiH    31 log10_lipids_to_amino_sugars -1.173679
# 5 HamAva    14 log10_lipids_to_amino_sugars -1.193000
# 6 HamMin    40 log10_lipids_to_amino_sugars -1.069219

unique( df.zones.melt$group )
# 14 41 11 31 40 23 NA 13 10 48 28 33

sort( unique( df.zones.melt$group ) )
# 10 11 13 14 23 28 31 33 40 41 48

sel.na <- which(is.na(df.zones.melt$group))
df.zones.melt$group[sel.na] <- "Rem"

df.zones.melt$group <- factor(df.zones.melt$group, levels = c(as.character(1:50),"Rem"), ordered = TRUE)


str(df.zones.melt)
# 'data.frame':	209 obs. of  4 variables:
# $ sample  : chr  "ChrMrs" "ChrRic" "DunIsp" "DunSiH" ...
# $ group   : Ord.factor w/ 51 levels "1"<"2"<"3"<"4"<..: 14 41 11 31 14 40 23 51 13 31 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -1.29 -1.17 -1.22 -1.17 -1.19 ...

df.zones.melt$study <- "PCaN\nRestoration"

dat <- df.zones.melt

temp1 <- dat




## Sun-Badgley post-mining restoration

#saveRDS(object = df.zones, file = "df.zones--SunBadgley.RDS")

df.zones <- readRDS("df.zones--SunBadgley.RDS")

str(df.zones)
# 'data.frame':	15 obs. of  15 variables:
# $ sample      : chr  "20C" "30B" "30A" "UMA" ...
# $ group       : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ lipids      : num  0.202 0.292 0.264 0.276 0.217 ...
# $ proteins    : num  2.67 2.52 2.56 2.62 2.6 ...
# $ amino_sugars: num  4.51 4.41 4.42 4.51 4.52 ...
# $ carbs       : num  9.49 9.26 9.33 9.16 9.37 ...
# $ cond_aromtx : num  1.08 1.33 1.19 1.25 1.11 ...
# $ lignin      : num  8.25 8.01 8.12 8.11 8.22 ...
# $ tannins     : num  16.2 15.9 16.2 16 16.3 ...
# $ methylated  : num  0.427 0.407 0.418 0.405 0.443 ...
# $ hydrated    : num  4.67 4.81 4.68 4.71 4.61 ...
# $ demethylated: num  13.3 13.3 13.4 13.3 13.2 ...
# $ condensed   : num  0.891 0.827 0.854 0.778 0.949 ...
# $ k2_b12      : num  0.45 0.501 0.488 0.513 0.467 ...
# $ all         : num  62.1 61.6 61.8 61.6 62.1 ...


length(unique(df.zones$sample)) # 15

sel.nan <- which(df.zones$amino_sugars == 0) # none

sort(df.zones$amino_sugars)[1:15] # 
# [1] 4.411393 4.420543 4.441792 4.466144 4.506585 4.514712 4.523768 4.531974 4.546460 4.575702 4.604747 4.622936 4.640985
# [14] 4.654906 4.657028

# # remove cases with ZERO df.zones$amino_sugars
# df.zones <- df.zones[-sel.nan, ]


# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)


# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)


# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)



get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"     "log10_carbs_to_amino_sugars"       
# [4] "log10_cond_aromtx_to_amino_sugars"  "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"     "log10_demethylated_to_amino_sugars"
# [10] "log10_condensed_to_amino_sugars"    "log10_k2_b12_to_amino_sugars"   


df.zones.melt <- melt(df.zones[ ,c("sample","group", names(df.zones)[get_vars] )], id.vars = c("sample","group"))
head(df.zones.melt)
# sample group                     variable     value
# 1    20C    22 log10_lipids_to_amino_sugars -1.347743
# 2    30B    31 log10_lipids_to_amino_sugars -1.179442
# 3    30A    31 log10_lipids_to_amino_sugars -1.224114
# 4    UMA    UM log10_lipids_to_amino_sugars -1.214096
# 5    10B    12 log10_lipids_to_amino_sugars -1.318231
# 6     5A     6 log10_lipids_to_amino_sugars -1.381234

unique( df.zones.melt$group )
# [1] 22 31 UM 12 6 
# Levels: 6 < 12 < 22 < 31 < UM

str(df.zones.melt)
# 'data.frame':	165 obs. of  4 variables:
# $ sample  : chr  "20C" "30B" "30A" "UMA" ...
# $ group   : Ord.factor w/ 5 levels "6"<"12"<"22"<..: 3 4 4 5 2 1 2 3 3 1 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -1.35 -1.18 -1.22 -1.21 -1.32 ...


df.zones.melt$study <- "Post-mining\nRestoration"

dat <- rbind(dat, df.zones.melt)

temp2 <- dat




## AMI Disturbed vs Natural soils

#saveRDS(object = df.zones, file = "df.zones--AMI.RDS")

df.zones <- readRDS("df.zones--AMI.RDS")

str(df.zones)
# 'data.frame':	84 obs. of  15 variables:
# $ sample      : chr  "12465" "12469" "12471" "12473" ...
# $ group       : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ lipids      : num  0.283 0.288 0.3 0.304 0.297 ...
# $ proteins    : num  2.6 2.59 2.54 2.53 2.57 ...
# $ amino_sugars: num  4.67 4.8 4.74 4.75 4.74 ...
# $ carbs       : num  9.22 9.04 8.91 9.15 9.28 ...
# $ cond_aromtx : num  1.1 1.14 1.2 1.22 1.11 ...
# $ lignin      : num  8.08 8.04 8.04 8 8.02 ...
# $ tannins     : num  15.2 15.1 15.4 15.1 15.1 ...
# $ methylated  : num  0.421 0.415 0.411 0.409 0.418 ...
# $ hydrated    : num  4.4 4.25 4.29 4.31 4.34 ...
# $ demethylated: num  12.5 12.3 12.6 12.3 12.3 ...
# $ condensed   : num  0.493 0.466 0.486 0.469 0.465 ...
# $ k2_b12      : num  0.433 0.467 0.429 0.47 0.471 ...
# $ all         : num  59.5 58.9 59.3 59 59.1 ...


length(unique(df.zones$sample)) # 84

sel.nan <- which(df.zones$amino_sugars == 0) # none

sort(df.zones$amino_sugars)[1:20]
# [1] 4.279946 4.293305 4.299298 4.307610 4.308696 4.336715 4.339336 4.365910 4.423606 4.449702 4.458287 4.471072 4.500273
# [14] 4.503099 4.506155 4.508810 4.518687 4.521084 4.522212 4.522488

#df.zones <- df.zones[-sel.nan, ]

# lipids_to_amino_sugars
df.zones$lipids_to_amino_sugars <- df.zones$lipids / df.zones$amino_sugars
hist(df.zones$lipids_to_amino_sugars)
hist(log10(df.zones$lipids_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lipids_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lipids_to_amino_sugars[ -subsel.zero ])
  df.zones$lipids_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lipids_to_amino_sugars <- log10(df.zones$lipids_to_amino_sugars)
hist(df.zones$log10_lipids_to_amino_sugars)


# proteins_to_amino_sugars
df.zones$proteins_to_amino_sugars <- df.zones$proteins / df.zones$amino_sugars
hist(df.zones$proteins_to_amino_sugars)
hist(log10(df.zones$proteins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$proteins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$proteins_to_amino_sugars[ -subsel.zero ])
  df.zones$proteins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_proteins_to_amino_sugars <- log10(df.zones$proteins_to_amino_sugars)
hist(df.zones$log10_proteins_to_amino_sugars)


# carbs_to_amino_sugars
df.zones$carbs_to_amino_sugars <- df.zones$carbs / df.zones$amino_sugars
hist(df.zones$carbs_to_amino_sugars)
hist(log10(df.zones$carbs_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$carbs_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$carbs_to_amino_sugars[ -subsel.zero ])
  df.zones$carbs_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_carbs_to_amino_sugars <- log10(df.zones$carbs_to_amino_sugars)
hist(df.zones$log10_carbs_to_amino_sugars)


# cond_aromtx_to_amino_sugars
df.zones$cond_aromtx_to_amino_sugars <- df.zones$cond_aromtx / df.zones$amino_sugars
hist(df.zones$cond_aromtx_to_amino_sugars)
hist(log10(df.zones$cond_aromtx_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$cond_aromtx_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$cond_aromtx_to_amino_sugars[ -subsel.zero ])
  df.zones$cond_aromtx_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_cond_aromtx_to_amino_sugars <- log10(df.zones$cond_aromtx_to_amino_sugars)
hist(df.zones$log10_cond_aromtx_to_amino_sugars)


# lignin_to_amino_sugars
df.zones$lignin_to_amino_sugars <- df.zones$lignin / df.zones$amino_sugars
hist(df.zones$lignin_to_amino_sugars)
hist(log10(df.zones$lignin_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$lignin_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$lignin_to_amino_sugars[ -subsel.zero ])
  df.zones$lignin_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_lignin_to_amino_sugars <- log10(df.zones$lignin_to_amino_sugars)
hist(df.zones$log10_lignin_to_amino_sugars)


# tannins_to_amino_sugars
df.zones$tannins_to_amino_sugars <- df.zones$tannins / df.zones$amino_sugars
hist(df.zones$tannins_to_amino_sugars)
hist(log10(df.zones$tannins_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$tannins_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$tannins_to_amino_sugars[ -subsel.zero ])
  df.zones$tannins_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_tannins_to_amino_sugars <- log10(df.zones$tannins_to_amino_sugars)
hist(df.zones$log10_tannins_to_amino_sugars)


# methylated_to_amino_sugars
df.zones$methylated_to_amino_sugars <- df.zones$methylated / df.zones$amino_sugars
hist(df.zones$methylated_to_amino_sugars)
hist(log10(df.zones$methylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$methylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$methylated_to_amino_sugars[ -subsel.zero ])
  df.zones$methylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_methylated_to_amino_sugars <- log10(df.zones$methylated_to_amino_sugars)
hist(df.zones$log10_methylated_to_amino_sugars)


# hydrated_to_amino_sugars
df.zones$hydrated_to_amino_sugars <- df.zones$hydrated / df.zones$amino_sugars
hist(df.zones$hydrated_to_amino_sugars)
hist(log10(df.zones$hydrated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$hydrated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$hydrated_to_amino_sugars[ -subsel.zero ])
  df.zones$hydrated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_hydrated_to_amino_sugars <- log10(df.zones$hydrated_to_amino_sugars)
hist(df.zones$log10_hydrated_to_amino_sugars)


# demethylated_to_amino_sugars
df.zones$demethylated_to_amino_sugars <- df.zones$demethylated / df.zones$amino_sugars
hist(df.zones$demethylated_to_amino_sugars)
hist(log10(df.zones$demethylated_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$demethylated_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$demethylated_to_amino_sugars[ -subsel.zero ])
  df.zones$demethylated_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_demethylated_to_amino_sugars <- log10(df.zones$demethylated_to_amino_sugars)
hist(df.zones$log10_demethylated_to_amino_sugars)


# condensed_to_amino_sugars
df.zones$condensed_to_amino_sugars <- df.zones$condensed / df.zones$amino_sugars
hist(df.zones$condensed_to_amino_sugars)
hist(log10(df.zones$condensed_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$condensed_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$condensed_to_amino_sugars[ -subsel.zero ])
  df.zones$condensed_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_condensed_to_amino_sugars <- log10(df.zones$condensed_to_amino_sugars)
hist(df.zones$log10_condensed_to_amino_sugars)


# k2_b12_to_amino_sugars
df.zones$k2_b12_to_amino_sugars <- df.zones$k2_b12 / df.zones$amino_sugars
hist(df.zones$k2_b12_to_amino_sugars)
hist(log10(df.zones$k2_b12_to_amino_sugars)) # do log10 transform
# log10 - set zero-replacement value at 1/2 smallest non-zero value of that group
subsel.zero <- which(df.zones$k2_b12_to_amino_sugars == 0)
if (length(subsel.zero) > 0) {
  zero_replace <- 0.5*min(df.zones$k2_b12_to_amino_sugars[ -subsel.zero ])
  df.zones$k2_b12_to_amino_sugars[ subsel.zero ] <- zero_replace
}
df.zones$log10_k2_b12_to_amino_sugars <- log10(df.zones$k2_b12_to_amino_sugars)
hist(df.zones$log10_k2_b12_to_amino_sugars)



get_vars <- grep(pattern = "log10", x = names(df.zones))
names(df.zones)[get_vars]
# [1] "log10_lipids_to_amino_sugars"       "log10_proteins_to_amino_sugars"    
# [3] "log10_carbs_to_amino_sugars"        "log10_cond_aromtx_to_amino_sugars" 
# [5] "log10_lignin_to_amino_sugars"       "log10_tannins_to_amino_sugars"     
# [7] "log10_methylated_to_amino_sugars"   "log10_hydrated_to_amino_sugars"    
# [9] "log10_demethylated_to_amino_sugars" "log10_condensed_to_amino_sugars"   
# [11] "log10_k2_b12_to_amino_sugars" 


df.zones.melt <- melt(df.zones[ ,c("sample","group", names(df.zones)[get_vars] )], id.vars = c("sample","group"))
head(df.zones.melt)
# sample   group                     variable     value
# 1  12465 Natural log10_lipids_to_amino_sugars -1.217148
# 2  12469 Natural log10_lipids_to_amino_sugars -1.221775
# 3  12471 Natural log10_lipids_to_amino_sugars -1.197939
# 4  12473 Natural log10_lipids_to_amino_sugars -1.193423
# 5  12475 Natural log10_lipids_to_amino_sugars -1.203628
# 6  12477 Natural log10_lipids_to_amino_sugars -1.188753

unique( df.zones.melt$group )
# [1] Natural   Disturbed
# Levels: Disturbed < Natural

str(df.zones.melt)
# 'data.frame':	924 obs. of  4 variables:
#   $ sample  : chr  "12465" "12469" "12471" "12473" ...
# $ group   : Ord.factor w/ 2 levels "Disturbed"<"Natural": 2 2 2 2 2 2 2 2 2 2 ...
# $ variable: Factor w/ 11 levels "log10_lipids_to_amino_sugars",..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value   : num  -1.22 -1.22 -1.2 -1.19 -1.2 ...


df.zones.melt$study <- "AMI Disturbed\nvs Natural"

dat <- rbind(dat, df.zones.melt)

temp3 <- dat


saveRDS(object = dat, file = "dat--all-Soil-datasets-vk-zone-AminoSugars-ratios.RDS")

dat <- readRDS("dat--all-Soil-datasets-vk-zone-AminoSugars-ratios.RDS")


unique(dat$variable)
# [1] log10_lipids_to_amino_sugars       log10_proteins_to_amino_sugars    
# [3] log10_carbs_to_amino_sugars        log10_cond_aromtx_to_amino_sugars 
# [5] log10_lignin_to_amino_sugars       log10_tannins_to_amino_sugars     
# [7] log10_methylated_to_amino_sugars   log10_hydrated_to_amino_sugars    
# [9] log10_demethylated_to_amino_sugars log10_condensed_to_amino_sugars   
# [11] log10_k2_b12_to_amino_sugars      
# 11 Levels: log10_lipids_to_amino_sugars ... log10_k2_b12_to_amino_sugars

dat$variable <- factor(dat$variable, 
                       levels = c("log10_lipids_to_amino_sugars", "log10_proteins_to_amino_sugars", "log10_carbs_to_amino_sugars",
                                  "log10_cond_aromtx_to_amino_sugars", "log10_lignin_to_amino_sugars", "log10_tannins_to_amino_sugars",
                                  "log10_methylated_to_amino_sugars", "log10_hydrated_to_amino_sugars", "log10_demethylated_to_amino_sugars",
                                  "log10_condensed_to_amino_sugars", "log10_k2_b12_to_amino_sugars" ),
                       # labels = c("Lipids\n:Amino sugars", "Proteins\n:Amino sugars", "Carbohydrates\n:Amino sugars",
                       #            "Condensed\nAromatics\n:Amino sugars", "Lignin\n:Amino sugars", "Tannins\n:Amino sugars",
                       #            "Other\nMethylated\n:Amino sugars", "Other\nHydrated\n:Amino sugars", "Other\nDemethylated\n:Amino sugars",
                       #            "Other\nCondensed\n:Amino sugars","Near\nVitK2-VitB12\n:Amino sugars"),
                       labels = c("Lipids", "Proteins", "Carbo-\nhydrates",
                                  "Condensed\nAromatics", "Lignin", "Tannins",
                                  "Other\nMethyl\n-ated", "Other\nHydrated", "Other\nDemethyl\n-ated",
                                  "Other\nCondensed","Near\nVitK2\n-VitB12"),
                       ordered = TRUE)
levels( dat$variable )
# [1] "Lipids"                 "Proteins"               "Carbo-\nhydrates"      
# [4] "Condensed\nAromatics"   "Lignin"                 "Tannins"               
# [7] "Other\nMethyl\n-ated"   "Other\nHydrated"        "Other\nDemethyl\n-ated"
# [10] "Other\nCondensed"       "Near\nVitK2\n-VitB12"  


unique(dat$group)
# [1] 14        41        11        31        40        23        Rem       13       
# [9] 10        48        28        33        22        UM        12        6        
# [17] Natural   Disturbed
# 54 Levels: 1 < 2 < 3 < 4 < 5 < 6 < 7 < 8 < 9 < 10 < 11 < 12 < 13 < 14 < ... < Natural


## create named numeric vector for x-axis display in facet_grid

dat$age_vec_plot <- as.character( dat$group )

sort(unique( dat$age_vec_plot ))
# [1] "10"        "11"        "12"        "13"        "14"        "22"        "23"       
# [8] "28"        "31"        "33"        "40"        "41"        "48"        "6"        
# [15] "Disturbed" "Natural"   "Rem"       "UM" 

dat$age_vec_plot <- factor(dat$age_vec_plot,
                      levels = c(
                        "6",
                        "10", "11", "12", "13", "14","22", "23", 
                        "28",   "31", "33", "40", "41", "48", 
                        
                        "Rem",
                        "UM",
                        "Disturbed", "Natural"
                        ),
                      labels = c(
                        "6",
                        "10", "11", "12", "13", "14","22", "23", 
                        "28",   "31", "33", "40", "41", "48", 
                        
                        "58",
                        "42",
                        "2010", "2020"
                      ), 
                      ordered = TRUE)

dat$age_vec_plot_num <- as.integer(as.character(dat$age_vec_plot))
sort(unique( dat$age_vec_plot_num))
#   6   10   11   12   13   14   22   23   28   31   33   40   41   42   48   58 2010 2020

sel <- which(dat$study == "Post-mining\nRestoration")
dat$age_vec_plot_num[sel] <- dat$age_vec_plot_num[sel]+1000

sort(unique( dat$age_vec_plot_num))
#  10   11   13   14   23   28   31   33   40   41   48   58 1006 1012 1022 1031 1042 2010 2020

age_vec_plot_num <- c(1:2025)

x <- c(
  rep(NA, times=9),"10", # 1:10
  rep(NA, times=9),"20", # 11:20
  rep(NA, times=9),"30", # 21:30
  rep(NA, times=9),"40", # 31:40
  rep(NA, times=10),     # 41:50
  NA,NA,NA,NA,NA,NA,NA,"Rem",NA,NA,  # 51:60
  rep(NA, times = 940),   # 61:1000
  NA,NA,NA,NA,NA,"6",NA,NA,NA,NA, # 1001:1010
  NA,"12",NA,NA,NA,NA,NA,NA,NA,NA, # 1011:1020
  NA,"22",NA,NA,NA,NA,NA,NA,NA,NA, # 1021:1030
  "31",NA,NA,NA,NA,NA,NA,NA,NA,NA, # 1031:1040
  NA,"UM",NA,NA,NA,NA,NA,NA,NA,NA, # 1041:1050
  rep(NA, times = 950),   # 1051:2000
  NA,NA,NA,NA,"",NA,NA,NA,NA,"Disturbed", # 2001:2010
  NA,NA,NA,NA,NA,NA,NA,NA,NA,"Natural", # 2011:2020
  NA,NA,NA,NA,"" # 2021:2025
)

names(age_vec_plot_num) <- x

label_breaks <- which( is.na( names(age_vec_plot_num) ) == FALSE)

age_vec_plot_num[ label_breaks ]
# 10        20        30        40       Rem         6        12        22        31        UM 
# 10        20        30        40        58      1006      1012      1022      1031      1042 
# Disturbed   Natural           
# 2005      2010      2020      2025

names( age_vec_plot_num[ label_breaks ] )
# [1] "10"        "20"        "30"        "40"        "Rem"       "6"         "12"        "22"       
# [9] "31"        "UM"        ""          "Disturbed" "Natural"   ""

unique(dat$study)
# [1] "PCaN\nRestoration"         "Post-mining\nRestoration" 
# [3] "AMI Disturbed\nvs Natural"


dat$study <- factor(dat$study, 
                    levels = c(
                      "PCaN\nRestoration",
                      "Post-mining\nRestoration",
                      "AMI Disturbed\nvs Natural"),
                    ordered = TRUE)

str(dat)
# 'data.frame':	1298 obs. of  7 variables:
# $ sample          : chr  "ChrMrs" "ChrRic" "DunIsp" "DunSiH" ...
# $ group           : Ord.factor w/ 54 levels "1"<"2"<"3"<"4"<..: 14 41 11 31 14 40 23 51 13 31 ...
# $ variable        : Ord.factor w/ 11 levels "Lipids"<"Proteins"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ value           : num  -1.29 -1.17 -1.22 -1.17 -1.19 ...
# $ study           : Ord.factor w/ 3 levels "PCaN\nRestoration"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ age_vec_plot    : Ord.factor w/ 18 levels "6"<"10"<"11"<..: 6 13 3 10 6 12 8 15 5 10 ...
# $ age_vec_plot_num: num  14 41 11 31 14 40 23 53 13 31 ...




p <- ggplot()+ # y = value   y = win95abun
  
  geom_smooth(data = filter(dat, study == "PCaN\nRestoration"), aes(x = age_vec_plot_num, y = value), method = "loess")+
  geom_point(data = filter(dat, study == "PCaN\nRestoration"), aes(x = age_vec_plot_num, y = value), alpha = 0.4)+
  
  geom_smooth(data = filter(dat, study == "Post-mining\nRestoration"), aes(x = age_vec_plot_num, y = value), method = "loess")+
  geom_point(data = filter(dat, study == "Post-mining\nRestoration"), aes(x = age_vec_plot_num, y = value), alpha = 0.4)+
  
  geom_boxplot(data = filter(dat, study == "AMI Disturbed\nvs Natural"), aes(x = age_vec_plot_num, y = value, group = age_vec_plot_num), outlier.shape=NA)+
  
  facet_grid(rows = vars(variable), cols = vars(study), scales = "free" )+ #, space = "free_x")+
  theme_classic() +
  labs(x = NULL, y = "Log10 ratio of CPP (cf. Amino sugars)") +
  
  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+
  
  theme(
    #legend.position = "bottom",
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )
p


dev.print(tiff, file = paste0(workdir,"/plots/","All-Soil-log10-CPP-ratio-cf-Amino-sugars-ROUGH2--v9.tiff"), width = 15, height = 24, units = "cm", res=450, compression="lzw",type="cairo")



#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "PCaN\nRestoration"
## separate analysis of CPP-ASALR by soil pH: 1) Strongly acidic; 2) Acidic to neutral
#-------------------------

dat.temp <- dat

unique(dat.temp$study)
# [1] PCaN\nRestoration         Post-mining\nRestoration  AMI Disturbed\nvs Natural
# Levels: PCaN\nRestoration < Post-mining\nRestoration < AMI Disturbed\nvs Natural


this_study <- "PCaN\nRestoration"


sel <- which(dat.temp$study == this_study)
length(unique( dat.temp$sample[sel] )) # 19


# add pH data
dat.temp$pH <- NA

for (i in 1:dim(dat.temp[sel, ])[1]) {
  #i<-210
  this_samp <- dat.temp$sample[sel[i]]
  sel.phy <- which(phy@sam_data$sample == this_samp)
  dat.temp$pH[sel[i]] <- phy@sam_data$pH[sel.phy]
  print(paste0("completed ",i))
}

dat.temp$pH_class <- NA
subsel1 <- which(dat.temp$pH[sel] <4.5) # qty 88
dat.temp$pH_class[sel[subsel1]] <- "low"
subsel2 <- which(dat.temp$pH[sel] >=4.5) # qty 110
dat.temp$pH_class[sel[subsel2]] <- "mod"

ok <- complete.cases(dat.temp)
sel.ok <- which(ok==TRUE) # qty 198
dat.temp <- dat.temp[sel.ok, ] # Cut back to PCaN only !!!

dim(dat.temp) # 198  9



pos_anotations <- list()
sig_anotations <- list()
Kendall_Tau <- list()



## "Lipids"
this_var <- "Lipids"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.1947 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 1.7974, p-value = 0.07227
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.5400617

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = -0.72627, p-value = 0.4677
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.1840175

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Proteins"
this_var <- "Proteins"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.020 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0.77033, p-value = 0.4411
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.231455 

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = -0.72627, p-value = 0.4677
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.1840175

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Carbo-\nhydrates"
this_var <- "Carbo-\nhydrates"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # -0.235 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = -0.51355, p-value = 0.6076
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.1543033

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = -1.9973, p-value = 0.0458
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.5060481

pos_anotations[[this_var]] <- c("right", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Condensed\nAromatics"
this_var <- "Condensed\nAromatics"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.1276 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# 
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
# tau 
#   0 

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = 1.4525, p-value = 0.1463
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.368035

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Lignin"
this_var <- "Lignin"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.0873 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Tannins"
this_var <- "Tannins"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.275 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0.51355, p-value = 0.6076
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1543033

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = 0.72627, p-value = 0.4677
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1840175

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other\nMethyl\n-ated"
this_var <- "Other\nMethyl\n-ated"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # -0.047 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0.25678, p-value = 0.7974
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.07715167

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = -0.90784, p-value = 0.364
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.2300219

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other\nHydrated"
this_var <- "Other\nHydrated"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.0201 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0, p-value = 1
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = -1.4525, p-value = 0.1463
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.368035

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other\nDemethyl\n-ated"
this_var <- "Other\nDemethyl\n-ated"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.1813 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0.25678, p-value = 0.7974
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.07715167

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = 0.18157, p-value = 0.8559
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.04600437

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Other\nCondensed"
this_var <- "Other\nCondensed"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.1813 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 0.51355, p-value = 0.6076
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1543033

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = 1.6341, p-value = 0.1022
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.4140393 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

sel <- which(dat.temp$variable == this_var & dat.temp$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat.temp$age_vec_plot_num[sel], y = dat.temp[sel ,"value"], method = "kendall")
ktcor # 0.0604 (ns)

sel1 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "low")
# Kendall Tau correlation
ktcor1<- cor.test(x = dat.temp$age_vec_plot_num[sel1], y = dat.temp[sel1 ,"value"], method = "kendall")
ktcor1
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel1] and dat.temp[sel1, "value"]
# z = 1.5407, p-value = 0.1234
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.46291

sel2 <- which(dat.temp$variable == this_var  & dat.temp$study == this_study & dat.temp$pH_class == "mod")
# Kendall Tau correlation
ktcor2<- cor.test(x = dat.temp$age_vec_plot_num[sel2], y = dat.temp[sel2 ,"value"], method = "kendall")
ktcor2
# Kendall's rank correlation tau
# data:  dat.temp$age_vec_plot_num[sel2] and dat.temp[sel2, "value"]
# z = -1.8157, p-value = 0.06942
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.4600437 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- c( sig_symb(ktcor1$p.value), sig_symb(ktcor2$p.value))
Kendall_Tau[[this_var]] <- c( round(ktcor1$estimate, 3) , round(ktcor2$estimate, 3) )




annot_df.pcan.pH <- data.frame(
  study = rep("PCaN\nRestoration", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  npcx = c( pos_anotations[["Lipids"]][1],
            pos_anotations[["Proteins"]][1],
            pos_anotations[["Carbo-\nhydrates"]][1],
            pos_anotations[["Condensed\nAromatics"]][1],
            pos_anotations[["Lignin"]][1],
            pos_anotations[["Tannins"]][1],
            pos_anotations[["Other\nMethyl\n-ated"]][1],
            pos_anotations[["Other\nHydrated"]][1],
            pos_anotations[["Other\nDemethyl\n-ated"]][1],
            pos_anotations[["Other\nCondensed"]][1],
            pos_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  npcy = c( pos_anotations[["Lipids"]][2],
            pos_anotations[["Proteins"]][2],
            pos_anotations[["Carbo-\nhydrates"]][2],
            pos_anotations[["Condensed\nAromatics"]][2],
            pos_anotations[["Lignin"]][2],
            pos_anotations[["Tannins"]][2],
            pos_anotations[["Other\nMethyl\n-ated"]][2],
            pos_anotations[["Other\nHydrated"]][2],
            pos_anotations[["Other\nDemethyl\n-ated"]][2],
            pos_anotations[["Other\nCondensed"]][2],
            pos_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  tau1 = c(Kendall_Tau[["Lipids"]][1],
          Kendall_Tau[["Proteins"]][1],
          Kendall_Tau[["Carbo-\nhydrates"]][1],
          Kendall_Tau[["Condensed\nAromatics"]][1],
          Kendall_Tau[["Lignin"]][1],
          Kendall_Tau[["Tannins"]][1],
          Kendall_Tau[["Other\nMethyl\n-ated"]][1],
          Kendall_Tau[["Other\nHydrated"]][1],
          Kendall_Tau[["Other\nDemethyl\n-ated"]][1],
          Kendall_Tau[["Other\nCondensed"]][1],
          Kendall_Tau[["Near\nVitK2\n-VitB12"]][1]
  ),
  tau2 = c(Kendall_Tau[["Lipids"]][2],
           Kendall_Tau[["Proteins"]][2],
           Kendall_Tau[["Carbo-\nhydrates"]][2],
           Kendall_Tau[["Condensed\nAromatics"]][2],
           Kendall_Tau[["Lignin"]][2],
           Kendall_Tau[["Tannins"]][2],
           Kendall_Tau[["Other\nMethyl\n-ated"]][2],
           Kendall_Tau[["Other\nHydrated"]][2],
           Kendall_Tau[["Other\nDemethyl\n-ated"]][2],
           Kendall_Tau[["Other\nCondensed"]][2],
           Kendall_Tau[["Near\nVitK2\n-VitB12"]][2]
  ),
  siglabel1 = c(sig_anotations[["Lipids"]][1],
               sig_anotations[["Proteins"]][1],
               sig_anotations[["Carbo-\nhydrates"]][1],
               sig_anotations[["Condensed\nAromatics"]][1],
               sig_anotations[["Lignin"]][1],
               sig_anotations[["Tannins"]][1],
               sig_anotations[["Other\nMethyl\n-ated"]][1],
               sig_anotations[["Other\nHydrated"]][1],
               sig_anotations[["Other\nDemethyl\n-ated"]][1],
               sig_anotations[["Other\nCondensed"]][1],
               sig_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  siglabel2 = c(sig_anotations[["Lipids"]][2],
                sig_anotations[["Proteins"]][2],
                sig_anotations[["Carbo-\nhydrates"]][2],
                sig_anotations[["Condensed\nAromatics"]][2],
                sig_anotations[["Lignin"]][2],
                sig_anotations[["Tannins"]][2],
                sig_anotations[["Other\nMethyl\n-ated"]][2],
                sig_anotations[["Other\nHydrated"]][2],
                sig_anotations[["Other\nDemethyl\n-ated"]][2],
                sig_anotations[["Other\nCondensed"]][2],
                sig_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  stringsAsFactors = FALSE)
str(annot_df.pcan.pH)

# ensure no conflict in factor levels
annot_df.pcan.pH$variable <- factor(annot_df.pcan.pH$variable,
                                 levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                            "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.pcan.pH$study <- factor(annot_df.pcan.pH$study,
                              levels = c("PCaN\nRestoration", "Post-mining\nRestoration", "AMI Disturbed\nvs Natural"),
                              ordered = TRUE)

#annot_df.pcan.pH$displaylabel <- paste0("Kendall's\ntau: ", annot_df.pcan.pH$tau, annot_df.pcan.pH$siglabel)
annot_df.pcan.pH$displaylabel <- paste0("Kendall's tau\n",
                                         "1) ",annot_df.pcan.pH$tau1, annot_df.pcan.pH$siglabel1,
                                         "; 2) ",annot_df.pcan.pH$tau2, annot_df.pcan.pH$siglabel2)

annot_df.pcan.pH$displaylabel <- gsub(pattern = "ns", replacement = "", x = annot_df.pcan.pH$displaylabel)


str( annot_df.pcan.pH )
# 'data.frame':	11 obs. of  9 variables:
# $ study       : Ord.factor w/ 3 levels "PCaN\nRestoration"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ variable    : Ord.factor w/ 11 levels "Lipids"<"Proteins"<..: 1 2 3 4 5 6 7 8 9 10 ...
# $ npcx        : chr  "right" "left" "right" "left" ...
# $ npcy        : chr  "bottom" "bottom" "top" "top" ...
# $ tau1        : num  0.54 0.231 -0.154 0 0 0.154 0.077 0 0.077 0.154 ...
# $ tau2        : num  -0.184 -0.184 -0.506 0.368 0 0.184 -0.23 -0.368 0.046 0.414 ...
# $ siglabel1   : chr  "ns" "ns" "ns" "ns" ...
# $ siglabel2   : chr  "ns" "ns" "*" "ns" ...
# $ displaylabel: chr  "Kendall's tau\n1) 0.54; 2) -0.184" "Kendall's tau\n1) 0.231; 2) -0.184" "Kendall's tau\n1) -0.154; 2) -0.506*" "Kendall's tau\n1) 0; 2) 0.368" ...

saveRDS(object = annot_df.pcan.pH, file = "annot_df.pcan.pH.RDS" )



#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "Post-mining\nRestoration"
#-------------------------

unique(dat$study)
# [1] PCaN\nRestoration         Post-mining\nRestoration  AMI Disturbed\nvs Natural
# Levels: PCaN\nRestoration < Post-mining\nRestoration < AMI Disturbed\nvs Natural


this_study <- "Post-mining\nRestoration"


pos_anotations <- list()
sig_anotations <- list()
Kendall_Tau <- list()



## "Lipids"
this_var <- "Lipids"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 3.0382, p-value = 0.00238
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.6172134 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Proteins"
this_var <- "Proteins"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = -1.9242, p-value = 0.05433
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.3909018 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Carbo-\nhydrates"
this_var <- "Carbo-\nhydrates"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = -2.4306, p-value = 0.01507
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.4937707 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Condensed\nAromatics"
this_var <- "Condensed\nAromatics"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 3.342, p-value = 0.0008317
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.6789347

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Lignin"
this_var <- "Lignin"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = -0.20255, p-value = 0.8395
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.04114756 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Tannins"
this_var <- "Tannins"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 0.81019, p-value = 0.4178
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.1645902 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other\nMethyl\n-ated"
this_var <- "Other\nMethyl\n-ated"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = -4.051, p-value = 5.101e-05
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# -0.8229512 

pos_anotations[[this_var]] <- c("left", "top") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other\nHydrated"
this_var <- "Other\nHydrated"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 1.114, p-value = 0.2653
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.2263116 

pos_anotations[[this_var]] <- c("right", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other\nDemethyl\n-ated"
this_var <- "Other\nDemethyl\n-ated"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 1.114, p-value = 0.2653
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.2263116 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Other\nCondensed"
this_var <- "Other\nCondensed"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 1.5191, p-value = 0.1287
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.3086067 

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

sel <- which(dat$variable == this_var & dat$study == this_study)
# Kendall Tau correlation
ktcor<- cor.test(x = dat$age_vec_plot_num[sel], y = dat[sel ,"value"], method = "kendall")
ktcor
# data:  dat$age_vec_plot_num[sel] and dat[sel, "value"]
# z = 2.4306, p-value = 0.01507
# alternative hypothesis: true tau is not equal to 0
# sample estimates:
#   tau 
# 0.4937707

pos_anotations[[this_var]] <- c("left", "bottom") # npcx = "left", npcy = "top"
sig_anotations[[this_var]] <- sig_symb(ktcor$p.value)
Kendall_Tau[[this_var]] <- round(ktcor$estimate, 3)



annot_df.postmining <- data.frame(
  study = rep("Post-mining\nRestoration", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  npcx = c( pos_anotations[["Lipids"]][1],
            pos_anotations[["Proteins"]][1],
            pos_anotations[["Carbo-\nhydrates"]][1],
            pos_anotations[["Condensed\nAromatics"]][1],
            pos_anotations[["Lignin"]][1],
            pos_anotations[["Tannins"]][1],
            pos_anotations[["Other\nMethyl\n-ated"]][1],
            pos_anotations[["Other\nHydrated"]][1],
            pos_anotations[["Other\nDemethyl\n-ated"]][1],
            pos_anotations[["Other\nCondensed"]][1],
            pos_anotations[["Near\nVitK2\n-VitB12"]][1]
  ),
  npcy = c( pos_anotations[["Lipids"]][2],
            pos_anotations[["Proteins"]][2],
            pos_anotations[["Carbo-\nhydrates"]][2],
            pos_anotations[["Condensed\nAromatics"]][2],
            pos_anotations[["Lignin"]][2],
            pos_anotations[["Tannins"]][2],
            pos_anotations[["Other\nMethyl\n-ated"]][2],
            pos_anotations[["Other\nHydrated"]][2],
            pos_anotations[["Other\nDemethyl\n-ated"]][2],
            pos_anotations[["Other\nCondensed"]][2],
            pos_anotations[["Near\nVitK2\n-VitB12"]][2]
  ),
  tau = c(Kendall_Tau[["Lipids"]],
          Kendall_Tau[["Proteins"]],
          Kendall_Tau[["Carbo-\nhydrates"]],
          Kendall_Tau[["Condensed\nAromatics"]],
          Kendall_Tau[["Lignin"]],
          Kendall_Tau[["Tannins"]],
          Kendall_Tau[["Other\nMethyl\n-ated"]],
          Kendall_Tau[["Other\nHydrated"]],
          Kendall_Tau[["Other\nDemethyl\n-ated"]],
          Kendall_Tau[["Other\nCondensed"]],
          Kendall_Tau[["Near\nVitK2\n-VitB12"]]
          ),
  siglabel = c(sig_anotations[["Lipids"]],
               sig_anotations[["Proteins"]],
               sig_anotations[["Carbo-\nhydrates"]],
               sig_anotations[["Condensed\nAromatics"]],
               sig_anotations[["Lignin"]],
               sig_anotations[["Tannins"]],
               sig_anotations[["Other\nMethyl\n-ated"]],
               sig_anotations[["Other\nHydrated"]],
               sig_anotations[["Other\nDemethyl\n-ated"]],
               sig_anotations[["Other\nCondensed"]],
               sig_anotations[["Near\nVitK2\n-VitB12"]]
  ), stringsAsFactors = FALSE)
str(annot_df.postmining)

# ensure no conflict in factor levels
annot_df.postmining$variable <- factor(annot_df.postmining$variable,
                                levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                           "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.postmining$study <- factor(annot_df.postmining$study,
                             levels = c("PCaN\nRestoration", "Post-mining\nRestoration", "AMI Disturbed\nvs Natural"),
                             ordered = TRUE)

annot_df.postmining$displaylabel <- paste0("Kendall's\ntau: ", annot_df.postmining$tau, annot_df.postmining$siglabel)
annot_df.postmining$displaylabel <- gsub(pattern = "ns", replacement = "", x = annot_df.postmining$displaylabel)


str( annot_df.postmining )
saveRDS(object = annot_df.postmining, file = "annot_df.postmining.RDS" )


#-------------------------

### log10 ratio CPP-ASALR (normalized wrt Amino sugars) - STATS - "AMI Disturbed\nvs Natural"
#-------------------------

unique(dat$study)
# [1] PCaN\nRestoration         Post-mining\nRestoration  AMI Disturbed\nvs Natural
# Levels: PCaN\nRestoration < Post-mining\nRestoration < AMI Disturbed\nvs Natural


this_study <- "AMI Disturbed\nvs Natural"


y_anotations <- list()
sig_anotations <- list()



## "Lipids"
this_var <- "Lipids"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.1014, num df = 28, denom df = 54, p-value = 0.743

tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -5.2943, df = 82, p-value = 4.899e-07

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.358  -1.293  -1.229  -1.233  -1.189  -1.036 

y_anotations[[this_var]] <- -1.11
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Proteins"
this_var <- "Proteins"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.54189, num df = 28, denom df = 54, p-value = 0.08107

tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 4.6532, df = 82, p-value = 6.203e-06

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.2881 -0.2670 -0.2516 -0.2537 -0.2428 -0.2217 

y_anotations[[this_var]] <- -0.23
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Carbo-\nhydrates"
this_var <- "Carbo-\nhydrates"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.1378, num df = 28, denom df = 54, p-value = 0.6694

tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 4.5568, df = 82, p-value = 8.961e-06

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.2699  0.2898  0.3070  0.3097  0.3328  0.3534

y_anotations[[this_var]] <- 0.346
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Condensed\nAromatics"
this_var <- "Condensed\nAromatics"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.8162, num df = 28, denom df = 54, p-value = 0.5681

tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -3.4379, df = 82, p-value = 0.0004617

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.7550 -0.6673 -0.6478 -0.6475 -0.6253 -0.5602 

y_anotations[[this_var]] <- -0.58
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Lignin"
this_var <- "Lignin"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.69304, num df = 28, denom df = 54, p-value = 0.2947

tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 3.9459, df = 82, p-value = 8.348e-05

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.2092  0.2283  0.2487  0.2486  0.2716  0.2861 

y_anotations[[this_var]] <- 0.28
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Tannins"
this_var <- "Tannins"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.28662, num df = 28, denom df = 54, p-value = 0.0006215

# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 995, p-value = 0.03191

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.4831  0.5030  0.5231  0.5230  0.5384  0.5741 

y_anotations[[this_var]] <- 0.56
#sig_anotations[[this_var]] <- sig_symb( tt$p.value )
sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Other\nMethyl\n-ated"
this_var <- "Other\nMethyl\n-ated"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 1.1028, num df = 28, denom df = 54, p-value = 0.7402

tt <- t.test(x,y,var.equal = TRUE, alternative = "greater") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = 5.1257, df = 82, p-value = 9.691e-07

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.0886 -1.0504 -1.0189 -1.0009 -0.9541 -0.8501 

y_anotations[[this_var]] <- -0.89
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Other\nHydrated"
this_var <- "Other\nHydrated"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.44175, num df = 28, denom df = 54, p-value = 0.02128

# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 1178, p-value = 0.0001751

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -0.065602 -0.033996 -0.003328 -0.004731  0.023379  0.045045 

y_anotations[[this_var]] <- 0.036
#sig_anotations[[this_var]] <- sig_symb( tt$p.value )
sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Other\nDemethyl\n-ated"
this_var <- "Other\nDemethyl\n-ated"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.27105, num df = 28, denom df = 54, p-value = 0.0003705

# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 999, p-value = 0.02931

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.3957  0.4196  0.4436  0.4427  0.4600  0.5017  

y_anotations[[this_var]] <- 0.48
#sig_anotations[[this_var]] <- sig_symb( tt$p.value )
sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Other\nCondensed"
this_var <- "Other\nCondensed"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.68586, num df = 28, denom df = 54, p-value = 0.2811

tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
tt
# Two Sample t-test
# data:  x and y
# t = -2.5272, df = 82, p-value = 0.00671

# wt <- wilcox.test(x, y, alternative = "greater") # "less" "two.sided"  "greater"
# wt
# # Wilcoxon rank sum test with continuity correction
# # data:  x and y

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.3416 -1.1013 -1.0355 -1.0325 -0.9749 -0.6814 

y_anotations[[this_var]] <- -0.8
sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )



## "Near\nVitK2\n-VitB12"
this_var <- "Near\nVitK2\n-VitB12"

x=na.omit( filter(dat, group == "Disturbed" & variable == this_var & study == this_study)$value )  # 29
y=na.omit( filter(dat, group == "Natural" & variable == this_var & study == this_study)$value ) # 55

# http://www.sthda.com/english/wiki/unpaired-two-samples-t-test-in-r#check-your-data
# test for homogeneity of variances
var.test(x,y) # F = 0.42881, num df = 28, denom df = 54, p-value = 0.01719

# tt <- t.test(x,y,var.equal = TRUE, alternative = "less") # alternative = "two.sided"  alternative = "less"   alternative = "greater"
# tt
# # Two Sample t-test
# # data:  x and y

wt <- wilcox.test(x, y, alternative = "less") # "less" "two.sided"  "greater"
wt
# Wilcoxon rank sum test with continuity correction
# data:  x and y
# W = 652, p-value = 0.08626

summary(c(x,y))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# -1.1832 -1.0592 -1.0430 -1.0332 -1.0113 -0.8938 

y_anotations[[this_var]] <- NA
sig_anotations[[this_var]] <- NA
#sig_anotations[[this_var]] <- sig_symb( tt$p.value )
#sig_anotations[[this_var]] <- sig_symb( wt$p.value )


annot_df.ami <- data.frame(
  study = rep("AMI Disturbed\nvs Natural", times = 11),
  #sex = rep("Female", times = 11),
  variable=c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
             "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"),
  y_val = c( y_anotations[["Lipids"]],
             y_anotations[["Proteins"]],
             y_anotations[["Carbo-\nhydrates"]],
             y_anotations[["Condensed\nAromatics"]],
             y_anotations[["Lignin"]],
             y_anotations[["Tannins"]],
             y_anotations[["Other\nMethyl\n-ated"]],
             y_anotations[["Other\nHydrated"]],
             y_anotations[["Other\nDemethyl\n-ated"]],
             y_anotations[["Other\nCondensed"]],
             y_anotations[["Near\nVitK2\n-VitB12"]]
  ),
  x_label = rep(2015, times = 11),      # Disturbed == 2010; Natural == 2020
  x_seg_start = rep(2010.5, times = 11),
  x_seg_end = rep(2019.5, times = 11),
  label = c(sig_anotations[["Lipids"]],
            sig_anotations[["Proteins"]],
            sig_anotations[["Carbo-\nhydrates"]],
            sig_anotations[["Condensed\nAromatics"]],
            sig_anotations[["Lignin"]],
            sig_anotations[["Tannins"]],
            sig_anotations[["Other\nMethyl\n-ated"]],
            sig_anotations[["Other\nHydrated"]],
            sig_anotations[["Other\nDemethyl\n-ated"]],
            sig_anotations[["Other\nCondensed"]],
            sig_anotations[["Near\nVitK2\n-VitB12"]]
  ), stringsAsFactors = FALSE)
str(annot_df.ami)

# ensure no conflict in factor levels
annot_df.ami$variable <- factor(annot_df.ami$variable,
                                        levels = c("Lipids", "Proteins", "Carbo-\nhydrates", "Condensed\nAromatics", "Lignin", "Tannins", "Other\nMethyl\n-ated",
                                                   "Other\nHydrated", "Other\nDemethyl\n-ated", "Other\nCondensed", "Near\nVitK2\n-VitB12"), ordered = TRUE)
annot_df.ami$study <- factor(annot_df.ami$study,
                                     levels = c("PCaN\nRestoration", "Post-mining\nRestoration", "AMI Disturbed\nvs Natural"),
                                     ordered = TRUE)
str( annot_df.ami )
saveRDS(object = annot_df.ami, file = "annot_df.ami.RDS" )


# round(annot_df.ami$y_val, 3) # -0.483  1.041  1.854  0.432  0.843  1.420  0.542  0.459  1.163  0.525 -0.141
# annot_df.ami$y_val2 <- c(-0.75, 0.05, 1.1, -0.9, 0.35, 0.87,   -0.3, 0.45, 0.80, -0.025, -0.25)


#-------------------------

#### Soils - log10 ratio CPP-ASALR (normalized wrt Amino sugars) - FINAL PLOT
#-------------------------


p <- ggplot()+

  geom_smooth(data = filter(dat.temp, study == "PCaN\nRestoration"), aes(x = age_vec_plot_num, y = value, color=pH_class), method = "lm", se = FALSE)+  # method = "loess"
  geom_point(data = filter(dat.temp, study == "PCaN\nRestoration"), aes(x = age_vec_plot_num, y = value, color=pH_class), alpha = 0.4)+
  
  geom_smooth(data = filter(dat, study == "Post-mining\nRestoration"), aes(x = age_vec_plot_num, y = value), method = "lm", se = FALSE)+ # method = "loess"
  geom_point(data = filter(dat, study == "Post-mining\nRestoration"), aes(x = age_vec_plot_num, y = value), alpha = 0.4)+
  
  geom_boxplot(data = filter(dat, study == "AMI Disturbed\nvs Natural"), aes(x = age_vec_plot_num, y = value, group = age_vec_plot_num), outlier.shape=NA)+
  
  # PCaN
  geom_text_npc(data = annot_df.pcan.pH, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 2.75, lineheight=0.9 )+ # size = 3

  # Post-mining
  geom_text_npc(data = annot_df.postmining, aes(npcx = npcx, npcy = npcy, label = displaylabel), size = 2.75, lineheight=0.9 )+ # size = 3

  # AMI Disturbed vs Natural
  geom_text(data= filter(annot_df.ami, label != "ns"), mapping = aes(x = x_label, y = y_val, label = label), size = 5, vjust = 0.4)+ #  label = "**", size = 6   vjust = "bottom"
  geom_segment(data=filter(annot_df.ami, label != "ns"), mapping = aes(x = x_seg_start, y = y_val, xend = x_seg_end, yend = y_val))+


  facet_grid(rows = vars(variable), cols = vars(study), scales = "free" )+ #, space = "free_x")+
  theme_classic() +
  labs(x = NULL, y = "Log10 ( CPP[class] / CPP[amino sugars] )") +
  
  scale_color_discrete(labels=c("1) Strongly acidic", "2) Acidic to neutral"), name = "pH level" )+

  scale_x_continuous(breaks = label_breaks, labels= names( age_vec_plot_num[ label_breaks ] ) )+

  theme(
    legend.position = "bottom",
    legend.justification="left",
    legend.margin=margin(0,0,0,0),
    legend.box.margin=margin(-5,0,-5,0), 
    legend.title = element_text(size = rel(0.9)),
    strip.background = element_rect(fill="white", linetype = "blank"),
    #axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
    strip.text = element_text(margin=margin(t = 0,r = 0,b = 0,l = 0,"pt"), size = rel(0.8))
  )
p


dev.print(tiff, file = paste0(workdir,"/plots/","All-Soil-log10-CPP-ratio-cf-Amino-sugars-WITHSIGSTATS--v9d-Extra-pH-level.tiff"), width = 15, height = 24, units = "cm", res=450, compression="lzw",type="cairo")


#-------------------------

### END of code
